{
 "cells": [
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Load required modules"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 1,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "LOADED\n"
     ]
    }
   ],
   "source": [
    "# External libraries\n",
    "import numpy as np\n",
    "import pandas as pd\n",
    "import pygal\n",
    "import math\n",
    "import IPython\n",
    "import matplotlib\n",
    "import matplotlib.pyplot as plt\n",
    "import matplotlib.font_manager as font_manager\n",
    "import importlib\n",
    "import svgutils\n",
    "from svgutils.compose import *\n",
    "import os.path\n",
    "#import svgwrite\n",
    "\n",
    "# Custom code\n",
    "from fx_RNAseq import *\n",
    "def findgene(genename):\n",
    "    dPath = \"~/Documents/LiMolec/zfRNAseq/20190827/20190827_DESeq2/\" #directory path\n",
    "#     fPath = \"00_rodsVcones/rodVCones_raw.csv\" #file path\n",
    "    fPath = \"00_rodsVcones/rodVCones_fpkm.csv\" #file path\n",
    "    tempPath = \"temp.txt\" # temporary file to store values\n",
    "    !head -1 {dPath + fPath} > {dPath + tempPath}\n",
    "    # EXACT MATCH ONLY by matching word (-w) and only at the start of each line (^) which corresponds to the symbol\n",
    "    !grep -w ^$genename {dPath + fPath} >> {dPath + tempPath}\n",
    "    gdf = pd.read_csv(dPath + tempPath)\n",
    "    return gdf\n",
    "\n",
    "def findgenelist(genename):\n",
    "    dPath = \"~/Documents/LiMolec/zfRNAseq/20190827/20190827_DESeq2/\" #directory path\n",
    "#     fPath = \"00_rodsVcones/rodVCones_raw.csv\" #file path\n",
    "    fPath = \"00_rodsVcones/rodVCones_fpkm.csv\" #file path\n",
    "    tempPath = \"temp.txt\" # temporary file to store values\n",
    "    !head -1 {dPath + fPath} > {dPath + tempPath}\n",
    "    # EXACT MATCH ONLY by matching word (-w) and only at the start of each line (^) which corresponds to the symbol\n",
    "    !grep ^$genename {dPath + fPath} >> {dPath + tempPath}\n",
    "    gdf = pd.read_csv(dPath + tempPath)\n",
    "    return gdf\n",
    "\n",
    "# def findgene(genename):\n",
    "#     !head -1 ~/Dropbox/LiLab/RNAseq/20190827_DESeq2/00_rodsVcones/rodVCones_raw.csv > ~/Dropbox/LiLab/RNAseq/20190827_DESeq2/temp.txt\n",
    "#     # EXACT MATCH ONLY by matching word (-w) and only at the start of each line (^) which corresponds to the symbol\n",
    "#     !grep -w ^$genename ~/Dropbox/LiLab/RNAseq/20190827_DESeq2/00_rodsVcones/rodVCones_raw.csv >> ~/Dropbox/LiLab/RNAseq/20190827_DESeq2/temp.txt\n",
    "#     gdf = pd.read_csv(\"~/Dropbox/LiLab/RNAseq/20190827_DESeq2/temp.txt\")\n",
    "    \n",
    "#     return gdf\n",
    "\n",
    "# def findgenelist(genename):\n",
    "#     !head -1 ~/Dropbox/LiLab/RNAseq/20190827_DESeq2/00_rodsVcones/rodVCones_raw.csv > ~/Dropbox/LiLab/RNAseq/20190827_DESeq2/temp.txt\n",
    "#     # EXACT MATCH ONLY by matching word (-w) and only at the start of each line (^) which corresponds to the symbol\n",
    "#     !grep ^$genename ~/Dropbox/LiLab/RNAseq/20190827_DESeq2/00_rodsVcones/rodVCones_raw.csv >> ~/Dropbox/LiLab/RNAseq/20190827_DESeq2/temp.txt\n",
    "#     gdf = pd.read_csv(\"~/Dropbox/LiLab/RNAseq/20190827_DESeq2/temp.txt\")\n",
    "#     return gdf\n",
    "\n",
    "def plotNReads():\n",
    "    rdf = pd.read_csv(\"~/Documents/LiMolec/zfRNAseq/20190827/20190827_DESeq2/nMappedReads.csv\")\n",
    "    h = rdf.iloc[:,1]\n",
    "    h = h / 1e6\n",
    "    n = np.arange(1,7) # Rods\n",
    "    n = np.append(n, 6.5 + np.arange(1,6)) # UV\n",
    "    n = np.append(n, 12 + np.arange(1,7)) # S\n",
    "    n = np.append(n, 18.5 + np.arange(1,8)) # M\n",
    "    n = np.append(n, 26 + np.arange(1,7)) # L\n",
    "    return n, h\n",
    "\n",
    "def plotBars(genename):\n",
    "    gdf = findgene(genename)\n",
    "    n = np.arange(1,7) # Rods\n",
    "    n = np.append(n, 6.5 + np.arange(1,6)) # UV\n",
    "    n = np.append(n, 12 + np.arange(1,7)) # S\n",
    "    n = np.append(n, 18.5 + np.arange(1,8)) # M\n",
    "    n = np.append(n, 26 + np.arange(1,7)) # L\n",
    "    h_start = 7\n",
    "    h_end = 37\n",
    "    h = gdf.iloc[0,h_start:h_end]\n",
    "    return n, h\n",
    "\n",
    "def plotBars_list(genelist):\n",
    "    n=[]\n",
    "    h=[]\n",
    "    for i,gene in enumerate(genelist, start=0):\n",
    "        tempn, temph = plotBars(gene)\n",
    "        n = np.append(n,tempn+(30*i))\n",
    "        h = np.append(h,temph)\n",
    "    return n, h\n",
    "\n",
    "def svgreload():\n",
    "    !osascript -e 'tell application \"Google Chrome\" to tell the active tab of second window to reload'\n",
    "\n",
    "from pygal.style import Style\n",
    "custom_style = Style(\n",
    "    background = 'black',\n",
    "    plot_background = '#111111',\n",
    "    foreground = '#999999',\n",
    "    foreground_strong = '#eeeeee',\n",
    "    foreground_subtle = '#555555',\n",
    "    opacity = '.8',\n",
    "    opacity_hover = '.4',\n",
    "    transition = '10ms',\n",
    "    colors = ('#747474','#B540B7', '#4669F2','#04CD22','#CC2C2A'),\n",
    "    label_font_size = 24,\n",
    "    major_label_font_size = 24,\n",
    "    title_font_size = 24,\n",
    "    legend_font_size = 18\n",
    ")\n",
    "\n",
    "#photoreceptor Colors\n",
    "pC = {\n",
    "    'r' : '#747474',\n",
    "    'u' : '#B540B7',\n",
    "    's' : '#4669F2',\n",
    "    'm' : '#04CD22',\n",
    "    'l' : '#CC2C2A',\n",
    "    'plt' : '',\n",
    "}\n",
    "\n",
    "pC['plt']=[\n",
    "    pC['r'],pC['r'],pC['r'],pC['r'],pC['r'],pC['r'],\n",
    "    pC['u'],pC['u'],pC['u'],pC['u'],pC['u'],\n",
    "    pC['s'],pC['s'],pC['s'],pC['s'],pC['s'],pC['s'],\n",
    "    pC['m'],pC['m'],pC['m'],pC['m'],pC['m'],pC['m'],pC['m'],\n",
    "    pC['l'],pC['l'],pC['l'],pC['l'],pC['l'],pC['l']\n",
    "]\n",
    "\n",
    "def findgene_glia(genename):\n",
    "    dPath = \"~/Documents/LiLab/RNAseq/zfMullerGlia_2018Charlton/\" #directory path\n",
    "#     fPath = \"glia_23615_sup-0006-tables2.csv\" #file path\n",
    "    fPath = \"glia8dpf_DEGcounts.csv\" #file path\n",
    "    tempPath = \"temp.txt\" # temporary file to store values\n",
    "    !head -1 {dPath + fPath} > {dPath + tempPath}\n",
    "    # EXACT MATCH ONLY by matching word (-w) and only at the start of each line (^) which corresponds to the symbol\n",
    "    !grep -w ^$genename {dPath + fPath} >> {dPath + tempPath}\n",
    "    gdf = pd.read_csv(dPath + tempPath)\n",
    "    return gdf\n",
    "\n",
    "def findgenelist_glia(genename):\n",
    "    dPath = \"~/Documents/LiLab/RNAseq/zfMullerGlia_2018Charlton/\" #directory path\n",
    "#     fPath = \"glia_23615_sup-0006-tables2.csv\" #file path\n",
    "    fPath = \"glia8dpf_DEGcounts.csv\" #file path\n",
    "    tempPath = \"temp.txt\" # temporary file to store values\n",
    "    !head -1 {dPath + fPath} > {dPath + tempPath}\n",
    "    # EXACT MATCH ONLY by matching word (-w) and only at the start of each line (^) which corresponds to the symbol\n",
    "    !grep ^$genename {dPath + fPath} >> {dPath + tempPath}\n",
    "    gdf = pd.read_csv(dPath + tempPath)\n",
    "    return gdf\n",
    "\n",
    "def plotBars_glia(genename):\n",
    "    gdf = findgene_glia(genename)\n",
    "    n = np.arange(1,19) # GFP+\n",
    "    n = np.append(n, 19.5 + np.arange(1,19)) # GFP-\n",
    "    h_start = 5\n",
    "    h_end = 41\n",
    "#     h = np.power(2,(gdf.iloc[0,h_start:h_end]))\n",
    "    h = gdf.iloc[0,h_start:h_end]\n",
    "    return n, h\n",
    "\n",
    "#glia Colors\n",
    "gC = {\n",
    "    'm48' : '#c7ffec',\n",
    "    'm60' : '#94ffdb',\n",
    "    'm72' : '#57ffc7',\n",
    "    'm96' : '#2effb9',\n",
    "    'm120': '#05ffac',\n",
    "    'm192': '#00d68f',\n",
    "    \n",
    "    'c48' : '#e3e3e3',\n",
    "    'c60' : '#c9c9c9',\n",
    "    'c72' : '#ababab',\n",
    "    'c96' : '#969696',\n",
    "    'c120': '#828282',\n",
    "    'c192': '#6b6b6b',\n",
    "    \n",
    "    'plt' : '',\n",
    "}\n",
    "\n",
    "gC['plt']=[\n",
    "    gC['m48'],gC['m48'],gC['m48'],\n",
    "    gC['m60'],gC['m60'],gC['m60'],\n",
    "    gC['m72'],gC['m72'],gC['m72'],\n",
    "    gC['m96'],gC['m96'],gC['m96'],\n",
    "    gC['m120'],gC['m120'],gC['m120'],\n",
    "    gC['m192'],gC['m192'],gC['m192'],\n",
    "    gC['c48'],gC['c48'],gC['c48'],\n",
    "    gC['c60'],gC['c60'],gC['c60'],\n",
    "    gC['c72'],gC['c72'],gC['c72'],\n",
    "    gC['c96'],gC['c96'],gC['c96'],\n",
    "    gC['c120'],gC['c120'],gC['c120'],\n",
    "    gC['c192'],gC['c192'],gC['c192'],\n",
    "]\n",
    "\n",
    "def findgene_rods(genename):\n",
    "    dPath = \"~/Documents/LiLab/RNAseq/zfRods_Sun2018/\" #directory path\n",
    "#     fPath = \"glia_23615_sup-0006-tables2.csv\" #file path\n",
    "    fPath = \"zfRods.csv\" #file path\n",
    "    tempPath = \"temp.txt\" # temporary file to store values\n",
    "    !head -1 {dPath + fPath} > {dPath + tempPath}\n",
    "    # EXACT MATCH ONLY by matching word (-w) and only at the start of each line (^) which corresponds to the symbol\n",
    "    !grep -w ^$genename {dPath + fPath} >> {dPath + tempPath}\n",
    "    gdf = pd.read_csv(dPath + tempPath)\n",
    "    return gdf\n",
    "\n",
    "def findgenelist_rods(genename):\n",
    "    dPath = \"~/Documents/LiLab/RNAseq/zfRods_Sun2018/\" #directory path\n",
    "#     fPath = \"glia_23615_sup-0006-tables2.csv\" #file path\n",
    "    fPath = \"zfRods.csv\" #file path\n",
    "    tempPath = \"temp.txt\" # temporary file to store values\n",
    "    !head -1 {dPath + fPath} > {dPath + tempPath}\n",
    "    # EXACT MATCH ONLY by matching word (-w) and only at the start of each line (^) which corresponds to the symbol\n",
    "    !grep ^$genename {dPath + fPath} >> {dPath + tempPath}\n",
    "    gdf = pd.read_csv(dPath + tempPath)\n",
    "    return gdf\n",
    "\n",
    "def plotBars_rods(genename):\n",
    "    gdf = findgene_rods(genename)\n",
    "    n = np.arange(1,5) # GFP+\n",
    "    n = np.append(n, 4.5 + np.arange(1,5)) # GFP-\n",
    "    h_start = 7\n",
    "    h_end = 15\n",
    "#     h = np.power(2,(gdf.iloc[0,h_start:h_end]))\n",
    "    h = gdf.iloc[0,h_start:h_end].values\n",
    "    return n, h\n",
    "\n",
    "#rod Colors\n",
    "rC = {\n",
    "    'r' : '#747474',\n",
    "    'c': '#dac910',\n",
    "    'plt' : '',\n",
    "}\n",
    "\n",
    "rC['plt']=[\n",
    "    rC['r'],rC['r'],rC['r'],rC['r'],\n",
    "    rC['c'],rC['c'],rC['c'],rC['c'],\n",
    "]\n",
    "\n",
    "def findgene_sqcones(genename):\n",
    "    dPath = \"~/Documents/LiLab/RNAseq/sqCones_Kunze/\" #directory path\n",
    "    fPath = \"sqConesAwHib_Adult.csv\" #file path\n",
    "    tempPath = \"temp.txt\" # temporary file to store values\n",
    "    !head -1 {dPath + fPath} > {dPath + tempPath}\n",
    "    # EXACT MATCH ONLY by matching word (-w) and only at the start of each line (^) which corresponds to the symbol\n",
    "    !grep -w ^$genename {dPath + fPath} >> {dPath + tempPath}\n",
    "    gdf = pd.read_csv(dPath + tempPath)\n",
    "    return gdf\n",
    "\n",
    "def findgenelist_sqcones(genename):\n",
    "    dPath = \"~/Documents/LiLab/RNAseq/sqCones_Kunze/\" #directory path\n",
    "    fPath = \"sqConesAwHib_Adult.csv\" #file path\n",
    "    tempPath = \"temp.txt\" # temporary file to store values\n",
    "    !head -1 {dPath + fPath} > {dPath + tempPath}\n",
    "    # EXACT MATCH ONLY by matching word (-w) and only at the start of each line (^) which corresponds to the symbol\n",
    "    !grep ^$genename {dPath + fPath} >> {dPath + tempPath}\n",
    "    gdf = pd.read_csv(dPath + tempPath)\n",
    "    return gdf\n",
    "\n",
    "def plotBars_sqcones(genename):\n",
    "    gdf = findgene_sqcones(genename)\n",
    "    n = np.arange(1,8) # GFP+\n",
    "    n = np.append(n, 7.5 + np.arange(1,7)) # GFP-\n",
    "    h_start = 8\n",
    "    h_end = 21\n",
    "#     h = np.power(2,(gdf.iloc[0,h_start:h_end]))\n",
    "    h = gdf.iloc[0,h_start:h_end].values\n",
    "    return n, h\n",
    "\n",
    "#squirrel Colors\n",
    "sqC = {\n",
    "    's' : '#4669F2',\n",
    "    'sHib' : '#548ced',\n",
    "    'm' : '#04CD22',\n",
    "    'mHib' : '#53e477',\n",
    "    'plt' : '',\n",
    "}\n",
    "\n",
    "sqC['plt']=[\n",
    "    sqC['s'],sqC['s'],sqC['s'],sqC['s'],sqC['s'],\n",
    "    sqC['sHib'],sqC['sHib'],\n",
    "    sqC['m'],sqC['m'],sqC['m'],\n",
    "    sqC['mHib'],sqC['mHib'],sqC['mHib'],\n",
    "]\n",
    "\n",
    "\n",
    "def findgene_haircell(genename):\n",
    "    dPath = \"~/Documents/LiLab/RNAseq/zfHairCells_Mattern2018/\" #directory path\n",
    "    fPath = \"zfHairCells_rpkm.csv\" #file path\n",
    "    tempPath = \"temp.txt\" # temporary file to store values\n",
    "    !head -1 {dPath + fPath} > {dPath + tempPath}\n",
    "    # EXACT MATCH ONLY by matching word (-w) and only at the start of each line (^) which corresponds to the symbol\n",
    "    !grep -w ^$genename {dPath + fPath} >> {dPath + tempPath}\n",
    "    gdf = pd.read_csv(dPath + tempPath)\n",
    "    return gdf\n",
    "\n",
    "def findgenelist_haircell(genename):\n",
    "    dPath = \"~/Documents/LiLab/RNAseq/zfHairCells_Mattern2018/\" #directory path\n",
    "    fPath = \"zfHairCells_rpkm.csv\" #file path\n",
    "    tempPath = \"temp.txt\" # temporary file to store values\n",
    "    !head -1 {dPath + fPath} > {dPath + tempPath}\n",
    "    # EXACT MATCH ONLY by matching word (-w) and only at the start of each line (^) which corresponds to the symbol\n",
    "    !grep ^$genename {dPath + fPath} >> {dPath + tempPath}\n",
    "    gdf = pd.read_csv(dPath + tempPath)\n",
    "    return gdf\n",
    "\n",
    "def plotBars_haircell(genename):\n",
    "    gdf = findgene_sqcones(genename)\n",
    "    n = np.arange(1,4) # riboTagged\n",
    "    n = np.append(n, 3.5 + np.arange(1,4)) # Whatever this is\n",
    "    h_start = 2\n",
    "    h_end = 8\n",
    "#     h = np.power(2,(gdf.iloc[0,h_start:h_end]))\n",
    "    h = gdf.iloc[0,h_start:h_end].values\n",
    "    return n, h\n",
    "\n",
    "#hairCell Colors\n",
    "hcC = {\n",
    "    'h' : '#e147c0',\n",
    "    'c' : '#dac910',\n",
    "    'plt' : '',\n",
    "}\n",
    "\n",
    "hcC['plt']=[\n",
    "    hcC['h'],hcC['h'],hcC['h'],\n",
    "    hcC['c'],hcC['c'],hcC['c'],\n",
    "]\n",
    "\n",
    "print('LOADED')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "/Users/angueyraaristjm/Documents/LiMolec/zfRNAseq/20190827/20190827_DESeq2/genePlots/fpkm\n"
     ]
    }
   ],
   "source": [
    "# %cd ~/Dropbox/Presentations/RevealReboot/resources/20191002_RNAseq/\n",
    "%cd ~/Documents/LiMolec/zfRNAseq/20190827/20190827_DESeq2/genePlots/fpkm/"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "metadata": {},
   "outputs": [],
   "source": [
    "# dark background\n",
    "params = {\"ytick.color\" : \"w\",\n",
    "          \"xtick.color\" : \"w\",\n",
    "          \"axes.labelcolor\" : \"w\",\n",
    "          \"axes.edgecolor\" : \"w\",\n",
    "         \"axes.linewidth\" : 3,\n",
    "         \"xtick.major.width\" : 3,\n",
    "         \"ytick.major.width\" : 3,\n",
    "         \"xtick.major.size\" : 8,\n",
    "         \"ytick.major.size\" : 8,\n",
    "         \"text.color\" : \"w\"}\n",
    "plt.rcParams.update(params)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# white background\n",
    "params = {\"ytick.color\" : \"k\",\n",
    "          \"xtick.color\" : \"k\",\n",
    "          \"axes.labelcolor\" : \"k\",\n",
    "          \"axes.edgecolor\" : \"k\",\n",
    "         \"axes.linewidth\" : 3,\n",
    "         \"xtick.major.width\" : 3,\n",
    "         \"ytick.major.width\" : 3,\n",
    "         \"xtick.major.size\" : 8,\n",
    "         \"ytick.major.size\" : 8,\n",
    "         \"text.color\" : \"w\"}\n",
    "plt.rcParams.update(params)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "***\n",
    "# Index <a id='Index'>∮</a>\n",
    "***\n",
    "- [Plot Total Number of Reads](#nReads)\n",
    "- [PCA Plots](#pcaPlots)\n",
    "- [BarPlot: Opsins](#barOpsins)\n",
    "- [BarPlot: Custom](#barCustom)\n",
    "- [BarPlot: Many Genes](#barMany)\n",
    "- [BarPlot_subPlotsAutomatic: Custom](#barManyAuto)\n",
    "- [Heatmap: Opsins](#heatOpsins)\n",
    "- [Heatmap: Custom](#heatCustom)\n",
    "- [Heatmap: Other Data Sets](#heatOther)\n",
    "- [Sparkline plots](#sparkline)\n",
    "- [PCA](#pca)\n",
    "- [GO clustering](https://go.princeton.edu/) \n",
    "***"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "***\n",
    "## Total number of reads<a id='nReads'>∮</a>\n",
    "***\n",
    "[Back to Index](#Index)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### pygal version (does not import well into Intaglio)\n",
    "> hist = pygal.Histogram(x_title='',y_title='million of mapped reads',show_x_labels=False,title='',style=custom_style,legend_at_bottom=True,legend_at_bottom_columns=4,show_legend=False)  \n",
    "> hist.add('Rod',[(11.27,0.2,0.8),(11.77,1.2,1.8),(9.4,2.2,2.8),(14.99,3.2,3.8),(9.75,4.2,4.8),(9.08,5.2,5.8),])  \n",
    "> hist.add('UV',[(9.42,6.2,6.8),(6.52,7.2,7.8),(10.45,8.2,8.8),(9.46,9.2,9.8),(11.02,10.2,10.8)])  \n",
    "> hist.add('S',[(9.23,11.2,11.8),(9.33,12.2,12.8),(9.37,13.2,13.8),(14.95,14.2,14.8),(12.85,15.2,15.8),(10.96,16.2,16.8)])  \n",
    "> hist.add('M',[(10.40,18.2,18.8),(8.634,19.2,19.8),(7.36,20.2,20.8),(10.32,21.2,21.8),(8.33,22.2,22.8),(9.53,23.2,23.8),(9.526,17.2,17.8)])  \n",
    "> hist.add('L',[(8.99,24.2,24.8),(10.50,25.2,25.8),(10.03,26.2,26.8),(11.08,27.2,27.8),(10.44,28.2,28.8),(10.72,29.2,29.8)])  \n",
    "> hist.render_to_file('test.svg')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "#pyplot version (Sep_2019)\n",
    "plotname = ''\n",
    "n,h = plotNReads()\n",
    "fH, ax = plt.subplots()\n",
    "pH = plt.bar(n, h, width=0.8, bottom=None, align='center', data=None, color=pC['plt'])\n",
    "formatFigure_nReads(plotname, fH, ax, pH)\n",
    "plt.tight_layout()\n",
    "plt.savefig(\"{0}.svg\".format('nReads'), transparent=True, format=\"svg\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "np.mean(h)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": []
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "***\n",
    "## pcaPlots<a id='pcaPlots'>∮</a>\n",
    "***\n",
    "[Back to Index](#Index)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "def formatFigure_pca(plotname, figH, axH, plotH):\n",
    "    font_path = '/System/Library/Fonts/Avenir.ttc'\n",
    "    fontTicks = font_manager.FontProperties(fname=font_path, size=18)\n",
    "    fontLabels = font_manager.FontProperties(fname=font_path, size=22)\n",
    "    fontTitle = font_manager.FontProperties(fname=font_path, size=28)\n",
    "    \n",
    "    axH.ticklabel_format(style='sci',axis='y',scilimits=(0,4))\n",
    "    axH.ticklabel_format(style='sci',axis='x',scilimits=(0,4))\n",
    "    axH.set_title(plotname, fontproperties=fontTitle)\n",
    "    axH.spines['top'].set_visible(False)\n",
    "    axH.spines['right'].set_visible(False)\n",
    "    \n",
    "    for label in (axH.get_xticklabels() + axH.get_yticklabels()):\n",
    "        label.set_fontproperties(fontTicks)\n",
    "    axH.tick_params(axis = 'both', which = 'major', labelsize = 18)\n",
    "    axH.xaxis.offsetText.set_fontsize(18)\n",
    "    axH.yaxis.offsetText.set_fontsize(18)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "#pyplot version (Sep_2019)\n",
    "plotname = 'PCA_RodsCones'\n",
    "pc_df = pd.read_csv(\"~/Documents/LiMolec/zfRNAseq/20190827/20190827_DESeq2/00_rodsVcones/pcaData.csv\")\n",
    "pc1 = pc_df.iloc[:,0]\n",
    "pc2 = pc_df.iloc[:,1]\n",
    "fH, ax = plt.subplots(figsize=[5,5])\n",
    "pH = plt.scatter(pc1, pc2, color=pC['plt'])\n",
    "formatFigure_pca(plotname, fH, ax, pH)\n",
    "ax.set_xlabel('PC1 (var = 21%)', fontproperties=font_manager.FontProperties(fname='/System/Library/Fonts/Avenir.ttc', size=22))\n",
    "ax.set_ylabel('PC2 (var = 9%)', fontproperties=font_manager.FontProperties(fname='/System/Library/Fonts/Avenir.ttc', size=22))\n",
    "plt.xlim(-100,50)\n",
    "plt.ylim(-60,60)\n",
    "plt.tight_layout()\n",
    "plt.savefig(\"{0}.svg\".format(plotname), transparent=True, format=\"svg\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "#pyplot version (Sep_2019)\n",
    "plotname = 'PCA_UVS_ML'\n",
    "pc_df = pd.read_csv(\"~/Documents/LiMolec/zfRNAseq/20190827/20190827_DESeq2/00_USvsLM/pcaData.csv\")\n",
    "pc1 = pc_df.iloc[:,0]\n",
    "pc2 = pc_df.iloc[:,1]\n",
    "fH, ax = plt.subplots(figsize=[5,5])\n",
    "pH = plt.scatter(pc1, pc2, color=[\n",
    "    pC['u'],pC['u'],pC['u'],pC['u'],pC['u'],\n",
    "    pC['s'],pC['s'],pC['s'],pC['s'],pC['s'],pC['s'],\n",
    "    pC['m'],pC['m'],pC['m'],pC['m'],pC['m'],pC['m'],pC['m'],\n",
    "    pC['l'],pC['l'],pC['l'],pC['l'],pC['l'],pC['l']\n",
    "]\n",
    "                )\n",
    "formatFigure_pca(plotname, fH, ax, pH)\n",
    "ax.set_xlabel('PC1 (var = 13%)', fontproperties=font_manager.FontProperties(fname='/System/Library/Fonts/Avenir.ttc', size=22))\n",
    "ax.set_ylabel('PC2 (var = 10%)', fontproperties=font_manager.FontProperties(fname='/System/Library/Fonts/Avenir.ttc', size=22))\n",
    "plt.xlim(-80,80)\n",
    "plt.ylim(-60,60)\n",
    "plt.tight_layout()\n",
    "plt.savefig(\"{0}.svg\".format(plotname), transparent=True, format=\"svg\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "#pyplot version (Sep_2019)\n",
    "plotname = 'PCA_ML'\n",
    "pc_df = pd.read_csv(\"~/Documents/LiMolec/zfRNAseq/20190827/20190827_DESeq2/00_LvsM/pcaData.csv\")\n",
    "pc1 = pc_df.iloc[:,0]\n",
    "pc2 = pc_df.iloc[:,1]\n",
    "fH, ax = plt.subplots(figsize=[5,5])\n",
    "pH = plt.scatter(pc1, pc2, color=[\n",
    "    pC['m'],pC['m'],pC['m'],pC['m'],pC['m'],pC['m'],pC['m'],\n",
    "    pC['l'],pC['l'],pC['l'],pC['l'],pC['l'],pC['l']\n",
    "]\n",
    "                )\n",
    "formatFigure_pca(plotname, fH, ax, pH)\n",
    "ax.set_xlabel('PC1 (var = 23%)', fontproperties=font_manager.FontProperties(fname='/System/Library/Fonts/Avenir.ttc', size=22))\n",
    "ax.set_ylabel('PC2 (var = 10%)', fontproperties=font_manager.FontProperties(fname='/System/Library/Fonts/Avenir.ttc', size=22))\n",
    "plt.xlim(-80,80)\n",
    "plt.ylim(-80,80)\n",
    "plt.tight_layout()\n",
    "plt.savefig(\"{0}.svg\".format(plotname), transparent=True, format=\"svg\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "#pyplot version (Sep_2019)\n",
    "plotname = 'PCA_U_S'\n",
    "pc_df = pd.read_csv(\"~/Documents/LiMolec/zfRNAseq/20190827/20190827_DESeq2/00_UvsS/pcaData.csv\")\n",
    "pc1 = pc_df.iloc[:,0]\n",
    "pc2 = pc_df.iloc[:,1]\n",
    "fH, ax = plt.subplots(figsize=[5,5])\n",
    "pH = plt.scatter(pc1, pc2, color=[\n",
    "    pC['u'],pC['u'],pC['u'],pC['u'],pC['u'],\n",
    "    pC['s'],pC['s'],pC['s'],pC['s'],pC['s'],pC['s'],\n",
    "]\n",
    "                )\n",
    "formatFigure_pca(plotname, fH, ax, pH)\n",
    "ax.set_xlabel('PC1 (var = 13%)', fontproperties=font_manager.FontProperties(fname='/System/Library/Fonts/Avenir.ttc', size=22))\n",
    "ax.set_ylabel('PC2 (var = 10%)', fontproperties=font_manager.FontProperties(fname='/System/Library/Fonts/Avenir.ttc', size=22))\n",
    "plt.xlim(-100,100)\n",
    "plt.ylim(-80,80)\n",
    "plt.tight_layout()\n",
    "plt.savefig(\"{0}.svg\".format(plotname), transparent=True, format=\"svg\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# US vs LM: PC1 = 13% and PC2 = 10%\n",
    "# M vs L: PC1 = 23% and PC2 = 10%\n",
    "# UV vs S: PC1 = 15% and PC2 = 13%"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### tSNE plot"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "#pyplot version (Sep_2019)\n",
    "plotname = 'tSNE'\n",
    "pc_df = pd.read_csv(\"~/Documents/LiMolec/zfRNAseq/20190827/20190827_DESeq2/tSNEData.csv\")\n",
    "pc1 = pc_df.iloc[:,0]\n",
    "pc2 = pc_df.iloc[:,1]\n",
    "fH, ax = plt.subplots(figsize=[6,6])\n",
    "pH = plt.scatter(pc1, pc2, color=pC['plt'])\n",
    "formatFigure_pca(plotname, fH, ax, pH)\n",
    "ax.set_xlabel('tSNE1', fontproperties=font_manager.FontProperties(fname='/System/Library/Fonts/Avenir.ttc', size=22))\n",
    "ax.set_ylabel('tSNE2', fontproperties=font_manager.FontProperties(fname='/System/Library/Fonts/Avenir.ttc', size=22))\n",
    "ax.set_xticks([-1000,-500,0,500,1000])\n",
    "ax.set_yticks([-1000,-500,0,500,1000])\n",
    "plt.xlim(-1000,1000)\n",
    "plt.ylim(-1000,1000)\n",
    "plt.tight_layout()\n",
    "plt.savefig(\"{0}.svg\".format(plotname), transparent=True, format=\"svg\")"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "***\n",
    "## barPlot: Opsins<a id='barOpsins'>∮</a>\n",
    "***\n",
    "[Back to Index](#Index)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# from fx_RNAseq import *"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "genename = 'opn1lw1'\n",
    "n,h = plotBars(genename)\n",
    "h = h\n",
    "fH, ax = plt.subplots()\n",
    "pH = plt.bar(n, h, width=0.8, bottom=None, align='center', data=None, color=pC['plt'])\n",
    "formatFigure_Opsins(genename, fH, ax, pH)\n",
    "plt.savefig(\"{0}.svg\".format(genename), transparent=True, format=\"svg\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "findgene('rho')"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "***\n",
    "## barPlot: Custom Gene<a id='barCustom'>∮</a>\n",
    "***\n",
    "[Back to Index](#Index)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "> Genes to pay attention to but still haven't given interesting results\n",
    "> cdh (cadherins)\n",
    "> protocadherins\n",
    "> semaphorins\n",
    "> plexins (plx)\n",
    "> laminins (lamb) and integrins (itg)\n",
    "> dystroglycan (dag), dystrophin (dmd), pikachurin\n",
    "> neurexins\n",
    "> cadm\n",
    "> dscam and sidekicks (sdk)\n",
    "> pappaa\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "gdf = findgenelist('lrit')\n",
    "gdf"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "genename = 'foxq2'\n",
    "# genename = 'ncdn'\n",
    "# genename = 'chn2'\n",
    "# genename = 'grin1b'\n",
    "# genename = 'cdhr1a'\n",
    "# genename = 'lrit1a'\n",
    "n,h = plotBars(genename)\n",
    "fH, ax = plt.subplots()\n",
    "pH = plt.bar(n, h, width=0.8, bottom=None, align='center', data=None, color=pC['plt'])\n",
    "formatFigure(genename, fH, ax, pH)\n",
    "plt.savefig(\"{0}.svg\".format(genename), transparent=True, format=\"svg\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "***\n",
    "## barPlot: Many Genes Layout<a id='barMany'>∮</a>\n",
    "***\n",
    "[Back to Index](#Index)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "cd '~/Dropbox/LiLab/RNAseq/genePlots/combos'"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "genelist = ['gnat1','gnat2']\n",
    "n, h = plotBars_list(genelist)\n",
    "fH, ax = plt.subplots()\n",
    "pH = plt.bar(n, h, width=0.8, bottom=None, align='center', data=None, color=pC['plt'])\n",
    "formatFigure_list(genelist, fH, ax, pH)\n",
    "# plt.savefig(\"{0}.svg\".format('gnat'), transparent=True, format=\"svg\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "genelist = ['rho','opn1sw1','opn1sw2','opn1mw1','opn1mw2','opn1mw3','opn1mw4','opn1lw1','opn1lw2']\n",
    "n, h = plotBars_list(genelist)\n",
    "fH, ax = plt.subplots(figsize=[6*len(genelist),4])\n",
    "pH = plt.bar(n, h, width=0.8, bottom=None, align='center', data=None, color=pC['plt'])\n",
    "formatFigure_list(genelist, fH, ax, pH)\n",
    "plt.xlim(0,len(genelist)*10)\n",
    "# plt.savefig(\"{0}.svg\".format('opn1'), transparent=True, format=\"svg\")"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "***\n",
    "## barPlot: Many Genes Layout Using Subplots (proper way)<a id='barManySubPlots'>∮</a>\n",
    "***\n",
    "[Back to Index](#Index)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "shrink = 1#2/3"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "genelist = ['gnat1','gnat2']\n",
    "fH, axes = plt.subplots(1,len(genelist),sharey=True,figsize=[6*shrink*len(genelist),4*shrink])\n",
    "for i,g in enumerate(genelist,start=0):\n",
    "    n, h = plotBars(g)\n",
    "    plt.sca(axes[i]) \n",
    "    pH = plt.bar(n, h, width=0.8, bottom=None, align='center', data=None,\n",
    "                 color=pC['plt'])\n",
    "    if  i == 0:\n",
    "        formatFigure(g, fH, axes[i], pH)\n",
    "    else:\n",
    "        formatFigure_right(g, fH, axes[i], pH)\n",
    "plt.savefig(\"{0}.svg\".format('gnat'), transparent=True, format=\"svg\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "genelist = ['rho','rhol']\n",
    "\n",
    "fH, axes = plt.subplots(1,len(genelist),sharey=True,figsize=[6*shrink*len(genelist),4*shrink])\n",
    "for i,g in enumerate(genelist,start=0):\n",
    "    n, h = plotBars(g)\n",
    "    plt.sca(axes[i])\n",
    "    pH = plt.bar(n, h, width=0.8, bottom=None, align='center', data=None,\n",
    "                 color=pC['plt'])\n",
    "    if  i == 0:\n",
    "        formatFigure(g, fH, axes[i], pH)\n",
    "#         plt.ylim(0,3.5e6)\n",
    "#         axes[0].set_yticks(np.arange(0, 4.0e6, step=0.5e6))\n",
    "    else:\n",
    "        formatFigure_right(g, fH, axes[i], pH)\n",
    "plt.savefig(\"{0}.svg\".format('rho'), transparent=True, format=\"svg\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "genelist = ['opn1sw1','opn1sw1']\n",
    "\n",
    "fH, axes = plt.subplots(1,len(genelist),sharey=True,figsize=[6*shrink*len(genelist),4*shrink])\n",
    "for i,g in enumerate(genelist,start=0):\n",
    "    n, h = plotBars(g)\n",
    "    plt.sca(axes[i])\n",
    "    pH = plt.bar(n, h, width=0.8, bottom=None, align='center', data=None,\n",
    "                 color=pC['plt'])\n",
    "    if  i == 0:\n",
    "        formatFigure(g, fH, axes[i], pH)\n",
    "#         plt.ylim(0,2e6)\n",
    "#         axes[0].set_yticks(np.arange(0, 2.5e6, step=0.5e6))\n",
    "    else:\n",
    "        formatFigure_right(g, fH, axes[i], pH)\n",
    "plt.savefig(\"{0}.svg\".format('opn1sw1'), transparent=True, format=\"svg\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "genelist = ['opn1mw1','opn1mw2']\n",
    "\n",
    "fH, axes = plt.subplots(1,len(genelist),sharey=True,figsize=[6*shrink*len(genelist),4*shrink])\n",
    "for i,g in enumerate(genelist,start=0):\n",
    "    n, h = plotBars(g)\n",
    "    plt.sca(axes[i])\n",
    "    pH = plt.bar(n, h, width=0.8, bottom=None, align='center', data=None,\n",
    "                 color=pC['plt'])\n",
    "    if  i == 0:\n",
    "        formatFigure(g, fH, axes[i], pH)\n",
    "    else:\n",
    "        formatFigure_right(g, fH, axes[i], pH)\n",
    "plt.savefig(\"{0}.svg\".format('opn1mwA'), transparent=True, format=\"svg\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "genelist = ['opn1mw3','opn1mw4']\n",
    "\n",
    "fH, axes = plt.subplots(1,len(genelist),sharey=True,figsize=[6*shrink*len(genelist),4*shrink])\n",
    "for i,g in enumerate(genelist,start=0):\n",
    "    n, h = plotBars(g)\n",
    "    plt.sca(axes[i])\n",
    "    pH = plt.bar(n, h, width=0.8, bottom=None, align='center', data=None,\n",
    "                 color=pC['plt'])\n",
    "    if  i == 0:\n",
    "        formatFigure(g, fH, axes[i], pH)\n",
    "        plt.ylim(0,1.0e5)\n",
    "        axes[0].set_yticks(np.arange(0, 1.5e5, step=0.5e5))\n",
    "    else:\n",
    "        formatFigure_right(g, fH, axes[i], pH)\n",
    "\n",
    "\n",
    "plt.savefig(\"{0}.svg\".format('opn1mwB'), transparent=True, format=\"svg\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "genelist = ['opn1lw1','opn1lw2']\n",
    "\n",
    "fH, axes = plt.subplots(1,len(genelist),sharey=True,figsize=[6*shrink*len(genelist),4*shrink])\n",
    "for i,g in enumerate(genelist,start=0):\n",
    "    n, h = plotBars(g)\n",
    "    plt.sca(axes[i])\n",
    "    pH = plt.bar(n, h, width=0.8, bottom=None, align='center', data=None,\n",
    "                 color=pC['plt'])\n",
    "    if  i == 0:\n",
    "        formatFigure(g, fH, axes[i], pH)\n",
    "        plt.ylim(0,1.0e5)\n",
    "        axes[0].set_yticks(np.arange(0, 1.5e5, step=0.5e5))\n",
    "    else:\n",
    "        formatFigure_right(g, fH, axes[i], pH)\n",
    "#         axes[i].set_yticks(np.arange(0, 2.0e6, step=0.5e6))\n",
    "\n",
    "plt.savefig(\"{0}.svg\".format('opn1lw'), transparent=True, format=\"svg\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Using svgutils to automatically tile these figures the right way\n",
    "Figure(\"2400\", \"2000\",\n",
    "       SVG(\"rho.svg\").move(0, 290*0),\n",
    "       SVG(\"opn1sw.svg\").move(0, 290*1),\n",
    "       SVG(\"opn1mwA.svg\").move(0, 290*2),\n",
    "       SVG(\"opn1mwB.svg\").move(0, 290*3),\n",
    "       SVG(\"opn1lw.svg\").move(0, 290*4)\n",
    "        ).save(\"OpsinsAll.svg\")\n",
    "svgreload()"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Phototransduction"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "gdf = findgenelist('gngt')\n",
    "gdf = gdf.sort_values(by=['symbol'])\n",
    "gdf"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# genelist = ['saga','sagb','arr3a','arr3b']\n",
    "# plotName = 'arrestins'\n",
    "\n",
    "# genelist = ['gnat1','gnat2']\n",
    "# plotName = 'gnat'\n",
    "\n",
    "# genelist = ['gngt1','gngt2a','gngt2b']\n",
    "# plotName = 'gngt'\n",
    "\n",
    "# genelist = ['gnb1a','gnb1b','gnb3a','gnb3b','gnb4b','gnb5a','gnb5b']\n",
    "# plotName = 'gnb'\n",
    "\n",
    "# genelist = ['pde6a','pde6b','pde6c','pde6d','pde6ga','pde6gb','pde6c','pde6ha']\n",
    "# plotName = 'pde'\n",
    "\n",
    "# genelist = ['gucy1a1','gucy1b1','gucy1b2','gucy2c','gucy2d','gucy2f','gucy2g']\n",
    "# plotName = 'GC'\n",
    "\n",
    "# genelist = ['gucy2d','gucy2f']\n",
    "# plotName = 'GC_retinaSpecific'\n",
    "\n",
    "# genelist = ['guca1a','guca1b','guca1c','guca1d','guca1e','guca1g']\n",
    "# plotName = 'GCAP'\n",
    "\n",
    "# genelist = ['cnga1a','cnga1b','cnga3a','cnga3b','cnga4','cnga2a','cnga2b']\n",
    "# plotName = 'CNGalpha'\n",
    "\n",
    "# genelist = ['cnga1a','cnga1b','cnga3a','cnga3b']\n",
    "# plotName = 'CNGalpha_retinaSpecific'\n",
    "\n",
    "# genelist = ['cngb1a','cngb3.1','cngb3.2']\n",
    "# plotName = 'CNGbeta'\n",
    "\n",
    "\n",
    "# genelist = ['rgs9a','rgs9b','rgs9bp',\n",
    "#             'rgs3a','rgs6',\n",
    "#             'rgs11','rgs12a','rgs16','rgs20']\n",
    "# plotName = 'RGS'\n",
    "\n",
    "# genelist = ['grk1a','grk1b',\n",
    "#             'grk3',\n",
    "#             'grk4','grk5','grk5l',\n",
    "#             'grk6',\n",
    "#             'grk7a','grk7b']\n",
    "# plotName = 'GRK'\n",
    "\n",
    "# genelist = ['rcvrna','rcvrnb','rcvrn2','rcvrn3','ncalda','ncaldb']\n",
    "# plotName = 'rcvrn'\n",
    "\n",
    "\n",
    "\n",
    "fH, axes = plt.subplots(1,len(genelist),sharey=True,figsize=[6*len(genelist),4])\n",
    "for i,g in enumerate(genelist,start=0):\n",
    "    n, h = plotBars(g)\n",
    "    plt.sca(axes[i])\n",
    "    pH = plt.bar(n, h, width=0.8, bottom=None, align='center', data=None,\n",
    "                 color=pC['plt'])\n",
    "    if  i == 0:\n",
    "        formatFigure(g, fH, axes[i], pH)\n",
    "    else:\n",
    "        formatFigure_right(g, fH, axes[i], pH)\n",
    "plt.savefig(\"{0}.svg\".format(plotName), transparent=True, format=\"svg\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Others"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "gdf = findgenelist('cacn')\n",
    "gdf = gdf.sort_values(by=['symbol'])\n",
    "gdf"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "genelist = ['ctbp1','ctbp2a']\n",
    "plotName = 'ctbp'\n",
    "\n",
    "genelist = ['cadm1a','cadm1b','cadm2a','cadm2b','cadm3','cadm4']\n",
    "plotName = 'cadm'\n",
    "\n",
    "genelist = ['tjp1a','tjp1b','tjp2a','tjp2b','tjp3']\n",
    "plotName = 'tjp'\n",
    "\n",
    "genelist = ['nrxn1a','nrxn1b','nrxn2a','nrxn2b','nrxn3a','nrxn3b',]\n",
    "plotName = 'nrxn'\n",
    "\n",
    "genelist = ['eml1','eml2','eml3']\n",
    "plotName = 'eml'\n",
    "\n",
    "genelist = ['elfn1b','elfn2a']\n",
    "plotName = 'elfn'\n",
    "\n",
    "genelist = ['cpne1','cpne2','cpne3','cpne4a','cpne4b','cpne5a','cpne5b','cpne7','cpne8','cpne9']\n",
    "plotName = 'cpne'\n",
    "\n",
    "# genelist = ['tbx2a','tbx2b','tbxas1','tbx21']\n",
    "genelist = ['tbx2a','tbx2b']\n",
    "plotName = 'tbx2'\n",
    "\n",
    "genelist = ['chn1','chn2']\n",
    "plotName = 'chn'\n",
    "\n",
    "genelist = ['ncdn','ncdn']\n",
    "plotName = 'ncdn'\n",
    "\n",
    "genelist = ['sema7a','sema6a','sema6d','sema4ab','sema4ba','sema3fa','sema3fb']\n",
    "plotName = 'sema'\n",
    "\n",
    "genelist = ['cacna1aa','cacna1ab',\n",
    "            'cacna1da','cacna1db',\n",
    "            'cacna1fa','cacna1fb',\n",
    "            'cacna2d1a','cacna2d2a',\n",
    "            'cacna2d4a','cacna2d4b',\n",
    "            'cacna1sb',\n",
    "           ]\n",
    "plotName = 'CaV'\n",
    "\n",
    "# genelist = ['stxbp1b','stxbp4','stxbp2','stxbp5a','stxbp5b','stxbp5l','stxbp3']\n",
    "genelist = ['stxbp1a','stxbp1b',\n",
    "            'stxbp2','stxbp3','stxbp4',\n",
    "            'stxbp5a','stxbp5b','stxbp5l',\n",
    "            'stxbp6','stxbp6l']\n",
    "plotName = 'stxBP'\n",
    "\n",
    "genelist = ['stx1a','stx1b','stx2a','stx2b','stx3a',\n",
    "            'stx4','stx5a','stx5al','stx6','stx7l',\n",
    "            'stx8','stx10','stx11a','stx11b.1','stx11b.2',\n",
    "            'stx12','stx12l','stx16','stx17','stx18']\n",
    "plotName = 'stx'\n",
    "\n",
    "genelist = ['prickle1a','prickle1b','prickle2a','prickle2b','prickle3',]\n",
    "plotName = 'prickle'\n",
    "\n",
    "genelist = ['pcdh15a','pcdh15b','pcdh8','pcdh11','pcdh10a']\n",
    "plotName = 'pdch'\n",
    "\n",
    "genelist = ['chn1','chn2']\n",
    "plotName = 'chn'\n",
    "\n",
    "\n",
    "genelist = ['lrit1a','lrit1b','lrit2','lrit3a','lrit3b']\n",
    "plotName = 'lrit'\n",
    "\n",
    "genelist = ['pou4f1','pou4f2','pou4f3','rbpms','grm6a','grm6b','vsx2','gfap','prkca','syt2a','rpe65a', 'opn4']\n",
    "plotName = 'controls'\n",
    "\n",
    "\n",
    "# genelist = ['gpx1a','gpx1b','gpx4a','gpx4b']\n",
    "# genelist = ['rgs9bp','RGS9BP']\n",
    "# genelist = ['chchd1','chchd10']\n",
    "# genelist = ['yipf6','yipf1']\n",
    "# genelist = ['sema3fb','sema3fa','sema6d']\n",
    "# genelist = ['esama','esamb']\n",
    "\n",
    "# genelist = ['pcdh1gc6','tex264a']\n",
    "# plotName = 'test'\n",
    "\n",
    "            \n",
    "fH, axes = plt.subplots(1,len(genelist),sharey=False,figsize=[6*len(genelist),4])\n",
    "for i,g in enumerate(genelist,start=0):\n",
    "    n, h = plotBars(g)\n",
    "    plt.sca(axes[i])\n",
    "    pH = plt.bar(n, h, width=0.8, bottom=None, align='center', data=None,\n",
    "                 color=pC['plt'])\n",
    "    if  i == 0:\n",
    "        formatFigure(g, fH, axes[i], pH)\n",
    "    else:\n",
    "        formatFigure_right(g, fH, axes[i], pH)\n",
    "plt.savefig(\"{0}.svg\".format(plotName), transparent=True, format=\"svg\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "gdf = findgenelist('opn1')\n",
    "# gdf = gdf.sort_values(by=['symbol'])\n",
    "# gdf = gdf.sort_values(by=['baseMean'], ascending=False)\n",
    "# gdf['symbol']\n",
    "gdf"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### According to Neuhauss recoverin1a in enriched in rods and UV-cones"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "***\n",
    "## heatMap: Opsins<a id='barManySubPlots'>∮</a>\n",
    "***\n",
    "[Back to Index](#Index)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Not sure this is helpful.  \n",
    "Maybe use this to show differential expression by normalizing by mean expression in correct type"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "gdf = findgenelist('opn1')\n",
    "gdf2 = findgene('rho')\n",
    "gdf2\n",
    "gdf = gdf.append(gdf2)\n",
    "gdf.index = range(len(gdf))\n",
    "gdf\n",
    "\n",
    "Opsins = [\"Rhodopsin (rho)\",\"UV-opsin (opn1sw1)\", \"S-opsin (opn1sw2)\",\"M-opsin (opn1mw1)\",\"M-opsin (opn1mw2)\",\"M-opsin (opn1mw3)\",\"M-opsin (opn1mw4)\",\"L-opsin (opn1lw1)\",\"L-opsin (opn1lw2)\"]\n",
    "Samples = [\"\",\"\",\"R\",\"\",\"\",\"\",\n",
    "           \"\",\"\",\"U\",\"\",\"\",\n",
    "           \"\",\"\",\"S\",\"\",\"\",\"\",\n",
    "           \"\",\"\",\"\",\"M\",\"\",\"\",\"\",\n",
    "           \"\",\"\",\"L\",\"\",\"\",\"\",\n",
    "          ]\n",
    "Samples = [\n",
    "          ]\n",
    "\n",
    "data = gdf.iloc[0:9,7:37]\n",
    "data = data.reindex([8,4,5,7,0,2,6,1,3])\n",
    "data.index = range(len(data))\n",
    "# np.around(np.divide(data.values,1e6),2)\n",
    "# opsinData = np.divide(data.values,1e6)\n",
    "opsinData = np.divide(data.values,1)\n",
    "\n",
    "# # log2 representation does not work well\n",
    "# log2matrix = np.vectorize(log2)\n",
    "# opsinData = log2matrix(opsinData)\n",
    "\n",
    "params = {\"ytick.color\" : \"w\",\n",
    "          \"xtick.color\" : \"w\",\n",
    "          \"axes.labelcolor\" : \"w\",\n",
    "          \"axes.edgecolor\" : \"w\",\n",
    "         \"axes.linewidth\" : 2,\n",
    "         \"xtick.major.width\" : 2,\n",
    "         \"ytick.major.width\" : 2,\n",
    "         \"xtick.major.size\" : 0,\n",
    "         \"ytick.major.size\" : 0}\n",
    "plt.rcParams.update(params)\n",
    "\n",
    "plotName = \"heatMap_opsins\"\n",
    "fig, ax = plt.subplots()\n",
    "\n",
    "im, cbar = heatmap(opsinData, Opsins, Samples, ax=ax,\n",
    "                   cmap=\"inferno\", cbarlabel=\"fpkm\")\n",
    "\n",
    "\n",
    "# texts = annotate_heatmap(im, valfmt=\"{x:.1f} t\")\n",
    "\n",
    "# fig.tight_layout()\n",
    "plt.savefig(\"{0}.svg\".format(plotName), transparent=True, format=\"svg\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "***\n",
    "## heatMap: Custom<a id='heatCustom'>∮</a>\n",
    "***\n",
    "[Back to Index](#Index)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "plotName = \"test\"\n",
    "\n",
    "# gdf = findgene('rho')\n",
    "# gdf2 = findgenelist('opn1sw')\n",
    "# gdf = gdf.append(gdf2)\n",
    "# gdf2 = findgenelist('opn1mw')\n",
    "# gdf2 = gdf2.sort_values(by=[\"symbol\"])\n",
    "# gdf = gdf.append(gdf2)\n",
    "# gdf2 = findgenelist('opn1lw')\n",
    "# gdf2 = gdf2.sort_values(by=[\"symbol\"])\n",
    "# gdf = gdf.append(gdf2)\n",
    "\n",
    "# gdf.index = range(len(gdf))\n",
    "# plotName = \"heatMap_opsins\"\n",
    "\n",
    "# gdf = findgenelist('slc24') -> NCKX transporter; slc25a5 is mitochondrial but known to express in eye\n",
    "# plotName = \"slc24\"\n",
    "\n",
    "# gdf = findgenelist('sema')\n",
    "# gdf = gdf.sort_values(by=[\"symbol\"])\n",
    "# plotName = \"heatMap_sema\"\n",
    "\n",
    "# gdf = findgenelist('cdh')\n",
    "# gdf = gdf.sort_values(by=[\"baseMean\"],ascending=False)\n",
    "# plotName = \"heatMap_cdh\"\n",
    "\n",
    "# gdf = findgenelist('eph')\n",
    "# gdf = gdf.drop([1,14,15,16]) #not ephrin receptors\n",
    "# gdf = gdf.sort_values(by=[\"symbol\"])\n",
    "# plotName = \"heatMap_eph\"\n",
    "\n",
    "\n",
    "# gdf = findgenelist('efn')\n",
    "# gdf = gdf.sort_values(by=[\"symbol\"])\n",
    "# plotName = \"heatMap_efn\"\n",
    "\n",
    "# gdf = findgenelist('pdgf')\n",
    "# gdf = findgenelist('rgr')\n",
    "# gdf = findgenelist('lmo')\n",
    "# gdf.index = range(len(gdf))\n",
    "# gdf.sort_values(by=[\"log2FoldChange\"])\n",
    "\n",
    "\n",
    "## genes in 2019 Papal, Swaroop: # igsf11, epha10, tuft1, prtg, lrfn2, vasn\n",
    "\n",
    "# gdf = findgenelist('igsf')\n",
    "# gdf = findgenelist('tuft')\n",
    "# gdf = findgenelist('prtg')\n",
    "# gdf = findgenelist('lrfn')\n",
    "# gdf = findgenelist('vasn')\n",
    "\n",
    "# gdf = findgenelist('rdh')\n",
    "# gdf = gdf.sort_values(by=[\"symbol\"])\n",
    "# plotName = \"rdhHeatMap\"\n",
    "\n",
    "# gdf = findgenelist('cyp2')\n",
    "# gdf = findgenelist('cyb')\n",
    "\n",
    "# gdf = findgenelist('sag')\n",
    "# plotName = \"heatMap_sag\"\n",
    "\n",
    "# gdf = findgenelist('dio')\n",
    "# gdf = gdf.sort_values(by=[\"symbol\"])\n",
    "# plotName = \"heatMap_dio\"\n",
    "\n",
    "# gdf = findgenelist('lrrn')\n",
    "# gdf = gdf.sort_values(by=[\"symbol\"])\n",
    "# plotName = \"heatMap_lrrn\"\n",
    "\n",
    "# gdf = findgenelist('dnmt')\n",
    "# gdf = gdf.sort_values(by=[\"symbol\"])\n",
    "# plotName = \"heatMap_dnmt\"\n",
    "\n",
    "# gdf = findgenelist('tet')\n",
    "\n",
    "# gdf = findgenelist('isl')\n",
    "# gdf = gdf.drop([1])\n",
    "\n",
    "# gdf = findgenelist('gj')\n",
    "# gdf = gdf.sort_values(by=[\"symbol\"])\n",
    "# plotName = \"heatMap_gj\"\n",
    "\n",
    "# gdf = findgenelist('fgf')\n",
    "# gdf = findgenelist('bmp')\n",
    "# gdf = findgenelist('fzd')\n",
    "# gdf = findgenelist('edar')\n",
    "# gdf = findgenelist('axin')\n",
    "\n",
    "# gdf = findgenelist('cadm') # cadm2a in S-Cones probably(?) and cadm3 in all photoreceptors (Hunter, JCN, 2011)\n",
    "# gdf = gdf.sort_values(by=[\"symbol\"])\n",
    "# plotName = \"heatMap_cadm\"\n",
    "\n",
    "# gdf = findgenelist('lrit') # lrit1b for rods and cones; lrit2 for cones; lrit3 for rods\n",
    "# gdf = gdf.sort_values(by=[\"symbol\"])\n",
    "# plotName = \"heatMap_lrit\"\n",
    "\n",
    "# gdf = findgenelist('elfn') # not detected\n",
    "# gdf = gdf.sort_values(by=[\"symbol\"])\n",
    "# plotName = \"heatMap_elfn\"\n",
    "\n",
    "# gdf = findgenelist('tenm') # for RGCs in zebrafish (Antinucci, Cell Reports, 2013)\n",
    "\n",
    "# gdf = findgenelist('foxo')\n",
    "\n",
    "# gdf = findgenelist('sdk')\n",
    "\n",
    "# gdf = findgenelist('crmp')\n",
    "\n",
    "\n",
    "\n",
    "# Even More\n",
    "# gdf = findgenelist('thr')\n",
    "# gdf = gdf.sort_values(by=[\"symbol\"])\n",
    "# plotName = \"heatMap_thr\"\n",
    "\n",
    "# gdf = findgenelist('lingo')\n",
    "# gdf = gdf.sort_values(by=[\"symbol\"])\n",
    "# plotName = \"heatMap_lingo\"\n",
    "\n",
    "# gdf = findgenelist('mitf')\n",
    "\n",
    "# gdf = findgenelist('prickle')\n",
    "# gdf = gdf.sort_values(by=[\"symbol\"])\n",
    "# plotName = \"heatMap_prickle\"\n",
    "\n",
    "# gdf = findgenelist('rdh')\n",
    "# gdf = findgenelist('sdh')\n",
    "# lingo, rx, mitf\n",
    "\n",
    "# gdf = findgenelist('dag')\n",
    "# can't find pikachurin: # gdf = findgenelist('egflam') gdf = findgenelist('si:dkeyp-100d11.2-001')\n",
    "\n",
    "# gdf = findgenelist('kcn')\n",
    "# gdf = gdf.sort_values(by=[\"baseMean\"], ascending=False)\n",
    "# gdf = gdf.drop(np.arange(20,109))\n",
    "# gdf\n",
    "\n",
    "# gdf = findgenelist('kcnj')\n",
    "# gdf = gdf.sort_values(by=[\"baseMean\"], ascending=False)\n",
    "# gdf\n",
    "\n",
    "# gdf = findgenelist('slc')\n",
    "# gdf = gdf.sort_values(by=[\"baseMean\"], ascending=False)\n",
    "# gdf = gdf.drop(np.arange(40,410))\n",
    "\n",
    "# gdf = findgenelist('rcv')\n",
    "# gdf = gdf.sort_values(by=[\"symbol\"])\n",
    "# plotName = \"heatMap_recoverin\"\n",
    "\n",
    "# gdf = findgenelist('sag')\n",
    "# gdf = gdf.append(findgenelist('arrb'))\n",
    "# gdf = gdf.append(findgenelist('arr3'))\n",
    "# gdf = gdf.sort_values(by=[\"symbol\"])\n",
    "# plotName = \"heatMap_arr\"\n",
    "\n",
    "\n",
    "# gdf = findgenelist('cng')\n",
    "# gdf = gdf.sort_values(by=[\"symbol\"])\n",
    "# plotName = \"heatMap_cng\"\n",
    "\n",
    "# gdf = findgenelist('cnga')\n",
    "# gdf = gdf.sort_values(by=[\"symbol\"])\n",
    "# plotName = \"heatMap_cnga\"\n",
    "\n",
    "# gdf = findgenelist('cngb')\n",
    "# gdf = gdf.sort_values(by=[\"symbol\"])\n",
    "# plotName = \"heatMap_cngb\"\n",
    "\n",
    "# gdf = findgenelist('myo')\n",
    "# gdf = gdf.sort_values(by=[\"symbol\"])\n",
    "\n",
    "# gdf = findgenelist('crx')\n",
    "# gdf = gdf.sort_values(by=[\"baseMean\"], ascending=False)\n",
    "# plotName = \"heatMap_crx\"\n",
    "\n",
    "\n",
    "# gdf = findgenelist('neurod')\n",
    "# gdf = gdf.sort_values(by=[\"symbol\"])\n",
    "# plotName = \"heatMap_neurod\"\n",
    "\n",
    "# gdf = findgenelist('ror')\n",
    "# gdf = gdf.drop([5,6])\n",
    "# gdf = gdf.sort_values(by=[\"symbol\"])\n",
    "# plotName = \"heatMap_ror\"\n",
    "\n",
    "\n",
    "# gdf = findgenelist('nr')\n",
    "# gdf = gdf.sort_values(by=[\"baseMean\"], ascending=False)\n",
    "# gdf = gdf.drop(np.arange(5,76))\n",
    "# plotName = \"heatMap_nrl\"\n",
    "\n",
    "\n",
    "# gdf = findgenelist('pcdh')\n",
    "# gdf = gdf.drop([1]) #highest ones that obscure others.\n",
    "# # gdf = gdf.drop([1,15]) #highest ones that obscure others.\n",
    "# gdf = gdf[gdf.baseMean > .45]\n",
    "# gdf = gdf.sort_values(by=[\"baseMean\"], ascending=False)\n",
    "# plotName = \"heatMap_pcdh_partial\"\n",
    "\n",
    "# gdf = findgenelist('hcn')\n",
    "# gdf = gdf.sort_values(by=[\"symbol\"])\n",
    "\n",
    "# gdf = findgenelist('gabr')\n",
    "# gdf = gdf.drop([0,2]) #highest ones that obscure others.\n",
    "# gdf = gdf.sort_values(by=[\"symbol\"])\n",
    "\n",
    "# gdf = findgenelist('opn4')\n",
    "# gdf = gdf.drop([1,2]) #highest ones that obscure others.\n",
    "# gdf = gdf.sort_values(by=[\"baseMean\"], ascending=False)\n",
    "# gdf = gdf.drop(np.arange(5,76))\n",
    "# gdf = gdf.sort_values(by=[\"symbol\"])\n",
    "# plotName = \"heatMap_opn4\"\n",
    "\n",
    "# CSNB causes: Nyx, cacna1f, GPR179 (Kamermans, 2019)\n",
    "# gdf = findgenelist('nyx') # bipolar cell -> not expressed in photoreceptors\n",
    "# gdf = findgenelist('gpr1') # bipolar cell -> not expressed in photoreceptors\n",
    "# gdf = findgenelist('cacna') # bipolar cell -> not expressed in photoreceptors\n",
    "\n",
    "\n",
    "# For Karen Carleton\n",
    "\n",
    "# gdf = findgenelist('rx')\n",
    "# gdf = gdf.drop([2,3,4,5,7,8,9,10,11,12,13,14,15,16,17,18,19])\n",
    "# gdf = gdf.sort_values(by=[\"symbol\"])\n",
    "# # gdf = gdf.drop([1])\n",
    "# plotName = \"heatMap_rx\"\n",
    "\n",
    "# gdf = findgenelist('tbx')\n",
    "# gdf = gdf.drop([3,17])\n",
    "# plotName = \"heatMap_tbx\"\n",
    "# gdf = gdf.drop([1,2])\n",
    "\n",
    "# gdf = findgenelist('tbx2')\n",
    "# gdf = gdf.drop([0,3,4])\n",
    "# gdf = gdf.sort_values(by=[\"symbol\"])\n",
    "# plotName = \"heatMap_tbx2\"\n",
    "\n",
    "# gdf = findgenelist('six')\n",
    "# gdf = gdf.sort_values(by=[\"symbol\"])\n",
    "# plotName = \"heatMap_six\"\n",
    "\n",
    "# gdf = findgenelist('lin')\n",
    "# gdf = gdf.drop([1,3,4,5,10,11,13,16])\n",
    "# plotName = \"heatMap_lin\"\n",
    "\n",
    "\n",
    "# gdf = findgenelist('sall') #sall4 mutations cause Duane syndrome; sall4 seems to be related to sall1 in zebrafish\n",
    "# plotName = \"heatMap_sall\"\n",
    "\n",
    "# For Mark Hankins\n",
    "# gdf = findgenelist('opn')\n",
    "# gdf = gdf.sort_values(by=[\"symbol\"])\n",
    "# gdf = gdf.drop([1,4,24,0,3,23,5,11,2])\n",
    "\n",
    "# gdf = findgenelist('rgr')\n",
    "# gdf = findgenelist('lrit')\n",
    "\n",
    "# gdf = findgenelist('ndrg') #potential target of tbx2\n",
    "\n",
    "# from YiRong Peng's talk: ON SACs: slit2 and fezf1; OFF SACs: tenm3 and rnd1?; contactins; Foveal cones: vtn and cacna1f\n",
    "# gdf = findgenelist('cacna2d2b')\n",
    "# gdf = gdf.drop([0])\n",
    "# gdf = gdf.sort_values(by=[\"symbol\"])\n",
    "\n",
    "# gdf = findgenelist('nrxn')\n",
    "# gdf = gdf.sort_values(by=[\"symbol\"])\n",
    "# plotName = \"heatMap_nrxn\"\n",
    "\n",
    "\n",
    "# gdf = findgenelist('frmpd')\n",
    "# gdf = gdf.sort_values(by=[\"symbol\"])\n",
    "# plotName = \"heatMap_frmpd\"\n",
    "\n",
    "# gdf = findgenelist('ntng')\n",
    "# gdf = gdf.sort_values(by=[\"symbol\"])\n",
    "# plotName = \"heatMap_netrins\"\n",
    "\n",
    "# gdf = findgenelist('dscam')\n",
    "# gdf = gdf.sort_values(by=[\"symbol\"])\n",
    "# plotName = \"heatmap_dscams\"\n",
    "\n",
    "\n",
    "# gdf = findgenelist('dag')\n",
    "# gdf = gdf.sort_values(by=[\"symbol\"])\n",
    "# plotName = \"heatmap_dag\"\n",
    "\n",
    "# gdf = findgenelist('syne') #involved in photoreceptor migration to ONL\n",
    "# gdf = gdf.sort_values(by=[\"symbol\"])\n",
    "# plotName = \"heatmap_syne\"\n",
    "# gdf = findgenelist('rasgrf') #also involved in phtoreceptor migration\n",
    "\n",
    "# gdf = findgenelist('syt')\n",
    "# plotName = \"heatmap_syt\"\n",
    "\n",
    "# gdf = findgenelist('rfx')\n",
    "# plotName = \"heatmap_rfx\"\n",
    "\n",
    "# gdf = findgenelist('anks') #anks3-/- causes RP\n",
    "# gdf = gdf.sort_values(by=[\"symbol\"])\n",
    "# plotName = \"heatmap_anks\"\n",
    "\n",
    "# gdf = findgenelist('spock') #https://zfin.org/action/image/view/ZDB-IMAGE-050208-655 -> GCL enriched?\n",
    "# gdf = gdf.sort_values(by=[\"symbol\"])\n",
    "# plotName = \"heatmap_anks\"\n",
    "\n",
    "# gdf = findgenelist('hcls') # Laura Campello: downstream of nrl and crx in mouse\n",
    "\n",
    "# gdf = findgenelist('slc32')  #from GO analysis: slc32a1 = L-cone specific + synaptic vesicle transporter.\n",
    "# gdf = findgenelist('slc') \n",
    "# gdf = gdf.sort_values(by=[\"baseMean\"],ascending=False)\n",
    "# gdf = gdf[gdf.baseMean > 100]\n",
    "\n",
    "# gdf = findgenelist('frmpd') #frmpd2 interacts with lrit1 in cones \n",
    "\n",
    "# from sqRNAseq + antibodies: rab32,ell2, ankrd33, pex5l, egr1, lamb1, chst3? or 7?, flna, cdh20, synj1\n",
    "# gdf = findgenelist('ell') # ell2 in sqRNAseql seems cone specific\n",
    "# gdf = findgenelist('ankrd33') #it's there, 33aa cone enriched; 33ba rods and 33bb cones.\n",
    "# gdf = findgenelist('pex5') #pek5l in sqRNAseq; pex5 in rods\n",
    "# gdf = findgenelist('egr') # egr1 looks interesting\n",
    "\n",
    "\n",
    "\n",
    "# gdf = findgenelist('gnat')\n",
    "# gdf2 = findgenelist('gnb')\n",
    "# gdf = gdf.append(gdf2)\n",
    "# gdf2 = findgenelist('gng')\n",
    "# gdf = gdf.append(gdf2)\n",
    "# gdf.index = range(len(gdf))\n",
    "# plotName = \"heatMap_transducin\"\n",
    "\n",
    "# gdf = findgenelist('aldh1')\n",
    "gdf = findgenelist('elovl')\n",
    "gdf = gdf.sort_values(by=[\"baseMean\"],ascending=False)\n",
    "# gdf = findgenelist('gnat')\n",
    "gdf = gdf[gdf.baseMean > 0]\n",
    "# gdf = gdf.drop(np.arange(20,93))\n",
    "# gdf = gdf.drop([4,24])\n",
    "# gdf = gdf.sort_values(by=[\"symbol\"])\n",
    "# gdf = gdf.sort_values(by=[\"padj\"])\n",
    "\n",
    "gdf\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "genenames = gdf['symbol'].values\n",
    "samplenames = [] #skipping names to replace in intaglio\n",
    "data = gdf.iloc[0:,7:37].values\n",
    "\n",
    "# log2matrix = np.vectorize(log2)\n",
    "# data = log2matrix(gdf.iloc[0:,7:37].values)\n",
    "\n",
    "fig, ax = plt.subplots(figsize=[12,12])\n",
    "im, cbar = heatmap(data, genenames, samplenames, ax=ax,\n",
    "                   cmap=\"inferno\", cbarlabel=\"fpkm\")\n",
    "# texts = annotate_heatmap(im, valfmt=\"{x:.1f} t\")\n",
    "\n",
    "plt.savefig(\"{0}.svg\".format(plotName), transparent=True, format=\"svg\")\n",
    "plt.show()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "plotName = 'test'\n",
    "\n",
    "# foxq2 = S specific;\n",
    "\n",
    "# genelist = ['wrb','get4'] # disrupts photoreceptors synapses (Danielle, 2016)\n",
    "# genelist = ['pappaa','igf2bp2b'] # In Miller, 2018 model pappaa cleaves igf-bp to increase igf1 availability. unfortunately, this is supposed to bind mRNA not IGF itself\n",
    "# genelist = ['gpx1a','gpx1b','gpx4a','gpx4b']\n",
    "# genelist = ['rgs9bp','RGS9BP']\n",
    "# genelist = ['chchd1','chchd10']\n",
    "# genelist = ['yipf6','yipf1','meis1b','pbx1a']\n",
    "# genelist = ['sema3fb','sema3fa','sema6d']\n",
    "# genelist = ['esama','esamb']\n",
    "# genelist = ['il34','il7r']\n",
    "# genelist = ['cxxc4','cxxc1a']\n",
    "# genelist = ['mbd3b','mbd3a']\n",
    "# genelist = ['cdr2a','cdr2l']\n",
    "# genelist = ['coro6','coro2a','coro1ca']\n",
    "# genelist = ['auts2a','auts2b']\n",
    "# genelist = ['plxnb1a','plxna2']\n",
    "# genelist = ['cux1a','cux1b','cux2b']\n",
    "# genelist = ['arih1','arih2','arih1l']\n",
    "# genelist = ['crhbp','crhr2','crhr1']\n",
    "# genelist = ['chn1','chn2']\n",
    "# genelist = ['robo1','robo2','robo3','robo4'] #nothing here\n",
    "# genelist = ['ntng1a','ntng2a','ntng2b'] #low counts but interesting. Netrins are important for RGC axons going towards optic disc. ntng2b in mouse mediates rod-hc recognition!\n",
    "# genelist = ['fat1a','fat1b','yap1','scrib','cc2d2a'] #fat1 important for renal development along yap1/hippo and scribble; cc2d2a involved in cilium development\n",
    "# genelist = ['c1qbp','c1qa','c1qb'] #keep this in mind. Beth Stevens work shows that microglia engulf synapses through activation by c1q; maybe c1qbp is produced in healthy cells (inside mitochondria) to keep c1q bound and gets released upon damage.\n",
    "# genelist = ['cc2d2a','rab8'] #Bahcman 2011: cc2d2a-/- causes Joubert syndrom. Important for cilium development by association with rab8. Not expressed in adults?\n",
    "# genelist = ['per2','per3']\n",
    "# genelist = ['gpsm2','gpsm1b','gpr107','gpr108'] #\n",
    "# genelist = ['tbx2a','tbx2b','gdf6a','cnbpa','egr1', 'ndrg1a','ndrg1b'] # fate genes\n",
    "\n",
    "#Hensley, 2011 smarca4\n",
    "# genelist = ['smarca4a','guk1a','guk1b','dtnbp1a','elovl4b','id2b','ckmt2a','ckmt2b','aanat2','pbx1a'] #smarca4 regulated genes: follow up elovl4b, ckmt2a/b, pbx1a\n",
    "# genelist = ['nme2a','nme3',''] #smarca4 regulated genes: follow up elovl4b, ckmt2a/b, pbx1a\n",
    "\n",
    "# genelist = ['adgrl3.1','ntng2a','ntng2b'] #low counts but interesting. \n",
    "# genelist = ['cplx4a','cplx4b','cplx4c'] #low counts but interesting. \n",
    "\n",
    "# genelist = ['frmpd2', 'lrit3a'] #low counts but interesting. \n",
    "genelist = ['si:busm1-57f23.1 ', 'mhc1uba','serinc2'] # \n",
    "\n",
    "# genelist = ['ccdc28a','ccdc25','ccdc173','sox6','prdm9','prdm1b']\n",
    "\n",
    "# genelist = ['per2','bhlhe40','nr1d2b','tefa','aanat2','aanat1'] #circadian genes. Correlate with collection time?\n",
    "\n",
    "# genelist = ['elovl4b','elovl8a'] #a cone specific gene. \n",
    "# genelist = ['pbx1a','egr1','egr3','egr4'] # fishing for new targets\n",
    "# genelist = ['foxg1b','foxq2','foxo3b','lmo4b','nfia'] # fishing for new targets\n",
    "# genelist = ['anxa13','prdx5','mpp5b','mpp6a','serpina10a'] # fishing for new targets\n",
    "# genelist = ['rbp4l','rbp3','impg1a','impg2b','ntm'] # fishing for new targets\n",
    "# genelist = ['tapt1b','tapt1a','frmpd2','wdr37','fabp7a','tle3a'] # fishing for new targets\n",
    "# genelist = ['ankrd33aa','ankrd33ab','dhx32a','dhx32b','wdr17']\n",
    "\n",
    "\n",
    "# genelist = ['slc1a8a','slc1a8b','slc1a4','slc1a7a'] # EAATs (glutamate uptake)\n",
    "# genelist = ['pmela','slc6a6b','naa35','ush2a'] # sex-associated rod genes in Sun, 2018 (all enriched in females, except pmela)\n",
    "\n",
    "# genelist = ['pcloa','pclob']\n",
    "# genelist = ['mtdha','mtdhb']\n",
    "# genelist = ['igsf9b','igsf21a','igsf21b']\n",
    "# genelist = ['rx3','lin9','igsf21b','foxq2','foxo3b','foxg1b','rxrga','smad5']\n",
    "# genelist = ['sall1a','sall1b','sall2','sall4']\n",
    "# genelist = ['rbm8a','rbm39a','rbm39b','rbmx','rbm22','rbm5']\n",
    "# genelist = ['rbm5','rbm4.1','rbm4.2']\n",
    "# genelist = ['kif5ba','kif1aa','kif1ab','kif1bp'] #kif5-/- : Charcot-MarieTooth; Campbell 2013: kif5s in all retinal layers, but only kif5ba and bb are in ONL\n",
    "# genelist = ['kif3a','kif3b','kif16ba','kif19','kif25'] #\n",
    "# genelist = ['robo2','slit1a','nbeaa','prdm1b'] #\n",
    "\n",
    "\n",
    "# genelist = ['pcdh2ab9','pcdh2g28','pcdh15a','pcdh15b','pcdh11','pcdh10a'] #interesting ones\n",
    "# plotName = 'pcdhs'\n",
    "\n",
    "# genelist = ['sema7a','sema6d','sema3fb','sema3fa','sema4ab','sema4ba'] #the good ones\n",
    "# genelist = ['sema6a','sema3aa','sema6e','sema5bb','sema4c','sema6bb','sema6dl','sema5ba'] # the inconsistent ones\n",
    "# genelist = ['sema6ba','sema5a','sema3c','sema3b','sema3e','sema3d','sema3h','sema3ab','sema4e','sema4aa','sema3gb','sema3ga','sema4bb','sema3bl','sema4ga','sema4gb'] # the negative ones\n",
    "\n",
    "# genelist = ['ncdn','flrt3'] \n",
    "# genelist = ['odf2a','odf2b']\n",
    "# genelist = ['syne1a','syne1b','syne2b','syne3']\n",
    "# genelist = ['rasgrf2a','rasgrf2b']\n",
    "# genelist = ['rasgrf2a','rasgrf2b','nmnat2','rbx1','nrf1','unc119.2','unc119b'] # involved in migration\n",
    "\n",
    "# genelist = ['syt5a','syt5b','syt1a','syt1b','syt14a']\n",
    "\n",
    "\n",
    "# genelist = ['myo19','myo16','myo7bb','myo5b','myo1cb','myo5aa','myo7aa','myo6a','myo6b']\n",
    "# genelist = ['fgfr1op','fgfr1op2','fgf1b','fgf13a','fgf11b']\n",
    "# genelist = ['prdm1b','fgfr1op2']\n",
    "# genelist = ['frmpd1a','gpsm2'] # Ulisse + Pahlberg: these proteins inteact with rod transducin \n",
    "\n",
    "# genelist = ['epha7','frmpd2','pkp4','chd7','hook1']\n",
    "# genelist = ['thrb','samd7','samd1a','samd11','dscaml1'] #samd11:https://www.ncbi.nlm.nih.gov/pubmed/27734943\n",
    "# genelist = ['thrb','nrl']\n",
    "# genelist = ['opn3','opn4.1','opn4a','opn4b','opn5','opn6a','opn9'] #\n",
    "            \n",
    "fH, axes = plt.subplots(1,len(genelist),sharey=False,figsize=[6*len(genelist),4])\n",
    "for i,g in enumerate(genelist,start=0):\n",
    "    n, h = plotBars(g)\n",
    "    plt.sca(axes[i])\n",
    "    pH = plt.bar(n, h, width=0.8, bottom=None, align='center', data=None,\n",
    "                 color=pC['plt'])\n",
    "#     if  i == 0:\n",
    "    formatFigure(g, fH, axes[i], pH)\n",
    "#     else:\n",
    "#         formatFigure_right(g, fH, axes[i], pH)\n",
    "plt.savefig(\"{0}.svg\".format(plotName), transparent=True, format=\"svg\")\n",
    "fH.tight_layout()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "***\n",
    "## heatMap: photoreceptors, rods, Müller Glia, microglia<a id='heatOther'>∮</a>\n",
    "***\n",
    "[Back to Index](#Index)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 227,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>symbol</th>\n",
       "      <th>genename</th>\n",
       "      <th>baseMean</th>\n",
       "      <th>log2FoldChange</th>\n",
       "      <th>lfcSE</th>\n",
       "      <th>pvalue</th>\n",
       "      <th>padj</th>\n",
       "      <th>R1</th>\n",
       "      <th>R2</th>\n",
       "      <th>R3</th>\n",
       "      <th>...</th>\n",
       "      <th>M4</th>\n",
       "      <th>M5</th>\n",
       "      <th>M6</th>\n",
       "      <th>S7</th>\n",
       "      <th>L1</th>\n",
       "      <th>L2</th>\n",
       "      <th>L3</th>\n",
       "      <th>L4</th>\n",
       "      <th>L5</th>\n",
       "      <th>L6</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>5</th>\n",
       "      <td>rxraa</td>\n",
       "      <td>retinoid X receptor alpha a</td>\n",
       "      <td>3.193433</td>\n",
       "      <td>5.106533e-07</td>\n",
       "      <td>0.001443</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>...</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.230013</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>rxrab</td>\n",
       "      <td>retinoid x receptor alpha b</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>...</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>rxrba</td>\n",
       "      <td>retinoid x receptor beta a</td>\n",
       "      <td>1.932289</td>\n",
       "      <td>9.706625e-07</td>\n",
       "      <td>0.001443</td>\n",
       "      <td>0.546298</td>\n",
       "      <td>0.770803</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>1.990136</td>\n",
       "      <td>8.686704</td>\n",
       "      <td>...</td>\n",
       "      <td>1.803009</td>\n",
       "      <td>1.034601</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>1.379058</td>\n",
       "      <td>0.718297</td>\n",
       "      <td>1.124813</td>\n",
       "      <td>1.832872</td>\n",
       "      <td>0.680537</td>\n",
       "      <td>13.279284</td>\n",
       "      <td>0.331964</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>rxrbb</td>\n",
       "      <td>retinoid x receptor beta b</td>\n",
       "      <td>2.451623</td>\n",
       "      <td>5.821352e-07</td>\n",
       "      <td>0.001443</td>\n",
       "      <td>0.524443</td>\n",
       "      <td>0.755996</td>\n",
       "      <td>9.156169</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>3.506605</td>\n",
       "      <td>...</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.908774</td>\n",
       "      <td>1.870303</td>\n",
       "      <td>3.914267</td>\n",
       "      <td>0.429391</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>1.882712</td>\n",
       "      <td>2.685764</td>\n",
       "      <td>0.452056</td>\n",
       "      <td>0.301923</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>rxrga</td>\n",
       "      <td>retinoid x receptor gamma a</td>\n",
       "      <td>6.981624</td>\n",
       "      <td>4.293240e-07</td>\n",
       "      <td>0.001443</td>\n",
       "      <td>0.846805</td>\n",
       "      <td>0.938479</td>\n",
       "      <td>47.819299</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>...</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>12.851492</td>\n",
       "      <td>20.433176</td>\n",
       "      <td>37.204805</td>\n",
       "      <td>30.856741</td>\n",
       "      <td>33.704397</td>\n",
       "      <td>12.122965</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>rxrgb</td>\n",
       "      <td>retinoid X receptor gamma b</td>\n",
       "      <td>2.709465</td>\n",
       "      <td>1.455723e-07</td>\n",
       "      <td>0.001443</td>\n",
       "      <td>0.525238</td>\n",
       "      <td>0.756586</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>...</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>6.828193</td>\n",
       "      <td>11.548887</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>13.175123</td>\n",
       "      <td>4.733322</td>\n",
       "      <td>0.000000</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>6 rows × 37 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "  symbol                     genename  baseMean  log2FoldChange     lfcSE  \\\n",
       "5  rxraa  retinoid X receptor alpha a  3.193433    5.106533e-07  0.001443   \n",
       "4  rxrab  retinoid x receptor alpha b  0.000000             NaN       NaN   \n",
       "2  rxrba   retinoid x receptor beta a  1.932289    9.706625e-07  0.001443   \n",
       "0  rxrbb   retinoid x receptor beta b  2.451623    5.821352e-07  0.001443   \n",
       "3  rxrga  retinoid x receptor gamma a  6.981624    4.293240e-07  0.001443   \n",
       "1  rxrgb  retinoid X receptor gamma b  2.709465    1.455723e-07  0.001443   \n",
       "\n",
       "     pvalue      padj         R1        R2        R3  ...        M4        M5  \\\n",
       "5       NaN       NaN   0.000000  0.000000  0.000000  ...  0.000000  0.000000   \n",
       "4       NaN       NaN   0.000000  0.000000  0.000000  ...  0.000000  0.000000   \n",
       "2  0.546298  0.770803   0.000000  1.990136  8.686704  ...  1.803009  1.034601   \n",
       "0  0.524443  0.755996   9.156169  0.000000  3.506605  ...  0.000000  0.908774   \n",
       "3  0.846805  0.938479  47.819299  0.000000  0.000000  ...  0.000000  0.000000   \n",
       "1  0.525238  0.756586   0.000000  0.000000  0.000000  ...  0.000000  0.000000   \n",
       "\n",
       "         M6        S7         L1         L2         L3         L4         L5  \\\n",
       "5  0.000000  0.000000   0.000000   0.000000   0.000000   0.230013   0.000000   \n",
       "4  0.000000  0.000000   0.000000   0.000000   0.000000   0.000000   0.000000   \n",
       "2  0.000000  1.379058   0.718297   1.124813   1.832872   0.680537  13.279284   \n",
       "0  1.870303  3.914267   0.429391   0.000000   1.882712   2.685764   0.452056   \n",
       "3  0.000000  0.000000  12.851492  20.433176  37.204805  30.856741  33.704397   \n",
       "1  0.000000  0.000000   6.828193  11.548887   0.000000  13.175123   4.733322   \n",
       "\n",
       "          L6  \n",
       "5   0.000000  \n",
       "4   0.000000  \n",
       "2   0.331964  \n",
       "0   0.301923  \n",
       "3  12.122965  \n",
       "1   0.000000  \n",
       "\n",
       "[6 rows x 37 columns]"
      ]
     },
     "execution_count": 227,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "geneTag = 'rxr'\n",
    "gdf = findgenelist(geneTag)\n",
    "# gdf = gdf.sort_values(by=[\"baseMean\"],ascending=False)\n",
    "# gdf = gdf[gdf.baseMean > 0]\n",
    "# gdf = gdf.drop(np.arange(20,93))\n",
    "# gdf = gdf.drop([4,24])\n",
    "gdf = gdf.sort_values(by=[\"symbol\"])\n",
    "# gdf = gdf.sort_values(by=[\"padj\"])\n",
    "\n",
    "rdf = findgenelist_rods(geneTag)\n",
    "rdf = rdf.sort_values(by=[\"symbol\"])\n",
    "# rdf = rdf[rdf.baseMean > 0]\n",
    "\n",
    "mdf = findgenelist_glia(geneTag)\n",
    "mdf = mdf.sort_values(by=[\"symbol\"])\n",
    "# mdf = mdf[mdf.baseMean > 0]\n",
    "\n",
    "sqdf = findgenelist_sqcones(geneTag)\n",
    "sqdf = sqdf.sort_values(by=[\"symbol\"])\n",
    "# sqdf = sqdf[sqdf.baseMean > 0]\n",
    "\n",
    "hcdf = findgenelist_haircell(geneTag)\n",
    "hcdf = hcdf.sort_values(by=[\"symbol\"])\n",
    "# hcdf = hcdf[hcdf.baseMean > 0]\n",
    "gdf"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 228,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "image/png": "iVBORw0KGgoAAAANSUhEUgAAA7MAAAEeCAYAAABYA0J2AAAABHNCSVQICAgIfAhkiAAAAAlwSFlzAAALEgAACxIB0t1+/AAAADh0RVh0U29mdHdhcmUAbWF0cGxvdGxpYiB2ZXJzaW9uMy4xLjAsIGh0dHA6Ly9tYXRwbG90bGliLm9yZy+17YcXAAAgAElEQVR4nO3dfZgcVZ3o8W/PTIgQBojIS0yUIK4IkReD+vCy10t4MWF0Ua8oIOwmKubRcMU3iuyy17th1V2gXCCog9e7KEHFBAP4giEBTFy4vChhXBJiEIkEQmANK0uICQkhqfvHqUl6ZronPTOZrq6u7+d5+unqqlP1O3WmeqZ/fc6pKSVJgiRJkiRJedKSdQUkSZIkSRook1lJkiRJUu6YzEqSJEmScsdkVpIkSZKUOyazkiRJkqTcacu6Asq/KIr2Az4HXBPH8YtZ10eSJElS87NnVrvDfsA/pM+SJEmSNOzsmW1SURSdDCwpW5UAzwK/Af45juP7s6iXVMmCSQtnAe/rWDLlHf1tB1YAx3YsmXJMP8eaAvwcGN+xZMqa3V9bFd3kGRuOAv4XcBLwOuBJ4CYgXtTZvjnLuknDYUTXYScTPlM8B7xx68RVr1Yp1wL8ATgE+KutE1fdXrdKSnX08FETTia8J9qPW77izxlXp9DsmW1+E4BDgbcCHwdeAe6NomhyprWSBucG4OgFkxYe3U+ZqcDdJrIaDpNnbDgauB94GTgbeDvwTSACZmZYNakeDiJ8sVjNZGBsneoiSfbMFsDTcRx3f2P0eBRFdwEPAF8EFmVXLWlQfgk8BZwPXNJ744JJC/cFPgBMq2utVCT/DPx6UWf7tLJ1KyfP2HAn8PtsqiTVzQJgOvDjKtunA7cBH65bjSQVmj2zBRPHcQLcC7wx67pIA9WxZEoC3Aicu2DSwkq/v84GNlH9g5Y0VG8mTNfoYVFn++OLOtuTDOoj1VMnMHlE12F9PkOM6DrsYEKv7XV1r5WkwrJntpiOAlZ3v4ii6ATgS4T5X22ED2pfi+O4R0IQRdFIwjC68wnzYdYSEou5vcq1ALMIwz0PAh4FojiOy+fwSoM1h3C9ngws7rVtKnBTx5IpW+pdKRXGU8D7Js/YcPmizvb/zLoyUp3dC/wW+AThxo/lPg48sXXiqiUjug6re8UkFZM9swURRVFLFEVviKLocsKclmvT9acC/0ZIbicB7wLuBm6Ooujisv1LwE+BCwg3PjmKMJzoNOB7vcJ9kfCH7lPARMIfvuujKBoxXOen4uhYMmUVcB/hS5UdFkxa+GbgROA7WdRLhfF3hJEtKyfP2BBNnrHBUS4qmm8BHx/RdVhr94oRXYeVCH/3v5VZrSQVkj2zze8/oyiC8LNuAZYBH47jeEEURXsA3waujuO4/MYlK6Ioehr4VhRFN8dx/DThj9QJwFvjOH42Lfd4FEWLgbt6xfxL4OdxHN8BEEXRx4HRcRxvHaZzVPHcAPzLgkkLZ3QsmdJ999hpwCMdS6b0GQIq7S6LOtsfTm8C9b8JPVNXTJ6x4W7gikWd7b/ItnZSXdwIXA50AD9L150OjCGMnJGkurFntvmdCBxL6K16BjgpjuP56baTgDcBV1XY77vABuDM9PVfAz8oS2QBiON4O3BZr33vB86JomhmFEVvA7bFcfz87jgZNa1a5hqWl7kZGEF6fS6YtLBEuEbtldWwW9TZ/sSizva/AQ4APgKMBu6ePGPD+f3vKeXf1omrNgA/AD5ZtvqTwA+3Tlz1Yja1klRUJrPN7/E4jh8j3Pl1JKE3odsY4OU4jv/Ye6f0RlF/AF5fVvbJajF67w58hTB/cTnwTBRFn+yzl7TTekJCUM3rgB0fkjqWTHmJcMfM7uRhEnAw4QOWVBeLOttfXtTZPp/wpeESwr0CpCLoBDpGdB02dkTXYQcC78cbP0nKgMlsQcRx/CLh5k2fj6JoQrr6OWDPKIoO7F0+nSN7aFqmu+z4Kof/i16xtsdxfGUcx0cC4wjDjv5PFEXvGPKJqFmtBsYtmLRwnyrbJ1B207LUHGDKgkkL9yd8cfKTjiVT/jRsNZSqWNTZvhWYT/h9JzW9rRNXLQN+RZiC9DHgka0TVy3NtlaSishktljmAL8GrkuT1fsJCcLnK5T9G2Bfwk2fIPR4fTSKojHlhdLjfKnXukOiKNoXII7jtXEcX0roeTtm952KmswSYAvwud4bFkxa+E7gv9P33+3cDawjfJj6EA4xVh1MnrFh8uQZG46ssOm/Af9e7/pIGbqO8Pv3AuyVlZQRbwBVIHEcJ1EUXQg8DEyN4/iGKIqmAz+LomgUMI+QUHQAlwJ/H8fxU+nu/5fwPzzvj6Lo7wk3kjqEMHy5912KfwRsT++G/AJhKOiewP8b1hNUbnUsmbJ+waSFnwe+vWDSwv2Am4CNwCmEOdm3dCyZ8vNe+2xfMGnh94B/BJ4H7qxztVUwk2dsaCMMJT5s8owNXyL8m5J2wgf69xFuficVxY+Aq4F96PUv+qQCeePDR03Y1GtdctzyFU9VLK3dzp7Zgonj+BHgm0AcRdFr4zi+C3gn8BbgDuAeYArw3jiOryzbLyH8S59vA18mJMRfJ/yfz7/uFeYDwCrgJ8BDhPmMZ8Zx/LthPDXlXMeSKdcDf0X4d06/ALqAGcAVwDlVdptDmAs+p2PJlO31qKeKa1Fn+6vAqYR/bfY/Cb8HbwH2Ao5b1Nn+SIbVk+pq68RVWwgjYuZsnbiq94d5qShWEO4pU/5YlWmNCqaUJLXcRFSSJEmSpMZR1GHGs6osGzffMY3bvDGN27wxjdu8MY3b3HGziGnc5o1p3OaNOaxxi9ozW37SJeM2TUzjNm9M4zZvTOM2b0zjNnfcIp1r0eIW6VyLFrfpztU5s5IkSZKk3DGZlSRJkiTljsmsJEmSJCl3TGYlSZIkSbljMitJkiRJyh2TWUmSJElS7uT6X/PMmzcvWbp0adbVkIZs0tJTs66CmszsI07KugrSkC2+4NisqyDl1oNTX5N1FaTd4rjlKz4G3FBpW657ZkePHp11FSRJkiRJw2d8tQ25TmYlSZIkScVkMitJkiRJyp1cz5kFklKprc4ht6XPrQWIW6Rz3Rn31e3fr2vUtpbzSZKErK7l1486uW4Rn934y3SpWNdUa8tr6xt1+wsZXFPF/H1RjLjZnWuWvxv92Ro333F9/zRv3CKda4ibJMllwKxKW+2ZlSRJkiTljsmsJEmSJCl3TGYlSZIkSbljMitJkiRJyh2TWUmSJElS7pjMSpIkSZJyx2RWkiRJkpQ7JrOSJEmSpNwZbDLbBpR2UWbEII8tSZIkSVK/BpLMjgduB+YCK4EF6fq9gS5gLDANuAZYDMwGDkrLLwIeBi4oO95M4C7gHuBWYGSFmLOApJ+HJEmSJKmA2gZY/hTgeGAZcCMwHTgGuAJYm5Y5HTgRWE/onb0EeDqNtTzd7xVgfrofwJeBs9NtkiRJkiT1a6DJbBchkQWYASxNX88rK3MbIZEF2A6cBbwHaAXGAAcQEt8xwOXAfsA44PsDr74kSZIkqYgGOmd2c9nyBGAdcCQwukqZiwlJ63sJPbaPEebajgO+AXw2XT+bynNwZ6Xrqz0kSZIkSQU02BtAtQOdwHnAlcD1VcodDDwCbAMOJwxJBtgfeB54ltA7/MFB1kOSJEmSVEADHWbc7TrgamANYZ7rGYRhx5t6lfs6cBMwFXgKuBsYRRia/ATwEPAc8EC6XpIkSZKkXRpIMrsaOC1dPr/XtnOr7PMHwg2jKvn0AGJLkiRJkrTDYIcZS5IkSZKUGZNZSZIkSVLumMxKkiRJknLHZFaSJEmSlDsms5IkSZKk3DGZlSRJkiTljsmsJEmSJCl3TGYlSZIkSblTSpIk6zoMRVIqlbKugzRkSZLgtazdyWtKzcDrWBo83z9qFkmSXAbMqrTNnllJkiRJUu60ZV2BoWutc7xtBYpbpHM1bvPGNG7zxjRu88Y0bvPGNG7zxjRu88ZshLiV2TMrSZIkScodk1lJkiRJUu6YzEqSJEmScsdkVpIkSZKUOyazkiRJkqTcMZmVJEmSJOWOyawkSZIkKXdMZiVJkiRJuZNVMntRRnElSZIkSU3AZFaSJEmSlDsOM5YkSZIk5c5gktk2oLSLMiMGcVxJkiRJkmpSazI7HrgdmAusBBak6/cGuoCxwDTgGmAxMBs4KC2/CHgYuKDXMWcCDwD3pvvtWSHuLCDp5yFJkiRJKqC2AZQ9BTgeWAbcCEwHjgGuANamZU4HTgTWE3pnLwGeTuMsT/d7BTgU2A6ckO43E/gq8IXBn4okSZIkqSgGMsy4i5DIAswgJJ4HAPPKytxGSGQhJKtnAQuBO4AxaXmAPwFx2X4x0DGQikuSJEmSimsgyezmsuUJwDrgSGB0lTIXExLY9xJ6bB9j13Nte5uV7lPtIUmSJEkqoMHcAKod6ATOA64Erq9S7mDgEWAbcDhhSHK3/QnJbreL2TkPV5IkSZKkfg1kzmy364CrgTWEObBnEIYdb+pV7uvATcBU4CngbmBUuu3JNPZ9hOHIjwDRIOoiSZIkSSqgUpLk+qbASak0mHx8KLalz60FiFukc802bpIkFONaLt7PNqu49b+mitfGxYlbpOs4xA382Ro3z3F9/zRv3CKda4ibJMllhOmnfQxmmLEkSZIkSZkymZUkSZIk5Y7JrCRJkiQpd0xmJUmSJEm5YzIrSZIkScodk1lJkiRJUu6YzEqSJEmScsdkVpIkSZKUO/X+T8rDYNuuixg3hzGN29xxi3SuRYtbpHMtWtwinWvR4hbpXIsWt0jnWrS4RTrX6uyZlSRJkiTlThP0zLbWOV74NqKltE9do25PXgLgoFEn1S3mHzfely5l08b1PFfI/nxhjzrHfSV9ruf5dp9rvX/1vJo+ey0Pn+6fbVbvn2LFLZX2rFvEJHkZgJEj3lC3mABbtq5Jl4rwuxGyuaa6r6dRdYwJSbIxxGVkfeOyJV0qzvs2q/dPVj/brK7l+rZzaOOj9/poHWPCsk03pUtZfYarzJ5ZSZIkSVLumMxKkiRJknLHZFaSJEmSlDsms5IkSZKk3DGZlSRJkiTljsmsJEmSJCl3TGYlSZIkSbkznMnsscC7h/H4kiRJkqSCMpmVJEmSJOWOw4wlSZIkSbkzmGS2DSjtosyIQRxXkiRJkqSa1JrMjgduB+YCK4EF6fq9gS5gLDANuAZYDMxOtx8IzAfuA+4FPlh2zJnAXcA9wK3AyApxZwFJPw9JkiRJUgG1DaDsKcDxwDLgRmA6cAxwBbA2LXM6cCKwnpDcnkOYN/sYsA9wB7Aa+A0hyb0i3e/LwNnpcSVJkiRJ6tdAhhl3ERJZgBnAF4ADgHllZW4jJLLd5hASWYCXgGuBD6WvxwA/IvTOngW8YSAVlyRJkiQV10CS2c1lyxOAdcCRwOgqZSopEYYHjwO+AXyW0Js7m8rzcGel66s9JEmSJEkFNJgbQLUDncB5wJXA9f2UnQocXrbfZwi9t/sDzwPPEoY6f7Di3pIkSZIkVTCQObPdrgOuBtYQ5rieQRh2vKlC2bnAVwhDigGuIgxXLgFPAA8BzwEPAKMGURdJkiRJUgHVmsyuBk5Ll8/vte3cKvvckD4qSYBP1xhbkiRJkqQeBjPMWJIkSZKkTJnMSpIkSZJyx2RWkiRJkpQ7JrOSJEmSpNwxmZUkSZIk5Y7JrCRJkiQpd0xmJUmSJEm5YzIrSZIkScqdUpIkWddhKJJSqZR1HaQhS5IEr2XtTl5TagZex9Lg+f5Rs0iS5DJgVqVt9sxKkiRJknKnLesKDNXrR51c13jPbvxlutRa17iwDYC21gPrFvHVbesAaG15bd1iAmzb/kK6tEdd48Ir6XM2P9vxe3fUNerqPy8AYK+Rh9Ut5qYtq9KlbNp41Mi31DXqxi2PAzBu71PrGveZP/8iXapnO2/LIObOuIfsPbmuUZ/68yIASoysa9yELelSPf98v5o+Z/Ozza6NsznfIr1vs/o7XyqNqmvUJNkIwB5tY+sW85VX16ZL9f6on/Hvi4x+tvX8rLzzc3KxfrbV2DMrSZIkScodk1lJkiRJUu6YzEqSJEmScsdkVpIkSZKUOyazkiRJkqTcMZmVJEmSJOWOyawkSZIkKXdMZiVJkiRJuTNcyeyxwLv72X7RMMWVJEmSJBWAyawkSZIkKXccZixJkiRJyp2BJrNtQGkXZUYMsi6SJEmSJNWklmR2PHA7MBdYCSxI1+8NdAFjgWnANcBiYHa6/UBgPnAfcC/wwV7HnQk8kG67BtizQuxZQNLPQ5IkSZJUQG01ljsFOB5YBtwITAeOAa4A1qZlTgdOBNYTkttzCPNmHwP2Ae4AVgO/AQ4FtgMnpPvOBL4KfGEI5yJJkiRJKohahxl3ERJZgBmEpPMAYF5ZmdsIiWy3OYREFuAl4FrgQ+nrPwFxWdkY6Ki51pIkSZKkQqs1md1ctjwBWAccCYyuUqaSEgMfGjwr3a/aQ5IkSZJUQAO9AVQ70AmcB1wJXN9P2anA4WX7fYbQewuwP3BxWdmL2TkXV5IkSZKkftU6Z7bbdcDVwBrC3NkzCMOON1UoOxf4CjAmfX0VYbgywJNp7PsIc2cfAaIB1kWSJEmSVFC1JLOrgdPS5fN7bTu3yj43pI9q3pw+X15DfEmSJEmSehjoMGNJkiRJkjJnMitJkiRJyh2TWUmSJElS7pjMSpIkSZJyx2RWkiRJkpQ7JrOSJEmSpNwxmZUkSZIk5Y7JrCRJkiQpd0pJkmRdh6FISqVS1nWQhixJEryWtTt5TakZeB1Lg+f7R80iSZLLgFmVttkzK0mSJEnKnbasKzBUj06ZVNd4b1u4JF1qrWtc2JZB3CxiFjnuHnWO+0r6XJxrasyod9c16nMb70mXinAtZ/3+KVbckSPeULeIW7auSZeK1cbFiFukcy1a3BDzqL3OqWNMWL5pLgAf2uczdY17y0tfB+Cmt02ta9yPPjoHgH86dEbdYl76ZCcA33zL9LrFBLjw8W8DcMSos+oad+XG+f1ut2dWkiRJkpQ7JrOSJEmSpNwxmZUkSZIk5Y7JrCRJkiQpd0xmJUmSJEm5YzIrSZIkScodk1lJkiRJUu6YzEqSJEmScsdkVpIkSZKUOyazkiRJkqTcMZmVJEmSJOXOYJLZNqC0izIjBnFcSZIkSZJqUmsyOx64HZgLrAQWpOv3BrqAscA04BpgMTA73X4W8CBwL3AjcC1wcrptJnAXcA9wKzCyQtxZQNLPQ5IkSZJUQG0DKHsKcDywjJCYTgeOAa4A1qZlTgdOBNYDE4FLgMnAfwETCEntrWnZ+em+AF8Gzk6PK0mSJElSvwaSzHYRElmAGcDS9PW8sjK3ERJZgA8DVxISWYAVwM1lZccAlwP7AeOA7w+k4pIkSZKk4hrInNnNZcsTgHXAkcDoKmVKVB8KPA74BvBZQm/ubCrPw52Vrq/2kCRJkiQV0GBuANUOdALnEXper69Sbj7wRWDf9PURwEfS5f2B54FnCb3DHxxEPSRJkiRJBTWQYcbdrgOuBtYQ5rieQRh2vKlXuaXA14CfA1sJw43nEXprlwFPAA8BzwEPAKMGURdJkiRJUgHVmsyuBk5Ll8/vte3cfva7lZ03fIIwtPgFQkL76RpjS5IkSZLUw2CGGddqInBp2etD0nUrhjGmJEmSJKkABjPMuFaPAZ8AHgZeTh+fArYPY0xJkiRJUgEMZzK7CbhwGI8vSZIkSSqo4RxmLEmSJEnSsDCZlSRJkiTljsmsJEmSJCl3TGYlSZIkSbljMitJkiRJyp1SkiRZ12EoklKplHUdpCFLkgSvZe1OXlNqBl7H0uD5/lGzSJLkMmBWpW32zEqSJEmScmc4/89snbTWOd62AsUNMd886sw6xoQnNv40XcqmjS886KK6Rv3mH69Nl4pzTRXjXIsWt0jnWrS4RTrXnXH3ec2RdY360ubfAtDSsl/dYm7f/mK6VKyfbTHihpj77nlUHWPC+peXA/CWUR+oa9zHN/44XcrmZ3v0Xh+tW8Rlm24C4DV7HFK3mACbX3kqXcrq/VOZPbOSJEmSpNwxmZUkSZIk5Y7JrCRJkiQpd0xmJUmSJEm5YzIrSZIkScodk1lJkiRJUu6YzEqSJEmScsdkVpIkSZKUO1klsxdlFFeSJEmS1ARMZiVJkiRJueMwY0mSJElS7gw0mW0DSrsoM2KQdZEkSZIkqSa1JLPjgduBucBKYEG6fm+gCxgLTAOuARYDs9PtZwEPAvcCNwLXAieXHXcm8EC6/RpgzwqxZwFJPw9JkiRJUgG11VjuFOB4YBkhMZ0OHANcAaxNy5wOnAisByYClwCTgf8CJhCS1lvTsocC24ET0tczga8CXxj8qUiSJEmSiqLWYcZdhEQWYAYh6TwAmFdW5jZCIgvwYeBKQiILsAK4uazsn4C47HUMdNRca0mSJElSodWazG4uW54ArAOOBEZXKVNi9wwDnpUeq9pDkiRJklRAA70BVDvQCZxH6Hm9vkq5+cAXgX3T10cAHynbvj9wcdnri9k5F1eSJEmSpH7VOme223XA1cAawtzZMwjDjjf1KrcU+Brwc2ArYbjxPHb21j6Zxr6PMHf2ESAaePUlSZIkSUVUSzK7GjgtXT6/17Zz+9nvVnbe8AngG8AL6fKb0+fLa4gvSZIkSVIPAx1mXKuJwKVlrw9J160YpniSJEmSpAIZ6DDjWj0GfAJ4GHg5fXyKMKRYkiRJkqQhGa5kdhNw4TAdW5IkSZJUcMM1zFiSJEmSpGFjMitJkiRJyp1SkiS7LtWg7rzzzuSuu+7KuhrSkE1aemrWVVCTmX3ESVlXQRqyxRccm3UVpNx6cOprsq6CtFsct3zFZcCsStvsmZUkSZIk5Y7JrCRJkiQpd0xmJUmSJEm5k+s5s8A04IaM6yBJkiRJGh7TqJLz5T2ZlSRJkiQVkMOMJUmSJEm5YzIrSZIkScodk1lJkiRJUu6YzEqSJEmScsdkVpIkSZKUOyazkiRJkqTcMZmVJEmSJOWOyawkSZIkKXdMZiVJkiRJuWMyK0mSJEnKHZNZSZIkSVLumMxKkiRJknLHZFaSJEmSlDsms5IkSZKk3GnLugLSbpZkXQFJkiQNSSnrCigf7JmVJEmSJOWOyawkSZIkKXccZqym1VLaY+eLUvdolfD9Tan8e5xSS49t5dtLpfLve7rXtfY5Rne5iuvS8tWOW6K1x7qWHsfoG6tlR6y+x22pEKuFCseg7zFaqBA/3V6+riXpea4tFY7bXSasK1WIX+pRrqVsNFGpe1vZup3HKFtXqlSu1zFKFbbVuK6lbIDTzuN1v95pZz3KypcqHaPvupYhlq9Up12Wr3iMpMcxyo9T+RhJ1Xq0lI30L/U6RliX9FjXHbvaOfQuX21d73PocdwayvdYt6N8hXPpUd+kT313Hjfpu66/c6lY3wrrKpXrdfyedavtGOX79j5ef9tqPW6ppfwY26seo8e6lkrH7blv+XH7O0b3fj3XVdi3xvLUetzu+rVUOOde26qdV+99+9vWo241ltuxrbx8d7mWSvUtP0avbVD2Juz1DDveLKUK63q9kehTcEe5lp7PlbZVKZfsWNdaY/n+yu38+5mUeh+3bFul8hXL7XpdUraNlrYK5SusK7X1LF+2vKN8qe+28vKlHccY0Xdd+lyqUL5UdtyWHeta+5RrSevbynlIA2XPrCRJkiQpd0xmJUmSJEm54zBjNbpW4L3AeHper1dlUhtJkiRJDcFkVo3uZ8BmYDmwfRdlJUmSJBWEyawa3Tjg6KwrIUmSJKmxOGdWje4O4D1ZV0KSJElSY7FnVo3uQeA2whcvWwk3/U+AfbKslCRJkqRsmcyq0f0LcAJhzmyyi7KSJEmSCsJhxmp0vwcexURWkiRJUhl7ZtXongN+SZg7u6Vsvf+aR5IkSSowk1k1uifTxx7pA+yllSRJkgrPZFaNbg6wute6d2ZQD0mSJEkNxDmzanS3AGPLXr8b+E5GdZEkSZLUIExm1eg+BfwYOBjoAK5NnyVJkiQVmMOM1egeAi4C7gQ2A6cDz2daI0mSJEmZM5lVo/oZPW/0tBewHrg+fX1m3WskSZIkqWGYzKpRfS3rCkiSJElqXCazalT/VrZ8MPAuQk/tQ8B/ZFIjSZIkSQ3DG0Cp0V0A/Br4H8BZwIPAxzOtkSRJkqTM2TOrRhcBbwf+lL7eH7gf/z2PJEmSVGj2zKrRPQNsKHu9AViTUV0kSZIkNQh7ZtXo1gK/An5CmDP7fsKw4y+k26/KqF6SJEmSMmTPrBrV99LnM4Efs/Pf9PwEeA5oTx+SJEmSCsieWTWq44BDgKeBr1fY/kJ9qyNJkiSpkZjMqlF9C1gIHAosLVtfIvTSvimLSkmSJElqDA4zVqO6FjgC+C4hce1+HIqJrCRJklR4JrNqdJ/OugKSJEmSGk8pSZJdl5LywwtakiQp30pZV0D5YM+sJEmSJCl3TGYlSZIkSbnj3YzVzByi0r/yIdm2Vf9sq9rZVrWzrWpnW9XOtqqdbVU720oNyZ5ZSZIkSVLumMxKkiRJknLHZFaSJEmSlDsms5IkSZKk3DGZlSRJkiTljnczVrO5LOsK5IhtVTvbqna2Ve1sq9rZVrWzrWpnW9XOtlJDKiVJsutSkiRJkiQ1EIcZS5IkSZJyx2RWkiRJkpQ7JrNqNlOA3wFPAH+bcV0azXeAdcCjZeteC9wF/D59Hp1BvRrNG4AlwEpgBfDZdL1t1ddrgF8DjxDaqntOlW1VXSvwG+D29LVtVdlqYDnw78DSdJ1tVdl+wHzgMcLvrROwrSo5nHA9dT9eAj6HbVXN5wm/1x8Ffkj4fW9bqeGYzKqZtALfBM4AjgTOTZ8V3EBI9sv9LfAL4C/SZ78AgFeBLwJHAMcDFxKuI9uqry3AKcAxwLGE6+t4bKv+fJaQcHSzraqbRLiu3pG+tq0qmw0sBN5KeC+uxLaq5HeE6+lY4DhgE3AbtlUlY4GLCO+9txE+X52DbaUGZDKrZvIuQo/sH4BXgLnA+zOtUWO5B3ih17r3A3PS5TnAB+pao8b0HNCVLm8gfDAci21VSQL8OV0ekT4SbKtqxgHvBf61bBKH/IcAAALOSURBVJ1tVTvbqq99gHcD16evXwFexLbalVOBVcBT2FbVtAF7ps97Ac9iW6kBmcyqmYwF1pS9fiZdp+oOIiRvpM8HZliXRjQeeDvwK2yraloJQ/bWEYad2VbVXQNcAmwvW2dbVZYAdwIPA9PTdbZVX28Cnge+Sxi+/q/AKGyrXTmHMHQWbKtK1gJfA54mtMl6wvvRtlLDMZlVMylVWOf/ntJg7Q3cQphT9VLGdWlk2wjD9sYRRke8LdvqNKz3ERL+h7OuSE6cBEwkTBu5kND7qL7aCO10HeGLt4049HNX9gDOBH6UdUUa2GhCL+yhwOsJX5Ccn2mNpCpMZtVMniHcvKfbOMKwGFX3R2BMujyG8GFbYbjsLcAPgFvTdbZV/14EfkmYN2tb9XUS4QP0asIUiFOA72NbVdP9u3sdYV7ju7CtKnkmffwqfT2fkNzaVtWdQZhK8sf0tW3V12nAk4Re/62Ev4MnYlupAZnMqpk8RLgpwaGEb17PAX6aaY0a30+BqenyVOAnGdalUZQI889WAleVrbet+jqAcCdVCHOrTiPcUdW26uvvCF+wjSf8blpM6OmwrfoaBbSXLb+HcEdV26qv/yBMrzk8fX0q8Ftsq/6cy84hxmBbVfI04WZ+exH+Jp5K+JtoW6nhlJLEUZhqKh2EeWmthH9F89Vsq9NQfgicDLyO8O3qPwA/Bm4G3kj44/Vh+t4kqmj+EriX8G9Buuc2Xkro+bCtejqacBOQVsKXozcD/wjsj23Vn5OBiwlDj22rvt5E6I2FMIz2JsLvctuqsmMJc2X3INwA8WPsfD/aVj3tRUj+30SYBwpeV9VcBpxNuMP/b4ALCNNvbCs1FJNZSZIkSVLuOMxYkiRJkpQ7JrOSJEmSpNwxmZUkSZIk5Y7JrCRJkiQpd0xmJUmSJEm5YzIrSZIkScodk1lJkiRJUu6YzEqSJEmScuf/A59dwH4zSBv/AAAAAElFTkSuQmCC\n",
      "text/plain": [
       "<Figure size 1152x1152 with 2 Axes>"
      ]
     },
     "metadata": {
      "needs_background": "dark"
     },
     "output_type": "display_data"
    },
    {
     "data": {
      "image/png": "iVBORw0KGgoAAAANSUhEUgAABXEAAAHuCAYAAADZbk+wAAAABHNCSVQICAgIfAhkiAAAAAlwSFlzAAALEgAACxIB0t1+/AAAADh0RVh0U29mdHdhcmUAbWF0cGxvdGxpYiB2ZXJzaW9uMy4xLjAsIGh0dHA6Ly9tYXRwbG90bGliLm9yZy+17YcXAAAgAElEQVR4nOz9e5ydZ33f/X4E5uBgSCYOpyAcp7EChjSmO83UgTwxJcAjIOnTxMrGBJ5AGkJiLMAlDOweXESd7MRMAQMyanYwGwildhDwUCgRkAMiZWMWYB5DG5uOkxghIBjMlIPBONiz/7iXnLE8I+swo7Usvd+v13rpnrWu+75+a0k/hL6+1nVvWFpaCgAAAACA6XS3SRcAAAAAAMDqhLgAAAAAAFNMiAsAAAAAMMWEuAAAAAAAU0yICwAAAAAwxYS4AAAAAABT7IRJFwAAAAAcP/Zs3v3B6oOn7Dpr24RLOWJ7Nu8+tfqb6odP2XXWdePn3lh1yq6znjWhsjgGLIxmPlh9cNPs4rYJl7ImFkYzpzbulU2zi9eNn3tj1abZxWdNqq67EiEuAAAAcJezZ/PuM6v3VptP2XXW6CDPmam2Vv9Htam6R3Vt9a7qVafsOuur61QuTMzCaOa2Xtk0u3hQvbIwmjlgr2yaXdQrR5ntFAAAAIC7oq9Vn6y+fjCD92ze/U+r/9kQSl1S/Ux1ZvX71a9U/2PP5t2PWJ9SYaIOqVcWRjN32isLoxm9cpRZiQsAAADc5Zyy66yrq589mLF7Nu/+R9V7GkKo3zpl11lLy17+1J7Nu/+wuqD62zUvFCZs0+ziQffKwmjmdr2yaXbxdr2yMJrRKxMixAUAAACOdf+x+nB3DHCrOmXXWV+v5o56VTB9buuV/QLcqjbNLuqVCRHiAgAAAEfdns27f7n6N9UPVZ+pXnLKrrP+ZPza/RtW+z2l2lhdX11a/ftTdp1163jMqd3xpmJL1VkNX/n+xerjDYHTbPXYlQLcA9T3PdW/q365ekDD18vnT9l11h8e4vs8sZof1/N91ceq55+y66yrDuU6HJ8WRjN36JNNs4t/suz1VXtl0+zireMxp3bHm4odsFdWCnAPUOOKvbJpdvGQemVhNLNir2yaXdQr2RMXAAAAOPrOrn6zek71uIaA6V17Nu9+0Pj1X6xOrs6tzqieW/2L6gUHce1XVXevHjs+56eqm6r/drDF7dm8e0PDjaB+ofqt6h9Xr6xesWfz7t842OuM/YdxLU9bVsv2Q7wGx6cV+2RhNPOgZWMm2isLo5lVe2VhNKNX1pCVuAAAAMDRdnL1U6fsOuubVXs27/7Vhj02N1dvPGXXWb/fsCfnPtfs2bz7tIbA6lV3cu2vnLLrrF/d98Oezbu/v/rqKbvOuuUQ6nt29aPVPzxl11k3jJ/773s27/589dY9m3e/9ZRdZ33jIK/109VbT9l11u5xPWcfQh0cYxZGM/+yen51/+qj1R9U/7X6gerihm0Mrh0PP7n6qU2zi98cn3u7PqnaNLt4h15ZGM0cdK9sml28rVcWRjPfX3110+ziYfXKptnF23plYTTz+eqtC6OZt26aXTykXtk0u7h7XI9eWUaICwAAABxt79kX4Fadsuusb+zZvHuh4evg+1bC/nLD6r6fbPhq9YnV5w/i2v95v5+/Wn3/ns27734IQe4zqj9cFuDuq/MD4+0RHlu9+yCv9f+rnrdn8+6vV7tO2XXWtXd2AsemhdHMwxu2L3hqw7YHT61eUb21WqpeV/31slPesy/Ardo0u/iNhdHMbX0yvuaa98rCaObuhxDkPqP6w2UB7r5aPzDeHuGxHWKvLIxmvl7tWhZmkxAXAAAAOPpWurP9/+rvc4pXVL9W/e74eLEhpDqYr2d/Zb+fP1Ldu3pM9aGDrO/B1Qv3bN690lfS71U95CCvU/XChvf2wuq1ezbv/quGG6y96xCuwTFg0+ziNQujmZ/Yt1dtddHCaOY/VD9c3bBpdnFxv1PurE9qSnplYTSz5r2yMJr5q4aVyXolIS4AAABw9N262gt7Nu++d/W86pxTdp319mXPr3rOgZyy66yr9mzePar+dQcfTH2x+kD12lVeXylcW23+b1f/qvpXezbv3lS9uLp8z+bd/+CUXWd94WCvw7FhWYC77+dbqtVWnB7wz/zCaOa2Xtk0u/j2Zc8fVq9sml28amE0M7Fe2TS7eFuvLIxmbuuVhdHMP9g0u3jc94obmwEAAADT5MSGRWf7r0r8+SO45rnV/7Zn8+758VYNt7Nn8+4NezbvfsOezbufO37qP1X/e/X5U3addc2+R/VX1Ymn7Drrfx3sxHs27z59z+bd96w6ZddZC9WLGlYonn4E7wdqHXtlYTQzP96q4XYWRjMbFkYzb1gYzdyhVzbNLl6z79G4VzbNLh50ryyMZk5fGM3cs2rT7KJe2Y+VuAAAAMDUOGXXWYt7Nu/+ZHXReDuDm6vfrO7ZYeYYp+w668o9m3c/pfqj6rF7Nu++pLqy+k71yOr8hpszbR+f8gcN+5V+aM/m3S+rrmn4yvu/qu6xZ/Pux5yy66w7Xe04XlX8/uqqPZt3b6v+ruGmVl+tPnk47wX22TS7uLgwmvlkw7YMa9Irm2YXr1wYzdzWKwujmYPulYXRzB16ZWE085j9Vx+vZLyq+P3VVQujmW3plTuwEhcAAACYNr/UEN58oOGmSF9u+Gr19x/uBU/ZddYHG8Kndzd8Bf2/VZ+oXtbw1fFHnrLrrCvHY5eqzdU7qourq6pLqg9XTziYAHd8nZuqJzTctOpPq7+ofqj630/ZddZXD/e9wDJr3iubZhc/2J30yqbZxSvHYw/YKwcT4I6vs2qvbJpd1CvVhqWlpUnXAAAAAADAKqzEBQAAAACYYkJcAAAAAIApJsQFAAAAAJhiQlwAAAAAgCkmxAUAAAAAmGJCXAAAAACAKSbEBQAAAACYYkJcAAAAAIApJsQFAAAAAJhiQlwAAAAAgCl2wqQLAAAAAI6KbascH2+2rXJ8vNm2yvHxZNsqx8ebbascH2+2rXI8FTYsLS1NugYAAABg/S0PADZMrIrJ8zkMfA4+g318DoOp/hxspwAAAAAAMMWEuAAAAAAAU0yICwAAAAAwxYS4AAAAAABTTIgLAAAAADDFhLgAAAAAAFNsw9LS0qRrAAAAgOPCDa+4ZunGD3xp0mXAVPjOv//nky4BpsqV//m1PfVVz9iw0mtW4gIAAMBRcsID7z3pEgCYUt//0JNXfU2ICwAAAAAwxYS4AAAAAABT7IRJFwAAAADHi+99xqmNXvviSZcxMU8cva2ql/3QeROuZLJe+tlLqvrn99064Uom5//6xvaWdi317Pu/YNKlTNSlX3lNVec/+PkTrmSyLv7i8DnMPeR5E65kcuY//9re/4rVX7cSFwAAAABgiglxAQAAAACmmBAXAAAAAGCKCXEBAAAAAKaYEBcAAAAAYIoJcQEAAAAAppgQFwAAAABgiglxAQAAAACmmBAXAAAAAGCKCXEBAAAAAKaYEBcAAAAAYIoJcQEAAAAAppgQFwAAAABgiglxAQAAAACmmBAXAAAAAGCKCXEBAAAAAKaYEBcAAAAAYIoJcQEAAAAAppgQFwAAAABgiglxAQAAAACmmBAXAAAAAGCKCXEBAAAAAKaYEBcAAAAAYIoJcQEAAAAAppgQFwAAAABgiglxAQAAAACmmBAXAACAY90J1YY7GXOPo1EIABwOIS4AAADHolOr91SXVVdX7x0/f1J1ZfWQ6lnVxdWfVa+uHjge/77qE9Wzl13vJdUHqg9V76jutcq826qlAzwA4JCdMOkCAAAAYJ08rjqz+lT15uo51RnVRdXnx2OeUD26+lrDatwXV3sa/r386fF5N1c7x+dVXVg9dfwaAKw7IS4AAADHqisbAtyq51YfH/98+bIx72wIcKturbZUT6zuXj24un9D4Pvg6veq76s2Vm9Z59oB4Da2UwAAAOBYddOy40dW11ePqGZWGfOihrD2KQ0rdK9p2Et3Y7W9esH4+Ve3+h6728avrfYAgEMmxAUAAOBYd9/qddXTq5dXl64y7kHVVdUt1cMatl6oOrn6cvWFhm+0/sJ6FgsA+7OdAgAAAMe6HdWrqs817GP7pIbtFb6137jXVm+tnll9tvqT6j4NWzBcW32s+mL1kfHzAHBUCHEBAAA4Fl1XPX58/Iz9XnvaKuf8dcON0FZy7hrUBACHxXYKAAAAAABTTIgLAAAAADDFhLgAAAAAAFNMiAsAAAAAMMWEuAAAAAAAU0yICwAAAAAwxYS4AAAAAABTTIgLAAAAADDFhLgAAAAAAFNMiAsAAAAAMMWEuAAAAAAAU0yICwAAAAAwxYS4AAAAAABTTIgLAAAAADDFhLgAAAAAAFNMiAsAAAAAMMWEuAAAAAAAU0yICwAAAAAwxYS4AAAAAABTTIgLAAAAADDFhLgAAAAAAFNMiAsAAAAAMMWEuAAAAAAAU0yICwAAAAAwxYS4AAAAAABTTIgLAAAAADDFNiwtLU26BgAAADheLG3YsGHSNcBUWFpaSj/A33v/K97bE174pBWbwkpcAAAAAIAJ+/DXP7bqayccxToAAADguHe3DfebdAkTc+vS18dH95xoHZN3c1UbNpw44TomZ2np2+Oj4z2a+u7417tPtIrJu2X86/H8OdxywFetxAUAAAAAmGJCXAAAAACAKSbEBQAAAACYYkJcAAAAAIApJsQFAAAAAJhiQlwAAAAAgCkmxAUAAAAAmGJCXAAAAACAKSbEBQAAAACYYkJcAAAAAIApJsQFAAAAAJhiQlwAAAAAgCkmxAUAAAAAmGJCXAAAAACAKSbEBQAAAACYYkJcAAAAAIApJsQFAAAAAJhiQlwAAAAAgCkmxAUAAAAAmGJCXAAAAACAKSbEBQAAAACYYkJcAAAAAIApJsQFAAAAAJhiQlwAAAAAgCkmxAUAAAAAmGJCXAAAAACAKSbEBQAAgAN7/qQLAOD4JsQFAACAAxPiAjBRQlwAAAAAgCkmxAUAAOBYdkK14U7G3ONoFAIAh0uICwAAwLHm1Oo91WXV1dV7x8+fVF1ZPaR6VnVx9WfVq6sHjse/r/pE9ez9rvmS6iPVX4zPO3GVubdVSwd4AMAhO2HSBQAAAMA6eFx1ZvWp6s3Vc6ozqouqz4/HPKF6dPW1htW4L672NPxb+dPj826ufri6tfqp8XkvqX6neuFReB8AYCUuAAAAx6QrGwLcquc2BK73ry5fNuadDQFuDSHtlmpX9cfVg8fjq26o5pedN189eV2qBoAVCHEBAAA4Ft207PiR1fXVI6qZVca8qCG4fUrDCt1ruvO9dFeybXzeag8AOGRCXAAAAI5l961eVz29enl16SrjHlRdVd1SPaxh64V9Tm4Iefd5UX+/zy4ArDt74gIAAHAs21G9qvpcwx63T2rYXuFb+417bfXW6pnVZ6s/qe4zfu1vGv79/OGGbReuqubWu3AA2EeICwAAwLHmuurx4+Nn7Pfa01Y5568bboS2ktPGv/7ekZUFAIfHdgoAAAAAAFNMiAsAAAAAMMWEuAAAAAAAU0yICwAAAAAwxYS4AAAAAABTTIgLAAAAADDFhLgAAAAAAFNMiAsAAAAAMMWEuAAAAAAAU0yICwAAAAAwxYS4AAAAAABTTIgLAAAAADDFhLgAAAAAAFNMiAsAAAAAMMWEuKyrubm58+fm5h416TrgrmBhNHP+wmhGvwAAAAC3I8RlvZ1fCaXg4OgXAAAA4A6EuAAAAAAAU0yICwAAAAAwxYS4AAAAAABTTIgLAAAAADDFTph0AUze3NzcG6tvVldU/676keqhDSH/y6qnVDPVtdUbq1fNz89/d79r/HLDTZn+4fha76n+zQpzPa36f1U/Wn1hfK3t6/C2YM0tjGbe2Aq9sml28QsLo5mNrdIvm2YXv7vfdQ6qXxZGM3fol02zi/oFAAAAjjNW4rLPYxpCqfOqH6u+W32s+gfVU6sfr15ezVVvn5ubu+3Pztzc3L+pXl+9rfp/NIRYt1b/rTpp2bh/Ur2p2lH9o2p79Yq5ubnT1/m9wVq6Xa+MA9wHdIB+WRjN3NYvC6OZg+qXhdHMiv2yMJrRLwAAAHCcsRKXfR5RPWx+fv66qrm5ubdU11U/Oz8/f+t4zGfm5uY+Wn2qIaj6z+MA9qXVL87Pz79n2fVGc3NzL622LXvu0dV18/Pz/3H88zVzc3OXzc/Pf3Gd3hMclIXRzL+snl/dv/po9QfVf61+oLq4+q1Ns4vXjoc/onrYptnF65Zd4pWN+2XT7OJt/bIwmrldv4wD2JdWv7hpdvF2/bIwmlmxXzbNLt7WLwujmcs2zS7qFwAAADjOWInLPh9ZFuDeq/ql6uJlAW5V8/Pz1zR89fvs8VPnVH+1X4C7z+9W31n28xXVj8zNzV0yNzf303Nzc/cU4DJpC6OZh1e/0hC0/lj1/uoV1dcatkTYU/31slM+sjzAXRjN3NYvywLcqjbNLq7YL/sFuPus2C8Lo5lLFkYzP70wmrmnABcAAACOT0Jc9vnKsuOTq3tWf7PK2L+qfnB8/ODVxs3Pz9/csDpx388fqX6xOqP68+qrc3Nzr56bm7vHEVUOR2ActP7EptnF0abZxes2zS5eVJ3SsA/tD2yaXXzefuHsV/a7xJr0y6bZxdv1y6bZxTv0y8Jo5tULoxn9AgAAAMcZIS4ruaG6uTp1ldd/pNq3IvCLq40bh7M/tPy5+fn5d83Pz/90w42fXlA9q3rekRYMR2KFFbS3bJpdvHbT7OLiQZy+Jv0yDmdv1y+bZhfftWl2Ub8AAADAcc6euNzB/Pz8d+bm5t5evWBubu5t8/PzS/tem5ub+9GGGzH96vipP6r+9dzc3JPn5+ffu9+lXlLde9m531P94Pz8/LXz8/PfrC6dm5v7uYabNsFd0qbZxe8sjGbeXr1gYTTztk2zi7f1y8JoZsV+WRjNPHnT7OIB+2VhNPM91Q+Ow+RvVpcujGb0CwDAMeDWpa9PuoQpcPOkC5gKS0vfnnQJU+C7ky5gStwy6QKmxPH9Ofzj+/zkqq8JcVnNC6tPVu+bm5t7RfWF6h837Nv5vuqyqvn5+f8xNzf329Xbxjcy+9PqftUvVz9bXb/smi+rnjk3N/ec6urqzOoJ1XlH5R3B+rmtXxZGM6v2y6bZxf+xMJr57ept4xuZ3Wm/LIxm9AsAAAAc54S4rGh+fv5v5+bmfqy6qHpzw9e5/7r67eqS5atz5+fnXzY3N/ep6t+OX/9G9e7qf6v+Ytll/231d9VrqgdUn63+9fz8/JvW/x3B+tk0u/i3C6OZVftl+ercTbOLL1sYzRx2v2yaXdQvAAB3cY/6nmdMuoSJ+b+/9ZaqHnafX5xwJZP1mRvfUdXZ9zt+d0t7+9dfW9VHz3ryhCuZrH+ye/iS5osfcvz+Wah6+eeHPw+/+YDnT7iSyfmP17/mgK9vWFpaOuAAAAAAYM0s/aP7/J+TrmFihLgDIe4Q4i4tLTV67FMmXcpECXEHQtwhxH33y9/bz809acNKr7uxGQAAAADAFBPiAgAAAABMMSEuAAAAAMAUE+ICAAAAAEwxIS4AAAAAwBQ7YdIFTLltqxzflWxb5fiuZNsqx3c121Y5vivZtsrxXcm2VY7vSratcnxXsm2VYwAAAGA/QtwDe+my422TKuIIeQ/T41h4H97DdPAeAAAA4DhiOwUAAAAAgCkmxAUAAAAAmGJCXAAAAACAKSbEBQAAAACYYkJcAAAAAIApJsQFAAAAAJhiJ0y6gCNx+eWXL3384x8/WtMtHa2J1pH3MD3W5X0855devx6XXc2x8HvhPUyHdXsPm2YXf7V643pdHwAAAI6Gu/RK3JmZmUmXAMB0O3XSBQAAAMCRukuHuAAAAAAAxzohLgAAAADAFLtL74n7xCc+sQ//+rsnXcZh+/d7tlf1wPs8ZsKVHJkv3fjhqi7/8V+ZcCWH76mfenNVP3nisyZbyBH42Lff2Pz8Ul/+9dMnXcphe8Drr6nqa1990YQrOTLf+/3/oarv3vqWCVdy+E642zOq+spv/OiEKzl8P/D7/7OlY2HHYAAAAI57VuICAAAAAEwxIS4AAAAAwBQT4gIAAAAATDEhLgAAAADAFBPiAgAAAABMMSEuAAAAAMAUE+ICAAAAAEwxIS4AAADUo6qfmXQRALASIS4AAAAIcQGYYkJcAAAAAIApJsQFAADgWHZCteFOxtxjDefbVi0d4AEAh+yESRcAAAAAa+zUanv1zeonqmurJ1UnVR+qfr56QsMWCj9eXVONqgdUO6sHV7dWr6zeOb7mS6rHV/eqvlI9rfrO0XgzACDEBQAA4Fj0uOrM6lPVm6vnVGdUF1WfH495QvXo6mvVs6pzGvbFvaa6X/XH1XXVJxvC3YvG511YPXV8XQBYd7ZTAAAA4Fh0ZUOAW/Xc6oXV/avLl415Z0OAu8+bGgLcqq9Xr6nOHv/84Opt1QeqLdVDV5l3W8P2Das9AOCQWYkLAADAseimZcePrK6vHlHNVIsrjFnJhoZ9bDc2bM/w5OoL1W9WP7CWxQLAgViJCwAAwLHsvtXrqqdXL68uPcDYZ1YPW3be8xpW655cfbkhwD2h+oX1KhYAVmIlLgAAAMeyHdWrqs817GH7pIbtFb61wtjLqt9u2DqhhhubXdmwIvfa6mPVF6uPVPdZ16oBYBkhLgAAAMea66rHj4+fsd9rT1vlnDeOHytZqs490qIA4HAd7nYKJ3TnG7Lf4zCvDQAAAADA2KGEuKdW72n4esnV1XvHz5/U8PWSh1TPqi6u/qx6dfXA8fj3VZ+onr3sei9puKvnh6p3VPdaYc5tDf/Fc7UHAAAAAMAx7VC3U3hcdWb1qYa9hJ5TnVFdVH1+POYJ1aOrrzWsxn1xtWc816fH591c7RyfV3Vh9dTxawAAAAAAjB1qiHtlQ4Bbw0bwHx//fPmyMe9sCHCrbq22VE+s7t6wOfz9GwLfB1e/V31ftbF6y6GXDwAAAABwbDvUEPemZcePrK6vHlHNVIsrjHlR9QPVU6pbqisa9tLdWG2vnlx9ofrN8bj9bRs/VmNLBQCYIld0w49Vv1M9puHv/Cur3zmzkz+4VnNc0rUrznFep606xyVd+9jqz1d5+efP67T3rFV9Ved0xa9XW6sfrb5S/Wm17bLOvG4t5zkYZ/T+32vYxup5V/XE7Udjznv29sd2gM/75s5e0897JRuuvvgRDd/2+umGxQQfqf7d0unnf3K9515u7969K/553bhx4wfXao7RaLTiHLOzs6vOMRqNHtsBfo9mZ2fX9Pdo586dK/bEli1brlvLeQ7Gjh07buuJc88996j0xIUXXvjYDvB5X3DBBeveE1u3bl2xJ7Zv335UewIAODyHe2Oz+1avq55evby6dJVxD6quaghwH9aw9ULVydWXGwLcE6pfOMw6AIApcUU3PLz6i/GP/6zhruBXVh+4ohsetxZzXNK1q85xSdcezByPrH54v8efrkVt+5zTFS+p5qsd1T9q2DJqpvrTc7rie9dyrjtzRu/f1HDPgq8ezXmXWffPeyUbrr74odXu8Y//vNpc3VDt3nD1xQ9f7/n32bt376p/Xvfu3bsmPTEajVadYzQaTUVP7Ny5c9We2Llz51HtiR07dhyXPbF169ZVe2Lr1q1HrScAgMN3qCtx99lRvar6XMM+tk9q2F7hW/uNe2311uqZ1WerP6nu07AFw7XVx6ovNvxX4PscZi0AwHR4UbW3OvvMTv7u+LlPXtENP1Rd0HDj0zWb47xOu22OS7r2YOfYc16nfXMN6jiQ11bvuKwzF/Y9cU5X/FLDt5Z+rvpP6zz/cq+ufrf6l0dxzuX23NzZ6/15r+T/bPh22FOXTj//u1Ubrr74V6ufqH69+q2jVMdtf143btx425/XvXv3rktPzM7O3jbHaDQ66J6YnZ09Kj2xZcuW23pi586dx21PXHDBBRPtie3bt3+3auvWrZPoCQDgMB1KiHtdw3/Zr3rGfq89bZVz/rrhRmgrOfcQ5gYApt8LqnsvC3D3+b+74/93OKI5lgW46zHHEbmsM79VLez39N0avr581MKbM3r/P2tY5XdJkwusJuXm6sZ9AW7V0unnL224+uKvjV87Wl5Q3XtZgLvPmvfEsgB3PeY4Ilu2bJmKntixY8dx3xP7Atyq7du3L23duvVo9wQAcJgOdyUuAMDtnNnJN1Y3rvDSo6v/sRZznNdpRzrHRy/p2h9pWAF4efVv13Nl7jldsaEhNPp/N3zF/b+u11zLndH7793wranzruqJ3z2j9x+NaVfy0Xv29tt93kdpZe4fVnMbrr74txu2/rqlYQHBIxpWJB4VGzduXPeemJ2dPeKeGI1Gt/s9Ws+VuTt37pxIT+zYseO2njj33HO/u2PHjqMx7Uo+euGFF97u8z5KK3P/sJrbunXrRHsCADh8h7snLgDAnbqiG36x4evSl6zXHJd07cHMcVPDdlDPr36yemlDcPHO9arrnK74+ervqr+qHlg98bLO3H+15HqZq665qifuOkrz7e+of97LLZ1+/peqJ1a/UX2t+sa4hp9bOv38vzkaNaxm7969694To9FoKnti586dd+iJLVu2HNWeOPfcc4/Lnti+ffuqPbF9+/aJ9gQAcHCsxAUA1sUV3fDY6i3VK8/s5A+txxyXdO1tc5zXaavOcV6nXVFdseypT1/StZ+oRpd07dnnddrb16G8P28Ia86oXlz9X+d0xZMv68x1/eryGb3/lIb9Lf/Jes5zIDd39h0+73v29k9Uo3v29rNv7uz1+Lxvs+Hqi3+4em/D/Rj+oGHV4a9V795w9cWPWTr9/GvWc/7V7N2797GN/7xu3LhxXXpiNBrdNsfs7Oyqc8zOzt7h92g0Gn1iuMTo7NnZ2aPSEzt37nzyli1b1rUnduzYMfGeuOCCC+7weV944YWfqEYXXnjh2RdccMG69sTWrVtX7YmtW7c+Zvv27RPpCQDg4AlxAYA1d0U3PKZ6d/X2hrBmzV3StUc0x3md9olLunZ39Yvja6ypyzrzm9Unq0+e0xXvbbix669Ur1/rufbzyuoNV/XEz6zzPIfk5mhwXB4AACAASURBVM7+xD17+7p93vv53eq/L51+/vL7NuzecPXF76r+Q8Mq1aNq7969694To9HoiOaYnZ39xGg0Wrffoy1bttzWEzt37jzqPXHuuedOVU9ccMEFn7jwwguPak9s3779dj2xdevWifUEAHBohLgAwJq6oht+smHF17uqZ53Zybeu9RyXdO3t5jiv0w53js9UP7Jmha3iss68/pyueE/1hNYxsDqj9/909djq4Wf0/v3/f97dxs/dclVPXFqvGu7EUfm8q59t5RDzP1eXHoX5b2fv3r23+/O6cePGNe+J0Wh0uzlmZ2enuie2bNly/c6dO9e9J3bs2HFbT+zYseMOPTF+7pZzzz1XTwAAU82euADAmrmiGx5Vva9hNeAzz+zkW9Z6jku69nZznNdpRzLHD1VfWJPCxs7pinuc0xV3X+GlE6qT1nKuFZxWnVx9uWHv0X2PH6pePT7+2XWu4UDW/PNexX2qlUK5W6sTN1x98VFbyLB3797b/XnduHHjmvfEaDS63Ryzs7NT1RM7d+68x86dO/XEyqaiJ7Zu3WpxDwBMOX9ZAwBr4opu+LHqAw2rAdcrwL3dHAcT4F7StSc2fFV553md9oZlzz+8enz1S2tV3zld8eDq09X5DfuS7nv+vg0rDnes1VyreHfDnqP7+y/jev6oYeXfurlnb7/t8765s9+w7Pk1/7wP4CMNX1F/437Pb6k+unT6+UflZlp79+693Z/XdQpwbzfHwQS4o9Hott+j2dnZNyx7fs1/j3bu3LliT+zcufO46YkLL7zwts/7ggsueMOy56emJ7Zv3360bjAHABwmIS4AcMSu6Ibva7hhzqcb7nj+0Cu64XZjzuzk645kjku69g5zXNK1txtzXqddd0nX/lx1efXU8zrtPed12rcv6dovVdsv6dp7V7sbbqw0X/1xQ5izJi7rzC+e0xX/pdp+TlfcqyE4+cHqZdU3G/bmXDdX9cQbar8Pvjqj999c7bmqJ358Peevurmzv33P3v6lavs9e/u6ft4H8LzqYxuuvvgPGr4qfkv1q9U/q37qKMzf3r177/Dnde/evbcbs3HjxuuOZI7RaHSHOUaj0e3GzM7OXjcajW7ridnZ2ffMzs5+ezQafanaPhqN1vX3aMuWLV/cuXPnf6m279y586j3xLnnnrtiT+zYsePmas+555677j1xwQUXfPvCCy/8UrX9wgsvnGhPbN26dWI9AQAcGSEuALAWvq964Phx7SpjNhylOZYaAorlXx1+dvU/q39Zvar6XMMKwIvO67S13gvz2dU11W9Vp1ZfqnZVF1zWmd9a47mm1aqf982dve57jy6dfv5fbrj64n9cXdSwSnVD9aFqdun08z+13vOP3WV7YnZ29qj1xJYtW477nrjgggvWvSe2b9/+l1u3bl2xJ7Zv3360egIAOAJCXADgiI1X2R5pIHVA53XaQc1xXqf91+p++z13S8Pd2X93XYpb5rLOvLV6+fgxFa7qiacezflu7uyj9nmvZun0869uWGU4EeNVtuvaE7Ozswc1x+zs7B16YrztwlH5PdqyZcvU9cS555576tGc74ILLph4T2zfvn2iPQEAHBkhLgCw1ratcqwGNUyihknPrwY1TFsNk54fADgMQlwAYK29dNnxNjWoYcI1THp+Nahh2mqY9PwAwGG426QLAAAAAABgdUJcAAAAAIApJsQFAAAAAJhiQlwAAAAAgCkmxAUAAAAAmGIblpaWJl3DkVjasGHDpGuAqbG0tJSegL+3tLT0stx5exKW/5+LSf2PkhrUMC3zq0EN01bDpOfHv2PhNv4NC7f37pe/t5+be9KKTXHC0S4GAJh+f9U3l77cd9biUof9X4s/0eJazH9ENfxFX5loDVf39bWa/whq+MbEa/i7qz830fmrPnffLROv4Qtf+MJEa9izZ89azX/YNXz5y1+eeA1f+cqa/e/CYdfwpS99aaLzV23fvv1XqzeuVSEAQF1/89+t+tpdPsT9wq/8w0mXcNh+8M2fruqp3/u8CVdyZC7/2mur+o8P+/UJV3L4fvMzf1DVPz3xNyZcyeH782//flVfeMaPT7iSw/eDb/lUVc9/0PMnXMmRec3fvqaqZ37/Xfd9vOmrw3v40GP+jwlXcvh+5sPvmnQJd2n3suMSAAd26qQLuCu7/tkPn3QJE/OA119T1a6f/H9OuJLJ2vyxP6rqrT/2zAlXMjm//N/fVNUbTv+1CVcyWf/i6kur+jcP3TrhSibrdz63varHnPjsCVcyOR/+9usP+Lp/oQEAAAAATDEhLgAAAADAFLvLb6cAAKy9jX1P9/utyX3V83tf+bmWlpa634mnT6yGb9x0TUtLS92y9J8mMv8Jd3tGS0tLPfCkR09k/qrrb/xIS0tL3XjTKydWw0kn/lZLS0udet/NE6vhs998XxuXzu/hJ509sRo+c+M7Wlpa6iEn/dOJzP+FGz/Y0tJS1//aIycyf9UD3/CXLS0t9aSTzptYDbtufF1LS0v9yskvmFgNf/jV17S0tNQjTvqlidVw9Y072759+8TmB4DjkZW4AAAAAABTTIgLAAAAADDFhLgAAAAAAFNMiAsAAAAAMMWEuAAAAAAAU0yICwAAAAAwxYS4AAAAAABTTIgLAAAAADDFhLgAAAAAAFNMiAsAAAAAMMWEuAAwXU6oNtzJmHscjUIAAACYDkJcAJi8U6v3VJdVV1fvHT9/UnVl9ZDqWdXF1Z9Vr64eOB7/vuoT1bOXXe8l1QeqD1XvqO61wpzbqqUDPAAAAJgSJ0y6AACgqsdVZ1afqt5cPac6o7qo+vx4zBOqR1dfa1iN++JqT8Pf558en3dztXN8XtWF1VPHrwEAAHAXJMQFgOlwZUOAW/Xc6uPjny9fNuadDQFu1a3VluqJ1d2rB1f3bwh8H1z9XvV91cbqLetcOwAAAOvIdgoAMB1uWnb8yOr66hHVzCpjXtQQ1j6lYYXuNQ176W6stlcvGD//6lbeY3fb+PnVHgAAAEyJ9QxxH1X9zDpeHwCORfetXlc9vXp5dekq4x5UXVXdUj2sYeuFqpOrL1dfaPjGzS+sZ7EAAACsv/XcTuFRDauBPrSOcwDAsWZH9arqcw372D6pYXuFb+037rXVW6tnVp+t/qS6T8MWDNdWH6u+WH1k/DwAAAB3UfbEBYDJu656/Pj4Gfu99rRVzvnrhhuhreTcNagJAACAKXE42ymc0J3vlXePw7guAAAAAAD7OdgQ99TqPdVl1dXVe8fPn9RwN+2HVM+qLq7+rOEmKlUPqHZWH67+otvvy/eS6gMN2y28o7rXCvNuq5YO8AAAAAAAOKYdynYKj2v42uanGvboe07DTVQuqj4/HvOE6tHV1xpC3XMabm52TXW/6o8bvjL6yYZw96LxeRdWTx1fFwAAAACAsUPZTuHKhgC3hhusvLC6f3X5sjHvbAhw93lTQ4Bb9fXqNdXZ458fXL2tYTXuluqhh1I4AAAAAMDx4FBC3JuWHT+yur56RDWzypiVbGjYBmFjtb16QcPq3Ve38j6728bPr/YAAAAAADimHc6Nze5bva56evXy6tIDjH1m9bBl5z2vYbXuydWXqy80bOnwCyueDQAAAOvvUQ1bAa7m+UerEABYyeGEuDuqV1Wfa9jD9jsN2yus5LLqt6v/1rAf7iv7+20Zrq0+1nBTs49U9zmMWgAAAOBICXEBmGoHe2Oz66rHj4+fsd9rT1vlnDeOHytZqs49yLkBAAAAAI5bh7MSFwAAAO4KTujO76dyj6NRCAAciYNdiQsAHGe+95Wfm3QJfeOmayZdQifcbf8vIR1d19/4kYnOX3XSib816RL67DffN+kS+syN75h0CX3hxg9OdP4HvuEvJzp/1a4bXzfpEvrDr75m0iV09Y07J13CNDu14Uba36x+omErvydVJ1Ufqn6+4Qbbj6p+vLqmGlUPqHZWD65ubdgO8J3LrvuS6p9X360+Uf2r6tsrzL+teumaviMAjntCXAAAAI41j6vObLgfy5ur51RnVBdVnx+PeUL16Opr1bOqcxr2xb2mul/DfV2uqz5Z/XBDsPtT43NfUv1O9cL1fiMAUEJcAGBVd5/g3LeoYeLzq+GONdxzgjXcPP7Vn0c1TFMNU23fDbVruBH3x8c/X75szDsbAtx93tQQ4FZ9vXpNdXZDiHtDNb9s7Hz1lwlxAThK7IkLAADAseamZcePrK6vHlHNrDJmJRsabsp9qLaNz13tAQCHTIgLAADAseq+1euqp1cvry49wNhnVg9bdt7z+vs9cU+uXrRs7Iuq965ppQBwALZTAAAA4Fi1o3pV9bmGvXGf1LC9wrdWGHtZ9dsNNzar4cZmV46P/6bh388fbtgb96pqbt2qBoD9CHEBAAA4llxXPX58/Iz9XnvaKue8cfxYzWnjX3/vcIsCgCNhOwUAAAAAgCkmxAUAAAAAmGJCXAAAAACAKSbEBQAAAACYYkJcAAAAAIApJsQFAAAAAJhiQlwAOLY8f9IFAAAAsLaEuABwbBHiAgAAHGOEuAAAAAAAU+yESRdwpH7wzZ+edAlH7PKvvXbSJayJ3/zMH0y6hCP259/+/UmXcMR+8C2fmnQJR+w1f/uaSZewJt701bv++/iZD79r0iUcb06obqmWDjDmHtXfrcFc26qXrsF1AAAAWGdW4gLAZJ1avae6rLq6eu/4+ZOqK6uHVM+qLq7+rHp19cDx+PdVn6ievd81X1J9pPqL8XknrmP9AAAArLO7/Ercc77veZMu4bBd9r+GFbhPvs9zJ1zJkXnvja+r6vWn/9qEKzl8z7760qo2nvSzE67k8O395p9W9U9O/NUJV3L4Pvrt/29Vjz3xOROu5Mh88Nv/n6rud+9HTLiSw/f1m/6yqgfe5zETruTwfenGD0+6hEPxuOrM6lPVm6vnVGdUF1WfH495QvXo6msNq3FfXO1p+Lv80+Pzbq5+uLq1+qnxeS+pfqd64VF4HwAAAKwDK3EBYPKubAhwq57bELjev7p82Zh3NgS4NYS0W6pd1R9XDx6Pr7qhml923nz15BXm3FZtOMADAACAKXGXX4kLAMeAm5YdP7K6vnpENVMtrjDmRdUPVE9p2EP3igSvAAAAxywrcQFgety3el319Orl1aWrjHtQdVVDgPuwhq0X9jm5IeTd50X9/T67AAAA3AVZiQsA02NH9arqcw173D6pYXuFb+037rXVW6tnVp+t/qS6z/i1v2n4+/3DDdsuXFXNrXfhAAAArB8hLgBM1nXV48fHz9jvtaetcs5fN9wIbSWnjX/9vSMrCwAAgGlhOwUAAAAAgCkmxAUAAAAAmGJCXAAAAACAKSbEBQAAAACYYkJcAAAAAIApJsQFAAAAAJhiQlwAAAAAgCkmxAUAAAAAmGJCXAAAAACAKXbCpAsAAKbVLZMuIDVMw/ylhn1unnQBTf5zmPT8pYZ9pqEGDtcDXn/NpEuYuM0f+6NJlzAVfvm/v2nSJUzcv7j60kmXMBV+53PbJ13CVPjwt18/6RKmlpW4AAAAAAATduKG1dfbWokLAKzobhvuN7G5b136elUbutfEaljqO9XkPod9n8EJd3/AROav+u4t11e1YcN9JlbD0tKNVd3jhAdNrIa/++7fVnWfe/3oxGq48Tv/s6r73vvhE5n/GzcNqwanoSfvdY+HTqyG7/zd58ZH95xYDftWhN/9bt8/sQpuufWrE5v7WHHqSU+edAkTc90331tN9u+3abDv79h7nvCQCVcyOTd/9/NV/dMTf2PClUzWn3/796v6tR94/oQrmaxLv/Kaqn7+pPMmXMnkvPublxzwdStxAQAAAACm2HqFuI+qfuYArx/f/3kBAAAAAOAgCXEBAAAAAKaY7RQAAAAAAKbYoYa4J1Qb7mTMPQ6zFgAAAAAA9nMwIe6p1Xuqy6qrq/eOnz+purJ6SPWs6uLqz6pXj19/QLWz+nD1F9Uv7Hfdl1QfGb92cXXiCnNvq5YO8AAAAAAAOKadcJDjHledWX2qenP1nOqM6qLq8+MxT6geXX2tIdQ9p2Ff3Guq+1V/XF1XfbL64erW6qfG576k+p3qhUfwXgAAAAAAjjkHu53ClQ0BbtVzG8LW+1eXLxvzzoYAd583NQS4VV+vXlOdPf75hmp+2dj56skHXTUAAAAAwHHiYEPcm5YdP7K6vnpENbPKmJVs6NC3QNg2Pm+1BwAAAADAMe1Qb2x23+p11dOrl1eXHmDsM6uHLTvveQ2rdatOrl60bOyL+vu9dgEAAAAAGDvYPXH32VG9qvpcw964T2rYXuFbK4y9rPrt6sHjn1/ZsC1D1d+M5/5ww964V1Vzh1gLAByvHtWw3/yHJl0IAAAA6+9gQtzrqsePj5+x32tPW+WcN44fqzlt/OvvHcT8AMDtParamBAXAADguHCo2ykAAAAAAHAUCXEBYHqc0J3fuPMeazTXtoYbjq72AAAAYEoc6p64AMDaOrXaXn2z+onq2oY9509q2C7h56snNGyh8OPVNdWoekC1s2Hv+Vsb9p7fdwPRlzRshXSv6isN2x9952i8GQAAANaeEBcAJu9x1ZnVpxpuHPqc6ozqourz4zFPqB5dfa16VnVO9TMNoe79qj9u2Mf+kw3h7kXj8y6snjq+LgAAAHdBtlMAgMm7siHArXpu9cLq/tXly8a8syHA3edNDQFu1der11Rnj39+cPW26gPVluqhK8y5rWHrhtUeAAAATAkrcQFg8m5advzI6vrqEdVMtbjCmJVsaNjLdmPD9gxP7v/f3v3HS1Xfdx5/neEiIqAiauJKKiZmUyGLFpPUmD6yCQ+ogY2rJqKh2Ghrl6JsTB9opLqPR6RJ3forGi3FtA1pNLZVwuKuVdzVbkOwxrb8SCRBSZcEGvwREfkl3MvPO/vH9wx3GGbunZl7Z75n5r6ej8c87jlnzpnzPufeO985nznne+B1YA5w6kCGlSRJkiQ1l2fiSpKUHaOARcAs4G5gcS/zXgN8oGi5LxDO1h0DvEUo4HYAlzcqrCRJkiSpOTwTV5Kk7HgIuB/YQujDdhqhe4XOMvM+BvwxoesECDc2W0s4I3cjsAp4A3gRGNHQ1JIkSZKkhrKIK0lSXJuBKenw1SXPzaywzLfTRzl54Pr+hpIkSZIkZYfdKUiSJEmSJElShlnElSRJkiRJkqQMs4grSZIkSZIkSRlmEVeSJEmSJEmSMswiriRJkiRJkiRlmEVcSZIkSZIkScowi7iSJEmSJEmSlGEWcSVJkiRJkiQpwyziSpIkSZLaWQeQ9DHP0GYEkSSpXh2xA/TXYzv/NHaEflu+d1HsCAPi915ZHDtCv7265//GjtBv/9z1V7Ej9NuKrr+IHWFA7N73cuwI/fbm3hdiR1BE3fndsSOQZ3/sCNH3w6HDW6OuHyCf3xs7AgcP/TJ2BPbu/9fYEXhn34ao68/C/+T+g1tiRwAOxA7A4e7tsSNk3ThgIbAHuADYCEwDRgIrgUuAqcD5wERgA3ADcAVwM3AQ2ATsBJYBK4D5wBRgGLANmAll/ykWALc3YJskSYNYyxdxJUmSJEkqYzJwIbAOeASYDZwH3AW8ls4zFbgI2AVMAm4BLgZ2ABOA5wlFXICl6bIAXwWuSl9XkqSGa/ki7snDz4sdoW47u14C4EPDr4mcpH9Wdz0MQC45MXKS+hXOsuoYcnrkJPUrnKn13hGfjpykfj/f+xQA00fcEDlJ/xTOrl/1iWmRk9TvwyueAeC+c+ZETlK/eRu/ETtCy0uSEdHWXTjz88Tjx0fLUDibPkmGR1l/Pt8FwNCOd0dZPxSf/XpctAyFMx6HH3d2tARdBzYB2fhdxNoPhX2Qhb+FU06YFC3B9s61QLz3Beh5bzhh2PuiZejc/7No667RWkIBF8JZtqvT8ceL5nmCUMAFmAHcTSjgAqwHlhTNewZwJ3AyMBZ4tCGpJUkqwz5xJUmSJEntaF/R8ARgKzAeGF1hngTIV3itsYTuGb5IOHv3ASr3s7sgfa7SQ5KkmlnElSRJkiS1s1HAImAW4UzbSjfzWArcBJyUjp8LXJkOjwHeAl4nXNF6eaPCSpJUTst3pyBJkiRJUi8eAu4HthD6sJ1G6F6hs2S+1cC9wNOEG5vtIHS9kCd0w7ARWAW8AbwIxOt3SJI06FjElSRJkiS1m83AlHT46pLnZvay3DJ6bmQGoQuF7YRC7vUDFU6SpFrZnYIkSZIkSTAJuK1o/Kx02vo4cSRJ6uGZuJIkSZIkwQbgOmAN0JU+5gDdMUNJkgQWcSVJkiRJgtBH7tzYISRJKsfuFCRJkiRJkiQpwyziSpIkSZIkSVKGWcSVJKm1nA98vJfnb2xWEEmSJElSc1jElSSptVjElSRJkqRBxiKuJEmSJEmSJGWYRVxJkrKhA0j6mGdoM4JIkiRJkrLFIq4kSfGMA54CHgNeAZan00cCa4EzgWuBrwP/ADyQPn86sBR4AXgeuLzkdecDL6bPfR0YXmbdC4B8Lw9JkiRJUkZ0xA4gSdIgNxm4EFgHPALMBs4D7gJeS+eZClwE7CIUdT9H6Bd3A3Ai8AywGfghcDbQDXw0XXY+cAcwr9EbIkmSJElqDM/ElSQprrWEAi7ADYRi62nA40XzPEEo4BY8TCjgAuwGHgQ+m46/DdxTNO89wPSBjSxJkiRJaiaLuJIkxbWvaHgCsBUYD4yuME85CbV3gbAgXa7SQ5IkSZKUERZxJUnKhlHAImAWcDewuJd5rwE+ULTcFwhn6wKMAW4umvdmevralSRJkiS1IPvElSQpGx4C7ge2EPrGnUboXqGzzLyPAX8MnJGO30folgFgE6F9f4HQN+5LwJcallqSJEmS1HAWcSVJimczMCUdvrrkuZkVlvl2+qjknPTnnfWGkiRJkiRli90pSJIkSZIkSVKG1VPE7aDvG54MreN1y1lAuFFLpYckSZIkSZIktbVqu1MYBywE9gAXABsJffWNBFYClwBTgfOBicAGQj9+VxBuqHKQ0EffTmAZsAKYT7iEdBiwjXDZ6P5+b5EkSZIkSZIktZFa+sSdDFwIrCPccGU2cB5wF/BaOs9U4CJgFzAJuAW4GNgBTACeJxRxAZamywJ8FbgqfV1JkiRJkiRJUqqW7hTWEgq4EM6ynQecBjxeNM8ThAIuwAzgbkIBF2A9sKRo3jOA7wLPEc7YfU+ZdS4gdN1Q6SFJkiRJkiRJba2WM3H3FQ1PALYC44HR9BRqi+dJqNxv7VhC9wzTgdeBOcCpNWSRJEmSJEmSpEGhnhubjQIWAbMIZ9ourjDfUuAm4KR0/FzgynR4DPAWoYDbAVxeRw5JkiRJkiRJanu1nIlb8BBwP7CF0IftNEL3Cp0l860G7gWeJtzYbAeh64U8oVuGjcAq4A3gRWBEHVkkSZIkSZIkqa1VW8TdDExJh68ueW5mL8sto+dGZhC6UNhOKOReX+W6JUmSJEmSJGnQqudM3GpNAj4F/Pd0/Kx02voGrlOSJA2QfH5v7Ajs3vdy7Ajk811R13/w0C+jrj84EDsAXQc2xY6Qid9F/P0Q/29he+fa2BGivy8AdO7/WewI6ofNe5bHjhDdocNbY0fIhAOHXosdIbrvdf157AiZsHjbg7EjZMLf7fmz2BEyq5FF3A3AdcAaoCt9zAG6G7hOSZIkSZIkSWo5V906tuJzjSzidgJzG/j6kiSpgZIkXnf1hbOAO4acHi1D4QyhhGFR1p9nPwCnnDApyvqh54zHLPwezhjx8WgZ3ti7EoDjOs6MlqFwptaJx4+Psv7CWfG55MQo6wfozu8GsvF7SJLh0TIUzgIeftzZ0TLEPyO89f3NB6+JHSGa3/rJwwBMGv7bkZPEtbbrOwB8fPjsyEniWdn1FwDs/fKpkZPENeIr2wA41P1o5CRxdeRC762DeT8U9kEluSblkCRJkiRJkiTVwSKuJEmSJEmSJGWYRVxJkiRJkiRJyjCLuJIkSZIkSZKUYRZxJUmSJEmSJCnDLOJKkiRJkiRJUoZZxJUkSZIkSZKkDLOIK0mSJEmSJEkZZhFXkiRJkiRJkjLMIq4kSZIkSZIkZZhFXEmSJEmSJEnKMIu4kiRJkiRJkpRhFnElSZIkSZIkKcMs4kqSlB0dQNLHPEObEUSSJEmSlB0dsQP0186ul2JH6LfVXQ/HjjAguvO7Y0fot0OHt8aO0G8/3/tU7Aj9tnzvotgRBsSHVzwTO0K/zdv4jdgRBoNxwEJgD3ABsBGYBowEVgKXAFOB84GJwAbgBuAK4GbgILAJ2AksA1YA84EpwDBgGzAT2F+y3gXA7Q3aJkmSJEnSAGr5Iq4kSW1gMnAhsA54BJgNnAfcBbyWzjMVuAjYBUwCbgEuBnYAE4DnCUVcgKXpsgBfBa5KX1eSJEmS1IJavog7bOh7Ykeo2/6DWwA4d8QVkZP0zyt7l6ZDQ6Lm6J/DABzXcWbkHPU7cCjUeRKGRU5Sv3x6ouApJ0yKnKR/tneuBSCXOzlykvp1d+8EYMIJV0ZOUr/1nUtiR6jFWkIBF8JZtqvT8ceL5nmCUMAFmAHcTSjgAqwHijf4DOBO4GRgLPBoQ1JLkiRJkprCPnElSYpvX9HwBGArMB4YXWGeBMhXeK2xhO4Zvkg4e/cByvezuyCdXukhSZIkScoIi7iSJGXHKGARMItwpu3iCvMtBW4CTkrHzwUKp02PAd4CXidccXN5o8JKkiRJkpqj5btTkCSpjTwE3A9sIfRhO43QvUJnyXyrgXuBpwk3NttB6HohT+iGYSOwCngDeBEY0YTskiRJkqQGsYgrSVJcm4Ep6fDVJc/N7GW5ZfTcyAxCFwrbCYXc6wcqnCRJkiQpPrtTkCSp9UwCbisaPyudtj5OHEmSJElSI3kmriRJrWcDcB2wBuhKH3OA7pihJEmSJEmNYRFXkqTW0wnMjR1CkiRJktQcdqcgSZIkSZIkSRlmEVeSJEmSJEmSbY2YbwAAFblJREFUMswiriRJkiRJkiRlmEVcSZIkSZJ6d2PsAJKkwc0iriRJkiRJvbOIK0mKyiKuJEmSJEmSJGVYR+wAkiQpm/L5vbEjcOjw1tgRyLM/6vq3d66Nun7Ixu/hjb0rY0fgwKHXYkdg976Xo66/O7876vohG7+HfL4rdgS6DmyKHaFVdACHgXwv8wwFDg7gOhcAtw/g60mSZBFXkiRJktRWxgELgT3ABcBGYBowElgJXAJMBc4HJgIbgBuAK4CbCQXdTcBOYBmwIn3d+cBlwCFgDXArEL+iL0kaFCziSpKkssaNnB5t3Zv3LE+HjouWAQ4A8LHhvxdl7S90fROAS0bOjbJ+gL/b82cADO14d7QMBw/9EoAZJ30hWobv7vpTAJJkRLQMhTPjhw19T5T17z+4BYAkGR5l/dBz9utxHWdGy1A4CzhhWLQMhasDTjlhUrQMWbhCoAqTgQuBdcAjwGzgPOAuoHA691TgImAXMAm4BbgY2AFMAJ4nFHEBzga6gY+m4/OBO4B5Dd4OSZIA+8SVJEmSJLWftYQCLoSzbOcBpwGPF83zBKGACzADuJtQwAVYDywpmvdt4J6i8XuASt92LgCSXh6SJNXMIq4kSZIkqd3sKxqeAGwFxgOjK8yT0Hu/uZIkRRWriHtjpPVKkiRJkgaPUcAiYBbhTNvFFeZbCtwEnJSOnwtcWfT8GEJ/uQU3A8uRJKlJYvWJeyPwYKR1S5IkSZIGh4eA+4EthL5xpxG6V+gsmW81cC/wNOHGZjsIXS8Uzs7dRDh+foHQN+5LwJcanF2SpCO8sZkkSZIkqZ1sBqakw1eXPDezl+WW0XMjM4CFwPZ0+Jz05539DSdJUj1q7U6hg747Yh9aZxZJkiRJkmKYBNxWNH5WOm19nDiSJB2tmjNxxxG+gdwDXABsJFyCMhJYCVwCTAXOByYCGwiXp1xB6CfoIOHSk52EbzVXpK87H7gMOASsAW4FukrWvQC4vfbNkiRJkiSpahuA6wjHpl3pYw6h6wRJkqKrtjuFycCFwDpCP0KzgfOAu4DX0nmmAhcBuwjfWN4CXEzoS2gC8Dw9l6acTWgMP5qOzwfuAObVvymSJEmSJNWlE5gbO4QkSZVU253CWkIBF8JZtvOA0wgdvRc8QSjgAswg3PlzRzq+HlhSNO/bwD1F4/cA06tOLUmSJEmSJEmDRLVF3H1FwxOArcB4YHSFeRJ67uLZHwvS16r0kCRJkiRJkqS2VuuNzUYBi4BZhDNtF1eYbylwE3BSOn4ucGXR82MI/eUW3AwsrzGLJEk61o2xA0iSJEmSBla1feIWPATcD2wh9I07jdC9QmfJfKuBe4GnCTc220HoeqFwdu6mdN0vEPrGfQn4Uu3xJUlSiRuBB2OHkCRJkiQNnGqKuJuBKenw1SXPzexluWX03MgMYCGwPR0+J/15ZxXrlyRJkiRJkqRBq9buFKo1CbitaPysdNr6Bq1PkqRW10Hf/b0PbUYQSZIkSVK21NqdQrU2ANcBa4Cu9DGH0HWCJEkKxhGuVNkDXABsJHRVNBJYCVwCTAXOByYS2tcbgCsI/ckfJHRRtJNw9cuK9HXnA5cBhwht8a2EtrjYAuD2BmyTJEmSJGmANaqI2wnMbdBrS5LUTiYDFwLrCP3NzwbOA+4CXkvnmQpcBOwiXNlyC3Axoc/5CcDz9HRhdDbhS9OPpuPzgTuAeQ3eDkmSJElSgzSqOwVJklSdtYQCLoSzbOcBpxFuCFrwBKGACzADuJtQwIXQVdGSonnfBu4pGr8HmD6wkSVJkiRJzWQRV5KkuPYVDU8AtgLjgdEV5kmA/ACsd0H6WpUekiRJkqSMsIgrSVI2jAIWAbMIZ9ourjDfUuAm4KR0/FzgyqLnxxD6yy24GVg+oEklSZIkSU3VqD5xJUlSbR4C7ge2EPrGnUboXqGzZL7VwL3A04Qbm+0gdL1QODt3E6F9f4HQN+5LwJcanF2SJEmS1EAWcSVJimczMCUdvrrkuZm9LLeMnhuZASwEtqfD56Q/7+xvOEmSJElSNtidgiRJrWUScFvR+FnptPVx4kiSJEmSGs0zcSVJai0bgOuANUBX+phD6DpBkiRJktSGLOJKktRaOoG5sUNIkiRJkprH7hQkSZIkSZIkKcMs4kqSJEmSJElShlnElSRJkiRJkqQMS/L5fOwMdXv22Wfzzz33XOwYUmbMnvHN2BGkTHn/R3b8EbAgdo5W9Cqd+VfpipphDTuirh/gebZFXf8r7I66/pDhndgROPjKltgR2DLqitgReP3116Ou/xe/+EXU9QO89dZbsSOwbVvc9wWAN998M3YEFi5caBtbp12Pbs7vevTfYseQMmH/Vy6LHUHKlPd9+PvkkolJuec8E1eSJEmSJEmSMswiriRJkiRJkiRlmEVcSZIkSZIkScqwlu4TF7gW+HbkDJKk7LoW24l6XYv7TpJU2bXYTtTrWtx3kqTyrqVCG9HqRVxJkiRJkiRJamt2pyBJkiRJkiRJGWYRV5IkSZIkSZIyzCKuJEmSJEmSJGWYRVxJkiRJkiRJyjCLuJIkSZIkSZKUYRZxJUmSJEmSJCnDLOJKkiRJkiRJUoZZxJUkSZIkSZKkDLOIK0mSJEmSJEkZZhFXkiRJkiRJkjLMIq4kSZIkSZIkZZhFXEmSJEmSJEnKMIu4kiRJkiRJkpRhFnElSZIkSVJfPg38ENgO7AbeSX9KkpogyefzsTNIkiRJkqRs2wh8BvgxYCFBkpqsI3aAQexk4PPAOI7+PdxYZl4bSEmNkMQOIGWAbaykRrCNVTvaAvyE6ttO21hJA21Qt68WceNZDvwT4VvM7shZJEmSJEnqzS2E49jvA/uLpt8XJ44kDS4WceM5HpgXO4QkSW2mlitdJElS9e4A9hCOZY+LnEWSBh2LuPF8B/gvwFMc/S3m9t4WyuWGATkScpDkjgwn6TBAkgw5Mu3Iz6OmDSFJcuTomQ6QKzxHjlw6PVcYJzyXKxkGyOVzR+bJkSOXD3Mk6Zw940khbchJQi4pTCseh6TkZy4Jz3NkmKJ5i6dRdhqUn35k2pHn8kc9x5Hn80eWPTIMJEk+zdezXJLki9ZX+ny+l2n5nvUleRLyYX1HhnuWTZJ8meWKpqU/j1quMK1ouHS50uEklyeXdPeMp9M4krv7yLTi55Oku+w04Mh0SufL5SHXXX4Yyj5H7thxcuE1SQrLFebJE/5t8ukfQOGRpP9KSc8jBE3Hc+mjeDhHPpeD3JCjpgHkj5o2hHxSmG/IkWXyR8bDz57lCtM60uH0Z9KRTusI05L0rTudliQdkBsafiYdJOm0JOkgl3SE//Okg1xuCEOY1dtbjNQO6r7Spa82trR9BapqY0vbV6CqNra0fQWqamNL29ewvr7b2NL2NWxf321s6TRKp1O+jS1tX6G6Nrb0ubB9fbexpe0rFI9XbmNL21fSefpqY0uHKcrTWxtb2r4eWV8fbWxp+wpU18aWDBfW12cbW9K+FtbXZxtb2r4e+YPpvY0tbV+BqtrYY9pX6LuNLde+gm2sBKcAv1nPgrnc8KL2FfpqY8u1r0CfbWy59jWsrfc2tlz7WkjZWxtbrn2FvtvYcm0pFab3dQwLfbex5drX0uXKtbHl2lcobXePbWPLta/Qdxtbrn0tLFdte9tzvNp3G9vb8WpvbWxp+3okYxXtbXH7Wliurzb2mPY1bGDfbWxJ+wpU1cYe075CdW1safsK1bWxJe1r2LyeNnZo7vN9vMMMHhZx4zkA3AP8N3r6CsoD742WSJKk1ueVLpIkNcbfE4q4z8YOIkmDkUXceOYB5wDbYgeRJKmN1HWliyRJ6tNcQr+4+4GDhHMD88CJMUNJ0mBhETee9UBn7BCSJLUZr3SRJKkxRpWZNqjvFC9JzWQRN57DwI+A73H0mULeeEWSpPp5pYskSY3xFeDLReM5whUwdggtSU1gETee/5k+JEnSwPFKF0mSGuNXgFuBPwGGAd8F1kZNJEmDiEXceB4GjgN+lXCZ508Jl4BKkqT6eaWLJEmN8TvAXxMKuZ8EngHuj5pIkgYRi7jxTAf+HPgZoR+hs4HfJzSEkiSpPuWudMmXm1GSJFVlUtHwA4Tj2BeA76fPeTauJDWBRdx47iN8e7kxHX8f8DQWcSVJ6o+TCQeYxb4YI4gkSW3iayXjO4Dx6fQ8MLnpiSRpELKIG89Wegq4AD9Pp0mSpPpdw7FF3GvLTJMkSdX5ZOwAkiSLuDGtB5YDSwjfXs4AVgGfSZ9fVm6h7u795SZLkjTYzQR+C3gv8GTR9FHAtmpewDZWkqRejQFuB36DcAz7j8BXgLf7WrC7u6uxySRpELCIG8/xwJvAf0zH3wJGA5cQGsSyRVxJklTWD4A3gHdx9GWfewjFXUmS1D+PASuBz6bjs4DHgSnREknSIJLk897rI5KHCX307UzHRxMOOn+3zLz+kiQ1QhI7gNQAazn6BiwA64CJFea3jZXUCLaxakdrgAtKpq0GPlRhfttYSQNtULevnokbz0R6CrgQOof/tSqWy8IfbHFjbJ7yspbJPL0zj9T6rgduIHSnsK5o+ijCWbqVtOv/WLu+j7TjdrXjNoHbJbWj7wGfI3QJCHAF4ebclWTlfySr/7fmql4WM4G5apXVXC3DM3HjeQn4BKF4C3AK8H3gP5SZN2t/6ObpW9Yymad35pFa30mEq1r+BPjDounvANujJIqrXd9H2nG72nGbwO2S2k0CHE6HCz+HAHvT4TxwYrNDVSmr/7fmql4WM4G5apXVXC3DM3Hj+RrhzKClhD/kK4E7oiaSJKl17UofM2MHkSSpDeWBH3Fsl0WSpCaxiBvPI4T+gyYTvoH4DPBy1ESSJEmSJJX3IvBhYFXsIJI0GNmdQmvI2inn5ulb1jKZp3fmkdRu2vV9pB23qx23CdwuqR29DHwA2EzoRiEh/E9UunloVmT1/9Zc1ctiJjBXrbKaq2V4Jq4kSZIkSerLtNgBJGkws4jbGv4odoAS5ulb1jKZp3fmkdRu2vV9pB23qx23CdwuqR39W+wAdcrq/625qpfFTGCuWmU1V8uwOwVJkiRJkiRJyrBc7ACSJEmSJEmSpMos4kqSJEmSJElShlnEbQ2fAn4KbAT+MHIWCHcj/THwI2B1hPV/C9gK/KRo2inAc8D/S3+OjpxnAfAaYR/9CJjexDzvAb4HvAKsB76YTo+1jyrlWUCcfXQ88C/AS2meQr88sfZPpTwLiPc3JCn7am0LbyV8jvgpcHGTMtajnjasFbatnranFbYLYAjwQ+CpdLwdtmkzx37WbYftktpRX8fKCfBg+vw6YFLRc5tp3HFtX7l+FXgR2A/cXOOysXJtJt7+mkX4/a0DfgCcV8OysXJtpjH7q69Ml6Z5Cuv9jRqWjZVrM3FrTC3DPnGzbwjwr8BU4FVgFTATeDlips3Ah4Btkdb/cWAP8AjwwXTa3cB24E7CG8ZoYH7EPAvSafc2KUOxM9LHWmAUsAa4DLiWOPuoUp4ribOPEmBEuu6hwD8SigSfIc7+qZTnU8T7G5KUfbW0heOBvwU+Avw74O+Bfw8cbm7kqtTahrXKttXa9rTKdgHMI3wuPBH4NO3xd7iZYz/rtsN2Se2mmmPl6cAX0p+/DjyQ/oTGHddWk+t04CxCG7eDns/8jTz+708uiLu/LiJ8wbsDmEY43v71KpeNkQsas7+qyTQS2AvkgYnAEkJxPva+qpQL4teYWoZn4mbfRwjfZPwcOAA8RvgGYzBbSfgQXexS4OF0+GFCoxMzT0xvEA5+Ad4hNCpnEm8fVcoTS55wEA3hQHpoOi3W/qmUR5J6U0tbeCnh88N+YBPhc8VHmpCxHrW2Ya2ybbW2Pa2yXWOB/wR8s2haq29TJe26XVIrq+ZY+VLCF5554J+AkwlfFsbOtZVQ6DpYx7IxcjVSNbl+QCiUQvg9jq1h2Ri5GqWaTHvoOZ4cUTQce19VyqUaWMTNvjOBLUXjrxK3AAbhn+1ZwtkxsyNnKXgX4cCP9OfpEbMU/FfC5QLforndOxQbB/wa8M9kYx8V54F4+2gI4VKNrYRLImPvn3J5IBt/Q5JaR6X3sSx+lqjGOPpuw1pp22ppe1plu74O3AJ0F01r9W2C8p9122G7pHZTzf9fb/M06ri2P+8LjXxP6e9rZ2V/XQc8U+eyzcoFjdlf1Wa6HNgAPA38bo3LNjsXZLPGlEkWcbMvKTMt9jcWHyP0JTQNmEu4pFNHewh4H3A+4YP+1yJkGAn8D+APgN0R1l+qNE/MfXQ4Xe9YwreGH+x99oYrlycLf0OS2kMWP0v0pdo2rJW2rZa2pxW269OEgvSaKudvhW0qqOWzbittl9Ruqvn/622eRh3X9ud9oZHvKf197Szsr08SiqWFbu+ysr9Kc0Fj9le1mZ4gdFVwGfDVGpdtdi6wxlQ1i7jZ9yrhJh8FY4HXI2UpKKx/K+GfMAuXjL1Jz2UxZxCyxfQm4WCtG/hLmr+PhhIOfv8aWFaUKdY+qpQn5j4C2AmsIPQ/m4W/odI8sfePpNZS6X0si58lelNLG9Zq2wbVtT2tsF0fA/4zoR+7x4DJwKO09jYVlPus2w7bJbWbav7/epunUce1/XlfaOR7Sn9fO/b+mkjovudS4O0al212LmjM/qp1e1cSTgw6tY5lm5ULslljyiSLuNm3Cng/cDZwHPA54MmIeUYQbjRSGP5Njr4zdixPAtekw9cA/ytiFji6n6XLae4+SoDFhH4E7yuaHmsfVcoTax+dRugLC2A4MIVwSUes/VMpT8y/IUmtqdL72JOEzw/DCJ8n3g/8S9PTVafWNqxVtq3WtqcVtutWwkHaOELWfwCuprW3CSp/1m317ZLaUTXHyk8Cnye0LxcCuwhXuTXyuLY/x/CNPP7vz2vH3l+/Qvhi97cJN9CqZdkYuRq1v6rJdA49Z8ZOSud7u8plY+TKao0pkzpiB1CfDhH6xfw/hL7UvgWsj5jnXYRvRiD8/fwN8L+bnOFvgU/Q823S7YQ7BS8hXMLwC2BG5DyfIFwymSecofL7TczzMUIj8mNC33sAtxFvH1XKM5M4++gMwg1JhhC+yFoCPAW8SJz9UynPd4j3NyQp+2ppC9en018mfK6YSzjTP4tqbcNaZdtqbXtaZbvKafXfVaXPuqto7e2S2lGlY+U56fPfAJYD0wk3XeoEfid9rpHHtdXkejewGjiRcOXdHwDjCV0INer4vz+5TiXu/voyMAZYVLTMh3pZNnauRv19VZPps4QvLg4CXcBVhGPK2PuqUq4s1JhaRpLP22WTJEmSJEmSJGWV3SlIkiRJkiRJUoZZxJUkSZIkSZKkDLOIK0mSJEmSJEkZZhFXkiRJkiRJkjLMIq4kSZIkSZIkZZhFXEmSJEmSJEnKMIu4kiRJkiRJkpRhFnElSZIkSZIkKcP+P0MkduZQNqXSAAAAAElFTkSuQmCC\n",
      "text/plain": [
       "<Figure size 1728x1152 with 6 Axes>"
      ]
     },
     "metadata": {
      "needs_background": "dark"
     },
     "output_type": "display_data"
    },
    {
     "data": {
      "image/png": "iVBORw0KGgoAAAANSUhEUgAAA6sAAAFFCAYAAAAKM3teAAAABHNCSVQICAgIfAhkiAAAAAlwSFlzAAALEgAACxIB0t1+/AAAADh0RVh0U29mdHdhcmUAbWF0cGxvdGxpYiB2ZXJzaW9uMy4xLjAsIGh0dHA6Ly9tYXRwbG90bGliLm9yZy+17YcXAAAfwElEQVR4nO3deXhtZ10v8G/OOaHQkmpDB0oY6g3ax+K1EKUiXLGUQqQMrXiCYCuUolwIeuGaCx64Fwmo16BEBiXgfQBbRCwkMrQFDJXBgcHahpmWQqClBKRigMYWTtOeff9YOz0708lwhvUm+/N5nv1k5V3vWvuXrP3uvb97Dbuj0WgEAAAASrKj7gIAAABgKWEVAACA4girAAAAFEdYBQAAoDjCKgAAAMURVgFgEwZGZ88bGJ09r+46ADi8ds/sOXP3zJ4L666jHQmrUJD+wbnh/sG5j9ZdB7STgdHZGwZGZy/cxKLnNW+wpXVO9V7YOdV7Q911wOG2e2bPR3fP7BnexKJnJrnwkBbDuuyquwA40voH5/5rkv+T5BFJjk/ytSRvT/Ink2NdP6yzNmBjBkZnVx3P40Pdh3Q8D4zOvjtJY3yo+8mHcr1sH51TvWcm+UiSbyW5/3zf9B2r9NuR5KtJHpDkifN901ccsSKhYLtn9pyZljE00TOy4hjaPbNn0Ria6Bk5pGNo98yeVyd55ETPyM8cyvWycfas0lb6B+d+OsnHk/wgya8meUiS1yd5YZLfrbE0YIMGRmeP9Hj+UpLrDsN62X5OSvKEA8zvT9JzhGqBrajuMfS1JJ8/jOtnnexZpd38UZKrJse6Lmxpu7Z/cO6DSb5cT0nAJv1RkqvGh7ovbGm7dmB09rCM5/Gh7j2Hep1sW+9P8uwk71ll/rOTvDvJwBGrCLaWWsfQRM/I6w7Hetk4e1ZpNw9M8qmljZNjXddPjnU1aqgH2LwVx/P4UPf140PdxjN1GkvS3znVe/+lMzqneu+dao/RG454VbB1jCXp3z2zZ9kY2j2zxxhqI/as0m5uTPKE/sG5kcmxru9sdiX9g3M/luSlSR6T5MQk30gyOjnWNdacf36Si5N0T451zTXbjknynVTnxv5ey7r+JMmTJse6Tl3hfu6e5L1JTk3yqMmxrq9ttmbYhm5M8oSB0dmR8aHuTY/npo6B0dn/keQFqc59/XSSF4wPdU8tdBgYnb04SZbsyc3A6OzPJPmDJL+Q5M4kH0vyivGh7k8eZE1sXf+U5ItJnpXkZUvmXZTkK/N90x/pnOrd0Eo7p3pPSPXa8/gk901yc5I3J3nFfN/0vs6p3kck+eckp833TV/bstzXk3x0vm/66S1tz0t1dMLxK9xPR5K3JHlSksfO901fs6FC4eCtOYYmekY+sntmcwe87J7Z82tJ/neqc16/lOR3J3pG/r5l/nCSMyd6Rs5cstwDU42bs5PcLcnVSUYmekY+sKlCWJM9q7SbFye5f6pDf1/YPzi37BO7dboo1Xlyv57k9FRPeH/UPzj3y835H0jSkerqcQv6k+xNcu6SdZ2dZNmFAfoH5+6W5F1JepM8UlCFZe4azwOjsy8cGJ3d7HhOkqFU43UgyTnNtssGRmfvscZyD07ywVRvrH4uyaNSfXj1jwOjs088iHrY+t6Y5KLOqd6dCw3NEPis5rzNeHKSeyV5bqrXnsFUr0fPb87/RKoPRc9uuc++JMcmeXznVG/rToqzk1w53zd9e+sdNGv8f6n2XJ0lqFKjNya5aPfMnrvG0O6ZPQc7hpLkV5I8J9WhxGelOj/1vc09tgfSk+q5/oZUz/UPS/Xh5GW7Z/Y89yDq4QCEVdrK5FjXNUl+OlWYfFmSG/oH5z7YPzj36A2u56WTY13Pmxzr+ujkWNd1k2Ndl6bak/rk5vzZVG8azm5Z7Nwkr0pySv/g3ClJ0j84d3yqNxyLwmr/4FxnkolUQfUXJ8e6vr7RvxW2u/Gh7mXjeWB09oMDo7MbGs9N80kGxoe6rxkf6v7nJM9M9cbk59ZY7vQkTx0f6v6/40PdXxgf6p4aH+p+dpI/T/IXA6Ozd9tELWwPb03yo9n/4UdSHY1zcpJLNrPC+b7pv5jvmz5/vm/6g/N909fN901fnuTVab72zPdN70vyd1n+2vP2VHth/1uSNAP0mVnhg9JUFyl7UpJHzfdNf2YzdcIhcsjHUNO9kpwz0TPysYmekatSPd8nyS+tsdwDk7x0omfkhRM9I5+e6Bn53ETPyEtSfdg5untmz4kHUROrEFZpO5NjXV+ZHOt6epITkjwlyXFJ/r5/cO6C9a6jf3Cus39w7rn9g3NX9A/OfaN/cO6WJM9LdVjWgivSfMPQPzi3M9VhW3+bZDL7966eneSWVIdtLdiV5B2pnhR/cXKsa2YTfya0hfGh7q+MD3UvG88Do7PrHs8Lqxof6r6zZb3TSb6fxWN6JR8fH+q+coX2V6V6Q/XwDdbBNjHfNz2X5K+T/GZL828m+Zv5vunvbWadnVO9HZ1Tved3TvVOdE713tg51fv9JK/M8teeM1v26J6b5PLmbeG156FJfiTVRWxa1/+6VN8dfOZ837QroVKriZ6RVcfQRM/IpsZQ0xUTPSP/ueR+vpy1n++/OdEz8qYV2seaP89ZYR4HSVilbU2Odf1gcqxrItWbyY8kGd7A4u9I8nup9uicl+SMVIekdLT0uSLJaf2Dcyen+jT7u5NjXdcmuSyLw+rk5FjXfMtyJ6Q6tOS6JP++wT8L2tL4UPcPxoe6Nzuek+TfVmj7Xta+tsP1q9TzzSQ/THKfDdbB9jKW5JzOqd6ezqneE1M99x/MRWFGm+u8OslTU+35f1kWv/b8XZKjk5zRvMBTb5IPZ/lrz7/O901/u2W5o5M8OtW54DceRI1wKI0lOWf3zJ6e5p7Lgx1DySF+vm9+F+yN8Xx/WAirtL1mUJzI2p+oJUn6B+cemOSXkzxhcqzr9ZNjXVdPjnVdl2Rna7/Jsa4vpDoP4tGpDqm6vDnr/Ul+vn9w7risfL7qt1Kd33pmkjf1D851BFiX8aHuDY3nFvs2eZc/vlLjwOjsfZLcPdV4pk3N901/Nsm/pDrH7plJPjPfN331ZtbVOdV79yS/neSi+b7pkfm+6U/M901flyWP3fm+6e+nOlrn0ane2F853ze9N9WpKcd2TvX+VFZ+7bktyWNTHRHwns6p3qM2UyccShM9I8vG0ETPyKbGUItD+nzfPKf2AfF8f1gIq7SV/sG5/v7BudNWmPULqa7+uR5dzZ/fbVnvrqx8rsP7Ur0puCusNs9n/ddUF8S4X6q9s4tMjnV9LtXFLZ6SxHd9wQoGRmf7B0ZnD3Y8H6yHD4zOnrVC+/9M9en9x49QHZTrDaneaP9GDm6P0D1S7fn57pL2lS7ktXAaysIhwJnvm74z1Yelv5bk57PC+arzfdMzqc4JPD3JpUsuyAR1OVRj6GD17J7Z88wV2p+T6uiG968wj4PkSYi20QyUw0l6+wfnXprqim5dqZ4An5DmhSfW4Yup3oSO9Q/OvSTVG4gXpTpkd+mYuiLVFX1vb97fgsuSvCLJJ1f7Cp3Jsa6P9w/O7U7y3v7Buf+cHOt68Trrg21vYHT2rvE8MDp7MOP5YH02yTsHRmdfleTKVM8Hv5bqCq2/Oj7UvfcI1UG5xlNdBOnYJJdudiXzfdPf7Zzq/VSSV3ZO9T4/1evKc1J9fcbS157LU329xr5Uj8cFl6W6aM2/z/dNL/uO4ub9fLlzqveXknw0ycWdU71Pb164CepySMbQIfCVJH+4e2bPqam+VvDOVDsjXpjkdyZ6Rr59oIXZHHtWaRuTY113pDos6nVJfivJNakueHR0kp+ZHOta11UPJ8e69qZ6M3xsqkuWvz3JPzTX272k+0dSvVn4QPP+F1ye6k3tSldibL2vDyR5epIXNYMxkGR8qPuA43l8qPtIXcX009l/2P4/pPpk/b8keej4UPd7j1ANFKx5CO5bklwy3zd920GubiDJbKoPRi5P9SHpi7LktWe+b/r6VF+vMTXfN31zy6zJVKesvG+Nmj+V6k34r6TePVmQiZ6Ru8bQRM/IwY6hgzGT5JGpzgP/QKr3eGcmOXuiZ+T1Nda1rXU0Go26awAAAIBFHAYMTf2Dcw9I8qU1ut04OdZ16pGoB9i8gdHZdY3n8aFu45nadU71/nCtPvN903c/ErXAVrR7Zs+aY2iiZ8QY2oKEVdjvm0kevEaf+TXmA2UwntlK1nqsAgdmDG1TDgMGAACgOC6wBAAAQHGEVQAAAIojrAIAAFAcYRUAAIDiCKsAAAAUR1gFAACgOMIqAAAAxRFWAQAAKI6wCgAAQHF21V3AFjG8ynQ7G15lup0NrzLdroZXmW5nw6tMt7PhVabb1fAq0+1seJXpdja8ynS7Gl5lup0NrzLdzoZXmW5Xw6tMF6Wj0WjUXcNW0PpP6qitirL4nyznf7KY/8dy/ifL+Z8s5v+xnP/Jcv4ni/l/LOd/spz/yWJb4v/hMGAAAACKI6wCAABQHGEVAACA4girAAAAFEdYBQAAoDjCKgAAAMUp/qtrXvXWHzSu/OQddZcBsK0c2ztfdwlsAe951EPrLgG2nHNPGqi7BNhyJnpGnpnk4qXtxe9ZPelexZcIAADA5p2yUqMkCAAAQHGEVQAAAIpT/DmrSRoX3uv5dddQjEtmX5ckOfGYn6+5kjLcfOsnmlM7a62jLHcmSY4/5oya6yjHd269qjnlcVK5M41GI527Tqq7kGLccefNzSmPkf2qx8kJ9/y5ugspwsLzyFGd96u5knLsnb8pSfJ79/+tmispxyu+/udpNBo5qvO+dZdShNvvmGlOeW7dr3qf5rlkv73zN6XRaLw8yfDSefasAgAAUBxhFQAAgOIIqwAAABRHWAUAAKA4wioAAADFEVYBAAAojrAKAABAcYRVAAAAiiOsAgAAUBxhFQAAgOIIqwAAABRHWAUAAKA4wioAAADFEVYBAAAojrAKAABAcYRVAAAAiiOsAgAAUBxhFQAAgOIIqwAAABRHWAUAAKA4wioAAADFEVYBAAAojrAKAABAcYRVAAAAiiOsAgAAUBxhFQAAgOIIqwAAABRnM2F1V5KONfp0bmK9AAAAkGT9YfWUJFckuTTJtUne32y/Z5KpJD1JLkzymiQfTvLaJCc1+08muSbJb6yy7uEkjQPcAAAAaDO7NtD3rCQPS/LZJG9N8uwkpyd5ZZKZZp/HJHl4ku+n2rv6oiRfb97P55rL3X4oCgcAAGD72khYnUoVVJNkMMnVzd/f0dLn3amCapLsS7I7yWOT7ExycpITsj/YAgAAwIo2cs7qD1umH5Tk5iSnJTlulT7/K1VAfXyqPa7XZeVzXYeb7avdAAAAaDObucBSV5KxJOcn+eMkb16l372TfCbJnUlOTXXIMAAAAKxpI4cBL3hDklcnuSnVOaiPS3VY8G1L+v1ZkrcneUaSG5P8fZJjNl0pAAAAbWO9YfWGJGc3py9YMu9pqyzz1VQXZAIAAIAN2cxhwAAAAHBYCasAAAAUR1gFAACgOMIqAAAAxRFWAQAAKI6wCgAAQHGEVQAAAIojrAIAAFAcYRUAAIDiCKsAAAAUR1gFAACgOMIqAAAAxRFWAQAAKI6wCgAAQHGEVQAAAIojrAIAAFAcYRUAAIDiCKsAAAAUR1gFAACgOMIqAAAAxRFWAQAAKI6wCgAAQHGEVQAAAIojrAIAAFAcYRUAAIDiCKsAAAAUR1gFAACgOB2NRqPuGtbS6OjoqLsGgG2l0WjEcytr8TiBjTNuYOMajcbLkwwvbbdnFQAAgOLsqruA9Xj00c+pu4RifOi2NyZJfuKY82qupAzX3/qe5tSWeCgfIXckSd72UxfWW0ZBLvj8xUmSozrvV28hhdg7f1OS5JijfqLmSspx697rkyQ/e49n1FxJOa7+wSVJkh0dx9ZcSRn2NW5JknTkqJorKUcje5Mk377otJorKcdJb/lic2pnrXWU484kyU8es7vmOspx7a0TSZKjj+qtuZJy3LZ3etV59qwCAABQHGEVAACA4girAAAAFEdYBQAAoDjCKgAAAMURVgEAACiOsAoAAEBxhFUAAACKI6wCAABQHGEVAACA4girAAAAFEdYBQAAoDjCKgAAAMURVgEAACiOsAoAAEBxhFUAAACKI6wCAABQHGEVAACA4girAAAAFEdYBQAAoDjCKgAAAMURVgEAACiOsAoAAEBxhFUAAACKI6wCAABQHGEVAACA4girAAAAFEdYBQAAoDiHK6w+OMkjD9O6AQAA2OaEVQAAAIrjMGAAAACKs9GwuitJxxp9OjdZCwAAACRZX1g9JckVSS5Ncm2S9zfb75lkKklPkguTvCbJh5O8tjn/xCQTST6W5J+S/PIq6x9O0jjADQAAgDaza539zkrysCSfTfLWJM9OcnqSVyaZafZ5TJKHJ/l+qvD61FTnrV6X5NgkH0hyQ5JPHZLKAQAA2LbWexjwVKqgmiSDSX4nyQlJ3tHS592pguqCS1IF1SS5JcnrkvzKpisFAACgbaw3rP6wZfpBSW5OclqS41bps5KOrHxY73Bz3mo3AAAA2sxGL7DUlWQsyflJ/jjJmw/Q9xlJTm1Z7rdT7X0FAACAA1rvOasL3pDk1UluSnXu6uNSHRZ82wp9L03yB0lObv7+p6kOJwYAAIADWk9YvSHJ2c3pC5bMe9oqy1zcvAEAAMCGbfQwYAAAADjshFUAAACKI6wCAABQHGEVAACA4girAAAAFEdYBQAAoDjCKgAAAMURVgEAACiOsAoAAEBxhFUAAACKI6wCAABQHGEVAACA4girAAAAFEdYBQAAoDjCKgAAAMURVgEAACiOsAoAAEBxhFUAAACKI6wCAABQHGEVAACA4girAAAAFEdYBQAAoDjCKgAAAMURVgEAACiOsAoAAEBxhFUAAACKI6wCAABQnI5Go1F3DWtpdHR01F0DwLbSaDTiuZW1eJzAxhk3sHGNRuPlSYaXttuzCgAAQHF21V3Aehxz1E/UXUIxbt17fZJk184Ta66kDHfceXOS5Phjzqi5knJ859arkiRHH9VbcyXluG3vdJLkMUc/t+ZKynDlbW9Ikjz66OfUXEk5PnTbG5MkJx3ziJorKce3b/1YkuSOfW+ruZIy7NpxQZLk1GOeXHMl5fjSre9KknR0HFNzJeVoNG5Nksyc/+CaKylDz19/Oknytd3epy34sYnqfdr85feuuZJydD7x31adZ88qAAAAxRFWAQAAKI6wCgAAQHGEVQAAAIojrAIAAFAcYRUAAIDiCKsAAAAUR1gFAACgOMIqAAAAxRFWAQAAKI6wCgAAQHGEVQAAAIojrAIAAFAcYRUAAIDiCKsAAAAUR1gFAACgOMIqAAAAxRFWAQAAKI6wCgAAQHGEVQAAAIojrAIAAFAcYRUAAIDiCKsAAAAUR1gFAACgOMIqAAAAxRFWAQAAKI6wCgAAQHGEVQAAAIojrAIAAFAcYRUAAIDibDSs7krSsUafzk3WAgAAAEnWF1ZPSXJFkkuTXJvk/c32eyaZStKT5MIkr0ny4SSvbc7fneSTSf4pyVuTvC7JmSusfzhJ4wA3AAAA2syudfY7K8nDknw2VfB8dpLTk7wyyUyzz2OSPDzJ95P0JXlRkv4k303yoFSh9V2HqnAAAAC2r/WG1alUQTVJBpNc3fz9HS193p0qqCbJQJI/ThVUk+QLSd55UJUCAADQNtZ7zuoPW6YflOTmJKclOW6VPh1Z/yG8w83+q90AAABoMxu9wFJXkrEk56fac/rmVfpNJBlK8iPN338yyVM2UyAAAADtZ72HAS94Q5JXJ7kp1bmrj0t1WPBtS/pdneRVSd6XZD7V4cDviAsmAQAAsA7rCas3JDm7OX3BknlPO8By78riCyr9eZLZdVcGAABA29roYcDr1ZfkJS2/P6DZ9oXDdH8AAABsIxs9DHi9rkvyrCTXJPlB8/acJPsO0/0BAACwjRyusHpbkucdpnUDAACwzR2uw4ABAABg04RVAAAAiiOsAgAAUBxhFQAAgOIIqwAAABRHWAUAAKA4wioAAADFEVYBAAAojrAKAABAcYRVAAAAiiOsAgAAUBxhFQAAgOIIqwAAABRHWAUAAKA4wioAAADFEVYBAAAojrAKAABAcYRVAAAAiiOsAgAAUBxhFQAAgOIIqwAAABRHWAUAAKA4HY1Go+4aDuiv3re38bb33V53GQDbyrG983WXwBbwnkc9tO4SYMs596SBukuALWeiZ+TlSYaXttuzCgAAQHGEVQAAAIojrAIAAFCc4s9ZTXJhkotrrgEAAIDD48KskPm2QlgFAACgzTgMGAAAgOIIqwAAABRHWAUAAKA4wioAAADFEVYBAAAojrAKAABAcYRVAAAAiiOsAgAAUBxhFQAAgOIIqwAAABRHWAUAAKA4wioAAADFEVYBAAAojrAKAABAcXbVXQBsQqPuAgAAOCgddRdA+exZBQAAoDjCKgAAAMVxGDAleH6Sv0wyl+RNSR6SZE+SD6614I6Ou1UTHa1HklSfwXQsfBbTsWP1eUk67prf2rZz1X4LbR0d6+2/c1nbjgPc147W9Wb5enescF87lvRbtP7mvB2L+i+vbaFtR6N12cXrW7SOxkJbxwr9O9bo17GobcV5Ldt0f7+s0a85r2OF/iu07ehYvK7FbQu/r7T+ldaxvG3x/R9c/8U1tSy7tP+K69h/5Px61lG1NVatY+m8xetoLGtbuP+V/uaV+h9oHYv/hsam+rcusyOtbUvrbVnviv/LxgHWm+VtK9a7QttK/Za0La5tfetYWsei9a1j3lrr7djR2m/fqutYuf9Cv30H7Ld0fYvn7Vuzf+v8A/XPetfb0pYd+9Y9b9H8HQf+m5ctu2N968hK62htW9Jv8X0uzGvpvtC2/OW2deC19F/e1jKQsqzjon47lvw8wLyW6caitp2r9tvff+eq81rX0ehYab37l73rfhf6L1rvJtt27Fph3gptHa1tu5Yte9cyHcvnLUx3LFpH5/K25nRHy7IL8xd+7uhonbdz2Tp2NOvdmfMDG2HPKiW4KMktSR6b5IQkz0wyUmtFAABArYRVSrDwkek5qfawfiaLP7sFAADajLBKCa5JdcjvOUkmk3Ql2XfAJQAAgG3NOauU4FlJHpzkq0luS3KvVIcCAwAAbcqeVUpwZZKpJN9r/v4fSV5dXzkAAEDd7FmlTndPcnSS45Mcl/3nqR6b5D51FQUAANRPWKVO/z3JC1IF02uyP6zekuT1dRUFAADUT1ilTq9t3n47yZ/VXAsAAFAQYZUS/FmShyc5JYsfk2+tpRoAAKB2wiol+KskvUk+neTOZlsjwioAALQtYZUS/GyS01IFVAAAAF9dQxE+n+TedRcBAACUw55VSnB8ki8muSrJ3pb2J9VTDgAAUDdhlRIM110AAABQFmGVEvxD3QUAAABlEVYpwVz2X1zpbkk6k9ya5NjaKgIAAGolrFKCriW/n5fkjDoKAQAAyuBqwJToPUnOqrsIAACgPvasUoInt0zvSPW9q75zFQAA2piwSgme2DJ9R5IbkpxbTykAAEAJhFVK8My6CwAAAMrinFVKcN8k705yc5JvJ/nbZhsAANCmhFVK8JdJLktynyQ9SS5vtgEAAG1KWKUEJ6QKp3c0bxc32wAAgDYlrFKC7yS5IMnO5u2CJP9Ra0UAAECthFVKcFGSpyT5tyTfSrI7LroEAABtzdWAKcHvJ3lGku82f+9O8qpUIRYAAGhDwiol+OnsD6pJMpvkIetZcF/j9sNSEAAAUC+HAVOCHUmOa/m9Oz5IAQCAtiYQUILRJB9PMpGkker81T+stSIAAKBWHY1Go+4aIElOS3JWko4kH0ryxQP0bX3QdhzOoiiGbd5+bPP2ZLu3H9scWJWwylbkha392ObtxzZvT7Z7+7HNgVU5ZxUAAIDiCKsAAAAUR1gFAACgOMIqAAAAxfHVNWxFL6+7AI4427z92ObtyXZvP7Y5sCpXAwYAAKA4DgMGAACgOMIqAAAAxRFW2Wp+KcmXknwlyZ6aa+HQekuSm5N8vqWtO8mVSb7c/Hlcy7wXp3ocfClJ/xGqkUPrfkk+kuTaJF9I8vxmu+2+fd09yVVJPpNqmy+cr2ibb387k3wqyRXN321zYE3CKlvJziSvT/K4JKcleVrzJ9vDxak+jGi1J8mHkvx48+fCBxSnJXlqkgc1lxlL9fhga7kjyVCSn0zysCTPS7Vtbffta2+Ss5KcnuTBqbbjw2Kbt4Pnp/pgaoFtDqxJWGUrOSPVJ61fTXJ7kkuTnFtrRRxK/5hkdknbuUkuaU5fkuS8lvZLU73x/Vqqx8UZR6BGDq1vJZlqTs+leiPbE9t9O2sk+c/mdGfz1ohtvt3dN8njk7yppc02B9YkrLKV9CS5qeX3bzTb2L5OShVo0vx5YnPaY2H7OSXJQ5L8S2z37W5nkk+nOuz/ytjm7eA1SV6UZF9Lm20OrElYZSvpWKHNdy+1J4+F7eWeSf42yQuS3HKAfrb79nBnqkOA75tqj9lPHaCvbb71PSHVBxPXrLO/bQ7cRVhlK/lGqguyLLhvkm/WVAtHxreTnNycPjnVG57EY2E76UwVVP86ybuabbZ7e/heko+mOi/RNt++HpHkSUluSHV471lJ3hbbHFgHYZWt5F9TXYjhx5LcLdUFGC6rtSIOt8uSPKM5/Ywk721pf2qSo1I9Hn481RVG2Vo6krw51bmqf9rSbrtvXyck+dHm9D2SnJ3kutjm29mLUwXOU1Jtyw8nuSC2ObAOu+ouADbgjiS/lWQy1TlPb0n11QdsD3+T5Mwkx6f6ZP1lSUaSvDPJs5J8PclAs+8Xmu1fTPW4eF6qQwvZWh6R5NeTfC7VOYxJ8pLY7tvZyakuprMz1Qfm70z1VSafiG3eboxzYE0djYbTAAAAACiLw4ABAAAojrAKAABAcYRVAAAAiiOsAgAAUBxhFQAAgOIIqwAAABRHWAUAAKA4wioAAADF+f8SgKGmzLDDRAAAAABJRU5ErkJggg==\n",
      "text/plain": [
       "<Figure size 1152x1152 with 2 Axes>"
      ]
     },
     "metadata": {
      "needs_background": "dark"
     },
     "output_type": "display_data"
    }
   ],
   "source": [
    "plotName = 'test'\n",
    "genenames = gdf['symbol'].values\n",
    "samplenames = [] #skipping names to replace in intaglio\n",
    "data = gdf.iloc[0:,7:37].values\n",
    "fig, ax = plt.subplots(figsize=[16,16])\n",
    "im, cbar = heatmap(data, genenames, samplenames, ax=ax)\n",
    "# texts = annotate_heatmap(im, valfmt=\"{x:.1f} t\")\n",
    "plt.show()\n",
    "\n",
    "\n",
    "genenamesr = rdf['symbol'].values\n",
    "samplenamesr = [] #skipping names to replace in intaglio\n",
    "datar = rdf.iloc[0:,7:16]\n",
    "genenamesm = mdf['symbol'].values\n",
    "samplenamesm = [] #skipping names to replace in intaglio\n",
    "datam = mdf.iloc[0:,5:41]\n",
    "genenameshc = mdf['symbol'].values\n",
    "samplenameshc = [] #skipping names to replace in intaglio\n",
    "datahc = hcdf.iloc[0:,2:8]\n",
    "\n",
    "figr, (axr,axm, axhc) = plt.subplots(1,3,figsize=[24,16])\n",
    "# figr = plt.figure(figsize=[16,2],constrained_layout=True)\n",
    "# fig_grid = figr.add_gridspec(1,28)\n",
    "# axr = figr.add_subplot(fig_grid[0,0:8])\n",
    "# axm = figr.add_subplot(fig_grid[0,9:])\n",
    "imr, cbarr = heatmap_rods(datar, genenamesr, samplenamesr, ax=axr)\n",
    "imm, cbarm = heatmap_glia(datam, genenamesm, samplenamesm, ax=axm)\n",
    "imhc, cbarhc = heatmap_haircell(datahc, genenameshc, samplenameshc, ax=axhc)\n",
    "plt.show()\n",
    "\n",
    "\n",
    "genenamessq = sqdf['symbol'].values\n",
    "samplenamessq = [] #skipping names to replace in intaglio\n",
    "datasq = sqdf.iloc[0:,8:21]\n",
    "figsq, axsq = plt.subplots(figsize=[16,16])\n",
    "imsq, cbarsq = heatmap_sqcones(datasq, genenamessq, samplenamessq, ax=axsq)\n",
    "plt.show()\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# genenames = gdf['symbol'].values\n",
    "# samplenames = [] #skipping names to replace in intaglio\n",
    "# data = gdf.iloc[0:,7:37].values\n",
    "\n",
    "# genenamesr = rdf['symbol'].values\n",
    "# samplenamesr = [] #skipping names to replace in intaglio\n",
    "# datar = rdf.iloc[0:,7:16]\n",
    "# genenamesm = mdf['symbol'].values\n",
    "# samplenamesm = [] #skipping names to replace in intaglio\n",
    "# datam = mdf.iloc[0:,5:41]\n",
    "\n",
    "# genenamessq = sqdf['symbol'].values\n",
    "# samplenamessq = [] #skipping names to replace in intaglio\n",
    "# datasq = sqdf.iloc[0:,8:21]\n",
    "\n",
    "# figr = plt.figure(figsize=[8,24],constrained_layout=True)\n",
    "# fig_grid = figr.add_gridspec(3,2)\n",
    "# ax = figr.add_subplot(fig_grid[0,:])\n",
    "# axr = figr.add_subplot(fig_grid[1,0])\n",
    "# axm = figr.add_subplot(fig_grid[1,1])\n",
    "# axsq = figr.add_subplot(fig_grid[2,:])\n",
    "# # axm = plt.subplot(2,2,2)\n",
    "# im, cbar = heatmap(data, genenames, samplenames, ax=ax,cmap=\"inferno\", cbarlabel=\"fpkm\")\n",
    "# imr, cbarr = heatmap_rods(datar, genenamesr, samplenamesr, ax=axr, cmap=\"inferno\", cbarlabel=\"cpm\")\n",
    "# imm, cbarm = heatmap_glia(datam, genenamesm, samplenamesm, ax=axm, cmap=\"inferno\", cbarlabel=\"tmm\")\n",
    "# imsq, cbarsq = heatmap_sqcones(datasq, genenamessq, samplenamessq, ax=axsq, cmap=\"inferno\", cbarlabel=\"counts\")\n",
    "\n",
    "# plt.show()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "***\n",
    "## sparkline plot for markdown file<a id='sparkline'>∮</a>\n",
    "***\n",
    "[Back to Index](#Index)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "genename = 'grk7a'\n",
    "n,h = plotBars(genename)\n",
    "fH, ax = plt.subplots(figsize=[2.25,.25])\n",
    "ax.axhline(y = np.mean(h), color = 'white', linewidth = .2, alpha = 1, solid_capstyle='butt', zorder = 0)\n",
    "pH = ax.bar(n, h, width=.8, bottom=None, align='center', data=None, color=pC['plt'], zorder = 1)\n",
    "ax.text(.90, 1, genename,horizontalalignment='left',verticalalignment='top',transform=ax.transAxes, fontsize=7, zorder = 0)\n",
    "ax.text(.97, 0.2, (np.around(np.mean(h),decimals=0)),horizontalalignment='left',verticalalignment='top',transform=ax.transAxes, fontsize=6, zorder = 0)\n",
    "ax.set_axis_off()\n",
    "# plt.savefig(\"/Users/angueyraaristjm/Documents/LiMolec/zfGenotyping/targetGenes/202001_zfDEGs/resources/{0}.svg\".format(genename), transparent=True, format=\"svg\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "'{0}fpkm'.format(np.around(np.mean(h),decimals=0))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# genename = 'kcnk1'\n",
    "# n,h = plotBars_sqcones(genename)\n",
    "# fH, ax = plt.subplots(figsize=[2.2,.25])\n",
    "# ax.axhline(y = np.mean(h), color = 'white', linewidth = .2, alpha = 1, solid_capstyle='butt')\n",
    "# pH = ax.bar(n, h, width=.8, bottom=None, align='center', data=None, color=sqC['plt'])\n",
    "# ax.text(.92, 1, genename,horizontalalignment='left',verticalalignment='top',transform=ax.transAxes, fontsize=6)\n",
    "# ax.text(.97, 0.2, (np.around(np.mean(h),decimals=0)),horizontalalignment='left',verticalalignment='top',transform=ax.transAxes, fontsize=4)\n",
    "# ax.set_axis_off()\n",
    "# plt.savefig(\"/Users/angueyraaristjm/Documents/LiMolec/zfGenotyping/targetGenes/202001_zfDEGs/resources/{0}sq.svg\".format(genename), transparent=True, format=\"svg\")"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "***\n",
    "## PCA <a id='pca'>∮</a>\n",
    "***\n",
    "[Back to Index](#Index)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### This is supposed to run only once. Creates sparklines and a markdown file with entries for these genes"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 175,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "/Users/angueyraaristjm/Documents/LiMolec/zfRNAseq/20190827/20190827_DESeq2/PCA_genes\n"
     ]
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>symbol</th>\n",
       "      <th>genename</th>\n",
       "      <th>PC1</th>\n",
       "      <th>PC2</th>\n",
       "      <th>PC3</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>opn1mw2</td>\n",
       "      <td>opsin 1 (cone pigments) medium-wave-sensitive 2</td>\n",
       "      <td>-0.048248</td>\n",
       "      <td>0.014348</td>\n",
       "      <td>0.004286</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>ntf3</td>\n",
       "      <td>neurotrophin 3</td>\n",
       "      <td>-0.047881</td>\n",
       "      <td>0.015787</td>\n",
       "      <td>0.002621</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>opn1mw1</td>\n",
       "      <td>opsin 1 (cone pigments) medium-wave-sensitive 1</td>\n",
       "      <td>-0.047535</td>\n",
       "      <td>-0.001411</td>\n",
       "      <td>-0.013467</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>opn1mw3</td>\n",
       "      <td>opsin 1 (cone pigments) medium-wave-sensitive 3</td>\n",
       "      <td>-0.043933</td>\n",
       "      <td>0.019464</td>\n",
       "      <td>0.014728</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>hs3st3l</td>\n",
       "      <td>heparan sulfate (glucosamine) 3-O-sulfotransfe...</td>\n",
       "      <td>-0.042757</td>\n",
       "      <td>0.020579</td>\n",
       "      <td>0.004224</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>22920</th>\n",
       "      <td>opn1lw2</td>\n",
       "      <td>opsin 1 (cone pigments) long-wave-sensitive 2</td>\n",
       "      <td>0.048579</td>\n",
       "      <td>-0.019831</td>\n",
       "      <td>0.006714</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>22921</th>\n",
       "      <td>s100v2</td>\n",
       "      <td>S100 calcium binding protein V2</td>\n",
       "      <td>0.048884</td>\n",
       "      <td>-0.015291</td>\n",
       "      <td>-0.005555</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>22922</th>\n",
       "      <td>thrb</td>\n",
       "      <td>thyroid hormone receptor beta</td>\n",
       "      <td>0.049103</td>\n",
       "      <td>-0.009038</td>\n",
       "      <td>0.001115</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>22923</th>\n",
       "      <td>slc32a1</td>\n",
       "      <td>solute carrier family 32 member 1</td>\n",
       "      <td>0.051255</td>\n",
       "      <td>-0.015804</td>\n",
       "      <td>-0.005305</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>22924</th>\n",
       "      <td>opn1lw1</td>\n",
       "      <td>opsin 1 (cone pigments) long-wave-sensitive 1</td>\n",
       "      <td>0.052043</td>\n",
       "      <td>-0.011839</td>\n",
       "      <td>0.012526</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>22925 rows × 5 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "        symbol                                           genename       PC1  \\\n",
       "0      opn1mw2    opsin 1 (cone pigments) medium-wave-sensitive 2 -0.048248   \n",
       "1         ntf3                                     neurotrophin 3 -0.047881   \n",
       "2      opn1mw1    opsin 1 (cone pigments) medium-wave-sensitive 1 -0.047535   \n",
       "3      opn1mw3    opsin 1 (cone pigments) medium-wave-sensitive 3 -0.043933   \n",
       "4      hs3st3l  heparan sulfate (glucosamine) 3-O-sulfotransfe... -0.042757   \n",
       "...        ...                                                ...       ...   \n",
       "22920  opn1lw2      opsin 1 (cone pigments) long-wave-sensitive 2  0.048579   \n",
       "22921   s100v2                    S100 calcium binding protein V2  0.048884   \n",
       "22922     thrb                      thyroid hormone receptor beta  0.049103   \n",
       "22923  slc32a1                  solute carrier family 32 member 1  0.051255   \n",
       "22924  opn1lw1      opsin 1 (cone pigments) long-wave-sensitive 1  0.052043   \n",
       "\n",
       "            PC2       PC3  \n",
       "0      0.014348  0.004286  \n",
       "1      0.015787  0.002621  \n",
       "2     -0.001411 -0.013467  \n",
       "3      0.019464  0.014728  \n",
       "4      0.020579  0.004224  \n",
       "...         ...       ...  \n",
       "22920 -0.019831  0.006714  \n",
       "22921 -0.015291 -0.005555  \n",
       "22922 -0.009038  0.001115  \n",
       "22923 -0.015804 -0.005305  \n",
       "22924 -0.011839  0.012526  \n",
       "\n",
       "[22925 rows x 5 columns]"
      ]
     },
     "execution_count": 175,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# deseqdir= \"~/Documents/LiMolec/zfRNAseq/20190827/20190827_DESeq2/00_rodsVcones/\"\n",
    "# fname = \"fromPython_PCgenes_RvC_fromPython.md\"\n",
    "\n",
    "# deseqdir= \"~/Documents/LiMolec/zfRNAseq/20190827/20190827_DESeq2/00_USvsLM/\"\n",
    "# fname = \"fromPython_PCgenes_USvLM.md\"\n",
    "\n",
    "deseqdir= \"~/Documents/LiMolec/zfRNAseq/20190827/20190827_DESeq2/00_LvsM/\"\n",
    "fname = \"fromPython_PCgenes_LvM.md\"\n",
    "\n",
    "# deseqdir= \"~/Documents/LiMolec/zfRNAseq/20190827/20190827_DESeq2/00_UvsS/\"\n",
    "# fname = \"fromPython_PCgenes_UvS.md\"\n",
    "\n",
    "\n",
    "\n",
    "pcadir= \"~/Documents/LiMolec/zfRNAseq/20190827/20190827_DESeq2/PCA_genes/\"\n",
    "%cd {pcadir}\n",
    "pcr = pd.read_csv(deseqdir + \"pcaWeights.csv\")\n",
    "pcr.sort_values(by=['PC1'])\n",
    "pcr"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 176,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "image/png": "iVBORw0KGgoAAAANSUhEUgAAAJkAAAAeCAYAAAA/6bzaAAAABHNCSVQICAgIfAhkiAAAAAlwSFlzAAALEgAACxIB0t1+/AAAADh0RVh0U29mdHdhcmUAbWF0cGxvdGxpYiB2ZXJzaW9uMy4xLjAsIGh0dHA6Ly9tYXRwbG90bGliLm9yZy+17YcXAAAC6ElEQVR4nO3bT4hVVRzA8c/TRmgxSNsxDBoVcvqHbtoE2maGWYW4ElwFZlKbYlr1ZyCx6CER1AQGQWi40I2KoinYRiKiWSS2k/6hUatiaChGnBbn+rhzGe+bN74zvjf9vpt3f+fce37n8b7vdx7nzDTm5+cFQU7W3O8JBKufkCzITkgWZCckC7ITkgXZCcmC7DxwvycQ9Dw78Dg+KuI1+AQPYhb72w3Q85Kd23m+tZE3fnmssZR72933P2SPJMsMtuISNkmyzGIS/+AMzmKquPcqfirGGMczeBsvFm2fSdLdrkve85IFXeFhfI9TOIIPMYj3JJnexI84gTl8i0+LZ3fgeVzEW0XbVhzEb9oIRp9J1klVCxbwPp5CE+uwFgOYR6N4VYqr4vyCRyVfbuEH7JIq3iP4uS55X0m2mhiYHm59Yea2Xc/9hdmHzZI8s3ijiA9Jy+Q7RftxfImPsUWqfjcwja8lqQ5iQlomb+HXdsn7WrKobEvmSOn6pLQ8ltlbifdV4q8q7a90kryvJetlVrhSdcLulU4Ykt2F0QMzLUkuTA3esySdSNfDgi6LdpINrcgslkd1bkM1fd3OBQamh2/cuZ7bdn1DB8+3i5fal4ub3Rys5yrZuZ3nWx/c+OWxug/unhk9MNPKdWFqcEM5DrpHO8m6avQyqMtf7btZ09fizjK4yBJYm2sJy+dy51qtiJ3Mqy/ouUrWbbr926qO8m+pPmMEo9K2xhVp83aLtLP/Ov7FY3ih9MyktJH7B47WDb7qJVstfPfESEvg7VevNcpxmWpfOV6sr7i8hqfxrHS09Jd0VrlO2lsbwt+lNA8V7YfxgRySTUxMLPoGm81mo9zXbDYXVI5qX929wYrzBf7ERjyH05JYDbyKd7FeEpCFpwS1NOIfSQKM4UkMSzv4I/gGn+M16VRgCC/jJelEYFJaLn/HsbrBQ7IgO/FHi0F2QrIgOyFZkJ2QLMhOSBZkJyQLshOSBdn5D2j92BtCOZbGAAAAAElFTkSuQmCC\n",
      "text/plain": [
       "<Figure size 162x18 with 1 Axes>"
      ]
     },
     "metadata": {
      "needs_background": "dark"
     },
     "output_type": "display_data"
    }
   ],
   "source": [
    "# Sparkline plots\n",
    "genename = 'spock3'\n",
    "n,h = plotBars(genename)\n",
    "fH, ax = plt.subplots(figsize=[2.25,.25])\n",
    "ax.axhline(y = np.mean(h), color = 'white', linewidth = .2, alpha = 1, solid_capstyle='butt', zorder = 0)\n",
    "pH = ax.bar(n, h, width=.8, bottom=None, align='center', data=None, color=pC['plt'], zorder = 1)\n",
    "ax.text(.90, 1, genename,horizontalalignment='left',verticalalignment='top',transform=ax.transAxes, fontsize=7, zorder = 0)\n",
    "ax.text(.97, 0.2, (np.around(np.mean(h),decimals=0)),horizontalalignment='left',verticalalignment='top',transform=ax.transAxes, fontsize=6, zorder = 0)\n",
    "ax.set_axis_off()\n",
    "plt.savefig(\"./resources/{0}.svg\".format(genename), transparent=True, format=\"svg\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "\n",
    "# nTop = 100;\n",
    "# print(\"Started...\")\n",
    "# for i in range(nTop):\n",
    "#     genename = pcr['symbol'].iloc[i]\n",
    "#     if os.path.isfile(\"./resources/{0}.svg\".format(genename)) == False:\n",
    "#         n,h = plotBars(genename)\n",
    "#         fH, ax = plt.subplots(figsize=[2.25,.25])\n",
    "#         ax.axhline(y = np.mean(h), color = 'white', linewidth = .2, alpha = 1, solid_capstyle='butt', zorder = 0)\n",
    "#         pH = ax.bar(n, h, width=.8, bottom=None, align='center', data=None, color=pC['plt'], zorder = 1)\n",
    "#         ax.text(.90, 1, genename,horizontalalignment='left',verticalalignment='top',transform=ax.transAxes, fontsize=7, zorder = 0)\n",
    "#         ax.text(.97, 0.2, (np.around(np.mean(h),decimals=0)),horizontalalignment='left',verticalalignment='top',transform=ax.transAxes, fontsize=6, zorder = 0)\n",
    "#         ax.set_axis_off()\n",
    "#         plt.savefig(\"./resources/{0}.svg\".format(genename), transparent=True, format=\"svg\")\n",
    "#         plt.close(fH)\n",
    "# print(\"Up plots created...\")\n",
    "\n",
    "# for i in -np.add(range(nTop),1):\n",
    "#     genename = pcr['symbol'].iloc[i]\n",
    "#     if os.path.isfile(\"./resources/{0}.svg\".format(genename)) == False:\n",
    "#         n,h = plotBars(genename)\n",
    "#         fH, ax = plt.subplots(figsize=[2.25,.25])\n",
    "#         ax.axhline(y = np.mean(h), color = 'white', linewidth = .2, alpha = 1, solid_capstyle='butt', zorder = 0)\n",
    "#         pH = ax.bar(n, h, width=.8, bottom=None, align='center', data=None, color=pC['plt'], zorder = 1)\n",
    "#         ax.text(.90, 1, genename,horizontalalignment='left',verticalalignment='top',transform=ax.transAxes, fontsize=7, zorder = 0)\n",
    "#         ax.text(.97, 0.2, (np.around(np.mean(h),decimals=0)),horizontalalignment='left',verticalalignment='top',transform=ax.transAxes, fontsize=6, zorder = 0)\n",
    "#         ax.set_axis_off()\n",
    "#         plt.savefig(\"./resources/{0}.svg\".format(genename), transparent=True, format=\"svg\")\n",
    "#         plt.close(fH)\n",
    "# print(\"Down plots created...\")\n",
    "        \n",
    "# fid = open(fname, 'w+t') \n",
    "# fid.write(\"_________________________________________________________\\n\")\n",
    "# fid.write(\"## Top \" + str(nTop) + \" genes enriched in rods \\n\\n\\n\")\n",
    "# for i in range(nTop):\n",
    "#     fid.write(\"#### _\" + pcr['symbol'].iloc[i] + \"_ ![\" + pcr['symbol'].iloc[i] + \"](./resources/\" + pcr['symbol'].iloc[i] + \".svg)\\n\")\n",
    "#     fid.write(\"- _\" + str(pcr['genename'].iloc[i]) + \"_\\n\")\n",
    "#     fid.write(\"- [ ] &#10060;?&#9989;?&#8505;?&#9770;? RANK = \" + str(i+1) + \" ![R](./resources/R.svg)\\n\")\n",
    "#     fid.write(\"- [ZFIN link]()\" + \"\\n\")\n",
    "#     fid.write(\"- NOTESHERE\" + \"\\n\")\n",
    "#     fid.write(\"\\n\")\n",
    "\n",
    "\n",
    "# fid.write(\"_________________________________________________________\\n\")\n",
    "# fid.write(\"## Top \" + str(nTop) + \" genes enriched in cones \\n\\n\\n\")\n",
    "# for i in -np.add(range(nTop),1):\n",
    "#     fid.write(\"#### _\" + pcr['symbol'].iloc[i] + \"_ ![\" + pcr['symbol'].iloc[i] + \"](./resources/\" + pcr['symbol'].iloc[i] + \".svg)\\n\")\n",
    "#     fid.write(\"- _\" + str(pcr['genename'].iloc[i]) + \"_\\n\")\n",
    "#     fid.write(\"- [ ] &#10060;?&#9989;?&#8505;?&#9770;? RANK = \" + str(-i) + \" ![Cs](./resources/Cs.svg)\\n\")\n",
    "#     fid.write(\"- [ZFIN link]()\" + \"\\n\")\n",
    "#     fid.write(\"- NOTESHERE\" + \"\\n\")\n",
    "#     fid.write(\"\\n\")\n",
    "\n",
    "# fid.close()\n",
    "\n",
    "\n",
    "# print(\"MD file created. \")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "help(plt.subplots_adjust)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# histogram of PC weights\n",
    "pcr.sort_values(by=['PC1'])\n",
    "a = pcr['PC1']\n",
    "# a = np.divide(a,pcr['PC1'].sum())\n",
    "a = a[a != 0] #there are too many nil-values\n",
    "fH, ax = plt.subplots(figsize=[12,6])\n",
    "n, h, patches = ax.hist(np.multiply(a,100), bins=100, log=True) #use bins='fd' for optimal binning"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "***\n",
    "## misc. things<a id='misc'>∮</a>\n",
    "***\n",
    "[Back to Index](#Index)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "#### Adding genenames to pcaWeights and using opportunity to sort by PC1 value and remove NaN"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "deseqdir= \"~/Documents/LiMolec/zfRNAseq/20190827/20190827_DESeq2\"\n",
    "# pcr = pd.read_csv(deseqdir + \"/00_rodsVcones/pcaWeights_Rout.csv\")\n",
    "# pcr = pd.read_csv(deseqdir + \"/00_USvsLM/pcaWeights_Rout.csv\")\n",
    "# pcr = pd.read_csv(deseqdir + \"/00_LvsM/pcaWeights_Rout.csv\")\n",
    "pcr = pd.read_csv(deseqdir + \"/00_UvsS/pcaWeights_Rout.csv\")\n",
    "# pcr.head\n",
    "gdf = pd.read_csv(deseqdir + \"/00_rodsVcones/rodVCones_fpkm.csv\")\n",
    "pcr = pcr.set_index('symbol')\n",
    "pcr = pcr.reindex(index=gdf['symbol'])\n",
    "pcr = pcr.reset_index()\n",
    "pcr.insert(1,'genename',gdf['genename'],True)\n",
    "pcr.dropna(subset=['PC1'], inplace=True)\n",
    "pcr.sort_values(by=['PC1'], inplace=True)\n",
    "# pcr.to_csv(path_or_buf=deseqdir + \"/00_rodsVcones/pcaWeights.csv\", index=False)\n",
    "# pcr.to_csv(path_or_buf=deseqdir + \"/00_USvsLM/pcaWeights.csv\", index=False)\n",
    "# pcr.to_csv(path_or_buf=deseqdir + \"/00_LvsM/pcaWeights.csv\", index=False)\n",
    "# pcr.to_csv(path_or_buf=deseqdir + \"/00_UvsS/pcaWeights.csv\", index=False)\n",
    "pcr"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "help(gdf.to_csv)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Trying to draw counts for single gene by wedges in a circle \n",
    "> Prob not worth it, difficult to interpret"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "fig, ax = plt.subplots(figsize=[12,12])\n",
    "\n",
    "n = 30;\n",
    "r = 0;\n",
    "wdg_theta = 360/30;\n",
    "\n",
    "patches = []\n",
    "circles = []\n",
    "\n",
    "pM = matplotlib.patches.Wedge(0, np.mean(data[r,0:6]), wdg_theta * (0), (wdg_theta * (6)), edgecolor=\"none\", facecolor = pC['r'], alpha = 0.4)\n",
    "patches.append(pM)\n",
    "\n",
    "pM = matplotlib.patches.Wedge(0, np.mean(data[r,7:11]), wdg_theta * (6), (wdg_theta * (11)), edgecolor=\"none\", facecolor = pC['u'], alpha = 0.4)\n",
    "patches.append(pM)\n",
    "\n",
    "pM = matplotlib.patches.Wedge(0, np.mean(data[r,12:17]), wdg_theta * (11), (wdg_theta * (17)), edgecolor=\"none\", facecolor = pC['s'], alpha = 0.4)\n",
    "patches.append(pM)\n",
    "\n",
    "pM = matplotlib.patches.Wedge(0, np.mean(data[r,18:24]), wdg_theta * (17), (wdg_theta * (24)), edgecolor=\"none\", facecolor = pC['m'], alpha = 0.4)\n",
    "patches.append(pM)\n",
    "\n",
    "pM = matplotlib.patches.Wedge(0, np.mean(data[r,25:30]), wdg_theta * (24), (wdg_theta * (30)), edgecolor=\"none\", facecolor = pC['l'], alpha = 0.4)\n",
    "patches.append(pM)\n",
    "\n",
    "for i in range(n):\n",
    "    p = matplotlib.patches.Wedge(0, data[r,i], wdg_theta * (i), (wdg_theta * (i+1)) -2, edgecolor=\"none\", facecolor = pC['plt'][i], alpha = 0.4)\n",
    "    patches.append(p)\n",
    "\n",
    "for i in [0.001,.2,.4,.6,.8,1]:\n",
    "    c = matplotlib.patches.Circle((0,0), np.max(data[r,:])*i, edgecolor = '#000000', facecolor='none', alpha=0.59)\n",
    "    circles.append(c)\n",
    "\n",
    "plt.xlim(-np.max(data[r,:])*1, np.max(data[r,:])*1)\n",
    "plt.ylim(-np.max(data[r,:])*1, np.max(data[r,:])*1)\n",
    "\n",
    "\n",
    "circleColl = matplotlib.collections.PatchCollection(circles, match_original=True);\n",
    "ax.add_collection(circleColl)\n",
    "\n",
    "patchColl = matplotlib.collections.PatchCollection(patches, match_original=True);\n",
    "ax.add_collection(patchColl)\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "np.max(data[r,:])"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "np.max(data[4,i])"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "data"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "range(0,4)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "range?"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### From 2019, Whitaker ,..., Swaroop\n",
    "> genes involved in rod synapse size and localization\n",
    "\n",
    "> [link to biorXiv](https://www.biorxiv.org/content/biorxiv/early/2019/08/30/753012.full.pdf)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "plotName = \"test\"\n",
    "\n",
    "# gdf = findgenelist('dpf') # dpf3 -> not this one but dpf2 in all photoreceptors\n",
    "# gdf = findgenelist('epb') # epb4.1l2 -> not this one but some interesting patterns in this family\n",
    "# gdf = findgenelist('grtp') # grtp1 -> this one is interesting!\n",
    "# gdf = findgenelist('kcnj') # kncj14 -> squirrel too but not zebrafish\n",
    "# gdf = gdf.sort_values(by=[\"baseMean\"],ascending=False)\n",
    "\n",
    "# gdf = findgenelist('llgl') # llgl2 -> nope\n",
    "# gdf = findgenelist('rab') # rab28 -> nope; this family is too big\n",
    "# gdf = gdf[gdf.baseMean > 2]\n",
    "# gdf = gdf.sort_values(by=[\"baseMean\"],ascending=False)\n",
    "\n",
    "# gdf = findgenelist('rom1') #rom1 -> rod-specific; what about outer segments?\n",
    "\n",
    "# gdf = findgenelist('arl') # arl2bp -> nope but arl3l2 in all photoreceptors, and arl3l1 downregulated in uv and s. Both confirmed eye specific (https://zfin.org/ZDB-GENE-040426-1678; https://zfin.org/ZDB-GENE-040426-1649)\n",
    "# These genes are supponsed to interact with unc119c (https://www.ncbi.nlm.nih.gov/pubmed/27079236) and act as their GDF\n",
    "# gdf = gdf[gdf.baseMean > 5]\n",
    "# # gdf = gdf[gdf.baseMean < 300]\n",
    "# gdf = gdf.sort_values(by=[\"baseMean\"],ascending=False)\n",
    "\n",
    "# gdf = findgenelist('unc119') # unc119c is photoreceptor specific (https://www.ncbi.nlm.nih.gov/pubmed/27079236) and CoIPs with arl3l2\n",
    "\n",
    "\n",
    "gdf = findgenelist('ppar') #ppargc2\n",
    "# gdf = findgenelist('dpf')\n",
    "# gdf = findgenelist('dpf')\n",
    "# gdf = findgenelist('dpf')\n",
    "# gdf = findgenelist('dpf')\n",
    "# gdf = findgenelist('dpf')\n",
    "# gdf = findgenelist('dpf')\n",
    "# gdf = findgenelist('dpf')\n",
    "# gdf = findgenelist('dpf')\n",
    "# gdf = findgenelist('dpf')\n",
    "# gdf = findgenelist('dpf')\n",
    "# gdf = findgenelist('dpf')\n",
    "# gdf = findgenelist('dpf')\n",
    "# gdf = findgenelist('dpf')\n",
    "\n",
    "### https://www.sciencedirect.com/science/article/pii/S2211124718303218\n",
    "# gdf = findgenelist('frmpd') #frmpd2 interacts with lrit1 in cones \n",
    "# gdf = findgenelist('sdc') #sdcbp interacts with lrit 1\n",
    "# gdf = findgenelist('grip') #grip1 interacts with lrit1\n",
    "\n",
    "# gdf = findgenelist('dpf')\n",
    "# gdf = gdf.sort_values(by=[\"baseMean\"],ascending=False)\n",
    "# gdf = gdf.drop(np.arange(20,93))\n",
    "# gdf = gdf.drop([0])\n",
    "# gdf = gdf.sort_values(by=[\"symbol\"])\n",
    "# gdf = gdf[gdf.baseMean > .45]\n",
    "\n",
    "gdf\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "genenames = gdf['symbol'].values\n",
    "samplenames = [] #skipping names to replace in intaglio\n",
    "log2matrix = np.vectorize(log2)\n",
    "# data = log2matrix(gdf.iloc[0:,7:15].values)\n",
    "data = gdf.iloc[0:,7:37].values\n",
    "\n",
    "\n",
    "fig, ax = plt.subplots(figsize=[12,12])\n",
    "im, cbar = heatmap(data, genenames, samplenames, ax=ax,\n",
    "                   cmap=\"inferno\", cbarlabel=\"fpkm\")\n",
    "# texts = annotate_heatmap(im, valfmt=\"{x:.1f} t\")\n",
    "\n",
    "plt.savefig(\"{0}.svg\".format(plotName), transparent=True, format=\"svg\")\n",
    "plt.show()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# gdf = gdf[gdf.baseMean > .45]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.7.4"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 4
}
