{
 "cells": [
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Load required modules"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 1,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "LOADED\n"
     ]
    }
   ],
   "source": [
    "# External libraries\n",
    "import numpy as np\n",
    "import pandas as pd\n",
    "import pygal\n",
    "import math\n",
    "import IPython\n",
    "import matplotlib\n",
    "import matplotlib.pyplot as plt\n",
    "import matplotlib.font_manager as font_manager\n",
    "import importlib\n",
    "import svgutils\n",
    "from svgutils.compose import *\n",
    "#import svgwrite\n",
    "\n",
    "# Custom code\n",
    "from fx_RNAseq import *\n",
    "def findgene(genename):\n",
    "    dPath = \"~/Documents/LiMolec/zfRNAseq/20190827/20190827_DESeq2/\" #directory path\n",
    "#     fPath = \"00_rodsVcones/rodVCones_raw.csv\" #file path\n",
    "    fPath = \"00_rodsVcones/rodVCones_fpkm.csv\" #file path\n",
    "    tempPath = \"temp.txt\" # temporary file to store values\n",
    "    !head -1 {dPath + fPath} > {dPath + tempPath}\n",
    "    # EXACT MATCH ONLY by matching word (-w) and only at the start of each line (^) which corresponds to the symbol\n",
    "    !grep -w ^$genename {dPath + fPath} >> {dPath + tempPath}\n",
    "    gdf = pd.read_csv(dPath + tempPath)\n",
    "    return gdf\n",
    "\n",
    "def findgenelist(genename):\n",
    "    dPath = \"~/Documents/LiMolec/zfRNAseq/20190827/20190827_DESeq2/\" #directory path\n",
    "#     fPath = \"00_rodsVcones/rodVCones_raw.csv\" #file path\n",
    "    fPath = \"00_rodsVcones/rodVCones_fpkm.csv\" #file path\n",
    "    tempPath = \"temp.txt\" # temporary file to store values\n",
    "    !head -1 {dPath + fPath} > {dPath + tempPath}\n",
    "    # EXACT MATCH ONLY by matching word (-w) and only at the start of each line (^) which corresponds to the symbol\n",
    "    !grep ^$genename {dPath + fPath} >> {dPath + tempPath}\n",
    "    gdf = pd.read_csv(dPath + tempPath)\n",
    "    return gdf\n",
    "\n",
    "# def findgene(genename):\n",
    "#     !head -1 ~/Dropbox/LiLab/RNAseq/20190827_DESeq2/00_rodsVcones/rodVCones_raw.csv > ~/Dropbox/LiLab/RNAseq/20190827_DESeq2/temp.txt\n",
    "#     # EXACT MATCH ONLY by matching word (-w) and only at the start of each line (^) which corresponds to the symbol\n",
    "#     !grep -w ^$genename ~/Dropbox/LiLab/RNAseq/20190827_DESeq2/00_rodsVcones/rodVCones_raw.csv >> ~/Dropbox/LiLab/RNAseq/20190827_DESeq2/temp.txt\n",
    "#     gdf = pd.read_csv(\"~/Dropbox/LiLab/RNAseq/20190827_DESeq2/temp.txt\")\n",
    "    \n",
    "#     return gdf\n",
    "\n",
    "# def findgenelist(genename):\n",
    "#     !head -1 ~/Dropbox/LiLab/RNAseq/20190827_DESeq2/00_rodsVcones/rodVCones_raw.csv > ~/Dropbox/LiLab/RNAseq/20190827_DESeq2/temp.txt\n",
    "#     # EXACT MATCH ONLY by matching word (-w) and only at the start of each line (^) which corresponds to the symbol\n",
    "#     !grep ^$genename ~/Dropbox/LiLab/RNAseq/20190827_DESeq2/00_rodsVcones/rodVCones_raw.csv >> ~/Dropbox/LiLab/RNAseq/20190827_DESeq2/temp.txt\n",
    "#     gdf = pd.read_csv(\"~/Dropbox/LiLab/RNAseq/20190827_DESeq2/temp.txt\")\n",
    "#     return gdf\n",
    "\n",
    "def plotNReads():\n",
    "    rdf = pd.read_csv(\"~/Documents/LiMolec/zfRNAseq/20190827/20190827_DESeq2/nMappedReads.csv\")\n",
    "    h = rdf.iloc[:,1]\n",
    "    h = h / 1e6\n",
    "    n = np.arange(1,7) # Rods\n",
    "    n = np.append(n, 6.5 + np.arange(1,6)) # UV\n",
    "    n = np.append(n, 12 + np.arange(1,7)) # S\n",
    "    n = np.append(n, 18.5 + np.arange(1,8)) # M\n",
    "    n = np.append(n, 26 + np.arange(1,7)) # L\n",
    "    return n, h\n",
    "\n",
    "def plotBars(genename):\n",
    "    gdf = findgene(genename)\n",
    "    n = np.arange(1,7) # Rods\n",
    "    n = np.append(n, 6.5 + np.arange(1,6)) # UV\n",
    "    n = np.append(n, 12 + np.arange(1,7)) # S\n",
    "    n = np.append(n, 18.5 + np.arange(1,8)) # M\n",
    "    n = np.append(n, 26 + np.arange(1,7)) # L\n",
    "    h_start = 7\n",
    "    h_end = 37\n",
    "    h = gdf.iloc[0,h_start:h_end]\n",
    "    return n, h\n",
    "\n",
    "def plotBars_list(genelist):\n",
    "    n=[]\n",
    "    h=[]\n",
    "    for i,gene in enumerate(genelist, start=0):\n",
    "        tempn, temph = plotBars(gene)\n",
    "        n = np.append(n,tempn+(30*i))\n",
    "        h = np.append(h,temph)\n",
    "    return n, h\n",
    "\n",
    "def svgreload():\n",
    "    !osascript -e 'tell application \"Google Chrome\" to tell the active tab of second window to reload'\n",
    "\n",
    "from pygal.style import Style\n",
    "custom_style = Style(\n",
    "    background = 'black',\n",
    "    plot_background = '#111111',\n",
    "    foreground = '#999999',\n",
    "    foreground_strong = '#eeeeee',\n",
    "    foreground_subtle = '#555555',\n",
    "    opacity = '.8',\n",
    "    opacity_hover = '.4',\n",
    "    transition = '10ms',\n",
    "    colors = ('#747474','#B540B7', '#4669F2','#04CD22','#CC2C2A'),\n",
    "    label_font_size = 24,\n",
    "    major_label_font_size = 24,\n",
    "    title_font_size = 24,\n",
    "    legend_font_size = 18\n",
    ")\n",
    "\n",
    "#photoreceptor Colors\n",
    "pC = {\n",
    "    'r' : '#747474',\n",
    "    'u' : '#B540B7',\n",
    "    's' : '#4669F2',\n",
    "    'm' : '#04CD22',\n",
    "    'l' : '#CC2C2A',\n",
    "    'plt' : '',\n",
    "}\n",
    "\n",
    "pC['plt']=[\n",
    "    pC['r'],pC['r'],pC['r'],pC['r'],pC['r'],pC['r'],\n",
    "    pC['u'],pC['u'],pC['u'],pC['u'],pC['u'],\n",
    "    pC['s'],pC['s'],pC['s'],pC['s'],pC['s'],pC['s'],\n",
    "    pC['m'],pC['m'],pC['m'],pC['m'],pC['m'],pC['m'],pC['m'],\n",
    "    pC['l'],pC['l'],pC['l'],pC['l'],pC['l'],pC['l']\n",
    "]\n",
    "\n",
    "def findgene_glia(genename):\n",
    "    dPath = \"~/Documents/LiLab/RNAseq/zfMullerGlia_2018Charlton/\" #directory path\n",
    "#     fPath = \"glia_23615_sup-0006-tables2.csv\" #file path\n",
    "    fPath = \"glia8dpf_DEGcounts.csv\" #file path\n",
    "    tempPath = \"temp.txt\" # temporary file to store values\n",
    "    !head -1 {dPath + fPath} > {dPath + tempPath}\n",
    "    # EXACT MATCH ONLY by matching word (-w) and only at the start of each line (^) which corresponds to the symbol\n",
    "    !grep -w ^$genename {dPath + fPath} >> {dPath + tempPath}\n",
    "    gdf = pd.read_csv(dPath + tempPath)\n",
    "    return gdf\n",
    "\n",
    "def findgenelist_glia(genename):\n",
    "    dPath = \"~/Documents/LiLab/RNAseq/zfMullerGlia_2018Charlton/\" #directory path\n",
    "#     fPath = \"glia_23615_sup-0006-tables2.csv\" #file path\n",
    "    fPath = \"glia8dpf_DEGcounts.csv\" #file path\n",
    "    tempPath = \"temp.txt\" # temporary file to store values\n",
    "    !head -1 {dPath + fPath} > {dPath + tempPath}\n",
    "    # EXACT MATCH ONLY by matching word (-w) and only at the start of each line (^) which corresponds to the symbol\n",
    "    !grep ^$genename {dPath + fPath} >> {dPath + tempPath}\n",
    "    gdf = pd.read_csv(dPath + tempPath)\n",
    "    return gdf\n",
    "\n",
    "def plotBars_glia(genename):\n",
    "    gdf = findgene_glia(genename)\n",
    "    n = np.arange(1,19) # GFP+\n",
    "    n = np.append(n, 19.5 + np.arange(1,19)) # GFP-\n",
    "    h_start = 5\n",
    "    h_end = 41\n",
    "#     h = np.power(2,(gdf.iloc[0,h_start:h_end]))\n",
    "    h = gdf.iloc[0,h_start:h_end]\n",
    "    return n, h\n",
    "\n",
    "#glia Colors\n",
    "gC = {\n",
    "    'm48' : '#c7ffec',\n",
    "    'm60' : '#94ffdb',\n",
    "    'm72' : '#57ffc7',\n",
    "    'm96' : '#2effb9',\n",
    "    'm120': '#05ffac',\n",
    "    'm192': '#00d68f',\n",
    "    \n",
    "    'c48' : '#e3e3e3',\n",
    "    'c60' : '#c9c9c9',\n",
    "    'c72' : '#ababab',\n",
    "    'c96' : '#969696',\n",
    "    'c120': '#828282',\n",
    "    'c192': '#6b6b6b',\n",
    "    \n",
    "    'plt' : '',\n",
    "}\n",
    "\n",
    "gC['plt']=[\n",
    "    gC['m48'],gC['m48'],gC['m48'],\n",
    "    gC['m60'],gC['m60'],gC['m60'],\n",
    "    gC['m72'],gC['m72'],gC['m72'],\n",
    "    gC['m96'],gC['m96'],gC['m96'],\n",
    "    gC['m120'],gC['m120'],gC['m120'],\n",
    "    gC['m192'],gC['m192'],gC['m192'],\n",
    "    gC['c48'],gC['c48'],gC['c48'],\n",
    "    gC['c60'],gC['c60'],gC['c60'],\n",
    "    gC['c72'],gC['c72'],gC['c72'],\n",
    "    gC['c96'],gC['c96'],gC['c96'],\n",
    "    gC['c120'],gC['c120'],gC['c120'],\n",
    "    gC['c192'],gC['c192'],gC['c192'],\n",
    "]\n",
    "\n",
    "def findgene_rods(genename):\n",
    "    dPath = \"~/Documents/LiLab/RNAseq/zfRods_Sun2018/\" #directory path\n",
    "#     fPath = \"glia_23615_sup-0006-tables2.csv\" #file path\n",
    "    fPath = \"zfRods.csv\" #file path\n",
    "    tempPath = \"temp.txt\" # temporary file to store values\n",
    "    !head -1 {dPath + fPath} > {dPath + tempPath}\n",
    "    # EXACT MATCH ONLY by matching word (-w) and only at the start of each line (^) which corresponds to the symbol\n",
    "    !grep -w ^$genename {dPath + fPath} >> {dPath + tempPath}\n",
    "    gdf = pd.read_csv(dPath + tempPath)\n",
    "    return gdf\n",
    "\n",
    "def findgenelist_rods(genename):\n",
    "    dPath = \"~/Documents/LiLab/RNAseq/zfRods_Sun2018/\" #directory path\n",
    "#     fPath = \"glia_23615_sup-0006-tables2.csv\" #file path\n",
    "    fPath = \"zfRods.csv\" #file path\n",
    "    tempPath = \"temp.txt\" # temporary file to store values\n",
    "    !head -1 {dPath + fPath} > {dPath + tempPath}\n",
    "    # EXACT MATCH ONLY by matching word (-w) and only at the start of each line (^) which corresponds to the symbol\n",
    "    !grep ^$genename {dPath + fPath} >> {dPath + tempPath}\n",
    "    gdf = pd.read_csv(dPath + tempPath)\n",
    "    return gdf\n",
    "\n",
    "def plotBars_rods(genename):\n",
    "    gdf = findgene_rods(genename)\n",
    "    n = np.arange(1,5) # GFP+\n",
    "    n = np.append(n, 4.5 + np.arange(1,5)) # GFP-\n",
    "    h_start = 7\n",
    "    h_end = 15\n",
    "#     h = np.power(2,(gdf.iloc[0,h_start:h_end]))\n",
    "    h = gdf.iloc[0,h_start:h_end].values\n",
    "    return n, h\n",
    "\n",
    "#rod Colors\n",
    "rC = {\n",
    "    'r' : '#747474',\n",
    "    'c': '#dac910',\n",
    "    'plt' : '',\n",
    "}\n",
    "\n",
    "rC['plt']=[\n",
    "    rC['r'],rC['r'],rC['r'],rC['r'],\n",
    "    rC['c'],rC['c'],rC['c'],rC['c'],\n",
    "]\n",
    "\n",
    "print('LOADED')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "/Users/angueyraaristjm/Documents/LiMolec/zfRNAseq/20190827/20190827_DESeq2/genePlots/fpkm\n"
     ]
    }
   ],
   "source": [
    "# %cd ~/Dropbox/Presentations/RevealReboot/resources/20191002_RNAseq/\n",
    "%cd ~/Documents/LiMolec/zfRNAseq/20190827/20190827_DESeq2/genePlots/fpkm/"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "***\n",
    "# Index <a id='Index'>∮</a>\n",
    "***\n",
    "- [Plot Total Number of Reads](#nReads)\n",
    "- [PCA Plots](#pcaPlots)\n",
    "- [BarPlot: Opsins](#barOpsins)\n",
    "- [BarPlot: Custom](#barCustom)\n",
    "- [BarPlot: Many Genes](#barMany)\n",
    "- [BarPlot_subPlotsAutomatic: Custom](#barManyAuto)\n",
    "- [Heatmap: Opsins](#heatOpsins)\n",
    "- [Heatmap: Custom](#heatCustom)\n",
    "- [Heatmap: Other Data Sets](#heatOther)\n",
    "- [GO clustering](https://go.princeton.edu/) \n",
    "***"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "***\n",
    "## Total number of reads<a id='nReads'>∮</a>\n",
    "***\n",
    "[Back to Index](#Index)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### pygal version (does not import well into Intaglio)\n",
    "> hist = pygal.Histogram(x_title='',y_title='million of mapped reads',show_x_labels=False,title='',style=custom_style,legend_at_bottom=True,legend_at_bottom_columns=4,show_legend=False)  \n",
    "> hist.add('Rod',[(11.27,0.2,0.8),(11.77,1.2,1.8),(9.4,2.2,2.8),(14.99,3.2,3.8),(9.75,4.2,4.8),(9.08,5.2,5.8),])  \n",
    "> hist.add('UV',[(9.42,6.2,6.8),(6.52,7.2,7.8),(10.45,8.2,8.8),(9.46,9.2,9.8),(11.02,10.2,10.8)])  \n",
    "> hist.add('S',[(9.23,11.2,11.8),(9.33,12.2,12.8),(9.37,13.2,13.8),(14.95,14.2,14.8),(12.85,15.2,15.8),(10.96,16.2,16.8)])  \n",
    "> hist.add('M',[(10.40,18.2,18.8),(8.634,19.2,19.8),(7.36,20.2,20.8),(10.32,21.2,21.8),(8.33,22.2,22.8),(9.53,23.2,23.8),(9.526,17.2,17.8)])  \n",
    "> hist.add('L',[(8.99,24.2,24.8),(10.50,25.2,25.8),(10.03,26.2,26.8),(11.08,27.2,27.8),(10.44,28.2,28.8),(10.72,29.2,29.8)])  \n",
    "> hist.render_to_file('test.svg')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# dark background\n",
    "params = {\"ytick.color\" : \"w\",\n",
    "          \"xtick.color\" : \"w\",\n",
    "          \"axes.labelcolor\" : \"w\",\n",
    "          \"axes.edgecolor\" : \"w\",\n",
    "         \"axes.linewidth\" : 3,\n",
    "         \"xtick.major.width\" : 3,\n",
    "         \"ytick.major.width\" : 3,\n",
    "         \"xtick.major.size\" : 8,\n",
    "         \"ytick.major.size\" : 8,\n",
    "         \"text.color\" : \"w\"}\n",
    "plt.rcParams.update(params)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# white background\n",
    "params = {\"ytick.color\" : \"k\",\n",
    "          \"xtick.color\" : \"k\",\n",
    "          \"axes.labelcolor\" : \"k\",\n",
    "          \"axes.edgecolor\" : \"k\",\n",
    "         \"axes.linewidth\" : 3,\n",
    "         \"xtick.major.width\" : 3,\n",
    "         \"ytick.major.width\" : 3,\n",
    "         \"xtick.major.size\" : 8,\n",
    "         \"ytick.major.size\" : 8,\n",
    "         \"text.color\" : \"w\"}\n",
    "plt.rcParams.update(params)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "#pyplot version (Sep_2019)\n",
    "plotname = ''\n",
    "n,h = plotNReads()\n",
    "fH, ax = plt.subplots()\n",
    "pH = plt.bar(n, h, width=0.8, bottom=None, align='center', data=None, color=pC['plt'])\n",
    "formatFigure_nReads(plotname, fH, ax, pH)\n",
    "plt.tight_layout()\n",
    "plt.savefig(\"{0}.svg\".format('nReads'), transparent=True, format=\"svg\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "np.mean(h)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": []
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "***\n",
    "## pcaPlots<a id='pcaPlots'>∮</a>\n",
    "***\n",
    "[Back to Index](#Index)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "def formatFigure_pca(plotname, figH, axH, plotH):\n",
    "    font_path = '/System/Library/Fonts/Avenir.ttc'\n",
    "    fontTicks = font_manager.FontProperties(fname=font_path, size=18)\n",
    "    fontLabels = font_manager.FontProperties(fname=font_path, size=22)\n",
    "    fontTitle = font_manager.FontProperties(fname=font_path, size=28)\n",
    "    \n",
    "    axH.ticklabel_format(style='sci',axis='y',scilimits=(0,4))\n",
    "    axH.ticklabel_format(style='sci',axis='x',scilimits=(0,4))\n",
    "    axH.set_title(plotname, fontproperties=fontTitle)\n",
    "    axH.spines['top'].set_visible(False)\n",
    "    axH.spines['right'].set_visible(False)\n",
    "    \n",
    "    for label in (axH.get_xticklabels() + axH.get_yticklabels()):\n",
    "        label.set_fontproperties(fontTicks)\n",
    "    axH.tick_params(axis = 'both', which = 'major', labelsize = 18)\n",
    "    axH.xaxis.offsetText.set_fontsize(18)\n",
    "    axH.yaxis.offsetText.set_fontsize(18)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "#pyplot version (Sep_2019)\n",
    "plotname = 'PCA_RodsCones'\n",
    "pc_df = pd.read_csv(\"~/Documents/LiMolec/zfRNAseq/20190827/20190827_DESeq2/00_rodsVcones/pcaData.csv\")\n",
    "pc1 = pc_df.iloc[:,0]\n",
    "pc2 = pc_df.iloc[:,1]\n",
    "fH, ax = plt.subplots(figsize=[5,5])\n",
    "pH = plt.scatter(pc1, pc2, color=pC['plt'])\n",
    "formatFigure_pca(plotname, fH, ax, pH)\n",
    "ax.set_xlabel('PC1 (var = 21%)', fontproperties=font_manager.FontProperties(fname='/System/Library/Fonts/Avenir.ttc', size=22))\n",
    "ax.set_ylabel('PC2 (var = 9%)', fontproperties=font_manager.FontProperties(fname='/System/Library/Fonts/Avenir.ttc', size=22))\n",
    "plt.xlim(-100,50)\n",
    "plt.ylim(-60,60)\n",
    "plt.tight_layout()\n",
    "plt.savefig(\"{0}.svg\".format(plotname), transparent=True, format=\"svg\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "#pyplot version (Sep_2019)\n",
    "plotname = 'PCA_UVS_ML'\n",
    "pc_df = pd.read_csv(\"~/Documents/LiMolec/zfRNAseq/20190827/20190827_DESeq2/00_USvsLM/pcaData.csv\")\n",
    "pc1 = pc_df.iloc[:,0]\n",
    "pc2 = pc_df.iloc[:,1]\n",
    "fH, ax = plt.subplots(figsize=[5,5])\n",
    "pH = plt.scatter(pc1, pc2, color=[\n",
    "    pC['u'],pC['u'],pC['u'],pC['u'],pC['u'],\n",
    "    pC['s'],pC['s'],pC['s'],pC['s'],pC['s'],pC['s'],\n",
    "    pC['m'],pC['m'],pC['m'],pC['m'],pC['m'],pC['m'],pC['m'],\n",
    "    pC['l'],pC['l'],pC['l'],pC['l'],pC['l'],pC['l']\n",
    "]\n",
    "                )\n",
    "formatFigure_pca(plotname, fH, ax, pH)\n",
    "ax.set_xlabel('PC1 (var = 13%)', fontproperties=font_manager.FontProperties(fname='/System/Library/Fonts/Avenir.ttc', size=22))\n",
    "ax.set_ylabel('PC2 (var = 10%)', fontproperties=font_manager.FontProperties(fname='/System/Library/Fonts/Avenir.ttc', size=22))\n",
    "plt.xlim(-80,80)\n",
    "plt.ylim(-60,60)\n",
    "plt.tight_layout()\n",
    "plt.savefig(\"{0}.svg\".format(plotname), transparent=True, format=\"svg\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "#pyplot version (Sep_2019)\n",
    "plotname = 'PCA_ML'\n",
    "pc_df = pd.read_csv(\"~/Documents/LiMolec/zfRNAseq/20190827/20190827_DESeq2/00_LvsM/pcaData.csv\")\n",
    "pc1 = pc_df.iloc[:,0]\n",
    "pc2 = pc_df.iloc[:,1]\n",
    "fH, ax = plt.subplots(figsize=[5,5])\n",
    "pH = plt.scatter(pc1, pc2, color=[\n",
    "    pC['m'],pC['m'],pC['m'],pC['m'],pC['m'],pC['m'],pC['m'],\n",
    "    pC['l'],pC['l'],pC['l'],pC['l'],pC['l'],pC['l']\n",
    "]\n",
    "                )\n",
    "formatFigure_pca(plotname, fH, ax, pH)\n",
    "ax.set_xlabel('PC1 (var = 23%)', fontproperties=font_manager.FontProperties(fname='/System/Library/Fonts/Avenir.ttc', size=22))\n",
    "ax.set_ylabel('PC2 (var = 10%)', fontproperties=font_manager.FontProperties(fname='/System/Library/Fonts/Avenir.ttc', size=22))\n",
    "plt.xlim(-80,80)\n",
    "plt.ylim(-80,80)\n",
    "plt.tight_layout()\n",
    "plt.savefig(\"{0}.svg\".format(plotname), transparent=True, format=\"svg\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "#pyplot version (Sep_2019)\n",
    "plotname = 'PCA_U_S'\n",
    "pc_df = pd.read_csv(\"~/Documents/LiMolec/zfRNAseq/20190827/20190827_DESeq2/00_UvsS/pcaData.csv\")\n",
    "pc1 = pc_df.iloc[:,0]\n",
    "pc2 = pc_df.iloc[:,1]\n",
    "fH, ax = plt.subplots(figsize=[5,5])\n",
    "pH = plt.scatter(pc1, pc2, color=[\n",
    "    pC['u'],pC['u'],pC['u'],pC['u'],pC['u'],\n",
    "    pC['s'],pC['s'],pC['s'],pC['s'],pC['s'],pC['s'],\n",
    "]\n",
    "                )\n",
    "formatFigure_pca(plotname, fH, ax, pH)\n",
    "ax.set_xlabel('PC1 (var = 13%)', fontproperties=font_manager.FontProperties(fname='/System/Library/Fonts/Avenir.ttc', size=22))\n",
    "ax.set_ylabel('PC2 (var = 10%)', fontproperties=font_manager.FontProperties(fname='/System/Library/Fonts/Avenir.ttc', size=22))\n",
    "plt.xlim(-100,100)\n",
    "plt.ylim(-80,80)\n",
    "plt.tight_layout()\n",
    "plt.savefig(\"{0}.svg\".format(plotname), transparent=True, format=\"svg\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# US vs LM: PC1 = 13% and PC2 = 10%\n",
    "# M vs L: PC1 = 23% and PC2 = 10%\n",
    "# UV vs S: PC1 = 15% and PC2 = 13%"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### tSNE plot"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "#pyplot version (Sep_2019)\n",
    "plotname = 'tSNE'\n",
    "pc_df = pd.read_csv(\"~/Documents/LiMolec/zfRNAseq/20190827/20190827_DESeq2/tSNEData.csv\")\n",
    "pc1 = pc_df.iloc[:,0]\n",
    "pc2 = pc_df.iloc[:,1]\n",
    "fH, ax = plt.subplots(figsize=[6,6])\n",
    "pH = plt.scatter(pc1, pc2, color=pC['plt'])\n",
    "formatFigure_pca(plotname, fH, ax, pH)\n",
    "ax.set_xlabel('tSNE1', fontproperties=font_manager.FontProperties(fname='/System/Library/Fonts/Avenir.ttc', size=22))\n",
    "ax.set_ylabel('tSNE2', fontproperties=font_manager.FontProperties(fname='/System/Library/Fonts/Avenir.ttc', size=22))\n",
    "ax.set_xticks([-1000,-500,0,500,1000])\n",
    "ax.set_yticks([-1000,-500,0,500,1000])\n",
    "plt.xlim(-1000,1000)\n",
    "plt.ylim(-1000,1000)\n",
    "plt.tight_layout()\n",
    "plt.savefig(\"{0}.svg\".format(plotname), transparent=True, format=\"svg\")"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "***\n",
    "## barPlot: Opsins<a id='barOpsins'>∮</a>\n",
    "***\n",
    "[Back to Index](#Index)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# from fx_RNAseq import *"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "genename = 'opn1lw1'\n",
    "n,h = plotBars(genename)\n",
    "h = h\n",
    "fH, ax = plt.subplots()\n",
    "pH = plt.bar(n, h, width=0.8, bottom=None, align='center', data=None, color=pC['plt'])\n",
    "formatFigure_Opsins(genename, fH, ax, pH)\n",
    "plt.savefig(\"{0}.svg\".format(genename), transparent=True, format=\"svg\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "findgene('rho')"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "***\n",
    "## barPlot: Custom Gene<a id='barCustom'>∮</a>\n",
    "***\n",
    "[Back to Index](#Index)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "> Genes to pay attention to but still haven't given interesting results\n",
    "> cdh (cadherins)\n",
    "> protocadherins\n",
    "> semaphorins\n",
    "> plexins (plx)\n",
    "> laminins (lamb) and integrins (itg)\n",
    "> dystroglycan (dag), dystrophin (dmd), pikachurin\n",
    "> neurexins\n",
    "> cadm\n",
    "> dscam and sidekicks (sdk)\n",
    "> pappaa\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "gdf = findgenelist('lrit')\n",
    "gdf"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "genename = 'prickle3'\n",
    "# genename = 'ncdn'\n",
    "# genename = 'chn2'\n",
    "# genename = 'grin1b'\n",
    "# genename = 'cdhr1a'\n",
    "# genename = 'lrit1a'\n",
    "n,h = plotBars(genename)\n",
    "fH, ax = plt.subplots()\n",
    "pH = plt.bar(n, h, width=0.8, bottom=None, align='center', data=None, color=pC['plt'])\n",
    "formatFigure(genename, fH, ax, pH)\n",
    "plt.savefig(\"{0}.svg\".format(genename), transparent=True, format=\"svg\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "***\n",
    "## barPlot: Many Genes Layout<a id='barMany'>∮</a>\n",
    "***\n",
    "[Back to Index](#Index)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "cd '~/Dropbox/LiLab/RNAseq/genePlots/combos'"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "genelist = ['gnat1','gnat2']\n",
    "n, h = plotBars_list(genelist)\n",
    "fH, ax = plt.subplots()\n",
    "pH = plt.bar(n, h, width=0.8, bottom=None, align='center', data=None, color=pC['plt'])\n",
    "formatFigure_list(genelist, fH, ax, pH)\n",
    "# plt.savefig(\"{0}.svg\".format('gnat'), transparent=True, format=\"svg\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "genelist = ['rho','opn1sw1','opn1sw2','opn1mw1','opn1mw2','opn1mw3','opn1mw4','opn1lw1','opn1lw2']\n",
    "n, h = plotBars_list(genelist)\n",
    "fH, ax = plt.subplots(figsize=[6*len(genelist),4])\n",
    "pH = plt.bar(n, h, width=0.8, bottom=None, align='center', data=None, color=pC['plt'])\n",
    "formatFigure_list(genelist, fH, ax, pH)\n",
    "plt.xlim(0,len(genelist)*10)\n",
    "# plt.savefig(\"{0}.svg\".format('opn1'), transparent=True, format=\"svg\")"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "***\n",
    "## barPlot: Many Genes Layout Using Subplots (proper way)<a id='barManySubPlots'>∮</a>\n",
    "***\n",
    "[Back to Index](#Index)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "shrink = 1#2/3"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "genelist = ['gnat1','gnat2']\n",
    "fH, axes = plt.subplots(1,len(genelist),sharey=True,figsize=[6*shrink*len(genelist),4*shrink])\n",
    "for i,g in enumerate(genelist,start=0):\n",
    "    n, h = plotBars(g)\n",
    "    plt.sca(axes[i]) \n",
    "    pH = plt.bar(n, h, width=0.8, bottom=None, align='center', data=None,\n",
    "                 color=pC['plt'])\n",
    "    if  i == 0:\n",
    "        formatFigure(g, fH, axes[i], pH)\n",
    "    else:\n",
    "        formatFigure_right(g, fH, axes[i], pH)\n",
    "plt.savefig(\"{0}.svg\".format('gnat'), transparent=True, format=\"svg\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "genelist = ['rho','rhol']\n",
    "\n",
    "fH, axes = plt.subplots(1,len(genelist),sharey=True,figsize=[6*shrink*len(genelist),4*shrink])\n",
    "for i,g in enumerate(genelist,start=0):\n",
    "    n, h = plotBars(g)\n",
    "    plt.sca(axes[i])\n",
    "    pH = plt.bar(n, h, width=0.8, bottom=None, align='center', data=None,\n",
    "                 color=pC['plt'])\n",
    "    if  i == 0:\n",
    "        formatFigure(g, fH, axes[i], pH)\n",
    "#         plt.ylim(0,3.5e6)\n",
    "#         axes[0].set_yticks(np.arange(0, 4.0e6, step=0.5e6))\n",
    "    else:\n",
    "        formatFigure_right(g, fH, axes[i], pH)\n",
    "plt.savefig(\"{0}.svg\".format('rho'), transparent=True, format=\"svg\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "genelist = ['opn1sw1','opn1sw1']\n",
    "\n",
    "fH, axes = plt.subplots(1,len(genelist),sharey=True,figsize=[6*shrink*len(genelist),4*shrink])\n",
    "for i,g in enumerate(genelist,start=0):\n",
    "    n, h = plotBars(g)\n",
    "    plt.sca(axes[i])\n",
    "    pH = plt.bar(n, h, width=0.8, bottom=None, align='center', data=None,\n",
    "                 color=pC['plt'])\n",
    "    if  i == 0:\n",
    "        formatFigure(g, fH, axes[i], pH)\n",
    "#         plt.ylim(0,2e6)\n",
    "#         axes[0].set_yticks(np.arange(0, 2.5e6, step=0.5e6))\n",
    "    else:\n",
    "        formatFigure_right(g, fH, axes[i], pH)\n",
    "plt.savefig(\"{0}.svg\".format('opn1sw1'), transparent=True, format=\"svg\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "genelist = ['opn1mw1','opn1mw2']\n",
    "\n",
    "fH, axes = plt.subplots(1,len(genelist),sharey=True,figsize=[6*shrink*len(genelist),4*shrink])\n",
    "for i,g in enumerate(genelist,start=0):\n",
    "    n, h = plotBars(g)\n",
    "    plt.sca(axes[i])\n",
    "    pH = plt.bar(n, h, width=0.8, bottom=None, align='center', data=None,\n",
    "                 color=pC['plt'])\n",
    "    if  i == 0:\n",
    "        formatFigure(g, fH, axes[i], pH)\n",
    "    else:\n",
    "        formatFigure_right(g, fH, axes[i], pH)\n",
    "plt.savefig(\"{0}.svg\".format('opn1mwA'), transparent=True, format=\"svg\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "genelist = ['opn1mw3','opn1mw4']\n",
    "\n",
    "fH, axes = plt.subplots(1,len(genelist),sharey=True,figsize=[6*shrink*len(genelist),4*shrink])\n",
    "for i,g in enumerate(genelist,start=0):\n",
    "    n, h = plotBars(g)\n",
    "    plt.sca(axes[i])\n",
    "    pH = plt.bar(n, h, width=0.8, bottom=None, align='center', data=None,\n",
    "                 color=pC['plt'])\n",
    "    if  i == 0:\n",
    "        formatFigure(g, fH, axes[i], pH)\n",
    "        plt.ylim(0,1.0e5)\n",
    "        axes[0].set_yticks(np.arange(0, 1.5e5, step=0.5e5))\n",
    "    else:\n",
    "        formatFigure_right(g, fH, axes[i], pH)\n",
    "\n",
    "\n",
    "plt.savefig(\"{0}.svg\".format('opn1mwB'), transparent=True, format=\"svg\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "genelist = ['opn1lw1','opn1lw2']\n",
    "\n",
    "fH, axes = plt.subplots(1,len(genelist),sharey=True,figsize=[6*shrink*len(genelist),4*shrink])\n",
    "for i,g in enumerate(genelist,start=0):\n",
    "    n, h = plotBars(g)\n",
    "    plt.sca(axes[i])\n",
    "    pH = plt.bar(n, h, width=0.8, bottom=None, align='center', data=None,\n",
    "                 color=pC['plt'])\n",
    "    if  i == 0:\n",
    "        formatFigure(g, fH, axes[i], pH)\n",
    "        plt.ylim(0,1.0e5)\n",
    "        axes[0].set_yticks(np.arange(0, 1.5e5, step=0.5e5))\n",
    "    else:\n",
    "        formatFigure_right(g, fH, axes[i], pH)\n",
    "#         axes[i].set_yticks(np.arange(0, 2.0e6, step=0.5e6))\n",
    "\n",
    "plt.savefig(\"{0}.svg\".format('opn1lw'), transparent=True, format=\"svg\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Using svgutils to automatically tile these figures the right way\n",
    "Figure(\"2400\", \"2000\",\n",
    "       SVG(\"rho.svg\").move(0, 290*0),\n",
    "       SVG(\"opn1sw.svg\").move(0, 290*1),\n",
    "       SVG(\"opn1mwA.svg\").move(0, 290*2),\n",
    "       SVG(\"opn1mwB.svg\").move(0, 290*3),\n",
    "       SVG(\"opn1lw.svg\").move(0, 290*4)\n",
    "        ).save(\"OpsinsAll.svg\")\n",
    "svgreload()"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Phototransduction"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "gdf = findgenelist('gngt')\n",
    "gdf = gdf.sort_values(by=['symbol'])\n",
    "gdf"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# genelist = ['saga','sagb','arr3a','arr3b']\n",
    "# plotName = 'arrestins'\n",
    "\n",
    "# genelist = ['gnat1','gnat2']\n",
    "# plotName = 'gnat'\n",
    "\n",
    "# genelist = ['gngt1','gngt2a','gngt2b']\n",
    "# plotName = 'gngt'\n",
    "\n",
    "# genelist = ['gnb1a','gnb1b','gnb3a','gnb3b','gnb4b','gnb5a','gnb5b']\n",
    "# plotName = 'gnb'\n",
    "\n",
    "# genelist = ['pde6a','pde6b','pde6c','pde6d','pde6ga','pde6gb','pde6c','pde6ha']\n",
    "# plotName = 'pde'\n",
    "\n",
    "# genelist = ['gucy1a1','gucy1b1','gucy1b2','gucy2c','gucy2d','gucy2f','gucy2g']\n",
    "# plotName = 'GC'\n",
    "\n",
    "# genelist = ['gucy2d','gucy2f']\n",
    "# plotName = 'GC_retinaSpecific'\n",
    "\n",
    "# genelist = ['guca1a','guca1b','guca1c','guca1d','guca1e','guca1g']\n",
    "# plotName = 'GCAP'\n",
    "\n",
    "# genelist = ['cnga1a','cnga1b','cnga3a','cnga3b','cnga4','cnga2a','cnga2b']\n",
    "# plotName = 'CNGalpha'\n",
    "\n",
    "# genelist = ['cnga1a','cnga1b','cnga3a','cnga3b']\n",
    "# plotName = 'CNGalpha_retinaSpecific'\n",
    "\n",
    "# genelist = ['cngb1a','cngb3.1','cngb3.2']\n",
    "# plotName = 'CNGbeta'\n",
    "\n",
    "\n",
    "# genelist = ['rgs9a','rgs9b','rgs9bp',\n",
    "#             'rgs3a','rgs6',\n",
    "#             'rgs11','rgs12a','rgs16','rgs20']\n",
    "# plotName = 'RGS'\n",
    "\n",
    "# genelist = ['grk1a','grk1b',\n",
    "#             'grk3',\n",
    "#             'grk4','grk5','grk5l',\n",
    "#             'grk6',\n",
    "#             'grk7a','grk7b']\n",
    "# plotName = 'GRK'\n",
    "\n",
    "# genelist = ['rcvrna','rcvrnb','rcvrn2','rcvrn3','ncalda','ncaldb']\n",
    "# plotName = 'rcvrn'\n",
    "\n",
    "\n",
    "\n",
    "fH, axes = plt.subplots(1,len(genelist),sharey=True,figsize=[6*len(genelist),4])\n",
    "for i,g in enumerate(genelist,start=0):\n",
    "    n, h = plotBars(g)\n",
    "    plt.sca(axes[i])\n",
    "    pH = plt.bar(n, h, width=0.8, bottom=None, align='center', data=None,\n",
    "                 color=pC['plt'])\n",
    "    if  i == 0:\n",
    "        formatFigure(g, fH, axes[i], pH)\n",
    "    else:\n",
    "        formatFigure_right(g, fH, axes[i], pH)\n",
    "plt.savefig(\"{0}.svg\".format(plotName), transparent=True, format=\"svg\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Others"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "gdf = findgenelist('cacn')\n",
    "gdf = gdf.sort_values(by=['symbol'])\n",
    "gdf"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "genelist = ['ctbp1','ctbp2a']\n",
    "plotName = 'ctbp'\n",
    "\n",
    "genelist = ['cadm1a','cadm1b','cadm2a','cadm2b','cadm3','cadm4']\n",
    "plotName = 'cadm'\n",
    "\n",
    "genelist = ['tjp1a','tjp1b','tjp2a','tjp2b','tjp3']\n",
    "plotName = 'tjp'\n",
    "\n",
    "genelist = ['nrxn1a','nrxn1b','nrxn2a','nrxn2b','nrxn3a','nrxn3b',]\n",
    "plotName = 'nrxn'\n",
    "\n",
    "genelist = ['eml1','eml2','eml3']\n",
    "plotName = 'eml'\n",
    "\n",
    "genelist = ['elfn1b','elfn2a']\n",
    "plotName = 'elfn'\n",
    "\n",
    "genelist = ['cpne1','cpne2','cpne3','cpne4a','cpne4b','cpne5a','cpne5b','cpne7','cpne8','cpne9']\n",
    "plotName = 'cpne'\n",
    "\n",
    "# genelist = ['tbx2a','tbx2b','tbxas1','tbx21']\n",
    "genelist = ['tbx2a','tbx2b']\n",
    "plotName = 'tbx2'\n",
    "\n",
    "genelist = ['chn1','chn2']\n",
    "plotName = 'chn'\n",
    "\n",
    "genelist = ['ncdn','ncdn']\n",
    "plotName = 'ncdn'\n",
    "\n",
    "genelist = ['sema7a','sema6a','sema6d','sema4ab','sema4ba','sema3fa','sema3fb']\n",
    "plotName = 'sema'\n",
    "\n",
    "genelist = ['cacna1aa','cacna1ab',\n",
    "            'cacna1da','cacna1db',\n",
    "            'cacna1fa','cacna1fb',\n",
    "            'cacna2d1a','cacna2d2a',\n",
    "            'cacna2d4a','cacna2d4b',\n",
    "            'cacna1sb',\n",
    "           ]\n",
    "plotName = 'CaV'\n",
    "\n",
    "# genelist = ['stxbp1b','stxbp4','stxbp2','stxbp5a','stxbp5b','stxbp5l','stxbp3']\n",
    "genelist = ['stxbp1a','stxbp1b',\n",
    "            'stxbp2','stxbp3','stxbp4',\n",
    "            'stxbp5a','stxbp5b','stxbp5l',\n",
    "            'stxbp6','stxbp6l']\n",
    "plotName = 'stxBP'\n",
    "\n",
    "genelist = ['stx1a','stx1b','stx2a','stx2b','stx3a',\n",
    "            'stx4','stx5a','stx5al','stx6','stx7l',\n",
    "            'stx8','stx10','stx11a','stx11b.1','stx11b.2',\n",
    "            'stx12','stx12l','stx16','stx17','stx18']\n",
    "plotName = 'stx'\n",
    "\n",
    "genelist = ['prickle1a','prickle1b','prickle2a','prickle2b','prickle3',]\n",
    "plotName = 'prickle'\n",
    "\n",
    "genelist = ['pcdh15a','pcdh15b','pcdh8','pcdh11','pcdh10a']\n",
    "plotName = 'pdch'\n",
    "\n",
    "genelist = ['chn1','chn2']\n",
    "plotName = 'chn'\n",
    "\n",
    "\n",
    "genelist = ['lrit1a','lrit1b','lrit2','lrit3a','lrit3b']\n",
    "plotName = 'lrit'\n",
    "\n",
    "genelist = ['pou4f1','pou4f2','pou4f3','rbpms','grm6a','grm6b','vsx2','gfap','prkca','syt2a','rpe65a', 'opn4']\n",
    "plotName = 'controls'\n",
    "\n",
    "\n",
    "# genelist = ['gpx1a','gpx1b','gpx4a','gpx4b']\n",
    "# genelist = ['rgs9bp','RGS9BP']\n",
    "# genelist = ['chchd1','chchd10']\n",
    "# genelist = ['yipf6','yipf1']\n",
    "# genelist = ['sema3fb','sema3fa','sema6d']\n",
    "# genelist = ['esama','esamb']\n",
    "\n",
    "# genelist = ['pcdh1gc6','tex264a']\n",
    "# plotName = 'test'\n",
    "\n",
    "            \n",
    "fH, axes = plt.subplots(1,len(genelist),sharey=False,figsize=[6*len(genelist),4])\n",
    "for i,g in enumerate(genelist,start=0):\n",
    "    n, h = plotBars(g)\n",
    "    plt.sca(axes[i])\n",
    "    pH = plt.bar(n, h, width=0.8, bottom=None, align='center', data=None,\n",
    "                 color=pC['plt'])\n",
    "    if  i == 0:\n",
    "        formatFigure(g, fH, axes[i], pH)\n",
    "    else:\n",
    "        formatFigure_right(g, fH, axes[i], pH)\n",
    "plt.savefig(\"{0}.svg\".format(plotName), transparent=True, format=\"svg\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "gdf = findgenelist('opn1')\n",
    "# gdf = gdf.sort_values(by=['symbol'])\n",
    "# gdf = gdf.sort_values(by=['baseMean'], ascending=False)\n",
    "# gdf['symbol']\n",
    "gdf"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### According to Neuhauss recoverin1a in enriched in rods and UV-cones"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "***\n",
    "## heatMap: Opsins<a id='barManySubPlots'>∮</a>\n",
    "***\n",
    "[Back to Index](#Index)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Not sure this is helpful.  \n",
    "Maybe use this to show differential expression by normalizing by mean expression in correct type"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "gdf = findgenelist('opn1')\n",
    "gdf2 = findgene('rho')\n",
    "gdf2\n",
    "gdf = gdf.append(gdf2)\n",
    "gdf.index = range(len(gdf))\n",
    "gdf\n",
    "\n",
    "Opsins = [\"Rhodopsin (rho)\",\"UV-opsin (opn1sw1)\", \"S-opsin (opn1sw2)\",\"M-opsin (opn1mw1)\",\"M-opsin (opn1mw2)\",\"M-opsin (opn1mw3)\",\"M-opsin (opn1mw4)\",\"L-opsin (opn1lw1)\",\"L-opsin (opn1lw2)\"]\n",
    "Samples = [\"\",\"\",\"R\",\"\",\"\",\"\",\n",
    "           \"\",\"\",\"U\",\"\",\"\",\n",
    "           \"\",\"\",\"S\",\"\",\"\",\"\",\n",
    "           \"\",\"\",\"\",\"M\",\"\",\"\",\"\",\n",
    "           \"\",\"\",\"L\",\"\",\"\",\"\",\n",
    "          ]\n",
    "Samples = [\n",
    "          ]\n",
    "\n",
    "data = gdf.iloc[0:9,7:37]\n",
    "data = data.reindex([8,4,5,7,0,2,6,1,3])\n",
    "data.index = range(len(data))\n",
    "# np.around(np.divide(data.values,1e6),2)\n",
    "# opsinData = np.divide(data.values,1e6)\n",
    "opsinData = np.divide(data.values,1)\n",
    "\n",
    "# # log2 representation does not work well\n",
    "# log2matrix = np.vectorize(log2)\n",
    "# opsinData = log2matrix(opsinData)\n",
    "\n",
    "params = {\"ytick.color\" : \"w\",\n",
    "          \"xtick.color\" : \"w\",\n",
    "          \"axes.labelcolor\" : \"w\",\n",
    "          \"axes.edgecolor\" : \"w\",\n",
    "         \"axes.linewidth\" : 2,\n",
    "         \"xtick.major.width\" : 2,\n",
    "         \"ytick.major.width\" : 2,\n",
    "         \"xtick.major.size\" : 0,\n",
    "         \"ytick.major.size\" : 0}\n",
    "plt.rcParams.update(params)\n",
    "\n",
    "plotName = \"heatMap_opsins\"\n",
    "fig, ax = plt.subplots()\n",
    "\n",
    "im, cbar = heatmap(opsinData, Opsins, Samples, ax=ax,\n",
    "                   cmap=\"inferno\", cbarlabel=\"fpkm\")\n",
    "\n",
    "\n",
    "# texts = annotate_heatmap(im, valfmt=\"{x:.1f} t\")\n",
    "\n",
    "# fig.tight_layout()\n",
    "plt.savefig(\"{0}.svg\".format(plotName), transparent=True, format=\"svg\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "***\n",
    "## heatMap: Custom<a id='heatCustom'>∮</a>\n",
    "***\n",
    "[Back to Index](#Index)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "plotName = \"test\"\n",
    "\n",
    "# gdf = findgene('rho')\n",
    "# gdf2 = findgenelist('opn1sw')\n",
    "# gdf = gdf.append(gdf2)\n",
    "# gdf2 = findgenelist('opn1mw')\n",
    "# gdf2 = gdf2.sort_values(by=[\"symbol\"])\n",
    "# gdf = gdf.append(gdf2)\n",
    "# gdf2 = findgenelist('opn1lw')\n",
    "# gdf2 = gdf2.sort_values(by=[\"symbol\"])\n",
    "# gdf = gdf.append(gdf2)\n",
    "\n",
    "# gdf.index = range(len(gdf))\n",
    "# plotName = \"heatMap_opsins\"\n",
    "\n",
    "# gdf = findgenelist('slc24') -> NCKX transporter; slc25a5 is mitochondrial but known to express in eye\n",
    "# plotName = \"slc24\"\n",
    "\n",
    "# gdf = findgenelist('sema')\n",
    "# gdf = gdf.sort_values(by=[\"symbol\"])\n",
    "# plotName = \"heatMap_sema\"\n",
    "\n",
    "# gdf = findgenelist('cdh')\n",
    "# gdf = gdf.sort_values(by=[\"baseMean\"],ascending=False)\n",
    "# plotName = \"heatMap_cdh\"\n",
    "\n",
    "# gdf = findgenelist('eph')\n",
    "# gdf = gdf.drop([1,14,15,16]) #not ephrin receptors\n",
    "# gdf = gdf.sort_values(by=[\"symbol\"])\n",
    "# plotName = \"heatMap_eph\"\n",
    "\n",
    "\n",
    "# gdf = findgenelist('efn')\n",
    "# gdf = gdf.sort_values(by=[\"symbol\"])\n",
    "# plotName = \"heatMap_efn\"\n",
    "\n",
    "# gdf = findgenelist('pdgf')\n",
    "# gdf = findgenelist('rgr')\n",
    "# gdf = findgenelist('lmo')\n",
    "# gdf.index = range(len(gdf))\n",
    "# gdf.sort_values(by=[\"log2FoldChange\"])\n",
    "\n",
    "\n",
    "## genes in 2019 Papal, Swaroop: # igsf11, epha10, tuft1, prtg, lrfn2, vasn\n",
    "\n",
    "# gdf = findgenelist('igsf')\n",
    "# gdf = findgenelist('tuft')\n",
    "# gdf = findgenelist('prtg')\n",
    "# gdf = findgenelist('lrfn')\n",
    "# gdf = findgenelist('vasn')\n",
    "\n",
    "# gdf = findgenelist('rdh')\n",
    "# gdf = gdf.sort_values(by=[\"symbol\"])\n",
    "# plotName = \"rdhHeatMap\"\n",
    "\n",
    "# gdf = findgenelist('cyp2')\n",
    "# gdf = findgenelist('cyb')\n",
    "\n",
    "# gdf = findgenelist('sag')\n",
    "# plotName = \"heatMap_sag\"\n",
    "\n",
    "# gdf = findgenelist('dio')\n",
    "# gdf = gdf.sort_values(by=[\"symbol\"])\n",
    "# plotName = \"heatMap_dio\"\n",
    "\n",
    "# gdf = findgenelist('lrrn')\n",
    "# gdf = gdf.sort_values(by=[\"symbol\"])\n",
    "# plotName = \"heatMap_lrrn\"\n",
    "\n",
    "# gdf = findgenelist('dnmt')\n",
    "# gdf = gdf.sort_values(by=[\"symbol\"])\n",
    "# plotName = \"heatMap_dnmt\"\n",
    "\n",
    "# gdf = findgenelist('tet')\n",
    "\n",
    "# gdf = findgenelist('isl')\n",
    "# gdf = gdf.drop([1])\n",
    "\n",
    "# gdf = findgenelist('gj')\n",
    "# gdf = gdf.sort_values(by=[\"symbol\"])\n",
    "# plotName = \"heatMap_gj\"\n",
    "\n",
    "# gdf = findgenelist('fgf')\n",
    "# gdf = findgenelist('bmp')\n",
    "# gdf = findgenelist('fzd')\n",
    "# gdf = findgenelist('edar')\n",
    "# gdf = findgenelist('axin')\n",
    "\n",
    "# gdf = findgenelist('cadm') # cadm2a in S-Cones probably(?) and cadm3 in all photoreceptors (Hunter, JCN, 2011)\n",
    "# gdf = gdf.sort_values(by=[\"symbol\"])\n",
    "# plotName = \"heatMap_cadm\"\n",
    "\n",
    "# gdf = findgenelist('lrit') # lrit1b for rods and cones; lrit2 for cones; lrit3 for rods\n",
    "# gdf = gdf.sort_values(by=[\"symbol\"])\n",
    "# plotName = \"heatMap_lrit\"\n",
    "\n",
    "# gdf = findgenelist('elfn') # not detected\n",
    "# gdf = gdf.sort_values(by=[\"symbol\"])\n",
    "# plotName = \"heatMap_elfn\"\n",
    "\n",
    "# gdf = findgenelist('tenm') # for RGCs in zebrafish (Antinucci, Cell Reports, 2013)\n",
    "\n",
    "# gdf = findgenelist('foxo')\n",
    "\n",
    "# gdf = findgenelist('sdk')\n",
    "\n",
    "# gdf = findgenelist('crmp')\n",
    "\n",
    "\n",
    "\n",
    "# Even More\n",
    "# gdf = findgenelist('thr')\n",
    "# gdf = gdf.sort_values(by=[\"symbol\"])\n",
    "# plotName = \"heatMap_thr\"\n",
    "\n",
    "# gdf = findgenelist('lingo')\n",
    "# gdf = gdf.sort_values(by=[\"symbol\"])\n",
    "# plotName = \"heatMap_lingo\"\n",
    "\n",
    "# gdf = findgenelist('mitf')\n",
    "\n",
    "# gdf = findgenelist('prickle')\n",
    "# gdf = gdf.sort_values(by=[\"symbol\"])\n",
    "# plotName = \"heatMap_prickle\"\n",
    "\n",
    "# gdf = findgenelist('rdh')\n",
    "# gdf = findgenelist('sdh')\n",
    "# lingo, rx, mitf\n",
    "\n",
    "# gdf = findgenelist('dag')\n",
    "# can't find pikachurin: # gdf = findgenelist('egflam') gdf = findgenelist('si:dkeyp-100d11.2-001')\n",
    "\n",
    "# gdf = findgenelist('kcn')\n",
    "# gdf = gdf.sort_values(by=[\"baseMean\"], ascending=False)\n",
    "# gdf = gdf.drop(np.arange(20,109))\n",
    "# gdf\n",
    "\n",
    "# gdf = findgenelist('kcnj')\n",
    "# gdf = gdf.sort_values(by=[\"baseMean\"], ascending=False)\n",
    "# gdf\n",
    "\n",
    "# gdf = findgenelist('slc')\n",
    "# gdf = gdf.sort_values(by=[\"baseMean\"], ascending=False)\n",
    "# gdf = gdf.drop(np.arange(40,410))\n",
    "\n",
    "# gdf = findgenelist('rcv')\n",
    "# gdf = gdf.sort_values(by=[\"symbol\"])\n",
    "# plotName = \"heatMap_recoverin\"\n",
    "\n",
    "# gdf = findgenelist('sag')\n",
    "# gdf = gdf.append(findgenelist('arrb'))\n",
    "# gdf = gdf.append(findgenelist('arr3'))\n",
    "# gdf = gdf.sort_values(by=[\"symbol\"])\n",
    "# plotName = \"heatMap_arr\"\n",
    "\n",
    "\n",
    "# gdf = findgenelist('cng')\n",
    "# gdf = gdf.sort_values(by=[\"symbol\"])\n",
    "# plotName = \"heatMap_cng\"\n",
    "\n",
    "# gdf = findgenelist('cnga')\n",
    "# gdf = gdf.sort_values(by=[\"symbol\"])\n",
    "# plotName = \"heatMap_cnga\"\n",
    "\n",
    "# gdf = findgenelist('cngb')\n",
    "# gdf = gdf.sort_values(by=[\"symbol\"])\n",
    "# plotName = \"heatMap_cngb\"\n",
    "\n",
    "# gdf = findgenelist('myo')\n",
    "# gdf = gdf.sort_values(by=[\"symbol\"])\n",
    "\n",
    "# gdf = findgenelist('crx')\n",
    "# gdf = gdf.sort_values(by=[\"baseMean\"], ascending=False)\n",
    "# plotName = \"heatMap_crx\"\n",
    "\n",
    "\n",
    "# gdf = findgenelist('neurod')\n",
    "# gdf = gdf.sort_values(by=[\"symbol\"])\n",
    "# plotName = \"heatMap_neurod\"\n",
    "\n",
    "# gdf = findgenelist('ror')\n",
    "# gdf = gdf.drop([5,6])\n",
    "# gdf = gdf.sort_values(by=[\"symbol\"])\n",
    "# plotName = \"heatMap_ror\"\n",
    "\n",
    "\n",
    "# gdf = findgenelist('nr')\n",
    "# gdf = gdf.sort_values(by=[\"baseMean\"], ascending=False)\n",
    "# gdf = gdf.drop(np.arange(5,76))\n",
    "# plotName = \"heatMap_nrl\"\n",
    "\n",
    "\n",
    "# gdf = findgenelist('pcdh')\n",
    "# gdf = gdf.drop([1]) #highest ones that obscure others.\n",
    "# # gdf = gdf.drop([1,15]) #highest ones that obscure others.\n",
    "# gdf = gdf[gdf.baseMean > .45]\n",
    "# gdf = gdf.sort_values(by=[\"baseMean\"], ascending=False)\n",
    "# plotName = \"heatMap_pcdh_partial\"\n",
    "\n",
    "# gdf = findgenelist('hcn')\n",
    "# gdf = gdf.sort_values(by=[\"symbol\"])\n",
    "\n",
    "# gdf = findgenelist('gabr')\n",
    "# gdf = gdf.drop([0,2]) #highest ones that obscure others.\n",
    "# gdf = gdf.sort_values(by=[\"symbol\"])\n",
    "\n",
    "# gdf = findgenelist('opn4')\n",
    "# gdf = gdf.drop([1,2]) #highest ones that obscure others.\n",
    "# gdf = gdf.sort_values(by=[\"baseMean\"], ascending=False)\n",
    "# gdf = gdf.drop(np.arange(5,76))\n",
    "# gdf = gdf.sort_values(by=[\"symbol\"])\n",
    "# plotName = \"heatMap_opn4\"\n",
    "\n",
    "# CSNB causes: Nyx, cacna1f, GPR179 (Kamermans, 2019)\n",
    "# gdf = findgenelist('nyx') # bipolar cell -> not expressed in photoreceptors\n",
    "# gdf = findgenelist('gpr1') # bipolar cell -> not expressed in photoreceptors\n",
    "# gdf = findgenelist('cacna') # bipolar cell -> not expressed in photoreceptors\n",
    "\n",
    "\n",
    "# For Karen Carleton\n",
    "\n",
    "# gdf = findgenelist('rx')\n",
    "# gdf = gdf.drop([2,3,4,5,7,8,9,10,11,12,13,14,15,16,17,18,19])\n",
    "# gdf = gdf.sort_values(by=[\"symbol\"])\n",
    "# # gdf = gdf.drop([1])\n",
    "# plotName = \"heatMap_rx\"\n",
    "\n",
    "# gdf = findgenelist('tbx')\n",
    "# gdf = gdf.drop([3,17])\n",
    "# plotName = \"heatMap_tbx\"\n",
    "# gdf = gdf.drop([1,2])\n",
    "\n",
    "# gdf = findgenelist('tbx2')\n",
    "# gdf = gdf.drop([0,3,4])\n",
    "# gdf = gdf.sort_values(by=[\"symbol\"])\n",
    "# plotName = \"heatMap_tbx2\"\n",
    "\n",
    "# gdf = findgenelist('six')\n",
    "# gdf = gdf.sort_values(by=[\"symbol\"])\n",
    "# plotName = \"heatMap_six\"\n",
    "\n",
    "# gdf = findgenelist('lin')\n",
    "# gdf = gdf.drop([1,3,4,5,10,11,13,16])\n",
    "# plotName = \"heatMap_lin\"\n",
    "\n",
    "\n",
    "# gdf = findgenelist('sall') #sall4 mutations cause Duane syndrome; sall4 seems to be related to sall1 in zebrafish\n",
    "# plotName = \"heatMap_sall\"\n",
    "\n",
    "# For Mark Hankins\n",
    "# gdf = findgenelist('opn')\n",
    "# gdf = gdf.sort_values(by=[\"symbol\"])\n",
    "# gdf = gdf.drop([1,4,24,0,3,23,5,11,2])\n",
    "\n",
    "# gdf = findgenelist('rgr')\n",
    "# gdf = findgenelist('lrit')\n",
    "\n",
    "# gdf = findgenelist('ndrg') #potential target of tbx2\n",
    "\n",
    "# from YiRong Peng's talk: ON SACs: slit2 and fezf1; OFF SACs: tenm3 and rnd1?; contactins; Foveal cones: vtn and cacna1f\n",
    "# gdf = findgenelist('cacna2d2b')\n",
    "# gdf = gdf.drop([0])\n",
    "# gdf = gdf.sort_values(by=[\"symbol\"])\n",
    "\n",
    "# gdf = findgenelist('nrxn')\n",
    "# gdf = gdf.sort_values(by=[\"symbol\"])\n",
    "# plotName = \"heatMap_nrxn\"\n",
    "\n",
    "\n",
    "# gdf = findgenelist('frmpd')\n",
    "# gdf = gdf.sort_values(by=[\"symbol\"])\n",
    "# plotName = \"heatMap_frmpd\"\n",
    "\n",
    "# gdf = findgenelist('ntng')\n",
    "# gdf = gdf.sort_values(by=[\"symbol\"])\n",
    "# plotName = \"heatMap_netrins\"\n",
    "\n",
    "# gdf = findgenelist('dscam')\n",
    "# gdf = gdf.sort_values(by=[\"symbol\"])\n",
    "# plotName = \"heatmap_dscams\"\n",
    "\n",
    "\n",
    "# gdf = findgenelist('dag')\n",
    "# gdf = gdf.sort_values(by=[\"symbol\"])\n",
    "# plotName = \"heatmap_dag\"\n",
    "\n",
    "# gdf = findgenelist('syne') #involved in photoreceptor migration to ONL\n",
    "# gdf = gdf.sort_values(by=[\"symbol\"])\n",
    "# plotName = \"heatmap_syne\"\n",
    "# gdf = findgenelist('rasgrf') #also involved in phtoreceptor migration\n",
    "\n",
    "# gdf = findgenelist('syt')\n",
    "# plotName = \"heatmap_syt\"\n",
    "\n",
    "# gdf = findgenelist('rfx')\n",
    "# plotName = \"heatmap_rfx\"\n",
    "\n",
    "# gdf = findgenelist('anks') #anks3-/- causes RP\n",
    "# gdf = gdf.sort_values(by=[\"symbol\"])\n",
    "# plotName = \"heatmap_anks\"\n",
    "\n",
    "# gdf = findgenelist('spock') #https://zfin.org/action/image/view/ZDB-IMAGE-050208-655 -> GCL enriched?\n",
    "# gdf = gdf.sort_values(by=[\"symbol\"])\n",
    "# plotName = \"heatmap_anks\"\n",
    "\n",
    "# gdf = findgenelist('hcls') # Laura Campello: downstream of nrl and crx in mouse\n",
    "\n",
    "# gdf = findgenelist('slc32')  #from GO analysis: slc32a1 = L-cone specific + synaptic vesicle transporter.\n",
    "# gdf = findgenelist('slc') \n",
    "# gdf = gdf.sort_values(by=[\"baseMean\"],ascending=False)\n",
    "# gdf = gdf[gdf.baseMean > 100]\n",
    "\n",
    "# gdf = findgenelist('frmpd') #frmpd2 interacts with lrit1 in cones \n",
    "\n",
    "# from sqRNAseq + antibodies: rab32,ell2, ankrd33, pex5l, egr1, lamb1, chst3? or 7?, flna, cdh20, synj1\n",
    "# gdf = findgenelist('ell') # ell2 in sqRNAseql seems cone specific\n",
    "# gdf = findgenelist('ankrd33') #it's there, 33aa cone enriched; 33ba rods and 33bb cones.\n",
    "# gdf = findgenelist('pex5') #pek5l in sqRNAseq; pex5 in rods\n",
    "# gdf = findgenelist('egr') # egr1 looks interesting\n",
    "\n",
    "\n",
    "\n",
    "# gdf = findgenelist('gnat')\n",
    "# gdf2 = findgenelist('gnb')\n",
    "# gdf = gdf.append(gdf2)\n",
    "# gdf2 = findgenelist('gng')\n",
    "# gdf = gdf.append(gdf2)\n",
    "# gdf.index = range(len(gdf))\n",
    "# plotName = \"heatMap_transducin\"\n",
    "\n",
    "# gdf = findgenelist('aldh1')\n",
    "gdf = findgenelist('elovl')\n",
    "gdf = gdf.sort_values(by=[\"baseMean\"],ascending=False)\n",
    "# gdf = findgenelist('gnat')\n",
    "gdf = gdf[gdf.baseMean > 0]\n",
    "# gdf = gdf.drop(np.arange(20,93))\n",
    "# gdf = gdf.drop([4,24])\n",
    "# gdf = gdf.sort_values(by=[\"symbol\"])\n",
    "# gdf = gdf.sort_values(by=[\"padj\"])\n",
    "\n",
    "gdf\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "genenames = gdf['symbol'].values\n",
    "samplenames = [] #skipping names to replace in intaglio\n",
    "data = gdf.iloc[0:,7:37].values\n",
    "\n",
    "# log2matrix = np.vectorize(log2)\n",
    "# data = log2matrix(gdf.iloc[0:,7:37].values)\n",
    "\n",
    "fig, ax = plt.subplots(figsize=[12,12])\n",
    "im, cbar = heatmap(data, genenames, samplenames, ax=ax,\n",
    "                   cmap=\"inferno\", cbarlabel=\"fpkm\")\n",
    "# texts = annotate_heatmap(im, valfmt=\"{x:.1f} t\")\n",
    "\n",
    "plt.savefig(\"{0}.svg\".format(plotName), transparent=True, format=\"svg\")\n",
    "plt.show()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "plotName = 'test'\n",
    "\n",
    "# foxq2 = S specific;\n",
    "\n",
    "# genelist = ['wrb','get4'] # disrupts photoreceptors synapses (Danielle, 2016)\n",
    "# genelist = ['pappaa','igf2bp2b'] # In Miller, 2018 model pappaa cleaves igf-bp to increase igf1 availability. unfortunately, this is supposed to bind mRNA not IGF itself\n",
    "# genelist = ['gpx1a','gpx1b','gpx4a','gpx4b']\n",
    "# genelist = ['rgs9bp','RGS9BP']\n",
    "# genelist = ['chchd1','chchd10']\n",
    "# genelist = ['yipf6','yipf1','meis1b','pbx1a']\n",
    "# genelist = ['sema3fb','sema3fa','sema6d']\n",
    "# genelist = ['esama','esamb']\n",
    "# genelist = ['il34','il7r']\n",
    "# genelist = ['cxxc4','cxxc1a']\n",
    "# genelist = ['mbd3b','mbd3a']\n",
    "# genelist = ['cdr2a','cdr2l']\n",
    "# genelist = ['coro6','coro2a','coro1ca']\n",
    "# genelist = ['auts2a','auts2b']\n",
    "# genelist = ['plxnb1a','plxna2']\n",
    "# genelist = ['cux1a','cux1b','cux2b']\n",
    "# genelist = ['arih1','arih2','arih1l']\n",
    "# genelist = ['crhbp','crhr2','crhr1']\n",
    "# genelist = ['chn1','chn2']\n",
    "# genelist = ['robo1','robo2','robo3','robo4'] #nothing here\n",
    "# genelist = ['ntng1a','ntng2a','ntng2b'] #low counts but interesting. Netrins are important for RGC axons going towards optic disc. ntng2b in mouse mediates rod-hc recognition!\n",
    "# genelist = ['fat1a','fat1b','yap1','scrib','cc2d2a'] #fat1 important for renal development along yap1/hippo and scribble; cc2d2a involved in cilium development\n",
    "# genelist = ['c1qbp','c1qa','c1qb'] #keep this in mind. Beth Stevens work shows that microglia engulf synapses through activation by c1q; maybe c1qbp is produced in healthy cells (inside mitochondria) to keep c1q bound and gets released upon damage.\n",
    "# genelist = ['cc2d2a','rab8'] #Bahcman 2011: cc2d2a-/- causes Joubert syndrom. Important for cilium development by association with rab8. Not expressed in adults?\n",
    "# genelist = ['per2','per3']\n",
    "# genelist = ['gpsm2','gpsm1b','gpr107','gpr108'] #\n",
    "# genelist = ['tbx2a','tbx2b','gdf6a','cnbpa','egr1', 'ndrg1a','ndrg1b'] # fate genes\n",
    "\n",
    "#Hensley, 2011 smarca4\n",
    "# genelist = ['smarca4a','guk1a','guk1b','dtnbp1a','elovl4b','id2b','ckmt2a','ckmt2b','aanat2','pbx1a'] #smarca4 regulated genes: follow up elovl4b, ckmt2a/b, pbx1a\n",
    "# genelist = ['nme2a','nme3',''] #smarca4 regulated genes: follow up elovl4b, ckmt2a/b, pbx1a\n",
    "\n",
    "genelist = ['adgrl3.1','ntng2a','ntng2b'] #low counts but interesting. \n",
    "\n",
    "# genelist = ['ccdc28a','ccdc25','ccdc173','sox6','prdm9','prdm1b']\n",
    "\n",
    "# genelist = ['per2','bhlhe40','nr1d2b','tefa','aanat2','aanat1'] #circadian genes. Correlate with collection time?\n",
    "\n",
    "# genelist = ['elovl4b','elovl8a'] #a cone specific gene. \n",
    "# genelist = ['pbx1a','egr1','egr3','egr4'] # fishing for new targets\n",
    "# genelist = ['foxg1b','foxq2','foxo3b','lmo4b','nfia'] # fishing for new targets\n",
    "# genelist = ['anxa13','prdx5','mpp5b','mpp6a','serpina10a'] # fishing for new targets\n",
    "# genelist = ['rbp4l','rbp3','impg1a','impg2b','ntm'] # fishing for new targets\n",
    "# genelist = ['tapt1b','tapt1a','frmpd2','wdr37','fabp7a','tle3a'] # fishing for new targets\n",
    "# genelist = ['ankrd33aa','ankrd33ab','dhx32a','dhx32b','wdr17']\n",
    "\n",
    "\n",
    "# genelist = ['slc1a8a','slc1a8b','slc1a4','slc1a7a'] # EAATs (glutamate uptake)\n",
    "# genelist = ['pmela','slc6a6b','naa35','ush2a'] # sex-associated rod genes in Sun, 2018 (all enriched in females, except pmela)\n",
    "\n",
    "# genelist = ['pcloa','pclob']\n",
    "# genelist = ['mtdha','mtdhb']\n",
    "# genelist = ['igsf9b','igsf21a','igsf21b']\n",
    "# genelist = ['rx3','lin9','igsf21b','foxq2','foxo3b','foxg1b','rxrga','smad5']\n",
    "# genelist = ['sall1a','sall1b','sall2','sall4']\n",
    "# genelist = ['rbm8a','rbm39a','rbm39b','rbmx','rbm22','rbm5']\n",
    "# genelist = ['rbm5','rbm4.1','rbm4.2']\n",
    "# genelist = ['kif5ba','kif1aa','kif1ab','kif1bp'] #kif5-/- : Charcot-MarieTooth; Campbell 2013: kif5s in all retinal layers, but only kif5ba and bb are in ONL\n",
    "# genelist = ['kif3a','kif3b','kif16ba','kif19','kif25'] #\n",
    "# genelist = ['robo2','slit1a','nbeaa','prdm1b'] #\n",
    "\n",
    "\n",
    "# genelist = ['pcdh2ab9','pcdh2g28','pcdh15a','pcdh15b','pcdh11','pcdh10a'] #interesting ones\n",
    "# plotName = 'pcdhs'\n",
    "\n",
    "# genelist = ['sema7a','sema6d','sema3fb','sema3fa','sema4ab','sema4ba'] #the good ones\n",
    "# genelist = ['sema6a','sema3aa','sema6e','sema5bb','sema4c','sema6bb','sema6dl','sema5ba'] # the inconsistent ones\n",
    "# genelist = ['sema6ba','sema5a','sema3c','sema3b','sema3e','sema3d','sema3h','sema3ab','sema4e','sema4aa','sema3gb','sema3ga','sema4bb','sema3bl','sema4ga','sema4gb'] # the negative ones\n",
    "\n",
    "# genelist = ['ncdn','flrt3'] \n",
    "# genelist = ['odf2a','odf2b']\n",
    "# genelist = ['syne1a','syne1b','syne2b','syne3']\n",
    "# genelist = ['rasgrf2a','rasgrf2b']\n",
    "# genelist = ['rasgrf2a','rasgrf2b','nmnat2','rbx1','nrf1','unc119.2','unc119b'] # involved in migration\n",
    "\n",
    "# genelist = ['syt5a','syt5b','syt1a','syt1b','syt14a']\n",
    "\n",
    "\n",
    "# genelist = ['myo19','myo16','myo7bb','myo5b','myo1cb','myo5aa','myo7aa','myo6a','myo6b']\n",
    "# genelist = ['fgfr1op','fgfr1op2','fgf1b','fgf13a','fgf11b']\n",
    "# genelist = ['prdm1b','fgfr1op2']\n",
    "# genelist = ['frmpd1a','gpsm2'] # Ulisse + Pahlberg: these proteins inteact with rod transducin \n",
    "\n",
    "# genelist = ['epha7','frmpd2','pkp4','chd7','hook1']\n",
    "# genelist = ['thrb','samd7','samd1a','samd11','dscaml1'] #samd11:https://www.ncbi.nlm.nih.gov/pubmed/27734943\n",
    "# genelist = ['thrb','nrl']\n",
    "# genelist = ['opn3','opn4.1','opn4a','opn4b','opn5','opn6a','opn9'] #\n",
    "            \n",
    "fH, axes = plt.subplots(1,len(genelist),sharey=False,figsize=[6*len(genelist),4])\n",
    "for i,g in enumerate(genelist,start=0):\n",
    "    n, h = plotBars(g)\n",
    "    plt.sca(axes[i])\n",
    "    pH = plt.bar(n, h, width=0.8, bottom=None, align='center', data=None,\n",
    "                 color=pC['plt'])\n",
    "#     if  i == 0:\n",
    "    formatFigure(g, fH, axes[i], pH)\n",
    "#     else:\n",
    "#         formatFigure_right(g, fH, axes[i], pH)\n",
    "# plt.savefig(\"{0}.svg\".format(plotName), transparent=True, format=\"svg\")\n",
    "fH.tight_layout()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "***\n",
    "## heatMap: photoreceptors, rods, Müller Glia, microglia<a id='heatOther'>∮</a>\n",
    "***\n",
    "[Back to Index](#Index)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 5,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>symbol</th>\n",
       "      <th>genename</th>\n",
       "      <th>baseMean</th>\n",
       "      <th>log2FoldChange</th>\n",
       "      <th>lfcSE</th>\n",
       "      <th>pvalue</th>\n",
       "      <th>padj</th>\n",
       "      <th>R1</th>\n",
       "      <th>R2</th>\n",
       "      <th>R3</th>\n",
       "      <th>...</th>\n",
       "      <th>M4</th>\n",
       "      <th>M5</th>\n",
       "      <th>M6</th>\n",
       "      <th>S7</th>\n",
       "      <th>L1</th>\n",
       "      <th>L2</th>\n",
       "      <th>L3</th>\n",
       "      <th>L4</th>\n",
       "      <th>L5</th>\n",
       "      <th>L6</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <td>1</td>\n",
       "      <td>efna1a</td>\n",
       "      <td>ephrin-A1a</td>\n",
       "      <td>1.567270</td>\n",
       "      <td>-1.608322e-08</td>\n",
       "      <td>0.001443</td>\n",
       "      <td>0.961078</td>\n",
       "      <td>0.985671</td>\n",
       "      <td>0.185836</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>2.119078</td>\n",
       "      <td>...</td>\n",
       "      <td>0.073503</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>1.385170</td>\n",
       "      <td>0.0000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.260043</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>0</td>\n",
       "      <td>efna1b</td>\n",
       "      <td>ephrin-A1b</td>\n",
       "      <td>748.224327</td>\n",
       "      <td>-1.396069e-06</td>\n",
       "      <td>0.001443</td>\n",
       "      <td>0.028975</td>\n",
       "      <td>0.135968</td>\n",
       "      <td>745.161186</td>\n",
       "      <td>348.476739</td>\n",
       "      <td>10.123681</td>\n",
       "      <td>...</td>\n",
       "      <td>733.103043</td>\n",
       "      <td>241.950657</td>\n",
       "      <td>230.224561</td>\n",
       "      <td>403.251804</td>\n",
       "      <td>32.245952</td>\n",
       "      <td>0.198391</td>\n",
       "      <td>45.401299</td>\n",
       "      <td>98.8796</td>\n",
       "      <td>79.549634</td>\n",
       "      <td>88.273417</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>4</td>\n",
       "      <td>efna3a</td>\n",
       "      <td>ephrin-A3a</td>\n",
       "      <td>0.000413</td>\n",
       "      <td>2.960127e-07</td>\n",
       "      <td>0.001443</td>\n",
       "      <td>0.890184</td>\n",
       "      <td>NaN</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>...</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.0000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>5</td>\n",
       "      <td>efna3b</td>\n",
       "      <td>ephrin-A3b</td>\n",
       "      <td>0.138450</td>\n",
       "      <td>-1.627172e-07</td>\n",
       "      <td>0.001443</td>\n",
       "      <td>1.000000</td>\n",
       "      <td>NaN</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>...</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.0000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>4.153505</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>3</td>\n",
       "      <td>efna5a</td>\n",
       "      <td>ephrin-A5a</td>\n",
       "      <td>0.230419</td>\n",
       "      <td>8.321386e-07</td>\n",
       "      <td>0.001443</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>6.880405</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>...</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.0000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>5 rows × 37 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "   symbol    genename    baseMean  log2FoldChange     lfcSE    pvalue  \\\n",
       "1  efna1a  ephrin-A1a    1.567270   -1.608322e-08  0.001443  0.961078   \n",
       "0  efna1b  ephrin-A1b  748.224327   -1.396069e-06  0.001443  0.028975   \n",
       "4  efna3a  ephrin-A3a    0.000413    2.960127e-07  0.001443  0.890184   \n",
       "5  efna3b  ephrin-A3b    0.138450   -1.627172e-07  0.001443  1.000000   \n",
       "3  efna5a  ephrin-A5a    0.230419    8.321386e-07  0.001443       NaN   \n",
       "\n",
       "       padj          R1          R2         R3  ...          M4          M5  \\\n",
       "1  0.985671    0.185836    0.000000   2.119078  ...    0.073503    0.000000   \n",
       "0  0.135968  745.161186  348.476739  10.123681  ...  733.103043  241.950657   \n",
       "4       NaN    0.000000    0.000000   0.000000  ...    0.000000    0.000000   \n",
       "5       NaN    0.000000    0.000000   0.000000  ...    0.000000    0.000000   \n",
       "3       NaN    6.880405    0.000000   0.000000  ...    0.000000    0.000000   \n",
       "\n",
       "           M6          S7         L1        L2         L3       L4         L5  \\\n",
       "1    0.000000    0.000000   0.000000  0.000000   1.385170   0.0000   0.000000   \n",
       "0  230.224561  403.251804  32.245952  0.198391  45.401299  98.8796  79.549634   \n",
       "4    0.000000    0.000000   0.000000  0.000000   0.000000   0.0000   0.000000   \n",
       "5    0.000000    0.000000   0.000000  0.000000   0.000000   0.0000   0.000000   \n",
       "3    0.000000    0.000000   0.000000  0.000000   0.000000   0.0000   0.000000   \n",
       "\n",
       "          L6  \n",
       "1   0.260043  \n",
       "0  88.273417  \n",
       "4   0.000000  \n",
       "5   4.153505  \n",
       "3   0.000000  \n",
       "\n",
       "[5 rows x 37 columns]"
      ]
     },
     "execution_count": 5,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "geneTag = 'efna'\n",
    "gdf = findgenelist(geneTag)\n",
    "# gdf = gdf.sort_values(by=[\"baseMean\"],ascending=False)\n",
    "gdf = gdf[gdf.baseMean > 0]\n",
    "# gdf = gdf.drop(np.arange(20,93))\n",
    "# gdf = gdf.drop([4,24])\n",
    "gdf = gdf.sort_values(by=[\"symbol\"])\n",
    "# gdf = gdf.sort_values(by=[\"padj\"])\n",
    "\n",
    "rdf = findgenelist_rods(geneTag)\n",
    "rdf = rdf.sort_values(by=[\"symbol\"])\n",
    "\n",
    "mdf = findgenelist_glia(geneTag)\n",
    "mdf = mdf.sort_values(by=[\"symbol\"])\n",
    "\n",
    "gdf"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 6,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "image/png": "iVBORw0KGgoAAAANSUhEUgAAAtgAAADICAYAAADSmpa3AAAABHNCSVQICAgIfAhkiAAAAAlwSFlzAAALEgAACxIB0t1+/AAAADh0RVh0U29mdHdhcmUAbWF0cGxvdGxpYiB2ZXJzaW9uMy4xLjAsIGh0dHA6Ly9tYXRwbG90bGliLm9yZy+17YcXAAAgAElEQVR4nO3dfbhcVX3o8e/vkICAURAsRIiAQa+KUBqjtwpCIi9JU6v1XgR6Hy0WLA+klr7gXGrVCtY+SsfS2tJIAS0vtVZBWuESEoImApUKIZWXFMREfAOq+IZRIQL53T/2PslkzsxhTrL3mZmT7+d5Ts7ea6/122vNnnPml3XW7InMRJIkSVI1RvrdAUmSJGkqMcGWJEmSKmSCLUmSJFXIBFuSJEmqkAm2JEmSVKFp/e6A+qvRaOwB/CHwN81m88f97o8kSdKwcwZbewDvL79LkiRpOzmDPYAajcY8YGVLUQIPA/8JfKjZbH6pH/3S5Fo6f9m5wBsWrVw4d7zjwFrg8EUrF/7yOLEWAtcDBy5aufDb1fdWdVqweMOhwHuBI4C9gQeBfwaay5fMeKKffdNgmr5m9jyK15FHgBc+OWf9U13qjQBfBw4AfuPJOev/36R1UkPlzkMPmUfxnJrxynvW/rTP3Rl4zmAPtkOAg4CXAqcCvwBuaTQaC/raKw2ay4DDls5fdtg4dU4BbjK5Hj4LFm84DPgS8DhwEvArwN8DDeCcPnZNw2Efiv+Id7MA2G+S+iLtMJzBHmzfajabo/9LfKDRaKwAbgPOBpb3r1saMKuAbwJvBf5v+8Gl85c9F/hN4O2T2itV5UPA7cuXzHh7S9l9CxZvuBH4Wn+6pCGyFDgd+Lcux08H/hV4y6T1SNoBOIM9RJrNZgK3AC/sd180OBatXJjAFcBvLZ2/rNPP9EnAz+n+AqvBdjDF8rCtLF8y44HlS2ZkH/qj4bIEWDB9zewxrxvT18zel2J2+2OT3itpinMGe/gcCnxjdKfRaLwGeB/F2sxpFC/EH2k2m1slU41GYxeKPye/lWKt3UMUSdm/tNUbAc6lWFKwD3Av0Gg2m61rwjV4Lqd4HswDvtB27BTgnxetXLhxsjulSnwTeMOCxRs+vHzJjO/3uzMaOrcA/wWcRvGG9lanAuuenLN+5fQ1sye9Y9JU5gz2EGg0GiONRmNWo9H4MMV6ub8ty48BvkiRcM8HXg3cBHym0Wi8q6V9ANcC76B4o9ShFH8WPBa4su10Z1P8Ij4DmEPxi/njjUZjel3j0/ZbtHLheuDfKf4DtdnS+csOBl4LfKIf/VIl3k3xV6v7Fize0FiweIN/wdJEXQScOn3N7J1GC6avmR0Uv+sv6luvpCnMGezB9v1GowHFdRoB7gbe0mw2lzYajZ2Bi4G/bjabrW90WttoNL4FXNRoND7TbDa/RfFL9DXAS5vN5sNlvQcajcYXgBVt5zwSuL7ZbN4A0Gg0TgX2bDabT9Y0RlXnMuCvls5ftnjRyoWjd5Z4O3DXopULxywx0HBYvmTGneUbHf+MYgby/AWLN9wEnL98yYzP97d3GhJXAB8GFgHXlWXHATMp/volqWLOYA+21wKHU8w+fgc4otlsXl0eOwJ4EXBBh3b/CGwA3ljuvw34ZEtyDUCz2dwEnNfW9kvAyY1G45xGo/EK4Olms/loFYPRhPWyvra1zmeA6ZTXfen8ZUFx7Z29HnLLl8xYt3zJjN8Gng+cCOwJ3LRg8Ya3jt9SgifnrN8AfBL43Zbi3wU+9eSc9X7AmFQDE+zB9kCz2byf4s4Qu1DMYI2aCTzebDa/296ofDPk14EXtNR9sNs52psDH6RYt3sP8J1Go/G7Y1ppMjxGkUh1szew+cVx0cqFP6G4G8Bo0jUf2JfihVVTwPIlMx5fvmTG1RT/+V5J8X4JqRdLgEXT18zeb/qa2b8EvAnf3CjVxgR7CJQfYX4O8EeNRuOQsvgRYNdGo/FL7fXLNdcHlXVG6x7YJfyL2861qdls/mWz2Xw5sD/Fnw//odFodPywE9XqG8D+S+cve06X44fQ8obX0uXAwqXzl+1F8Z+kzy1aufAHtfVQfbF8yYwngaspfkalZ/TknPV3A1+mWDL4O8BdT85Zv7q/vZKmLhPs4XE5cDvwsTKB/hJFcvVHHer+NvBcijc2QjGD+X8ajcbM1kplnPe1lR3QaDSeC9BsNh9qNpt/SjGT2vVTAlWblcBG4A/bDyydv+xVwNGMvfXeTcD3KF5E/zcuDxl6CxZvWLBg8YaXdzj0OuArk90fDbWPUfxueAfOXku18k2OQ6LZbGaj0fg94E7glGazeVmj0TgduK7RaOwOfJoiGVsE/Cnwnmaz+c2y+SUU90L+UqPReA/FmyUPoFh60n53kKuATeVdSH5IsdxgV+DWWgeoMRatXPjY0vnL/gi4eOn8ZXtQfDT2z4DXU6yd/+yilQuvb2uzaen8ZVcCHwAeBW6c5G6rQgsWb5hGsQxk9oLFG95Hccu1GRRJ0hso3pQs9eoq4K+B59B2i1ZpAl5456GH/LytLF95z9pvdqy9g3IGe4g0m827KD4iudloNJ7XbDZXAK8CXgLcANwMLAR+vdls/mVLu6S4vd/FwJ9TJOl/R3G/5Le1neY3gfXA54A7KNbxvrHZbH61xqGpi0UrF34c+A2KWyZ+HlgDLAbOB07u0uxyijX7ly9auXDTZPRT9Vi+ZMZTwDEUt+Z8J8XP7meB3YBXLl8y464+dk9D5sk56zdS/FXr8ifnrG9PkKReraV4X1fr1/q+9mgARaYfBCZJkiRVxRlsSZIkqUIm2JIkSVKFTLAlSZKkCu3QCXZEnD5MceuMPWxx64w9bHHrjD1sceuMbdz6Yw9b3DpjD1vcOmMPW9w6Yw9b3DpjD3rcHTrBBup6QtX2RK0x9rDFrTP2sMWtM/awxa0ztnHrjz1sceuMPWxx64w9bHHrjD1sceuMPdBxd/QEW5IkSarUlLtN3xve8IZ82cte1u9uaMDM/u+DeeF3Duh3N7QNbjz4udw3/RX97oYGzPf/559y96FL+90N7UA+cMkLOP7WH/a7Gxogc++9j8xN0enYlJvB3nPPPfvdBQ2gAw40uR5WBx5wYL+7oAF04IEH9rsL2sH4nNNETLkEW5IkSeqnKbdE5Nxzz83zzvtgxVGfLr/vNCRx64w9bHGL2JlJxLTK4xaG67GoJ3Z9ceu5dkXswvA8FvXErTO2z4uxcavu81Pld58Xvo4Mc9w6Y9cTN/MpIkZ2nCUikiRJUj+ZYEuSJEkV2u4EOyLmRMSKiPhQFR2SJEmShlkVi4k+AizOzPu3NUBE7AZ8Cng0M99RQZ8kSZKkvqhiicis7Uyu9wJuAn5QQV8kSZKkvqoiwe747skJ+CnwPuCKCvoiSZIk9VXPCXa51vqGiLgpIq6NiMMiYhWwX0SsioijI+KWiDgrIm6MiLURsbil/TnlWu2bI+KaiNgFIDM3Zubnu5yzYxtJkiRpUPWUYEfEDOC9wImZeSxwGvDRzJwHPJSZ8zLzixQ3GzwgM48HjgTe35IUX52Zx2XmUcBa4KQeTt1Tm4g4PSJWR8TqXsYjSZIk1aXXGewjgMOB68pZ66uAmRGxR4e6lwJk5o+AB4B9yvKZEXFVRKwATgBm9XDentpk5sWZOTcz5/Y4HkmSJKkWvd5FJIDlmXnmmAMxZgn2hpbtjcBIROwPXAgsysyHI+IMYO9xT7gNbSRJkqR+63UG+1bgmIg4BCAiZkfE+RM4z14Ut+B7OIrPGX1zTW0kSZKkvuppBjszN0TEycCFZbK7CTh7Aue5G1gXEXcAjwC3AbvX0EaSJEnqq54/aCYz1wDzO5Qf3LI9r+3YsS27Y5aXtNVdBaxq2c9naiNJkiQNmirugy1JkiSpZIItSZIkVcgEW5IkSaqQCbYkSZJUoZ7f5DhcnjZu7bGHLW6dsYctbp2xhy1unbGHLW6dsYctbp2xn6op7jA+FsMWt87Yxq0/dh1xu6fRzmBLkiRJFZqSM9jzdj290nirHr8YgL12q/aT2H/w89Xl1k6Vxi2M/k+t6tjDFndL7LW/Nq/SqIfcsAqApzb9U6Vxp428FYDvnvrySuMC7POJ/6ol9mjcX3/271Ua9/qf/j0Ac3c9pdK4AKsfvxyARbsvrjTu0p8tAeBTh1bb59+6p+jvrUe+sdK4AEfeei0AFxx8RqVx/3jdRQDM2fVtlcZd8/iVAIyM7FFpXIBNm34MQLBLpXGTjQDstsvsSuP+fON6AI6q+HUP4Obyta+uPu+680GVxn38Fw8CMHP3oyqNC/DIz24ut3auOPIvyu/1vKbW9TweiedUGhdgU/4EgIhqP/Ik82e1xH0mzmBLkiRJFTLBliRJkipkgi1JkiRVyARbkiRJqtB2JdgRMSciVkTEh6rqUEvss6qOKUmSJNVte2ewPwL8fma+e1sDRMRuEfG5iLi07ZAJtiRJkobO9ibYszLz/m1tHBF7ATcBP9jOfkiSJEkDYXsT7NjO9j8F3gdcsZ1xJEmSpIHQU4JdrrW+ISJuiohrI+KwiFgF7BcRqyLi6Ii4JSLOiogbI2JtRCxuaX9OuVb75oi4JiJ2AcjMjZn5+XHOe05E3FbG/puI2HV7ByxJkiTV6RkT7IiYAbwXODEzjwVOAz6amfOAhzJzXmZ+keJjgw7IzOOBI4H3jybSwNWZeVxmHgWsBU7qoW8HAZsy8zWZ+TrgEeAvuvTx9IhYHRGrOx2XJEmSJksvM9hHAIcD15Wz1lcBMyOi02fUXgqQmT8CHgD2KctnRsRVEbECOAGY1cN5f5CZzZb9JrCoU8XMvDgz52ZmtZ9lLkmSJE3QtB7qBLA8M88ccyDGLMHe0LK9ERiJiP2BC4FFmflwRJwB7L2N/ZUkSZIGWi8z2LcCx0TEIQARMTsizp/AOfYCHi2T62nAm3ttFxHvatl/F7B0AueVJEmSJt0zzmBn5oaIOBm4sEyQNwFnT+AcdwPrIuIOinXUtwG799DuQWBaRPx7ec67gMYEzitJkiRNul6WiJCZa4D5HcoPbtme13bs2JbdMctL2uquAlZ1if3hXvooSZIkDYLtvQ+2JEmSpBYm2JIkSVKFTLAlSZKkCplgS5IkSRWKzOx3Hyp17rnn5nnnndfvbmjAZGan+7ZrCHjt1InPC002n3NqVTwfdibzFx2fFM5gS5IkSRP2VNcjPd2mb/jsVHG8p4csbp2xhy1unbGHLW6dsYctbp2xhy1unbGHLW6dsYctbp2xhy1unbGNW3/sOvvcnTPYkiRJUoVMsCVJkqQKmWBLkiRJFTLBliRJkiq03Ql2RMyJiBUR8aEqOiRJkiQNsyruIvIRYHFm3r8tjSPiBcDHgd2BTcD7MvOWCvolSZIkTboqEuxZ25pcl44GLsnMayLi+cBK4BUV9EuSJEmadFUk2Nv1sUaZ+amW7Ucj4omI2CMzf7z9XZMkSZImV89rsMu11jdExE0RcW1EHBYRq4D9ImJVRBwdEbdExFkRcWNErI2IxS3tzynXat8cEddExC4dzvFq4LHR5DoiTinPt7Jsu3cFY5YkSZJq09MMdkTMAN4LnJiZG8qlHJ/JzHkRsS4z55X1ngYOyMzjI2JP4P6I+HhmbgSuzszzy3p/DpwEXFHu7wrcArwIWNhy6puBKzNzU0ScBiwGPrD9w5YkSZLq0esM9hHA4cB15az1VcDMiNijQ91LATLzR8ADwD5l+cyIuCoiVgAnALNGG2Tm45k5F5gL/EOZwAPsBlxZtnlna5tWEXF6RKyOiNU9jkeSJEmqRa8JdgDLM3Ney9dLu6yT3tCyvREYiYj9gQuBP8jM44CP0mHtdmZ+neJNjq+NiGcBnwXOK9u8u1Obst3FmTm3TNIlSZKkvuk1wb4VOCYiDgGIiNkRcf4EzrMX8GhmPhwR04A3jx6IiOsj4iXl9m7A64C1wLOABNaXVd8ygfNJkiRJfdHTGuxy3fXJwIVlgrwJOHsC57kbWBcRdwCPALdR3Pca4DzgkojIsj8XZOY6gIj4JHBnRHybYj32nAmcU5IkSZp0Pd+mLzPXAPM7lB/csj2v7dixLbtndol7O8W9sDsd+yDwwV77KEmSJPXbdn9UuiRJkqQtTLAlSZKkCplgS5IkSRUywZYkSZIqZIItSZIkVajnu4gMl6eNW3vsYYtbZ+xhi1tn7GGLW2fsYYtbZ+xhi1tn7GGLW2fsYYtbZ2zj1h+7zj6P5Qy2JEmSVKEpOoO9U8XxRv/XMyxx64w9bHHrjD1sceuMPWxx64w9bHHrjD1sceuMPWxx64w9bHHrjG3c+mPX2efunMGWJEmSKmSCLUmSJFXIBFuSJEmqkAm2JEmSVKHtSrAjYk5ErIiID1XVoZbYZ1UdU5IkSarb9s5gfwT4/cx897Y0jogXRMQNEXFzRKyKiNe1HDbBliRJ0tDZ3tv0zcrM+7ej/dHAJZl5TUQ8H1gJvGI7+yRJkiT1zfYm2LE9jTPzUy3bj0bEExGxR2b+eDv7JUmSJPVFT0tEyrXWN0TETRFxbUQcFhGrgP3KpR1HR8QtEXFWRNwYEWsjYnFL+3PKtdo3R8Q1EbFLh3O8GnisNbmOiFMi4ray3WUR8bwKxixJkiTV5hkT7IiYAbwXODEzjwVOAz6amfOAhzJzXmZ+keKjcg7IzOOBI4H3tyTSV2fmcZl5FLAWOKkl/q4RsRpYBrSu5d4JmAMcUba7Abi0Sx9Pj4jVZRxJkiSpb3qZwT4COBy4rpy1vgqYGRF7dKh7KUBm/gh4ANinLJ8ZEVdFxArgBGDWaIPMfDwz5wJzgX8o12IDJPDezNxU1vs08OKIeFb7STPz4sycW8aRJEmS+qaXNdgBLM/MM8cciDFLsDe0bG8ERiJif+BCYFFmPhwRZwB7tzfMzK9HxErgtcDnxumLJEmSNLB6mcG+FTgmIg4BiIjZEXH+BM6xF/BomVxPA948eiAiro+Il5TbuwGvo1hCAkUy/ecRMVIePxF4IDOfmMC5JUmSpEn1jDPYmbkhIk4GLiwT5E3A2RM4x93Auoi4A3gEuA3YvTx2HnBJRGTZlwsyc1157GngK8AtEfE08C3gHRM4ryRJkjTperpNX2auAeZ3KD+4ZXte27FjW3bHLC8p69xOcS/sTsdGY1/WSx8lSZKkQbC9n+QoSZIkqYUJtiRJklQhE2xJkiSpQibYkiRJUoV6epPj8HnauLXHHra4dcYetrh1xh62uHXGHra4dcYetrh1xh62uHXGHra4dcY2bv2x6+zzWM5gS5IkSRWaojPYO1Ucb/R/PcMSt87Ywxa3ztjDFrfO2MMWt87Ywxa3ztjDFrfO2MMWt87Ywxa3ztjGrT92nX3uzhlsSZIkqUIm2JIkSVKFTLAlSZKkCplgS5IkSRXa7gQ7IuZExIqI+FAVHZIkSZKGWRV3EfkIsDgz79+WxhExD/gU8NWW4jdl5mMV9E2SJEmaVFUk2LO2NblucX1mvqOCvkiSJEl9VcUa7KgghiRJkjQl9Jxgl2utb4iImyLi2og4LCJWAftFxKqIODoibomIsyLixohYGxGLW9qfU67VvjkiromIXVrCHxcRt0bE6oh4fY9tJEmSpIHTU4IdETOA9wInZuaxwGnARzNzHvBQZs7LzC9SfFzOAZl5PHAk8P6WpPjqzDwuM48C1gInleVrgP+RmUcCbwEuiYg9n6FNe/9OL5Pz1RMbviRJklStXmewjwAOB64rZ62vAmZGxB4d6l4KkJk/Ah4A9inLZ0bEVRGxAjgBmFXW+0lmPlFuPwgsB147Xpt2mXlxZs7NzLk9jkeSJEmqRa9vcgxgeWaeOeZAjFmCvaFleyMwEhH7AxcCizLz4Yg4A9i7y7l+ATw1wTaSJEnSQOh1BvtW4JiIOAQgImZHxPkTOM9ewKNlojwNePPogYi4KCJOLLdnAa8HbhuvjSRJkjSoeprBzswNEXEycGGZ7G4Czp7Aee4G1kXEHcAjFAn07uWxJnBpRLwTeBI4LTN/EhHjtZEkSZIGUs/3wc7MNcD8DuUHt2zPazt2bMvumOUlZZ31XeJmtzaSJEnSoKriPtiSJEmSSibYkiRJUoVMsCVJkqQKRbHUeep429velvvuu2+/u6EB8/pnHUPe2u9eaFs8cMwvc8PDz+53NzRgnn3SxVw7Y0m/u6EdyBUrXstL/mlNv7uhATL33vvI3DTmftXgDLYkSZJUqSk3gx0Rb8/My/rdD0mSJE1d4+WcUy7BliRJkvrJJSKSJElShUywJUmSpAqZYEuSJEkVMsGWJEmSKmSCLUmSJFXIBFuSJEmqkAm2JEmSVCETbEmSJKlCJtiSJElShUywJUmSpAqZYEuSJEkVmtbvDmhqixhJSCA6He3Wasx2tB/bqmmXOh32Y0z78dqOH7dTeXQ5d291O5Rn93ON2ds8tG5xeyvbUl7+26VSt7Ydj3W6bL30ITqU9RgjxmzktsXpefzZPW6HA53ajykf53EryrNTYe/xx/RvbLwt9bd+MnbuU7bF6yzIroPact3G6UuHY5uPdzm2pd/dH7Ot+tehvPAMj0On80eXxz5a4nXpC2T3x7P9XFvVG28MXcrbH4exvw47VOpQ3l7W1iS3Ku/8O7ljJ8Y8EOO0baubm/d7id/2fav4sVXdMT/1LTGiW7yO5xv9rdvtNar4fuedDy7PzIVo4Jlgq2YJTCt+OWz+ZVL84SRG/4ASI2PKo70sdtqqzejx1rrd6+y0eX+kS7yR9rqMMNJeh/b9LW1GOpS17o/klv621x09NrL5F+nI5u2xx2Kr/RFiS1m0H2PrNhFbyqKtTtv+SLS2G40zeqy9bac6jKmzJe4z1xmN1bXOmDb5zHFbkr+RDmVbt8nN+yNtydHm/rXUGd3f8vi1t8kxbZ65Tm41vqJubj3GlvJO59hqv+X4lnGPPda+v+Wc47eJGNuHbakbI6NtNm3pQ9c6o+UtdUfGj7u57kj3uK11Nh8f71jLPiObOpZta11GctxjW5WPJLQd29J29DstddmqbKtfALT+am79oRn9ARjZen9kZMv3TmUt33Pz/k7j1NmpQ9yiLKOlfcv31rg55lj3/S3b09rqtO3HtJY607auE9PaYkwjNpdNL6rE1nWjrBuxpe7I6LHR15/R8rIP0+Kte6OhYIKtCYnip/7XgQNpef5k5gX96pMkSdIgMcHWRF0HPAHcA2zqc18kSZIGjgm2Jmr/zDys352QJEkaVN5FRBN1Q0Qc3+9OSJIkDSpnsDVR/wH8axTvIHyS4m0ymZnP6W+3JEmSBoMJtibqr4DXAPdkZpf7YUmSJO24XCKiifoacK/JtSRJUmfOYGuiHgFWRcQNwMbRQm/TJ0mSVDDB1kQ9WH7tXH5Bp49/kyRJ2kGZYGuiLs/Mb7QWRMSr+tQXSZKkgeMabE3UZyNiv9GdiDgK+EQf+yNJkjRQTLA1UWcA/xYR+0bEIuBvgUV97pMkSdLAcImIJiQz74iIs4AbKT4y/bjMfLTP3ZIkSRoYJtjqSURcx9ZvZtwNeAz4eESQmW/sT88kSZIGiwm2evWRfndAkiRpGJhgqyeZ+cXR7YjYF3g1xYz2HZn5333rmCRJ0oDxTY6akIh4B3A78L+AE4D/iIhT+9srSZKkweEMtiaqAfxKZv4AICL2Ar6Et+qTJEkCnMHWxH0H2NCyvwH4dp/6IkmSNHCcwdZEPQR8OSI+R7EG+03A7RHxxwCZeUE/OydJktRvzmCrJxFxZbn5RuDf2HLLvs8BjwAzyi9JkqQdmjPY6tUrI+IA4FvA37UfzMwfTn6XJEmSBo8Jtnp1EbAMOAhY3VIeFLPZL+rSbi089UTC1h9T07rbVj5F7A18v9+d6IMdddyw447dce9YHHd/DUIf1IPInJrZjeoRER/LzDMnUH91Zs6ts0+DyHHveHbUsTvuHYvjlnrjGmxNyESSa0mSpB2RCbYkSZJUIRNs1e3ifnegTxz3jmdHHbvj3rE4bqkHrsGWJEmSKuQMtiRJklQhE2zVJiIWRsRXI2JdRPxJv/tTtYj4RkTcExFfiYjVZdnzImJFRHyt/L5nWR4R8bflY3F3RMzpb+97FxGfiIjvRcS9LWUTHmdEnFLW/1pEnNKPsUxEl3GfGxEPldf8KxGxqOXYu8txfzUiFrSUD9XPQUTMioiVEXFfRKyNiD8oy6f0NR9n3FP6mkfEsyLi9oi4qxz3eWX5QRHx5fLafToidi7Ldyn315XHD2yJ1fHxGETjjPuyiHiw5XofXpZPiee5JlFm+uVX5V/ATsB6ivtj7wzcBby83/2qeIzfAPZuK/tL4E/K7T8Bzi+3FwE3UNw3/FeBL/e7/xMY51HAHODebR0n8Dzg6+X3PcvtPfs9tm0Y97nAuzrUfXn5HN+F4l7x68ufgaH7OQBmAnPK7RnAA+X4pvQ1H2fcU/qal9ft2eX2dODL5XX8DHByWX4RcGa5vRi4qNw+Gfj0eI9Hv8e3DeO+DDihQ/0p8Tz3a/K+nMFWXV4NrMvMr2fmL4B/Ad7U5z5NhjcBl5fblwO/2VJ+RRb+A9gjImb2o4MTlZk3A+2f1DnRcS4AVmTmDzPzR8AKYGH9vd92XcbdzZuAf8nMjZn5ILCO4mdg6H4OMvORzFxTbm8A7gP2Y4pf83HG3c2UuObldftpuTu9/Erg9cDVZXn79R59HlwNHBMRQffHYyCNM+5upsTzXJPHBFt12Q/4dsv+dxj/xWoYJXBjRNwZEaeXZftk5iNQvGADv1SWT7XHY6LjnErjf2f5J+JPjC6TYIqOu/zz/69QzO7tMNe8bdwwxa95ROwUEV8BvkeRIK4HfpyZT5VVWseweXzl8ceAvZgC487M0ev9F+X1/uuI2KUsmzLXW5PDBFt1iQ5lU+2WNUdk5hzg14Dfi4ijxqm7Izwe0H2cU2X8HwNmA4cDjwB/VZZPuXFHxLOBzwJ/mJk/Ga9qh7KhHXuHcU/5a56ZT2fm4cD+FLPOL+tUrfw+ZccdEa8A3g28FHgVxbKPc8rqU2bcmhwm2KrLd4BZLfv7Aw/3qS+1yMyHy+/fA/6V4oXpu6NLP8rv3yurT7XHY6LjnBLjz8zvli/Km4BL2PIn8Ck17oiYTqLGqbgAAAHpSURBVJFkfjIzrymLp/w17zTuHeWaA2Tmj4FVFGuM94iIaeWh1jFsHl95/LkUS6mmwrgXlkuFMjM3Av/IFL7eqpcJtupyB/Di8p3oO1O8GebaPvepMhGxe0TMGN0GjgfupRjj6LvITwE+V25fC/x2+U70XwUeG/1z+5Ca6DiXA8dHxJ7ln9iPL8uGStu6+TdTXHMoxn1yeYeFg4AXA7czhD8H5XrajwP3ZeYFLYem9DXvNu6pfs0j4vkRsUe5vStwLMX685XACWW19us9+jw4AfhCZibdH4+B1GXc97f8JzIo1p23Xu+hf55rEvXr3ZV+Tf0vinddP0Cxnu89/e5PxWN7EcU75u8C1o6Oj2It4ueBr5Xfn1eWB/D35WNxDzC332OYwFg/RfGn8ScpZmtO25ZxAqdSvPFpHfA7/R7XNo77ynJcd1O84M5sqf+ectxfBX6tpXyofg6AIyn+xH038JXya9FUv+bjjHtKX3PgMOA/y/HdC/xZWf4iigR5HXAVsEtZ/qxyf115/EXP9HgM4tc44/5Ceb3vBf6JLXcamRLPc78m78tPcpQkSZIq5BIRSZIkqUIm2JIkSVKFTLAlSZKkCplgS5IkSRUywZYkSZIqZIItSZIkVcgEW5IkSaqQCbYkSZJUof8PugwyBSUBrPgAAAAASUVORK5CYII=\n",
      "text/plain": [
       "<Figure size 864x864 with 2 Axes>"
      ]
     },
     "metadata": {
      "needs_background": "light"
     },
     "output_type": "display_data"
    },
    {
     "data": {
      "image/png": "iVBORw0KGgoAAAANSUhEUgAAAO0AAACXCAYAAAD59NwoAAAABHNCSVQICAgIfAhkiAAAAAlwSFlzAAALEgAACxIB0t1+/AAAADh0RVh0U29mdHdhcmUAbWF0cGxvdGxpYiB2ZXJzaW9uMy4xLjAsIGh0dHA6Ly9tYXRwbG90bGliLm9yZy+17YcXAAARQ0lEQVR4nO3de7Ab1X3A8e9P0rUvhgDCbsABBrARpJTQ4qGCBhMcDBnGgZI64ZEUCh0yNKENuCnbkJDUZYAhZIECAy1JYMaQpDxsPIECKU0Iz9QgHgFS85J5DAFc8xIGQ21j31//OEe6e++V7pWs1V2t9PvM7GgfZ4/OnqvfPdrV6idRVYwx6ZFJugHGmNZY0BqTMha0xqSMBa0xKWNBa0zKWNB2sSAIFgVB8CdJt6NflEv5ReVSvuv724K2uy0Cuv5F1ENS0d+5pBtgTCPlUn5f4HfAHoVi5eWEm9M1bKQ1JmUsaI1JGQtaY1LGzmnbFATBEmAd8BDwT8BsYFfcP8Rzgc8DeWAVsAT4lzAMN42q4yu4iyCf8nXdDpxT57m+DJwN7AW87uu6sgOH1ZXKpfwS6vR1oVh5vVzK70KD/i4UK5tG1dNUf5dL+TH9XShWEu9vG2njcTDuRfS3wL7AJuARYBZwPLAf8AMgAG4JgqDW70EQnANcAywF5uBedEPAg8A2kXIHAtcB/wbsD1wJXBIEwR92+Ni6zYi+9gH7ccbp73IpX+vvcinfVH+XS/m6/V0u5RPvbxtp47EPsHcYhi8DBEHwU+BlYH4YhkO+zHNBEDwMPIV7Yd3gA24xsDAMw9sj9ZWCIFgM/HNk3aeBl8MwvNovPxsEwY1hGK7u0DF1q32AvUddTb4U39+FYqXW3+VSfkR/+4BbDCwsFCsj+rtcytft70KxUuvvcil/Y6FYSby/baSNx4pIwE4FjgUuiwQsAGEYPot7K/ZFv+oE4IVRAVt1IbAhsvwQMDsIgquCIJgbBMGUPgxYgBXRgC2X8rX+jgQsAIVipW5/jwrYqrr9XS7lryqX8nPLpfyUbghYsKCNy1uR+enAFOClBmVfAD7h52c2KheG4Ubc6FFdXgEsBP4YuAd4JwiCy4MgGGir5enz1qjlWPq7UKyM6O9CsTKmv8ul/OXlUj7x/ragjd/bwEZg9wbbZwPV/9irG5XzwbhbdF0YhreGYTgXd6HlTOAU4BvtNjjlYulvH4wj+rtQrNxaKFa6rr8taGMWhuEG4BbgzCAIJLotCIK9cBc+bvGrbgYKQRAsqFPVt4DByL7TgiDY0z/HujAMrwV+jbtI0pMKxcr/FIoVGe9uqEKxUuvvcik/or/LpXzd/i6X8hP2d7mUn1Yu5ff0z7GuUKx0TX/bhajO+CbwW+CuIAguwX1ccADuvOku4EaAMAxXBkFwPrDUX3i6G9gW+AowH3gjUue5wMlBEJwGPAMcBByBu4rak8ql/FHATcDxDc5Dq2r9XS7lG/Z3oVhZWS7lzweW+gtPE/Z3uZTvuv62kbYDwjD8X9xHP68A1wOP4f6Tnw8cE71AFYbhucCJwJeBFcAyYAA4BHg/Uu13cR9VXAE8CXwH+E4Yhtd1+ngSpMBm/9hQoVgZt7+jF6gKxUpb/V0oVhLvb7HEbsaki420xqSMBa0xKWNBa0zK9HXQishpaaq3k3Wnrd5O1t3t9fZ10AKdekF17IXawbrTVm8n6+7qevs9aI1JnZ77yOess85SEZm4oNlixx79LvnBZUk3o6ftf8h2rNvwct0Xcs+NtGvWrEm6Cca0bfPQhobbei5ojel1PXfv8ezZs7nwgmtirXPjptcA2HZwn1jrfW/90wBsPXWvWOsF+GDD8wBMHdg11nrXb3wFgJ22eSbWetd88BsAdtz64FjrjdYdd19s+Oj3AOwwbU6s9b79wWO4bxHWZyOtMSljQWtMyrQdtCIyR0R+KSIXxtEgY8z44jinvRg4XVWf3dIKRGQacAPwpqp+NYY2GdOz4nh7vGubATsd+BUubYgxZgJxBG27dzKsA76H+/KyMWYCTQetP3f9hYj8SkRuE5H9ROReYGcRuVdEDhWRB0TkDBH5LxFZKSKnR/b/lj/3vV9ElovIVABV3aCqdzd4zrr7GNPPmgpaEfkYLv3Gcap6OHAqcLmqzgNeU9V5qnofLjXIbqr6OWAusDgSaMtU9QhV/QywEpdAeiJN7SMip4nIoyLyaDPHY0yaNTvSHoz7sd3/8KPrUmCmiGxfp+w1AKpaAZ4HdvTrZ4rIUhH5JfAl3O/dTKSpfVT1R6p6gKoe0OTxGJNazV49FuAuVf36mA1jb86PJsfaAGREZBfcb6EsUNXXReRrwIxxn3AL9jGmHzQ70j4IzBeRPwIQkdkiclELzzMd93HO6yKSA/6iQ/sY0/OaGmlV9X0ROQG40gfQEPAPLTzPU8AqEXkEl+V9BbB1B/Yxpuc1fXOFqj4OfLbO+j0j8/NGbTs8sjjmrfWosvcC90aWdaJ9jOlHdu+xMSljQWtMyljQGpMyFrTGpEzPZa6A4UwTcatmmohbNctEJ1SzK8Stmg0iLfVC5/rinQ8fj73O4Kg3G26zkdaYlOnJkXbRzDNire+y1VcAsGDr0yco2Zo7P/hXAE6ZHm97AZa87do8b6vO5N3eNPTTWOvLZU4EYO3bi2KtF2C76ZcB8J9/elys9R75yM0AfHT94AQl42UjrTEpY0FrTMpY0BqTMha0xqRMW0HbyUyMIhL/1RljekC7I+3FwDdU9dtbWoGITBORW0Vk9M8CWNAaU0e7QWuZGI2ZZO0GrWViNGaSNZvYbdIzMUb2W+HrvkxEtmr3gI1JuwmDNsFMjHsAQ6r6Z6p6CC57xQUN2mjZGE3faGakTSoT49uqGkaWQ2BBvYKWjdH0k2buPZ70TIzGmMaaGWmTyMQIMF1EzoosnwXc2cLzGtOTJhxpE8rECPASkBOR3/jnfBIIWnheY3pSsylUJzUT46i6v99MG43pF3bvsTEpY0FrTMpY0BqTMha0xqSNqvbUtHjxYgVs6tBUlXQ7enlSVf3eF7La6DVuI60xXei8nw813NaT2RghG3N9m1NWb6fr7kS91fZ24iW5yT+m6e/XmI20xqSMBa0xKWNBa0zKWNAakzKWjdGYlEk0G6OIfMKnsbnfp605JLLZgtaYOtq9vt5WNkbgUODHqrpcRP4AuAfYt802GdPT2g3atrIxquoNkfk3RWS9iGyvqu+22S5jelai2RhHPUcRWBsNWBE52WdjvF9ElojIDjEcszGpNuFIOyob4/v+bezNqjpPRFZVv/wuIrVsjCKSB54VkWtVdQMuG+NFvtx5uGyM1/vlrYAHgFnAkZGnzgJzgINVdUhEjscljlsYy5Ebk1KJZ2NU1f/zWRQPAH7o/ymAu3n6u6o65MvdBBREZMwv+FoKVdNPmgnaajbGeZHpkw3OO8fLxnimqh4BXE6dc2FVfRF3IerTE7RlDEuhavpJotkYReQOEdnLz08DDsElMwcXoOeJSMZvPw54XlXXt/DcxvScpLMxngv8WETUt+VSVV3lt20GngAe8OfLrwBfbeF5jelJiWZjVNUS7rPaetuqdS9ppo3G9Au799iYlLGgNSZlLGiNSRkLWmNSxoLWmJTp0cRumycu0hf1drLuTtW7aeIiWyxNfdH4uzg20hqTMj060qYlVWYa04ZWTYm5vo0dqjdad6f+fp1oc2M20hqTMha0xqSMBa0xKWNBa0zKtB20nUyjaowZK47LlhcDp29pVkYRmQfcADwXWX2Mqq6NoW3G9Jw4grbdNKoAd6iqfVfWmCbEcU7bVhpVY0xrmg7aDqdRPUJEHvTJ2Q5rch9j+lKzeY+jaVQPB04FLvfZKl7zyd7uw90ispuqfg6YCyyOBNoyVT1CVT+DywN1vF//OLC3qs4FjsWln8lPsM/o9lk2RtM3mh1pO5ZGVVXfqyZrU9WXgLsYzsjYMPVqlGVjNP2k2QtR1TSqY3I9iYw5pR0vjeoCn5Xxa8CMBs+1EdjU4j7G9I1mR9pOplG92qdHRUR2BQ7DZWxsuI8x/azZbIydTKMaAteIyN8BHwGnqup7IjLePsb0raY/p+1gGtUXGtSrjfYxpp/ZvcfGpIwFrTEpY0FrTMqIO3XsHSeddJLutNNOLe+3fv16nn76adauXcvQ0FBt/fz58+NsXk849uh3yQ8uS7oZPW2vA99DdXPdW4R7NEdU65YvX87MmTOZMWNGvc+ejekaPTfSisgpqrpkC/Z7XFXndKBJxrRsvNdxzwXtlhKRvwfWAbfj7uQCQFXfSaxRxtRhb4+HbcTd6HEOUP1PpsCsxFpkTB020noi8gJwoKq+lXRbjBmPjbTDVgIftrKDSEaHcwAIQgbEzeEnEaktRx/Hzvslrc7h963WTm1ddbm2XkaWcW2IlBEAjdQRLa+1Mm61jiqjkfqqZbW2rVZ3tbxoU49IdF59u3XU/MgyrhHD+7vt0YOuNlTcUUidya9XyUTWVeczfr8MIhlfNuP/ru7v9NhjL92lqkeSIAvaYZuBJ0TkHkae054x3k4ZGQTJkZEcGZnqHjMDfnmArAyQlRwZ/DxuypFz85ojR46cZsmS9Y8ZsmSYIhmykiErQlaEnFCbzwpkBbcuQ225tk4gm9HIeiUnSi7jHqvralNmyK9XsjLk963OD9XK1JYj87nsZjKZ4fWZzBBZvy6THSKb2UwmO+SX/XzWzUt2eJ1kNyM5v31gk19268m5spIbgpwiWYUcSE4gl4FsFnLVKYfmcpAb8PMDkBtABwbR3ACamxKZpropOxVyg2h2EHKDSHYrJDNIJjtINjNIJjOVTCZLTk5M/JtmFrTDfu4nY7qaBa2nqteJyBTgk7gLUM+p6sYJdjNm0lnQeiKyAPgh8ALuzGgPEfkbVf1Fsi0zZiQL2mGXAp9V1VXgvugP3AFY0JquYkE77I1qwHovAm+Mv4uuHNIP16MuK0CKzQB64aOuyTiOxPvJgnbYShG5E7gZd057LPCIiCwEUNXldfZZ3wvJ5ETkUTuO9LCgHTYIrAEO9ctvAnngaFwQ1wtaYyadBe2wDHCmqr4L4HMvX6Kqf51ss4wZyb4EP2y/asBCLW/z/hPs86PONmnS2HGkiN177InIk8A8H6yIyA7Afar6qWRbZsxI9vZ42CXAf4vIMtw57HHABck2yZix7O2xp6rXA1/EXYx6E1ioqj+pV1ZEjhSR50RklYicPZntbJWI7Coi94jIM/5H0c7063fwP25W9o95v15E5Ap/bE+JSFclBhCRrIj8VkRu98t7iMjD/jhu8ne1ISJT/fIqv333JNsdK1W1qYUJyOLumpoFTAGeBPZJul3jtHcmMMfPfwz3+0r7AD8AzvbrzwYu8vMLcDeUCHAQ8HDSxzDqeL4J/Dtwu1++GTjBz18NfN3Pnw5c7edPAG5Kuu1xTTbStq4IrFLVF9Xdm3wjcEzCbWpIVVerSzSPqr4PPAPsjGvzdb7YdcAX/PwxwPXqPARsLyIzJ7nZdfnfd/o8/kfexH1P8TCgmmVu9HFUj28Z7mdteiL5lwVt63YGfh9ZftWv63r+LeL+wMPAjqq6GlxgAx/3xbr5+C4D/pHhG9CmA++q6ia/HG1r7Tj89rW+fOpZ0Lau3n/rrr8ELyLbALcAi1T1vfGK1lmX+PGJyFG4W00fi66uU1Sb2JZqdvW4da8y8ndydwFeT6gtTRGRAVzA/kyHb8dcIyIzVXW1f/tbvc+6W4/vYODP/bexBoFtcSPv9iKS86NptK3V43jV/2jcdkBPJOmzkbZ1jwAFf9VyCu4ix20Jt6khfx53LfCMql4a2XQbcLKfPxm4NbL+r/xV5IOAtdW30UlS1W+r6i6qujuuz3+tqn8J3IP7wXEYexzV4/uSL98TI23iV8LSOOGusD6Pu4p8TtLtmaCtc3FvC58CnvDTAtz53d1A2T/u4MsLcJU/tt8BByR9DHWOaR7DV49nASVgFbAUmOrXD/rlVX77rKTbHddkd0QZkzL29tiYlLGgNSZlLGiNSRkLWmNSxoLWmJSxoDUmZSxojUkZC1pjUub/AcqM9pY8y/uAAAAAAElFTkSuQmCC\n",
      "text/plain": [
       "<Figure size 230.4x230.4 with 2 Axes>"
      ]
     },
     "metadata": {
      "needs_background": "light"
     },
     "output_type": "display_data"
    },
    {
     "data": {
      "image/png": "iVBORw0KGgoAAAANSUhEUgAAAaAAAAEaCAYAAABXZ4NKAAAABHNCSVQICAgIfAhkiAAAAAlwSFlzAAALEgAACxIB0t1+/AAAADh0RVh0U29mdHdhcmUAbWF0cGxvdGxpYiB2ZXJzaW9uMy4xLjAsIGh0dHA6Ly9tYXRwbG90bGliLm9yZy+17YcXAAAgAElEQVR4nO3df5xddX3n8df7ZkiAiEaDSyOTRSQqCAXEdDYCwgAh61JrSx0hdW3VxbKEjEgfYFl3yaKP1KVUHgrdSVOBUltbxWagBVTMD8nwo024BARqAGkUVgawQkCJwfycz/5xz52cmbkzuZN753xnMu/n43GT8/P7+Z5zv3M/93vOuecoIjAzMytaKXUFzMxscnICMjOzJJyAzMwsCScgMzNLwgnIzMyScAIyM7MknIDMzCyJltQVGA/Ws/k44AvAKYCAh4EvzGNmTyPlLmNTzXIXM2fYcpexqR1YO8zs31rMnG81UqeFrP9DoBN4B/AS8D3gc7cw75lGyh2NE1j1p8AVwKceZUHXWMWZyq3tjLAvd/ChhvblcPTEde8ClgKnAlOAdcD/jmMu/f5YxKvq7e2t2d5aW1t7Gim3XC7XLLetrW3Ycsvlcjsj7Pu2traG9n13d3fNdtzR0fFMI+WOxvLly/vb8aJFi8asHS9durSdEfblkiVLxqQdd3Z21mzHXV1dTWvHk74HtJ7NRwP3ZaMfBOZT+QNbvZ7NZ+5rucvYNGy5y9hUT7nHAkcOen1vX+sDsJD1VwBfBJYD7wbOB94IfG8h69/QSNn1OoFVbwc+DrxcRLxM0/flcPTEdbOBe7LR3wHeD2wG7tET1x09FjEBent7h21vvb29+9yOy+XysOWWy+Uk7bi7u3vYdtzd3V1IO16+fPl+3Y47OzuHbcednZ1Na8fuAcHlQC/woXnM3JVN+/56Nh8BLAHubrTcxczpL3cZm+ot9yeLmfPLfYw9nP8L3HYL8/6tOmEh6z8MvAJ8APj7Jser5XrgauCPCohV9ZMdfKjZ+3I4vw9sA86PYy7dBaAnrvsE8B7gD4HLxihuf3trbW3tb2+9vb1Na8dtbW395ZbL5brbcVtb25i0446Ojv523N3dPSna8ZIlSwpvx11dXbsAOjs7m96OJ30PCPg00J5LPlWPAG9utNxc8mlWufvsFua9lk8+mRKV7vWYN+wTWPVBKt/alo11rIR2AFuryQcgjrk0gF9k88bKp4H2XPKpako7ziWfZpW7zzo6Ol7LJ59MYe14+fLlk6YdV5MPQFdXV9Pb8aTvAc1j5lZga41ZJwMb97XcxcxptNwHlrHpKCrf6r4JXNnMHtFC1ovKH9H/oXJI5dvNKruWE1h1IPBlYPGjLNh1AqvGMtxgD0zl1gH7cgx7RF8DPqMnrvsT4M+A3cAi4F1UvlWOidbW1jFpx21tbQ2343K5PGDfN7NH1N3dXWg7Xr58eX87XrRo0a7ly5ePZbjBHli6dOmAfTmGPaKvAZ/p7Owc03bsHlAN69n8u1S68k39hrOMTfWUu43Kse1LgN8ArqLyhv9js+qxkPW/BewEfgQcBiy4hXmDv+E222eAJx9lwXfHOE7emO/LweKYS/8dWAD8dyrfFrdkcT8Qx1z69FjFraW3t3dM2nG5XB4X7bi7u3tIO+7o6CikHS9atGi/bsddXV3DtuOurq6mtWP5btgDrWdzO/Ad4C/mMfPyZpWbXd32HeAvFjNnVOUuY9N7gDJw3mLm3NpoXRay/nXA24ETgD8GngfOuYV5Y3KI6ARW/UfgMeA/PcqCH2bTngGuHcur4GqZyq39+3IHH2p4Xw6mJ647ErgfuBe4kco3xwuA3wROiWMufbLZMWvp7e1tJ2tvra2tTWvH2dVt3wH+oq2tbVTllsvl/n3f1tbW8L7v7u6u2Y47OjrGpB0vX768vx0vWrToh9m0Z4Brx/IquFqWLl3avy+XLFnS9Hbc2dk5Yjvu6upqSjue9Ifg8taz+RTgTuBWKg26KZaxqaFyFzPnoWVsugf43ayMhtzCvF8C3we+v5D136HyR/UHwE2Nlj2MLwE3V5NPSjv40ENTubVp+7KGq4EfxDGX/l5u2j164rrbgWup9BzGVG9v75i043K53FC5bW1tD5XL5abt+46Ojv523N3dXVg7riaflJYsWfLQ0qVLx7wdd3V1DWjHnZ2dTW3HTkCZ9Wz+DSrf7G4HPj6PmX3NKHcZmwaUu5g5+1ruD4GjmlGnvFuY97OFrP8WcDZj8Id7AqtOBdqBo09g1eD2Vsqm7X6UBUV2xcdkX2bOovaH8zeAvxqjmP16e3sHtLfW1tamtONyuTyg3La2tnHVjjs6On7W3d09Zu14+fLl/e14+fLlQ9pxNm33okWL3I5HweeAgPVsPhFYSeXb3cfmMXN3M8pdxqYB5S5mTiPlHkHlEMM+W8j6AxayfkqNWS3A6xopewRzgJnAi1SO11dfR1C5lHUnlcZepIb35QimA7U+hPqAg/TEdWP2pa+3t3dAe2ttbW1KOy6XywPKbWtrS9qOu7u7D+ju7nY7TtiOOzs7m9KOJ30PKLsLwmoq3+6amXwGlFtP8lnGpoOodKe7FzPn5tz0o6n8APDD+1qfhayfBfwrcCnwd7nph1D51jhWl/PcSeXE6WB3ZPX4Byrf5JpqKrf278sdfOjm3PSG9+VerKNyWOSrg6Z3AA/kL89upuwuCP3trYnJZ0C59SSfcrncv+/b2tpuzk1veN93d3fXbMfd3d37ZTteunRp/75csmTJzbnpSdtx/vLsRkzqixDWs3kG8CTwOJUfVw3545rHzGdGW+4yNu213MXMeWYZmz5A5XLK86u32FnGpr+m8svuy6n8EvkEKr/63gD8zmLm7PMbtpD1N1NpVJdRaWBvAT4PHAq8+xbmvbavZY9WERchTOXWEfflDj7U9Maf3YbnQeDrVA5V7AY+Afw34L1jcTue3t7evba31tbWZ0Zbbrlc3mu5bW1tz5TL5f52XL3FTrlcHnHft7W17fO+7+7uHrEdd3R0FNaOi7gIYenSpSPuyyVLljS9HWe34Rm2HTfrdjyTvQc0g8rlm4cBm4ZZRmNYblB5Y/MN6JPAU1R+Yf1l4Fkq3+quaST55Mp+ksof7luBfwe+CywpMvkUaNh9ORbJByCOufRxPXHdXOAaKj0HUbmSqC2OufSxsYjJBGvHjSSfXNk123GRyadAw+7LsUg+AF1dXY93dnbWbMddXV1Na8eTugdkZmbp+CIEMzNLwgnIzMyScAIyM7MknIDMzCwJJyAzM0vCCShH0oWTIabj7t9xJ9O2Tra4+9u2OgENlOLNTdKgHHe/jjuZtnWyxd2vttUJyMzMktjvfoj6pa9+JU7+eEfhcb/Vs4ZZ7e8pNOaanrVMa//1QmMCbOxZR6n92MLjPt3zINvajy487s4V/wLHzS405u0zzuak3W8oNCbAqlWrOO644wqPe/fddzNnzpzC45bLZY488shCYz7yyCO0trYWGhPgBz/4AW9+c/FPUf/ODY/w4x3317wTx37XA3r2mZ+kroKZmWW2x/BPDd/vEpCZmU0M+93NSN/AAZwypdiufF/fzwE4dHpbYTFf2loGYErpTYXFBNjd9zIAhxxY7KGwLdsqTwAW0wqNG2wH4KyDLyos5pqtlScKHNByWGExAXbt/hkA75j+O4XGfWrrPwHp3tvTDiruvP69v7oBSPd3+xsHfbzQuOXX/pp//MqJw853D8jMzJJwAjIzsyScgMzMLAknIDMzS8IJyMzMknACMjOzJBpOQJJOkrRa0tXNqJCZmU0Ozfgd0LXAxRHx5L4WIOlg4BvAixHxySbUyczMxrlmHIKb3WDymQmsATY3oS5mZjZBNCMB1bzJ3Cj8ElgC/G0T6mJmZhNE3QkoO9dzl6Q1ku6QdLykHuBwST2STpd0n6RLJK2StFHSxbn1r8jOFd0r6TZJ0wAiYntEfG+YmDXXMTOzia+uBCTpEOBK4LyImA9cAFwfEe3AcxHRHhH3ALuBIyJiAXAqcFUuaXRHxNkRcRqwETi/jtB1rSPpQkkbJG2oZ3vMzCy9entApwAnAndmvZ4VwCxJM2osexNARLwCPAVU76g4S9IKSauBDqCeB6zUtU5E3BARcyNibp3bY2ZmidV7FZyAlRGxaMgMDTkFtCU3vB0oSWoFuoBzIuJ5SRcBh44YcB/WMTOziaPeHtD9wFmSjgWQdJSka0YRZyaVS6yfl9QCnDtG65iZ2QRRVw8oIrZIWgh0ZcmgD7hsFHEeAzZJehB4AVgHTB+DdczMbIKo+4eoEfEwcEaN6XNyw+2D5s3PjQ45fDdo2R6gJzcee1vHzMwmLt8LzszMknACMjOzJJyAzMwsCScgMzNLwgnIzMyScAIyM7MknIDMzCwJJyAzM0sjIvar11VXXRWAX35N2FdV6nr45Vejr4iIhb92XAz3ee0ekJmZJVH3rXgmkmkH1POkh+bZvvNZAKaU3lRYzN19LwPQMqXYG4Tv2v0SAAdNPbLQuL/a8TQAJb2+0Lh98SoA75z+u4XGBTh42lGFxntt+48AuOZtxd4B64ofLwfg9Qe+q9C4r257HIBdfX9XWMyW0kcBmD7tHYXFBNi6/SkAbn/3RwqNuzfuAZmZWRJOQGZmloQTkJmZJeEEZGZmSTgBmZlZEk5AZmaWhBOQmZkl4QRkZmZJNJSAJJ0kabWkq5tVoVzZlzS7TDMzGz8a7QFdC3wqIj67rwVIOljS7ZJuGjTLCcjMbD/WaAKaHRFP7uvKkmYCa4DNDdbDzMwmmEYTkBpc/5fAEuBvGyzHzMwmmLoSUHau5y5JayTdIel4ST3A4ZJ6JJ0u6T5Jl0haJWmjpItz61+RnSu6V9JtkqYBRMT2iPjeCHGvkLQuK/s6SQc1usFmZjY+7DUBSToEuBI4LyLmAxcA10dEO/BcRLRHxD3AbuCIiFgAnApcVU00QHdEnB0RpwEbgfPrqNuRQF9EvDci3ge8AHxhmDpeKGmDpA11lGtmZuNAPT2gU4ATgTuzXs8KYJakGTWWvQkgIl4BngIOy6bPkrRC0mqgA6jneQmbI+KLufEvAufUWjAiboiIuRExt45yzcxsHKjneUACVkbEkIeESENOAW3JDW8HSpJagS7gnIh4XtJFQLEPsTEzs3Gnnh7Q/cBZko4FkHSUpGtGEWMm8GKWfFqAc+tdT9LlufHLge+MIq6ZmY1je+0BRcQWSQuBriyB9AGXjSLGY8AmSQ9SOY+zDphex3pPAy2S/jmL+SjwmVHENTOzcayuR3JHxMPAGTWmz8kNtw+aNz83OuIzfiOiB+gZpuw/raeOZmY2sfhecGZmloQTkJmZJeEEZGZmSTgBmZlZEk5AZmaWhBOQmZkl4QRkZmZJOAGZmVkSiojUdWiqz33uc/H5z38+dTXM9ln1b7LGvRbNJpSIQJpKxI6ajdk9IDMzS6KuW/FMNEU/ty7iVwAc0PJrhcXcueun2VDRb+EuAMS0vSzXXMF2oNh9DHv2s1TP7Quba/q0dxQab+v2pwAolWo9aWXs9PX9HEjXpor8vKh+VqTb1uLb8UjcAzIzsyScgMzMLAknIDMzS8IJyMzMknACMjOzJJyAzMwsCScgMzNLouEEJOkkSaslXd2MCpmZ2eTQjF8xXgtcHBFP7svKqvwK7AbgKCoJ8W8iYnkT6mVmZuNYMw7Bzd7X5JP5X8D/i4iTgfcBH5F0bBPqZWZm41gzekCN3jHxy8BrABGxU9IDwBxgY6MVMzOz8avuHlB2rucuSWsk3SHpeEk9wOGSeiSdLuk+SZdIWiVpo6SLc+tfkZ0rulfSbZKmAUTE5shukJQdjvtN4MGR1jEzs4mvrgQk6RDgSuC8iJgPXABcHxHtwHMR0R4R9wC7gSMiYgFwKnBVLml0R8TZEXEald7N+YNiTAG+AXwzIp6vZ53cuhdK2iBpQ/2bbmZmKdXbAzoFOBG4M+v1rABmSap129ybACLiFeAp4LBs+ixJKyStBjqA2YPWuxF4Gcg/zGdv65DFuiEi5kbE3Dq3x8zMEqv3HJCAlRGxaMiMoQ/N2pIb3g6UJLUCXcA5EfG8pIuAQ3Nl/BlwMPCRyJ7Gtbd1zMxsYqu3B3Q/cFb16jRJR0m6ZhRxZgIvZomkBTi3OkPSZcDRwEcjoq+edczMbOKrqwcUEVskLQS6smTQB1w2ijiPAZskPQi8AKwDqk9GuhJ4EliT6011AbeOsI6ZmU1wdV+GHREPA2fUmD4nN9w+aN783OiQw3fZMm8cIWzNdczMbOLzveDMzCwJJyAzM0vCCcjMzJJwAjIzsyScgMzMLAknIDMzS8IJyMzMknACMjOzJJrxPKBxJ3u6Q+F27vppgqi7EsSEYHuSuGn2MURsLTzm1u1PFR4ToK/v50nipmpTKT4v0m1r8e14JO4BmZnZGOobds5+2QOaNf20QuO9sPVeAEp6fWEx++JVAI4/+COFxQR47LWvZ0NTC40LOwB4y/T2QqM+v7UHAJHiWYhp9nHR21rtDbRMKfZm97t2v5QNFfkxWD1iUfRHbyVumnY8PPeAzMwsCScgMzNLwgnIzMyScAIyM7MknIDMzCwJJyAzM0vCCcjMzJJwAjIzsyQaTkCSTpK0WtLVzaiQmZlNDs34Oe61wMUR8eS+rCzpLcBfAdOp3LNhSUTc14R6mZnZONaMBDR7X5NP5nTgxoi4TdKbgbXAcU2ol5mZjWPNSEBqZOWI+EZu+EVJ2yTNiIg0t+Q1M7NC1H0OKDvXc5ekNZLukHS8pB7gcEk9kk6XdJ+kSyStkrRR0sW59a/IzhXdK+k2SUPuiiepDfhFNflI+lgWb222brF3KzQzszFTVwKSdAhwJXBeRMwHLgCuj4h24LmIaI+Ie4DdwBERsQA4Fbgql2i6I+LsiDgN2Aicnyv/IEkbgO8Cn82FvhdYEBFnALcAF1ODpAslbcjKMDOzCaDeHtApwInAnVmvZwUwS9KMGsveBBARrwBPAYdl02dJWiFpNdABzK6uEBG/ioi5wFzgK9m5IICDga9l63Tm18mLiBsiYm5WhpmZTQD1ngMSsDIiFg2ZoSGngLbkhrcDJUmtQBdwTkQ8L+kiYMjhtIj4saS1wMmSVgK3Ah+MiKckvZ9K4jIzs/1AvT2g+4GzJB0LIOkoSdeMIs5M4MUs+bQA51ZnSPq2pHdkwwcD76NyiO5AIIAfZYt+eBTxzMxsnKurBxQRWyQtBLqyBNIHXDaKOI8BmyQ9CLwArKPyux+AzwM3SoqsPl+KiE0Akv4eeEjSs1TOB500iphmZjaO1X0ZdkQ8DJxRY/qc3HD7oHnzc6NDDt9ly5Sp/Bao1rw/Af6k3jqamdnE4XvBmZlZEk5AZmaWhBOQmZkl4QRkZmZJOAGZmVkSTkBmZpaEE5CZmSXhBGRmZkk043lA484LW+9NErcvXi085mOvfb3wmBU7kkR9fmtPkrjB9gRR0+zjNNsKu3a/lCQu7JokMdO8t9LUYee5B2RmZknslz2gKaU3FRpvd9/L2dDwmb75Kt+OS3p9gTH39PKkgwqNG/GrLO70vSzZ7LhbgeLbFMDrD3xXofFe3fY4AIcceHShcbdsexKYHO9t9bMi1d9ty5Tx9UxP94DMzCwJJyAzM0vCCcjMzJJwAjIzsyScgMzMLAknIDMzS8IJyMzMknACMjOzJBpKQJJOkrRa0tXNqlCu7EuaXaaZmY0fjfaArgU+FRGf3ZeVJb1F0l2S7pXUI+l9udlOQGZm+7FGb8UzOyKebGD904EbI+I2SW8G1gLHNVgnMzObABpNQGpk5Yj4Rm74RUnbJM2IiJ83WC8zMxvn6joEl53ruUvSGkl3SDpeUg9weHbo7HRJ90m6RNIqSRslXZxb/4rsXNG9km6TNK1GjDbgF/nkI+ljktZl631VUvF3hDQzszGx1wQk6RDgSuC8iJgPXABcHxHtwHMR0R4R9wC7gSMiYgFwKnBVLtF0R8TZEXEasBE4P1f+QZI2AN8F8ueSpgAnAadk690F3DRMHS+UtCErx8zMJoB6ekCnACcCd2a9nhXALEkzaix7E0BEvAI8BRyWTZ8laYWk1UAHMLu6QkT8KiLmAnOBr2TnggACuDIi+rLlvgm8XdKBg4NGxA0RMTcrx8zMJoB6zgEJWBkRi4bM0JBTQFtyw9uBkqRWoAs4JyKel3QRMOShFBHxY0lrgZOB20eoi5mZ7Qfq6QHdD5wl6VgASUdJumYUMWYCL2bJpwU4tzpD0rclvSMbPhh4H5VDdFBJNksllbL55wFPRcS2UcQ2M7Nxaq89oIjYImkh0JUlkD7gslHEeAzYJOlB4AVgHVB99OHngRslRVaXL0XEpmzebuAR4D5Ju4GfAJ8cRVwzMxvH6roMOyIeBs6oMX1Obrh90Lz5udEhh++yZcpUfgtUa1617K/WU0czM5tYfC84MzNLwgnIzMyScAIyM7MknIDMzCwJJyAzM0vCCcjMzJJwAjIzsyScgMzMLIlGnwc0Lu3uezlR5B2FR+yLVwuPCRDxq0RxtyaJm6JNvbrt8cJjAmzZ1sgzJvfdZHpvU/3d7tr9UuExI3YPO889IDMzS2K/7AFNKRX73LrqN6hDDjy6sJjVb6klvb6wmJD/5ja10Lh7epdTCo5b+fYmhjxDcczNft3ZhcZ79perAZh2wOy9LNlc23c+C4A0fS9LNle1x1Xk50X1syLVthb9ebE37gGZmVkSTkBmZpaEE5CZmSXhBGRmZkk4AZmZWRJOQGZmloQTkJmZJeEEZGZmSTScgCSdJGm1pKubUSEzM5scmnEnhGuBiyNin24gJakd+Abww9zk346IXzShbmZmNk41IwHN3tfkk/PtiPhkE+piZmYTRDPOAakJZZiZ2SRTdwLKzvXcJWmNpDskHS+pBzhcUo+k0yXdJ+kSSaskbZR0cW79K7JzRfdKuk1S/u6OZ0u6X9IGSWfWuY6ZmU1gdSUgSYcAVwLnRcR84ALg+ohoB56LiPaIuIfKrYOPiIgFwKnAVbmk0R0RZ0fEacBG4Pxs+sPAOyPiVODDwI2S3riXdQbX78IseW0Y3eabmVkq9faATgFOBO7Mej0rgFmSZtRY9iaAiHgFeAo4LJs+S9IKSauBDmB2ttyrEbEtG34aWAmcPNI6g0XEDRExNyLm1rk9ZmaWWL0XIQhYGRGLhszQkFNAW3LD24GSpFagCzgnIp6XdBFw6DCxdgC7RrmOmZlNMPX2gO4HzpJ0LICkoyRdM4o4M4EXs0TSApxbnSHpLyWdlw3PBs4E1o20jpmZTXx19YAiYoukhUBXlgz6gMtGEecxYJOkB4EXqCSY6iMBvwjcJKkT2AlcEBGvShppHTMzm+Dq/h1QRDwMnFFj+pzccPugefNzo0MO32XL/GiYcmO4dczMbOLzveDMzCwJJyAzM0vCCcjMzJJwAjIzsyScgMzMLAknIDMzS8IJyMzMknACMjOzJJrxQLpxZ3ffy0nibtnW6HP5Rq8vXi08ZsWORHF3J4kabC885rO/XF14TIDtO59NEjdia5K4KT4vUm1rms+LncPOcQ/IzMyS2C97QDCl4HjVb+VFxk0RMx+36KazK/s/1fZOLThuipjVXu1ke2+L3N7JtK175x6QmZkl4QRkZmZJOAGZmVkSTkBmZpaEE5CZmSXhBGRmZkk4AZmZWRINJSBJJ0laLenqZlUoV/YlzS7TzMzGj0Z7QNcCn4qIz+7LypLaJb0gqSf3ekM22wnIzGw/1ujPYmdHRKM3QPt2RHyywTLMzGyCabQHpKbUwszMJp26ElB2rucuSWsk3SHpeEk9wOHZYbPTJd0n6RJJqyRtlHRxbv0rsnNF90q6TdK0XPFnS7pf0gZJZw6Ke4WkdVnZ10k6qClbbWZmye01AUk6BLgSOC8i5gMXANdHRDvwXES0R8Q9VO52d0RELABOBa7KJZruiDg7Ik4DNgLnZ9MfBt4ZEacCHwZulPTGbN6RQF9EvDci3ge8AHxhmDpemCWwDaPeA2ZmlkQ9PaBTgBOBO7NezwpglqQZNZa9CSAiXgGeAg7Lps+StELSaqADmJ0t92pEbMuGnwZWAidn62yOiC/myv4icE6tCkbEDRExNyLm1rE9ZmY2DtRzEYKAlRGxaMgMDTkFtCU3vB0oSWoFuoBzIuJ5SRcBhw4Tawd77lduZmb7sXp6QPcDZ0k6FkDSUZKuGUWMmcCLWfJpAc6tzpD0l5LOy4ZnA2cC66rrSbo8V87lwHdGEdfMzMaxvfaAImKLpIVAV5ZA+oDLRhHjMWCTpAepnMdZB0zP5n0RuElSJ5Xntl4Q0f/M2KeBFkn/nMV8FPjMKOKamdk4VtfvgCLiYeCMGtPn5IbbB82bnxsdcvguW+ZHtcodVPaf1lNHMzObWHwvODMzS8IJyMzMknACMjOzJJyAzMwsCScgMzNLwgnIzMyScAIyM7MkGn0e0LjzC3ayLv698Ljf6lnDrPb3FBpzTc9aprX/eqExATb2rKPUfmzhcZ/ueZBt7UcXHnfnin+B42YXFk9PXMftM87m2Wd/VFjMqlWrVnHccccVHvfuu+9mzpw5e1+wycrlMkceeWShMR955BFaW1sLjQmwdu1a1q69r/C4I9nvEpCZmQ1nfN1qUxGRug5NJenjEfHV1PUwM7ORP5P3uwRkZmYTgy9CMDOzJJyAzMwsCV+EkIik44G3knsPIuK2ZBUyMyuYE1ACkm4Gjgc2UnnWEUAATkBmNmn4IoQEJD0eEe9KXQ8zs5TcA0pjnaR3RcTj9SwslQZ9S1CNYWVDys3OTesfqs7Pzxs6PvDf4ZbZU4/+4ahVQrVa+eVrbUFWkobOGzKu2mX0D2v4OAOXiZGXG1KX2Mv8QcsMqWcMWKHmOv3lRn5Rqjt34D6JAXXoj5Hffg0qZ4TxSvkDxweXWY07sKnl1xm4/f3boT3r9tdZQ+eN/Kbl9vmQN3rQcH+T1KBpgwrKF1hjfqhUY3kNWl4g7XkXs2WUX3bQugP/FsVDDz29MiLezyTiBJTG31BJQj8FtlNphhERx9dePBAHZI23hCiBSv3Dqg5rSv94//+aQkktiClIJUq55UoqVaZTolSdVh2nMq80eKfdrx0AAAWcSURBVDhK/fOr4yWUTVNuXNXaVYZVHSc3Dsr936LqvMoO2bMMuWmMflr/vKgxL/rX6x8GpKDEwHWkYMqQ6VFj2eq0yMWNSoz+4T3rSTFondx49v+AdarTRhieUuqrDJeCkvr6p++ZFkh7lsnPk/pGnMbgZUoBuXh7G6aUG5/SV/m/FJVk1D8vqDTzLCGVqi9lTV97Xiplw6XslR8uEaUSlKYMmpYfn1JJMqUpleHqtOp4y9Q9w6UpUGrJhrP/1ZJNa6lMy41LLVA6oPK/WlA2TWqp/F1qSmW4NIUWffTQpn7KTABOQGncDPw+8K/sOQdkZjapOAGl8ZOIuCN1JczMUnICSuNJSV8H7qRyCA7wZdhmNrk4AaVxEJXEsyA3zZdhm9mk4gSUQER8InUdzMxScwJKQNKRwKeAtzLwTggfTFUnM7OiOQGl8U/AX1E5B+Sr4MxsUnICSmNbRPx56kqYmaXkBJTG9ZKuAlYx8Cq4h9NVycysWE5Aafw6lR+insnAm5GemaxGZmYFcwJK41zgbRGxI3VFzMxScQJK41FgBvCzOpdfGew8tHpPx+j/Z0I6FHgpdSWawNsx/kz0bZnIdd8nfhxDApJ6qDwP6EEGngPa7y/DlrQhIuamrkejvB3jz/60LZOFe0BpHEzlMFyVgGsS1cXMLAknoDRaIuKe/ARJB6WqjJlZCk5ABZK0CLgYeJukx3KzDgH+OU2tCndD6go0ibdj/NmftmVS8DmgAkl6A/BG4Grgf+RmbYmIl9PUyswsDScgMzNLopS6Arb/kDRb0lpJT0jaKOnT2fQ3SVot6d+y/9+YTZekP5e0SdJjkk5KuwUDSZoi6fuSvpWNHynpgWw7vilpajZ9Wja+KZv/1pT1HkzSDEndkp7M3pv3TsT3RNIfZe3qB5K+IenAifqeWIUTkDXTLuCyiDgGmAcslvQuKocbvxcRbwe+x57Dj/8FeHv2uhBYXnyVR/Rp4Inc+DXAl7PteAW4IJt+AfBKRMwBvsz4u6LxeuC7EXE0cAKVbZpQ74mkw4FLgLkRcRwwBVjIxH1PDCAi/PJrTF7A7cDZwA+BWdm0WcAPs+GvAL+XW75/udQvoJXKB/OZwLeoXCr/EpUrGAHeC6zMhlcC782GW7LllHobsvq8Hnh6cH0m2nsCHA48C7wp28ffAv7zRHxP/Nrzcg/IxkR2yOPdwAPAYRHxAkD2/3/IFqt+qFT1ZtPGg+uAP2bPvfpmAj+PiF3ZeL6u/duRzf9Ftvx48DbgReCvs8OJN0mazgR7TyLiOeBa4CfAC1T28UNMzPfEMk5A1nSSXgfcClwaEa+OtGiNacmvipH0AeBnEfFQfnKNRaOOeam1ACcByyPi3cBWBl6BOdi43JbsHNVvA0cCbwGmUzlcONhEeE8s4wRkTSXpACrJ5+8j4rZs8r9LmpXNn8Wee+D1ArNzq7cCzxdV1xGcAnxQ0jPALVQOw10HzJBU/e1cvq7925HNfwMwXi6r7wV6I+KBbLybSkKaaO/JfODpiHgxInYCtwEnMzHfE8s4AVnTSBKVJ70+ERFfys26A/hYNvwxKueGqtP/ILvyah7wi+phoZQi4rMR0RoRb6VyovvuiPivwFqgI1ts8HZUt68jW35cfNuOiJ8Cz0p6ZzbpLOBxJth7QuXQ2zxJB2ftrLodE+49sT38OyBrGkmnAvcB/8qecyf/k8p5oH8A/iOVD5IPR8TL2QdJF/B+4DXgExGxofCKj0BSO3B5RHxA0tuo9IjeBHwf+GhEbJd0IPA1Kue8XgYWRsSPU9V5MEknAjcBU4EfA5+g8uVzQr0nkj4PnE/lasvvA5+kcq5nwr0nVuEEZGZmSfgQnJmZJeEEZGZmSTgBmZlZEk5AZmaWhBOQmZkl4QRkZmZJOAGZmVkSTkBmZpbE/wf45CBmRUyLZgAAAABJRU5ErkJggg==\n",
      "text/plain": [
       "<Figure size 460.8x460.8 with 2 Axes>"
      ]
     },
     "metadata": {
      "needs_background": "light"
     },
     "output_type": "display_data"
    }
   ],
   "source": [
    "plotName = 'test'\n",
    "\n",
    "genenames = gdf['symbol'].values\n",
    "samplenames = [] #skipping names to replace in intaglio\n",
    "data = gdf.iloc[0:,7:37].values\n",
    "fig, ax = plt.subplots(figsize=[12,12])\n",
    "im, cbar = heatmap(data, genenames, samplenames, ax=ax,cmap=\"inferno\", cbarlabel=\"fpkm\")\n",
    "# texts = annotate_heatmap(im, valfmt=\"{x:.1f} t\")\n",
    "plt.show()\n",
    "genenamesr = rdf['symbol'].values\n",
    "samplenamesr = [] #skipping names to replace in intaglio\n",
    "datar = rdf.iloc[0:,7:16]\n",
    "figr, axr = plt.subplots(figsize=[3.2,3.2])\n",
    "imr, cbarr = heatmap_rods(datar, genenamesr, samplenamesr, ax=axr, cmap=\"inferno\", cbarlabel=\"cpm\")\n",
    "plt.show()\n",
    "\n",
    "genenamesm = mdf['symbol'].values\n",
    "samplenamesm = [] #skipping names to replace in intaglio\n",
    "datam = mdf.iloc[0:,5:41]\n",
    "figm, axm = plt.subplots(figsize=[6.4,6.4])\n",
    "imm, cbarm = heatmap_glia(datam, genenamesm, samplenamesm, ax=axm, cmap=\"inferno\", cbarlabel=\"tmm\")\n",
    "plt.show()\n",
    "\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "np.zeros([1,8])\n",
    "datar"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Trying to draw counts for single gene by wedges in a circle \n",
    "> Prob not worth it, difficult to interpret"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "fig, ax = plt.subplots(figsize=[12,12])\n",
    "\n",
    "n = 30;\n",
    "r = 0;\n",
    "wdg_theta = 360/30;\n",
    "\n",
    "patches = []\n",
    "circles = []\n",
    "\n",
    "pM = matplotlib.patches.Wedge(0, np.mean(data[r,0:6]), wdg_theta * (0), (wdg_theta * (6)), edgecolor=\"none\", facecolor = pC['r'], alpha = 0.4)\n",
    "patches.append(pM)\n",
    "\n",
    "pM = matplotlib.patches.Wedge(0, np.mean(data[r,7:11]), wdg_theta * (6), (wdg_theta * (11)), edgecolor=\"none\", facecolor = pC['u'], alpha = 0.4)\n",
    "patches.append(pM)\n",
    "\n",
    "pM = matplotlib.patches.Wedge(0, np.mean(data[r,12:17]), wdg_theta * (11), (wdg_theta * (17)), edgecolor=\"none\", facecolor = pC['s'], alpha = 0.4)\n",
    "patches.append(pM)\n",
    "\n",
    "pM = matplotlib.patches.Wedge(0, np.mean(data[r,18:24]), wdg_theta * (17), (wdg_theta * (24)), edgecolor=\"none\", facecolor = pC['m'], alpha = 0.4)\n",
    "patches.append(pM)\n",
    "\n",
    "pM = matplotlib.patches.Wedge(0, np.mean(data[r,25:30]), wdg_theta * (24), (wdg_theta * (30)), edgecolor=\"none\", facecolor = pC['l'], alpha = 0.4)\n",
    "patches.append(pM)\n",
    "\n",
    "for i in range(n):\n",
    "    p = matplotlib.patches.Wedge(0, data[r,i], wdg_theta * (i), (wdg_theta * (i+1)) -2, edgecolor=\"none\", facecolor = pC['plt'][i], alpha = 0.4)\n",
    "    patches.append(p)\n",
    "\n",
    "for i in [0.001,.2,.4,.6,.8,1]:\n",
    "    c = matplotlib.patches.Circle((0,0), np.max(data[r,:])*i, edgecolor = '#000000', facecolor='none', alpha=0.59)\n",
    "    circles.append(c)\n",
    "\n",
    "plt.xlim(-np.max(data[r,:])*1, np.max(data[r,:])*1)\n",
    "plt.ylim(-np.max(data[r,:])*1, np.max(data[r,:])*1)\n",
    "\n",
    "\n",
    "circleColl = matplotlib.collections.PatchCollection(circles, match_original=True);\n",
    "ax.add_collection(circleColl)\n",
    "\n",
    "patchColl = matplotlib.collections.PatchCollection(patches, match_original=True);\n",
    "ax.add_collection(patchColl)\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "np.max(data[r,:])"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "np.max(data[4,i])"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "data"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "range(0,4)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "range?"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### From 2019, Whitaker ,..., Swaroop\n",
    "> genes involved in rod synapse size and localization\n",
    "\n",
    "> [link to biorXiv](https://www.biorxiv.org/content/biorxiv/early/2019/08/30/753012.full.pdf)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "plotName = \"test\"\n",
    "\n",
    "# gdf = findgenelist('dpf') # dpf3 -> not this one but dpf2 in all photoreceptors\n",
    "# gdf = findgenelist('epb') # epb4.1l2 -> not this one but some interesting patterns in this family\n",
    "# gdf = findgenelist('grtp') # grtp1 -> this one is interesting!\n",
    "# gdf = findgenelist('kcnj') # kncj14 -> squirrel too but not zebrafish\n",
    "# gdf = gdf.sort_values(by=[\"baseMean\"],ascending=False)\n",
    "\n",
    "# gdf = findgenelist('llgl') # llgl2 -> nope\n",
    "# gdf = findgenelist('rab') # rab28 -> nope; this family is too big\n",
    "# gdf = gdf[gdf.baseMean > 2]\n",
    "# gdf = gdf.sort_values(by=[\"baseMean\"],ascending=False)\n",
    "\n",
    "# gdf = findgenelist('rom1') #rom1 -> rod-specific; what about outer segments?\n",
    "\n",
    "# gdf = findgenelist('arl') # arl2bp -> nope but arl3l2 in all photoreceptors, and arl3l1 downregulated in uv and s. Both confirmed eye specific (https://zfin.org/ZDB-GENE-040426-1678; https://zfin.org/ZDB-GENE-040426-1649)\n",
    "# These genes are supponsed to interact with unc119c (https://www.ncbi.nlm.nih.gov/pubmed/27079236) and act as their GDF\n",
    "# gdf = gdf[gdf.baseMean > 5]\n",
    "# # gdf = gdf[gdf.baseMean < 300]\n",
    "# gdf = gdf.sort_values(by=[\"baseMean\"],ascending=False)\n",
    "\n",
    "# gdf = findgenelist('unc119') # unc119c is photoreceptor specific (https://www.ncbi.nlm.nih.gov/pubmed/27079236) and CoIPs with arl3l2\n",
    "\n",
    "\n",
    "gdf = findgenelist('ppar') #ppargc2\n",
    "# gdf = findgenelist('dpf')\n",
    "# gdf = findgenelist('dpf')\n",
    "# gdf = findgenelist('dpf')\n",
    "# gdf = findgenelist('dpf')\n",
    "# gdf = findgenelist('dpf')\n",
    "# gdf = findgenelist('dpf')\n",
    "# gdf = findgenelist('dpf')\n",
    "# gdf = findgenelist('dpf')\n",
    "# gdf = findgenelist('dpf')\n",
    "# gdf = findgenelist('dpf')\n",
    "# gdf = findgenelist('dpf')\n",
    "# gdf = findgenelist('dpf')\n",
    "# gdf = findgenelist('dpf')\n",
    "\n",
    "### https://www.sciencedirect.com/science/article/pii/S2211124718303218\n",
    "# gdf = findgenelist('frmpd') #frmpd2 interacts with lrit1 in cones \n",
    "# gdf = findgenelist('sdc') #sdcbp interacts with lrit 1\n",
    "# gdf = findgenelist('grip') #grip1 interacts with lrit1\n",
    "\n",
    "# gdf = findgenelist('dpf')\n",
    "# gdf = gdf.sort_values(by=[\"baseMean\"],ascending=False)\n",
    "# gdf = gdf.drop(np.arange(20,93))\n",
    "# gdf = gdf.drop([0])\n",
    "# gdf = gdf.sort_values(by=[\"symbol\"])\n",
    "# gdf = gdf[gdf.baseMean > .45]\n",
    "\n",
    "gdf\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "genenames = gdf['symbol'].values\n",
    "samplenames = [] #skipping names to replace in intaglio\n",
    "log2matrix = np.vectorize(log2)\n",
    "# data = log2matrix(gdf.iloc[0:,7:15].values)\n",
    "data = gdf.iloc[0:,7:37].values\n",
    "\n",
    "\n",
    "fig, ax = plt.subplots(figsize=[12,12])\n",
    "im, cbar = heatmap(data, genenames, samplenames, ax=ax,\n",
    "                   cmap=\"inferno\", cbarlabel=\"fpkm\")\n",
    "# texts = annotate_heatmap(im, valfmt=\"{x:.1f} t\")\n",
    "\n",
    "plt.savefig(\"{0}.svg\".format(plotName), transparent=True, format=\"svg\")\n",
    "plt.show()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# gdf = gdf[gdf.baseMean > .45]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.7.4"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 4
}
