{
 "cells": [
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Load required modules"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 38,
   "metadata": {},
   "outputs": [],
   "source": [
    "# External libraries\n",
    "import numpy as np\n",
    "import pandas as pd\n",
    "import pygal\n",
    "import math\n",
    "import IPython\n",
    "import matplotlib\n",
    "import matplotlib.pyplot as plt\n",
    "import matplotlib.font_manager as font_manager\n",
    "import importlib\n",
    "import svgutils \n",
    "from svgutils.compose import *\n",
    "#import svgwrite\n",
    "\n",
    "# Custom code\n",
    "from fx_RNAseq import *\n",
    "def findgene(genename):\n",
    "    !head -1 ~/Documents/LiMolec/zfRNAseq/20190827/20190827_DESeq2/00_rodsVcones/rodVCones_raw.csv > ~/Documents/LiMolec/zfRNAseq/20190827/20190827_DESeq2/temp.txt\n",
    "    # EXACT MATCH ONLY by matching word (-w) and only at the start of each line (^) which corresponds to the symbol\n",
    "    !grep -w ^$genename ~/Documents/LiMolec/zfRNAseq/20190827/20190827_DESeq2/00_rodsVcones/rodVCones_raw.csv >> ~/Documents/LiMolec/zfRNAseq/20190827/20190827_DESeq2/temp.txt\n",
    "    gdf = pd.read_csv(\"~/Documents/LiMolec/zfRNAseq/20190827/20190827_DESeq2/temp.txt\")\n",
    "    \n",
    "    return gdf\n",
    "\n",
    "def findgenelist(genename):\n",
    "    !head -1 ~/Documents/LiMolec/zfRNAseq/20190827/20190827_DESeq2/00_rodsVcones/rodVCones_raw.csv > ~/Documents/LiMolec/zfRNAseq/20190827/20190827_DESeq2/temp.txt\n",
    "    # EXACT MATCH ONLY by matching word (-w) and only at the start of each line (^) which corresponds to the symbol\n",
    "    !grep ^$genename ~/Documents/LiMolec/zfRNAseq/20190827/20190827_DESeq2/00_rodsVcones/rodVCones_raw.csv >> ~/Documents/LiMolec/zfRNAseq/20190827/20190827_DESeq2/temp.txt\n",
    "    gdf = pd.read_csv(\"~/Documents/LiMolec/zfRNAseq/20190827/20190827_DESeq2/temp.txt\")\n",
    "    return gdf\n",
    "\n",
    "# def findgene(genename):\n",
    "#     !head -1 ~/Dropbox/LiLab/RNAseq/20190827_DESeq2/00_rodsVcones/rodVCones_raw.csv > ~/Dropbox/LiLab/RNAseq/20190827_DESeq2/temp.txt\n",
    "#     # EXACT MATCH ONLY by matching word (-w) and only at the start of each line (^) which corresponds to the symbol\n",
    "#     !grep -w ^$genename ~/Dropbox/LiLab/RNAseq/20190827_DESeq2/00_rodsVcones/rodVCones_raw.csv >> ~/Dropbox/LiLab/RNAseq/20190827_DESeq2/temp.txt\n",
    "#     gdf = pd.read_csv(\"~/Dropbox/LiLab/RNAseq/20190827_DESeq2/temp.txt\")\n",
    "    \n",
    "#     return gdf\n",
    "\n",
    "# def findgenelist(genename):\n",
    "#     !head -1 ~/Dropbox/LiLab/RNAseq/20190827_DESeq2/00_rodsVcones/rodVCones_raw.csv > ~/Dropbox/LiLab/RNAseq/20190827_DESeq2/temp.txt\n",
    "#     # EXACT MATCH ONLY by matching word (-w) and only at the start of each line (^) which corresponds to the symbol\n",
    "#     !grep ^$genename ~/Dropbox/LiLab/RNAseq/20190827_DESeq2/00_rodsVcones/rodVCones_raw.csv >> ~/Dropbox/LiLab/RNAseq/20190827_DESeq2/temp.txt\n",
    "#     gdf = pd.read_csv(\"~/Dropbox/LiLab/RNAseq/20190827_DESeq2/temp.txt\")\n",
    "#     return gdf\n",
    "\n",
    "def plotNReads():\n",
    "    rdf = pd.read_csv(\"~/Documents/LiMolec/zfRNAseq/20190827/20190827_DESeq2/nMappedReads.csv\")\n",
    "    h = rdf.iloc[:,1]\n",
    "    h = h / 1e6\n",
    "    n = np.arange(1,7) # Rods\n",
    "    n = np.append(n, 6.5 + np.arange(1,6)) # UV\n",
    "    n = np.append(n, 12 + np.arange(1,7)) # S\n",
    "    n = np.append(n, 18.5 + np.arange(1,8)) # M\n",
    "    n = np.append(n, 26 + np.arange(1,7)) # L\n",
    "    return n, h\n",
    "\n",
    "def plotBars(genename):\n",
    "    gdf = findgene(genename)\n",
    "    n = np.arange(1,7) # Rods\n",
    "    n = np.append(n, 6.5 + np.arange(1,6)) # UV\n",
    "    n = np.append(n, 12 + np.arange(1,7)) # S\n",
    "    n = np.append(n, 18.5 + np.arange(1,8)) # M\n",
    "    n = np.append(n, 26 + np.arange(1,7)) # L\n",
    "    h_start = 7\n",
    "    h_end = 37\n",
    "    h = gdf.iloc[0,h_start:h_end]\n",
    "    return n, h\n",
    "\n",
    "def plotBars_list(genelist):\n",
    "    n=[]\n",
    "    h=[]\n",
    "    for i,gene in enumerate(genelist, start=0):\n",
    "        tempn, temph = plotBars(gene)\n",
    "        n = np.append(n,tempn+(30*i))\n",
    "        h = np.append(h,temph)\n",
    "    return n, h\n",
    "\n",
    "def svgreload():\n",
    "    !osascript -e 'tell application \"Google Chrome\" to tell the active tab of second window to reload'\n",
    "\n",
    "from pygal.style import Style\n",
    "custom_style = Style(\n",
    "    background = 'black',\n",
    "    plot_background = '#111111',\n",
    "    foreground = '#999999',\n",
    "    foreground_strong = '#eeeeee',\n",
    "    foreground_subtle = '#555555',\n",
    "    opacity = '.8',\n",
    "    opacity_hover = '.4',\n",
    "    transition = '10ms',\n",
    "    colors = ('#747474','#B540B7', '#4669F2','#04CD22','#CC2C2A'),\n",
    "    label_font_size = 24,\n",
    "    major_label_font_size = 24,\n",
    "    title_font_size = 24,\n",
    "    legend_font_size = 18\n",
    ")\n",
    "\n",
    "#photoreceptor Colors\n",
    "pC = {\n",
    "    'r' : '#747474',\n",
    "    'u' : '#B540B7',\n",
    "    's' : '#4669F2',\n",
    "    'm' : '#04CD22',\n",
    "    'l' : '#CC2C2A',\n",
    "    'plt' : '',\n",
    "}\n",
    "\n",
    "pC['plt']=[\n",
    "    pC['r'],pC['r'],pC['r'],pC['r'],pC['r'],pC['r'],\n",
    "    pC['u'],pC['u'],pC['u'],pC['u'],pC['u'],\n",
    "    pC['s'],pC['s'],pC['s'],pC['s'],pC['s'],pC['s'],\n",
    "    pC['m'],pC['m'],pC['m'],pC['m'],pC['m'],pC['m'],pC['m'],\n",
    "    pC['l'],pC['l'],pC['l'],pC['l'],pC['l'],pC['l']\n",
    "]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 39,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "[Errno 2] No such file or directory: '/Users/angueyraaristjm/Dropbox/LiLab/RNAseq/genePlots/'\n",
      "/Users/angueyraaristjm/Documents/Python/PythonCode/RNAseq\n"
     ]
    }
   ],
   "source": [
    "cd '~/Dropbox/LiLab/RNAseq/genePlots/'"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "***\n",
    "# Index <a id='Index'>∮</a>\n",
    "***\n",
    "- [Plot Total Number of Reads](#nReads)\n",
    "- [BarPlot: Opsins](#barOpsins)\n",
    "- [BarPlot: Custom](#barCustom)\n",
    "- [BarPlot: Many Genes](#barMany)\n",
    "- [BarPlot_subPlotsAutomatic: Custom](#barManyAuto)\n",
    "- [Heatmap: Opsins](#heatOpsins)\n",
    "- [Heatmap: Custom](#heatCustom)\n",
    "***"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "***\n",
    "## Total number of reads<a id='nReads'>∮</a>\n",
    "***\n",
    "[Back to Index](#Index)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### pygal version (does not import well into Intaglio)\n",
    "> hist = pygal.Histogram(x_title='',y_title='million of mapped reads',show_x_labels=False,title='',style=custom_style,legend_at_bottom=True,legend_at_bottom_columns=4,show_legend=False)  \n",
    "> hist.add('Rod',[(11.27,0.2,0.8),(11.77,1.2,1.8),(9.4,2.2,2.8),(14.99,3.2,3.8),(9.75,4.2,4.8),(9.08,5.2,5.8),])  \n",
    "> hist.add('UV',[(9.42,6.2,6.8),(6.52,7.2,7.8),(10.45,8.2,8.8),(9.46,9.2,9.8),(11.02,10.2,10.8)])  \n",
    "> hist.add('S',[(9.23,11.2,11.8),(9.33,12.2,12.8),(9.37,13.2,13.8),(14.95,14.2,14.8),(12.85,15.2,15.8),(10.96,16.2,16.8)])  \n",
    "> hist.add('M',[(10.40,18.2,18.8),(8.634,19.2,19.8),(7.36,20.2,20.8),(10.32,21.2,21.8),(8.33,22.2,22.8),(9.53,23.2,23.8),(9.526,17.2,17.8)])  \n",
    "> hist.add('L',[(8.99,24.2,24.8),(10.50,25.2,25.8),(10.03,26.2,26.8),(11.08,27.2,27.8),(10.44,28.2,28.8),(10.72,29.2,29.8)])  \n",
    "> hist.render_to_file('test.svg')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "params = {\"ytick.color\" : \"w\",\n",
    "          \"xtick.color\" : \"w\",\n",
    "          \"axes.labelcolor\" : \"w\",\n",
    "          \"axes.edgecolor\" : \"w\",\n",
    "         \"axes.linewidth\" : 3,\n",
    "         \"xtick.major.width\" : 3,\n",
    "         \"ytick.major.width\" : 3,\n",
    "         \"xtick.major.size\" : 8,\n",
    "         \"ytick.major.size\" : 8}\n",
    "plt.rcParams.update(params)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "#pyplot version (Sep_2019)\n",
    "plotname = ''\n",
    "n,h = plotNReads()\n",
    "fH, ax = plt.subplots()\n",
    "pH = plt.bar(n, h, width=0.8, bottom=None, align='center', data=None, color=pC['plt'])\n",
    "formatFigure_nReads(plotname, fH, ax, pH)\n",
    "plt.savefig(\"{0}.svg\".format('nReads'), transparent=True, format=\"svg\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "np.mean(h)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": []
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "***\n",
    "## barPlot: Opsins<a id='barOpsins'>∮</a>\n",
    "***\n",
    "[Back to Index](#Index)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# from fx_RNAseq import *"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "genename = 'opn1lw2'\n",
    "n,h = plotBars(genename)\n",
    "h = h / 1e6\n",
    "fH, ax = plt.subplots()\n",
    "pH = plt.bar(n, h, width=0.8, bottom=None, align='center', data=None, color=pC['plt'])\n",
    "formatFigure_Opsins(genename, fH, ax, pH)\n",
    "# plt.savefig(\"{0}.svg\".format(genename), transparent=True, format=\"svg\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "findgene('rho')"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "***\n",
    "## barPlot: Custom Gene<a id='barCustom'>∮</a>\n",
    "***\n",
    "[Back to Index](#Index)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "> Genes to pay attention to but still haven't given interesting results\n",
    "> cdh (cadherins)\n",
    "> protocadherins\n",
    "> semaphorins\n",
    "> plexins (plx)\n",
    "> laminins (lamb) and integrins (itg)\n",
    "> dystroglycan (dag), dystrophin (dmd), pikachurin\n",
    "> neurexins\n",
    "> cadm\n",
    "> dscam and sidekicks (sdk)\n",
    "> pappaa\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "gdf = findgenelist('opn1sw1')\n",
    "gdf"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# genename = 'prickle3'\n",
    "# genename = 'grin1b'\n",
    "genename = 'cdhr1a'\n",
    "genename = 'sema7a'\n",
    "n,h = plotBars(genename)\n",
    "fH, ax = plt.subplots()\n",
    "pH = plt.bar(n, h, width=0.8, bottom=None, align='center', data=None, color=pC['plt'])\n",
    "formatFigure(genename, fH, ax, pH)\n",
    "# plt.savefig(\"{0}.svg\".format(genename), transparent=True, format=\"svg\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "***\n",
    "## barPlot: Many Genes Layout<a id='barMany'>∮</a>\n",
    "***\n",
    "[Back to Index](#Index)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "cd '~/Dropbox/LiLab/RNAseq/genePlots/combos'"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "genelist = ['gnat1','gnat2']\n",
    "n, h = plotBars_list(genelist)\n",
    "fH, ax = plt.subplots()\n",
    "pH = plt.bar(n, h, width=0.8, bottom=None, align='center', data=None, color=pC['plt'])\n",
    "formatFigure_list(genelist, fH, ax, pH)\n",
    "# plt.savefig(\"{0}.svg\".format('gnat'), transparent=True, format=\"svg\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "genelist = ['rho','opn1sw1','opn1sw2','opn1mw1','opn1mw2','opn1mw3','opn1mw4','opn1lw1','opn1lw2']\n",
    "n, h = plotBars_list(genelist)\n",
    "fH, ax = plt.subplots(figsize=[6*len(genelist),4])\n",
    "pH = plt.bar(n, h, width=0.8, bottom=None, align='center', data=None, color=pC['plt'])\n",
    "formatFigure_list(genelist, fH, ax, pH)\n",
    "plt.xlim(0,len(genelist)*10)\n",
    "# plt.savefig(\"{0}.svg\".format('opn1'), transparent=True, format=\"svg\")"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "***\n",
    "## barPlot: Many Genes Layout Using Subplots (proper way)<a id='barManySubPlots'>∮</a>\n",
    "***\n",
    "[Back to Index](#Index)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "shrink = 2/3"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "genelist = ['gnat1','gnat2']\n",
    "fH, axes = plt.subplots(1,len(genelist),sharey=True,figsize=[6*shrink*len(genelist),4*shrink])\n",
    "for i,g in enumerate(genelist,start=0):\n",
    "    n, h = plotBars(g)\n",
    "    plt.sca(axes[i]) \n",
    "    pH = plt.bar(n, h, width=0.8, bottom=None, align='center', data=None,\n",
    "                 color=pC['plt'])\n",
    "    if  i == 0:\n",
    "        formatFigure(g, fH, axes[i], pH)\n",
    "    else:\n",
    "        formatFigure_right(g, fH, axes[i], pH)\n",
    "plt.savefig(\"{0}.svg\".format('gnat'), transparent=True, format=\"svg\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "genelist = ['rho','rhol']\n",
    "\n",
    "fH, axes = plt.subplots(1,len(genelist),sharey=True,figsize=[6*shrink*len(genelist),4*shrink])\n",
    "for i,g in enumerate(genelist,start=0):\n",
    "    n, h = plotBars(g)\n",
    "    plt.sca(axes[i])\n",
    "    pH = plt.bar(n, h, width=0.8, bottom=None, align='center', data=None,\n",
    "                 color=pC['plt'])\n",
    "    if  i == 0:\n",
    "        formatFigure(g, fH, axes[i], pH)\n",
    "        plt.ylim(0,3.5e6)\n",
    "        axes[0].set_yticks(np.arange(0, 4.0e6, step=0.5e6))\n",
    "    else:\n",
    "        formatFigure_right(g, fH, axes[i], pH)\n",
    "plt.savefig(\"{0}.svg\".format('rho'), transparent=True, format=\"svg\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "genelist = ['opn1sw1','opn1sw2']\n",
    "\n",
    "fH, axes = plt.subplots(1,len(genelist),sharey=True,figsize=[6*shrink*len(genelist),4*shrink])\n",
    "for i,g in enumerate(genelist,start=0):\n",
    "    n, h = plotBars(g)\n",
    "    plt.sca(axes[i])\n",
    "    pH = plt.bar(n, h, width=0.8, bottom=None, align='center', data=None,\n",
    "                 color=pC['plt'])\n",
    "    if  i == 0:\n",
    "        formatFigure(g, fH, axes[i], pH)\n",
    "        plt.ylim(0,2e6)\n",
    "        axes[0].set_yticks(np.arange(0, 2.5e6, step=0.5e6))\n",
    "    else:\n",
    "        formatFigure_right(g, fH, axes[i], pH)\n",
    "plt.savefig(\"{0}.svg\".format('opn1sw'), transparent=True, format=\"svg\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "genelist = ['opn1mw1','opn1mw2']\n",
    "\n",
    "fH, axes = plt.subplots(1,len(genelist),sharey=True,figsize=[6*shrink*len(genelist),4*shrink])\n",
    "for i,g in enumerate(genelist,start=0):\n",
    "    n, h = plotBars(g)\n",
    "    plt.sca(axes[i])\n",
    "    pH = plt.bar(n, h, width=0.8, bottom=None, align='center', data=None,\n",
    "                 color=pC['plt'])\n",
    "    if  i == 0:\n",
    "        formatFigure(g, fH, axes[i], pH)\n",
    "    else:\n",
    "        formatFigure_right(g, fH, axes[i], pH)\n",
    "plt.savefig(\"{0}.svg\".format('opn1mwA'), transparent=True, format=\"svg\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "genelist = ['opn1mw3','opn1mw4']\n",
    "\n",
    "fH, axes = plt.subplots(1,len(genelist),sharey=True,figsize=[6*shrink*len(genelist),4*shrink])\n",
    "for i,g in enumerate(genelist,start=0):\n",
    "    n, h = plotBars(g)\n",
    "    plt.sca(axes[i])\n",
    "    pH = plt.bar(n, h, width=0.8, bottom=None, align='center', data=None,\n",
    "                 color=pC['plt'])\n",
    "    if  i == 0:\n",
    "        formatFigure(g, fH, axes[i], pH)\n",
    "        plt.ylim(0,1.5e6)\n",
    "        axes[0].set_yticks(np.arange(0, 2.0e6, step=0.5e6))\n",
    "    else:\n",
    "        formatFigure_right(g, fH, axes[i], pH)\n",
    "\n",
    "\n",
    "plt.savefig(\"{0}.svg\".format('opn1mwB'), transparent=True, format=\"svg\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "genelist = ['opn1lw1','opn1lw2']\n",
    "\n",
    "fH, axes = plt.subplots(1,len(genelist),sharey=True,figsize=[6*shrink*len(genelist),4*shrink])\n",
    "for i,g in enumerate(genelist,start=0):\n",
    "    n, h = plotBars(g)\n",
    "    plt.sca(axes[i])\n",
    "    pH = plt.bar(n, h, width=0.8, bottom=None, align='center', data=None,\n",
    "                 color=pC['plt'])\n",
    "    if  i == 0:\n",
    "        formatFigure(g, fH, axes[i], pH)\n",
    "        plt.ylim(0,1e6)\n",
    "        axes[0].set_yticks(np.arange(0, 1.5e6, step=0.5e6))\n",
    "    else:\n",
    "        formatFigure_right(g, fH, axes[i], pH)\n",
    "#         axes[i].set_yticks(np.arange(0, 2.0e6, step=0.5e6))\n",
    "\n",
    "plt.savefig(\"{0}.svg\".format('opn1lw'), transparent=True, format=\"svg\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Using svgutils to automatically tile these figures the right way\n",
    "Figure(\"2400\", \"2000\",\n",
    "       SVG(\"rho.svg\").move(0, 290*0),\n",
    "       SVG(\"opn1sw.svg\").move(0, 290*1),\n",
    "       SVG(\"opn1mwA.svg\").move(0, 290*2),\n",
    "       SVG(\"opn1mwB.svg\").move(0, 290*3),\n",
    "       SVG(\"opn1lw.svg\").move(0, 290*4)\n",
    "        ).save(\"OpsinsAll.svg\")\n",
    "svgreload()"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Phototransduction"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "gdf = findgenelist('gngt')\n",
    "gdf = gdf.sort_values(by=['symbol'])\n",
    "gdf"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# genelist = ['saga','sagb','arr3a','arr3b']\n",
    "# plotName = 'arrestins'\n",
    "\n",
    "# genelist = ['gngt1','gngt2a','gngt2b']\n",
    "# plotName = 'gngt'\n",
    "\n",
    "# genelist = ['gnb1a','gnb1b','gnb3a','gnb3b','gnb4b','gnb5a','gnb5b']\n",
    "# plotName = 'gnb'\n",
    "\n",
    "# genelist = ['pde6a','pde6b','pde6c','pde6d','pde6ga','pde6gb','pde6c','pde6ha']\n",
    "# plotName = 'pde'\n",
    "\n",
    "# genelist = ['gucy1a1','gucy1b1','gucy1b2','gucy2c','gucy2d','gucy2f','gucy2g']\n",
    "# plotName = 'GC'\n",
    "\n",
    "# genelist = ['gucy2d','gucy2f']\n",
    "# plotName = 'GC_retinaSpecific'\n",
    "\n",
    "# genelist = ['guca1a','guca1b','guca1c','guca1d','guca1e','guca1g']\n",
    "# plotName = 'GCAP'\n",
    "\n",
    "# genelist = ['cnga1a','cnga1b','cnga3a','cnga3b','cnga4','cnga2a','cnga2b']\n",
    "# plotName = 'CNGalpha'\n",
    "\n",
    "# genelist = ['cnga1a','cnga1b','cnga3a','cnga3b']\n",
    "# plotName = 'CNGalpha_retinaSpecific'\n",
    "\n",
    "# genelist = ['cngb1a','cngb3.1','cngb3.2']\n",
    "# plotName = 'CNGbeta'\n",
    "\n",
    "\n",
    "# genelist = ['rgs9a','rgs9b','rgs9bp',\n",
    "#             'rgs3a','rgs6',\n",
    "#             'rgs11','rgs12a','rgs16','rgs20']\n",
    "# plotName = 'RGS'\n",
    "\n",
    "genelist = ['grk1a','grk1b',\n",
    "            'grk3',\n",
    "            'grk4','grk5','grk5l',\n",
    "            'grk6',\n",
    "            'grk7a','grk7b']\n",
    "plotName = 'GRK'\n",
    "\n",
    "# genelist = ['rcvrna','rcvrnb','rcvrn2','rcvrn3','ncalda','ncaldb']\n",
    "# plotName = 'rcvrn'\n",
    "\n",
    "fH, axes = plt.subplots(1,len(genelist),sharey=True,figsize=[6*len(genelist),4])\n",
    "for i,g in enumerate(genelist,start=0):\n",
    "    n, h = plotBars(g)\n",
    "    plt.sca(axes[i])\n",
    "    pH = plt.bar(n, h, width=0.8, bottom=None, align='center', data=None,\n",
    "                 color=pC['plt'])\n",
    "    if  i == 0:\n",
    "        formatFigure(g, fH, axes[i], pH)\n",
    "    else:\n",
    "        formatFigure_right(g, fH, axes[i], pH)\n",
    "plt.savefig(\"{0}.svg\".format(plotName), transparent=True, format=\"svg\")"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Others"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "gdf = findgenelist('prickle')\n",
    "gdf = gdf.sort_values(by=['symbol'])\n",
    "gdf"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# genelist = ['ctbp1','ctbp2a']\n",
    "# plotName = 'ctbp'\n",
    "\n",
    "# genelist = ['cadm1a','cadm1b','cadm2a','cadm2b','cadm3','cadm4']\n",
    "# plotName = 'cadm'\n",
    "\n",
    "# genelist = ['tjp1a','tjp1b','tjp2a','tjp2b','tjp3']\n",
    "# plotName = 'tjp'\n",
    "\n",
    "# genelist = ['nrxn1a','nrxn1b','nrxn2a','nrxn2b','nrxn3a','nrxn3b',]\n",
    "# plotName = 'nrxn'\n",
    "\n",
    "# genelist = ['eml1','eml2','eml3']\n",
    "# plotName = 'eml'\n",
    "\n",
    "# genelist = ['cpne1','cpne2','cpne3','cpne4a','cpne4b','cpne5a','cpne5b','cpne7','cpne8','cpne9']\n",
    "# plotName = 'cpne'\n",
    "\n",
    "# genelist = ['tbx2a','tbx2b','tbxas1','tbx21']\n",
    "# genelist = ['tbx2a','tbx2b']\n",
    "# plotName = 'tbx2'\n",
    "\n",
    "# genelist = ['chn1','chn2']\n",
    "# plotName = 'chn'\n",
    "\n",
    "# genelist = ['ncdn','ncdn']\n",
    "# plotName = 'ncdn'\n",
    "\n",
    "# genelist = ['sema7a','sema6a','sema6d','sema4ab','sema4ba','sema3fa','sema3fb']\n",
    "# plotName = 'sema'\n",
    "\n",
    "# genelist = ['cacna1aa','cacna1ab',\n",
    "#             'cacna1da','cacna1db',\n",
    "#             'cacna1fa','cacna1fb',\n",
    "#             'cacna2d1a','cacna2d2a',\n",
    "#             'cacna2d4a','cacna2d4b',\n",
    "#             'cacna1sb',\n",
    "#            ]\n",
    "# plotName = 'CaV'\n",
    "\n",
    "# genelist = ['stxbp1b','stxbp4','stxbp2','stxbp5a','stxbp5b','stxbp5l','stxbp3']\n",
    "# genelist = ['stxbp1a','stxbp1b',\n",
    "#             'stxbp2','stxbp3','stxbp4',\n",
    "#             'stxbp5a','stxbp5b','stxbp5l',\n",
    "#             'stxbp6','stxbp6l']\n",
    "# plotName = 'stxBP'\n",
    "\n",
    "# genelist = ['stx1a','stx1b','stx2a','stx2b','stx3a',\n",
    "#             'stx4','stx5a','stx5al','stx6','stx7l',\n",
    "#             'stx8','stx10','stx11a','stx11b.1','stx11b.2',\n",
    "#             'stx12','stx12l','stx16','stx17','stx18']\n",
    "# plotName = 'stx'\n",
    "\n",
    "genelist = ['prickle1a','prickle1b','prickle2a','prickle2b','prickle3',]\n",
    "plotName = 'prickle'\n",
    "\n",
    "            \n",
    "fH, axes = plt.subplots(1,len(genelist),sharey=False,figsize=[6*len(genelist),4])\n",
    "for i,g in enumerate(genelist,start=0):\n",
    "    n, h = plotBars(g)\n",
    "    plt.sca(axes[i])\n",
    "    pH = plt.bar(n, h, width=0.8, bottom=None, align='center', data=None,\n",
    "                 color=pC['plt'])\n",
    "    if  i == 0:\n",
    "        formatFigure(g, fH, axes[i], pH)\n",
    "    else:\n",
    "        formatFigure_right(g, fH, axes[i], pH)\n",
    "plt.savefig(\"{0}.svg\".format(plotName), transparent=True, format=\"svg\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "gdf = findgenelist('opn1')\n",
    "# gdf = gdf.sort_values(by=['symbol'])\n",
    "# gdf = gdf.sort_values(by=['baseMean'], ascending=False)\n",
    "# gdf['symbol']\n",
    "gdf"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### According to Neuhauss recoverin1a in enriched in rods and UV-cones"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "***\n",
    "## heatMap: Opsins<a id='barManySubPlots'>∮</a>\n",
    "***\n",
    "[Back to Index](#Index)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Not sure this is helpful.  \n",
    "Maybe use this to show differential expression by normalizing by mean expression in correct type"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "gdf = findgenelist('opn1')\n",
    "gdf2 = findgene('rho')\n",
    "gdf2\n",
    "gdf = gdf.append(gdf2)\n",
    "gdf.index = range(len(gdf))\n",
    "gdf\n",
    "\n",
    "Opsins = [\"Rhodopsin (rho)\",\"UV-opsin (opn1sw1)\", \"S-opsin (opn1sw2)\",\"M-opsin (opn1mw1)\",\"M-opsin (opn1mw2)\",\"M-opsin (opn1mw3)\",\"M-opsin (opn1mw4)\",\"L-opsin (opn1lw1)\",\"L-opsin (opn1lw2)\"]\n",
    "Samples = [\"\",\"\",\"R\",\"\",\"\",\"\",\n",
    "           \"\",\"\",\"U\",\"\",\"\",\n",
    "           \"\",\"\",\"S\",\"\",\"\",\"\",\n",
    "           \"\",\"\",\"\",\"M\",\"\",\"\",\"\",\n",
    "           \"\",\"\",\"L\",\"\",\"\",\"\",\n",
    "          ]\n",
    "\n",
    "data = gdf.iloc[0:9,7:37]\n",
    "data = data.reindex([8,4,5,7,0,2,6,1,3])\n",
    "data.index = range(len(data))\n",
    "np.around(np.divide(data.values,1e6),2)\n",
    "opsinData = np.divide(data.values,1e6)\n",
    "\n",
    "# # log2 representation does not work well\n",
    "# log2matrix = np.vectorize(log2)\n",
    "# opsinData = log2matrix(opsinData)\n",
    "\n",
    "params = {\"ytick.color\" : \"w\",\n",
    "          \"xtick.color\" : \"w\",\n",
    "          \"axes.labelcolor\" : \"w\",\n",
    "          \"axes.edgecolor\" : \"w\",\n",
    "         \"axes.linewidth\" : 2,\n",
    "         \"xtick.major.width\" : 2,\n",
    "         \"ytick.major.width\" : 2,\n",
    "         \"xtick.major.size\" : 0,\n",
    "         \"ytick.major.size\" : 0}\n",
    "plt.rcParams.update(params)\n",
    "\n",
    "plotName = \"heatMap_opsins\"\n",
    "fig, ax = plt.subplots()\n",
    "\n",
    "im, cbar = heatmap(opsinData, Opsins, Samples, ax=ax,\n",
    "                   cmap=\"inferno\", cbarlabel=\"counts per million\")\n",
    "\n",
    "\n",
    "# texts = annotate_heatmap(im, valfmt=\"{x:.1f} t\")\n",
    "\n",
    "# fig.tight_layout()\n",
    "plt.savefig(\"{0}.svg\".format(plotName), transparent=True, format=\"svg\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "***\n",
    "## heatMap: Custom<a id='heatCustom'>∮</a>\n",
    "***\n",
    "[Back to Index](#Index)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 144,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>symbol</th>\n",
       "      <th>genename</th>\n",
       "      <th>baseMean</th>\n",
       "      <th>log2FoldChange</th>\n",
       "      <th>lfcSE</th>\n",
       "      <th>pvalue</th>\n",
       "      <th>padj</th>\n",
       "      <th>R1</th>\n",
       "      <th>R2</th>\n",
       "      <th>R3</th>\n",
       "      <th>...</th>\n",
       "      <th>M4</th>\n",
       "      <th>M5</th>\n",
       "      <th>M6</th>\n",
       "      <th>S7</th>\n",
       "      <th>L1</th>\n",
       "      <th>L2</th>\n",
       "      <th>L3</th>\n",
       "      <th>L4</th>\n",
       "      <th>L5</th>\n",
       "      <th>L6</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <td>0</td>\n",
       "      <td>nrxn3a</td>\n",
       "      <td>neurexin 3a</td>\n",
       "      <td>48.609829</td>\n",
       "      <td>-5.507985e-07</td>\n",
       "      <td>0.001443</td>\n",
       "      <td>0.183900</td>\n",
       "      <td>0.45802</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>...</td>\n",
       "      <td>227.843299</td>\n",
       "      <td>350.963367</td>\n",
       "      <td>6.140467</td>\n",
       "      <td>33.887915</td>\n",
       "      <td>0</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>98.667355</td>\n",
       "      <td>9.084001</td>\n",
       "      <td>12.011948</td>\n",
       "      <td>27.457751</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>1</td>\n",
       "      <td>nrxn3b</td>\n",
       "      <td>neurexin 3b</td>\n",
       "      <td>57.079152</td>\n",
       "      <td>1.203286e-07</td>\n",
       "      <td>0.001443</td>\n",
       "      <td>0.693287</td>\n",
       "      <td>0.85899</td>\n",
       "      <td>190.283702</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>...</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0</td>\n",
       "      <td>2.378322</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>167.704627</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>177.298618</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>2 rows × 37 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "   symbol     genename   baseMean  log2FoldChange     lfcSE    pvalue  \\\n",
       "0  nrxn3a  neurexin 3a  48.609829   -5.507985e-07  0.001443  0.183900   \n",
       "1  nrxn3b  neurexin 3b  57.079152    1.203286e-07  0.001443  0.693287   \n",
       "\n",
       "      padj          R1  R2  R3  ...          M4          M5        M6  \\\n",
       "0  0.45802    0.000000   0   0  ...  227.843299  350.963367  6.140467   \n",
       "1  0.85899  190.283702   0   0  ...    0.000000    0.000000  0.000000   \n",
       "\n",
       "          S7  L1        L2         L3          L4         L5          L6  \n",
       "0  33.887915   0  0.000000  98.667355    9.084001  12.011948   27.457751  \n",
       "1   0.000000   0  2.378322   0.000000  167.704627   0.000000  177.298618  \n",
       "\n",
       "[2 rows x 37 columns]"
      ]
     },
     "execution_count": 144,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "plotName = \"test\"\n",
    "\n",
    "# gdf = findgenelist('sema')\n",
    "# gdf = gdf.sort_values(by=[\"symbol\"])\n",
    "# plotName = \"heatMap_sema\"\n",
    "\n",
    "# gdf = findgenelist('cdh')\n",
    "# gdf = gdf.sort_values(by=[\"baseMean\"],ascending=False)\n",
    "# plotName = \"heatMap_cdh\"\n",
    "\n",
    "# gdf = findgenelist('eph')\n",
    "# gdf = gdf.drop([1,14,15,16])\n",
    "# gdf = gdf.sort_values(by=[\"symbol\"])\n",
    "# plotName = \"heatMap_eph\"\n",
    "\n",
    "\n",
    "# gdf = findgenelist('efn')\n",
    "# gdf = gdf.sort_values(by=[\"symbol\"])\n",
    "# plotName = \"heatMap_efn\"\n",
    "\n",
    "# gdf = findgenelist('pdgf')\n",
    "# gdf = findgenelist('rgr')\n",
    "# gdf = findgenelist('lmo')\n",
    "# gdf.index = range(len(gdf))\n",
    "# gdf.sort_values(by=[\"log2FoldChange\"])\n",
    "\n",
    "\n",
    "## genes in 2019 Papal, Swaroop: # igsf11, epha10, tuft1, prtg, lrfn2, vasn\n",
    "\n",
    "# gdf = findgenelist('igsf')\n",
    "# gdf = findgenelist('tuft')\n",
    "# gdf = findgenelist('prtg')\n",
    "# gdf = findgenelist('lrfn')\n",
    "# gdf = findgenelist('vasn')\n",
    "\n",
    "# gdf = findgenelist('rdh')\n",
    "# gdf = gdf.sort_values(by=[\"symbol\"])\n",
    "# plotName = \"rdhHeatMap\"\n",
    "\n",
    "# gdf = findgenelist('cyp2')\n",
    "# gdf = findgenelist('cyb')\n",
    "\n",
    "# gdf = findgenelist('sag')\n",
    "# plotName = \"heatMap_sag\"\n",
    "\n",
    "# gdf = findgenelist('dio')\n",
    "# plotName = \"heatMap_dio\"\n",
    "\n",
    "# gdf = findgenelist('lrrn')\n",
    "# plotName = \"heatMap_lrrn\"\n",
    "# gdf = findgenelist('dnmt')\n",
    "# plotName = \"heatMap_dnmt\"\n",
    "# gdf = findgenelist('tet')\n",
    "\n",
    "# gdf = findgenelist('isl')\n",
    "# gdf = gdf.drop([1])\n",
    "# gdf = findgenelist('gj')\n",
    "# plotName = \"heatMap_gj\"\n",
    "\n",
    "# gdf = findgenelist('fgf')\n",
    "# gdf = findgenelist('bmp')\n",
    "# gdf = findgenelist('fzd')\n",
    "# gdf = findgenelist('edar')\n",
    "# gdf = findgenelist('axin')\n",
    "\n",
    "# gdf = findgenelist('cadm') # cadm2a in S-Cones probably(?) and cadm3 in all photoreceptors (Hunter, JCN, 2011)\n",
    "# gdf = gdf.sort_values(by=[\"symbol\"])\n",
    "# plotName = \"heatMap_cadm\"\n",
    "\n",
    "# gdf = findgenelist('lrit') # lrit1b for rods and cones; lrit2 for cones\n",
    "# gdf = gdf.sort_values(by=[\"symbol\"])\n",
    "# plotName = \"heatMap_lrit\"\n",
    "\n",
    "# gdf = findgenelist('elfn') # not detected\n",
    "# gdf = gdf.sort_values(by=[\"symbol\"])\n",
    "# plotName = \"heatMap_elfn\"\n",
    "\n",
    "# gdf = findgenelist('tenm') # for RGCs in zebrafish (Antinucci, Cell Reports, 2013)\n",
    "\n",
    "# gdf = findgenelist('foxo')\n",
    "\n",
    "# gdf = findgenelist('sdk')\n",
    "\n",
    "# gdf = findgenelist('crmp')\n",
    "\n",
    "\n",
    "\n",
    "# Even More\n",
    "# gdf = findgenelist('thr')\n",
    "# gdf = gdf.sort_values(by=[\"symbol\"])\n",
    "# plotName = \"heatMap_thr\"\n",
    "\n",
    "# gdf = findgenelist('lingo')\n",
    "# gdf = gdf.sort_values(by=[\"symbol\"])\n",
    "# plotName = \"heatMap_lingo\"\n",
    "\n",
    "# gdf = findgenelist('mitf')\n",
    "\n",
    "# gdf = findgenelist('prickle')\n",
    "# gdf = gdf.sort_values(by=[\"symbol\"])\n",
    "# plotName = \"heatMap_prickle\"\n",
    "\n",
    "# gdf = findgenelist('rdh')\n",
    "# gdf = findgenelist('sdh')\n",
    "# lingo, rx, mitf\n",
    "\n",
    "# gdf = findgenelist('dag')\n",
    "# can't find pikachurin: # gdf = findgenelist('egflam') gdf = findgenelist('si:dkeyp-100d11.2-001')\n",
    "\n",
    "# gdf = findgenelist('kcn')\n",
    "# gdf = gdf.sort_values(by=[\"baseMean\"], ascending=False)\n",
    "# gdf = gdf.drop(np.arange(20,109))\n",
    "# gdf\n",
    "\n",
    "# gdf = findgenelist('kcnj')\n",
    "# gdf = gdf.sort_values(by=[\"baseMean\"], ascending=False)\n",
    "# gdf\n",
    "\n",
    "# gdf = findgenelist('slc')\n",
    "# gdf = gdf.sort_values(by=[\"baseMean\"], ascending=False)\n",
    "# gdf = gdf.drop(np.arange(40,410))\n",
    "\n",
    "# gdf = findgenelist('rcv')\n",
    "# gdf = gdf.sort_values(by=[\"symbol\"])\n",
    "# plotName = \"heatMap_recoverin\"\n",
    "\n",
    "# gdf = findgenelist('sag')\n",
    "# gdf = gdf.append(findgenelist('arrb'))\n",
    "# gdf = gdf.append(findgenelist('arr3'))\n",
    "# gdf = gdf.sort_values(by=[\"symbol\"])\n",
    "# plotName = \"heatMap_arr\"\n",
    "\n",
    "\n",
    "# gdf = findgenelist('cng')\n",
    "# gdf = gdf.sort_values(by=[\"symbol\"])\n",
    "# plotName = \"heatMap_cng\"\n",
    "\n",
    "# gdf = findgenelist('cnga')\n",
    "# gdf = gdf.sort_values(by=[\"symbol\"])\n",
    "# plotName = \"heatMap_cnga\"\n",
    "\n",
    "# gdf = findgenelist('cngb')\n",
    "# gdf = gdf.sort_values(by=[\"symbol\"])\n",
    "# plotName = \"heatMap_cngb\"\n",
    "\n",
    "# gdf = findgenelist('myo')\n",
    "# gdf = gdf.sort_values(by=[\"symbol\"])\n",
    "\n",
    "# gdf = findgenelist('crx')\n",
    "# gdf = gdf.sort_values(by=[\"baseMean\"], ascending=False)\n",
    "# plotName = \"heatMap_crx\"\n",
    "\n",
    "\n",
    "# gdf = findgenelist('neurod')\n",
    "# gdf = gdf.sort_values(by=[\"symbol\"])\n",
    "# plotName = \"heatMap_neurod\"\n",
    "\n",
    "# gdf = findgenelist('ror')\n",
    "# gdf = gdf.drop([5,6])\n",
    "# gdf = gdf.sort_values(by=[\"symbol\"])\n",
    "# plotName = \"heatMap_ror\"\n",
    "\n",
    "\n",
    "# gdf = findgenelist('nr')\n",
    "# gdf = gdf.sort_values(by=[\"baseMean\"], ascending=False)\n",
    "# gdf = gdf.drop(np.arange(5,76))\n",
    "# plotName = \"heatMap_nrl\"\n",
    "\n",
    "\n",
    "# gdf = findgenelist('pcdh')\n",
    "# gdf = gdf.drop([1,15]) #highest ones that obscure others.\n",
    "\n",
    "# gdf = findgenelist('hcn')\n",
    "# gdf = gdf.sort_values(by=[\"symbol\"])\n",
    "\n",
    "# gdf = findgenelist('gabr')\n",
    "# gdf = gdf.drop([0,2]) #highest ones that obscure others.\n",
    "# gdf = gdf.sort_values(by=[\"symbol\"])\n",
    "\n",
    "# gdf = findgenelist('opn4')\n",
    "# gdf = gdf.drop([1,2]) #highest ones that obscure others.\n",
    "# gdf = gdf.sort_values(by=[\"baseMean\"], ascending=False)\n",
    "# gdf = gdf.drop(np.arange(5,76))\n",
    "# gdf = gdf.sort_values(by=[\"symbol\"])\n",
    "# plotName = \"heatMap_ror\"\n",
    "\n",
    "# CSNB causes: Nyx, cacna1f, GPR179 (Kamermans, 2019)\n",
    "# gdf = findgenelist('nyx') # bipolar cell -> not expressed in photoreceptors\n",
    "# gdf = findgenelist('gpr1') # bipolar cell -> not expressed in photoreceptors\n",
    "# gdf = findgenelist('cacna') # bipolar cell -> not expressed in photoreceptors\n",
    "\n",
    "\n",
    "# For Karen Carleton\n",
    "\n",
    "# gdf = findgenelist('rx')\n",
    "# gdf = gdf.drop([2,3,4,5,7,8,9,10,11,12,13,14,15,16,17,18,19])\n",
    "# gdf = gdf.sort_values(by=[\"symbol\"])\n",
    "# # gdf = gdf.drop([1])\n",
    "# plotName = \"heatMap_rx\"\n",
    "\n",
    "# gdf = findgenelist('tbx')\n",
    "# gdf = gdf.drop([3,17])\n",
    "# plotName = \"heatMap_tbx\"\n",
    "\n",
    "# gdf = findgenelist('tbx2')\n",
    "# gdf = gdf.drop([0,3,4])\n",
    "# gdf = gdf.sort_values(by=[\"symbol\"])\n",
    "# plotName = \"heatMap_tbx2\"\n",
    "\n",
    "# gdf = findgenelist('six')\n",
    "# gdf = gdf.sort_values(by=[\"symbol\"])\n",
    "# plotName = \"heatMap_six\"\n",
    "\n",
    "# gdf = findgenelist('lin')\n",
    "# gdf = gdf.drop([1,3,4,5,10,11,13,16])\n",
    "# plotName = \"heatMap_lin\"\n",
    "\n",
    "# For Mark Hankins\n",
    "# gdf = findgenelist('opn')\n",
    "# gdf = gdf.sort_values(by=[\"symbol\"])\n",
    "# gdf = gdf.drop([1,4,24,0,3,23,5,11,2])\n",
    "\n",
    "# gdf = findgenelist('rgr')\n",
    "# gdf = findgenelist('rrh')\n",
    "\n",
    "# gdf = findgenelist('ndrg') #potential target of tbx2\n",
    "\n",
    "gdf = findgenelist('nrxn3')\n",
    "# gdf = gdf.drop([0])\n",
    "# gdf = gdf.sort_values(by=[\"symbol\"])\n",
    "\n",
    "gdf"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 145,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "image/png": "iVBORw0KGgoAAAANSUhEUgAAAtgAAACFCAYAAABhXI8FAAAABHNCSVQICAgIfAhkiAAAAAlwSFlzAAALEgAACxIB0t1+/AAAADh0RVh0U29mdHdhcmUAbWF0cGxvdGxpYiB2ZXJzaW9uMy4xLjAsIGh0dHA6Ly9tYXRwbG90bGliLm9yZy+17YcXAAAZ/UlEQVR4nO3de5wcdZnv8c/TyYAEJlwMN4kQTmAVAhjGqFxUMiDM7CwKuqi4C4YERV+jh8V167DicQ2u7qLlgso6eOAFJCiKiKKAY4YAI+CJgiQbLuEm4WaEJRu5DSSEJPPsH1U9U+npnunJVk315ft+veY1Xb/61dNP/+bX3c9UV1WbuyMiIiIiIuko5J2AiIiIiEgjUYEtIiIiIpIiFdgiIiIiIilSgS0iIiIikiIV2CIiIiIiKZqcdwKSnyAIdgHOAb4VhuGLeecjIiIi0gi0B7u57QJ8Of4tIiIiIinQHuwaEwTBXKA/0eTAM8B/AP8ahuGyPPKSidfbvmQhcGJXf+ec0dYDq4DZXf2dbxslVifwS2BGV3/nH9PPVrLU0T1wKPB/gaOBacATwA+BsK+n9bU8c5Pa1LJi5lyi95JngX03ta3eXKFfAXgc2A94/6a21TdNWJJSN5YfOmsu0Xxqffv9q17JOZ26oD3YtWsWsD/wVmAB8DpwZxAEHblmJbVoEXBYb/uSw0bpMw+4RcV1/enoHjgMWAZsAD4KHA58FwiAc3NMTerDnkT/iFfSAewzQbmINA3twa5dT4dhWPwv8dEgCJYCvwU+D/Tll5bUoF8DTwGnAf+ndGVv+5KdgZOBMyY0K0nLvwJ39/W0npFoe6ije+Bm4A/5pCR1pBc4C/h5hfVnAdcDH56wjESagPZg14kwDB24E9g371yktnT1dzpwFfCx3vYl5Z7THwXWU/kNVmrbAUSHiG2lr6f10b6eVs8hH6kvPUBHy4qZI947WlbM3Ito7/YlE56VSIPTHuz6cijwZHEhCIIjgS8RHZc5mehN+JthGG5VSAVBsD3RR8mnER1n9yeiguyakn4FYCHR4QR7Ag8AQRiGyWPCpTYtJpoLc4HbStbNA37Y1d+5caKTklQ8BZzY0T1wQV9P67q8k5G6cyfwIHAm0UntSQuAxza1re5vWTFzwhMTaWTag13jgiAoBEHw5iAILiA6Vu47cftxwO1EBXc78E7gFuDaIAj+IbG9ATcAnyA6SepQoo8E3wd8v+TuPk/0IvxpoI3oRfnyIAhasnp8ko6u/s7VwP8n+idqSG/7kgOAo4Ar8shLUvEFok+uHuroHgg6ugf0KZaM1/eABS0rZk4qNrSsmGlEr/ffyy0rkQamPdi1a10QBBD9jQrAfcCHwzDsDYJgO+BS4KIwDJMnOa0KguBp4HtBEFwbhuHTRC+gRwJvDcPwmbjfo0EQ3AYsLbnPdwO/DMPwVwBBECwAdg3DcFNGj1HStQj4t972Jd1d/Z3FK0ucAdzb1d854hADqQ99Pa3L4xMd/4loD+TXO7oHbgG+3tfTemu+2UmduAq4AOgCbozbjgf2Jvr0S0RSpj3YtesoYDbRnsc1wNFhGF4Xrzsa+F/AhWW2uxIYAD4QL58OXJ0orgEIw3AQOL9k22XAqUEQnBsEwSHAljAM/yuNByPbpJrja5N9rgVaiP/2ve1LjOjvr73Xda6vp/Wxvp7WjwO7Ax8BdgVu6egeOG30LUVgU9vqAeBq4JOJ5k8CP9rUtlpfMiaSARXYtevRMAwfJroqxPZEe6+K9gY2hGH4XOlG8cmQjwNvSvR9otJ9lG4OfJXomN37gTVBEHxyxFYyUV4iKqQqmQYMvTl29Xe+THQ1gGLR1Q7sRfTGKg2gr6d1Q19P63VE/4D3E50zIVKNHqCrZcXMfVpWzNwDOAmd3CiSGRXYNS7+CvNzgc8FQTArbn4W2CEIgj1K+8fHXO8f9yn2nVEh/IEl9zUYhuE3wjA8GJhO9NHh/wuCoOwXnUjmngSm97YvmVph/SwSJ73GFgOdve1L3kj0j9Ivuvo7/5xZhpKLvp7WTcB1RM9TkTFtalt9H3AX0WGD84F7N7WtviffrEQalwrs+rAYuBu4JC6glxEVVp8r0/fjwM5EJzZCtPfyb4Ig2DvZKY7zpZK2/YIg2BkgDMM/hWF4HtFe1IrfECiZ6gc2AueUruhtX/IO4BhGXnrvFmAt0ZvoX6PDQ+peR/dAR0f3wMFlVr0HWDnR+Uhdu4ToteETaO+1SKZ0kmMdCMPQgyD4DLAcmBeG4aIgCM4CbgyCYEfgx0SFWBdwHvDFMAyfije/jOg6yMuCIPgi0cmS+xEdelJ6dZCfAIPxVUieJzrUYAfgN5k+QCmrq7/zpd72JZ8DLu1tX7IL0VdjvwocS3T8/E+7+jt/WbLNYG/7ku8DXwH+C7h5gtOWFHV0D0wmOgxkZkf3wJeILrnWSlQknUh0YrJItX4CXARMpeQyrSJV2nf5obPWl7T52+9f9VTZ3k1Me7DrRBiG9xJ9PXIYBMFuYRguBd4B/AXwK+AOoBP4qzAMv5HYzoku73cp8M9ERfrFRNdKPr3kbk4GVgO/AH5PdAzvB8IwfCTDhyaj6OrvvBx4P9FlE28FVgDdwNeBUytstpjouP3FXf2dgxORp2Sjr6d1M3Ac0eU5P0v0/P0pMAV4e19P6705pid1ZlPb6o1En2ot3tS2urRIEqnGKqLzupI/q3PNqEaZu74ITEREREQkLdqDLSIiIiKSIhXYIiIiIiIpUoEtIiIiIpKipi6wzewsxc02dr3FzTJ2vcXNMna9xc0ydr3FzTK24mYfu97iZhm73uJmGbve4mYZO624TV1gA1n94estbpax6y1ulrHrLW6Wsestbpax6y1ulrEVN/vY9RY3y9j1FjfL2PUWN8vYKrBFRERERGpNw12m78QTT/SDDjoo7zSkhsz8zwPYd81+each2+jmA3bmoZZD8k5Dasy6d53HfYf25p2GNImvXPYmTvjN83mnITXmPY+8zPrX/2jl1jXcHuxdd9017xSkxuw3Q8V1PZux34y8U5AaNGPGjLxTkCai+SblDPrrFdc1XIEtIiIiIpKnhjtEZOHChX7++V9NOeqW+PekOombZewo7rt2mJ9q1Ls2XAmA2Q6pxnXfgLtjNjnVuJFsx3iH7fZPNeqG15+Ib9XXfMvm71e/z739dupINepTr/QBsP789D/9m/LlFwCYPeW0VOOuXP8D3J17Tzg+1bizl94CwFt2/FCqcQEeefVn8a1s5sVzCw5ONeqeVzwIwObBH6QaF2ByIZoPWbzeA0zb8Z2pxl336t11+z6yYNrZqUa9Yt13ACjY1FTjAgz6ywBMKuyWatwtg9GhPTvvcGiqcV9cfx9vaNmT1zY91xyHiIiIiIiI5EkFtoiIiIhIilRgi4iIiIikKPcC28zebWbLzOzXZnajme2Td04iIiIiItsq9wIb6ALmuftc4FtAmG86IiIiIiLbLrUC28xatmU7dz/P3f8QL94FpHsqtIiIiIjIBKrqmjNmNhO4DHgYmAVMAU4HjgBmA4cBD5vZdsDD7v5NMzsc+DbQDiwCVgHHANOBG9z9i2XuagFwU3yfk4BvAIcAU4H7gE97o11XUEREREQaSrV7sLcARwNXuPsxQA9wbrzueOCD7t4NnA38rZm1ARcD8919S7z9HKLDQeYAp5jZjGJwMzvSzB4BPgVcCBBv1+PuHe5+JFFR/55yyZnZWWZ2j5ndU/UjFxERERHJwHgOEVnp7sUC9g6g+P3T17v7SwDuvh6YD9wOXOPuqxPbX+6RjcDdwL7FFe7+W3d/C/D3wM8S28wys5+b2VLg3cCbyyXm7pe6+xx3nzOOxyMiIiIikrrxFNgDidsbE9u+VtLvWGAZcEKV2w9x9z5gqpntbGZHAOcAZ7j78cCPgbLfliMiIiIiUitSvYpIfNz1yUSHgqwzs1G/o9PM9jCzpWY2JV4+ENg+3iO+J/C4u79oZjvGMUVEREREalpVJzlWIy6CLwdOdfctcXH9OzO7s9I27r7WzK4GbjOzDcBmopMnAZYA88zsN8BaYCmwY1r5ioiIiIhkoaoC292fBN6XWF4DzC3p8yrQllh+hegKIABnlPT9ROL2IqKrjJTe50bgQ9XkJyIiIiJSK2rhi2ZERERERBqGCmwRERERkRSpwBYRERERSZEKbBERERGRFFmjffP4woUL/fzzz887Dakh7o6ZLqFer/T3k3I0L2Qiab5JKXdn3vSDWbzmwbITQ3uwRURERERSlNp1sGvJXcek+50077q9N741KdW4sCWjuFnGVtzS2K88dlKqUXc64Bfxrfobi/ULd0s16pSFz8e3shmLOTvMSzku3LNhMQDPzjtkjJ7js/fiB+Jb9Tcv/mX/7lSjnvdET3wr7bewzQD89r0nphwXjrzjpvhWvfz99P6Ufdxk7GzmclZjcUPbx1KOCx9Y8aP4VjY53/quv0457ui0B1tEREREJEUqsEVEREREUqQCW0REREQkRSqwRURERERSVJMFtpntYmYfzzsPEREREZHxyrXANrN3m9kyM/u1md1oZvvEq3YBVGCLiIiISN3Jew92FzDP3ecC3wLCfNMREREREfmfSaXANrOWbdnO3c9z9z/Ei3cBB6eRj4iIiIhIXsa8srmZzQQuAx4GZgFTgNOBI4DZwGHAw2a2HfCwu3/TzA4Hvg20A4uAVcAxwHTgBnf/Ypm7WgDclFgumNmF8f0MAn3A19x9cBsep4iIiIjIhKhmD/YW4GjgCnc/BugBzo3XHQ980N27gbOBvzWzNuBiYL67b4m3n0N0OMgc4BQzm1EMbmZHmtkjwKeACxP3ewzwO3c/CngvsBfw2XIJmtlZZnaPmd1T1aMWEREREclItYeIrHT3YvF6B7BffPt6d38JwN3XA/OB24Fr3H11YvvLPbIRuBvYt7jC3X/r7m8B/h74WWKbFe5+bdxnEPgK8JFyybn7pe4+x93nVPl4REREREQyUW2BPZC4vTGx3Wsl/Y4FlgEnVLn9EHfvA6aa2c6j5OFVZSsiIiIikpPUriISH3d9MtGhIOvM7Owx+u9hZkvNbEq8fCCwfXGPONBmZqfE6wz4EvCTtPIVEREREcnCmCc5VsPMdgQuB0519y1xcf07M7uz0jbuvtbMrgZuM7MNwGaikyeLbgeONLPPxctLgX9PI18RERERkayMWWC7+5PA+xLLa4C5JX1eBdoSy68Ah8SLZ5T0/UTi9iKiq4yUu89jx8pNRERERKTW5P1FMyIiIiIiDUUFtoiIiIhIilRgi4iIiIikyNwb68p3p59+uu+11155pyE15Ng3HIf/Ju8sZFs9etzb+NUzO+WdhtSYnT56KTe09uSdhjSJq5YexV/8YEXeaUiN+c4LzuI1D1q5ddqDLSIiIiKSoobbg21mZ8RXJxERERERycRoNWfDFdgiIiIiInnSISIiIiIiIilSgS0iIiIikiIV2CIiIiIiKVKBLSIiIiKSIhXYIiIiIiIpUoEtIiIiIpIiFdgiIiIiIilSgS0iIiIikiIV2CIiIiIiKVKBLeNiZn9nZlMtcrmZrTCzE/LOS0RERKRWqMCW8Vrg7i8DJwC7A/OBC/JNSURERKR2TM47Aak7Fv/uAq5093vNzCp2toKPHWq0NitptTKblelTIZ4Vl23r9aURxopb/r4S8cu2jNW3TJtXe8/DDeXjljf6XyB+/BU2rvhHr5DbNuVgFdqryGPrGF5+XTW5VPX4ffS4ZVaMbCoTo3S6johR5ulVcRuvIr+R8Ya32Xoylo/lJfHKpTdiUo+8Pyv/sjH8Nx0lzwrbRnOw8roROVZ8DKOMQ8X7rvy3HTHmW3UaYzyT9zeizyjjXOUTYKv7HfNJaZXbRrys2Sh9knFGS6h029H7+tByNfETv0fkZVv19bLxiq+dZeKNdl+Uf49KvqYvX/5En7t3IjVPBbaM13IzuxnYH/iCmbUCg5W7O0ZLdHPoBSX64MSKH6BYYai92GaJtmh50tA2xXWlfZN9iu3GpK36FMrEAygk+w61lSwn1g/FidsKlMRPtBe8UBKn5LcXly3Rx0asi2JYYqRK2sy23jYeQcMS6+K2UZYLJYV6oeT1Phmr+Ccd2YeS+On1GXnfPuY2hZJCZbhPSTtgcbyhGVghvpkn7mOsbbyqPsNxS/Mt9vWt8zdPtA1vv/XjH14/vM5HrKvUt2L8xHKlONvS1wpOwQbLr0v0Gc5zsOy6ctsM9R2lT7Q8OKIPiXWlMYb2IxTK55JsH0/fofussA2J9qE+pfkWSvsm2rZ+SU5OznhbkpNt69/F1+ih9kL0U9pW5rcXClCYNEafSWXWRdu4bb1MYVIUM9knsa50ufK6ySXtkxP3Obmkz+ThPgA2ecQ6G9qmZetlm4yV9Cn+LgwtTxpqLxTzAybbadOQuqACW8brTGA28Li7rzezNxIdJiIiIiIi6BhsGb+l7r7C3V8EcPc/AxflnJOIiIhIzdAebKmKmb0BmAJMM7NdGf5wcSrwptwSExEREakxKrClWp8CziEqppczXGC/DHw3r6REREREao0KbKmKu38b+LaZ/W93vzjvfERERERqlQpsGRd3v9jMjgJmkJg/7n5VbkmJiIiI1BAV2DIuZvZ9YCawEtgSNzugAltEREQEFdgyfnOAg919lC+QEREREWleukyfjNcDwF55JyEiIiJSq7QHW8ZrGvCgmd0NbCw2uvsH8ktJREREpHaowJbxWph3AiIiIiK1TAW2jIu73553DiIiIiK1TAW2jIuZDRBdNQRgO6AFeNXdp+aXlYiIiEjtUIEt4+LurcllMzsZeGdO6YiIiIjUHF1FRP5H3P3nwLF55yEiIiJSK7QHW8bFzD6UWCwQXRdb18QWERERianAlvF6f+L2ZuBJ4KR8UhERERGpPSqwZVzcfX7eOYiIiIjUMh2DLeNiZtPN7HozW2tmz5nZT81set55iYiIiNQKFdgyXlcCNwBvAvYBbozbRERERAQV2DJ+u7v7le6+Of5ZBOyed1IiIiIitUIFtozXOjM7zcwmxT+nAX/OOykRERGRWmHuusKaVM/M9gX+HTiS6PJ8y4Cz3f3pCv0fAF6buAxr2jRgXd5J1AiNRUTjMExjMUxjMUxjESmOwzp378w7GRmbriIi4/XPwDx3fwHAzHYDvgksqND/NXefM1HJ1TIzu0djEdFYRDQOwzQWwzQWwzQWEY1D/dEhIjJehxWLawB3fx44PMd8RERERGqKCmwZr4KZ7VpciPdg65MQERERkZgKIxmvfwOWmdl1RMdgfwT42ij9L52QrOqDxmKYxiKicRimsRimsRimsYhoHOqMTnKUcTOzg4FjAQNudfcHc05JREREpGaowBYRERERSZGOwZbMmFmnmT1iZo+Z2T/mnc9EMrMnzex+M1tpZvfEbbuZ2VIz+0P8e9ex4tQjM7vCzNbGl2gstpV97Bb5TjxH7jOztvwyT1+FsVhoZn+K58ZKM+tKrPtCPBaPmFlHPlmnz8zebGb9ZvaQma0ys7+L25tuXowyFs04L95gZneb2b3xWJwft+9vZnfF8+LHZrZd3L59vPxYvH5GnvmnaZSxWGRmTyTmxey4vWGfI41CBbZkwswmAd8F/hI4GPhYfGhJM2l399mJSyv9I9EhNQcCt8bLjWgRUHqd1kqP/S+BA+Ofs4BLJijHibKIkWMBcFE8N2a7ey8MHXp1KjAr3qYnfh41gs3A5939IOAI4DPx423GeVFpLKD55sVG4Fh3fxswG+g0syOArxONxYHAC8CZcf8zgRfc/QDgorhfo6g0FgBBYl6sjNsa+TnSEFRgS1beCTzm7o+7++vANcBJOeeUt5OAxfHtxcDJOeaSGXe/A3i+pLnSYz8JuMojvwN2MbO9JybT7FUYi0pOAq5x943u/gTwGNHzqO65+7PuviK+PQA8BOxDE86LUcaikkaeF+7ur8SLLfGPE53jc13cXjovivPlOuA4M7MJSjdTo4xFJQ37HGkUKrAlK/sAf0wsr2H0N5FG48DNZrbczM6K2/Z092chepMF9sgtu4lX6bE36zz5bPyx7hWJQ4WaYizij/UPB+6iyedFyVhAE84LM5tkZiuBtcBSYDXwortvjrskH+/QWMTrXwLeOLEZZ6d0LNy9OC++Fs+Li8xs+7itoedFI1CBLVkpt1ehmc6oPdrd24g+xvuMmb0374RqVDPOk0uAmUQfAz9LdOlLaIKxMLOdgJ8C57j7y6N1LdPW6GPRlPPC3be4+2xgOtGe+YPKdYt/N9VYmNkhwBeAtwLvAHYDzo27N/RYNAIV2JKVNcCbE8vTgWdyymXCufsz8e+1wPVEbxzPFT/Ci3+vzS/DCVfpsTfdPHH35+I30kHgMoY/7m/osTCzFqKC8mp3/1nc3JTzotxYNOu8KHL3F4FfEx2XvouZFb+nI/l4h8YiXr8z1R+CVTcSY9EZH1Lk7r4RuJImmxf1TAW2ZOX3wIHx2eDbEZ2kc0POOU0IM9vRzFqLt4ETgAeIHv+8uNs84Bf5ZJiLSo/9BuDj8RnxRwAvFQ8ZaFQlx0l+kGhuQDQWp8ZXStif6OSluyc6vyzEx8leDjzk7hcmVjXdvKg0Fk06L3Y3s13i2zsA7yM6Jr0fOCXuVjovivPlFOA2b5BrDVcYi4cT/4Aa0bHoyXnRkM+RRqFvcpRMuPtmM/ss0AdMAq5w91U5pzVR9gSuj8+9mQz80N2XmNnvgWvN7EzgaeDDOeaYGTP7ETAXmGZma4AvAxdQ/rH3Al1EJ26tB+ZPeMIZqjAWc+NLbTnwJPApAHdfZWbXAg8SXWniM+6+JY+8M3A0cDpwf3yMKcB5NOe8qDQWH2vCebE3sDi+KkoBuNbdbzKzB4FrzOyrwH8Q/UNC/Pv7ZvYY0Z7rU/NIOiOVxuI2M9ud6JCQlcCn4/6N/BxpCPqiGRERERGRFOkQERERERGRFKnAFhERERFJkQpsEREREZEUqcAWEREREUmRCmwRERERkRSpwBYRERERSZEKbBERERGRFKnAFhERERFJ0X8DZzFBO5T6i5EAAAAASUVORK5CYII=\n",
      "text/plain": [
       "<Figure size 864x864 with 2 Axes>"
      ]
     },
     "metadata": {
      "needs_background": "light"
     },
     "output_type": "display_data"
    }
   ],
   "source": [
    "genenames = gdf['symbol'].values\n",
    "samplenames = [] #skipping names to replace in intaglio\n",
    "log2matrix = np.vectorize(log2)\n",
    "# data = log2matrix(gdf.iloc[0:,7:15].values)\n",
    "data = gdf.iloc[0:,7:37].values\n",
    "\n",
    "\n",
    "fig, ax = plt.subplots(figsize=[12,12])\n",
    "im, cbar = heatmap(data, genenames, samplenames, ax=ax,\n",
    "                   cmap=\"inferno\", cbarlabel=\"counts\")\n",
    "# texts = annotate_heatmap(im, valfmt=\"{x:.1f} t\")\n",
    "\n",
    "plt.savefig(\"{0}.svg\".format(plotName), transparent=True, format=\"svg\")\n",
    "plt.show()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "#### Trying to draw counts for single gene by wedges in a circle "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 32,
   "metadata": {},
   "outputs": [
    {
     "ename": "AttributeError",
     "evalue": "'list' object has no attribute 'get_label'",
     "output_type": "error",
     "traceback": [
      "\u001b[0;31m---------------------------------------------------------------------------\u001b[0m",
      "\u001b[0;31mAttributeError\u001b[0m                            Traceback (most recent call last)",
      "\u001b[0;32m<ipython-input-32-5ddde20404c1>\u001b[0m in \u001b[0;36m<module>\u001b[0;34m\u001b[0m\n\u001b[1;32m      5\u001b[0m \u001b[0mp2\u001b[0m \u001b[0;34m=\u001b[0m \u001b[0mmatplotlib\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mpatches\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mWedge\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0;36m0\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0;36m1\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0;36m20\u001b[0m\u001b[0;34m,\u001b[0m\u001b[0;36m30\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0mec\u001b[0m\u001b[0;34m=\u001b[0m\u001b[0;34m\"none\"\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m      6\u001b[0m \u001b[0mpatches\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mappend\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0mp2\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0;32m----> 7\u001b[0;31m \u001b[0max\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0madd_collection\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0mpatches\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0m",
      "\u001b[0;32m~/anaconda3/lib/python3.7/site-packages/matplotlib/axes/_base.py\u001b[0m in \u001b[0;36madd_collection\u001b[0;34m(self, collection, autolim)\u001b[0m\n\u001b[1;32m   1858\u001b[0m         \u001b[0mAdd\u001b[0m \u001b[0ma\u001b[0m\u001b[0;31m \u001b[0m\u001b[0;31m`\u001b[0m\u001b[0;34m~\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mCollection\u001b[0m\u001b[0;31m`\u001b[0m \u001b[0mto\u001b[0m \u001b[0mthe\u001b[0m \u001b[0maxes\u001b[0m\u001b[0;31m'\u001b[0m \u001b[0mcollections\u001b[0m\u001b[0;34m;\u001b[0m \u001b[0;32mreturn\u001b[0m \u001b[0mthe\u001b[0m \u001b[0mcollection\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m   1859\u001b[0m         \"\"\"\n\u001b[0;32m-> 1860\u001b[0;31m         \u001b[0mlabel\u001b[0m \u001b[0;34m=\u001b[0m \u001b[0mcollection\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mget_label\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0m\u001b[1;32m   1861\u001b[0m         \u001b[0;32mif\u001b[0m \u001b[0;32mnot\u001b[0m \u001b[0mlabel\u001b[0m\u001b[0;34m:\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m   1862\u001b[0m             \u001b[0mcollection\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mset_label\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0;34m'_collection%d'\u001b[0m \u001b[0;34m%\u001b[0m \u001b[0mlen\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0mself\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mcollections\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n",
      "\u001b[0;31mAttributeError\u001b[0m: 'list' object has no attribute 'get_label'"
     ]
    },
    {
     "data": {
      "image/png": "iVBORw0KGgoAAAANSUhEUgAAAXwAAAD8CAYAAAB0IB+mAAAABHNCSVQICAgIfAhkiAAAAAlwSFlzAAALEgAACxIB0t1+/AAAADh0RVh0U29mdHdhcmUAbWF0cGxvdGxpYiB2ZXJzaW9uMy4xLjAsIGh0dHA6Ly9tYXRwbG90bGliLm9yZy+17YcXAAANgElEQVR4nO3ccYjfd33H8efLxE6mtY7lBEmi7Vi6Gsqg7ug6hFnRjbR/JP8USaC4SmnArQ5mETocKvWvKUMQsmm2iVPQWv1DD4nkD1fpECO50lmalMAtOnNE6Fm7/lO0Znvvj99P77hcct/e/e4u3vv5gMDv+/t9fr9758PdM798f/f7paqQJG1/r9rqASRJm8PgS1ITBl+SmjD4ktSEwZekJgy+JDWxavCTfC7Jc0meucLtSfLpJHNJnk7ytsmPKUlaryHP8D8PHLjK7XcB+8Z/jgL/tP6xJEmTtmrwq+oJ4GdXWXII+EKNnALekORNkxpQkjQZOyfwGLuBC0uO58fX/WT5wiRHGf0vgNe+9rV/dMstt0zgy0tSH08++eRPq2pqLfedRPCzwnUrfl5DVR0HjgNMT0/X7OzsBL68JPWR5L/Xet9J/JbOPLB3yfEe4OIEHleSNEGTCP4M8N7xb+vcAbxYVZedzpEkba1VT+kk+TJwJ7AryTzwUeDVAFX1GeAEcDcwB7wEvG+jhpUkrd2qwa+qI6vcXsBfTWwiSdKG8J22ktSEwZekJgy+JDVh8CWpCYMvSU0YfElqwuBLUhMGX5KaMPiS1ITBl6QmDL4kNWHwJakJgy9JTRh8SWrC4EtSEwZfkpow+JLUhMGXpCYMviQ1YfAlqQmDL0lNGHxJasLgS1ITBl+SmjD4ktSEwZekJgy+JDVh8CWpCYMvSU0YfElqwuBLUhMGX5KaMPiS1ITBl6QmDL4kNTEo+EkOJDmXZC7Jwyvc/uYkjyd5KsnTSe6e/KiSpPVYNfhJdgDHgLuA/cCRJPuXLfs74LGqug04DPzjpAeVJK3PkGf4twNzVXW+ql4GHgUOLVtTwOvHl28ALk5uREnSJAwJ/m7gwpLj+fF1S30MuDfJPHAC+MBKD5TkaJLZJLMLCwtrGFeStFZDgp8Vrqtlx0eAz1fVHuBu4ItJLnvsqjpeVdNVNT01NfXKp5UkrdmQ4M8De5cc7+HyUzb3A48BVNX3gNcAuyYxoCRpMoYE/zSwL8lNSa5j9KLszLI1PwbeBZDkrYyC7zkbSbqGrBr8qroEPAicBJ5l9Ns4Z5I8kuTgeNlDwANJfgB8Gbivqpaf9pEkbaGdQxZV1QlGL8Yuve4jSy6fBd4+2dEkSZPkO20lqQmDL0lNGHxJasLgS1ITBl+SmjD4ktSEwZekJgy+JDVh8CWpCYMvSU0YfElqwuBLUhMGX5KaMPiS1ITBl6QmDL4kNWHwJakJgy9JTRh8SWrC4EtSEwZfkpow+JLUhMGXpCYMviQ1YfAlqQmDL0lNGHxJasLgS1ITBl+SmjD4ktSEwZekJgy+JDVh8CWpCYMvSU0MCn6SA0nOJZlL8vAV1rwnydkkZ5J8abJjSpLWa+dqC5LsAI4BfwbMA6eTzFTV2SVr9gF/C7y9ql5I8saNGliStDZDnuHfDsxV1fmqehl4FDi0bM0DwLGqegGgqp6b7JiSpPUaEvzdwIUlx/Pj65a6Gbg5yXeTnEpyYKUHSnI0yWyS2YWFhbVNLElakyHBzwrX1bLjncA+4E7gCPAvSd5w2Z2qjlfVdFVNT01NvdJZJUnrMCT488DeJcd7gIsrrPlGVf2yqn4InGP0D4Ak6RoxJPingX1JbkpyHXAYmFm25uvAOwGS7GJ0iuf8JAeVJK3PqsGvqkvAg8BJ4Fngsao6k+SRJAfHy04Czyc5CzwOfKiqnt+ooSVJr1yqlp+O3xzT09M1Ozu7JV9bkn5TJXmyqqbXcl/faStJTRh8SWrC4EtSEwZfkpow+JLUhMGXpCYMviQ1YfAlqQmDL0lNGHxJasLgS1ITBl+SmjD4ktSEwZekJgy+JDVh8CWpCYMvSU0YfElqwuBLUhMGX5KaMPiS1ITBl6QmDL4kNWHwJakJgy9JTRh8SWrC4EtSEwZfkpow+JLUhMGXpCYMviQ1YfAlqQmDL0lNGHxJasLgS1ITg4Kf5ECSc0nmkjx8lXX3JKkk05MbUZI0CasGP8kO4BhwF7AfOJJk/wrrrgf+Gvj+pIeUJK3fkGf4twNzVXW+ql4GHgUOrbDu48AngJ9PcD5J0oQMCf5u4MKS4/nxdb+W5DZgb1V982oPlORoktkkswsLC694WEnS2g0Jfla4rn59Y/Iq4FPAQ6s9UFUdr6rpqpqempoaPqUkad2GBH8e2LvkeA9wccnx9cCtwHeS/Ai4A5jxhVtJurYMCf5pYF+Sm5JcBxwGZn51Y1W9WFW7qurGqroROAUcrKrZDZlYkrQmqwa/qi4BDwIngWeBx6rqTJJHkhzc6AElSZOxc8iiqjoBnFh23UeusPbO9Y8lSZo032krSU0YfElqwuBLUhMGX5KaMPiS1ITBl6QmDL4kNWHwJakJgy9JTRh8SWrC4EtSEwZfkpow+JLUhMGXpCYMviQ1YfAlqQmDL0lNGHxJasLgS1ITBl+SmjD4ktSEwZekJgy+JDVh8CWpCYMvSU0YfElqwuBLUhMGX5KaMPiS1ITBl6QmDL4kNWHwJakJgy9JTRh8SWpiUPCTHEhyLslckodXuP2DSc4meTrJt5O8ZfKjSpLWY9XgJ9kBHAPuAvYDR5LsX7bsKWC6qv4Q+BrwiUkPKklanyHP8G8H5qrqfFW9DDwKHFq6oKoer6qXxoengD2THVOStF5Dgr8buLDkeH583ZXcD3xrpRuSHE0ym2R2YWFh+JSSpHUbEvyscF2tuDC5F5gGPrnS7VV1vKqmq2p6ampq+JSSpHXbOWDNPLB3yfEe4OLyRUneDXwYeEdV/WIy40mSJmXIM/zTwL4kNyW5DjgMzCxdkOQ24LPAwap6bvJjSpLWa9XgV9Ul4EHgJPAs8FhVnUnySJKD42WfBF4HfDXJfyaZucLDSZK2yJBTOlTVCeDEsus+suTyuyc8lyRpwnynrSQ1YfAlqQmDL0lNGHxJasLgS1ITBl+SmjD4ktSEwZekJgy+JDVh8CWpCYMvSU0YfElqwuBLUhMGX5KaMPiS1ITBl6QmDL4kNWHwJakJgy9JTRh8SWrC4EtSEwZfkpow+JLUhMGXpCYMviQ1YfAlqQmDL0lNGHxJasLgS1ITBl+SmjD4ktSEwZekJgy+JDVh8CWpCYMvSU0MCn6SA0nOJZlL8vAKt/9Wkq+Mb/9+khsnPagkaX1WDX6SHcAx4C5gP3Akyf5ly+4HXqiq3wc+Bfz9pAeVJK3PkGf4twNzVXW+ql4GHgUOLVtzCPi38eWvAe9KksmNKUlar50D1uwGLiw5ngf++EprqupSkheB3wV+unRRkqPA0fHhL5I8s5aht6FdLNurxtyLRe7FIvdi0R+s9Y5Dgr/SM/Vawxqq6jhwHCDJbFVND/j62557sci9WOReLHIvFiWZXet9h5zSmQf2LjneA1y80pokO4EbgJ+tdShJ0uQNCf5pYF+Sm5JcBxwGZpatmQH+Ynz5HuDfq+qyZ/iSpK2z6imd8Tn5B4GTwA7gc1V1JskjwGxVzQD/CnwxyRyjZ/aHB3zt4+uYe7txLxa5F4vci0XuxaI170V8Ii5JPfhOW0lqwuBLUhMbHnw/lmHRgL34YJKzSZ5O8u0kb9mKOTfDanuxZN09SSrJtv2VvCF7keQ94++NM0m+tNkzbpYBPyNvTvJ4kqfGPyd3b8WcGy3J55I8d6X3KmXk0+N9ejrJ2wY9cFVt2B9GL/L+F/B7wHXAD4D9y9b8JfCZ8eXDwFc2cqat+jNwL94J/Pb48vs778V43fXAE8ApYHqr597C74t9wFPA74yP37jVc2/hXhwH3j++vB/40VbPvUF78afA24BnrnD73cC3GL0H6g7g+0Med6Of4fuxDItW3YuqeryqXhofnmL0noftaMj3BcDHgU8AP9/M4TbZkL14ADhWVS8AVNVzmzzjZhmyFwW8fnz5Bi5/T9C2UFVPcPX3Mh0CvlAjp4A3JHnTao+70cFf6WMZdl9pTVVdAn71sQzbzZC9WOp+Rv+Cb0er7kWS24C9VfXNzRxsCwz5vrgZuDnJd5OcSnJg06bbXEP24mPAvUnmgRPABzZntGvOK+0JMOyjFdZjYh/LsA0M/nsmuReYBt6xoRNtnavuRZJXMfrU1fs2a6AtNOT7Yiej0zp3Mvpf338kubWq/meDZ9tsQ/biCPD5qvqHJH/C6P0/t1bV/238eNeUNXVzo5/h+7EMi4bsBUneDXwYOFhVv9ik2TbbantxPXAr8J0kP2J0jnJmm75wO/Rn5BtV9cuq+iFwjtE/ANvNkL24H3gMoKq+B7yG0QerdTOoJ8ttdPD9WIZFq+7F+DTGZxnFfruep4VV9qKqXqyqXVV1Y1XdyOj1jINVteYPjbqGDfkZ+TqjF/RJsovRKZ7zmzrl5hiyFz8G3gWQ5K2Mgr+wqVNeG2aA945/W+cO4MWq+slqd9rQUzq1cR/L8Btn4F58Engd8NXx69Y/rqqDWzb0Bhm4Fy0M3IuTwJ8nOQv8L/Chqnp+66beGAP34iHgn5P8DaNTGPdtxyeISb7M6BTervHrFR8FXg1QVZ9h9PrF3cAc8BLwvkGPuw33SpK0At9pK0lNGHxJasLgS1ITBl+SmjD4ktSEwZekJgy+JDXx/4aZaro1YsjCAAAAAElFTkSuQmCC\n",
      "text/plain": [
       "<Figure size 432x288 with 1 Axes>"
      ]
     },
     "metadata": {
      "needs_background": "light"
     },
     "output_type": "display_data"
    }
   ],
   "source": [
    "fH, ax = plt.subplots()\n",
    "patches = []\n",
    "p1 = matplotlib.patches.Wedge(0, 1, 0,10, ec=\"none\")\n",
    "patches.append(p1)\n",
    "p2 = matplotlib.patches.Wedge(0, 1, 20,30, ec=\"none\")\n",
    "patches.append(p2)\n",
    "ax.add_collection(patches)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.7.4"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 4
}
