{
 "cells": [
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Load required modules"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 1,
   "metadata": {},
   "outputs": [
    {
     "ename": "NameError",
     "evalue": "name 'np' is not defined",
     "output_type": "error",
     "traceback": [
      "\u001b[0;31m---------------------------------------------------------------------------\u001b[0m",
      "\u001b[0;31mNameError\u001b[0m                                 Traceback (most recent call last)",
      "\u001b[0;32m<ipython-input-1-4122a2ff3e2d>\u001b[0m in \u001b[0;36m<module>\u001b[0;34m\u001b[0m\n\u001b[1;32m      1\u001b[0m \u001b[0ma\u001b[0m \u001b[0;34m=\u001b[0m \u001b[0;34m[\u001b[0m\u001b[0;36m2.5\u001b[0m\u001b[0;34m,\u001b[0m\u001b[0;36m5\u001b[0m\u001b[0;34m,\u001b[0m\u001b[0;36m.5\u001b[0m\u001b[0;34m,\u001b[0m\u001b[0;36m.5\u001b[0m\u001b[0;34m]\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m      2\u001b[0m \u001b[0ma\u001b[0m \u001b[0;34m=\u001b[0m \u001b[0;34m[\u001b[0m\u001b[0;36m0.5\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0;36m5\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0;36m.5\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0;36m.5\u001b[0m \u001b[0;34m]\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0;32m----> 3\u001b[0;31m \u001b[0mnp\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mmultiply\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0ma\u001b[0m\u001b[0;34m,\u001b[0m\u001b[0;36m4\u001b[0m\u001b[0;34m*\u001b[0m\u001b[0;36m1.05\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0m",
      "\u001b[0;31mNameError\u001b[0m: name 'np' is not defined"
     ]
    }
   ],
   "source": [
    "a = [2.5,5,.5,.5]\n",
    "a = [0.5, 5, .5, .5 ]\n",
    "np.multiply(a,4*1.05)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 1,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "LOADED\n"
     ]
    }
   ],
   "source": [
    "# External libraries\n",
    "import numpy as np\n",
    "import pandas as pd\n",
    "import pygal\n",
    "import math\n",
    "import IPython\n",
    "import matplotlib\n",
    "import matplotlib.pyplot as plt\n",
    "import matplotlib.font_manager as font_manager\n",
    "import importlib\n",
    "import svgutils\n",
    "from svgutils.compose import *\n",
    "import os.path\n",
    "#import svgwrite\n",
    "\n",
    "# Custom code\n",
    "from fx_RNAseq import *\n",
    "def findgene(genename):\n",
    "    dPath = \"~/Documents/LiMolec/zfRNAseq/20190827/20190827_DESeq2/\" #directory path\n",
    "#     fPath = \"00_rodsVcones/rodVCones_raw.csv\" #file path\n",
    "    fPath = \"00_rodsVcones/rodVCones_fpkm.csv\" #file path\n",
    "    tempPath = \"temp.txt\" # temporary file to store values\n",
    "    !head -1 {dPath + fPath} > {dPath + tempPath}\n",
    "    # EXACT MATCH ONLY by matching word (-w) and only at the start of each line (^) which corresponds to the symbol\n",
    "    !grep -w ^$genename {dPath + fPath} >> {dPath + tempPath}\n",
    "    gdf = pd.read_csv(dPath + tempPath)\n",
    "    return gdf\n",
    "\n",
    "def findgenelist(genename):\n",
    "    dPath = \"~/Documents/LiMolec/zfRNAseq/20190827/20190827_DESeq2/\" #directory path\n",
    "#     fPath = \"00_rodsVcones/rodVCones_raw.csv\" #file path\n",
    "    fPath = \"00_rodsVcones/rodVCones_fpkm.csv\" #file path\n",
    "    tempPath = \"temp.txt\" # temporary file to store values\n",
    "    !head -1 {dPath + fPath} > {dPath + tempPath}\n",
    "    # EXACT MATCH ONLY by matching word (-w) and only at the start of each line (^) which corresponds to the symbol\n",
    "    !grep ^$genename {dPath + fPath} >> {dPath + tempPath}\n",
    "    gdf = pd.read_csv(dPath + tempPath)\n",
    "    return gdf\n",
    "\n",
    "# def findgene(genename):\n",
    "#     !head -1 ~/Dropbox/LiLab/RNAseq/20190827_DESeq2/00_rodsVcones/rodVCones_raw.csv > ~/Dropbox/LiLab/RNAseq/20190827_DESeq2/temp.txt\n",
    "#     # EXACT MATCH ONLY by matching word (-w) and only at the start of each line (^) which corresponds to the symbol\n",
    "#     !grep -w ^$genename ~/Dropbox/LiLab/RNAseq/20190827_DESeq2/00_rodsVcones/rodVCones_raw.csv >> ~/Dropbox/LiLab/RNAseq/20190827_DESeq2/temp.txt\n",
    "#     gdf = pd.read_csv(\"~/Dropbox/LiLab/RNAseq/20190827_DESeq2/temp.txt\")\n",
    "    \n",
    "#     return gdf\n",
    "\n",
    "# def findgenelist(genename):\n",
    "#     !head -1 ~/Dropbox/LiLab/RNAseq/20190827_DESeq2/00_rodsVcones/rodVCones_raw.csv > ~/Dropbox/LiLab/RNAseq/20190827_DESeq2/temp.txt\n",
    "#     # EXACT MATCH ONLY by matching word (-w) and only at the start of each line (^) which corresponds to the symbol\n",
    "#     !grep ^$genename ~/Dropbox/LiLab/RNAseq/20190827_DESeq2/00_rodsVcones/rodVCones_raw.csv >> ~/Dropbox/LiLab/RNAseq/20190827_DESeq2/temp.txt\n",
    "#     gdf = pd.read_csv(\"~/Dropbox/LiLab/RNAseq/20190827_DESeq2/temp.txt\")\n",
    "#     return gdf\n",
    "\n",
    "def plotNReads():\n",
    "    rdf = pd.read_csv(\"~/Documents/LiMolec/zfRNAseq/20190827/20190827_DESeq2/nMappedReads.csv\")\n",
    "    h = rdf.iloc[:,1]\n",
    "    h = h / 1e6\n",
    "    n = np.arange(1,7) # Rods\n",
    "    n = np.append(n, 6.5 + np.arange(1,6)) # UV\n",
    "    n = np.append(n, 12 + np.arange(1,7)) # S\n",
    "    n = np.append(n, 18.5 + np.arange(1,8)) # M\n",
    "    n = np.append(n, 26 + np.arange(1,7)) # L\n",
    "    return n, h\n",
    "\n",
    "def plotBars(genename):\n",
    "    gdf = findgene(genename)\n",
    "    n = np.arange(1,7) # Rods\n",
    "    n = np.append(n, 6.5 + np.arange(1,6)) # UV\n",
    "    n = np.append(n, 12 + np.arange(1,7)) # S\n",
    "    n = np.append(n, 18.5 + np.arange(1,8)) # M\n",
    "    n = np.append(n, 26 + np.arange(1,7)) # L\n",
    "    h_start = 7\n",
    "    h_end = 37\n",
    "    h = gdf.iloc[0,h_start:h_end]\n",
    "    return n, h\n",
    "\n",
    "def plotBars_list(genelist):\n",
    "    n=[]\n",
    "    h=[]\n",
    "    for i,gene in enumerate(genelist, start=0):\n",
    "        tempn, temph = plotBars(gene)\n",
    "        n = np.append(n,tempn+(30*i))\n",
    "        h = np.append(h,temph)\n",
    "    return n, h\n",
    "\n",
    "def svgreload():\n",
    "    !osascript -e 'tell application \"Google Chrome\" to tell the active tab of second window to reload'\n",
    "\n",
    "from pygal.style import Style\n",
    "custom_style = Style(\n",
    "    background = 'black',\n",
    "    plot_background = '#111111',\n",
    "    foreground = '#999999',\n",
    "    foreground_strong = '#eeeeee',\n",
    "    foreground_subtle = '#555555',\n",
    "    opacity = '.8',\n",
    "    opacity_hover = '.4',\n",
    "    transition = '10ms',\n",
    "    colors = ('#747474','#B540B7', '#4669F2','#04CD22','#CC2C2A'),\n",
    "    label_font_size = 24,\n",
    "    major_label_font_size = 24,\n",
    "    title_font_size = 24,\n",
    "    legend_font_size = 18\n",
    ")\n",
    "\n",
    "#photoreceptor Colors\n",
    "pC = {\n",
    "    'r' : '#747474',\n",
    "    'u' : '#B540B7',\n",
    "    's' : '#4669F2',\n",
    "    'm' : '#04CD22',\n",
    "    'l' : '#CC2C2A',\n",
    "    'plt' : '',\n",
    "}\n",
    "\n",
    "pC['plt']=[\n",
    "    pC['r'],pC['r'],pC['r'],pC['r'],pC['r'],pC['r'],\n",
    "    pC['u'],pC['u'],pC['u'],pC['u'],pC['u'],\n",
    "    pC['s'],pC['s'],pC['s'],pC['s'],pC['s'],pC['s'],\n",
    "    pC['m'],pC['m'],pC['m'],pC['m'],pC['m'],pC['m'],pC['m'],\n",
    "    pC['l'],pC['l'],pC['l'],pC['l'],pC['l'],pC['l']\n",
    "]\n",
    "\n",
    "def findgene_glia(genename):\n",
    "    dPath = \"~/Documents/LiLab/RNAseq/zfMullerGlia_2018Charlton/\" #directory path\n",
    "#     fPath = \"glia_23615_sup-0006-tables2.csv\" #file path\n",
    "    fPath = \"glia8dpf_DEGcounts.csv\" #file path\n",
    "    tempPath = \"temp.txt\" # temporary file to store values\n",
    "    !head -1 {dPath + fPath} > {dPath + tempPath}\n",
    "    # EXACT MATCH ONLY by matching word (-w) and only at the start of each line (^) which corresponds to the symbol\n",
    "    !grep -w ^$genename {dPath + fPath} >> {dPath + tempPath}\n",
    "    gdf = pd.read_csv(dPath + tempPath)\n",
    "    return gdf\n",
    "\n",
    "def findgenelist_glia(genename):\n",
    "    dPath = \"~/Documents/LiLab/RNAseq/zfMullerGlia_2018Charlton/\" #directory path\n",
    "#     fPath = \"glia_23615_sup-0006-tables2.csv\" #file path\n",
    "    fPath = \"glia8dpf_DEGcounts.csv\" #file path\n",
    "    tempPath = \"temp.txt\" # temporary file to store values\n",
    "    !head -1 {dPath + fPath} > {dPath + tempPath}\n",
    "    # EXACT MATCH ONLY by matching word (-w) and only at the start of each line (^) which corresponds to the symbol\n",
    "    !grep ^$genename {dPath + fPath} >> {dPath + tempPath}\n",
    "    gdf = pd.read_csv(dPath + tempPath)\n",
    "    return gdf\n",
    "\n",
    "def plotBars_glia(genename):\n",
    "    gdf = findgene_glia(genename)\n",
    "    n = np.arange(1,19) # GFP+\n",
    "    n = np.append(n, 19.5 + np.arange(1,19)) # GFP-\n",
    "    h_start = 5\n",
    "    h_end = 41\n",
    "#     h = np.power(2,(gdf.iloc[0,h_start:h_end]))\n",
    "    h = gdf.iloc[0,h_start:h_end]\n",
    "    return n, h\n",
    "\n",
    "#glia Colors\n",
    "gC = {\n",
    "    'm48' : '#c7ffec',\n",
    "    'm60' : '#94ffdb',\n",
    "    'm72' : '#57ffc7',\n",
    "    'm96' : '#2effb9',\n",
    "    'm120': '#05ffac',\n",
    "    'm192': '#00d68f',\n",
    "    \n",
    "    'c48' : '#e3e3e3',\n",
    "    'c60' : '#c9c9c9',\n",
    "    'c72' : '#ababab',\n",
    "    'c96' : '#969696',\n",
    "    'c120': '#828282',\n",
    "    'c192': '#6b6b6b',\n",
    "    \n",
    "    'plt' : '',\n",
    "}\n",
    "\n",
    "gC['plt']=[\n",
    "    gC['m48'],gC['m48'],gC['m48'],\n",
    "    gC['m60'],gC['m60'],gC['m60'],\n",
    "    gC['m72'],gC['m72'],gC['m72'],\n",
    "    gC['m96'],gC['m96'],gC['m96'],\n",
    "    gC['m120'],gC['m120'],gC['m120'],\n",
    "    gC['m192'],gC['m192'],gC['m192'],\n",
    "    gC['c48'],gC['c48'],gC['c48'],\n",
    "    gC['c60'],gC['c60'],gC['c60'],\n",
    "    gC['c72'],gC['c72'],gC['c72'],\n",
    "    gC['c96'],gC['c96'],gC['c96'],\n",
    "    gC['c120'],gC['c120'],gC['c120'],\n",
    "    gC['c192'],gC['c192'],gC['c192'],\n",
    "]\n",
    "\n",
    "def findgene_rods(genename):\n",
    "    dPath = \"~/Documents/LiLab/RNAseq/zfRods_Sun2018/\" #directory path\n",
    "#     fPath = \"glia_23615_sup-0006-tables2.csv\" #file path\n",
    "    fPath = \"zfRods.csv\" #file path\n",
    "    tempPath = \"temp.txt\" # temporary file to store values\n",
    "    !head -1 {dPath + fPath} > {dPath + tempPath}\n",
    "    # EXACT MATCH ONLY by matching word (-w) and only at the start of each line (^) which corresponds to the symbol\n",
    "    !grep -w ^$genename {dPath + fPath} >> {dPath + tempPath}\n",
    "    gdf = pd.read_csv(dPath + tempPath)\n",
    "    return gdf\n",
    "\n",
    "def findgenelist_rods(genename):\n",
    "    dPath = \"~/Documents/LiLab/RNAseq/zfRods_Sun2018/\" #directory path\n",
    "#     fPath = \"glia_23615_sup-0006-tables2.csv\" #file path\n",
    "    fPath = \"zfRods.csv\" #file path\n",
    "    tempPath = \"temp.txt\" # temporary file to store values\n",
    "    !head -1 {dPath + fPath} > {dPath + tempPath}\n",
    "    # EXACT MATCH ONLY by matching word (-w) and only at the start of each line (^) which corresponds to the symbol\n",
    "    !grep ^$genename {dPath + fPath} >> {dPath + tempPath}\n",
    "    gdf = pd.read_csv(dPath + tempPath)\n",
    "    return gdf\n",
    "\n",
    "def plotBars_rods(genename):\n",
    "    gdf = findgene_rods(genename)\n",
    "    n = np.arange(1,5) # GFP+\n",
    "    n = np.append(n, 4.5 + np.arange(1,5)) # GFP-\n",
    "    h_start = 7\n",
    "    h_end = 15\n",
    "#     h = np.power(2,(gdf.iloc[0,h_start:h_end]))\n",
    "    h = gdf.iloc[0,h_start:h_end].values\n",
    "    return n, h\n",
    "\n",
    "#rod Colors\n",
    "rC = {\n",
    "    'r' : '#747474',\n",
    "    'c': '#dac910',\n",
    "    'plt' : '',\n",
    "}\n",
    "\n",
    "rC['plt']=[\n",
    "    rC['r'],rC['r'],rC['r'],rC['r'],\n",
    "    rC['c'],rC['c'],rC['c'],rC['c'],\n",
    "]\n",
    "\n",
    "def findgene_sqcones(genename):\n",
    "    dPath = \"~/Documents/LiLab/RNAseq/sqCones_Kunze/\" #directory path\n",
    "    fPath = \"sqConesAwHib_Adult.csv\" #file path\n",
    "    tempPath = \"temp.txt\" # temporary file to store values\n",
    "    !head -1 {dPath + fPath} > {dPath + tempPath}\n",
    "    # EXACT MATCH ONLY by matching word (-w) and only at the start of each line (^) which corresponds to the symbol\n",
    "    !grep -w ^$genename {dPath + fPath} >> {dPath + tempPath}\n",
    "    gdf = pd.read_csv(dPath + tempPath)\n",
    "    return gdf\n",
    "\n",
    "def findgenelist_sqcones(genename):\n",
    "    dPath = \"~/Documents/LiLab/RNAseq/sqCones_Kunze/\" #directory path\n",
    "    fPath = \"sqConesAwHib_Adult.csv\" #file path\n",
    "    tempPath = \"temp.txt\" # temporary file to store values\n",
    "    !head -1 {dPath + fPath} > {dPath + tempPath}\n",
    "    # EXACT MATCH ONLY by matching word (-w) and only at the start of each line (^) which corresponds to the symbol\n",
    "    !grep ^$genename {dPath + fPath} >> {dPath + tempPath}\n",
    "    gdf = pd.read_csv(dPath + tempPath)\n",
    "    return gdf\n",
    "\n",
    "def plotBars_sqcones(genename):\n",
    "    gdf = findgene_sqcones(genename)\n",
    "    n = np.arange(1,8) # GFP+\n",
    "    n = np.append(n, 7.5 + np.arange(1,7)) # GFP-\n",
    "    h_start = 8\n",
    "    h_end = 21\n",
    "#     h = np.power(2,(gdf.iloc[0,h_start:h_end]))\n",
    "    h = gdf.iloc[0,h_start:h_end].values\n",
    "    return n, h\n",
    "\n",
    "#squirrel Colors\n",
    "sqC = {\n",
    "    's' : '#4669F2',\n",
    "    'sHib' : '#548ced',\n",
    "    'm' : '#04CD22',\n",
    "    'mHib' : '#53e477',\n",
    "    'plt' : '',\n",
    "}\n",
    "\n",
    "sqC['plt']=[\n",
    "    sqC['s'],sqC['s'],sqC['s'],sqC['s'],sqC['s'],\n",
    "    sqC['sHib'],sqC['sHib'],\n",
    "    sqC['m'],sqC['m'],sqC['m'],\n",
    "    sqC['mHib'],sqC['mHib'],sqC['mHib'],\n",
    "]\n",
    "\n",
    "\n",
    "def findgene_haircell(genename):\n",
    "    dPath = \"~/Documents/LiLab/RNAseq/zfHairCells_Mattern2018/\" #directory path\n",
    "    fPath = \"zfHairCells_rpkm.csv\" #file path\n",
    "    tempPath = \"temp.txt\" # temporary file to store values\n",
    "    !head -1 {dPath + fPath} > {dPath + tempPath}\n",
    "    # EXACT MATCH ONLY by matching word (-w) and only at the start of each line (^) which corresponds to the symbol\n",
    "    !grep -w ^$genename {dPath + fPath} >> {dPath + tempPath}\n",
    "    gdf = pd.read_csv(dPath + tempPath)\n",
    "    return gdf\n",
    "\n",
    "def findgenelist_haircell(genename):\n",
    "    dPath = \"~/Documents/LiLab/RNAseq/zfHairCells_Mattern2018/\" #directory path\n",
    "    fPath = \"zfHairCells_rpkm.csv\" #file path\n",
    "    tempPath = \"temp.txt\" # temporary file to store values\n",
    "    !head -1 {dPath + fPath} > {dPath + tempPath}\n",
    "    # EXACT MATCH ONLY by matching word (-w) and only at the start of each line (^) which corresponds to the symbol\n",
    "    !grep ^$genename {dPath + fPath} >> {dPath + tempPath}\n",
    "    gdf = pd.read_csv(dPath + tempPath)\n",
    "    return gdf\n",
    "\n",
    "def plotBars_haircell(genename):\n",
    "    gdf = findgene_sqcones(genename)\n",
    "    n = np.arange(1,4) # riboTagged\n",
    "    n = np.append(n, 3.5 + np.arange(1,4)) # Whatever this is\n",
    "    h_start = 2\n",
    "    h_end = 8\n",
    "#     h = np.power(2,(gdf.iloc[0,h_start:h_end]))\n",
    "    h = gdf.iloc[0,h_start:h_end].values\n",
    "    return n, h\n",
    "\n",
    "#hairCell Colors\n",
    "hcC = {\n",
    "    'h' : '#e147c0',\n",
    "    'c' : '#dac910',\n",
    "    'plt' : '',\n",
    "}\n",
    "\n",
    "hcC['plt']=[\n",
    "    hcC['h'],hcC['h'],hcC['h'],\n",
    "    hcC['c'],hcC['c'],hcC['c'],\n",
    "]\n",
    "\n",
    "print('LOADED')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "/Users/angueyraaristjm/Documents/LiMolec/zfRNAseq/20190827/20190827_DESeq2/genePlots/fpkm\n"
     ]
    }
   ],
   "source": [
    "# %cd ~/Dropbox/Presentations/RevealReboot/resources/20191002_RNAseq/\n",
    "%cd ~/Documents/LiMolec/zfRNAseq/20190827/20190827_DESeq2/genePlots/fpkm/"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "metadata": {},
   "outputs": [],
   "source": [
    "# dark background\n",
    "params = {\"ytick.color\" : \"w\",\n",
    "          \"xtick.color\" : \"w\",\n",
    "          \"axes.labelcolor\" : \"w\",\n",
    "          \"axes.edgecolor\" : \"w\",\n",
    "         \"axes.linewidth\" : 3,\n",
    "         \"xtick.major.width\" : 3,\n",
    "         \"ytick.major.width\" : 3,\n",
    "         \"xtick.major.size\" : 8,\n",
    "         \"ytick.major.size\" : 8,\n",
    "         \"text.color\" : \"w\"}\n",
    "plt.rcParams.update(params)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# white background\n",
    "params = {\"ytick.color\" : \"k\",\n",
    "          \"xtick.color\" : \"k\",\n",
    "          \"axes.labelcolor\" : \"k\",\n",
    "          \"axes.edgecolor\" : \"k\",\n",
    "         \"axes.linewidth\" : 3,\n",
    "         \"xtick.major.width\" : 3,\n",
    "         \"ytick.major.width\" : 3,\n",
    "         \"xtick.major.size\" : 8,\n",
    "         \"ytick.major.size\" : 8,\n",
    "         \"text.color\" : \"w\"}\n",
    "plt.rcParams.update(params)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "***\n",
    "# Index <a id='Index'>∮</a>\n",
    "***\n",
    "- [Plot Total Number of Reads](#nReads)\n",
    "- [PCA Plots](#pcaPlots)\n",
    "- [BarPlot: Opsins](#barOpsins)\n",
    "- [BarPlot: Custom](#barCustom)\n",
    "- [BarPlot: Many Genes](#barMany)\n",
    "- [BarPlot_subPlotsAutomatic: Custom](#barManyAuto)\n",
    "- [Heatmap: Opsins](#heatOpsins)\n",
    "- [Heatmap: Custom](#heatCustom)\n",
    "- [Heatmap: Other Data Sets](#heatOther)\n",
    "- [Sparkline plots](#sparkline)\n",
    "- [PCA](#pca)\n",
    "- [GO clustering](https://go.princeton.edu/) \n",
    "***"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "***\n",
    "## Total number of reads<a id='nReads'>∮</a>\n",
    "***\n",
    "[Back to Index](#Index)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### pygal version (does not import well into Intaglio)\n",
    "> hist = pygal.Histogram(x_title='',y_title='million of mapped reads',show_x_labels=False,title='',style=custom_style,legend_at_bottom=True,legend_at_bottom_columns=4,show_legend=False)  \n",
    "> hist.add('Rod',[(11.27,0.2,0.8),(11.77,1.2,1.8),(9.4,2.2,2.8),(14.99,3.2,3.8),(9.75,4.2,4.8),(9.08,5.2,5.8),])  \n",
    "> hist.add('UV',[(9.42,6.2,6.8),(6.52,7.2,7.8),(10.45,8.2,8.8),(9.46,9.2,9.8),(11.02,10.2,10.8)])  \n",
    "> hist.add('S',[(9.23,11.2,11.8),(9.33,12.2,12.8),(9.37,13.2,13.8),(14.95,14.2,14.8),(12.85,15.2,15.8),(10.96,16.2,16.8)])  \n",
    "> hist.add('M',[(10.40,18.2,18.8),(8.634,19.2,19.8),(7.36,20.2,20.8),(10.32,21.2,21.8),(8.33,22.2,22.8),(9.53,23.2,23.8),(9.526,17.2,17.8)])  \n",
    "> hist.add('L',[(8.99,24.2,24.8),(10.50,25.2,25.8),(10.03,26.2,26.8),(11.08,27.2,27.8),(10.44,28.2,28.8),(10.72,29.2,29.8)])  \n",
    "> hist.render_to_file('test.svg')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "#pyplot version (Sep_2019)\n",
    "plotname = ''\n",
    "n,h = plotNReads()\n",
    "fH, ax = plt.subplots()\n",
    "pH = plt.bar(n, h, width=0.8, bottom=None, align='center', data=None, color=pC['plt'])\n",
    "formatFigure_nReads(plotname, fH, ax, pH)\n",
    "plt.tight_layout()\n",
    "plt.savefig(\"{0}.svg\".format('nReads'), transparent=True, format=\"svg\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "np.mean(h)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": []
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "***\n",
    "## pcaPlots<a id='pcaPlots'>∮</a>\n",
    "***\n",
    "[Back to Index](#Index)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "def formatFigure_pca(plotname, figH, axH, plotH):\n",
    "    font_path = '/System/Library/Fonts/Avenir.ttc'\n",
    "    fontTicks = font_manager.FontProperties(fname=font_path, size=18)\n",
    "    fontLabels = font_manager.FontProperties(fname=font_path, size=22)\n",
    "    fontTitle = font_manager.FontProperties(fname=font_path, size=28)\n",
    "    \n",
    "    axH.ticklabel_format(style='sci',axis='y',scilimits=(0,4))\n",
    "    axH.ticklabel_format(style='sci',axis='x',scilimits=(0,4))\n",
    "    axH.set_title(plotname, fontproperties=fontTitle)\n",
    "    axH.spines['top'].set_visible(False)\n",
    "    axH.spines['right'].set_visible(False)\n",
    "    \n",
    "    for label in (axH.get_xticklabels() + axH.get_yticklabels()):\n",
    "        label.set_fontproperties(fontTicks)\n",
    "    axH.tick_params(axis = 'both', which = 'major', labelsize = 18)\n",
    "    axH.xaxis.offsetText.set_fontsize(18)\n",
    "    axH.yaxis.offsetText.set_fontsize(18)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "#pyplot version (Sep_2019)\n",
    "plotname = 'PCA_RodsCones'\n",
    "pc_df = pd.read_csv(\"~/Documents/LiMolec/zfRNAseq/20190827/20190827_DESeq2/00_rodsVcones/pcaData.csv\")\n",
    "pc1 = pc_df.iloc[:,0]\n",
    "pc2 = pc_df.iloc[:,1]\n",
    "fH, ax = plt.subplots(figsize=[5,5])\n",
    "pH = plt.scatter(pc1, pc2, color=pC['plt'])\n",
    "formatFigure_pca(plotname, fH, ax, pH)\n",
    "ax.set_xlabel('PC1 (var = 21%)', fontproperties=font_manager.FontProperties(fname='/System/Library/Fonts/Avenir.ttc', size=22))\n",
    "ax.set_ylabel('PC2 (var = 9%)', fontproperties=font_manager.FontProperties(fname='/System/Library/Fonts/Avenir.ttc', size=22))\n",
    "plt.xlim(-100,50)\n",
    "plt.ylim(-60,60)\n",
    "plt.tight_layout()\n",
    "plt.savefig(\"{0}.svg\".format(plotname), transparent=True, format=\"svg\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "#pyplot version (Sep_2019)\n",
    "plotname = 'PCA_UVS_ML'\n",
    "pc_df = pd.read_csv(\"~/Documents/LiMolec/zfRNAseq/20190827/20190827_DESeq2/00_USvsLM/pcaData.csv\")\n",
    "pc1 = pc_df.iloc[:,0]\n",
    "pc2 = pc_df.iloc[:,1]\n",
    "fH, ax = plt.subplots(figsize=[5,5])\n",
    "pH = plt.scatter(pc1, pc2, color=[\n",
    "    pC['u'],pC['u'],pC['u'],pC['u'],pC['u'],\n",
    "    pC['s'],pC['s'],pC['s'],pC['s'],pC['s'],pC['s'],\n",
    "    pC['m'],pC['m'],pC['m'],pC['m'],pC['m'],pC['m'],pC['m'],\n",
    "    pC['l'],pC['l'],pC['l'],pC['l'],pC['l'],pC['l']\n",
    "]\n",
    "                )\n",
    "formatFigure_pca(plotname, fH, ax, pH)\n",
    "ax.set_xlabel('PC1 (var = 13%)', fontproperties=font_manager.FontProperties(fname='/System/Library/Fonts/Avenir.ttc', size=22))\n",
    "ax.set_ylabel('PC2 (var = 10%)', fontproperties=font_manager.FontProperties(fname='/System/Library/Fonts/Avenir.ttc', size=22))\n",
    "plt.xlim(-80,80)\n",
    "plt.ylim(-60,60)\n",
    "plt.tight_layout()\n",
    "plt.savefig(\"{0}.svg\".format(plotname), transparent=True, format=\"svg\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "#pyplot version (Sep_2019)\n",
    "plotname = 'PCA_ML'\n",
    "pc_df = pd.read_csv(\"~/Documents/LiMolec/zfRNAseq/20190827/20190827_DESeq2/00_LvsM/pcaData.csv\")\n",
    "pc1 = pc_df.iloc[:,0]\n",
    "pc2 = pc_df.iloc[:,1]\n",
    "fH, ax = plt.subplots(figsize=[5,5])\n",
    "pH = plt.scatter(pc1, pc2, color=[\n",
    "    pC['m'],pC['m'],pC['m'],pC['m'],pC['m'],pC['m'],pC['m'],\n",
    "    pC['l'],pC['l'],pC['l'],pC['l'],pC['l'],pC['l']\n",
    "]\n",
    "                )\n",
    "formatFigure_pca(plotname, fH, ax, pH)\n",
    "ax.set_xlabel('PC1 (var = 23%)', fontproperties=font_manager.FontProperties(fname='/System/Library/Fonts/Avenir.ttc', size=22))\n",
    "ax.set_ylabel('PC2 (var = 10%)', fontproperties=font_manager.FontProperties(fname='/System/Library/Fonts/Avenir.ttc', size=22))\n",
    "plt.xlim(-80,80)\n",
    "plt.ylim(-80,80)\n",
    "plt.tight_layout()\n",
    "plt.savefig(\"{0}.svg\".format(plotname), transparent=True, format=\"svg\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "#pyplot version (Sep_2019)\n",
    "plotname = 'PCA_U_S'\n",
    "pc_df = pd.read_csv(\"~/Documents/LiMolec/zfRNAseq/20190827/20190827_DESeq2/00_UvsS/pcaData.csv\")\n",
    "pc1 = pc_df.iloc[:,0]\n",
    "pc2 = pc_df.iloc[:,1]\n",
    "fH, ax = plt.subplots(figsize=[5,5])\n",
    "pH = plt.scatter(pc1, pc2, color=[\n",
    "    pC['u'],pC['u'],pC['u'],pC['u'],pC['u'],\n",
    "    pC['s'],pC['s'],pC['s'],pC['s'],pC['s'],pC['s'],\n",
    "]\n",
    "                )\n",
    "formatFigure_pca(plotname, fH, ax, pH)\n",
    "ax.set_xlabel('PC1 (var = 13%)', fontproperties=font_manager.FontProperties(fname='/System/Library/Fonts/Avenir.ttc', size=22))\n",
    "ax.set_ylabel('PC2 (var = 10%)', fontproperties=font_manager.FontProperties(fname='/System/Library/Fonts/Avenir.ttc', size=22))\n",
    "plt.xlim(-100,100)\n",
    "plt.ylim(-80,80)\n",
    "plt.tight_layout()\n",
    "plt.savefig(\"{0}.svg\".format(plotname), transparent=True, format=\"svg\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# US vs LM: PC1 = 13% and PC2 = 10%\n",
    "# M vs L: PC1 = 23% and PC2 = 10%\n",
    "# UV vs S: PC1 = 15% and PC2 = 13%"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### tSNE plot"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "#pyplot version (Sep_2019)\n",
    "plotname = 'tSNE'\n",
    "pc_df = pd.read_csv(\"~/Documents/LiMolec/zfRNAseq/20190827/20190827_DESeq2/tSNEData.csv\")\n",
    "pc1 = pc_df.iloc[:,0]\n",
    "pc2 = pc_df.iloc[:,1]\n",
    "fH, ax = plt.subplots(figsize=[6,6])\n",
    "pH = plt.scatter(pc1, pc2, color=pC['plt'])\n",
    "formatFigure_pca(plotname, fH, ax, pH)\n",
    "ax.set_xlabel('tSNE1', fontproperties=font_manager.FontProperties(fname='/System/Library/Fonts/Avenir.ttc', size=22))\n",
    "ax.set_ylabel('tSNE2', fontproperties=font_manager.FontProperties(fname='/System/Library/Fonts/Avenir.ttc', size=22))\n",
    "ax.set_xticks([-1000,-500,0,500,1000])\n",
    "ax.set_yticks([-1000,-500,0,500,1000])\n",
    "plt.xlim(-1000,1000)\n",
    "plt.ylim(-1000,1000)\n",
    "plt.tight_layout()\n",
    "plt.savefig(\"{0}.svg\".format(plotname), transparent=True, format=\"svg\")"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "***\n",
    "## barPlot: Opsins<a id='barOpsins'>∮</a>\n",
    "***\n",
    "[Back to Index](#Index)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# from fx_RNAseq import *"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "genename = 'opn1lw1'\n",
    "n,h = plotBars(genename)\n",
    "h = h\n",
    "fH, ax = plt.subplots()\n",
    "pH = plt.bar(n, h, width=0.8, bottom=None, align='center', data=None, color=pC['plt'])\n",
    "formatFigure_Opsins(genename, fH, ax, pH)\n",
    "plt.savefig(\"{0}.svg\".format(genename), transparent=True, format=\"svg\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "findgene('rho')"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "***\n",
    "## barPlot: Custom Gene<a id='barCustom'>∮</a>\n",
    "***\n",
    "[Back to Index](#Index)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "> Genes to pay attention to but still haven't given interesting results\n",
    "> cdh (cadherins)\n",
    "> protocadherins\n",
    "> semaphorins\n",
    "> plexins (plx)\n",
    "> laminins (lamb) and integrins (itg)\n",
    "> dystroglycan (dag), dystrophin (dmd), pikachurin\n",
    "> neurexins\n",
    "> cadm\n",
    "> dscam and sidekicks (sdk)\n",
    "> pappaa\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "gdf = findgenelist('lrit')\n",
    "gdf"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "genename = 'foxq2'\n",
    "# genename = 'ncdn'\n",
    "# genename = 'chn2'\n",
    "# genename = 'grin1b'\n",
    "# genename = 'cdhr1a'\n",
    "# genename = 'lrit1a'\n",
    "n,h = plotBars(genename)\n",
    "fH, ax = plt.subplots()\n",
    "pH = plt.bar(n, h, width=0.8, bottom=None, align='center', data=None, color=pC['plt'])\n",
    "formatFigure(genename, fH, ax, pH)\n",
    "plt.savefig(\"{0}.svg\".format(genename), transparent=True, format=\"svg\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "***\n",
    "## barPlot: Many Genes Layout<a id='barMany'>∮</a>\n",
    "***\n",
    "[Back to Index](#Index)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "cd '~/Dropbox/LiLab/RNAseq/genePlots/combos'"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "genelist = ['gnat1','gnat2']\n",
    "n, h = plotBars_list(genelist)\n",
    "fH, ax = plt.subplots()\n",
    "pH = plt.bar(n, h, width=0.8, bottom=None, align='center', data=None, color=pC['plt'])\n",
    "formatFigure_list(genelist, fH, ax, pH)\n",
    "# plt.savefig(\"{0}.svg\".format('gnat'), transparent=True, format=\"svg\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "genelist = ['rho','opn1sw1','opn1sw2','opn1mw1','opn1mw2','opn1mw3','opn1mw4','opn1lw1','opn1lw2']\n",
    "n, h = plotBars_list(genelist)\n",
    "fH, ax = plt.subplots(figsize=[6*len(genelist),4])\n",
    "pH = plt.bar(n, h, width=0.8, bottom=None, align='center', data=None, color=pC['plt'])\n",
    "formatFigure_list(genelist, fH, ax, pH)\n",
    "plt.xlim(0,len(genelist)*10)\n",
    "# plt.savefig(\"{0}.svg\".format('opn1'), transparent=True, format=\"svg\")"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "***\n",
    "## barPlot: Many Genes Layout Using Subplots (proper way)<a id='barManySubPlots'>∮</a>\n",
    "***\n",
    "[Back to Index](#Index)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "shrink = 1#2/3"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "genelist = ['gnat1','gnat2']\n",
    "fH, axes = plt.subplots(1,len(genelist),sharey=True,figsize=[6*shrink*len(genelist),4*shrink])\n",
    "for i,g in enumerate(genelist,start=0):\n",
    "    n, h = plotBars(g)\n",
    "    plt.sca(axes[i]) \n",
    "    pH = plt.bar(n, h, width=0.8, bottom=None, align='center', data=None,\n",
    "                 color=pC['plt'])\n",
    "    if  i == 0:\n",
    "        formatFigure(g, fH, axes[i], pH)\n",
    "    else:\n",
    "        formatFigure_right(g, fH, axes[i], pH)\n",
    "plt.savefig(\"{0}.svg\".format('gnat'), transparent=True, format=\"svg\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "genelist = ['rho','rhol']\n",
    "\n",
    "fH, axes = plt.subplots(1,len(genelist),sharey=True,figsize=[6*shrink*len(genelist),4*shrink])\n",
    "for i,g in enumerate(genelist,start=0):\n",
    "    n, h = plotBars(g)\n",
    "    plt.sca(axes[i])\n",
    "    pH = plt.bar(n, h, width=0.8, bottom=None, align='center', data=None,\n",
    "                 color=pC['plt'])\n",
    "    if  i == 0:\n",
    "        formatFigure(g, fH, axes[i], pH)\n",
    "#         plt.ylim(0,3.5e6)\n",
    "#         axes[0].set_yticks(np.arange(0, 4.0e6, step=0.5e6))\n",
    "    else:\n",
    "        formatFigure_right(g, fH, axes[i], pH)\n",
    "plt.savefig(\"{0}.svg\".format('rho'), transparent=True, format=\"svg\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "genelist = ['opn1sw1','opn1sw1']\n",
    "\n",
    "fH, axes = plt.subplots(1,len(genelist),sharey=True,figsize=[6*shrink*len(genelist),4*shrink])\n",
    "for i,g in enumerate(genelist,start=0):\n",
    "    n, h = plotBars(g)\n",
    "    plt.sca(axes[i])\n",
    "    pH = plt.bar(n, h, width=0.8, bottom=None, align='center', data=None,\n",
    "                 color=pC['plt'])\n",
    "    if  i == 0:\n",
    "        formatFigure(g, fH, axes[i], pH)\n",
    "#         plt.ylim(0,2e6)\n",
    "#         axes[0].set_yticks(np.arange(0, 2.5e6, step=0.5e6))\n",
    "    else:\n",
    "        formatFigure_right(g, fH, axes[i], pH)\n",
    "plt.savefig(\"{0}.svg\".format('opn1sw1'), transparent=True, format=\"svg\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "genelist = ['opn1mw1','opn1mw2']\n",
    "\n",
    "fH, axes = plt.subplots(1,len(genelist),sharey=True,figsize=[6*shrink*len(genelist),4*shrink])\n",
    "for i,g in enumerate(genelist,start=0):\n",
    "    n, h = plotBars(g)\n",
    "    plt.sca(axes[i])\n",
    "    pH = plt.bar(n, h, width=0.8, bottom=None, align='center', data=None,\n",
    "                 color=pC['plt'])\n",
    "    if  i == 0:\n",
    "        formatFigure(g, fH, axes[i], pH)\n",
    "    else:\n",
    "        formatFigure_right(g, fH, axes[i], pH)\n",
    "plt.savefig(\"{0}.svg\".format('opn1mwA'), transparent=True, format=\"svg\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "genelist = ['opn1mw3','opn1mw4']\n",
    "\n",
    "fH, axes = plt.subplots(1,len(genelist),sharey=True,figsize=[6*shrink*len(genelist),4*shrink])\n",
    "for i,g in enumerate(genelist,start=0):\n",
    "    n, h = plotBars(g)\n",
    "    plt.sca(axes[i])\n",
    "    pH = plt.bar(n, h, width=0.8, bottom=None, align='center', data=None,\n",
    "                 color=pC['plt'])\n",
    "    if  i == 0:\n",
    "        formatFigure(g, fH, axes[i], pH)\n",
    "        plt.ylim(0,1.0e5)\n",
    "        axes[0].set_yticks(np.arange(0, 1.5e5, step=0.5e5))\n",
    "    else:\n",
    "        formatFigure_right(g, fH, axes[i], pH)\n",
    "\n",
    "\n",
    "plt.savefig(\"{0}.svg\".format('opn1mwB'), transparent=True, format=\"svg\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "genelist = ['opn1lw1','opn1lw2']\n",
    "\n",
    "fH, axes = plt.subplots(1,len(genelist),sharey=True,figsize=[6*shrink*len(genelist),4*shrink])\n",
    "for i,g in enumerate(genelist,start=0):\n",
    "    n, h = plotBars(g)\n",
    "    plt.sca(axes[i])\n",
    "    pH = plt.bar(n, h, width=0.8, bottom=None, align='center', data=None,\n",
    "                 color=pC['plt'])\n",
    "    if  i == 0:\n",
    "        formatFigure(g, fH, axes[i], pH)\n",
    "        plt.ylim(0,1.0e5)\n",
    "        axes[0].set_yticks(np.arange(0, 1.5e5, step=0.5e5))\n",
    "    else:\n",
    "        formatFigure_right(g, fH, axes[i], pH)\n",
    "#         axes[i].set_yticks(np.arange(0, 2.0e6, step=0.5e6))\n",
    "\n",
    "plt.savefig(\"{0}.svg\".format('opn1lw'), transparent=True, format=\"svg\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Using svgutils to automatically tile these figures the right way\n",
    "Figure(\"2400\", \"2000\",\n",
    "       SVG(\"rho.svg\").move(0, 290*0),\n",
    "       SVG(\"opn1sw.svg\").move(0, 290*1),\n",
    "       SVG(\"opn1mwA.svg\").move(0, 290*2),\n",
    "       SVG(\"opn1mwB.svg\").move(0, 290*3),\n",
    "       SVG(\"opn1lw.svg\").move(0, 290*4)\n",
    "        ).save(\"OpsinsAll.svg\")\n",
    "svgreload()"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Phototransduction"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "gdf = findgenelist('gngt')\n",
    "gdf = gdf.sort_values(by=['symbol'])\n",
    "gdf"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# genelist = ['saga','sagb','arr3a','arr3b']\n",
    "# plotName = 'arrestins'\n",
    "\n",
    "# genelist = ['gnat1','gnat2']\n",
    "# plotName = 'gnat'\n",
    "\n",
    "# genelist = ['gngt1','gngt2a','gngt2b']\n",
    "# plotName = 'gngt'\n",
    "\n",
    "# genelist = ['gnb1a','gnb1b','gnb3a','gnb3b','gnb4b','gnb5a','gnb5b']\n",
    "# plotName = 'gnb'\n",
    "\n",
    "# genelist = ['pde6a','pde6b','pde6c','pde6d','pde6ga','pde6gb','pde6c','pde6ha']\n",
    "# plotName = 'pde'\n",
    "\n",
    "# genelist = ['gucy1a1','gucy1b1','gucy1b2','gucy2c','gucy2d','gucy2f','gucy2g']\n",
    "# plotName = 'GC'\n",
    "\n",
    "# genelist = ['gucy2d','gucy2f']\n",
    "# plotName = 'GC_retinaSpecific'\n",
    "\n",
    "# genelist = ['guca1a','guca1b','guca1c','guca1d','guca1e','guca1g']\n",
    "# plotName = 'GCAP'\n",
    "\n",
    "# genelist = ['cnga1a','cnga1b','cnga3a','cnga3b','cnga4','cnga2a','cnga2b']\n",
    "# plotName = 'CNGalpha'\n",
    "\n",
    "# genelist = ['cnga1a','cnga1b','cnga3a','cnga3b']\n",
    "# plotName = 'CNGalpha_retinaSpecific'\n",
    "\n",
    "# genelist = ['cngb1a','cngb3.1','cngb3.2']\n",
    "# plotName = 'CNGbeta'\n",
    "\n",
    "\n",
    "# genelist = ['rgs9a','rgs9b','rgs9bp',\n",
    "#             'rgs3a','rgs6',\n",
    "#             'rgs11','rgs12a','rgs16','rgs20']\n",
    "# plotName = 'RGS'\n",
    "\n",
    "# genelist = ['grk1a','grk1b',\n",
    "#             'grk3',\n",
    "#             'grk4','grk5','grk5l',\n",
    "#             'grk6',\n",
    "#             'grk7a','grk7b']\n",
    "# plotName = 'GRK'\n",
    "\n",
    "# genelist = ['rcvrna','rcvrnb','rcvrn2','rcvrn3','ncalda','ncaldb']\n",
    "# plotName = 'rcvrn'\n",
    "\n",
    "\n",
    "\n",
    "fH, axes = plt.subplots(1,len(genelist),sharey=True,figsize=[6*len(genelist),4])\n",
    "for i,g in enumerate(genelist,start=0):\n",
    "    n, h = plotBars(g)\n",
    "    plt.sca(axes[i])\n",
    "    pH = plt.bar(n, h, width=0.8, bottom=None, align='center', data=None,\n",
    "                 color=pC['plt'])\n",
    "    if  i == 0:\n",
    "        formatFigure(g, fH, axes[i], pH)\n",
    "    else:\n",
    "        formatFigure_right(g, fH, axes[i], pH)\n",
    "plt.savefig(\"{0}.svg\".format(plotName), transparent=True, format=\"svg\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Others"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "gdf = findgenelist('cacn')\n",
    "gdf = gdf.sort_values(by=['symbol'])\n",
    "gdf"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "genelist = ['ctbp1','ctbp2a']\n",
    "plotName = 'ctbp'\n",
    "\n",
    "genelist = ['cadm1a','cadm1b','cadm2a','cadm2b','cadm3','cadm4']\n",
    "plotName = 'cadm'\n",
    "\n",
    "genelist = ['tjp1a','tjp1b','tjp2a','tjp2b','tjp3']\n",
    "plotName = 'tjp'\n",
    "\n",
    "genelist = ['nrxn1a','nrxn1b','nrxn2a','nrxn2b','nrxn3a','nrxn3b',]\n",
    "plotName = 'nrxn'\n",
    "\n",
    "genelist = ['eml1','eml2','eml3']\n",
    "plotName = 'eml'\n",
    "\n",
    "genelist = ['elfn1b','elfn2a']\n",
    "plotName = 'elfn'\n",
    "\n",
    "genelist = ['cpne1','cpne2','cpne3','cpne4a','cpne4b','cpne5a','cpne5b','cpne7','cpne8','cpne9']\n",
    "plotName = 'cpne'\n",
    "\n",
    "# genelist = ['tbx2a','tbx2b','tbxas1','tbx21']\n",
    "genelist = ['tbx2a','tbx2b']\n",
    "plotName = 'tbx2'\n",
    "\n",
    "genelist = ['chn1','chn2']\n",
    "plotName = 'chn'\n",
    "\n",
    "genelist = ['ncdn','ncdn']\n",
    "plotName = 'ncdn'\n",
    "\n",
    "genelist = ['sema7a','sema6a','sema6d','sema4ab','sema4ba','sema3fa','sema3fb']\n",
    "plotName = 'sema'\n",
    "\n",
    "genelist = ['cacna1aa','cacna1ab',\n",
    "            'cacna1da','cacna1db',\n",
    "            'cacna1fa','cacna1fb',\n",
    "            'cacna2d1a','cacna2d2a',\n",
    "            'cacna2d4a','cacna2d4b',\n",
    "            'cacna1sb',\n",
    "           ]\n",
    "plotName = 'CaV'\n",
    "\n",
    "# genelist = ['stxbp1b','stxbp4','stxbp2','stxbp5a','stxbp5b','stxbp5l','stxbp3']\n",
    "genelist = ['stxbp1a','stxbp1b',\n",
    "            'stxbp2','stxbp3','stxbp4',\n",
    "            'stxbp5a','stxbp5b','stxbp5l',\n",
    "            'stxbp6','stxbp6l']\n",
    "plotName = 'stxBP'\n",
    "\n",
    "genelist = ['stx1a','stx1b','stx2a','stx2b','stx3a',\n",
    "            'stx4','stx5a','stx5al','stx6','stx7l',\n",
    "            'stx8','stx10','stx11a','stx11b.1','stx11b.2',\n",
    "            'stx12','stx12l','stx16','stx17','stx18']\n",
    "plotName = 'stx'\n",
    "\n",
    "genelist = ['prickle1a','prickle1b','prickle2a','prickle2b','prickle3',]\n",
    "plotName = 'prickle'\n",
    "\n",
    "genelist = ['pcdh15a','pcdh15b','pcdh8','pcdh11','pcdh10a']\n",
    "plotName = 'pdch'\n",
    "\n",
    "genelist = ['chn1','chn2']\n",
    "plotName = 'chn'\n",
    "\n",
    "\n",
    "genelist = ['lrit1a','lrit1b','lrit2','lrit3a','lrit3b']\n",
    "plotName = 'lrit'\n",
    "\n",
    "genelist = ['pou4f1','pou4f2','pou4f3','rbpms','grm6a','grm6b','vsx2','gfap','prkca','syt2a','rpe65a', 'opn4']\n",
    "plotName = 'controls'\n",
    "\n",
    "\n",
    "# genelist = ['gpx1a','gpx1b','gpx4a','gpx4b']\n",
    "# genelist = ['rgs9bp','RGS9BP']\n",
    "# genelist = ['chchd1','chchd10']\n",
    "# genelist = ['yipf6','yipf1']\n",
    "# genelist = ['sema3fb','sema3fa','sema6d']\n",
    "# genelist = ['esama','esamb']\n",
    "\n",
    "# genelist = ['pcdh1gc6','tex264a']\n",
    "# plotName = 'test'\n",
    "\n",
    "            \n",
    "fH, axes = plt.subplots(1,len(genelist),sharey=False,figsize=[6*len(genelist),4])\n",
    "for i,g in enumerate(genelist,start=0):\n",
    "    n, h = plotBars(g)\n",
    "    plt.sca(axes[i])\n",
    "    pH = plt.bar(n, h, width=0.8, bottom=None, align='center', data=None,\n",
    "                 color=pC['plt'])\n",
    "    if  i == 0:\n",
    "        formatFigure(g, fH, axes[i], pH)\n",
    "    else:\n",
    "        formatFigure_right(g, fH, axes[i], pH)\n",
    "plt.savefig(\"{0}.svg\".format(plotName), transparent=True, format=\"svg\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "gdf = findgenelist('opn1')\n",
    "# gdf = gdf.sort_values(by=['symbol'])\n",
    "# gdf = gdf.sort_values(by=['baseMean'], ascending=False)\n",
    "# gdf['symbol']\n",
    "gdf"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### According to Neuhauss recoverin1a in enriched in rods and UV-cones"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "***\n",
    "## heatMap: Opsins<a id='barManySubPlots'>∮</a>\n",
    "***\n",
    "[Back to Index](#Index)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Not sure this is helpful.  \n",
    "Maybe use this to show differential expression by normalizing by mean expression in correct type"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "gdf = findgenelist('opn1')\n",
    "gdf2 = findgene('rho')\n",
    "gdf2\n",
    "gdf = gdf.append(gdf2)\n",
    "gdf.index = range(len(gdf))\n",
    "gdf\n",
    "\n",
    "Opsins = [\"Rhodopsin (rho)\",\"UV-opsin (opn1sw1)\", \"S-opsin (opn1sw2)\",\"M-opsin (opn1mw1)\",\"M-opsin (opn1mw2)\",\"M-opsin (opn1mw3)\",\"M-opsin (opn1mw4)\",\"L-opsin (opn1lw1)\",\"L-opsin (opn1lw2)\"]\n",
    "Samples = [\"\",\"\",\"R\",\"\",\"\",\"\",\n",
    "           \"\",\"\",\"U\",\"\",\"\",\n",
    "           \"\",\"\",\"S\",\"\",\"\",\"\",\n",
    "           \"\",\"\",\"\",\"M\",\"\",\"\",\"\",\n",
    "           \"\",\"\",\"L\",\"\",\"\",\"\",\n",
    "          ]\n",
    "Samples = [\n",
    "          ]\n",
    "\n",
    "data = gdf.iloc[0:9,7:37]\n",
    "data = data.reindex([8,4,5,7,0,2,6,1,3])\n",
    "data.index = range(len(data))\n",
    "# np.around(np.divide(data.values,1e6),2)\n",
    "# opsinData = np.divide(data.values,1e6)\n",
    "opsinData = np.divide(data.values,1)\n",
    "\n",
    "# # log2 representation does not work well\n",
    "# log2matrix = np.vectorize(log2)\n",
    "# opsinData = log2matrix(opsinData)\n",
    "\n",
    "params = {\"ytick.color\" : \"w\",\n",
    "          \"xtick.color\" : \"w\",\n",
    "          \"axes.labelcolor\" : \"w\",\n",
    "          \"axes.edgecolor\" : \"w\",\n",
    "         \"axes.linewidth\" : 2,\n",
    "         \"xtick.major.width\" : 2,\n",
    "         \"ytick.major.width\" : 2,\n",
    "         \"xtick.major.size\" : 0,\n",
    "         \"ytick.major.size\" : 0}\n",
    "plt.rcParams.update(params)\n",
    "\n",
    "plotName = \"heatMap_opsins\"\n",
    "fig, ax = plt.subplots()\n",
    "\n",
    "im, cbar = heatmap(opsinData, Opsins, Samples, ax=ax,\n",
    "                   cmap=\"inferno\", cbarlabel=\"fpkm\")\n",
    "\n",
    "\n",
    "# texts = annotate_heatmap(im, valfmt=\"{x:.1f} t\")\n",
    "\n",
    "# fig.tight_layout()\n",
    "plt.savefig(\"{0}.svg\".format(plotName), transparent=True, format=\"svg\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "***\n",
    "## heatMap: Custom<a id='heatCustom'>∮</a>\n",
    "***\n",
    "[Back to Index](#Index)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "plotName = \"test\"\n",
    "\n",
    "# gdf = findgene('rho')\n",
    "# gdf2 = findgenelist('opn1sw')\n",
    "# gdf = gdf.append(gdf2)\n",
    "# gdf2 = findgenelist('opn1mw')\n",
    "# gdf2 = gdf2.sort_values(by=[\"symbol\"])\n",
    "# gdf = gdf.append(gdf2)\n",
    "# gdf2 = findgenelist('opn1lw')\n",
    "# gdf2 = gdf2.sort_values(by=[\"symbol\"])\n",
    "# gdf = gdf.append(gdf2)\n",
    "\n",
    "# gdf.index = range(len(gdf))\n",
    "# plotName = \"heatMap_opsins\"\n",
    "\n",
    "# gdf = findgenelist('slc24') -> NCKX transporter; slc25a5 is mitochondrial but known to express in eye\n",
    "# plotName = \"slc24\"\n",
    "\n",
    "# gdf = findgenelist('sema')\n",
    "# gdf = gdf.sort_values(by=[\"symbol\"])\n",
    "# plotName = \"heatMap_sema\"\n",
    "\n",
    "# gdf = findgenelist('cdh')\n",
    "# gdf = gdf.sort_values(by=[\"baseMean\"],ascending=False)\n",
    "# plotName = \"heatMap_cdh\"\n",
    "\n",
    "# gdf = findgenelist('eph')\n",
    "# gdf = gdf.drop([1,14,15,16]) #not ephrin receptors\n",
    "# gdf = gdf.sort_values(by=[\"symbol\"])\n",
    "# plotName = \"heatMap_eph\"\n",
    "\n",
    "\n",
    "# gdf = findgenelist('efn')\n",
    "# gdf = gdf.sort_values(by=[\"symbol\"])\n",
    "# plotName = \"heatMap_efn\"\n",
    "\n",
    "# gdf = findgenelist('pdgf')\n",
    "# gdf = findgenelist('rgr')\n",
    "# gdf = findgenelist('lmo')\n",
    "# gdf.index = range(len(gdf))\n",
    "# gdf.sort_values(by=[\"log2FoldChange\"])\n",
    "\n",
    "\n",
    "## genes in 2019 Papal, Swaroop: # igsf11, epha10, tuft1, prtg, lrfn2, vasn\n",
    "\n",
    "# gdf = findgenelist('igsf')\n",
    "# gdf = findgenelist('tuft')\n",
    "# gdf = findgenelist('prtg')\n",
    "# gdf = findgenelist('lrfn')\n",
    "# gdf = findgenelist('vasn')\n",
    "\n",
    "# gdf = findgenelist('rdh')\n",
    "# gdf = gdf.sort_values(by=[\"symbol\"])\n",
    "# plotName = \"rdhHeatMap\"\n",
    "\n",
    "# gdf = findgenelist('cyp2')\n",
    "# gdf = findgenelist('cyb')\n",
    "\n",
    "# gdf = findgenelist('sag')\n",
    "# plotName = \"heatMap_sag\"\n",
    "\n",
    "# gdf = findgenelist('dio')\n",
    "# gdf = gdf.sort_values(by=[\"symbol\"])\n",
    "# plotName = \"heatMap_dio\"\n",
    "\n",
    "# gdf = findgenelist('lrrn')\n",
    "# gdf = gdf.sort_values(by=[\"symbol\"])\n",
    "# plotName = \"heatMap_lrrn\"\n",
    "\n",
    "# gdf = findgenelist('dnmt')\n",
    "# gdf = gdf.sort_values(by=[\"symbol\"])\n",
    "# plotName = \"heatMap_dnmt\"\n",
    "\n",
    "# gdf = findgenelist('tet')\n",
    "\n",
    "# gdf = findgenelist('isl')\n",
    "# gdf = gdf.drop([1])\n",
    "\n",
    "# gdf = findgenelist('gj')\n",
    "# gdf = gdf.sort_values(by=[\"symbol\"])\n",
    "# plotName = \"heatMap_gj\"\n",
    "\n",
    "# gdf = findgenelist('fgf')\n",
    "# gdf = findgenelist('bmp')\n",
    "# gdf = findgenelist('fzd')\n",
    "# gdf = findgenelist('edar')\n",
    "# gdf = findgenelist('axin')\n",
    "\n",
    "# gdf = findgenelist('cadm') # cadm2a in S-Cones probably(?) and cadm3 in all photoreceptors (Hunter, JCN, 2011)\n",
    "# gdf = gdf.sort_values(by=[\"symbol\"])\n",
    "# plotName = \"heatMap_cadm\"\n",
    "\n",
    "# gdf = findgenelist('lrit') # lrit1b for rods and cones; lrit2 for cones; lrit3 for rods\n",
    "# gdf = gdf.sort_values(by=[\"symbol\"])\n",
    "# plotName = \"heatMap_lrit\"\n",
    "\n",
    "# gdf = findgenelist('elfn') # not detected\n",
    "# gdf = gdf.sort_values(by=[\"symbol\"])\n",
    "# plotName = \"heatMap_elfn\"\n",
    "\n",
    "# gdf = findgenelist('tenm') # for RGCs in zebrafish (Antinucci, Cell Reports, 2013)\n",
    "\n",
    "# gdf = findgenelist('foxo')\n",
    "\n",
    "# gdf = findgenelist('sdk')\n",
    "\n",
    "# gdf = findgenelist('crmp')\n",
    "\n",
    "\n",
    "\n",
    "# Even More\n",
    "# gdf = findgenelist('thr')\n",
    "# gdf = gdf.sort_values(by=[\"symbol\"])\n",
    "# plotName = \"heatMap_thr\"\n",
    "\n",
    "# gdf = findgenelist('lingo')\n",
    "# gdf = gdf.sort_values(by=[\"symbol\"])\n",
    "# plotName = \"heatMap_lingo\"\n",
    "\n",
    "# gdf = findgenelist('mitf')\n",
    "\n",
    "# gdf = findgenelist('prickle')\n",
    "# gdf = gdf.sort_values(by=[\"symbol\"])\n",
    "# plotName = \"heatMap_prickle\"\n",
    "\n",
    "# gdf = findgenelist('rdh')\n",
    "# gdf = findgenelist('sdh')\n",
    "# lingo, rx, mitf\n",
    "\n",
    "# gdf = findgenelist('dag')\n",
    "# can't find pikachurin: # gdf = findgenelist('egflam') gdf = findgenelist('si:dkeyp-100d11.2-001')\n",
    "\n",
    "# gdf = findgenelist('kcn')\n",
    "# gdf = gdf.sort_values(by=[\"baseMean\"], ascending=False)\n",
    "# gdf = gdf.drop(np.arange(20,109))\n",
    "# gdf\n",
    "\n",
    "# gdf = findgenelist('kcnj')\n",
    "# gdf = gdf.sort_values(by=[\"baseMean\"], ascending=False)\n",
    "# gdf\n",
    "\n",
    "# gdf = findgenelist('slc')\n",
    "# gdf = gdf.sort_values(by=[\"baseMean\"], ascending=False)\n",
    "# gdf = gdf.drop(np.arange(40,410))\n",
    "\n",
    "# gdf = findgenelist('rcv')\n",
    "# gdf = gdf.sort_values(by=[\"symbol\"])\n",
    "# plotName = \"heatMap_recoverin\"\n",
    "\n",
    "# gdf = findgenelist('sag')\n",
    "# gdf = gdf.append(findgenelist('arrb'))\n",
    "# gdf = gdf.append(findgenelist('arr3'))\n",
    "# gdf = gdf.sort_values(by=[\"symbol\"])\n",
    "# plotName = \"heatMap_arr\"\n",
    "\n",
    "\n",
    "# gdf = findgenelist('cng')\n",
    "# gdf = gdf.sort_values(by=[\"symbol\"])\n",
    "# plotName = \"heatMap_cng\"\n",
    "\n",
    "# gdf = findgenelist('cnga')\n",
    "# gdf = gdf.sort_values(by=[\"symbol\"])\n",
    "# plotName = \"heatMap_cnga\"\n",
    "\n",
    "# gdf = findgenelist('cngb')\n",
    "# gdf = gdf.sort_values(by=[\"symbol\"])\n",
    "# plotName = \"heatMap_cngb\"\n",
    "\n",
    "# gdf = findgenelist('myo')\n",
    "# gdf = gdf.sort_values(by=[\"symbol\"])\n",
    "\n",
    "# gdf = findgenelist('crx')\n",
    "# gdf = gdf.sort_values(by=[\"baseMean\"], ascending=False)\n",
    "# plotName = \"heatMap_crx\"\n",
    "\n",
    "\n",
    "# gdf = findgenelist('neurod')\n",
    "# gdf = gdf.sort_values(by=[\"symbol\"])\n",
    "# plotName = \"heatMap_neurod\"\n",
    "\n",
    "# gdf = findgenelist('ror')\n",
    "# gdf = gdf.drop([5,6])\n",
    "# gdf = gdf.sort_values(by=[\"symbol\"])\n",
    "# plotName = \"heatMap_ror\"\n",
    "\n",
    "\n",
    "# gdf = findgenelist('nr')\n",
    "# gdf = gdf.sort_values(by=[\"baseMean\"], ascending=False)\n",
    "# gdf = gdf.drop(np.arange(5,76))\n",
    "# plotName = \"heatMap_nrl\"\n",
    "\n",
    "\n",
    "# gdf = findgenelist('pcdh')\n",
    "# gdf = gdf.drop([1]) #highest ones that obscure others.\n",
    "# # gdf = gdf.drop([1,15]) #highest ones that obscure others.\n",
    "# gdf = gdf[gdf.baseMean > .45]\n",
    "# gdf = gdf.sort_values(by=[\"baseMean\"], ascending=False)\n",
    "# plotName = \"heatMap_pcdh_partial\"\n",
    "\n",
    "# gdf = findgenelist('hcn')\n",
    "# gdf = gdf.sort_values(by=[\"symbol\"])\n",
    "\n",
    "# gdf = findgenelist('gabr')\n",
    "# gdf = gdf.drop([0,2]) #highest ones that obscure others.\n",
    "# gdf = gdf.sort_values(by=[\"symbol\"])\n",
    "\n",
    "# gdf = findgenelist('opn4')\n",
    "# gdf = gdf.drop([1,2]) #highest ones that obscure others.\n",
    "# gdf = gdf.sort_values(by=[\"baseMean\"], ascending=False)\n",
    "# gdf = gdf.drop(np.arange(5,76))\n",
    "# gdf = gdf.sort_values(by=[\"symbol\"])\n",
    "# plotName = \"heatMap_opn4\"\n",
    "\n",
    "# CSNB causes: Nyx, cacna1f, GPR179 (Kamermans, 2019)\n",
    "# gdf = findgenelist('nyx') # bipolar cell -> not expressed in photoreceptors\n",
    "# gdf = findgenelist('gpr1') # bipolar cell -> not expressed in photoreceptors\n",
    "# gdf = findgenelist('cacna') # bipolar cell -> not expressed in photoreceptors\n",
    "\n",
    "\n",
    "# For Karen Carleton\n",
    "\n",
    "# gdf = findgenelist('rx')\n",
    "# gdf = gdf.drop([2,3,4,5,7,8,9,10,11,12,13,14,15,16,17,18,19])\n",
    "# gdf = gdf.sort_values(by=[\"symbol\"])\n",
    "# # gdf = gdf.drop([1])\n",
    "# plotName = \"heatMap_rx\"\n",
    "\n",
    "# gdf = findgenelist('tbx')\n",
    "# gdf = gdf.drop([3,17])\n",
    "# plotName = \"heatMap_tbx\"\n",
    "# gdf = gdf.drop([1,2])\n",
    "\n",
    "# gdf = findgenelist('tbx2')\n",
    "# gdf = gdf.drop([0,3,4])\n",
    "# gdf = gdf.sort_values(by=[\"symbol\"])\n",
    "# plotName = \"heatMap_tbx2\"\n",
    "\n",
    "# gdf = findgenelist('six')\n",
    "# gdf = gdf.sort_values(by=[\"symbol\"])\n",
    "# plotName = \"heatMap_six\"\n",
    "\n",
    "# gdf = findgenelist('lin')\n",
    "# gdf = gdf.drop([1,3,4,5,10,11,13,16])\n",
    "# plotName = \"heatMap_lin\"\n",
    "\n",
    "\n",
    "# gdf = findgenelist('sall') #sall4 mutations cause Duane syndrome; sall4 seems to be related to sall1 in zebrafish\n",
    "# plotName = \"heatMap_sall\"\n",
    "\n",
    "# For Mark Hankins\n",
    "# gdf = findgenelist('opn')\n",
    "# gdf = gdf.sort_values(by=[\"symbol\"])\n",
    "# gdf = gdf.drop([1,4,24,0,3,23,5,11,2])\n",
    "\n",
    "# gdf = findgenelist('rgr')\n",
    "# gdf = findgenelist('lrit')\n",
    "\n",
    "# gdf = findgenelist('ndrg') #potential target of tbx2\n",
    "\n",
    "# from YiRong Peng's talk: ON SACs: slit2 and fezf1; OFF SACs: tenm3 and rnd1?; contactins; Foveal cones: vtn and cacna1f\n",
    "# gdf = findgenelist('cacna2d2b')\n",
    "# gdf = gdf.drop([0])\n",
    "# gdf = gdf.sort_values(by=[\"symbol\"])\n",
    "\n",
    "# gdf = findgenelist('nrxn')\n",
    "# gdf = gdf.sort_values(by=[\"symbol\"])\n",
    "# plotName = \"heatMap_nrxn\"\n",
    "\n",
    "\n",
    "# gdf = findgenelist('frmpd')\n",
    "# gdf = gdf.sort_values(by=[\"symbol\"])\n",
    "# plotName = \"heatMap_frmpd\"\n",
    "\n",
    "# gdf = findgenelist('ntng')\n",
    "# gdf = gdf.sort_values(by=[\"symbol\"])\n",
    "# plotName = \"heatMap_netrins\"\n",
    "\n",
    "# gdf = findgenelist('dscam')\n",
    "# gdf = gdf.sort_values(by=[\"symbol\"])\n",
    "# plotName = \"heatmap_dscams\"\n",
    "\n",
    "\n",
    "# gdf = findgenelist('dag')\n",
    "# gdf = gdf.sort_values(by=[\"symbol\"])\n",
    "# plotName = \"heatmap_dag\"\n",
    "\n",
    "# gdf = findgenelist('syne') #involved in photoreceptor migration to ONL\n",
    "# gdf = gdf.sort_values(by=[\"symbol\"])\n",
    "# plotName = \"heatmap_syne\"\n",
    "# gdf = findgenelist('rasgrf') #also involved in phtoreceptor migration\n",
    "\n",
    "# gdf = findgenelist('syt')\n",
    "# plotName = \"heatmap_syt\"\n",
    "\n",
    "# gdf = findgenelist('rfx')\n",
    "# plotName = \"heatmap_rfx\"\n",
    "\n",
    "# gdf = findgenelist('anks') #anks3-/- causes RP\n",
    "# gdf = gdf.sort_values(by=[\"symbol\"])\n",
    "# plotName = \"heatmap_anks\"\n",
    "\n",
    "# gdf = findgenelist('spock') #https://zfin.org/action/image/view/ZDB-IMAGE-050208-655 -> GCL enriched?\n",
    "# gdf = gdf.sort_values(by=[\"symbol\"])\n",
    "# plotName = \"heatmap_anks\"\n",
    "\n",
    "# gdf = findgenelist('hcls') # Laura Campello: downstream of nrl and crx in mouse\n",
    "\n",
    "# gdf = findgenelist('slc32')  #from GO analysis: slc32a1 = L-cone specific + synaptic vesicle transporter.\n",
    "# gdf = findgenelist('slc') \n",
    "# gdf = gdf.sort_values(by=[\"baseMean\"],ascending=False)\n",
    "# gdf = gdf[gdf.baseMean > 100]\n",
    "\n",
    "# gdf = findgenelist('frmpd') #frmpd2 interacts with lrit1 in cones \n",
    "\n",
    "# from sqRNAseq + antibodies: rab32,ell2, ankrd33, pex5l, egr1, lamb1, chst3? or 7?, flna, cdh20, synj1\n",
    "# gdf = findgenelist('ell') # ell2 in sqRNAseql seems cone specific\n",
    "# gdf = findgenelist('ankrd33') #it's there, 33aa cone enriched; 33ba rods and 33bb cones.\n",
    "# gdf = findgenelist('pex5') #pek5l in sqRNAseq; pex5 in rods\n",
    "# gdf = findgenelist('egr') # egr1 looks interesting\n",
    "\n",
    "\n",
    "\n",
    "# gdf = findgenelist('gnat')\n",
    "# gdf2 = findgenelist('gnb')\n",
    "# gdf = gdf.append(gdf2)\n",
    "# gdf2 = findgenelist('gng')\n",
    "# gdf = gdf.append(gdf2)\n",
    "# gdf.index = range(len(gdf))\n",
    "# plotName = \"heatMap_transducin\"\n",
    "\n",
    "# gdf = findgenelist('aldh1')\n",
    "gdf = findgenelist('elovl')\n",
    "gdf = gdf.sort_values(by=[\"baseMean\"],ascending=False)\n",
    "# gdf = findgenelist('gnat')\n",
    "gdf = gdf[gdf.baseMean > 0]\n",
    "# gdf = gdf.drop(np.arange(20,93))\n",
    "# gdf = gdf.drop([4,24])\n",
    "# gdf = gdf.sort_values(by=[\"symbol\"])\n",
    "# gdf = gdf.sort_values(by=[\"padj\"])\n",
    "\n",
    "gdf\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "genenames = gdf['symbol'].values\n",
    "samplenames = [] #skipping names to replace in intaglio\n",
    "data = gdf.iloc[0:,7:37].values\n",
    "\n",
    "# log2matrix = np.vectorize(log2)\n",
    "# data = log2matrix(gdf.iloc[0:,7:37].values)\n",
    "\n",
    "fig, ax = plt.subplots(figsize=[12,12])\n",
    "im, cbar = heatmap(data, genenames, samplenames, ax=ax,\n",
    "                   cmap=\"inferno\", cbarlabel=\"fpkm\")\n",
    "# texts = annotate_heatmap(im, valfmt=\"{x:.1f} t\")\n",
    "\n",
    "plt.savefig(\"{0}.svg\".format(plotName), transparent=True, format=\"svg\")\n",
    "plt.show()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "plotName = 'test'\n",
    "\n",
    "# foxq2 = S specific;\n",
    "\n",
    "# genelist = ['wrb','get4'] # disrupts photoreceptors synapses (Danielle, 2016)\n",
    "# genelist = ['pappaa','igf2bp2b'] # In Miller, 2018 model pappaa cleaves igf-bp to increase igf1 availability. unfortunately, this is supposed to bind mRNA not IGF itself\n",
    "# genelist = ['gpx1a','gpx1b','gpx4a','gpx4b']\n",
    "# genelist = ['rgs9bp','RGS9BP']\n",
    "# genelist = ['chchd1','chchd10']\n",
    "# genelist = ['yipf6','yipf1','meis1b','pbx1a']\n",
    "# genelist = ['sema3fb','sema3fa','sema6d']\n",
    "# genelist = ['esama','esamb']\n",
    "# genelist = ['il34','il7r']\n",
    "# genelist = ['cxxc4','cxxc1a']\n",
    "# genelist = ['mbd3b','mbd3a']\n",
    "# genelist = ['cdr2a','cdr2l']\n",
    "# genelist = ['coro6','coro2a','coro1ca']\n",
    "# genelist = ['auts2a','auts2b']\n",
    "# genelist = ['plxnb1a','plxna2']\n",
    "# genelist = ['cux1a','cux1b','cux2b']\n",
    "# genelist = ['arih1','arih2','arih1l']\n",
    "# genelist = ['crhbp','crhr2','crhr1']\n",
    "# genelist = ['chn1','chn2']\n",
    "# genelist = ['robo1','robo2','robo3','robo4'] #nothing here\n",
    "# genelist = ['ntng1a','ntng2a','ntng2b'] #low counts but interesting. Netrins are important for RGC axons going towards optic disc. ntng2b in mouse mediates rod-hc recognition!\n",
    "# genelist = ['fat1a','fat1b','yap1','scrib','cc2d2a'] #fat1 important for renal development along yap1/hippo and scribble; cc2d2a involved in cilium development\n",
    "# genelist = ['c1qbp','c1qa','c1qb'] #keep this in mind. Beth Stevens work shows that microglia engulf synapses through activation by c1q; maybe c1qbp is produced in healthy cells (inside mitochondria) to keep c1q bound and gets released upon damage.\n",
    "# genelist = ['cc2d2a','rab8'] #Bahcman 2011: cc2d2a-/- causes Joubert syndrom. Important for cilium development by association with rab8. Not expressed in adults?\n",
    "# genelist = ['per2','per3']\n",
    "# genelist = ['gpsm2','gpsm1b','gpr107','gpr108'] #\n",
    "# genelist = ['tbx2a','tbx2b','gdf6a','cnbpa','egr1', 'ndrg1a','ndrg1b'] # fate genes\n",
    "\n",
    "#Hensley, 2011 smarca4\n",
    "# genelist = ['smarca4a','guk1a','guk1b','dtnbp1a','elovl4b','id2b','ckmt2a','ckmt2b','aanat2','pbx1a'] #smarca4 regulated genes: follow up elovl4b, ckmt2a/b, pbx1a\n",
    "# genelist = ['nme2a','nme3',''] #smarca4 regulated genes: follow up elovl4b, ckmt2a/b, pbx1a\n",
    "\n",
    "# genelist = ['adgrl3.1','ntng2a','ntng2b'] #low counts but interesting. \n",
    "# genelist = ['cplx4a','cplx4b','cplx4c'] #low counts but interesting. \n",
    "\n",
    "# genelist = ['frmpd2', 'lrit3a'] #low counts but interesting. \n",
    "genelist = ['si:busm1-57f23.1 ', 'mhc1uba','serinc2'] # \n",
    "\n",
    "# genelist = ['ccdc28a','ccdc25','ccdc173','sox6','prdm9','prdm1b']\n",
    "\n",
    "# genelist = ['per2','bhlhe40','nr1d2b','tefa','aanat2','aanat1'] #circadian genes. Correlate with collection time?\n",
    "\n",
    "# genelist = ['elovl4b','elovl8a'] #a cone specific gene. \n",
    "# genelist = ['pbx1a','egr1','egr3','egr4'] # fishing for new targets\n",
    "# genelist = ['foxg1b','foxq2','foxo3b','lmo4b','nfia'] # fishing for new targets\n",
    "# genelist = ['anxa13','prdx5','mpp5b','mpp6a','serpina10a'] # fishing for new targets\n",
    "# genelist = ['rbp4l','rbp3','impg1a','impg2b','ntm'] # fishing for new targets\n",
    "# genelist = ['tapt1b','tapt1a','frmpd2','wdr37','fabp7a','tle3a'] # fishing for new targets\n",
    "# genelist = ['ankrd33aa','ankrd33ab','dhx32a','dhx32b','wdr17']\n",
    "\n",
    "\n",
    "# genelist = ['slc1a8a','slc1a8b','slc1a4','slc1a7a'] # EAATs (glutamate uptake)\n",
    "# genelist = ['pmela','slc6a6b','naa35','ush2a'] # sex-associated rod genes in Sun, 2018 (all enriched in females, except pmela)\n",
    "\n",
    "# genelist = ['pcloa','pclob']\n",
    "# genelist = ['mtdha','mtdhb']\n",
    "# genelist = ['igsf9b','igsf21a','igsf21b']\n",
    "# genelist = ['rx3','lin9','igsf21b','foxq2','foxo3b','foxg1b','rxrga','smad5']\n",
    "# genelist = ['sall1a','sall1b','sall2','sall4']\n",
    "# genelist = ['rbm8a','rbm39a','rbm39b','rbmx','rbm22','rbm5']\n",
    "# genelist = ['rbm5','rbm4.1','rbm4.2']\n",
    "# genelist = ['kif5ba','kif1aa','kif1ab','kif1bp'] #kif5-/- : Charcot-MarieTooth; Campbell 2013: kif5s in all retinal layers, but only kif5ba and bb are in ONL\n",
    "# genelist = ['kif3a','kif3b','kif16ba','kif19','kif25'] #\n",
    "# genelist = ['robo2','slit1a','nbeaa','prdm1b'] #\n",
    "\n",
    "\n",
    "# genelist = ['pcdh2ab9','pcdh2g28','pcdh15a','pcdh15b','pcdh11','pcdh10a'] #interesting ones\n",
    "# plotName = 'pcdhs'\n",
    "\n",
    "# genelist = ['sema7a','sema6d','sema3fb','sema3fa','sema4ab','sema4ba'] #the good ones\n",
    "# genelist = ['sema6a','sema3aa','sema6e','sema5bb','sema4c','sema6bb','sema6dl','sema5ba'] # the inconsistent ones\n",
    "# genelist = ['sema6ba','sema5a','sema3c','sema3b','sema3e','sema3d','sema3h','sema3ab','sema4e','sema4aa','sema3gb','sema3ga','sema4bb','sema3bl','sema4ga','sema4gb'] # the negative ones\n",
    "\n",
    "# genelist = ['ncdn','flrt3'] \n",
    "# genelist = ['odf2a','odf2b']\n",
    "# genelist = ['syne1a','syne1b','syne2b','syne3']\n",
    "# genelist = ['rasgrf2a','rasgrf2b']\n",
    "# genelist = ['rasgrf2a','rasgrf2b','nmnat2','rbx1','nrf1','unc119.2','unc119b'] # involved in migration\n",
    "\n",
    "# genelist = ['syt5a','syt5b','syt1a','syt1b','syt14a']\n",
    "\n",
    "\n",
    "# genelist = ['myo19','myo16','myo7bb','myo5b','myo1cb','myo5aa','myo7aa','myo6a','myo6b']\n",
    "# genelist = ['fgfr1op','fgfr1op2','fgf1b','fgf13a','fgf11b']\n",
    "# genelist = ['prdm1b','fgfr1op2']\n",
    "# genelist = ['frmpd1a','gpsm2'] # Ulisse + Pahlberg: these proteins inteact with rod transducin \n",
    "\n",
    "# genelist = ['epha7','frmpd2','pkp4','chd7','hook1']\n",
    "# genelist = ['thrb','samd7','samd1a','samd11','dscaml1'] #samd11:https://www.ncbi.nlm.nih.gov/pubmed/27734943\n",
    "# genelist = ['thrb','nrl']\n",
    "# genelist = ['opn3','opn4.1','opn4a','opn4b','opn5','opn6a','opn9'] #\n",
    "            \n",
    "fH, axes = plt.subplots(1,len(genelist),sharey=False,figsize=[6*len(genelist),4])\n",
    "for i,g in enumerate(genelist,start=0):\n",
    "    n, h = plotBars(g)\n",
    "    plt.sca(axes[i])\n",
    "    pH = plt.bar(n, h, width=0.8, bottom=None, align='center', data=None,\n",
    "                 color=pC['plt'])\n",
    "#     if  i == 0:\n",
    "    formatFigure(g, fH, axes[i], pH)\n",
    "#     else:\n",
    "#         formatFigure_right(g, fH, axes[i], pH)\n",
    "plt.savefig(\"{0}.svg\".format(plotName), transparent=True, format=\"svg\")\n",
    "fH.tight_layout()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "***\n",
    "## heatMap: photoreceptors, rods, Müller Glia, microglia<a id='heatOther'>∮</a>\n",
    "***\n",
    "[Back to Index](#Index)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>symbol</th>\n",
       "      <th>genename</th>\n",
       "      <th>baseMean</th>\n",
       "      <th>log2FoldChange</th>\n",
       "      <th>lfcSE</th>\n",
       "      <th>pvalue</th>\n",
       "      <th>padj</th>\n",
       "      <th>R1</th>\n",
       "      <th>R2</th>\n",
       "      <th>R3</th>\n",
       "      <th>...</th>\n",
       "      <th>M4</th>\n",
       "      <th>M5</th>\n",
       "      <th>M6</th>\n",
       "      <th>S7</th>\n",
       "      <th>L1</th>\n",
       "      <th>L2</th>\n",
       "      <th>L3</th>\n",
       "      <th>L4</th>\n",
       "      <th>L5</th>\n",
       "      <th>L6</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <td>2</td>\n",
       "      <td>ntng1a</td>\n",
       "      <td>netrin g1a</td>\n",
       "      <td>2.272131</td>\n",
       "      <td>-3.854488e-08</td>\n",
       "      <td>0.001443</td>\n",
       "      <td>0.852794</td>\n",
       "      <td>0.940516</td>\n",
       "      <td>3.848269</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>4.270796</td>\n",
       "      <td>...</td>\n",
       "      <td>8.871858</td>\n",
       "      <td>16.010263</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.625306</td>\n",
       "      <td>4.321891</td>\n",
       "      <td>14.157668</td>\n",
       "      <td>10.496059</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>1.117653</td>\n",
       "      <td>0.00000</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>0</td>\n",
       "      <td>ntng2a</td>\n",
       "      <td>netrin g2a</td>\n",
       "      <td>2.939897</td>\n",
       "      <td>-9.308646e-07</td>\n",
       "      <td>0.001443</td>\n",
       "      <td>0.000001</td>\n",
       "      <td>0.000019</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>...</td>\n",
       "      <td>1.057146</td>\n",
       "      <td>1.961797</td>\n",
       "      <td>1.531727</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>15.550220</td>\n",
       "      <td>2.619469</td>\n",
       "      <td>2.976087</td>\n",
       "      <td>19.422771</td>\n",
       "      <td>4.641261</td>\n",
       "      <td>1.60925</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>1</td>\n",
       "      <td>ntng2b</td>\n",
       "      <td>netrin g2b</td>\n",
       "      <td>9.467815</td>\n",
       "      <td>1.694546e-07</td>\n",
       "      <td>0.001443</td>\n",
       "      <td>0.822228</td>\n",
       "      <td>0.928472</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>30.756323</td>\n",
       "      <td>8.037255</td>\n",
       "      <td>...</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>8.868321</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.655206</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>2.423789</td>\n",
       "      <td>0.00000</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>3 rows × 37 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "   symbol    genename  baseMean  log2FoldChange     lfcSE    pvalue      padj  \\\n",
       "2  ntng1a  netrin g1a  2.272131   -3.854488e-08  0.001443  0.852794  0.940516   \n",
       "0  ntng2a  netrin g2a  2.939897   -9.308646e-07  0.001443  0.000001  0.000019   \n",
       "1  ntng2b  netrin g2b  9.467815    1.694546e-07  0.001443  0.822228  0.928472   \n",
       "\n",
       "         R1         R2        R3  ...        M4         M5        M6  \\\n",
       "2  3.848269   0.000000  4.270796  ...  8.871858  16.010263  0.000000   \n",
       "0  0.000000   0.000000  0.000000  ...  1.057146   1.961797  1.531727   \n",
       "1  0.000000  30.756323  8.037255  ...  0.000000   8.868321  0.000000   \n",
       "\n",
       "         S7         L1         L2         L3         L4        L5       L6  \n",
       "2  0.625306   4.321891  14.157668  10.496059   0.000000  1.117653  0.00000  \n",
       "0  0.000000  15.550220   2.619469   2.976087  19.422771  4.641261  1.60925  \n",
       "1  0.000000   0.000000   0.000000   0.655206   0.000000  2.423789  0.00000  \n",
       "\n",
       "[3 rows x 37 columns]"
      ]
     },
     "execution_count": 4,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "geneTag = 'ntng'\n",
    "gdf = findgenelist(geneTag)\n",
    "# gdf = gdf.sort_values(by=[\"baseMean\"],ascending=False)\n",
    "# gdf = gdf[gdf.baseMean > 0]\n",
    "# gdf = gdf.drop(np.arange(20,93))\n",
    "# gdf = gdf.drop([4,24])\n",
    "gdf = gdf.sort_values(by=[\"symbol\"])\n",
    "# gdf = gdf.sort_values(by=[\"padj\"])\n",
    "\n",
    "rdf = findgenelist_rods(geneTag)\n",
    "rdf = rdf.sort_values(by=[\"symbol\"])\n",
    "# rdf = rdf[rdf.baseMean > 0]\n",
    "\n",
    "mdf = findgenelist_glia(geneTag)\n",
    "mdf = mdf.sort_values(by=[\"symbol\"])\n",
    "# mdf = mdf[mdf.baseMean > 0]\n",
    "\n",
    "sqdf = findgenelist_sqcones(geneTag)\n",
    "sqdf = sqdf.sort_values(by=[\"symbol\"])\n",
    "# sqdf = sqdf[sqdf.baseMean > 0]\n",
    "\n",
    "hcdf = findgenelist_haircell(geneTag)\n",
    "hcdf = hcdf.sort_values(by=[\"symbol\"])\n",
    "# hcdf = hcdf[hcdf.baseMean > 0]\n",
    "gdf"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 5,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "image/png": "iVBORw0KGgoAAAANSUhEUgAAA70AAADECAYAAABTGEY4AAAABHNCSVQICAgIfAhkiAAAAAlwSFlzAAALEgAACxIB0t1+/AAAADh0RVh0U29mdHdhcmUAbWF0cGxvdGxpYiB2ZXJzaW9uMy4xLjAsIGh0dHA6Ly9tYXRwbG90bGliLm9yZy+17YcXAAAcEUlEQVR4nO3dfZgcVZ3o8W9PEkKAQUCEcAEDRL0oghjERzdcyCiScUDgcpcFVlggi7xkkatyS1b3Ksl67y5aCoiYoCxIVFgBFSMQE0AmQuDyEgIYEEQCAYK8yoMMLwl56fvHqc709HRPOjOdqpmq7+d5+unqU6fr9+uampn+9TlVXSqXy0iSJEmSlEdtWScgSZIkSdKmYtErSZIkScoti15JkiRJUm5Z9EqSJEmScsuiV5IkSZKUW6OzTkDFEkXRNsAXgAvjOH4163wkSZIk5ZsjvUrbNsC5yb0kSZIkbVKO9IooiqYA3VVNZeDPwP3Av8dxfGcWeUn1zOuYPwM4rKu78yMDrQceBvbt6u780ADb6gRuBHbr6u58pvXZqsimTu/ZG/jfwGRge+BJ4CogXjCrfWWWuUmbypglE6cQ3lM8B7x79aRlaxr0awOeACYAn1k9adkNqSUppeS+vfeaQvh9aN9v6cOvZ5xOoTnSq2p7AbsDewLTgLeB26MompppVtLgXAHsM69j/j4D9DkRuMWCV602dXrPPsCdwFvAMcCHge8DEXBOhqlJadmR8AFkI1OBnVPKRVLBOdKrak/HcVz5FOqxKIpuBv4fcDawILu0pEFZCDwFHA98uXblvI757wCOBE5KNSsVxb8D9yyY1X5SVdsjU6f33AT8KZuUpFTNA04FftVg/anAdcDRqWUkqbAc6VVDcRyXgduBd2edi7Sxuro7y8CPgePmdcyv97fuGOBNGr8hk4biPYRTRPpYMKv9sQWz2ssZ5COlbRYwdcySif3eQ4xZMnE8YRR4dupZSSokR3q1IXsDyysPoij6OPA1wjlqowlv6r4dx3GfwiGKorGEKXzHE87XeZZQgPyspl8bMIMwzXRH4CEgiuO4+hxjabDmEI7XKcCtNetOBK7q6u5clXZSKoSngMOmTu85b8Gs9pezTkbKwO3AH4B/JFzAsto04PHVk5Z1j1kyMfXEJBWPI73qJ4qitiiKdo2i6DzCOTcXJe2fBH5HKII7gI8CtwDXRFH0v6qeXwJ+DZxCuIjL3oRpTAcDP6kJdzbhH+LpwCTCP8jLoigas6len4qjq7tzGXAH4cOX9eZ1zH8P8DfA5VnkpUL4CmGWzCNTp/dEU6f3OGNGRXQJMG3MkomjKg1jlkwsEf7vX5JZVpIKx5FeVXs5iiIIx0Ub8Hvg6DiO50VRtBnwQ+CCOI6rL8LycBRFTwOXRFF0TRzHTxP+mX0c2DOO4z8n/R6LouhW4OaamAcAN8Zx/BuAKIqmAdvGcbx6E71GFc8VwHfmdcyf3tXdWbli7knAg13dnf2mn0qtsGBW+33Jxay+Thjl+ubU6T23AN9cMKv9t9lmJ6Xmx8B5QBdwfdL2KWAnwkwcSUqFI72q9jfAvoTRrxXA5DiOf56smwzsAZxf53k/AnqAw5PHJwBXVhW8AMRxvA6YWfPcO4Fjoyg6J4qiDwJr4zh+qRUvRrnVzPmQ1X2uAcaQHJ/zOuaXCMeoo7zapBbMan98waz2fwDeBfwdsC1wy9TpPccP/EwpH1ZPWtYDXAl8rqr5c8B/rp607NVsspJURBa9qvZYHMePEq50O5YwQlGxE/BWHMcv1D4pueDVE8B/qer7ZKMYtU8H/g/h/MqlwIooij7X71lSr78SiodGtgfWv5nq6u58jXCF0Eqh0QGMJ7wRkza5BbPa31owq/3nhA8WuwnXMZCKYhbQNWbJxJ3HLJm4A3AEXsBKUsosetVPHMevEi5C9cUoivZKmp8DxkVRtENt/+Qc3t2TPpW+uzXY/HtrYq2L4/hbcRx/ANiFMN3pB1EUfWTIL0R5tRzYZV7H/K0brN+LqouvJeYAnfM65r+T8AHL3K7uzr9ssgylOhbMal8N/Jzwt04qhNWTlv0euJtw6tPJwIOrJy1bnG1WkorGoleNzAHuAWYnRe2dhELii3X6/gPwDsLFqyCMoP19FEU7VXdKtvO1mrYJURS9AyCO42fjOP4qYSTvQ617KcqZbmAV8IXaFfM65u8PHET/ryG6BXiR8Kbrf+DUZm1iU6f3TJ06vecDdVb9N+CBtPORMjab8Pf3FBzllZQBL2SluuI4LkdR9E/AfcCJcRxfEUXRqcD1URRtCVxNKDy6gK8C/xLH8VPJ0y8lfAfqnVEU/QvhglgTCNOma6/KfC2wLrn68yuEKajjgEWb9AVqxOrq7vzrvI75XwR+OK9j/jbAVcAbwCcI54z/oqu788aa56yb1zH/J8C/Ai8BN6Wctgpk6vSe0YQpzBOnTu/5GuGrW9oJb/oPI1zATyqSa4ELgK2p+epCqSDefd/ee71Z01beb+nDT9XtrZZzpFcNxXH8IPB9II6iaLs4jm8G9gfeB/wGuA3oBA6N4/hbVc8rE77q6IfANwiF8/cI35N6Qk2YI4FlwFzgXsL5lofHcfzHTfjSNMJ1dXdeBnyG8DVXvwWWANOBbwLHNnjaHMK56nO6ujvXpZGnimnBrPY1wCcJX/d2JuFv4C+ALYD9FsxqfzDD9KTUrZ60bBVhhs2c1ZOW1b7xl4rgYcL1bqpvyzLNqGBK5XIzF0IttBkNlvMW07j5jWnc/MY0bn5jGjffcbOIadz8xjRufmMat0UsejesegeVchzTuPmNadz8xjRufmMaN99xi/Raixa3SK+1aHGL9FpzF9fpzZIkSZKk3LLolSRJkiTllkWvJEmSJCm3LHolSZIkSbll0StJkiRJyi2LXkmSJElSbhXmK4uuvvrq8uLFi7NOQxqSjsWfzDoF5cx33z856xSklrj1lH2zTkEase46cfOsU5BaYr+lD58MXFHbXpiR3m233TbrFCRJkiRJm85u9RoLU/RKkiRJkorHoleSJEmSlFuFOacXKI/f6oDUgr3wxh3J0qjUYgZrAdhlq3TP/Vzx+m+TpWxeb6m0ZapRy+U3kqU0X+9ayuUypdLoFGOGuEE2P9tixM3utWZ5TG23xaRUo77y5pJkqRg/26LFDcfy2BRjvp3cF2cfAxy19ZmpRv3laxcDcMUHpqUW86Q/XJ4sZfWeZlyqUcvltyiXy+zafnCqcSvvHeM9zkgtZvTEbAD+cfuzUosJcNnLFyVL2RxTbW3bpBp13bpXk6VM/hfMBGbUrnGkV5IkSZKUWxa9kiRJkqTcsuiVJEmSJOWWRa8kSZIkKbcseiVJkiRJuWXRK0mSJEnKLYteSZIkSVJuDbeitxN4Ddgl60QkSZIkSSNfK4reVn2788nA14AnWrQ9SZIkSVLBDaeidxkwFXh1kM+fAZQHuEmSJEmSCmZ0k/0mApcCjwJ7AVsQRmYvBnYGFgI3AkuAU4C1wARCsXk08EKynTOBE4C3gfuBHYFzgOXAbQ1ivw+IgXHAdsBM4Pom85YkSZIkFVizRe9aYDLwZWAxoeA9G5gCPJ7ck9wfAuxHKGRnAqcn90cCnwYOAlYmfRcQit6BPEcopF8CtgbuwKJXkiRJktSEjZne/ACh4IUwKjuhQb9bCAVvbb9jgHMJBS+E0eHuJuKuBj4P3ARcB+zWoN8MoDTATZIkSZJUMM2O9AL0VC2vonHB3KhficGdW3sB8AhhBBlgxSC2IUmSJEkqoFZcyGoN4RzfDbka+DowNnl8ENDRxPPGE84VhjAleoeNzE+SJEmSVFAbM9LbyJXAIuCn9Ban9VxHKGC7CVOcnwXmsuHR338DfgC8CDwELCUUzquGlLUkSZIkKfeaLXqXAwdXPV5B78WrvpHcKhZWLS9KbhWzk1vFDcArNbGm1Dy+F5jUZJ6SJEmSJK3XiunNzToUmFb1eF9gFH3PAZYkSZIkqWVaMb25WfcAFwJnEL6n92XgtBTjS5IkSZIKJs2i9yXgsynGkyRJkiQVXJrTmyVJkiRJSpVFryRJkiQptyx6JUmSJEm5ZdErSZIkScqtUrlczjqHtJRLpVLWOUhDUi6X8ThWK3lMKS88lqXB8/dHeVEul2cCM2rbHemVJEmSJOVWml9ZNAyMSjHW2gxiGjffcYv0WrOP29a2TWoR1617FYD9x52UWkyAe9+6Ilkq1s92s9E7pxbx7TXPAnDb5CNSiwlw4B1zk6W0/82vSe6z+dnuutWnUov4zOs3A7D9lh9NLSbAy2/ckyzl/28j9P59HDtm19Rirlr9DACjR+2QWkyANWtfTJay+dleuucpqUb93KP/AUD75numFrNn5aMAvGPc3qnFBPjrW0sBWHTA4anGPWDRrwHYbauuVOMuf30eAGNGj0817uo1zzdc50ivJEmSJCm3LHolSZIkSbll0StJkiRJyi2LXkmSJElSbln0SpIkSZJyy6JXkiRJkpRbFr2SJEmSpNyy6JUkSZIk5ZZFryRJkiQptyx6JUmSJEm5NZyK3k8DdwO/A24Eds42HUmSJEnSSNeKovesFmxja2A2cARwEPB94IIWbFeSJEmSVGDDpeh9E/gY8Hzy+DbgvRu5jRlAeYCbJEmSJKlgRjfZbyJwKfAosBewBXAycDFhGvJCwpTkJcApwFpgAqHYPBp4IdnOmcAJwNvA/cCOwDnAcnoLXoDjgDuT5fcBMTAO2A6YCVy/Ea9RkiRJklRQzRa9a4HJwJeBxYSC92xgCvB4ck9yfwiwH6GQnQmcntwfSThv9yBgZdJ3AaHorXYg8JUkHsBzhEL6JcI06Duw6JUkSZIkNWFjpjc/QCh4IUw/ntCg3y2Egre23zHAuYSCF8LocHfNcz8EzAGOIhS7AKuBzwM3AdcBuzWIOwMoDXCTJEmSJBVMsyO9AD1Vy6toXDA36ldi4HNr9wCuJUxtfqCq/QLgEcIIMsCKJvOVJEmSJBVcKy5ktYZwju+GXA18HRibPD4I6EiWdwTmAqcCd9U8bzzhXGEIU6J3GEKukiRJkqQC2ZiR3kauBBYBP6W3OK3nOkIB202Y4vwsodAtA8cC2xOmKFe8DhwG/BvwA+BF4CFgKaFwXtWC3CVJkiRJOdZs0bscOLjq8Qp6L171jeRWsbBqeVFyq5id3CpuAF4Bvpvc6rkXmNRknpIkSZIkrdeK6c3NOhSYVvV4X2AUfc8BliRJkiSpZVoxvblZ9wAXAmcQvqf3ZeC0FONLkiRJkgomzaL3JeCzKcaTJEmSJBVcmtObJUmSJElKlUWvJEmSJCm3LHolSZIkSbll0StJkiRJyq1SuVzOOoe0lEulUtY5SENSLpfxOFYreUwpLzyWpcHz90d5US6XZwIzatsd6ZUkSZIk5VaaX1k0DIxKMdZaAFZfPz7FmDDmM88D0Lnl9FTjzn9jFgA/2POUVOOe9uh/ADBt+7NSjXv5yxclS+kfU6+/9Z0UY8JW484G4Np9Tkg17tG//wkAa9b9NNW4o9uOB2D1VVumFnPM378BwKrzN08tJsDYL60EoK20dapx15VfA+DUd6X7e/vDl8Lv7UnvTC/uFX8JMc/bPd2/yf/8ZPibPGGrqanGfer1BQD81y2PSjXuH9/4ZbKU/t/keI8zUowJ0ROzk6U0XytUXm9WcY/a+szUIv7ytYuTpWLt42LEDTFLpXEpxoRy+a1kKZt9vMXYialGfXPVsmQpq2OqP0d6JUmSJEm5ZdErSZIkScoti15JkiRJUm5Z9EqSJEmScsuiV5IkSZKUWxa9kiRJkqTcsuiVJEmSJOWWRa8kSZIkKbeGe9E7Bdgn6yQkSZIkSSPTcCl6Pw3cDfwOuBHYOWmfgkWvJEmSJGmQhlr0ntWCHLYGZgNHAAcB3wcuaMF2JUmSJEkFNxyK3jeBjwHPJ49vA97bgu1KkiRJkgpudBN9JgKXAo8CewFbACcDFxOmIS8kTEleApwCrAUmAGXgaOCFZDtnAicAbwP3AzsC5wDL6S14AY4D7qx6/B7gN0A7sBr4OnB7nTxnAOc28XokSZIkSQXRTNG7FpgMfBlYTCh4zyacb/t4ck9yfwiwH6GQnQmcntwfSThv9yBgZdJ3AaHorXYg8JUkXsWxwCeAPwM7AfOBLuDZZl6gJEmSJKm4mp3e/ACh4IUw/XhCg363EAre2n7HEEZhVyaPFwLdNc/9EDAHOAp4rqr9e4SCl6T9KkIBLUmSJEnSgJotenuqllcN8LxG/UqE6c6N7AFcS5ja/MAGcmm0rRnJukY3SZIkSVLBDPVCVmsI5/huyNWEc3HHJo8PAjqS5R2BucCpwF11nvt5wrTmSt/jCFOcJUmSJEkaUDPn9A7kSmAR8FPChawauQ4YT5jSvJJwPu5cwojtscD2hJHaiteBw5LlnwGXES5ktZZQBHs+ryRJkiRpg5opepcDB1c9XkHvxau+kdwqFlYtL0puFbOTW8UNwCvAd5NbPTOayE+SJEmSpLqGOr25WYcC06oe7wuMou85wJIkSZIktdRQpzc36x7gQuAMwvf0vgycllJsSZIkSVJBpVX0vgR8NqVYkiRJkiQB6U1vliRJkiQpdRa9kiRJkqTcKpXL5axzSMVNN91Uvvnmm7NOQxqSjsWfzDoF5cx33z856xSklrj1lH2zTkEase46cfOsU5BaYr+lD8+kzjcAOdIrSZIkScoti15JkiRJUm5Z9EqSJEmScqsw5/QCJwFXZJyDJEmSJGnTOIk6NV+Ril5JkiRJUsE4vVmSJEmSlFsWvZIkSZKk3LLolSRJkiTllkWvJEmSJCm3LHolSZIkSbll0StJkiRJyi2LXkmSJElSbln0SpIkSZJyy6JXkiRJkpRbFr2SJEmSpNyy6JUkSZIk5ZZFryRJkiQptyx6JUmSJEm5ZdErSZIkScqt0VknIGWonHUCkiRJGpJS1glo+HOkV5IkSZKUWxa9kiRJkqTccnqzBLSVNgsLpcoMmd7Pg0qV5VL1Z0RtfdcBpfXrq9tG9elXKvXfbp+2mv7V60uM6tfWtn4bVeuStrY+sfpvt60mVht1tkH/bVS3rY9f9dxKW1u5/z5qq7PdSr+2qtlJvfFLDfuVqtcly323kbSV6vWjYb8+65psayvVbqt6+5U+1XlUtlVvG/3bBtu/+jn1cxqgf91t9J4RULuNenm2VZ1BUC+PyvpS3W2U++VRiV/vNZeqcqvdRnVb/W2UN9i/+jl92tbnWee1rI9Vtd26+7I8wHbrvJa6+dZpq+lXb/t9c9vwNuptb6B1fWI1ud1SW2Ub6wbcRm3/vtut89y25rZReW7ftvKg+tPsdittbY3zrl7f5zU089w66/rk1mS/9fErbaXqdfXyrWyjt6n3ufRbV/WLV9W/f1vVLxL9Oq5fV7Xhtrb+/de39e9X7tM2qon+o/q1UWcb5VK97Y5qGLPvdptoq1rX2za6Tv86baXqttF97vv0L/VdV71c6rONMf3bkuVSnf6V5bY+baP69WtL8h3FZ5E2hiO9kiRJkqTcsuiVJEmSJOWWRa8kSZIkKbc8p1d5Mgo4FNiNvsf2+ZlkI0mSJClzFr3Kk+uBlcBSYN0G+kqSJEkqAIte5ckuwD5ZJyFJkiRp+PCcXuXJb4BDsk5CkiRJ0vDhSK/y5C7gOsKHOasJ3wJYBrbOMilJkiRJ2bHoVZ58B/g44Zzecsa5SJIkSRoGnN6sPPkT8BAWvJIkSZISjvQqT54DFhLO7V1V1e5XFkmSJEkFZdGrPHkyuW2W3MBRX0mSJKnQLHqVJ3OA5TVt+2eQhyRJkqRhwnN6lSe/AHauenwgcHlGuUiSJEkaBix6lSenA78CxgNdwEXJvSRJkqSCcnqz8uRe4CzgJmAl8CngpUwzkiRJkpQpi17lwfX0vWDVFsBfgcuSx4ennpEkSZKkYcGiV3nw7awTkCRJkjQ8WfQqD35XtTwe+Chh5Pde4PlMMpIkSZI0LHghK+XJKcA9wFHA3wJ3AdMyzUiSJElSphzpVZ5EwIeBvySP3wnciV9bJEmSJBWWI73KkxVAT9XjHuCZjHKRJEmSNAw40qs8eRa4G5hLOKf3CMJ05y8l68/PKC9JkiRJGXGkV3nwk+T+cOBX9H590VzgOaA9uUmSJEkqGEd6lQf7AROAp4Hv1Vn/SrrpSJIkSRouLHqVB5cA84HdgcVV7SXCqO8eWSQlSZIkKXtOb1YeXAS8H/gRocCt3HbHgleSJEkqtFK5XN5wLymfPPglSZJGtlLWCWj4c6RXkiRJkpRbFr2SJEmSpNzyQlZS4NSYoaueLu7+HDr3Z+u5T1vL/dla7s/Wcn+2nvtUI5YjvZIkSZKk3LLolSRJkiTllkWvJEmSJCm3LHolSZIkSbll0StJkiRJyi2v3qwim5l1Ajnj/mwt92fruU9by/3ZWu7P1nJ/tp77VCNWqVwub7iXJEmSJEkjkNObJUmSJEm5ZdErSZIkScoti15JkiRJUm5Z9KqoOoE/Ao8D/5xxLnmxHFgKPAAszjaVEely4EXgoaq27YCbgT8l99tmkNdIVW9/zgCeJRyjDwBd6ac1Yu0KdAOPAA8D/zNp9xgdvEb7dAYep4OxOXAP8CBhf1YuurQ7cDfhGL0a2CyT7EaeRvvzCuBJeo/PfbNITtpYXshKRTQKeAz4FLACuBc4DvhDlknlwHLgI8DLGecxUh0IvA78GPhg0vYt4BXgPMKHM9sC52SS3chTb3/OSNq+nVFOI9lOyW0J0A7cBxwJnITH6GA12qd/h8fpYJSALQn7bgywiPBBwpeAXwI/Ay4hFHGzM8pxJGm0P08HbgB+nl1q0sZzpFdF9FHCCO8TwNuEf4RHZJqRBLcRiodqRwBzkuU5hDfEak69/anBe45QnAH0EEYnd8ZjdCga7VMNTplQoEEo0sYkbZ+gt0DzGG1eo/0pjUgWvSqinYFnqh6vwDcarVAGbiKMVpyacS55sSPhjTHJ/Q4Z5pIXZwK/J0x/diru4OwGfJgwZdRjtDV2o3efgsfpYI0iTLl9kTDdfhnwKrAmWe//+41Tuz8rx+f/JRyfFwBjs0lN2jgWvSqiUp02P70cusnAJODTwD8RppdKw8lsYCLhHLTngO9km86ItBXwC+ALwGsZ55IXtfvU43Tw1hL22y6EWV3vr9PH//fNq92fHwS+AuwJ7E84p9/TGTQiWPSqiFYQLiBSsQvw54xyyZPKPnwRuI7wD1JD8wLhnD+S+xczzCUPXiC8iVsHXIrH6MYaQyjOriScIwkeo0PVaJ96nA7Nq8BC4GPANsDopN3/94NT2Z+dhA9iysAq4Ed4fGqEsOhVEd0LvJdwRcfNgGOBX2ea0ci3JeFCLJXlQ+h71VwNzq+BE5PlE4G5GeaSBztVLf93PEY3Rgm4jHDe6flV7R6jg9don3qcDs67CAUuwDjgYMK+7Qb+Nmn3GG1evf35KL3HZ4lwfrTHp0YEr96souoCLiScr3I54fwUDd4ehNFdCJ+oX4X7dGP9JzAF2J4w0nMu8CvgGuDdwNPA0XhxpmbV259TCFP1yoSrjZ9G7/moGtgBwO2EryVbl7R9lXCOn8fo4DTap8fhcToY+xAuVDWKMKhzDfCvhP9PPyNMxb0fOJ4wSqmBNdqftxIK4hLhfN/T6b3glTRsWfRKkiRJknLL6c2SJEmSpNyy6JUkSZIk5ZZFryRJkiQptyx6JUmSJEm5ZdErSZIkScoti15JkiRJUm5Z9EqSJEmScuv/A2NgQtOQxyvfAAAAAElFTkSuQmCC\n",
      "text/plain": [
       "<Figure size 1152x1152 with 2 Axes>"
      ]
     },
     "metadata": {
      "needs_background": "dark"
     },
     "output_type": "display_data"
    },
    {
     "data": {
      "image/png": "iVBORw0KGgoAAAANSUhEUgAABXsAAAEpCAYAAADCofF1AAAABHNCSVQICAgIfAhkiAAAAAlwSFlzAAALEgAACxIB0t1+/AAAADh0RVh0U29mdHdhcmUAbWF0cGxvdGxpYiB2ZXJzaW9uMy4xLjAsIGh0dHA6Ly9tYXRwbG90bGliLm9yZy+17YcXAAAgAElEQVR4nOzdf3hcZZ3w//ek5afA1wAruNaC0LCCrrDgztbFNUWwBMEVITwUEQUeFug2YFeegYfdp5DaXX38ZpWC6Wa5WLlgBQQNIgu4paA0ULUMv0Vt3RTthvCj9EcECqU/5/njPpNO0plkkkxyJifv13Wda845c865PzPNp3fz6X3uk8rlckiSJEmSJEmSxreauAOQJEmSJEmSJI2cxV5JkiRJkiRJSgCLvZIkSZIkSZKUABZ7JUmSJEmSJCkBLPZKkiRJkiRJUgJY7JUkSZIkSZKkBJgcdwCSJEmSJEmFuho6lgJLpy6ub445lBHraug4FPg98IGpi+tXR/tuAZi6uP78mMJSQnRma5cCS+vSPc0xh1IRndnaQ4nypS7dszradwtAXbrn/LjiGk8s9kqSJEmSpETqauiYDvwYaJi6uD5b5jm1QBPwWaAO2A1YBdwLXDd1cf2GUQpXilVntrY3X+rSPWXlS2e2dsB8qUv3mC9jzGkcJEmSJElSUr0OPAO8Uc7BXQ0dJwD/RShcLQI+AUwHbgS+CPy6q6HjqNEJVYrdkPKlM1s7aL50ZmvNlzHmyF5JkiRJkpRIUxfXrwBOLOfYroaOPwPuJxSqrpi6uD5X8PYvuxo6vgvMA16teKBSFahL95SdL53Z2j75Upfu6ZMvndla8yUmFnslSZIkSZLgX4GfsWuhF4Cpi+vfADJjHpVUnXrzpV+hF4C6dI/5EhOLvZIkSZIkqSp1NXR8HvgH4BDgt8BVUxfXPxy990eEkYOnAlOA14DvAF+durh+R3TMoez6cLQcUE+4zfwM4ElCUSoNzChW6B0gvr2Ba4DPA+8h3NLeMnVx/XeH+Dn3AlqieN4NPAFcPnVx/XNDuY4mrs5s7S65Upfuebjg/ZL5Upfu2REdcyi7PhxtwHwpVugdIMai+VKX7hlSvnRma4vmS126x3zBOXslSZIkSVJ1OhO4FLgY+CShCHVvV0PHwdH7ZwAHALOBo4G/BS4EvlzGta8DJgEzonM+BrwDLCs3uK6GjhThYVafA64APgp8C/hmV0PHJeVeJ/LPUSznFMTSOsRraOIqmiud2dqDC46JNV86s7Ul86UzW2u+VJAjeyVJkiRJUjU6APjY1MX1GwG6GjouIMz/2QDcMnVx/Y2E+ULzVnY1dEwjFLWuG+Ta66Yurr8gv9HV0LE/sGHq4vrtQ4jvIuAI4E+nLq5fH+37VVdDx0vAHV0NHXdMXVz/ZpnX+jhwx9TF9R1RPGcOIQ4lUGe29u+Ay4E/Ah4HbgIeAA4EFhKmT1gVHX4A8LG6dM/G6Nw+uQJQl+7ZJV86s7Vl50tduqc3XzqztfsDG+rSPcPKl7p0T2++dGZrXwLu6MzW3lGX7hlSvtSlezqieMyXAhZ7JUmSJElSNbo/X+gFmLq4/s2uho5Owi3o+ZG1nyeMFPxzwu3cewEvlXHt7/Xb3gDs39XQMWkIBd8vAN8tKPTm43wompZhBnBfmdf6OXBZV0PHG8DiqYvrVw12gpKrM1v7QcK0CWcTpls4G/gmcAeQA/4F+F3BKffnC70AdemeNzuztb25El2z4vnSma2dNISC7xeA7xYUevOxPhRNyzCDIeZLZ7b2DWBxQdFbWOyVJEmSJEnV6dUi+/7AzlrGN4H/CXw9Wu8hFLLKuSV8Xb/tXwB7AscDj5YZ33uBr3Q1dBS7DX4P4H1lXgfgK4TP9hXg210NHS8QHhR37xCuoYSoS/es7MzWHpefSxf4Rme29p+BDwDr69I9Pf1OGSxXoErypTNbW/F86czWvkAY6Wy+YLFXkiRJkiRVpx2l3uhq6NgTuAyYNXVx/d0F+0ueM5Cpi+uf62royAJ/T/nFq1eAh4Bvl3i/WAGuVPubgKuBq7saOuqAK4G7uho6Dpu6uP7lcq+j5Cgo9Oa3twOlRrAO+HPfma3tzZe6dM/dBfuHlS916Z7nOrO1seVLXbqnN186s7W9+dKZrT2sLt0z4fPFB7RJkiRJkqTxZi/CALb+Ixw/M4Jrzgb+qquhoyWaIqKProaOVFdDx81dDR1/G+26HTgZeGnq4vqV+QV4Adhr6uL6P5TbcFdDx5FdDR27A0xdXN8J/C/CaMcjR/B5pLxRy5fObG1LNEVEH53Z2lRntvbmzmztLvlSl+5ZmV+I8qUu3VN2vnRma4/szNbuDlCX7jFf+nFkryRJkiRJGlemLq7v6WroeAb4RjSNwhbgUmB3hlnrmLq4/umuho5Tge8DM7oaOhYBTwObgQ8BcwkPmGqNTrmJMJfqo10NHfOBlYTb7K8Gdutq6Dh+6uL6QUdORqOUlwDPdTV0NANbCQ/m2gA8M5zPIhWqS/f0dGZrnyFMB1GRfKlL9zzdma3tzZfObG3Z+dKZrd0lXzqztcf3H81cTDRKeQnwXGe2thnzZReO7JUkSZIkSePRWYQCz0OEBzutJdzOvf9wLzh1cf1SQoHqPsJt78uAp4D5hNvVPzR1cf3T0bE5oAH4IbAQeA5YBPwM+FQ5hd7oOu8AnyI8eOsnwGPAIcDJUxfXbxjuZ5H6qXi+1KV7ljJIvtSle56Ojh0wX8op9EbXKZkvdeke8wVI5XK5uGOQJEmSJEmSJI2QI3slSZIkSZIkKQEs9kqSJEmSJElSAljslSRJkiRJkqQEsNgrSZIkSZIkSQlgsVeSJEmSJEmSEsBiryRJkiRJkiQlgMVeSZIkSZIkSUoAi72SJEmSJEmSlAAWeyVJkiRJkiQpASz2SpIkSZIkSVICTI47AEmSJEmSVFWaS6xPJM0l1iea5hLrE01zifWJpLnE+kTTXGK9aqRyuVzcMUiSJEmSpOpRWChIxRZFvPwOAr+HwO/B7yCv6r8Hp3GQJEmSJEmSpASw2CtJkiRJkiRJCWCxV5IkSZIkSZISwGKvJEmSJEmSJCWAxV5JkiRJkiRJSgCLvZIkSZIkSZKUAKlcLhd3DJIkSZIkKbL+mytzbz20Ju4wpKqx+aunxx2CVFXq0j0XALcUe8+RvZIkSZIkVZHJB+0ZdwiSpOp2aKk3LPZKkiRJkiRJUgJY7JUkSZIkSZKkBHDOXkmSJEmSqkvuzo+cH3cMsTrn+VsBmPf+ppgjic+CF1sB2PLQB2KOJF67f+r35HI5fnfmx+IOJVaH/3A5AJ2f/cuYI4lP3b0/B+A3n66POZJ4HfXjDnK53Hygudj7juyVJEmSJEmSpASw2CtJkiRJkiRJCWCxV5IkSZIkSZISwGKvJEmSJEmSJCWAxV5JkiRJkiRJSgCLvZIkSZIkSZKUABZ7JUmSJEmSJCkBLPZKkiRJkiRJUgJY7JUkSZIkSZKkBLDYK0mSJEmSJEkJYLFXkiRJkiRJkhLAYq8kSZIkSZIkJYDFXkmSJEmSJElKAIu9kiRJkiRJkpQAFnslSZIkSZIkKQEs9kqSJEmSJElSAljslSRJkiRJkqQEsNgrSZIkSZIkSQlgsVeSJEmSJEmSEsBiryRJkiRJkiQlgMVeSZIkSZIkSUoAi72SJEmSJEmSlAAWeyVJkiRJkiQpASz2SpIkSZIkSVICWOyVJEmSJGnoGoA3gClxByJJUp7FXkmSJEnSRHF5ha5zATAP+F2FridJUkVY7JUkSZIkTRSVKva+AJwM/KFC15MkqSIs9kqSJEmSkuRw4KfAvwAdwBPAh4GlwPui1wxwIvA94DbgMeBR4KCC6zQBj0fv3QDcBRwavfcosLFI20cA9wJLgCeBz5SIsRnIDbBIkjQsk+MOQJIkSZKkCtoOHA9cSSi4XgBcAcwAVkWvRK8zgeOA1cB84NLo9XTgFKAeeCc69kHgqkHafgW4CFgL7Af8DLhvxJ9IkqQyObJXkiRJkpQ0zxIKvRBG4R5S4riHCYXe/sedDVxLKPRCGA38SBntbgUuI4zsvYedI4ElSRoTFnslSZIkSUnzZsH6Zkr/7lvquBTDm07hOmAdYcTwicDrJY5rjtootUiSNCwWeyVJkiRJE8U2YO8yjrsLuAbYI9quB04o47yDgaej9RnAe4YYnyRJI+KcvZIkSZKkieJ2YBnhoWxPD3DcPYTC7SOEqRxeIjx4bbDRvl8DbgReA34FPE8oGG8eUdSSJJXJYq8kSZIkKUlWAycVbHez86FsC6Ilb2nB+rJoyWuLlrz7gQ392prRb/sJ4NghxCpJUkU5jYMkSZIkSX2dClxYsH0MMIm+c/xKklR1HNkrSZIkSVJfWWAhMBvYQnjo2iWxRiRJUhks9kqSJEmS1Nda4Ny4g5AkaaicxkGSJEmSJEmSEsBiryRJkiRJkiQlgMVeSZIkSZIkSUoAi72SJEmSJEmSlAAWeyVJkiRJkiQpASz2SpIkSZIkSVICWOyVJEmSJEmSpASw2CtJkiRJkiRJCWCxV5IkSZIkSZISwGKvJEmSJEmSJCWAxV5JkiRJkiRJSgCLvZIkSZIkSZKUABZ7JUmSJEmSJCkBLPZKkiRJkiRJUgJY7JUkSZIkSZKkBLDYK0mSJEmSJEkJYLFXkiRJkiRJkhLAYq8kSZIkSZIkJYDFXkmSJEmSJElKAIu9kiRJkiRJkpQAFnslSZIkSZIkKQEs9kqSJEmSJElSAljslSRJkiRJkqQESOVyubhjkCRJkiRJO+VSqVTcMUhVI5fLYU5IO+VyuflAc7H3HNkrSZIkSZIkSQkwOe4AJEmSJElSX7/5dH3cIcTqqB93ADD3vZfHHEl8Fr5yAwAXHjhxvwOAm9eF7+HNrnNijiRe+079HgDbdtwWcyTxmVzzBQA2P3ZUzJHEa4+/+s2A7zuyV5IkSZIkSZISwGKvJEmSJEmSJCWAxV5VhUwmMzeTyRwTdxxStevM1s7tzNaaK5IkSZIkaRcWe1Ut5gIWsKTBmSuSJEmSJKkoi72SJEmSJEmSlAAWeyVJkiRJkiQpASz2SpIkSZIkSVICWOyVJEmSJEmSpASYHHcAGj8ymcwtwEZgOXANcDjwfsJ/GswHTgVqgVXALcB1LS0t2/pd4/OEB0z9aXSt+4F/KNLWOcD/Bo4AXo6u1ToKH0uqqM5s7S0UyZO6dM/LndnaKZTIlbp0z7Z+1ykrVzqztbvkSl26x1yRJEmSJGkCstiroToemAnMAbqBbcBzwErgbOBVYDrQAnw8k8l8rqWlZQdAJpP5B0Kx6lrgS8C+wCXAMmCffAOZTOYvgFuBy4GlwCnANzOZzE9aWlpWjP5HlEasT55Ehd73AE9QIlc6s7Wfq0v37ADozNaWlSud2dqiudKZrf1JXbrHXFFVW876DwP/RMiXFPA08E/TOWBppdpYxKqibcxhWsk2FrFqBvBIibc/M4dp91cqvlks/xugifCfNeuAnwDNdzJ9daXaGIqjWfJ/gauAy55j5pj8p9Hu3D2DAb7vLZxZse+7lNSKhUcBC4CPA5OAXwDX5I6c+8xot12ou7u76M/rlClTllaqjWw2W7SNdDpdso1sNjuDAf6M0ul0xf6M2tvbi+ZEY2Pj6kq1MRRtbW29OTF79uwxyYkFCxbMYIDve968eaOeE01NTUVzorW1dUxzQpIkDY/FXg3VUcCftLS0rAbIZDK3AauBE/NFXeC3mUzmceCXhKLW9zKZzJGEwtUZLS0thf9IzWYymWuB5oJ9fwmsbmlp+ddoe2Umk7mzpaXllVH6TNKgOrO1f0coqv4R8DhwE/AAcCCwELiiLt2zKjr8KOBP6tI9qwsu8S2iXMkXdYHfdmZr++RKZ7a2N1fq0j19cqUzW1s0V+rSPb250pmtvbMu3WOuqKotZ/0HgceAR4G/BjYBs4CHlrP+5Okc8NORtrGIVSXbWMSqk+cwbbA2PgS83W/fmpHGlTeL5VcBVxNG5i8F9icUlX4yi+XH3sn01yvVVjmOZkkdcD6wYSzbLTCq33cpqRUL3w90EH5OTge2EoqNHakVC9O5I+euHO0YALq7u0v+vHZ3d588ZcqUEedENpst2UY2mz05nU7HmhPt7e0lc6K9vf3YxsbGMc2Jtra2CZkTTU1NJXOiqakp3draOiY5IUmShs9ir4bqFwWF3j2As4AvFhR6AWhpaVmZyWTuB84Evkf4ZeKFfoXevK8T/nGftxz450wmsyg6N2uhV3HqzNZ+EPgioSD7WvT6TeAOIAf8C/C7glN+UVjo7czW9uZKQaEXgLp0z8rObO0uudKv0JtXNFc6s7W9uWKhV+PE/yLcHXLmdA7IT2HyzHLWHwLMA0Zc2CpsYw7TettYxKpy2+iaw7SNFYijlG8DP7yT6Z35HbNYfhbQA5wG3D6KbRdzPeHvmL8b43bzurZw5mh+36WcB7wDnJ07cu42gNSKhRcAxwF/A1wxRnH0/rxOmTKl9+e1u7t7VHIinU73tpHNZsvOiXQ6Peo50djY2JsT7e3tEzon5s2bF2tOtLa2bgNoamqKIyckSdIw+YA2DdW6gvUDgN2B35c49gXgj6P195Y6rqWlZQthxGN++xfAGcDRhNvYNmQymeszmcxuI4pcGqa6dM9K4Li6dE+2Lt2zui7d8w1gKuE20wPr0j2X9Svirut3iYrkSl26p0+u1KV7dsmVzmzt9Z3ZWnNF1e7LwIyCQm/es4TR8xVro6DQOxptDNudTH+7sNAbqSHcMj2mBZ6jWfLXwAeARWPZbpXYAryVL/QC5I6cmwNej94bK18GZhQUevMqnhMFhd7RaGPYGhsb3y4s9EZiyYm2trYJnxP5Qi9Aa2trHDkhSZKGyZG9Gon1hH/0HQpki7x/OJAfZfgKYd6vXURF3EMK97W0tNwL3JvJZPYhjKL8FvDf0as05oqMyN1OeMBaOSqSK1ERt0+u1KV77gXu7czWmisaN6ZzwFvAW0Xe+kvg15VoYw7TRtrG44tYdThhVOFdwP8ZrZG+s1ieIhSWvkaYQ/WB0WinmKNZsidwHTDnOWZuO5olY9V0f4/vzt19vu8xGun7XSCTWrHwH4H/H9gOzCZMx3PeGLQPwJQpU0Y9J9Lp9IhzIpvN9vkzGq2Rvu3t7bHlRFtbW29OzJ49e1tbW9tYNd3f4wsWLOjzfY/RSN/vApmmpqZYc0KSJA2fI3s1bC0tLZuBu4EvZzKZVOF7mUzmCODU6H2A7wN1mUzm00UudRWwZ8G5e2cymWlRGxtbWlq+Q7i18M8q/ymk0VeX7unNlc5sbZ9c6czWFs2VzmztoLnSma3duzNbOy1qY2Ndusdc0bi1nPVnEG7VHrWRdItYVU4b7wBthDm6/5wwh/Z5wD2jEdMsln+GMCfmC8BBwMw7md5/5OVoygArn2Pm4jFss9CYft/95Y6cu4bwQM1LCCMX34xiOC135NxSd2OMie7u7lHPiWw2W3U50d7evktONDY2jnlOzJ49e0LmRGtra8mcaG1tjTUnJElSeRzZq5H6CvAM8GAmk/km8DLwUcIcZw8CdwK0tLT8OpPJ/CPwg+iBbD8B9gM+D5xImAc1bz7wpUwmczGwApgOfAqYMyafSBodvbnSma0tmSt16Z5fd2Zr/xH4QfRAtkFzpTNba65oXFvO+hnAbcC3pnPAo6PRxiJW9bYxh2kl25jDtOWE+bDznl/EqqeA7CJWnTmHaXeXOHW4HiEUdI4GrgR+NIvln76T6aN+u/TRLJlKmH/zL0a7rVK2cOYu3/fu3P0UkN2du8/cwpmV/r77SK1Y+AHgx8DDhAdvbgf+J3BfasXC48fqAW39dXd3zyD6eZ0yZcqo5EQ2m+1tI51Ol2wjnU7v8meUzWafCpfInplOp0c9J9rb2z/d2Ng46jnR1tYWe07Mmzdvl+97wYIFTwHZBQsWnDlv3rxRzYmmpqaSOdHU1HS8D2iTJKn6ObJXI9LS0vIq8GGgC/h34CnC6MN/BD5b+OC2lpaW+cAXgHOAXwDtwG7AXxFGDeT9H+DfgBuA54C/B/6+paXl1tH+PNJoqUv3DJgrhdNE1KV7RpQrdekec0XjxnLWHw/cRxjdfuVotLGIVSNqYw7TniI8nf6MCofGnUzfeCfTn7mT6bcAMwh/T3yx0u2U8C3g5ueY+dsxaq8sWzhz1L7vIr4O/Cp35NxzckfO/WnuyLkduSPnfhFYBvzzGLS/i+7u7lHPiWw2O6I20un0qP0ZNTY2bmxsbHymsbHxFmLKidmzZ1dVTsybN2/Mc6K1tfWc1tbWn7a2tna0trbGmhOSJGloHNmrsrW0tJxfYv964KIyr3EPxW9Dm1ZwzGaiotXQo5TiVZfuOX+A98rOlbp0z6C5Ek0PYa5o3FrO+j8njCC7Fzh/OgfsGOSUIVvEqj5tzGHacNv4LWF+7VFzJ9Nfm8Xy+wkj9P9tNNs6miUfJxTSPng0S/r/e7Am2rf9OWbmRjOOAYz69x05keLFzu8B3xmD9vvo7u7u8/M6ZcqUiudENpvt00Y6na7anGhsbHytvb19THKira2tNyfa2tp2yYlo3/bZs2ebE5Ikqao5sleSJI255aw/hjCFyX3Al6ZzwPZKt7GIVX3amMO0kbRxCGH6lYqYxfLdZrF8UpG3JgP7VKqdAUwDDgDWEuZHzS+HANdH6yeOQRylVPT7HsC7gGLFux3AXqkVC8dsYER3d3efn9cpU6ZUPCey2WyfNtLpdNXkRHt7+27t7e3mRGlVkRNNTU0OFpIkqcrZWUuSpDG1nPUfBh4ijC4crUJvnzbKKfQuYtVehNva2+cw7eaC/R8ETgLOqkRss1j+XuB5YC5hztT8/n0JIxjbKtHOIO4jzIva339EMX2fMJJw1OzO3b3f9xbOvLlgf0W/70H8gnBr/C399jcCj+eOnDsmDwbr7u7u8/M6SoXePm2UU+jNZrO9f0bpdPrmgv0V/TNqb28vmhPt7e0TKicWLFjQ+33Pmzfv5oL9VZMTra2tY/mwPEmSNAwWeyVJ0phZzvp3Ex788zzhCe/vX876PsdM54DVI2ljEat2aWMRq/ocM4dpqxex6jTgLuDsOUy7fw7TNi1i1RqgdRGr9iTMkXk00AL8J6HoM2J3Mv2VWSz/D6B1Fsv3IBRX/pjw0MWNhHlDR9VzzFwP/b544GiWbAG6nmPmk6MdwxbO3LQ7d68BWnfn7lH7vgdxGfBEasXCmwi3qG8HLgD+GvjYGLRPd3f3Lj+v3d3dfY6ZMmXK6pG0kc1md2kjm832OSadTq/OZrO9OZFOp+9Pp9ObstnsGqA1m82O2p9RY2PjK+3t7f8BtLa3t8eSE7Nnzy6aE21tbVuArtmzZ496TsybN2/TggUL1gCtCxYsiDUnmpqaYssJSZI0MhZ7JUnSWHo3cFC0rCpxTGqM2sgRChmFtyxfBPwX8HfAdcCLhFGF35jDtErO1XkRsBK4AjgUWAMsBubdyfS3K9hOtSv5fW/hzFGfGzV35NzfpFYs/CjwDcKo1xTwKJDOHTn3l6PdfmTc5kQ6nR6TnGhsbDQn4Bvz5s0b9ZxobW39TVNTU9GcaG1tHauckCRJI5DK5eJ6xoAkSZIkSSoit+LUGXHHEKujftwBwNz3Xh5zJPFZ+MoNAFx44MT9DgBuXncDuVyOjS9+Pu5QYrXv1O8BsG3HbYMcmVyTa74AwObHjoo5knjt8Ve/IZfLzQeai73vyN6Ray6xPt40l1gfT5pLrI8nzSXWx5PmEuvjSXOJ9fGmucT6eNJcYl3J0FxifSLFEHf7xmAM1RZD3O0bQ3XFIEmShsiRvSNX+AWO9Ba7OCXhc/gZqoOfoXok4XMk4TOotGr48407hrjbNwZjqLYY4m7fGKorhonMkb2O7HVkb8SRvYEjex3ZmzfYyN6asQ1HkiRJkiRJkjQaLPZKkiRJkiRJUgJY7JUkSZIkaWhOAR4HOoAHgPfFG44kSYHFXkmSJEnSRFGJyU/3A9qAzwL1wCLgugpcV5KkEbPYK0mSJEmaKCpR7H0bmA68Gm0/CtRV4LqSJI2YxV5JkiRJUpIcDvwU+BfCNAtPAB8GlhKmW1gKZIATge8BtwGPEYq2BxVcp4kwVcNjwA3AXcChwDZ2FnoBzgF+Hq0fAdwLLAGeBD5TIsZmIDfAIknSsEyOO4CxcNddd+WefPLJsWgqKZ1yEj6Hn2EAF5/1b6N16f78c6geSfgco/YZ6tI9FwC3jNb1k+oFNubWsrlSlxvWn+9T9FSq/WHH8BjrYm0fYAVvVEEMb8Yaw9YVL1aq/WHH8OK+jbHH8PLLL8ceQ1dXV6ztA6xduzb2GNati/fvhjVr1lSq/WHH0NraGlf/uh04HriSUHC9ALgCmAGsil6JXmcCxwGrgfnApdHr6YR5eeuBd6JjHwSu6tfWJ4Cro/YAXgEuAtYSpnv4GXBf5T6aJEkDmxAje2tra+MOQZJU3Q6NO4DxaI+J8c8ISdLwHRpj288SCr0QRuweUuK4hwmF3v7HnQ1cSyj0QhgN/Ei/c48GbgXOIBR5AbYClxFG9t6D/8aQJI0xf0uTJEmSJCVN4e0Gmyn9u2+p41IMPKL5MOAHhCkcni3Yfx2wjjBi+ETg9RLnN0dtlFokSRoWi72SJEmSpIliG7B3GcfdBVwD7BFt1wMnROsHEeblvRhY3u+8g4Gno/UZwHtGEKskSUM2IebsnTlzJud87uq4wxi2DW/n/60wKdY4RmY7ALtNPjjmOIZv67bwDIZUaq+YIxmZXG4TLS050ntfEHcow/bEplsAOGGvS+INZAQe2XQjML5/nnK5TQCcus+cmCMZmQc2LiKXhBmNYzCFvTlpn/Nia/+3b/2QXC7H7ru9N7YYtm57lVwux1/v2xRbDPdtXEQul2PvPQ6LLYZNW35PLpfjrVWnxdL+PnUPkMvl+OA+Z8bSPoSfxym5ueyz55/EFsNbm/+LXC7HuftfHlsMd/R8m1wuR+sRl8YWw2WdN5LL5dj8f/eLLYY9r36TXC7HlVPi+bNoeSn8OWxdekQs7QPsfkInra2tsbU/gNuBZYSHsj09wHH3EAq3jxCmcniJUODNAbOAAzUccoMAACAASURBVAmjc/M2AqcBXwNuBF4DfgU8TygYV2ySe0mSBjIhir2SJEmSpAljNXBSwXY3Ox/KtiBa8pYWrC+Llry2aMm7H9gAXB8txTwBHDvEeCVJqhincZAkSZIkqa9TgQsLto8h3Gr5ZvHDJUmqDo7slSRJkiSpryywEJgNbCE8dG38ziEmSZowLPZKkiRJktTXWuDcuIOQJGmonMZBkiRJkiRJkhLAYq8kSZIkSZIkJYDFXkmSJEmSJElKAIu9kiRJkiRJkpQAFnslSUq+BuANYErcgUiSJEmSRo/FXkmSqtPlFbrOBcA84HcVup4kSZIkqUpZ7JUkqTpVqtj7AnAy8IdhnNsM5AZYJEmSJElVZHLcAUiSNIEdDtwErAQ+BOxNGInbCrwPWAo8ADwNXARsBw4hFFrPAtZE12kCzgO2AM8ABwFXAauBR0u0fQTQAuwF7A/MB+6r3EeTJEmSJI01i72SJMVnO3A8cCXwJKHQewUwA1gVvRK9zgSOIxRw5wOXRq+nA6cA9cA70bEPEoq9A3mFUEBeC+wH/AyLvZIkSZI0rjmNgyRJ8XqWUOiFMAr3kBLHPUwo9PY/7mzgWkKhF8Jo4EfKaHcrcBmwBLgHOLTIMc1AaoBFkiRJklRFHNkrSVK83ixY30zp/4gtdVyK4c2fex2wgjBiGKB7GNeQJEmSJFURR/ZKklSdthHm8B3MXcA1wB7Rdj1wQhnnHUyYCxjC1A/vGWJ8kiRJkqQqU00jexuA7wNH4egiSZJuB5YBt7GzKFvMPYTC7SOEqRxeAu5l8NG+XwNuBF4DfgU8TygYbx5R1JIkSZKk2Iy02Hs5cEMF4riA8JCY31XgWpIkjRergZMKtrvZ+VC2BdGSt7RgfVm05LVFS979wIZ+bc3ot/0EcOwQYpUkSZIkVbmRTuNweUWigBeAk4E/DPP8ZsIIplKLJElJdSpwYcH2McAk+s7xK0mSJEmaAMoZ2Xs4cBOwEvgQYf7AC4BW4H2EkUYPEG4xvQjYTnhCeA44C1gTXacJOA/YAjwDHARcRRjV9GiJto8AWoC9gP2B+cB9ZX86SZKSLwssBGYT+th1wCWxRiRJkiRJikU5xd7twPHAlcCThELvFYTbQVex87bQGYQneh9HKODOBy6NXk8HTiE8NOad6NgHCcXegbxCKCCvBfYDfobFXkmSCq0Fzo07CEmSJElS/MqdxuFZQqEXwijcQ0oc9zCh0Nv/uLOBawmFXgijgR8po92twGXAEsIDaA4tcVwzkBpgkSRJkiRJkqREK/cBbYXz/m2mdJG41HEphjd37nXACsKIYQgPrpEkSZIkSZIk9TPSB7RtI8zhO5i7gGuAPaLteuCEMs47mDAXMISpH94zxPgkSZIkSZIkaUIod2RvKbcDy4Db2FmULeYeQuH2EcJUDi8B9zL4aN+vATcCrwG/Ap4nFIw3jyhqSZIkSZIkSUqYcoq9q4GTCra72flQtgXRkre0YH1ZtOS1RUve/cCGfm3N6Lf9BHBsGTFKkiRJkiRJ0oSWyuWGM5XukJ0KHATcHG0fA3wdOGUsGgdyB7zruDFqqvI2vJ0fND0p1jhGZjsAu00+OOY4hm/rtlcBSKX2ijmSkcnlNpHL5UjvfUHcoQzbE5tuAeCEvS6JN5AReGTTjcD4/nnK5TYBcOo+c2KOZGQe2LiIXC43n/CwTw1NLpWK9zmouVwOYzCGamjfGIyh2mKIu/0oBvvX4Yu9j5WqSTX8nSZVk4H62JFO41CuLLAQmA1sAdYB47dKI0mSJEmSJElVZqyKvWuBc8eoLUmSNEZO37cptrZ/9GYrAG/Pr40thr2v7QHg76dcFlsMX+v+NgDbdtwWWwyTa74AwJaOD8bS/u71KwG46yNfjKV9gLN/+e8ATKrZP7YYtu8IM6T9f3v9aWwxvL7peQBSqXfFFkMu9xYAe+z2/thi2Lz1RQCO3eu8WNp/etN3Afjjd82IpX2Al99aGlvbyTGe7+yshHB3aFx5VA3yubz/3hN7dsv83c5x/puvGuT/3fnUJ0+OOZL4HPfTBwH40v6XxxxJvG7dcMOA79eMURySJEmSJEmSpFFksVeSJEmSJEmSEsBiryRJkiRJkiQlgMVeSZIkSZIkSUoAi72SJEmSJEmSlAAWeyVJkiRJkiQpASz2SpIkSZIkSVICWOyVJEmSJEmSpASw2CtJkiRJkiRJCWCxV5IkSZIkSZISwGKvJEmSJEmSJCWAxV5JkpLtFOBxoAN4AHhfvOFIkiRJkkaLxV5JkqrT5RW4xn5AG/BZoB5YBFxXgetKkiRJkqqQxV5JkqpTJYq9bwPTgVej7UeBugpcV5IkSZJUhSbHHcBY2fD203GHUAHb4w5gxLZue3Xwg6pcLrcp7hAq4olNt8Qdwog9sunGuEMYsST8PD2wcVHcIYxnhwM3ASuBDwF7AxcArYTpFpYSpl54GriI0BEcAuSAs4A10XWagPOALcAzwEHAVcBqdhZ6Ac4Bfh6tHwG0AHsB+wPzgfv6xdcMXDviTylJkiRJGhMTptgrSVIV2g4cD1wJPEko9F4BzABWRa9ErzOB4wgF3PnApdHr6YR5eeuBd6JjHyQUewt9Arg6ag/gFUIBeS1huoefsWuxV5IkSZI0jkyYYu9+ex4VdwjD9sY7vwGgpubdMUcyfDt2/AGAA/b+aMyRDN/6t58E4JB9To45kpH5740PArDq9I/FHMnwTfvRLwB46C8aY45k+D71eDsA23bcFnMkwze55gsALJx2acyRjMzcVf8adwjPEgq9EKZZOK/EcQ8TCr35486N1s8mjL59J9peCjzS79yjgVuBzxGKvABbgQxhmodJwKHDC1+SJEmSVC2cs1eSpHi9WbC+mdJ9c6njUoRpHUo5DPgBYQqHZwv2XwesI4wYPhF4vci5zdH1Sy2SJEmSpCpisVeSpOq0jTCH72DuAq4B9oi264ETovWDgHuBi4Hl/c47mDAXMISpH94zglglSZIkSVVgwkzjIEnSOHM7sAy4jZ1F2WLuIRRuHyFM5fASocCbA2YBBxJG6OZtBE4DvgbcCLwG/Ap4nlAw3lzBzyBJkiRJGkMWeyVJis9q4KSC7W52PpRtQbTkLS1YXxYteW3Rknc/sAG4PlqKeQI4dojxSpIkSZKqmNM4SJI0vp0KXFiwfQzhgWtvFj9ckiRJkpRUjuyVJGl8ywILgdnAFsJD1y6JNSJJkiRJUiws9kqSNL6tBc6NOwhJkiRJUvycxkGSJEmSJEmSEsBiryRJkiRJkiQlgMVeSZIkSZIkSUoAi72SJEmSJI3cDOAjcQchSZrYLPZKkiRJklS+U4DHgQ7gAeB90f4ZWOyVJMXMYq8kSZIkaSK4vALX2A9oAz4L1AOLgOsqcF1JkipictwBSJKk8etHb7bGHQJ7X9sTdwh8rfvbcYfA5JovxB0Cu9evjLX9s3/577G2D7B9x4a4Q+D1Tc/HHQK53Ftxh8DmrS/GHQJPb/purO2//NbSWNuvQpcDN4zwGm8D04FXo+1HgX8axnWagWtHGIskSbuw2CtJkiRJSorDgZuAlcCHgL2BC4BWwnQLSwlTLzwNXARsBw4BcsBZwJroOk3AecAW4BngIOAqYDU7C70A5wA/L9ieBvwnsC+wFbgGeKySH1CSpIFY7JUkScNWU/Pu2NreseMPALx7r6Nji+EPm54DYK/dPxBbDJu2/B6A/fc+NrYYNrz9NBDf91BN3wFMii2GULOqjp/H3SYfHFsMW7eFOtz79/lUbDG8uPEhAFKpd8XSfn5k9YHvSsfSPsC6t7JxNb0dOB64EniSUOi9gjCf7qroleh1JnAcoYA7H7g0ej2dMC9vPfBOdOyDhGJvoU8AV0ft5c0CPgm8DLwXWAx8GnipIp9OkqRBOGevJEmSJClJniUUeiFMs3BIieMeJhR6+x93NmGKhXei7aXAI/3OPRq4FTgDeKVg/7cJhV6i/XcQCsf9NQOpARZJkobFYq8kSZIkKUneLFjfTOnfe0sdlyJM61DKYcAPCFM4PDtILINdS5KkirLYK0mSJEmaCLYR5vAdzF2EuXb3iLbrgROi9YOAe4GLgeVFzr2MMH1D/thzCFM5SJI0Jqplzt5TCLexvANsJHSczmkkSZIkSaqU24FlwG2EB7SVcg9wMGHqhncIv5veSxihOws4kPD7a95G4LRo/U7gO4QHtG0nFH/93VaSNGZGWuy9HLhhhNfYD2gDphOeavpp4Drgf4zwupIkSZKkiWU1cFLBdjc7H8q2IFrylhasL4uWvLZoybsf2ABcHy3FNA8xVkmSKm6k0zhcXoEY3mZnoRfCxPh1FbiuJEmSJElDdSpwYcH2McAk+s7xK0lSVSpnZO/hwE3ASuBDhDmOLgBagfcR/jf0AcJtMBcRblU5hHCLy1nAmug6TcB5wBbgGcL8RVcR/uc1X+iFMKfRz6P1I4AWYC9gf2A+cF+RGJsJT0uVJEmSJGkkssBCYDbh99d1wCWxRiRJUpnKKfZuB44HrgSeJBR6ryDcCrOKnbfEzABmAscRCrjzgUuj19MJ8/LWE+Y8mgE8SCj2FvoEcHXUHsArhALyWsJ0Dz+jeLFXkiRJkqRKWAucG3cQkiQNR7nTODxLKPRCmGbhkBLHPUwo9PY/7mzCyNt3ou2lhMnuCx0N3AqcQSjyAmwlTGi/hDBJ/qFlxitJkiRJkiRJE0q5xd7CuYk2D3BeqeNShGkdSjkM+AFhCodnC/ZfR7hlZiZwIvB6ifObozZKLZIkSZIkSZKUaCN9QNs2why+g7kLuAbYI9quB06I1g8C7gUuBpb3O+9gwlzAEKZ+eM8IYpUkScEM4CNxByFJkiRJqqxy5uwdyO3AMuA2dhZli7mHULh9hDCVw0uEAm8OmAUcSBidm7cROA34GnAj8BrwK+B5QsF48wjjliRpIjiF0L++Q+hbLyb0wTMI8+7/Mq7AJEmSJEmVV06xdzVwUsF2NzsfyrYgWvKWFqwvi5a8tmjJux/YAFwfLcU8ARxbRoySJCXJ5cANI7zGfoR+dzrwKvBpwvRI/2OE15UkSZIkVamRTuNQrlOBCwu2jwEm0XeOX0mSFFxegWu8zc5CL4QHp9ZV4LqSJEmSpCo10mkcypUFFgKzgS2Eh65dMkZtS5JUjQ4HbgJWAh8izIF/AdAKvI9wt8wDhGmSLgK2A4cQpkA6C1gTXacJOI/Qvz5DmAv/KsKdOflCL4SHoP68YHsa8J/AvsBWwtz6j/WLsRm4dkSfUpIkSZI0Zsaq2LsWOHeM2pIkaTzYDhwPXAk8SSj0XsHO+XRnRMfNAGYCxxEKuPOBS6PX0wnz8tYT5uWdATxIKPYW+gRwddRe3izgk8DLwHuBxYSpHl6qyKeTJEmSJI25sZrGQZIk7epZQqEXwjQLh5Q47mFCobf/cWcTRt6+E20vJTwMtdDRwK3AGcArBfu/TSj0Eu2/g1A4liRJkiSNUxZ7JUmKT+Hc9Zsp3S+XOi5FmNahlMOAHxCmcHh2kFiKXas52l9qkSRJkiRVEYu9kiRVn22EOXwHcxdhrt09ou164IRo/SDgXuBiYHmRcy8jTN+QP/YcwlQOkiRJkqRxaqzm7JUkSeW7HVgG3EZ4QFsp9wAHE6ZueIcw3+69hBG6s4ADCaNz8zYCp0XrdwLfITygbTuh+Ot8vZIkSZI0jlnslSQpHquBkwq2u9n5ULYF0ZK3tGB9WbTktUVL3v3ABuD6aCmmeYixSpIkSZLGAadxkCRp/DoVuLBg+xhgEn3n+JUkSZIkTRCO7JUkafzKAguB2cAWYB1wSawRSZIkSZJiY7FXkqTxay1wbtxBSJIkSZKqg9M4SJIkSZIkSVICWOyVJEmSJEmSpASw2CtJkiRJkiRJCZDK5XJxxzDqlixZknvooYfiDkOqGhef9W9xhyBVlbp0z3ygOe44xptu3s51synWGJ6iJ9b2AR5jXdwhsII34g6BFbwZa/tbV7wYa/sAL+7bGHcIvPzyy3GHQFdXV9whsHbt2rhDYN26eP9uWLNmTaztA7S2ttq/DtPrt63OvX7bf8cdhlQ1Nn/19LhDkKrKQL/DOrJXkiRJkiRJkhLAYq8kSZIkSZIkJYDFXkmSJEmSJElKgAkxZy9wPnBLzDFIkqrX+dhPDMf5+L1Jkko7H/uJ4TofvztJUmnnU6KfmCjFXkmSJEmSJElKNKdxkCRJkiRJkqQEsNgrSZIkSZIkSQlgsVeSJEmSJEmSEsBiryRJkiRJkiQlgMVeSZIkSZIkSUoAi72SJEmSJEmSlAAWeyVJkiRJkiQpASz2SpIkSZIkSVICWOyVJEmSJEmSpASw2CtJkiRJkiRJCWCxV5IkSZIkSZISwGKvJEmSJEmSJCWAxV5JkiRJkiRJSgCLvZIkSZIkqVJOA54BNgBvAG9Gr5KkMZDK5XJxxyBJkiRJkpJhFXAG8DxgwUGSxtjkuAPQgN4NfBE4lL5/VpeXON6OVNJoSMUdgBQz+1dJo8H+VUn1IvAryus/7WMljYYJ3cda7K1uPwaWE/5HdEfMsUiSJEmSNJgrCb/LdgCbC/Z/K55wJGlisdhb3fYEvhJ3EJIkJdBQ756RJEnl+SdgI+H32d1jjkWSJhyLvdXtu8DfAPfT939ENwx2Yk3NHkANKWogVdO7nsqvpyb1bve+piYBFOyfRCpVQ03B8TX5/dRQk9/HzvNqCO/X9F/P1fS+n9+uiUbVp6Ij8/tSIYKd66n8Nr3bEJ4umEql+rzWpPLHhDH7Nb3vh/dCe2G9cP9g+8jv730/t+s55HrPza/vjDMXxbfzvPy+Xd/LRe3tuj/syxXEkAvtRev0OS66Rp/zCraj113Oy+8vYz0fZ6omR01qR+/7O/flSOX31+x8D4i2d/TZ338f/bej86jZec3B1qnZ9Tzy6zU5yLdTk98fHZu/6aMmv6SiVErtXFI10XpNtETrIQmhpoZcTQ3UTCo4poZcn+1J5FLRMRC28/trJg3wOjlan7zzvNTkaP/ksL9gO5WaDDW7hdfUZFLRealUeK8mNTnkeWoyNTWTmMS5g/01I413w7p7plT/CgzaxxbrX8N5A/exxfpXYNA+tlj/Sm/EpfvYYv0rDN7HFutLQ5xl9Lu97+V2/vU7SB9brA/tjXOAPnbgfrd0H1usf+17XvE+tlj/SkEsA/W3/fvXfHsD9bHF+lfYtd/t38cW60cL2yvVxxbrc8vqY/v3r71/6IP0sf3715CYg/ex/ftXKK+P7de/AmX1sf371/BdTma3mi8O8reMlAj7AzOHelK5v8MCZfWx/ftXoKw+tn//CpTVx/bvX9n5aQbsY/v3r1BeH9t/H5Q4tvec4r/DUnDNgfrY/v1r+HyD97H9+9fw5zB4H9u/f6Xfe6X62P7r+fMG62P796+95w3Sx+7Sv0JZfewufSq7/s5a1u+wUXsD9rHF+teQdAP3scX6197zBuhji7wCg/axRftXGLSP9XfYXVnsrW5bgBbgH9g5l1EOOCy2iCRJSgbvnpEkaXQ8TCj2Lok7EEmaiCz2VrevANOAdXEHIklSwgz77hlJkjSgOYR5ezcDWwljDXPAfnEGJUkThcXe6vZr4O24g5AkKYG8e0aSpNGxb5F9qSL7JEmjwGJvddsOPAs8Qt9RRz48RpKkkfHuGUmSRsdXgWsKtmsId9Q4oaYkjQGLvdXtR9EiSZIqy7tnJEkaHVOBq4GvA3sAPwCejjUiSZpALPZWt1uB3YEPEm4t/S3htlNJkjQy3j0jSdLouAC4nVDwPQH4T+C6WCOSpAnEYm91+zRwI/ACYY6jDwCXEDpLSZI0fMXunskVO1CSJJXl2IL16wm/y/4M6Ijec3SvJI0Bi73V7VuE/wldFW0fDjyAxV5Jkkbq3YRfRAt9OY5AJElKiG/22+4Bjor254BPjnlEkjQBWeytbq+xs9AL8LtonyRJGpkvsWux9/wi+yRJUnlOiDsASZLF3mr3a+DHwPcJ/xN6FvAEcEb0/g9Lnbhjx+ZSb0mSNJGdA3weOAz4j4L9+wLrBjvZ/lWSpEEdAFwLfJzwe+wy4KvA+oFOso+VpMqw2Fvd9gTWAPXR9lqgFvgModMsWeyVJElF/Rx4BTiIvrebbiQUgSVJ0sjcCTwKnBltnwvcBZwUW0SSNIGkcjmfRVLFbiXMH/iHaLuW8IvphSWO9w9T0mhIxR2ANAqepu+DZAB+CXykyLH2r5JGg/2rkuop4Lh++54EPlrkWPtYSaNhQvexjuytbh9hZ6EXwgT3fzbA8eP1h7mwgx+vnwGS8zkgOZ/FzyGpv9nA3xKmcfhlwf59CaN+i6nmvKv2vx+Mb+SqPUbjk/5fe3ceI0lZxnH8O+xyCXIsAYIgWVbBsAREhQXDIaIci8qKooIQDknAqAkkHIIQBY0HoEIMyB8qQfDAsF4EUSCImBgURPZgF4EFV1ghbABFWBPkePzjqc50xumdafp4q3q/n6TS09XdM7+3puqtt96qelsT3QEcQw5HCHA0+UXjkxnEdlm37b5OeeqUBcwzFfOsXd3y1IZX9tbbYuAgspMXYBZwJ7B7qUADMiob6KiUA0anLJZD0kSbk3fKfBU4t23+88CzRRL1pu71g/l6V/eM5pPUbgx4pfq59TgDWFP9HMBmA85Qt+2+TnnqlAXMMxXzrF3d8tSGV/bW2zfIq4wWkivxR4EvF00kSVKzPVdNx5YOIknSCApgEf8/VJIkaUjs7K23a8mxjQ4mz1J8CFheNJEkSZIkSZ3dBewN3FM6iCStixzGQXUwKpfej0o5YHTKYjkkjbq61w/m613dM5pP0kTLgbcAK8nhG8bIbXGyL0EdhLpt93XKU6csYJ6pmGft6panNryyV5IkSZIk9cv80gEkaV1mZ6/q4KLSAfpkVMoBo1MWyyFp1NW9fjBf7+qe0XySJvp74b9ft+2+TnnqlAXMMxXzrF3d8tSGwzhIkiRJkiRJ0ghYr3QASZIkSZIkSVLv7OyVJEmSJEmSpBFgZ68kSZIkSZIkjQA7e1XC1cBq4P62ebOA24CHq8ctC+Tq1huBO4AHgGXA6dX8ppVlI+BuYDFZjtYg5zsBfyLL8RNggyLpujcDuA+4qXre1HKsBJYCi4A/V/Oatm5J6r+V1K9u6Ga/PgZ8C1gBLAHeXijfhcA/yOW4CDii7bXzqnwPAocNIV+37YlhL8NO+S6kHsuw23bMhtXzFdXrswecT1J/HU7WLSuAcyd5vdM2fghwL7kPvRc4uHCelh2BF4CzapBnD+Ausi5dStavpfKsD3y/yvEAuV/ph6nyHAj8BXgZOHrCayeS+5SHq59L5tmT8f/VEuBjhfO0bEa2Da6oQZ4dgVvJ9Wc569j+3s5elXANudG2Oxe4Hdi5epxsQ66bl4EzgV2BfYFPA3NpXlleJBs7byV3GoeT5bkYuIwsxz+BU0oF7NLpZIXe0tRyALyb/J/sVT1v2rolaTDqVjdcw/T36/OreTsDpwJXFcoHuW/Ys5purubNBY4Bdqs+823yJOIgddueGPYy7JQP6rEMu23HnFI9f3P1+sUDzCapv2YAV5L14FzgWMbro5ZO2/jTwAeA3cmOuusK52m5DPh1H7L0mmcm8APgk2T9fRDwUsE8HyE7gncH3gGcRu+dddPJ8xhwEvCjCfNnAV8A9gHmVT/3enK9lzz/AU5gfF97ObBFwTwtXwLu7DFHv/JcC1xKtl/mkSf+1xl29qqE3wPPTpi3gDxzR/X4waEmem2eJM8iATxPdjBuT/PKEuTZZMgzqOtX8w4GFlbzm1AOgB2A9wHfrZ6P0cxydNK0dUvScJSuG7rZry8gG98B/JE8MNmuQL5OFgDXkx2IfyOvJJk3oFwt3bYnhr0MO+XrZNjLsNt2TPtyXQi8h2wvSKq/eWSd8ijwX7KuWTDhPZ228fuAJ6r5y8irVjcsmAeyXnq0ytMPveQ5lLxCdHH12jPAKwXzBLAJ2Qm9cfX5fw8hz0pyObw6Yf5h5F02z5Kd07cx+YnkYeV5iLzCGHK9Xg1sXTAPZKf8tuTVtP3QS5655LpzW/X8BbKDfJ1hZ6/qYlvyYILqcZuCWV6L2cDbyFtPmliWGeQtmKvJCvER4F/k1TwAq1j7gV1dXA6cw3hlvxXNLAdkA+dW8jazU6t5TVy3JPVXU+qGTpm2Bx5ve1/Jevkz5AHC1YxfnVM632ymbk+UzNieD+qzDLtpx7Tnexl4jmwvSKq/6dQv09nGP0x2/r5YMM8mwGcZH3qmH3rJswvZxriFPMF3TuE8C4E15P7vMeDrTP/EbS95BvHZQf/OeeRQRY8UzLMe8A3g7B4z9CvPLmQ74Gfktn4pg79Tq1bs7JV6tynwU+AMej/bWMor5K2PO5A7i10neU8MNVH33k8e5N3bNm+yK3XqXo6W/chxGOeTt8weWDaOpJpoet1Ql3r5KuBN5L7vSfIABcrmm257olTGifnqtAy7acfUZR2U1L3pbL9TvWc3cqiA0wrnuYgctuCFSV4vkWcmsD9wXPV4FHmVbak888i6/Q3kGOxnAnOGkGcQnx3k79yOHJLkZCa/2nZYeT5FDuf0+FRv7EIveWYCB5BjYe9Nrjsn9SdWM9jZq7p4ivFbELejOeOprE8e+PyQPGsEzS0L5Nmv35Fj3W1BVpKQB09PdPhMXewHHEneynE9eftma+yiJpWjpZVzNfBzssHT5HVLUn80pW7olGkV+YVfLaXq5afIg8hXge8wPsxAqXzdtCdKZOyUr07LEKbXjmnPNxPYnN6vFpM0HNOpX9a2je9A7jtPoPerIHvNsw9wCXnscgbwOfJuiVJ5VpFjrT5N3u5+M71/AWgveT4O/IYcN3g18AfGv6tgkHkG8dlB/c7NgF8BF5DDOvWqlzzvJNffleRV2CcAXyuYZxV5Re+j5BXjv2A4XwpcG3b2qi5uZPwbLU8Eflkws9UPYAAAAoZJREFUy3SNAd8jx677Ztv8ppVla8YHc98YeC9ZpjsY/0bLJpTjPHIHMJv8Ypjfkmemm1YOyNu6Xt/286Hkt8g3bd2S1F9Nqhs6ZbqRPAAYIzvknmN8qIJhah/j9ihyOULmO4Ycx3En8su97h5wlm7bE8Nehp3y1WUZdtuOaV+uR5PtBa/slZrhHrJO2Ym8bf0Ycptu12kb34LsGDuP7DgsnecA8rhlNnmByleAKwrmuQXYA3gd2en6LmB5wTyPkRfvjJFtnn2Bvw4hTye3kO2uLavp0GpeqTwbkCcurgVu6DFHP/IcB+xIrs9nVbl6/cLgXvLcQ/6fWuMYH0zv63OzRIST07CnH0fEkxHxUkSsiohTImKriLg9Ih6uHmfVIOdU0/6RlkTEomo6ooFl2SMi7qvKcX9EfL6aPyci7o6IFRFxQ0RsWIOs050OioibGlyOORGxuJqWRcT51fymrVtOTk79nepaN3SzXx+LiCsj4pGIWBoRexXKd13195dExI0RsV3b+8+v8j0YEfOHkK/b9sSwl2GnfHVZht22Yzaqnq+oXp8zhP+xk5NT/6YjIuKhyDqmtR/8YkQcWf3caRu/ICLWxHg9tigitimYp326MCLOKrx8iIjjI9sX90fEJYXzbFrNXxYRyyPi7CHl2TuyrbAmIp6p/n7rs5+ocq6IiJML5zk+sl3Tvj7vWXj5tKaTIuKKGvy/DolsGyyNiGsiYoM+ZWrENBbhiWxJkiRJkiRJajqHcZAkSZIkSZKkEWBnryRJkiRJkiSNADt7JUmSJEmSJGkE2NkrSZIkSZIkSSPAzl5JkiRJkiRJGgF29kqSJEmSJEnSCLCzV5IkSZIkSZJGwP8AZVvkRAZiwxwAAAAASUVORK5CYII=\n",
      "text/plain": [
       "<Figure size 1728x1152 with 6 Axes>"
      ]
     },
     "metadata": {
      "needs_background": "dark"
     },
     "output_type": "display_data"
    },
    {
     "data": {
      "image/png": "iVBORw0KGgoAAAANSUhEUgAAA7YAAAEACAYAAACH9urZAAAABHNCSVQICAgIfAhkiAAAAAlwSFlzAAALEgAACxIB0t1+/AAAADh0RVh0U29mdHdhcmUAbWF0cGxvdGxpYiB2ZXJzaW9uMy4xLjAsIGh0dHA6Ly9tYXRwbG90bGliLm9yZy+17YcXAAAeu0lEQVR4nO3dfXhkVZ3g8W+lO7xuGAkgYlDQwPoM7grGgRFZpXmRAkQamURRfIHGQYiOupsHja5KdJw1rmSZcaSYWZ0e8IUBE3lp3gzo0LIKiBDwDQEJgmNEkAlIOUCTbmr/ODedSqoqSaeSunWqvp/nqSe3zr331C+n6lTV755zb2UKhQKSJEmSJMWqJe0AJEmSJEmqhomtJEmSJClqJraSJEmSpKiZ2EqSJEmSomZiK0mSJEmKmomtJEkrqGdo8qSeocmT0o5DkrTyuif613RP9J+WdhzNyMRWilS2Nz+Q7c1vTDsOqVn0DE0+1DM0edoSdj0puUnRax3rPK11rPOhtOOQVlr3RP/G7on+gSXsugY4bVmD0aKsTjsAqZ5le/P/FfgEcBiwO/Ar4BLgC6O5tmfTjE3S4vUMTVbsy8N97cval3uGJq8ACsN97ScvZ71qLK1jnWuAm4BHgJdOdY1vrrBdC/AgsA/w5qmu8WtqFqRUx7on+tdQ1IdGOgbL9qHuif5ZfWikY3BZ+1D3RP/5wBtGOgZfs5z1ats5YitVkO3Nvwq4BXgGeBvwauAC4BzgoymGJmkb9AxN1rov3wfcuwL1qjHtCZwwz/os0FGjWKQYpd2HfgX8bAXr1yI5YitV9jng9tFc22lFZb/I9uZvAH6ZTkiSluBzwO3Dfe2nFZX9omdockX68nBfe/9y16mGdh1wJnBlhfVnAlcAPTWLSIpLqn1opGPwiytRr7adI7ZSZfsBd80tHM213T+aayukEI+kpSnbl4f72u8f7mu3LyttOSDbOtb50rkrWsc6X0QYibqw5lFJ8cgB2e6J/pI+1D3Rbx9qIo7YSpU9DJyQ7c0PjubaHl9qJdne/MuATwJvBF4I/AYYGs215ZL1pwIXAe2jubZ8UrYz8DjhXN5PFdX1BeDE0VzbK8o8zg7AVcArgCNGc22/WmrMUoN5GDihZ2hycLivfcl9OZHpGZr8IPBhwrm6dwMfHu5rH5veoGdo8iKAOSPE9AxNvgb4LPB6YAvwA+Azw33tt1UZk+L2/4B7gDOAc+esWwc8MNU1flPrWOc2Vdo61rkH4bPnTcDewGPAPwGfmeoaf751rPMw4PvAAVNd478o2u/XwMaprvF3F5W9nzDzYfcyj5MB1gMnAsdMdY3fuU2BStVbsA+NdAze1D2xtMk03RP97wD+J+Ec3fuAj450DH6naP0AsGakY3DNnP32I/Sbo4HtgDuAwZGOweuXFIgW5IitVNnHgJcSph+fk+3NlxwJXKR1hHP73gUcSHhz/Fy2N/+WZP31QIZwFb1pWWATsHZOXUcDJRc9yPbmtwMuBzqBN5jUSrNs7cs9Q5Pn9AxNLrUvA/QR+moPcHxStqFnaHLHBfY7CLiB8AXsz4EjCAe5bu4ZmnxzFfGoMfwDsK51rHPVdEGSMJ6RrFuKk4HdgLMJnz29hM+jDyXrbyUcQD266DG7gF2AN7WOdRYPfhwN3DjVNf5c8QMkMf5fwojYkSa1StE/AOu6J/q39qHuif5q+xDAXwBnEaYzH0k4n/aqZCR4Ph2E9/uHCO/3ryUczNzQPdF/dhXxaB4mtlIFo7m2O4FXERLPc4GHsr35G7K9+aO2sZ5Pjuba3j+aa9s4mmu7dzTXdilhhPbkZP0k4QvG0UW7rQXOA/bN9ub3Bcj25ncnfDmZldhme/OtwAghqT18NNf26239X6VGNtzXXtKXe4Ymb+gZmtymvpyYAnqG+9rvHO5r/z5wOuELzJ8vsN+BwCnDfe3/a7iv/efDfe1jw33tZwJfAv6xZ2hyuyXEosbxVeAFzBwsgTDLZy/g4qVUONU1/o9TXeOnTnWN3zDVNX7vVNf41cD5JJ89U13jzwPfpvSz5xLC6O5/A0iS7TWUOahKuAjbicARU13jP15KnNIyWfY+lNgNOH6kY/AHIx2DtxPe8wGOXWC//YBPjnQMnjPSMXj3SMfgT0c6Bj9OODg61D3R/8IqYlIFJrbSPEZzbQ+M5treDewBvBXYFfhOtjf/zsXWke3Nt2Z782dne/PXZHvzv8n25p8C3k+YGjbtGpIvF9ne/CrC1LFvAaPMjNoeDTxFmDo2bTVwGeEN9PDRXNvEEv5NqeEN97U/MNzXXtKXe4YmF92Xp6sa7mvfUlTvOPAHZvfncm4Z7mu/sUz5eYQvXq/bxjjUQKa6xvPAN4C/LCr+S+BfprrGn1xKna1jnZnWsc5TW8c6R1rHOh9uHev8A/B5Sj971hSNFK8Frk5u0589BwN/QrhAT3H9XyT8PvOaqa5xrwirVI10DFbsQyMdg0vqQ4lrRjoG/zjncX7Jwu/5vx3pGPxKmfJc8vf4MutUJRNbaRFGc23PjObaRghfPm8CBrZh98uATxFGi04CDiFMi8kUbXMNcEC2N78X4Sj5E6O5tl8AG5id2I6O5tqmivbbgzC95V7g99v4b0lNZ7iv/Znhvval9mWA35Upe5KFr1lxf4V4fgs8C7x4G+NQ48kBx7eOdXa0jnW+kPDeX80Fb4aSOu8ATiHMKjiX2Z893wZ2Ag5JLl7VCfwrpZ89P5rqGn+0aL+dgKMI568/XEWM0nLKAcd3T/R3JCOi1fYhWOb3/OS3dh/G9/wVYWIrbYMkqRxh4SN1AGR78/sBbwFOGM21XTCaa7tjNNd2L7CqeLvRXNvPCedtHEWY1nV1suo64NBsb35Xyp9f+wjhfNw1wFeyvfkMkhY03Ne+TX25yPNLfMj9yxX2DE2+GNiB0JfVxKa6xn8C/JBwTuDpwI+nusbvWEpdrWOdOwB/Bayb6hofnOoav3Wqa/xe5rx+p7rG/0CYBXQUIQm4caprfBPh9JhdWsc6/wvlP3ueBo4hzDa4snWsc/ulxCktp5GOwZI+NNIxuKQ+VGRZ3/OTc4D3wff8FWFiK1WQ7c1ns735A8qsej3hSqiL0Zb8faKo3tWUPzfjWsIXiK2JbXL+7Y8IF/t4CWHUd5bRXNtPCRfueCvgb6lJc/QMTWZ7hiar7cvVel3P0OSRZcr/O2FE4JYaxaH6diHhS/l7qW6kaUfCiNITc8rLXahs+lSY6WnITHWNbyEcWH0HcChlzq+d6hqfIJzDeCBw6ZyLTUlpWa4+VK2O7on+08uUn0WYNXFdmXWqkm9CUhlJ8jkAdGZ7858kXNmujfBmeQLJRTUW4R7Cl9Zctjf/ccKXjY8Qpg3P7X/XEK5s/FzyeNM2AJ8Bbqv0s0OjubZbsr35buCqbG/+j6O5to8tMj6pofUMTW7tyz1Dk9X05Wr9BPhmz9DkecCNhPeCdxCuUvu24b72TTWKQ/VtmHCBp12AS5dayVTX+BOtY513AZ9vHev8EOFz5SzCT47M/ey5mvCTJM8TXpPTNhAuyPP7qa7xkt+BTh7nl61jnccCG4GLWsc6351clEpKy7L0oWXwAPA33RP9ryD8FOMWwsDFOcD/GOkYfHS+nbU0jthKZYzm2jYTpmZ9EfgAcCfhYk47Aa8ZzbUt6uqPo7m2TYQvz7sQLvN+CfC9pN72OZvfRPhicX3y+NOuJnwJLndFyuLHuh54N/CRJImWmt5wX/u8fXm4r71WV3K9m5nTBr5HOFr/cuDg4b72q2oUg+pcMg14PXDxVNf401VW1wNMEg6kXE04oPoR5nz2THWN30/4SZKxqa7xx4pWjRJOm7l2gZjvInxh/wvSHSGTGOkY3NqHRjoGq+1D1ZgA3kA4b/16wne8NcDRIx2DF6QYV0PLFAqFtGOQJEmSJGnJnIosLUG2N78PcN8Cmz08mmt7RS3ikbQ0PUOTi+rLw33t9mXVhdaxzmcX2maqa3yHWsQixah7on/BPjTSMWgfipCJrbQ0vwUOWmCbqQXWS0qffVmxWej1Kml+9qEG5VRkSZIkSVLUvHiUJEmSJClqJraSJEmSpKiZ2EqSJEmSomZiK0mSJEmKmomtJEmSJClqJraSJEmSpKiZ2EqSJEmSomZiK0mSJEmK2uq0A2hAAxWWm9VAheVmNlBhuZkNVFhuVgMVlpvZQIXlZjZQYblZDVRYbmYDFZab2UCF5WY1UGG5mQ1UWG5mAxWW60qmUCikHUOjKW7QTGpR1A/bo5RtUso2mc32KGWblLJNZrM9StkmpWyT2WyPUrZJqSjaxKnIkiRJkqSomdhKkiRJkqJmYitJkiRJipqJrSRJkiQpaia2kiRJkqSomdhKkiRJkqLWMD/3c95XnynceNvmtMOQpIazS+dU2iGozl15xMFphyBFZ+2ePWmHIEVnpGPwdOCicusaZsR2z90a5l+RJEmSJJXat9IKs0FJkiRJUtRMbCVJkiRJUWuYc2yBQiazOu0Y6sgWAA7dcV3KcdSPW59ZnyytSjWO+rEl+Wt7zLBNSm2hUCjg++s0XyOlfI2U8nUym+1RKvSbnXfYL+1A6sbTm8aTJV8ngf2m1BYKhcKngYFyax2xlSRJkiRFzcRWkiRJkhQ1E1tJkiRJUtRMbCVJkiRJUTOxlSRJkiRFzcRWkiRJkhQ1E1tJkiRJUtRMbCVJkiRJUTOxlSRJkiRFzcRWkiRJkhQ1E1tJkiRJUtRMbCVJkiRJUTOxlSRJkiRFzcRWkiRJkhQ1E1tJkiRJUtRMbCVJkiRJUTOxlSRJkiRFzcRWkiRJkhQ1E1tJkiRJUtRMbCVJkiRJUTOxlSRJkiRFzcRWkiRJkhQ1E1tJkiRJUtRMbCVJkiRJUTOxlSRJkiRFzcRWkiRJkhS1ekpsjwWeAvZOOxBJkiRJUjyqTWw/uCxRwOnAJ4EHl6k+SZIkSVKTqJfEdhzIAk8uU32SJEmSpCaxehHbdAJfBu4FXgnsRBhh/RLQAWwErgXGgPcCW4B9gALQAzya1PMB4F3Ac8BdwJ7AR4GHgJsXEccAcO4itpMkSZIkNZHFjNhuAQ4D1gOHAzmgD1gDTCR/v5BsdwzwCeD1wE3AWUkdJwHHJfu/Hrg8KZMkSZIkqSqLnYp8N3BHsnwzYUS2nO8QRmDnbvc2wmjrs8n9jYTEV5IkSZKkqiw2sc0XLW+aZ79K22UIU5OrMZDUU+kmSZIkSWpC1V48ajPhnNuFXAZ8Ctg+uX84cESVjy1JkiRJ0qIuHjWfbwDfB75OuHhUJVcALyJMP36WcG7uVVQ/iitJkiRJanKZQiG13PIa4O3Mnr5cjUImU22e3ki2AHDojutSjqN+3PrM+mRpVapx1I8tyV/bY4ZtUmoLhUIB31+n+Rop5WuklK+T2WyPUqHf7LzDfmkHUjee3jSeLPk6Cew3pbZQKBQ+TThFtUS1U5EX601AcYZ1EOFZWq6kVpIkSZLUpGp1ePV24G+Bswm/Y/s48L4aPbYkSZIkqYHVKrH9PXBqjR5LkiRJktREajUVWZIkSZKkFWFiK0mSJEmKmomtJEmSJClqJraSJEmSpKiZ2EqSJEmSomZiK0mSJEmKmomtJEmSJClqJraSJEmSpKiZ2EqSJEmSomZiK0mSJEmKmomtJEmSJClqJraSJEmSpKiZ2EqSJEmSomZiK0mSJEmKmomtJEmSJClqJraSJEmSpKiZ2EqSJEmSomZiK0mSJEmKmomtJEmSJClqJraSJEmSpKiZ2EqSJEmSomZiK0mSJEmKmomtJEmSJClqJraSJEmSpKitTjuA5bUl7QDqzq3PrE87hDrk62Q226OUbVLKNpnN9ihlm5SyTWazPeZ6etN42iHUIV8ns9kei+WIrSRJkiQpag01Yvuqnd6Rdgh14ydPXwLAMTudnXIk9eOGpy8E4I22CQA3Ju1x0QHrUo6kfpx2T5jhcOiOtsm06Vkf3z74rSlHUh+O/dE3AXj2f++cciT1Y4eP/AcA+QdPTjmS+tH28ssB2Pz811OOpD6sbnknAPefeFjKkdSP/7zhBwA8d+PLUo6kfmz3xl8B8Nx3O1OOpD5sd1QYzf+3U7pSjqR+vOTSsXnXO2IrSZIkSYqaia0kSZIkKWomtpIkSZKkqJnYSpIkSZKiZmIrSZIkSYqaia0kSZIkKWomtpIkSZKkqJnYSpIkSZKiZmIrSZIkSYqaia0kSZIkKWomtpIkSZKkqJnYSpIkSZKiZmIrSZIkSYqaia0kSZIkKWomtpIkSZKkqJnYSpIkSZKiZmIrSZIkSYqaia0kSZIkKWomtpIkSZKkqJnYSpIkSZKiZmIrSZIkSYqaia0kSZIkKWomtpIkSZKkqJnYSpIkSZKiZmIrSZIkSYqaia0kSZIkKWomtpIkSZKkqJnYSpIkSZKiZmIrSZIkSYpavSS2xwE/BL4HXAt0pBuOJEmSJCkW1Sa2H1yGGHYBLgTWAocDFwDnL0O9kiRJkqQmUA+J7dPAa4HfJfdvBvYvs90AUJjnJkmSJElqQqsXsU0n8GXgXuCVwE7A6cCXCFOGNxKmD48B7wW2APsQks0e4NGkng8A7wKeA+4C9gQ+CjzETFIL8HbgliX/R5IkSZKkprKYEdstwGHAesJU4RzQB6wBJpK/X0i2Owb4BPB64CbgrKSOkwjn0R6erLs8KZvrDcDHgM8u4X+RJEmSJDWhxU5Fvhu4I1m+mTAiW853CCOwc7d7G3Au8GxyfyMh8S12IHAxcDLwSJm6B4DMPDdJkiRJUhNabGKbL1reNM9+lbbLMP95sC8HhgnTkO9eZEySJEmSJFV98ajNhHNuF3IZ8Clg++T+4cARyfKewFXAmcBtVcYjSZIkSWoyi7l41Hy+AXwf+Drh4lGVXAG8iDD9+FnCublXEUZxTwF2J0w1nvZH4IQqY5MkSZIkNYHFJLYPAUcX3f8N4YJRAH+d3KZtLFr+fnKbdmFym3YNMAn8XXKTJEmSJGmbVTsVebHeBKwrun8QsIrZ5+RKkiRJkrTNqp2KvFi3A38LnE34HdvHgffV6LElSZIkSQ2sVont74FTa/RYkiRJkqQmUqupyJIkSZIkrQgTW0mSJElS1ExsJUmSJElRM7GVJEmSJEXNxFaSJEmSFDUTW0mSJElS1ExsJUmSJElRM7GVJEmSJEXNxFaSJEmSFDUTW0mSJElS1ExsJUmSJElRM7GVJEmSJEXNxFaSJEmSFDUTW0mSJElS1ExsJUmSJElRM7GVJEmSJEXNxFaSJEmSFDUTW0mSJElS1ExsJUmSJElRM7GVJEmSJEXNxFaSJEmSFLVMoVBIO4Zl8bVrNxW+fu1zaYchSQ1nl86ptENQnbvyiIPTDkGKzto9e9IOQYrOSMfgp4GBcuscsZUkSZIkRc3EVpIkSZIUNRNbSZIkSVLUGuYcW+A04KKUY5AkSZIkrYzTqJDzNVJiK0mSJElqQk5FliRJkiRFzcRWkiRJkhQ1E1tJkiRJUtRMbCVJkiRJUTOxlSRJkiRFzcRWkiRJkhQ1E1tJkiRJUtRMbCVJkiRJUTOxlSRJkiRFzcRWkiRJkhQ1E1tJkiRJUtRMbCVJkiRJUTOxlSRJkiRFzcRWkiRJkhS11WkHIK2gQtoBSJIkqSqZtANQHByxlSRJkiRFzcRWkiRJkhQ1pyIrNh8C/hnIA18BXg30AzfMt1NLZruZO5npGS0zx3Uy08uZlorrMpni40DTZasqbpcpriOz8PahbFVJWcucxyqut2XrY5XW21LmsVrKbUdpHS1z1hWvLy5rKZT+r3P3nd4mlGXKbJ+ZZ7vMrPvl1oX/tdx2zLNdsi5TZvsyZS1Fk6C2xjSrrHz9ob5ydZSWtczZvnj9UrcvLiu7fdk6CgvWUbxPSzLjv1wcLUVnA2TK1lEoKZv7+LParcz2i6ljVtkSty/+f2ae06L/L1nXUrYtC6Vls+qd878U17s13jJl5bYrUzYT2yK3z8wf79x951tXqd5My/R2z1eso/z2xduV2bdl9r7l1z1fcftZ+7aUeawy27PYeqfraynabu5jzbeuaP189ZaNbYE6KPf402XT27WUe8ziOuasg6KOTMl2051l9kdrmTfMrW9sLWW2Ky5rWcS6mbLC1rJVlbebtf2qeeqdqaOQKVfvqoqPOVNvNWWry6wrU5ZZPXv7ouVCcdnc7YrWZbauay0tyxRtl+yTyZTu25KZXreqdF0S7ypORdpWjtgqNuuAp4BjgD2A04HBVCOSJEmSlCoTW8Vm+jDu8YSR2x8XlUmSJElqQia2is2dhGnHxwOjQBvw/Lx7SJIkSWponmOr2JwBHAQ8CDwN7EaYjixJkiSpSTliq9jcCIwBTyb3/x04P71wJEmSJKXNEVvFYgdgJ2B3YFdmzqvdBXhxWkFJkiRJSp+JrWLxPuDDhCT2TmYS26eAC9IKSpIkSVL6TGwVi79Lbn8F/H3KsUiSJEmqIya2is3fA68D9mX26/erqUQjSZIkKXUmtorN14BO4G5gS1JWwMRWkiRJalomtorNnwEHEJJZSZIkSfLnfhSdnwEvSjsISZIkSfXDEVvFZnfgHuB2YFNR+YnphCNJkiQpbSa2is1A2gFIkiRJqi8mtorN99IOQJIkSVJ9MbFVbPLMXDhqO6AV+A9gl9QikiRJkpQqE1vFpm3O/ZOAQ9IIRJIkSVJ98KrIit2VwJFpByFJkiQpPY7YKjYnFy23EH7X1t+0lSRJkpqYia1i8+ai5c3AQ8DadEKRJEmSVA9MbBWb09MOQJIkSVJ98RxbxWZv4ArgMeBR4FtJmSRJkqQmZWKr2PwzsAF4MdABXJ2USZIkSWpSJraKzR6ERHZzcrsoKZMkSZLUpExsFZvHgXcCq5LbO4F/TzUiSZIkSakysVVs1gFvBX4HPAJ04wWlJEmSpKbmVZEVm78G3gM8kdxvB84jJLySJEmSmpCJrWLzKmaSWoBJ4NUL7fR84bkVC0iSJElSupyKrNi0ALsW3W/HAzSSJElSUzMhUGyGgFuAEaBAON/2b1KNSJIkSVKqMoVCIe0YpG11AHAkkAG+C9xTYbviF3dmpYNSWT4H6bL90+dzkC7bP30+B+my/dU0TGzVyHwzT5/PQbps//T5HKTL9k+fz0G6bH81Dc+xlSRJkiRFzcRWkiRJkhQ1E1tJkiRJUtRMbCVJkiRJUfPnftTIPp12API5SJntnz6fg3TZ/unzOUiX7a+m4VWRJUmSJElRcyqyJEmSJClqJraSJEmSpKiZ2KqRHQvcBzwA9KccS7N4CPgpcDdwR1LWDtwI/DL5u2sqkTWu9cBjwM+Kyiq1eQb4IqFP/AToql2YDatc+w8AE4R+cDdwfNG6jxHa/z4gW5sQG95LgJuAXwA/Bz6UlNsPaqNS+w9gP6iVHYDbgR8TnoPp82pfBvyQ0AcuA7ZLyrdP7j+QrN+3hrFKK8bEVo1qFXABcBxwAPD25K9W3hHAQcCfJff7ge8C+yd/PciwvC4iHMQpVqnNj0vK9gfOBC6sTYgN7SJK2x/gfEI/OAi4Lik7ADgFeGWyT47wXqXqbAb6gD8FXgu8n9DW9oPaqNT+YD+olU3AkcCBhLY+lvBcfJ7wHOwPPAGckWx/RnJ/v2T952scr7QiTGzVqA4hHIl8EHgOuBRYm2pEzWstcHGyfDFwUoqxNKKbgck5ZZXafC3wVaAA3Aa8ANirBjE2snLtX8lawnvRJuBXhPeoQ1YormbyCDCWLOcJI4cd2A9qpVL7V2I/WH4F4I/JcmtyKxCS3ZGkfG4fmO4bI8BRhJkMUtRMbNWoOoB/K7r/G+b/oNXyKAA3AHcSRkIA9iR88SH5+8IU4mo2ldrcflE7HyBMc13PzBRY23/l7Qu8mjC90n5Qe/sy0/5gP6ilVYQp348Rpt6PA08SRtRhdjsXPwebgT8Au9UsUmmFmNiqUZU78uhvW628wwjnqx1HmI72hnTD0Rz2i9q4EOgkTAl8BBhKym3/lfWfgG8BHwaemmc7n4eVMbf97Qe1tYXQ1nsTRsD/tMw20+3sc6CGZGKrRvUbwgUtpu0N/DalWJrJdBs/BlxB+HB9lJlpfnsl67SyKrW5/aI2HiV8yXwe+DIz0yxt/5XTSkiqvgFcnpTZD2qnUvvbD2rvSWAj4RzbFwCrk/Lidi5+DlYDf8LiT6mQ6paJrRrVjwgXS3gZ4SqApwAbUo2o8e0MtBUtH0O4UuwG4D1J+XuAq2ofWtOp1OYbgHcTjta/ljD97JGSvVWt4vM138LMFZM3EN6Ltie8N+1PuJKpqpMB/olwbuf/KSq3H9RGpfa3H9TOHoQkFmBH4GjC83ET0J2Uz+0D032jG/hXHLFVA1i98CZSlDYTzu0ZJZx3sp5wCXytnD0Jo7QQ3lsuAb5NOMjwTcJVGH8N9KQSXeP6F2ANsDvhKPy5wCDl2/w6wk9uPAA8DZxe41gbUbn2X0OYElgg/ATW+5Jtf054Xu4hvEe9nzCipeocBryLmZ8aA/g49oNaqdT+b8d+UCt7ES4GtYowaPVN4BpCG18KfBa4i3AAguTv1wh9YJJwoEGKXqZQ8ACNJEmSJCleTkWWJEmSJEXNxFaSJEmSFDUTW0mSJElS1ExsJUmSJElRM7GVJEmSJEXNxFaSJEmSFDUTW0mSJElS1ExsJUmSJElR+/9aQ8JHOBHPuAAAAABJRU5ErkJggg==\n",
      "text/plain": [
       "<Figure size 1152x1152 with 2 Axes>"
      ]
     },
     "metadata": {
      "needs_background": "dark"
     },
     "output_type": "display_data"
    }
   ],
   "source": [
    "plotName = 'test'\n",
    "genenames = gdf['symbol'].values\n",
    "samplenames = [] #skipping names to replace in intaglio\n",
    "data = gdf.iloc[0:,7:37].values\n",
    "fig, ax = plt.subplots(figsize=[16,16])\n",
    "im, cbar = heatmap(data, genenames, samplenames, ax=ax)\n",
    "# texts = annotate_heatmap(im, valfmt=\"{x:.1f} t\")\n",
    "plt.show()\n",
    "\n",
    "\n",
    "genenamesr = rdf['symbol'].values\n",
    "samplenamesr = [] #skipping names to replace in intaglio\n",
    "datar = rdf.iloc[0:,7:16]\n",
    "genenamesm = mdf['symbol'].values\n",
    "samplenamesm = [] #skipping names to replace in intaglio\n",
    "datam = mdf.iloc[0:,5:41]\n",
    "genenameshc = mdf['symbol'].values\n",
    "samplenameshc = [] #skipping names to replace in intaglio\n",
    "datahc = hcdf.iloc[0:,2:8]\n",
    "\n",
    "figr, (axr,axm, axhc) = plt.subplots(1,3,figsize=[24,16])\n",
    "# figr = plt.figure(figsize=[16,2],constrained_layout=True)\n",
    "# fig_grid = figr.add_gridspec(1,28)\n",
    "# axr = figr.add_subplot(fig_grid[0,0:8])\n",
    "# axm = figr.add_subplot(fig_grid[0,9:])\n",
    "imr, cbarr = heatmap_rods(datar, genenamesr, samplenamesr, ax=axr)\n",
    "imm, cbarm = heatmap_glia(datam, genenamesm, samplenamesm, ax=axm)\n",
    "imhc, cbarhc = heatmap_haircell(datahc, genenameshc, samplenameshc, ax=axhc)\n",
    "plt.show()\n",
    "\n",
    "\n",
    "genenamessq = sqdf['symbol'].values\n",
    "samplenamessq = [] #skipping names to replace in intaglio\n",
    "datasq = sqdf.iloc[0:,8:21]\n",
    "figsq, axsq = plt.subplots(figsize=[16,16])\n",
    "imsq, cbarsq = heatmap_sqcones(datasq, genenamessq, samplenamessq, ax=axsq)\n",
    "plt.show()\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# genenames = gdf['symbol'].values\n",
    "# samplenames = [] #skipping names to replace in intaglio\n",
    "# data = gdf.iloc[0:,7:37].values\n",
    "\n",
    "# genenamesr = rdf['symbol'].values\n",
    "# samplenamesr = [] #skipping names to replace in intaglio\n",
    "# datar = rdf.iloc[0:,7:16]\n",
    "# genenamesm = mdf['symbol'].values\n",
    "# samplenamesm = [] #skipping names to replace in intaglio\n",
    "# datam = mdf.iloc[0:,5:41]\n",
    "\n",
    "# genenamessq = sqdf['symbol'].values\n",
    "# samplenamessq = [] #skipping names to replace in intaglio\n",
    "# datasq = sqdf.iloc[0:,8:21]\n",
    "\n",
    "# figr = plt.figure(figsize=[8,24],constrained_layout=True)\n",
    "# fig_grid = figr.add_gridspec(3,2)\n",
    "# ax = figr.add_subplot(fig_grid[0,:])\n",
    "# axr = figr.add_subplot(fig_grid[1,0])\n",
    "# axm = figr.add_subplot(fig_grid[1,1])\n",
    "# axsq = figr.add_subplot(fig_grid[2,:])\n",
    "# # axm = plt.subplot(2,2,2)\n",
    "# im, cbar = heatmap(data, genenames, samplenames, ax=ax,cmap=\"inferno\", cbarlabel=\"fpkm\")\n",
    "# imr, cbarr = heatmap_rods(datar, genenamesr, samplenamesr, ax=axr, cmap=\"inferno\", cbarlabel=\"cpm\")\n",
    "# imm, cbarm = heatmap_glia(datam, genenamesm, samplenamesm, ax=axm, cmap=\"inferno\", cbarlabel=\"tmm\")\n",
    "# imsq, cbarsq = heatmap_sqcones(datasq, genenamessq, samplenamessq, ax=axsq, cmap=\"inferno\", cbarlabel=\"counts\")\n",
    "\n",
    "# plt.show()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "***\n",
    "## sparkline plot for markdown file<a id='sparkline'>∮</a>\n",
    "***\n",
    "[Back to Index](#Index)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "genename = 'grk7a'\n",
    "n,h = plotBars(genename)\n",
    "fH, ax = plt.subplots(figsize=[2.25,.25])\n",
    "ax.axhline(y = np.mean(h), color = 'white', linewidth = .2, alpha = 1, solid_capstyle='butt', zorder = 0)\n",
    "pH = ax.bar(n, h, width=.8, bottom=None, align='center', data=None, color=pC['plt'], zorder = 1)\n",
    "ax.text(.90, 1, genename,horizontalalignment='left',verticalalignment='top',transform=ax.transAxes, fontsize=7, zorder = 0)\n",
    "ax.text(.97, 0.2, (np.around(np.mean(h),decimals=0)),horizontalalignment='left',verticalalignment='top',transform=ax.transAxes, fontsize=6, zorder = 0)\n",
    "ax.set_axis_off()\n",
    "# plt.savefig(\"/Users/angueyraaristjm/Documents/LiMolec/zfGenotyping/targetGenes/202001_zfDEGs/resources/{0}.svg\".format(genename), transparent=True, format=\"svg\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "'{0}fpkm'.format(np.around(np.mean(h),decimals=0))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# genename = 'kcnk1'\n",
    "# n,h = plotBars_sqcones(genename)\n",
    "# fH, ax = plt.subplots(figsize=[2.2,.25])\n",
    "# ax.axhline(y = np.mean(h), color = 'white', linewidth = .2, alpha = 1, solid_capstyle='butt')\n",
    "# pH = ax.bar(n, h, width=.8, bottom=None, align='center', data=None, color=sqC['plt'])\n",
    "# ax.text(.92, 1, genename,horizontalalignment='left',verticalalignment='top',transform=ax.transAxes, fontsize=6)\n",
    "# ax.text(.97, 0.2, (np.around(np.mean(h),decimals=0)),horizontalalignment='left',verticalalignment='top',transform=ax.transAxes, fontsize=4)\n",
    "# ax.set_axis_off()\n",
    "# plt.savefig(\"/Users/angueyraaristjm/Documents/LiMolec/zfGenotyping/targetGenes/202001_zfDEGs/resources/{0}sq.svg\".format(genename), transparent=True, format=\"svg\")"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "***\n",
    "## PCA <a id='pca'>∮</a>\n",
    "***\n",
    "[Back to Index](#Index)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### This is supposed to run only once. Creates sparklines and a markdown file with entries for these genes"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# deseqdir= \"~/Documents/LiMolec/zfRNAseq/20190827/20190827_DESeq2/00_rodsVcones/\"\n",
    "# fname = \"fromPython_PCgenes_RvC_fromPython.md\"\n",
    "\n",
    "# deseqdir= \"~/Documents/LiMolec/zfRNAseq/20190827/20190827_DESeq2/00_USvsLM/\"\n",
    "# fname = \"fromPython_PCgenes_USvLM.md\"\n",
    "\n",
    "# deseqdir= \"~/Documents/LiMolec/zfRNAseq/20190827/20190827_DESeq2/00_LvsM/\"\n",
    "# fname = \"fromPython_PCgenes_LvM.md\"\n",
    "\n",
    "deseqdir= \"~/Documents/LiMolec/zfRNAseq/20190827/20190827_DESeq2/00_UvsS/\"\n",
    "fname = \"fromPython_PCgenes_UvS.md\"\n",
    "\n",
    "\n",
    "\n",
    "pcadir= \"~/Documents/LiMolec/zfRNAseq/20190827/20190827_DESeq2/PCA_genes/\"\n",
    "%cd {pcadir}\n",
    "pcr = pd.read_csv(deseqdir + \"pcaWeights.csv\")\n",
    "pcr.sort_values(by=['PC1'])\n",
    "pcr"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Sparkline plots\n",
    "genename = 'syt5b'\n",
    "n,h = plotBars(genename)\n",
    "fH, ax = plt.subplots(figsize=[2.25,.25])\n",
    "ax.axhline(y = np.mean(h), color = 'white', linewidth = .2, alpha = 1, solid_capstyle='butt', zorder = 0)\n",
    "pH = ax.bar(n, h, width=.8, bottom=None, align='center', data=None, color=pC['plt'], zorder = 1)\n",
    "ax.text(.90, 1, genename,horizontalalignment='left',verticalalignment='top',transform=ax.transAxes, fontsize=7, zorder = 0)\n",
    "ax.text(.97, 0.2, (np.around(np.mean(h),decimals=0)),horizontalalignment='left',verticalalignment='top',transform=ax.transAxes, fontsize=6, zorder = 0)\n",
    "ax.set_axis_off()\n",
    "plt.savefig(\"./resources/{0}.svg\".format(genename), transparent=True, format=\"svg\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "\n",
    "# nTop = 100;\n",
    "# print(\"Started...\")\n",
    "# for i in range(nTop):\n",
    "#     genename = pcr['symbol'].iloc[i]\n",
    "#     if os.path.isfile(\"./resources/{0}.svg\".format(genename)) == False:\n",
    "#         n,h = plotBars(genename)\n",
    "#         fH, ax = plt.subplots(figsize=[2.25,.25])\n",
    "#         ax.axhline(y = np.mean(h), color = 'white', linewidth = .2, alpha = 1, solid_capstyle='butt', zorder = 0)\n",
    "#         pH = ax.bar(n, h, width=.8, bottom=None, align='center', data=None, color=pC['plt'], zorder = 1)\n",
    "#         ax.text(.90, 1, genename,horizontalalignment='left',verticalalignment='top',transform=ax.transAxes, fontsize=7, zorder = 0)\n",
    "#         ax.text(.97, 0.2, (np.around(np.mean(h),decimals=0)),horizontalalignment='left',verticalalignment='top',transform=ax.transAxes, fontsize=6, zorder = 0)\n",
    "#         ax.set_axis_off()\n",
    "#         plt.savefig(\"./resources/{0}.svg\".format(genename), transparent=True, format=\"svg\")\n",
    "#         plt.close(fH)\n",
    "# print(\"Up plots created...\")\n",
    "\n",
    "# for i in -np.add(range(nTop),1):\n",
    "#     genename = pcr['symbol'].iloc[i]\n",
    "#     if os.path.isfile(\"./resources/{0}.svg\".format(genename)) == False:\n",
    "#         n,h = plotBars(genename)\n",
    "#         fH, ax = plt.subplots(figsize=[2.25,.25])\n",
    "#         ax.axhline(y = np.mean(h), color = 'white', linewidth = .2, alpha = 1, solid_capstyle='butt', zorder = 0)\n",
    "#         pH = ax.bar(n, h, width=.8, bottom=None, align='center', data=None, color=pC['plt'], zorder = 1)\n",
    "#         ax.text(.90, 1, genename,horizontalalignment='left',verticalalignment='top',transform=ax.transAxes, fontsize=7, zorder = 0)\n",
    "#         ax.text(.97, 0.2, (np.around(np.mean(h),decimals=0)),horizontalalignment='left',verticalalignment='top',transform=ax.transAxes, fontsize=6, zorder = 0)\n",
    "#         ax.set_axis_off()\n",
    "#         plt.savefig(\"./resources/{0}.svg\".format(genename), transparent=True, format=\"svg\")\n",
    "#         plt.close(fH)\n",
    "# print(\"Down plots created...\")\n",
    "        \n",
    "# fid = open(fname, 'w+t') \n",
    "# fid.write(\"_________________________________________________________\\n\")\n",
    "# fid.write(\"## Top \" + str(nTop) + \" genes enriched in rods \\n\\n\\n\")\n",
    "# for i in range(nTop):\n",
    "#     fid.write(\"#### _\" + pcr['symbol'].iloc[i] + \"_ ![\" + pcr['symbol'].iloc[i] + \"](./resources/\" + pcr['symbol'].iloc[i] + \".svg)\\n\")\n",
    "#     fid.write(\"- _\" + str(pcr['genename'].iloc[i]) + \"_\\n\")\n",
    "#     fid.write(\"- [ ] &#10060;?&#9989;?&#8505;?&#9770;? RANK = \" + str(i+1) + \" ![R](./resources/R.svg)\\n\")\n",
    "#     fid.write(\"- [ZFIN link]()\" + \"\\n\")\n",
    "#     fid.write(\"- NOTESHERE\" + \"\\n\")\n",
    "#     fid.write(\"\\n\")\n",
    "\n",
    "\n",
    "# fid.write(\"_________________________________________________________\\n\")\n",
    "# fid.write(\"## Top \" + str(nTop) + \" genes enriched in cones \\n\\n\\n\")\n",
    "# for i in -np.add(range(nTop),1):\n",
    "#     fid.write(\"#### _\" + pcr['symbol'].iloc[i] + \"_ ![\" + pcr['symbol'].iloc[i] + \"](./resources/\" + pcr['symbol'].iloc[i] + \".svg)\\n\")\n",
    "#     fid.write(\"- _\" + str(pcr['genename'].iloc[i]) + \"_\\n\")\n",
    "#     fid.write(\"- [ ] &#10060;?&#9989;?&#8505;?&#9770;? RANK = \" + str(-i) + \" ![Cs](./resources/Cs.svg)\\n\")\n",
    "#     fid.write(\"- [ZFIN link]()\" + \"\\n\")\n",
    "#     fid.write(\"- NOTESHERE\" + \"\\n\")\n",
    "#     fid.write(\"\\n\")\n",
    "\n",
    "# fid.close()\n",
    "\n",
    "\n",
    "# print(\"MD file created. \")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "help(plt.subplots_adjust)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# histogram of PC weights\n",
    "pcr.sort_values(by=['PC1'])\n",
    "a = pcr['PC1']\n",
    "# a = np.divide(a,pcr['PC1'].sum())\n",
    "a = a[a != 0] #there are too many nil-values\n",
    "fH, ax = plt.subplots(figsize=[12,6])\n",
    "n, h, patches = ax.hist(np.multiply(a,100), bins=100, log=True) #use bins='fd' for optimal binning"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "***\n",
    "## misc. things<a id='misc'>∮</a>\n",
    "***\n",
    "[Back to Index](#Index)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "#### Adding genenames to pcaWeights and using opportunity to sort by PC1 value and remove NaN"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "deseqdir= \"~/Documents/LiMolec/zfRNAseq/20190827/20190827_DESeq2\"\n",
    "# pcr = pd.read_csv(deseqdir + \"/00_rodsVcones/pcaWeights_Rout.csv\")\n",
    "# pcr = pd.read_csv(deseqdir + \"/00_USvsLM/pcaWeights_Rout.csv\")\n",
    "# pcr = pd.read_csv(deseqdir + \"/00_LvsM/pcaWeights_Rout.csv\")\n",
    "pcr = pd.read_csv(deseqdir + \"/00_UvsS/pcaWeights_Rout.csv\")\n",
    "# pcr.head\n",
    "gdf = pd.read_csv(deseqdir + \"/00_rodsVcones/rodVCones_fpkm.csv\")\n",
    "pcr = pcr.set_index('symbol')\n",
    "pcr = pcr.reindex(index=gdf['symbol'])\n",
    "pcr = pcr.reset_index()\n",
    "pcr.insert(1,'genename',gdf['genename'],True)\n",
    "pcr.dropna(subset=['PC1'], inplace=True)\n",
    "pcr.sort_values(by=['PC1'], inplace=True)\n",
    "# pcr.to_csv(path_or_buf=deseqdir + \"/00_rodsVcones/pcaWeights.csv\", index=False)\n",
    "# pcr.to_csv(path_or_buf=deseqdir + \"/00_USvsLM/pcaWeights.csv\", index=False)\n",
    "# pcr.to_csv(path_or_buf=deseqdir + \"/00_LvsM/pcaWeights.csv\", index=False)\n",
    "# pcr.to_csv(path_or_buf=deseqdir + \"/00_UvsS/pcaWeights.csv\", index=False)\n",
    "pcr"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "help(gdf.to_csv)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Trying to draw counts for single gene by wedges in a circle \n",
    "> Prob not worth it, difficult to interpret"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "fig, ax = plt.subplots(figsize=[12,12])\n",
    "\n",
    "n = 30;\n",
    "r = 0;\n",
    "wdg_theta = 360/30;\n",
    "\n",
    "patches = []\n",
    "circles = []\n",
    "\n",
    "pM = matplotlib.patches.Wedge(0, np.mean(data[r,0:6]), wdg_theta * (0), (wdg_theta * (6)), edgecolor=\"none\", facecolor = pC['r'], alpha = 0.4)\n",
    "patches.append(pM)\n",
    "\n",
    "pM = matplotlib.patches.Wedge(0, np.mean(data[r,7:11]), wdg_theta * (6), (wdg_theta * (11)), edgecolor=\"none\", facecolor = pC['u'], alpha = 0.4)\n",
    "patches.append(pM)\n",
    "\n",
    "pM = matplotlib.patches.Wedge(0, np.mean(data[r,12:17]), wdg_theta * (11), (wdg_theta * (17)), edgecolor=\"none\", facecolor = pC['s'], alpha = 0.4)\n",
    "patches.append(pM)\n",
    "\n",
    "pM = matplotlib.patches.Wedge(0, np.mean(data[r,18:24]), wdg_theta * (17), (wdg_theta * (24)), edgecolor=\"none\", facecolor = pC['m'], alpha = 0.4)\n",
    "patches.append(pM)\n",
    "\n",
    "pM = matplotlib.patches.Wedge(0, np.mean(data[r,25:30]), wdg_theta * (24), (wdg_theta * (30)), edgecolor=\"none\", facecolor = pC['l'], alpha = 0.4)\n",
    "patches.append(pM)\n",
    "\n",
    "for i in range(n):\n",
    "    p = matplotlib.patches.Wedge(0, data[r,i], wdg_theta * (i), (wdg_theta * (i+1)) -2, edgecolor=\"none\", facecolor = pC['plt'][i], alpha = 0.4)\n",
    "    patches.append(p)\n",
    "\n",
    "for i in [0.001,.2,.4,.6,.8,1]:\n",
    "    c = matplotlib.patches.Circle((0,0), np.max(data[r,:])*i, edgecolor = '#000000', facecolor='none', alpha=0.59)\n",
    "    circles.append(c)\n",
    "\n",
    "plt.xlim(-np.max(data[r,:])*1, np.max(data[r,:])*1)\n",
    "plt.ylim(-np.max(data[r,:])*1, np.max(data[r,:])*1)\n",
    "\n",
    "\n",
    "circleColl = matplotlib.collections.PatchCollection(circles, match_original=True);\n",
    "ax.add_collection(circleColl)\n",
    "\n",
    "patchColl = matplotlib.collections.PatchCollection(patches, match_original=True);\n",
    "ax.add_collection(patchColl)\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "np.max(data[r,:])"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "np.max(data[4,i])"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "data"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "range(0,4)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "range?"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### From 2019, Whitaker ,..., Swaroop\n",
    "> genes involved in rod synapse size and localization\n",
    "\n",
    "> [link to biorXiv](https://www.biorxiv.org/content/biorxiv/early/2019/08/30/753012.full.pdf)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "plotName = \"test\"\n",
    "\n",
    "# gdf = findgenelist('dpf') # dpf3 -> not this one but dpf2 in all photoreceptors\n",
    "# gdf = findgenelist('epb') # epb4.1l2 -> not this one but some interesting patterns in this family\n",
    "# gdf = findgenelist('grtp') # grtp1 -> this one is interesting!\n",
    "# gdf = findgenelist('kcnj') # kncj14 -> squirrel too but not zebrafish\n",
    "# gdf = gdf.sort_values(by=[\"baseMean\"],ascending=False)\n",
    "\n",
    "# gdf = findgenelist('llgl') # llgl2 -> nope\n",
    "# gdf = findgenelist('rab') # rab28 -> nope; this family is too big\n",
    "# gdf = gdf[gdf.baseMean > 2]\n",
    "# gdf = gdf.sort_values(by=[\"baseMean\"],ascending=False)\n",
    "\n",
    "# gdf = findgenelist('rom1') #rom1 -> rod-specific; what about outer segments?\n",
    "\n",
    "# gdf = findgenelist('arl') # arl2bp -> nope but arl3l2 in all photoreceptors, and arl3l1 downregulated in uv and s. Both confirmed eye specific (https://zfin.org/ZDB-GENE-040426-1678; https://zfin.org/ZDB-GENE-040426-1649)\n",
    "# These genes are supponsed to interact with unc119c (https://www.ncbi.nlm.nih.gov/pubmed/27079236) and act as their GDF\n",
    "# gdf = gdf[gdf.baseMean > 5]\n",
    "# # gdf = gdf[gdf.baseMean < 300]\n",
    "# gdf = gdf.sort_values(by=[\"baseMean\"],ascending=False)\n",
    "\n",
    "# gdf = findgenelist('unc119') # unc119c is photoreceptor specific (https://www.ncbi.nlm.nih.gov/pubmed/27079236) and CoIPs with arl3l2\n",
    "\n",
    "\n",
    "gdf = findgenelist('ppar') #ppargc2\n",
    "# gdf = findgenelist('dpf')\n",
    "# gdf = findgenelist('dpf')\n",
    "# gdf = findgenelist('dpf')\n",
    "# gdf = findgenelist('dpf')\n",
    "# gdf = findgenelist('dpf')\n",
    "# gdf = findgenelist('dpf')\n",
    "# gdf = findgenelist('dpf')\n",
    "# gdf = findgenelist('dpf')\n",
    "# gdf = findgenelist('dpf')\n",
    "# gdf = findgenelist('dpf')\n",
    "# gdf = findgenelist('dpf')\n",
    "# gdf = findgenelist('dpf')\n",
    "# gdf = findgenelist('dpf')\n",
    "\n",
    "### https://www.sciencedirect.com/science/article/pii/S2211124718303218\n",
    "# gdf = findgenelist('frmpd') #frmpd2 interacts with lrit1 in cones \n",
    "# gdf = findgenelist('sdc') #sdcbp interacts with lrit 1\n",
    "# gdf = findgenelist('grip') #grip1 interacts with lrit1\n",
    "\n",
    "# gdf = findgenelist('dpf')\n",
    "# gdf = gdf.sort_values(by=[\"baseMean\"],ascending=False)\n",
    "# gdf = gdf.drop(np.arange(20,93))\n",
    "# gdf = gdf.drop([0])\n",
    "# gdf = gdf.sort_values(by=[\"symbol\"])\n",
    "# gdf = gdf[gdf.baseMean > .45]\n",
    "\n",
    "gdf\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "genenames = gdf['symbol'].values\n",
    "samplenames = [] #skipping names to replace in intaglio\n",
    "log2matrix = np.vectorize(log2)\n",
    "# data = log2matrix(gdf.iloc[0:,7:15].values)\n",
    "data = gdf.iloc[0:,7:37].values\n",
    "\n",
    "\n",
    "fig, ax = plt.subplots(figsize=[12,12])\n",
    "im, cbar = heatmap(data, genenames, samplenames, ax=ax,\n",
    "                   cmap=\"inferno\", cbarlabel=\"fpkm\")\n",
    "# texts = annotate_heatmap(im, valfmt=\"{x:.1f} t\")\n",
    "\n",
    "plt.savefig(\"{0}.svg\".format(plotName), transparent=True, format=\"svg\")\n",
    "plt.show()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# gdf = gdf[gdf.baseMean > .45]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.7.4"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 4
}
