{
 "cells": [
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Load required modules"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 1,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "LOADED\n"
     ]
    }
   ],
   "source": [
    "# External libraries\n",
    "import numpy as np\n",
    "import pandas as pd\n",
    "import pygal\n",
    "import math\n",
    "import IPython\n",
    "import matplotlib\n",
    "import matplotlib.pyplot as plt\n",
    "import matplotlib.font_manager as font_manager\n",
    "import importlib\n",
    "import svgutils\n",
    "from svgutils.compose import *\n",
    "#import svgwrite\n",
    "\n",
    "# Custom code\n",
    "from fx_RNAseq import *\n",
    "def findgene(genename):\n",
    "    dPath = \"~/Documents/LiMolec/zfRNAseq/20190827/20190827_DESeq2/\" #directory path\n",
    "#     fPath = \"00_rodsVcones/rodVCones_raw.csv\" #file path\n",
    "    fPath = \"00_rodsVcones/rodVCones_fpkm.csv\" #file path\n",
    "    tempPath = \"temp.txt\" # temporary file to store values\n",
    "    !head -1 {dPath + fPath} > {dPath + tempPath}\n",
    "    # EXACT MATCH ONLY by matching word (-w) and only at the start of each line (^) which corresponds to the symbol\n",
    "    !grep -w ^$genename {dPath + fPath} >> {dPath + tempPath}\n",
    "    gdf = pd.read_csv(dPath + tempPath)\n",
    "    return gdf\n",
    "\n",
    "def findgenelist(genename):\n",
    "    dPath = \"~/Documents/LiMolec/zfRNAseq/20190827/20190827_DESeq2/\" #directory path\n",
    "#     fPath = \"00_rodsVcones/rodVCones_raw.csv\" #file path\n",
    "    fPath = \"00_rodsVcones/rodVCones_fpkm.csv\" #file path\n",
    "    tempPath = \"temp.txt\" # temporary file to store values\n",
    "    !head -1 {dPath + fPath} > {dPath + tempPath}\n",
    "    # EXACT MATCH ONLY by matching word (-w) and only at the start of each line (^) which corresponds to the symbol\n",
    "    !grep ^$genename {dPath + fPath} >> {dPath + tempPath}\n",
    "    gdf = pd.read_csv(dPath + tempPath)\n",
    "    return gdf\n",
    "\n",
    "# def findgene(genename):\n",
    "#     !head -1 ~/Dropbox/LiLab/RNAseq/20190827_DESeq2/00_rodsVcones/rodVCones_raw.csv > ~/Dropbox/LiLab/RNAseq/20190827_DESeq2/temp.txt\n",
    "#     # EXACT MATCH ONLY by matching word (-w) and only at the start of each line (^) which corresponds to the symbol\n",
    "#     !grep -w ^$genename ~/Dropbox/LiLab/RNAseq/20190827_DESeq2/00_rodsVcones/rodVCones_raw.csv >> ~/Dropbox/LiLab/RNAseq/20190827_DESeq2/temp.txt\n",
    "#     gdf = pd.read_csv(\"~/Dropbox/LiLab/RNAseq/20190827_DESeq2/temp.txt\")\n",
    "    \n",
    "#     return gdf\n",
    "\n",
    "# def findgenelist(genename):\n",
    "#     !head -1 ~/Dropbox/LiLab/RNAseq/20190827_DESeq2/00_rodsVcones/rodVCones_raw.csv > ~/Dropbox/LiLab/RNAseq/20190827_DESeq2/temp.txt\n",
    "#     # EXACT MATCH ONLY by matching word (-w) and only at the start of each line (^) which corresponds to the symbol\n",
    "#     !grep ^$genename ~/Dropbox/LiLab/RNAseq/20190827_DESeq2/00_rodsVcones/rodVCones_raw.csv >> ~/Dropbox/LiLab/RNAseq/20190827_DESeq2/temp.txt\n",
    "#     gdf = pd.read_csv(\"~/Dropbox/LiLab/RNAseq/20190827_DESeq2/temp.txt\")\n",
    "#     return gdf\n",
    "\n",
    "def plotNReads():\n",
    "    rdf = pd.read_csv(\"~/Documents/LiMolec/zfRNAseq/20190827/20190827_DESeq2/nMappedReads.csv\")\n",
    "    h = rdf.iloc[:,1]\n",
    "    h = h / 1e6\n",
    "    n = np.arange(1,7) # Rods\n",
    "    n = np.append(n, 6.5 + np.arange(1,6)) # UV\n",
    "    n = np.append(n, 12 + np.arange(1,7)) # S\n",
    "    n = np.append(n, 18.5 + np.arange(1,8)) # M\n",
    "    n = np.append(n, 26 + np.arange(1,7)) # L\n",
    "    return n, h\n",
    "\n",
    "def plotBars(genename):\n",
    "    gdf = findgene(genename)\n",
    "    n = np.arange(1,7) # Rods\n",
    "    n = np.append(n, 6.5 + np.arange(1,6)) # UV\n",
    "    n = np.append(n, 12 + np.arange(1,7)) # S\n",
    "    n = np.append(n, 18.5 + np.arange(1,8)) # M\n",
    "    n = np.append(n, 26 + np.arange(1,7)) # L\n",
    "    h_start = 7\n",
    "    h_end = 37\n",
    "    h = gdf.iloc[0,h_start:h_end]\n",
    "    return n, h\n",
    "\n",
    "def plotBars_list(genelist):\n",
    "    n=[]\n",
    "    h=[]\n",
    "    for i,gene in enumerate(genelist, start=0):\n",
    "        tempn, temph = plotBars(gene)\n",
    "        n = np.append(n,tempn+(30*i))\n",
    "        h = np.append(h,temph)\n",
    "    return n, h\n",
    "\n",
    "def svgreload():\n",
    "    !osascript -e 'tell application \"Google Chrome\" to tell the active tab of second window to reload'\n",
    "\n",
    "from pygal.style import Style\n",
    "custom_style = Style(\n",
    "    background = 'black',\n",
    "    plot_background = '#111111',\n",
    "    foreground = '#999999',\n",
    "    foreground_strong = '#eeeeee',\n",
    "    foreground_subtle = '#555555',\n",
    "    opacity = '.8',\n",
    "    opacity_hover = '.4',\n",
    "    transition = '10ms',\n",
    "    colors = ('#747474','#B540B7', '#4669F2','#04CD22','#CC2C2A'),\n",
    "    label_font_size = 24,\n",
    "    major_label_font_size = 24,\n",
    "    title_font_size = 24,\n",
    "    legend_font_size = 18\n",
    ")\n",
    "\n",
    "#photoreceptor Colors\n",
    "pC = {\n",
    "    'r' : '#747474',\n",
    "    'u' : '#B540B7',\n",
    "    's' : '#4669F2',\n",
    "    'm' : '#04CD22',\n",
    "    'l' : '#CC2C2A',\n",
    "    'plt' : '',\n",
    "}\n",
    "\n",
    "pC['plt']=[\n",
    "    pC['r'],pC['r'],pC['r'],pC['r'],pC['r'],pC['r'],\n",
    "    pC['u'],pC['u'],pC['u'],pC['u'],pC['u'],\n",
    "    pC['s'],pC['s'],pC['s'],pC['s'],pC['s'],pC['s'],\n",
    "    pC['m'],pC['m'],pC['m'],pC['m'],pC['m'],pC['m'],pC['m'],\n",
    "    pC['l'],pC['l'],pC['l'],pC['l'],pC['l'],pC['l']\n",
    "]\n",
    "\n",
    "print('LOADED')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "/Users/angueyraaristjm/Documents/LiMolec/zfRNAseq/20190827/20190827_DESeq2/genePlots/fpkm\n"
     ]
    }
   ],
   "source": [
    "# %cd ~/Dropbox/Presentations/RevealReboot/resources/20191002_RNAseq/\n",
    "%cd ~/Documents/LiMolec/zfRNAseq/20190827/20190827_DESeq2/genePlots/fpkm/"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "***\n",
    "# Index <a id='Index'>∮</a>\n",
    "***\n",
    "- [Plot Total Number of Reads](#nReads)\n",
    "- [PCA Plots](#pcaPlots)\n",
    "- [BarPlot: Opsins](#barOpsins)\n",
    "- [BarPlot: Custom](#barCustom)\n",
    "- [BarPlot: Many Genes](#barMany)\n",
    "- [BarPlot_subPlotsAutomatic: Custom](#barManyAuto)\n",
    "- [Heatmap: Opsins](#heatOpsins)\n",
    "- [Heatmap: Custom](#heatCustom)\n",
    "- [GO clustering](https://go.princeton.edu/) \n",
    "***"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "***\n",
    "## Total number of reads<a id='nReads'>∮</a>\n",
    "***\n",
    "[Back to Index](#Index)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### pygal version (does not import well into Intaglio)\n",
    "> hist = pygal.Histogram(x_title='',y_title='million of mapped reads',show_x_labels=False,title='',style=custom_style,legend_at_bottom=True,legend_at_bottom_columns=4,show_legend=False)  \n",
    "> hist.add('Rod',[(11.27,0.2,0.8),(11.77,1.2,1.8),(9.4,2.2,2.8),(14.99,3.2,3.8),(9.75,4.2,4.8),(9.08,5.2,5.8),])  \n",
    "> hist.add('UV',[(9.42,6.2,6.8),(6.52,7.2,7.8),(10.45,8.2,8.8),(9.46,9.2,9.8),(11.02,10.2,10.8)])  \n",
    "> hist.add('S',[(9.23,11.2,11.8),(9.33,12.2,12.8),(9.37,13.2,13.8),(14.95,14.2,14.8),(12.85,15.2,15.8),(10.96,16.2,16.8)])  \n",
    "> hist.add('M',[(10.40,18.2,18.8),(8.634,19.2,19.8),(7.36,20.2,20.8),(10.32,21.2,21.8),(8.33,22.2,22.8),(9.53,23.2,23.8),(9.526,17.2,17.8)])  \n",
    "> hist.add('L',[(8.99,24.2,24.8),(10.50,25.2,25.8),(10.03,26.2,26.8),(11.08,27.2,27.8),(10.44,28.2,28.8),(10.72,29.2,29.8)])  \n",
    "> hist.render_to_file('test.svg')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# dark background\n",
    "params = {\"ytick.color\" : \"w\",\n",
    "          \"xtick.color\" : \"w\",\n",
    "          \"axes.labelcolor\" : \"w\",\n",
    "          \"axes.edgecolor\" : \"w\",\n",
    "         \"axes.linewidth\" : 3,\n",
    "         \"xtick.major.width\" : 3,\n",
    "         \"ytick.major.width\" : 3,\n",
    "         \"xtick.major.size\" : 8,\n",
    "         \"ytick.major.size\" : 8,\n",
    "         \"text.color\" : \"w\"}\n",
    "plt.rcParams.update(params)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# white background\n",
    "params = {\"ytick.color\" : \"k\",\n",
    "          \"xtick.color\" : \"k\",\n",
    "          \"axes.labelcolor\" : \"k\",\n",
    "          \"axes.edgecolor\" : \"k\",\n",
    "         \"axes.linewidth\" : 3,\n",
    "         \"xtick.major.width\" : 3,\n",
    "         \"ytick.major.width\" : 3,\n",
    "         \"xtick.major.size\" : 8,\n",
    "         \"ytick.major.size\" : 8,\n",
    "         \"text.color\" : \"w\"}\n",
    "plt.rcParams.update(params)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "#pyplot version (Sep_2019)\n",
    "plotname = ''\n",
    "n,h = plotNReads()\n",
    "fH, ax = plt.subplots()\n",
    "pH = plt.bar(n, h, width=0.8, bottom=None, align='center', data=None, color=pC['plt'])\n",
    "formatFigure_nReads(plotname, fH, ax, pH)\n",
    "plt.tight_layout()\n",
    "plt.savefig(\"{0}.svg\".format('nReads'), transparent=True, format=\"svg\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "np.mean(h)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": []
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "***\n",
    "## pcaPlots<a id='pcaPlots'>∮</a>\n",
    "***\n",
    "[Back to Index](#Index)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "def formatFigure_pca(plotname, figH, axH, plotH):\n",
    "    font_path = '/System/Library/Fonts/Avenir.ttc'\n",
    "    fontTicks = font_manager.FontProperties(fname=font_path, size=18)\n",
    "    fontLabels = font_manager.FontProperties(fname=font_path, size=22)\n",
    "    fontTitle = font_manager.FontProperties(fname=font_path, size=28)\n",
    "    \n",
    "    axH.ticklabel_format(style='sci',axis='y',scilimits=(0,4))\n",
    "    axH.ticklabel_format(style='sci',axis='x',scilimits=(0,4))\n",
    "    axH.set_title(plotname, fontproperties=fontTitle)\n",
    "    axH.spines['top'].set_visible(False)\n",
    "    axH.spines['right'].set_visible(False)\n",
    "    \n",
    "    for label in (axH.get_xticklabels() + axH.get_yticklabels()):\n",
    "        label.set_fontproperties(fontTicks)\n",
    "    axH.tick_params(axis = 'both', which = 'major', labelsize = 18)\n",
    "    axH.xaxis.offsetText.set_fontsize(18)\n",
    "    axH.yaxis.offsetText.set_fontsize(18)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "#pyplot version (Sep_2019)\n",
    "plotname = 'PCA_RodsCones'\n",
    "pc_df = pd.read_csv(\"~/Documents/LiMolec/zfRNAseq/20190827/20190827_DESeq2/00_rodsVcones/pcaData.csv\")\n",
    "pc1 = pc_df.iloc[:,0]\n",
    "pc2 = pc_df.iloc[:,1]\n",
    "fH, ax = plt.subplots(figsize=[5,5])\n",
    "pH = plt.scatter(pc1, pc2, color=pC['plt'])\n",
    "formatFigure_pca(plotname, fH, ax, pH)\n",
    "ax.set_xlabel('PC1 (var = 21%)', fontproperties=font_manager.FontProperties(fname='/System/Library/Fonts/Avenir.ttc', size=22))\n",
    "ax.set_ylabel('PC2 (var = 9%)', fontproperties=font_manager.FontProperties(fname='/System/Library/Fonts/Avenir.ttc', size=22))\n",
    "plt.xlim(-100,50)\n",
    "plt.ylim(-60,60)\n",
    "plt.tight_layout()\n",
    "plt.savefig(\"{0}.svg\".format(plotname), transparent=True, format=\"svg\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "#pyplot version (Sep_2019)\n",
    "plotname = 'PCA_UVS_ML'\n",
    "pc_df = pd.read_csv(\"~/Documents/LiMolec/zfRNAseq/20190827/20190827_DESeq2/00_USvsLM/pcaData.csv\")\n",
    "pc1 = pc_df.iloc[:,0]\n",
    "pc2 = pc_df.iloc[:,1]\n",
    "fH, ax = plt.subplots(figsize=[5,5])\n",
    "pH = plt.scatter(pc1, pc2, color=[\n",
    "    pC['u'],pC['u'],pC['u'],pC['u'],pC['u'],\n",
    "    pC['s'],pC['s'],pC['s'],pC['s'],pC['s'],pC['s'],\n",
    "    pC['m'],pC['m'],pC['m'],pC['m'],pC['m'],pC['m'],pC['m'],\n",
    "    pC['l'],pC['l'],pC['l'],pC['l'],pC['l'],pC['l']\n",
    "]\n",
    "                )\n",
    "formatFigure_pca(plotname, fH, ax, pH)\n",
    "ax.set_xlabel('PC1 (var = 13%)', fontproperties=font_manager.FontProperties(fname='/System/Library/Fonts/Avenir.ttc', size=22))\n",
    "ax.set_ylabel('PC2 (var = 10%)', fontproperties=font_manager.FontProperties(fname='/System/Library/Fonts/Avenir.ttc', size=22))\n",
    "plt.xlim(-80,80)\n",
    "plt.ylim(-60,60)\n",
    "plt.tight_layout()\n",
    "plt.savefig(\"{0}.svg\".format(plotname), transparent=True, format=\"svg\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "#pyplot version (Sep_2019)\n",
    "plotname = 'PCA_ML'\n",
    "pc_df = pd.read_csv(\"~/Documents/LiMolec/zfRNAseq/20190827/20190827_DESeq2/00_LvsM/pcaData.csv\")\n",
    "pc1 = pc_df.iloc[:,0]\n",
    "pc2 = pc_df.iloc[:,1]\n",
    "fH, ax = plt.subplots(figsize=[5,5])\n",
    "pH = plt.scatter(pc1, pc2, color=[\n",
    "    pC['m'],pC['m'],pC['m'],pC['m'],pC['m'],pC['m'],pC['m'],\n",
    "    pC['l'],pC['l'],pC['l'],pC['l'],pC['l'],pC['l']\n",
    "]\n",
    "                )\n",
    "formatFigure_pca(plotname, fH, ax, pH)\n",
    "ax.set_xlabel('PC1 (var = 23%)', fontproperties=font_manager.FontProperties(fname='/System/Library/Fonts/Avenir.ttc', size=22))\n",
    "ax.set_ylabel('PC2 (var = 10%)', fontproperties=font_manager.FontProperties(fname='/System/Library/Fonts/Avenir.ttc', size=22))\n",
    "plt.xlim(-80,80)\n",
    "plt.ylim(-80,80)\n",
    "plt.tight_layout()\n",
    "plt.savefig(\"{0}.svg\".format(plotname), transparent=True, format=\"svg\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "#pyplot version (Sep_2019)\n",
    "plotname = 'PCA_U_S'\n",
    "pc_df = pd.read_csv(\"~/Documents/LiMolec/zfRNAseq/20190827/20190827_DESeq2/00_UvsS/pcaData.csv\")\n",
    "pc1 = pc_df.iloc[:,0]\n",
    "pc2 = pc_df.iloc[:,1]\n",
    "fH, ax = plt.subplots(figsize=[5,5])\n",
    "pH = plt.scatter(pc1, pc2, color=[\n",
    "    pC['u'],pC['u'],pC['u'],pC['u'],pC['u'],\n",
    "    pC['s'],pC['s'],pC['s'],pC['s'],pC['s'],pC['s'],\n",
    "]\n",
    "                )\n",
    "formatFigure_pca(plotname, fH, ax, pH)\n",
    "ax.set_xlabel('PC1 (var = 13%)', fontproperties=font_manager.FontProperties(fname='/System/Library/Fonts/Avenir.ttc', size=22))\n",
    "ax.set_ylabel('PC2 (var = 10%)', fontproperties=font_manager.FontProperties(fname='/System/Library/Fonts/Avenir.ttc', size=22))\n",
    "plt.xlim(-100,100)\n",
    "plt.ylim(-80,80)\n",
    "plt.tight_layout()\n",
    "plt.savefig(\"{0}.svg\".format(plotname), transparent=True, format=\"svg\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# US vs LM: PC1 = 13% and PC2 = 10%\n",
    "# M vs L: PC1 = 23% and PC2 = 10%\n",
    "# UV vs S: PC1 = 15% and PC2 = 13%"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### tSNE plot"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "#pyplot version (Sep_2019)\n",
    "plotname = 'tSNE'\n",
    "pc_df = pd.read_csv(\"~/Documents/LiMolec/zfRNAseq/20190827/20190827_DESeq2/tSNEData.csv\")\n",
    "pc1 = pc_df.iloc[:,0]\n",
    "pc2 = pc_df.iloc[:,1]\n",
    "fH, ax = plt.subplots(figsize=[6,6])\n",
    "pH = plt.scatter(pc1, pc2, color=pC['plt'])\n",
    "formatFigure_pca(plotname, fH, ax, pH)\n",
    "ax.set_xlabel('tSNE1', fontproperties=font_manager.FontProperties(fname='/System/Library/Fonts/Avenir.ttc', size=22))\n",
    "ax.set_ylabel('tSNE2', fontproperties=font_manager.FontProperties(fname='/System/Library/Fonts/Avenir.ttc', size=22))\n",
    "ax.set_xticks([-1000,-500,0,500,1000])\n",
    "ax.set_yticks([-1000,-500,0,500,1000])\n",
    "plt.xlim(-1000,1000)\n",
    "plt.ylim(-1000,1000)\n",
    "plt.tight_layout()\n",
    "plt.savefig(\"{0}.svg\".format(plotname), transparent=True, format=\"svg\")"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "***\n",
    "## barPlot: Opsins<a id='barOpsins'>∮</a>\n",
    "***\n",
    "[Back to Index](#Index)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# from fx_RNAseq import *"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "genename = 'opn1lw1'\n",
    "n,h = plotBars(genename)\n",
    "h = h\n",
    "fH, ax = plt.subplots()\n",
    "pH = plt.bar(n, h, width=0.8, bottom=None, align='center', data=None, color=pC['plt'])\n",
    "formatFigure_Opsins(genename, fH, ax, pH)\n",
    "plt.savefig(\"{0}.svg\".format(genename), transparent=True, format=\"svg\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "findgene('rho')"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "***\n",
    "## barPlot: Custom Gene<a id='barCustom'>∮</a>\n",
    "***\n",
    "[Back to Index](#Index)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "> Genes to pay attention to but still haven't given interesting results\n",
    "> cdh (cadherins)\n",
    "> protocadherins\n",
    "> semaphorins\n",
    "> plexins (plx)\n",
    "> laminins (lamb) and integrins (itg)\n",
    "> dystroglycan (dag), dystrophin (dmd), pikachurin\n",
    "> neurexins\n",
    "> cadm\n",
    "> dscam and sidekicks (sdk)\n",
    "> pappaa\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "gdf = findgenelist('lrit')\n",
    "gdf"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "genename = 'prickle3'\n",
    "# genename = 'ncdn'\n",
    "# genename = 'chn2'\n",
    "# genename = 'grin1b'\n",
    "# genename = 'cdhr1a'\n",
    "# genename = 'lrit1a'\n",
    "n,h = plotBars(genename)\n",
    "fH, ax = plt.subplots()\n",
    "pH = plt.bar(n, h, width=0.8, bottom=None, align='center', data=None, color=pC['plt'])\n",
    "formatFigure(genename, fH, ax, pH)\n",
    "plt.savefig(\"{0}.svg\".format(genename), transparent=True, format=\"svg\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "***\n",
    "## barPlot: Many Genes Layout<a id='barMany'>∮</a>\n",
    "***\n",
    "[Back to Index](#Index)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "cd '~/Dropbox/LiLab/RNAseq/genePlots/combos'"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "genelist = ['gnat1','gnat2']\n",
    "n, h = plotBars_list(genelist)\n",
    "fH, ax = plt.subplots()\n",
    "pH = plt.bar(n, h, width=0.8, bottom=None, align='center', data=None, color=pC['plt'])\n",
    "formatFigure_list(genelist, fH, ax, pH)\n",
    "# plt.savefig(\"{0}.svg\".format('gnat'), transparent=True, format=\"svg\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "genelist = ['rho','opn1sw1','opn1sw2','opn1mw1','opn1mw2','opn1mw3','opn1mw4','opn1lw1','opn1lw2']\n",
    "n, h = plotBars_list(genelist)\n",
    "fH, ax = plt.subplots(figsize=[6*len(genelist),4])\n",
    "pH = plt.bar(n, h, width=0.8, bottom=None, align='center', data=None, color=pC['plt'])\n",
    "formatFigure_list(genelist, fH, ax, pH)\n",
    "plt.xlim(0,len(genelist)*10)\n",
    "# plt.savefig(\"{0}.svg\".format('opn1'), transparent=True, format=\"svg\")"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "***\n",
    "## barPlot: Many Genes Layout Using Subplots (proper way)<a id='barManySubPlots'>∮</a>\n",
    "***\n",
    "[Back to Index](#Index)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "shrink = 1#2/3"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "genelist = ['gnat1','gnat2']\n",
    "fH, axes = plt.subplots(1,len(genelist),sharey=True,figsize=[6*shrink*len(genelist),4*shrink])\n",
    "for i,g in enumerate(genelist,start=0):\n",
    "    n, h = plotBars(g)\n",
    "    plt.sca(axes[i]) \n",
    "    pH = plt.bar(n, h, width=0.8, bottom=None, align='center', data=None,\n",
    "                 color=pC['plt'])\n",
    "    if  i == 0:\n",
    "        formatFigure(g, fH, axes[i], pH)\n",
    "    else:\n",
    "        formatFigure_right(g, fH, axes[i], pH)\n",
    "plt.savefig(\"{0}.svg\".format('gnat'), transparent=True, format=\"svg\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "genelist = ['rho','rhol']\n",
    "\n",
    "fH, axes = plt.subplots(1,len(genelist),sharey=True,figsize=[6*shrink*len(genelist),4*shrink])\n",
    "for i,g in enumerate(genelist,start=0):\n",
    "    n, h = plotBars(g)\n",
    "    plt.sca(axes[i])\n",
    "    pH = plt.bar(n, h, width=0.8, bottom=None, align='center', data=None,\n",
    "                 color=pC['plt'])\n",
    "    if  i == 0:\n",
    "        formatFigure(g, fH, axes[i], pH)\n",
    "#         plt.ylim(0,3.5e6)\n",
    "#         axes[0].set_yticks(np.arange(0, 4.0e6, step=0.5e6))\n",
    "    else:\n",
    "        formatFigure_right(g, fH, axes[i], pH)\n",
    "plt.savefig(\"{0}.svg\".format('rho'), transparent=True, format=\"svg\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "genelist = ['opn1sw1','opn1sw1']\n",
    "\n",
    "fH, axes = plt.subplots(1,len(genelist),sharey=True,figsize=[6*shrink*len(genelist),4*shrink])\n",
    "for i,g in enumerate(genelist,start=0):\n",
    "    n, h = plotBars(g)\n",
    "    plt.sca(axes[i])\n",
    "    pH = plt.bar(n, h, width=0.8, bottom=None, align='center', data=None,\n",
    "                 color=pC['plt'])\n",
    "    if  i == 0:\n",
    "        formatFigure(g, fH, axes[i], pH)\n",
    "#         plt.ylim(0,2e6)\n",
    "#         axes[0].set_yticks(np.arange(0, 2.5e6, step=0.5e6))\n",
    "    else:\n",
    "        formatFigure_right(g, fH, axes[i], pH)\n",
    "plt.savefig(\"{0}.svg\".format('opn1sw1'), transparent=True, format=\"svg\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "genelist = ['opn1mw1','opn1mw2']\n",
    "\n",
    "fH, axes = plt.subplots(1,len(genelist),sharey=True,figsize=[6*shrink*len(genelist),4*shrink])\n",
    "for i,g in enumerate(genelist,start=0):\n",
    "    n, h = plotBars(g)\n",
    "    plt.sca(axes[i])\n",
    "    pH = plt.bar(n, h, width=0.8, bottom=None, align='center', data=None,\n",
    "                 color=pC['plt'])\n",
    "    if  i == 0:\n",
    "        formatFigure(g, fH, axes[i], pH)\n",
    "    else:\n",
    "        formatFigure_right(g, fH, axes[i], pH)\n",
    "plt.savefig(\"{0}.svg\".format('opn1mwA'), transparent=True, format=\"svg\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "genelist = ['opn1mw3','opn1mw4']\n",
    "\n",
    "fH, axes = plt.subplots(1,len(genelist),sharey=True,figsize=[6*shrink*len(genelist),4*shrink])\n",
    "for i,g in enumerate(genelist,start=0):\n",
    "    n, h = plotBars(g)\n",
    "    plt.sca(axes[i])\n",
    "    pH = plt.bar(n, h, width=0.8, bottom=None, align='center', data=None,\n",
    "                 color=pC['plt'])\n",
    "    if  i == 0:\n",
    "        formatFigure(g, fH, axes[i], pH)\n",
    "        plt.ylim(0,1.0e5)\n",
    "        axes[0].set_yticks(np.arange(0, 1.5e5, step=0.5e5))\n",
    "    else:\n",
    "        formatFigure_right(g, fH, axes[i], pH)\n",
    "\n",
    "\n",
    "plt.savefig(\"{0}.svg\".format('opn1mwB'), transparent=True, format=\"svg\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "genelist = ['opn1lw1','opn1lw2']\n",
    "\n",
    "fH, axes = plt.subplots(1,len(genelist),sharey=True,figsize=[6*shrink*len(genelist),4*shrink])\n",
    "for i,g in enumerate(genelist,start=0):\n",
    "    n, h = plotBars(g)\n",
    "    plt.sca(axes[i])\n",
    "    pH = plt.bar(n, h, width=0.8, bottom=None, align='center', data=None,\n",
    "                 color=pC['plt'])\n",
    "    if  i == 0:\n",
    "        formatFigure(g, fH, axes[i], pH)\n",
    "        plt.ylim(0,1.0e5)\n",
    "        axes[0].set_yticks(np.arange(0, 1.5e5, step=0.5e5))\n",
    "    else:\n",
    "        formatFigure_right(g, fH, axes[i], pH)\n",
    "#         axes[i].set_yticks(np.arange(0, 2.0e6, step=0.5e6))\n",
    "\n",
    "plt.savefig(\"{0}.svg\".format('opn1lw'), transparent=True, format=\"svg\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Using svgutils to automatically tile these figures the right way\n",
    "Figure(\"2400\", \"2000\",\n",
    "       SVG(\"rho.svg\").move(0, 290*0),\n",
    "       SVG(\"opn1sw.svg\").move(0, 290*1),\n",
    "       SVG(\"opn1mwA.svg\").move(0, 290*2),\n",
    "       SVG(\"opn1mwB.svg\").move(0, 290*3),\n",
    "       SVG(\"opn1lw.svg\").move(0, 290*4)\n",
    "        ).save(\"OpsinsAll.svg\")\n",
    "svgreload()"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Phototransduction"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "gdf = findgenelist('gngt')\n",
    "gdf = gdf.sort_values(by=['symbol'])\n",
    "gdf"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# genelist = ['saga','sagb','arr3a','arr3b']\n",
    "# plotName = 'arrestins'\n",
    "\n",
    "# genelist = ['gnat1','gnat2']\n",
    "# plotName = 'gnat'\n",
    "\n",
    "# genelist = ['gngt1','gngt2a','gngt2b']\n",
    "# plotName = 'gngt'\n",
    "\n",
    "# genelist = ['gnb1a','gnb1b','gnb3a','gnb3b','gnb4b','gnb5a','gnb5b']\n",
    "# plotName = 'gnb'\n",
    "\n",
    "# genelist = ['pde6a','pde6b','pde6c','pde6d','pde6ga','pde6gb','pde6c','pde6ha']\n",
    "# plotName = 'pde'\n",
    "\n",
    "# genelist = ['gucy1a1','gucy1b1','gucy1b2','gucy2c','gucy2d','gucy2f','gucy2g']\n",
    "# plotName = 'GC'\n",
    "\n",
    "# genelist = ['gucy2d','gucy2f']\n",
    "# plotName = 'GC_retinaSpecific'\n",
    "\n",
    "# genelist = ['guca1a','guca1b','guca1c','guca1d','guca1e','guca1g']\n",
    "# plotName = 'GCAP'\n",
    "\n",
    "# genelist = ['cnga1a','cnga1b','cnga3a','cnga3b','cnga4','cnga2a','cnga2b']\n",
    "# plotName = 'CNGalpha'\n",
    "\n",
    "# genelist = ['cnga1a','cnga1b','cnga3a','cnga3b']\n",
    "# plotName = 'CNGalpha_retinaSpecific'\n",
    "\n",
    "# genelist = ['cngb1a','cngb3.1','cngb3.2']\n",
    "# plotName = 'CNGbeta'\n",
    "\n",
    "\n",
    "# genelist = ['rgs9a','rgs9b','rgs9bp',\n",
    "#             'rgs3a','rgs6',\n",
    "#             'rgs11','rgs12a','rgs16','rgs20']\n",
    "# plotName = 'RGS'\n",
    "\n",
    "# genelist = ['grk1a','grk1b',\n",
    "#             'grk3',\n",
    "#             'grk4','grk5','grk5l',\n",
    "#             'grk6',\n",
    "#             'grk7a','grk7b']\n",
    "# plotName = 'GRK'\n",
    "\n",
    "# genelist = ['rcvrna','rcvrnb','rcvrn2','rcvrn3','ncalda','ncaldb']\n",
    "# plotName = 'rcvrn'\n",
    "\n",
    "\n",
    "\n",
    "fH, axes = plt.subplots(1,len(genelist),sharey=True,figsize=[6*len(genelist),4])\n",
    "for i,g in enumerate(genelist,start=0):\n",
    "    n, h = plotBars(g)\n",
    "    plt.sca(axes[i])\n",
    "    pH = plt.bar(n, h, width=0.8, bottom=None, align='center', data=None,\n",
    "                 color=pC['plt'])\n",
    "    if  i == 0:\n",
    "        formatFigure(g, fH, axes[i], pH)\n",
    "    else:\n",
    "        formatFigure_right(g, fH, axes[i], pH)\n",
    "plt.savefig(\"{0}.svg\".format(plotName), transparent=True, format=\"svg\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Others"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "gdf = findgenelist('cacn')\n",
    "gdf = gdf.sort_values(by=['symbol'])\n",
    "gdf"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "genelist = ['ctbp1','ctbp2a']\n",
    "plotName = 'ctbp'\n",
    "\n",
    "genelist = ['cadm1a','cadm1b','cadm2a','cadm2b','cadm3','cadm4']\n",
    "plotName = 'cadm'\n",
    "\n",
    "genelist = ['tjp1a','tjp1b','tjp2a','tjp2b','tjp3']\n",
    "plotName = 'tjp'\n",
    "\n",
    "genelist = ['nrxn1a','nrxn1b','nrxn2a','nrxn2b','nrxn3a','nrxn3b',]\n",
    "plotName = 'nrxn'\n",
    "\n",
    "genelist = ['eml1','eml2','eml3']\n",
    "plotName = 'eml'\n",
    "\n",
    "genelist = ['elfn1b','elfn2a']\n",
    "plotName = 'elfn'\n",
    "\n",
    "genelist = ['cpne1','cpne2','cpne3','cpne4a','cpne4b','cpne5a','cpne5b','cpne7','cpne8','cpne9']\n",
    "plotName = 'cpne'\n",
    "\n",
    "# genelist = ['tbx2a','tbx2b','tbxas1','tbx21']\n",
    "genelist = ['tbx2a','tbx2b']\n",
    "plotName = 'tbx2'\n",
    "\n",
    "genelist = ['chn1','chn2']\n",
    "plotName = 'chn'\n",
    "\n",
    "genelist = ['ncdn','ncdn']\n",
    "plotName = 'ncdn'\n",
    "\n",
    "genelist = ['sema7a','sema6a','sema6d','sema4ab','sema4ba','sema3fa','sema3fb']\n",
    "plotName = 'sema'\n",
    "\n",
    "genelist = ['cacna1aa','cacna1ab',\n",
    "            'cacna1da','cacna1db',\n",
    "            'cacna1fa','cacna1fb',\n",
    "            'cacna2d1a','cacna2d2a',\n",
    "            'cacna2d4a','cacna2d4b',\n",
    "            'cacna1sb',\n",
    "           ]\n",
    "plotName = 'CaV'\n",
    "\n",
    "# genelist = ['stxbp1b','stxbp4','stxbp2','stxbp5a','stxbp5b','stxbp5l','stxbp3']\n",
    "genelist = ['stxbp1a','stxbp1b',\n",
    "            'stxbp2','stxbp3','stxbp4',\n",
    "            'stxbp5a','stxbp5b','stxbp5l',\n",
    "            'stxbp6','stxbp6l']\n",
    "plotName = 'stxBP'\n",
    "\n",
    "genelist = ['stx1a','stx1b','stx2a','stx2b','stx3a',\n",
    "            'stx4','stx5a','stx5al','stx6','stx7l',\n",
    "            'stx8','stx10','stx11a','stx11b.1','stx11b.2',\n",
    "            'stx12','stx12l','stx16','stx17','stx18']\n",
    "plotName = 'stx'\n",
    "\n",
    "genelist = ['prickle1a','prickle1b','prickle2a','prickle2b','prickle3',]\n",
    "plotName = 'prickle'\n",
    "\n",
    "genelist = ['pcdh15a','pcdh15b','pcdh8','pcdh11','pcdh10a']\n",
    "plotName = 'pdch'\n",
    "\n",
    "genelist = ['chn1','chn2']\n",
    "plotName = 'chn'\n",
    "\n",
    "\n",
    "genelist = ['lrit1a','lrit1b','lrit2','lrit3a','lrit3b']\n",
    "plotName = 'lrit'\n",
    "\n",
    "genelist = ['pou4f1','pou4f2','pou4f3','rbpms','grm6a','grm6b','vsx2','gfap','prkca','syt2a','rpe65a', 'opn4']\n",
    "plotName = 'controls'\n",
    "\n",
    "\n",
    "# genelist = ['gpx1a','gpx1b','gpx4a','gpx4b']\n",
    "# genelist = ['rgs9bp','RGS9BP']\n",
    "# genelist = ['chchd1','chchd10']\n",
    "# genelist = ['yipf6','yipf1']\n",
    "# genelist = ['sema3fb','sema3fa','sema6d']\n",
    "# genelist = ['esama','esamb']\n",
    "\n",
    "# genelist = ['pcdh1gc6','tex264a']\n",
    "# plotName = 'test'\n",
    "\n",
    "            \n",
    "fH, axes = plt.subplots(1,len(genelist),sharey=False,figsize=[6*len(genelist),4])\n",
    "for i,g in enumerate(genelist,start=0):\n",
    "    n, h = plotBars(g)\n",
    "    plt.sca(axes[i])\n",
    "    pH = plt.bar(n, h, width=0.8, bottom=None, align='center', data=None,\n",
    "                 color=pC['plt'])\n",
    "    if  i == 0:\n",
    "        formatFigure(g, fH, axes[i], pH)\n",
    "    else:\n",
    "        formatFigure_right(g, fH, axes[i], pH)\n",
    "plt.savefig(\"{0}.svg\".format(plotName), transparent=True, format=\"svg\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "gdf = findgenelist('opn1')\n",
    "# gdf = gdf.sort_values(by=['symbol'])\n",
    "# gdf = gdf.sort_values(by=['baseMean'], ascending=False)\n",
    "# gdf['symbol']\n",
    "gdf"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### According to Neuhauss recoverin1a in enriched in rods and UV-cones"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "***\n",
    "## heatMap: Opsins<a id='barManySubPlots'>∮</a>\n",
    "***\n",
    "[Back to Index](#Index)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Not sure this is helpful.  \n",
    "Maybe use this to show differential expression by normalizing by mean expression in correct type"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "gdf = findgenelist('opn1')\n",
    "gdf2 = findgene('rho')\n",
    "gdf2\n",
    "gdf = gdf.append(gdf2)\n",
    "gdf.index = range(len(gdf))\n",
    "gdf\n",
    "\n",
    "Opsins = [\"Rhodopsin (rho)\",\"UV-opsin (opn1sw1)\", \"S-opsin (opn1sw2)\",\"M-opsin (opn1mw1)\",\"M-opsin (opn1mw2)\",\"M-opsin (opn1mw3)\",\"M-opsin (opn1mw4)\",\"L-opsin (opn1lw1)\",\"L-opsin (opn1lw2)\"]\n",
    "Samples = [\"\",\"\",\"R\",\"\",\"\",\"\",\n",
    "           \"\",\"\",\"U\",\"\",\"\",\n",
    "           \"\",\"\",\"S\",\"\",\"\",\"\",\n",
    "           \"\",\"\",\"\",\"M\",\"\",\"\",\"\",\n",
    "           \"\",\"\",\"L\",\"\",\"\",\"\",\n",
    "          ]\n",
    "Samples = [\n",
    "          ]\n",
    "\n",
    "data = gdf.iloc[0:9,7:37]\n",
    "data = data.reindex([8,4,5,7,0,2,6,1,3])\n",
    "data.index = range(len(data))\n",
    "# np.around(np.divide(data.values,1e6),2)\n",
    "# opsinData = np.divide(data.values,1e6)\n",
    "opsinData = np.divide(data.values,1)\n",
    "\n",
    "# # log2 representation does not work well\n",
    "# log2matrix = np.vectorize(log2)\n",
    "# opsinData = log2matrix(opsinData)\n",
    "\n",
    "params = {\"ytick.color\" : \"w\",\n",
    "          \"xtick.color\" : \"w\",\n",
    "          \"axes.labelcolor\" : \"w\",\n",
    "          \"axes.edgecolor\" : \"w\",\n",
    "         \"axes.linewidth\" : 2,\n",
    "         \"xtick.major.width\" : 2,\n",
    "         \"ytick.major.width\" : 2,\n",
    "         \"xtick.major.size\" : 0,\n",
    "         \"ytick.major.size\" : 0}\n",
    "plt.rcParams.update(params)\n",
    "\n",
    "plotName = \"heatMap_opsins\"\n",
    "fig, ax = plt.subplots()\n",
    "\n",
    "im, cbar = heatmap(opsinData, Opsins, Samples, ax=ax,\n",
    "                   cmap=\"inferno\", cbarlabel=\"fpkm\")\n",
    "\n",
    "\n",
    "# texts = annotate_heatmap(im, valfmt=\"{x:.1f} t\")\n",
    "\n",
    "# fig.tight_layout()\n",
    "plt.savefig(\"{0}.svg\".format(plotName), transparent=True, format=\"svg\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "***\n",
    "## heatMap: Custom<a id='heatCustom'>∮</a>\n",
    "***\n",
    "[Back to Index](#Index)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 33,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>symbol</th>\n",
       "      <th>genename</th>\n",
       "      <th>baseMean</th>\n",
       "      <th>log2FoldChange</th>\n",
       "      <th>lfcSE</th>\n",
       "      <th>pvalue</th>\n",
       "      <th>padj</th>\n",
       "      <th>R1</th>\n",
       "      <th>R2</th>\n",
       "      <th>R3</th>\n",
       "      <th>...</th>\n",
       "      <th>M4</th>\n",
       "      <th>M5</th>\n",
       "      <th>M6</th>\n",
       "      <th>S7</th>\n",
       "      <th>L1</th>\n",
       "      <th>L2</th>\n",
       "      <th>L3</th>\n",
       "      <th>L4</th>\n",
       "      <th>L5</th>\n",
       "      <th>L6</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <td>0</td>\n",
       "      <td>kif9</td>\n",
       "      <td>kinesin family member 9</td>\n",
       "      <td>0.0</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>...</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>1 rows × 37 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "  symbol                 genename  baseMean  log2FoldChange  lfcSE  pvalue  \\\n",
       "0   kif9  kinesin family member 9       0.0             NaN    NaN     NaN   \n",
       "\n",
       "   padj   R1   R2   R3  ...   M4   M5   M6   S7   L1   L2   L3   L4   L5   L6  \n",
       "0   NaN  0.0  0.0  0.0  ...  0.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0  \n",
       "\n",
       "[1 rows x 37 columns]"
      ]
     },
     "execution_count": 33,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "plotName = \"test\"\n",
    "\n",
    "# gdf = findgene('rho')\n",
    "# gdf2 = findgenelist('opn1sw')\n",
    "# gdf = gdf.append(gdf2)\n",
    "# gdf2 = findgenelist('opn1mw')\n",
    "# gdf2 = gdf2.sort_values(by=[\"symbol\"])\n",
    "# gdf = gdf.append(gdf2)\n",
    "# gdf2 = findgenelist('opn1lw')\n",
    "# gdf2 = gdf2.sort_values(by=[\"symbol\"])\n",
    "# gdf = gdf.append(gdf2)\n",
    "\n",
    "# gdf.index = range(len(gdf))\n",
    "# plotName = \"heatMap_opsins\"\n",
    "\n",
    "# gdf = findgenelist('slc24') -> NCKX transporter; slc25a5 is mitochondrial but known to express in eye\n",
    "# plotName = \"slc24\"\n",
    "\n",
    "# gdf = findgenelist('sema')\n",
    "# gdf = gdf.sort_values(by=[\"symbol\"])\n",
    "# plotName = \"heatMap_sema\"\n",
    "\n",
    "# gdf = findgenelist('cdh')\n",
    "# gdf = gdf.sort_values(by=[\"baseMean\"],ascending=False)\n",
    "# plotName = \"heatMap_cdh\"\n",
    "\n",
    "# gdf = findgenelist('eph')\n",
    "# gdf = gdf.drop([1,14,15,16]) #not ephrin receptors\n",
    "# gdf = gdf.sort_values(by=[\"symbol\"])\n",
    "# plotName = \"heatMap_eph\"\n",
    "\n",
    "\n",
    "# gdf = findgenelist('efn')\n",
    "# gdf = gdf.sort_values(by=[\"symbol\"])\n",
    "# plotName = \"heatMap_efn\"\n",
    "\n",
    "# gdf = findgenelist('pdgf')\n",
    "# gdf = findgenelist('rgr')\n",
    "# gdf = findgenelist('lmo')\n",
    "# gdf.index = range(len(gdf))\n",
    "# gdf.sort_values(by=[\"log2FoldChange\"])\n",
    "\n",
    "\n",
    "## genes in 2019 Papal, Swaroop: # igsf11, epha10, tuft1, prtg, lrfn2, vasn\n",
    "\n",
    "# gdf = findgenelist('igsf')\n",
    "# gdf = findgenelist('tuft')\n",
    "# gdf = findgenelist('prtg')\n",
    "# gdf = findgenelist('lrfn')\n",
    "# gdf = findgenelist('vasn')\n",
    "\n",
    "# gdf = findgenelist('rdh')\n",
    "# gdf = gdf.sort_values(by=[\"symbol\"])\n",
    "# plotName = \"rdhHeatMap\"\n",
    "\n",
    "# gdf = findgenelist('cyp2')\n",
    "# gdf = findgenelist('cyb')\n",
    "\n",
    "# gdf = findgenelist('sag')\n",
    "# plotName = \"heatMap_sag\"\n",
    "\n",
    "# gdf = findgenelist('dio')\n",
    "# gdf = gdf.sort_values(by=[\"symbol\"])\n",
    "# plotName = \"heatMap_dio\"\n",
    "\n",
    "# gdf = findgenelist('lrrn')\n",
    "# gdf = gdf.sort_values(by=[\"symbol\"])\n",
    "# plotName = \"heatMap_lrrn\"\n",
    "\n",
    "# gdf = findgenelist('dnmt')\n",
    "# gdf = gdf.sort_values(by=[\"symbol\"])\n",
    "# plotName = \"heatMap_dnmt\"\n",
    "\n",
    "# gdf = findgenelist('tet')\n",
    "\n",
    "# gdf = findgenelist('isl')\n",
    "# gdf = gdf.drop([1])\n",
    "\n",
    "# gdf = findgenelist('gj')\n",
    "# gdf = gdf.sort_values(by=[\"symbol\"])\n",
    "# plotName = \"heatMap_gj\"\n",
    "\n",
    "# gdf = findgenelist('fgf')\n",
    "# gdf = findgenelist('bmp')\n",
    "# gdf = findgenelist('fzd')\n",
    "# gdf = findgenelist('edar')\n",
    "# gdf = findgenelist('axin')\n",
    "\n",
    "# gdf = findgenelist('cadm') # cadm2a in S-Cones probably(?) and cadm3 in all photoreceptors (Hunter, JCN, 2011)\n",
    "# gdf = gdf.sort_values(by=[\"symbol\"])\n",
    "# plotName = \"heatMap_cadm\"\n",
    "\n",
    "# gdf = findgenelist('lrit') # lrit1b for rods and cones; lrit2 for cones; lrit3 for rods\n",
    "# gdf = gdf.sort_values(by=[\"symbol\"])\n",
    "# plotName = \"heatMap_lrit\"\n",
    "\n",
    "# gdf = findgenelist('elfn') # not detected\n",
    "# gdf = gdf.sort_values(by=[\"symbol\"])\n",
    "# plotName = \"heatMap_elfn\"\n",
    "\n",
    "# gdf = findgenelist('tenm') # for RGCs in zebrafish (Antinucci, Cell Reports, 2013)\n",
    "\n",
    "# gdf = findgenelist('foxo')\n",
    "\n",
    "# gdf = findgenelist('sdk')\n",
    "\n",
    "# gdf = findgenelist('crmp')\n",
    "\n",
    "\n",
    "\n",
    "# Even More\n",
    "# gdf = findgenelist('thr')\n",
    "# gdf = gdf.sort_values(by=[\"symbol\"])\n",
    "# plotName = \"heatMap_thr\"\n",
    "\n",
    "# gdf = findgenelist('lingo')\n",
    "# gdf = gdf.sort_values(by=[\"symbol\"])\n",
    "# plotName = \"heatMap_lingo\"\n",
    "\n",
    "# gdf = findgenelist('mitf')\n",
    "\n",
    "# gdf = findgenelist('prickle')\n",
    "# gdf = gdf.sort_values(by=[\"symbol\"])\n",
    "# plotName = \"heatMap_prickle\"\n",
    "\n",
    "# gdf = findgenelist('rdh')\n",
    "# gdf = findgenelist('sdh')\n",
    "# lingo, rx, mitf\n",
    "\n",
    "# gdf = findgenelist('dag')\n",
    "# can't find pikachurin: # gdf = findgenelist('egflam') gdf = findgenelist('si:dkeyp-100d11.2-001')\n",
    "\n",
    "# gdf = findgenelist('kcn')\n",
    "# gdf = gdf.sort_values(by=[\"baseMean\"], ascending=False)\n",
    "# gdf = gdf.drop(np.arange(20,109))\n",
    "# gdf\n",
    "\n",
    "# gdf = findgenelist('kcnj')\n",
    "# gdf = gdf.sort_values(by=[\"baseMean\"], ascending=False)\n",
    "# gdf\n",
    "\n",
    "# gdf = findgenelist('slc')\n",
    "# gdf = gdf.sort_values(by=[\"baseMean\"], ascending=False)\n",
    "# gdf = gdf.drop(np.arange(40,410))\n",
    "\n",
    "# gdf = findgenelist('rcv')\n",
    "# gdf = gdf.sort_values(by=[\"symbol\"])\n",
    "# plotName = \"heatMap_recoverin\"\n",
    "\n",
    "# gdf = findgenelist('sag')\n",
    "# gdf = gdf.append(findgenelist('arrb'))\n",
    "# gdf = gdf.append(findgenelist('arr3'))\n",
    "# gdf = gdf.sort_values(by=[\"symbol\"])\n",
    "# plotName = \"heatMap_arr\"\n",
    "\n",
    "\n",
    "# gdf = findgenelist('cng')\n",
    "# gdf = gdf.sort_values(by=[\"symbol\"])\n",
    "# plotName = \"heatMap_cng\"\n",
    "\n",
    "# gdf = findgenelist('cnga')\n",
    "# gdf = gdf.sort_values(by=[\"symbol\"])\n",
    "# plotName = \"heatMap_cnga\"\n",
    "\n",
    "# gdf = findgenelist('cngb')\n",
    "# gdf = gdf.sort_values(by=[\"symbol\"])\n",
    "# plotName = \"heatMap_cngb\"\n",
    "\n",
    "# gdf = findgenelist('myo')\n",
    "# gdf = gdf.sort_values(by=[\"symbol\"])\n",
    "\n",
    "# gdf = findgenelist('crx')\n",
    "# gdf = gdf.sort_values(by=[\"baseMean\"], ascending=False)\n",
    "# plotName = \"heatMap_crx\"\n",
    "\n",
    "\n",
    "# gdf = findgenelist('neurod')\n",
    "# gdf = gdf.sort_values(by=[\"symbol\"])\n",
    "# plotName = \"heatMap_neurod\"\n",
    "\n",
    "# gdf = findgenelist('ror')\n",
    "# gdf = gdf.drop([5,6])\n",
    "# gdf = gdf.sort_values(by=[\"symbol\"])\n",
    "# plotName = \"heatMap_ror\"\n",
    "\n",
    "\n",
    "# gdf = findgenelist('nr')\n",
    "# gdf = gdf.sort_values(by=[\"baseMean\"], ascending=False)\n",
    "# gdf = gdf.drop(np.arange(5,76))\n",
    "# plotName = \"heatMap_nrl\"\n",
    "\n",
    "\n",
    "# gdf = findgenelist('pcdh')\n",
    "# gdf = gdf.drop([1]) #highest ones that obscure others.\n",
    "# # gdf = gdf.drop([1,15]) #highest ones that obscure others.\n",
    "# gdf = gdf[gdf.baseMean > .45]\n",
    "# gdf = gdf.sort_values(by=[\"baseMean\"], ascending=False)\n",
    "# plotName = \"heatMap_pcdh_partial\"\n",
    "\n",
    "# gdf = findgenelist('hcn')\n",
    "# gdf = gdf.sort_values(by=[\"symbol\"])\n",
    "\n",
    "# gdf = findgenelist('gabr')\n",
    "# gdf = gdf.drop([0,2]) #highest ones that obscure others.\n",
    "# gdf = gdf.sort_values(by=[\"symbol\"])\n",
    "\n",
    "# gdf = findgenelist('opn4')\n",
    "# gdf = gdf.drop([1,2]) #highest ones that obscure others.\n",
    "# gdf = gdf.sort_values(by=[\"baseMean\"], ascending=False)\n",
    "# gdf = gdf.drop(np.arange(5,76))\n",
    "# gdf = gdf.sort_values(by=[\"symbol\"])\n",
    "# plotName = \"heatMap_opn4\"\n",
    "\n",
    "# CSNB causes: Nyx, cacna1f, GPR179 (Kamermans, 2019)\n",
    "# gdf = findgenelist('nyx') # bipolar cell -> not expressed in photoreceptors\n",
    "# gdf = findgenelist('gpr1') # bipolar cell -> not expressed in photoreceptors\n",
    "# gdf = findgenelist('cacna') # bipolar cell -> not expressed in photoreceptors\n",
    "\n",
    "\n",
    "# For Karen Carleton\n",
    "\n",
    "# gdf = findgenelist('rx')\n",
    "# gdf = gdf.drop([2,3,4,5,7,8,9,10,11,12,13,14,15,16,17,18,19])\n",
    "# gdf = gdf.sort_values(by=[\"symbol\"])\n",
    "# # gdf = gdf.drop([1])\n",
    "# plotName = \"heatMap_rx\"\n",
    "\n",
    "# gdf = findgenelist('tbx')\n",
    "# gdf = gdf.drop([3,17])\n",
    "# plotName = \"heatMap_tbx\"\n",
    "# gdf = gdf.drop([1,2])\n",
    "\n",
    "# gdf = findgenelist('tbx2')\n",
    "# gdf = gdf.drop([0,3,4])\n",
    "# gdf = gdf.sort_values(by=[\"symbol\"])\n",
    "# plotName = \"heatMap_tbx2\"\n",
    "\n",
    "# gdf = findgenelist('six')\n",
    "# gdf = gdf.sort_values(by=[\"symbol\"])\n",
    "# plotName = \"heatMap_six\"\n",
    "\n",
    "# gdf = findgenelist('lin')\n",
    "# gdf = gdf.drop([1,3,4,5,10,11,13,16])\n",
    "# plotName = \"heatMap_lin\"\n",
    "\n",
    "\n",
    "# gdf = findgenelist('sall') #sall4 mutations cause Duane syndrome; sall4 seems to be related to sall1 in zebrafish\n",
    "# plotName = \"heatMap_sall\"\n",
    "\n",
    "# For Mark Hankins\n",
    "# gdf = findgenelist('opn')\n",
    "# gdf = gdf.sort_values(by=[\"symbol\"])\n",
    "# gdf = gdf.drop([1,4,24,0,3,23,5,11,2])\n",
    "\n",
    "# gdf = findgenelist('rgr')\n",
    "# gdf = findgenelist('lrit')\n",
    "\n",
    "# gdf = findgenelist('ndrg') #potential target of tbx2\n",
    "\n",
    "# from YiRong Peng's talk: ON SACs: slit2 and fezf1; OFF SACs: tenm3 and rnd1?; contactins; Foveal cones: vtn and cacna1f\n",
    "# gdf = findgenelist('cacna2d2b')\n",
    "# gdf = gdf.drop([0])\n",
    "# gdf = gdf.sort_values(by=[\"symbol\"])\n",
    "\n",
    "# gdf = findgenelist('nrxn')\n",
    "# gdf = gdf.sort_values(by=[\"symbol\"])\n",
    "# plotName = \"heatMap_nrxn\"\n",
    "\n",
    "\n",
    "# gdf = findgenelist('frmpd')\n",
    "# gdf = gdf.sort_values(by=[\"symbol\"])\n",
    "# plotName = \"heatMap_frmpd\"\n",
    "\n",
    "# gdf = findgenelist('ntng')\n",
    "# gdf = gdf.sort_values(by=[\"symbol\"])\n",
    "# plotName = \"heatMap_netrins\"\n",
    "\n",
    "# gdf = findgenelist('dscam')\n",
    "# gdf = gdf.sort_values(by=[\"symbol\"])\n",
    "# plotName = \"heatmap_dscams\"\n",
    "\n",
    "\n",
    "# gdf = findgenelist('dag')\n",
    "# gdf = gdf.sort_values(by=[\"symbol\"])\n",
    "# plotName = \"heatmap_dag\"\n",
    "\n",
    "# gdf = findgenelist('syne') #involved in photoreceptor migration to ONL\n",
    "# gdf = gdf.sort_values(by=[\"symbol\"])\n",
    "# plotName = \"heatmap_syne\"\n",
    "# gdf = findgenelist('rasgrf') #also involved in phtoreceptor migration\n",
    "\n",
    "# gdf = findgenelist('syt')\n",
    "# plotName = \"heatmap_syt\"\n",
    "\n",
    "# gdf = findgenelist('rfx')\n",
    "# plotName = \"heatmap_rfx\"\n",
    "\n",
    "# gdf = findgenelist('anks') #anks3-/- causes RP\n",
    "# gdf = gdf.sort_values(by=[\"symbol\"])\n",
    "# plotName = \"heatmap_anks\"\n",
    "\n",
    "# gdf = findgenelist('spock') #https://zfin.org/action/image/view/ZDB-IMAGE-050208-655 -> GCL enriched?\n",
    "# gdf = gdf.sort_values(by=[\"symbol\"])\n",
    "# plotName = \"heatmap_anks\"\n",
    "\n",
    "# gdf = findgenelist('hcls') # Laura Campello: downstream of nrl and crx in mouse\n",
    "\n",
    "# gdf = findgenelist('slc32')  #from GO analysis: slc32a1 = L-cone specific + synaptic vesicle transporter.\n",
    "# gdf = findgenelist('slc') \n",
    "# gdf = gdf.sort_values(by=[\"baseMean\"],ascending=False)\n",
    "# gdf = gdf[gdf.baseMean > 100]\n",
    "\n",
    "# gdf = findgenelist('frmpd') #frmpd2 interacts with lrit1 in cones \n",
    "\n",
    "# from sqRNAseq + antibodies: rab32,ell2, ankrd33, pex5l, egr1, lamb1, chst3? or 7?, flna, cdh20, synj1\n",
    "# gdf = findgenelist('ell') # ell2 in sqRNAseql seems cone specific\n",
    "# gdf = findgenelist('ankrd33') #it's there, 33aa cone enriched; 33ba rods and 33bb cones.\n",
    "# gdf = findgenelist('pex5') #pek5l in sqRNAseq; pex5 in rods\n",
    "# gdf = findgenelist('egr') # egr1 looks interesting\n",
    "\n",
    "\n",
    "\n",
    "# gdf = findgenelist('gnat')\n",
    "# gdf2 = findgenelist('gnb')\n",
    "# gdf = gdf.append(gdf2)\n",
    "# gdf2 = findgenelist('gng')\n",
    "# gdf = gdf.append(gdf2)\n",
    "# gdf.index = range(len(gdf))\n",
    "# plotName = \"heatMap_transducin\"\n",
    "\n",
    "# gdf = findgenelist('aldh1')\n",
    "gdf = findgenelist('kif9')\n",
    "# gdf = gdf.sort_values(by=[\"baseMean\"],ascending=False)\n",
    "# gdf = findgenelist('gnat')\n",
    "# gdf = gdf[gdf.baseMean > 20]\n",
    "# gdf = gdf.drop(np.arange(20,93))\n",
    "# gdf = gdf.drop([4,24])\n",
    "gdf = gdf.sort_values(by=[\"symbol\"])\n",
    "\n",
    "gdf\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 34,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "image/png": "iVBORw0KGgoAAAANSUhEUgAAAsYAAABuCAYAAAAkuNUfAAAABHNCSVQICAgIfAhkiAAAAAlwSFlzAAALEgAACxIB0t1+/AAAADh0RVh0U29mdHdhcmUAbWF0cGxvdGxpYiB2ZXJzaW9uMy4xLjAsIGh0dHA6Ly9tYXRwbG90bGliLm9yZy+17YcXAAAWuklEQVR4nO3de5QcZZnH8e8zSQDRAYKIASIEAh5uQQ1BwWuG28yOXDwLLLByk0WOzl6OqLWKyHo9B7C8ojuyuCqBVRTC4SbDDAlOEEVuCdeIQCZcTIjKZcVRIIbNs39U9UylpnqmM9M1VT39+5zT011vve/Tz9tvT/fb1VXV5u6IiIiIiDS7lqITEBEREREpA02MRURERETQxFhEREREBNDEWEREREQE0MRYRERERASA6UUnIJsvCILtgI8B3wzD8E9F5yMiIiIyFWiLcWPaDvhcfC0iIiIidaAtxnUSBMFCoD9R5MAzwH3ABWEY3lFEXjL5etp6Pw8c1dnfsWC09cBK4K2d/R1vGSVWB3ATMKezv+N39c9W8tTeNTgP+CzwLmAH4Angx0DY1936SpG5STnNWDF3IdF7yTpg1w3zB16tUq8FWA3sBhy9Yf7AzyYtSWkYy+ftt5Do+dR64EMr/1JwOg1BW4zrbz9gd2Bv4Ezgb8DtQRC0F5qVlNFlwAE9bb0HjFLndGCpJsWNp71r8ADgDuBl4ETgbcB/AgHwqQJTk8bwRqIP0NW0A7tMUi4iTUNbjOvv6TAMK5/KHguCYAnwa+ATQF9xaUkJLQOeAk4B/j29sqetd1vgA8AZk5qV1MsFwN193a1nJMoeae8avAV4vJiUpIH0AGcD11VZfzZwLXDCpGUk0gS0xThnYRg6cDuwa9G5SLl09nc4cDlwck9bb9b/4onAS1R/Y5Ry25NoV6pN9HW3PtbX3eoF5CONpRton7Fi7oj3jhkr5s4i2pr83UnPSmSK0xbjyTEPeLKyEATBIcD5RPsdTid68/xqGIabTICCINiS6CvXU4j2I1tLNJH6SapeC/B5oq/d3wg8DARhGCb3eZZyWkT0XFgI/Dy17nTgx539HesnOympi6eAo9q7Bi/s6259ruhkpOHcDvwG+Ceig62TzgRWbZg/0D9jxdxJT0xkKtMW45wEQdASBMGbgiC4kGhfsIvj8sOA24gmym3A24GlwFVBEHwy0d6AG4CziA7emUf01dnhwBWpu/sE0YvnR4D5RC+m3w+CYEZe/ZP66OzvGAB+RfThZ0hPW++ewDuBHxSRl9TFuUTfFD3S3jUYtHcN6lsj2VyXAGfOWDF3WqVgxoq5RvR6f0lhWYlMYdpiXH/PBUEA0WPbAjwInBCGYU8QBFsAlwLfCMMwefDNyiAIngYuCYLgqjAMnyZ64TsE2DsMw2fieo8FQfBzYEnqPt8N3BSG4c0AQRCcCcwMw3BDTn2U+roM+FpPW29XZ39H5UwFZwAPdPZ3jPgqXhpDX3fr8vgAvP8g2uJ3UXvX4FLgor7u1luLzU4axOXAhUAncGNcdgSwE9G3TSJSZ9piXH/vBN5KtKVvDfCuMAwXx+veBewBfD2j3Q+BQeCYePlU4EeJSTEAYRhuBL6QansHcFIQBJ8KgmB/4P/CMHy2Hp2Rcall/9FknauAGcRj39PWa0Tjr63FDa6vu3VVX3fracAbgH8AZgJL27sGTxm9pQhsmD8wCPwI+HCi+MPAlRvmD+jHnURyoIlx/T0WhuFvic4ysCXR1qKKnYCXwzD8Q7pRfJDeamDnRN0nqt1HujnwZaJ9Uh8C1gRB8OERrWSyvEg0AapmB2DoTa2zv+PPREeXVyZLbcAsojdEmQL6ultf7utuXUz0wbmf6JgAkVp0A50zVszdZcaKuTsCx6KD7kRyo4lxTuKfav4UcE4QBPvFxeuA1wRBsGO6frxP8e5xnUrdOVXC75W6r41hGH4lDMN9gdlEX7H9VxAEmT8wIbl7Epjd09a7TZX1+5E4GDO2COjoaet9PdEHnOs7+zuezy1DKURfd+sGYDHR/6nImDbMH3gQuIto97oPAQ9smD9wb7FZiUxdmhjnaxFwN/DdeOJ7B9GE6JyMuqcB2xIdcAfR1sJ/DIJgp2SlOM75qbLdgiDYFiAMw7VhGH6GaKtl1V9Uk1z1A+uBj6VX9LT1HgS8j5GnYFsK/JHoze84tBtFw2vvGmxv7xrcN2PVe4D7JzsfaWjfJXptOAttLRbJlQ6+y1EYhh4EwT8Dy4HTwzC8LAiCs4EbgyB4LfBToglUJ/AZ4LwwDJ+Km3+P6Dy2dwRBcB7RQXy7Ee2ikT7bxNXAxvisFi8QfSX/GuCXuXZQMnX2d7zY09Z7DnBpT1vvdkQ/AfxX4FCi/cOv6ezvuCnVZmNPW+8VwBeBZ4FbJjltqaP2rsHpRLtLzG3vGjyf6NRbrUSTm6OIDpgVqdXVwDeAbUidrlOkRrsun7ffS6kyP/ChlU9l1m5i2mKcszAMHyD6GdgwCILtwzBcAhwEvBm4GfgF0AG8PwzDryTaOdFp3i4FvkQ0uf420bluT03dzQeAAeB64B6ifVSPCcPw0Ry7JqPo7O/4PnA00enzbgVWAF3ARcBJVZotItovfVFnf8fGychT8tHX3foqcBjRaRr/hej/9xpga+DAvu7WBwpMTxrMhvkD64m+RVq0Yf5AenIjUouVRMctJS8DhWZUUuauH2ASEREREdEWYxERERERNDEWEREREQEacGJsZmc3WuxGi5tn7EaLm2fsRoubZ+xGi5tn7EaLm2fsRoubZ+xGi5tnbMXNP3ajxa1n7IabGAO5Pag5xm60uHnGbrS4ecZutLh5xm60uHnGbrS4ecZutLh5xm60uHnGVtz8Yzda3LrFbsSJsYiIiIhI3ZXirBRHHXWU77PPPkWnISUz9/d7suua3YpOQ8bplj235ZEZ+xedhpTMc+/4DA/O6yk6DWkSX/zezhz5yxeKTkNKZsHDv/mQu1+Wta4UW4xnzpxZdApSQrvN0aS4kc3ZbU7RKUgJzZkzp+gUpIno+SZVzKm2ohQTYxERERGRopXiJ6Hnzp3Lu3tX1j3uEXctBmDJO45v6rh5xs4zbughSw8+oa5xK7GhsR6LPOLmGfuIuxbj/c7+dR4/PcaNG7cS27udpQffWfe40HiPRR6xGy1unrGPuGsxJ/sAS69tjNehPGM3Wtw8Yx9+59VgVnW9thiLiIiIiKCJsYiIiIgIoImxiIiIiAhQw8TYzLYzs9PGqDPfzJaY2QX1S01EREREZPLUssV4O2CTibGZ7WRmX0oUfRX4V3c/18xmmtkVZvYLM7vTzE6pZ8IiIiIiInkY11kp3H0dcH6i6E3u/tv49reApe6+yMy2AZaZ2Up3v2+CuYqIiIiI5KZe+xgnz3txiLsvAnD3PwOXAQvrdD8iIiIiIrnYrImxmc0zs4fNbB8zW2pmrzOzZcAuZrbMzN4X10vG3QJ4PiPW2WZ2r5ndO5EOiIiIiIjUQ80TYzPbD7gCOA54GcDd/+LuC4G17r7Q3W8DlgBB3GYX4AygLx3P3S919wXuvmCinRARERERmahaJ8Y7A9cDJ7j7o2PU/SQwy8xuB3qAm9z9DxPIUUREREQkd7VOjH8P3A2cPFZFd3/J3c8BDgXWAl8ao4mIiIiISOFqnRhvBE4HDjazj9fY5nPAle7+l3FlJiIiIiIyiWo+XZu7bzCz44n2Fx51smtmBwHvYdNTuomIiIiIlNaYE2N3fxI4PL79EtGEF+DSRJ09U23uAd5XtyxFRERERHJWr/MYi4iIiIg0NE2MRUREREQAc/eic+DUU0/1WbNmFZ2GlMyhWx2G/7LoLGS8HjvsLdz8zOuKTkNK5nUnXsoNrd1FpyFN4vIl7+TN/7Oi6DSkZBY8/JsvuPvns9Zpi7GIiIiICCXZYmxmZ7j7ZUXnISIiIiJT22jzzlJMjEVEREREiqZdKURERERE0MRYRERERATQxFhEREREBNDEWEREREQE0MRYRERERATQxFhEREREBNDEWEREREQE0MRYRERERATQxFhEREREBNDEWEREREQE0MRYRERERASA6UUnIOVk1uKjrK2hzEb8HdnMUsU2Yt3wUjqGVWk3etxqbSzjvmurW6XMsyNlZTvctay42UYfgfheMypVi5e5vtqw1ZKDVSmvIRfbZKWPXFdrnJr679XjVgmcFaP6mGa1r/Kvldmmen5QyTE73nCb4Sdj9fH3RLxq6W3ypM6+Lxsjl4z1o7YdekxGf8zGyi8yyuNQJW+s+thu8riPqDTK45m+ryptN/+fhuz7zPqnHPMfdeST0UeUVVnOSiIrqao5pP4DzMasM7ycqDsivm1S1zPzrbx2ptqMUnf4VX5knyqv6cuXP9Hn7h1IQ9DEWKpwjBnRzaEXgugLBqMFrGVkGWDpcps2Yt2IZZs2VD5cZ9omdVoYWbclXTeu02KpZaYlbm/apiWjfKjMW1JxUtfeQsvQC2BlnW3StgUbenFsSV0PldtweeXRG7kuLk8sW2pdi6XbVpaz4qTbDC/XUmc4h7HrVGKNjOtV2wzVSU0wWjLKh+P6JveZFT9dZ7Q2Q7fHjOuJOul8K3V9OPdE2SbLifsZuj1iXWo5ed+pulnxR9Sp1nYz61Y+R7fYxsw2Q9ctybgbR6yrFneobkadSnm1OmTESN8nLdm50LKxtrqpsqH7zGqTLqssZ7UdqkuqLpuU08LQP4elXxDS19aSKGtJXWeUx7d9qGxalbYteNV1UblbYrlSlq4zFCOxPNo6gJbpQ+WeuB3d5/RUnelx2+lQZZ0Nlc8YXo7LLFWnct1i04ffnyplcQ7T7ZQdkIahXSlERERERNAW46Zg0cfY9wNzSIy5u3+9qJxEREREykYT4+ZwI/AK8BCwseBcREREREpJE+PmMNvdDyg6CREREZEy0z7GzeFmMzuy6CREREREykxbjJvDncC1Fp3iYQPRMc3u7tsUm5aIiIhIeWhi3By+BhwCPOTuo5yfWERERKR5aVeK5vA48LAmxSIiIiLVaYtxc1gHLDOzm4H1lUKdrk1ERERkmCbGzeGJ+LJFfIFqvyMrIiIi0qQ0MW4Oi9z9yWSBmR1UUC4iIiIipaR9jJvDNWa2S2XBzN4L/KDAfERERERKRxPj5vAR4Dozm2VmncDFQGfBOYmIiIiUinalaALufo+Z/RtwC9FPQx/h7s8WnJaIiIhIqWhiPIWZ2Y1sepDd1sCLwPfNDHc/ppjMRERERMpHE+Op7atFJyAiIiLSKDQxnsLc/bbKbTObBbydaAvyPe7++8ISExERESkhHXzXBMzsLOBu4O+B44E7zezMYrMSERERKRdtMW4OAfA2d38ewMxeD9yBTtkmIiIiMkRbjJvDGmAwsTwI/K6gXERERERKSVuMm8Na4C4zu55oH+NjgbvN7OMA7v71IpMTERERKQNtMZ7CzOyK+OYxwHUMn7rtemAd0BpfRERERJqethhPbQea2W7A08C30yvd/YVR2q50NrwSVUy1yyhrYDsAzxWdxCRRX6eeZuknNE9fm6Wf0Dx93aroBKR2mhhPbZcAvcDuwL2JciOa2u4xSttX3H1BjrmVgpnd2wz9BPV1KmqWfkLz9LVZ+gnN01czu3fsWlIW2pViCnP3i919H+CH7r5H4rK7u482KRYRERFpOpoYNwF3/2jROYiIiIiUnSbGUs2lRScwSZqln6C+TkXN0k9onr42Sz+hefraLP2cEsx96hxFJSIiIiIyXtpiLCIiIiKCJsYiIiIiIoAmxk3FzLY3syVm9nh8PbNKvV4z+5OZ/SxVvruZ3RW3/6mZbRGXbxkvr4rXz8m/N6PbjL6eHtd53MxOj8tazez+xOU5M/tmvO4MM3s2se6syexXlon0NS5fZmaPJvq0Y1xeqnGd4JhubWY3mdlvzWylmV2YqF+aMTWzjngsVpnZpzPWVx0TMzs3Ln/UzNprjVmE8fbTzI4ws+Vm9lB8fWiiTebzuGgT6OscM3s50Z9LEm0OjB+DVWZ2sZnZ5PUo2wT6+cHU6+1GM3trvK5Rx/S9ZrbCzF41s+NT66q9DpduTJuWu+vSJBfgK8Cn49ufBi6qUu8w4GjgZ6nyq4CT4tuXAB+Nb3cBl8S3TwJ+2gh9BbYHVsfXM+PbMzPqLQfeG98+A/hO0f2rZ1+BZcCCjDalGteJ9BPYGmiL62wB3A78XZnGFJgGDBCdX3wL4AFg31rGBNg3rr8l0XnLB+J4Y8ZssH6+Ddg5vr0/sDbRJvN53MB9nQM8XCXu3cAhROekv7nyXG7EfqbqzANWT4ExnQMcAFwOHJ8oH+11uFRj2swXbTFuLscCi+Lbi4APZFVy91uBwWRZ/On1UGBxRvtk3MXAYSX4tFtLX9uBJe7+grv/L7AE6EhWMLO9gB2JJlJlVZe+jhG3DOM67n66+0vu3g/g7n8DVgCzJyHnzfF2YJW7r45z/AlRn5OqjcmxwE/cfb27PwGsiuPVEnOyjbuf7n6fuz8Tl68EtjKzLScl6/GZyJhmMrOdgG3c/dcezagup8pr+SSqVz9PBq7MNdOJG7Ov7v6kuz8IbEy1zXx9KumYNi1NjJvLG919HUB8vTlfS70e+JO7vxovrwF2iW/vAvwujvsq8GJcv0i19HUo71iyTxUnE23ZSJ6+5Tgze9DMFpvZm+qZ9DjVo68/jL+qPD/xZlW2ca3LmJrZdkTfiNyaKC7DmNbyfKw2JtXa1hJzsk2kn0nHAfe5+/pEWdbzuEgT7evuZnafmd1mZu9J1F8zRszJVq8xPZGRE+NGHNPNbVvGMW1a+knoKcbMlgKzMladN9HQGWVew7rc1KGvteR9EnBqYvlG4Ep3X29mHyHaAnIoOcu5rx9097Vm1gpcQ9Tfy8dok4u8x9TMphO98V7s7qvj4kLGNEMtj3e1OtXKszZ+FH2Ozon0M1ppth9wEXBkYn2153GRJtLXdcCu7v68mR0IXBf3u5DX2zHUY0zfAbzk7g8n1jfqmG5u2zKOadPSxHiKcffDq60zsz+Y2U7uvi7+6uaPmxH6OWA7M5sef9qfDVS+0lwDvAlYE088tgVeGF8PaleHvq4BFiaWZxPt01aJ8RZgursvT9zn84n63yN6c85dnn1197Xx9aCZ/Zjoq8LLKWBc8x5TohPtP+7u30zcZyFjmqHyeFck/8fSddJjMlrbsWJOton0EzObDVwLnObuA5UGozyPizTuvsbfUq0HcPflZjYAvDmun9wNqOHHNHYSqa3FDTymo7VdmGq7jHKOadPSrhTN5QagchTs6cD1tTaMX6T7gcoRtsn2ybjHAz9P7XpQhFr62gccaWYzLTrDwZFxWcWI/d3iCVnFMcAjdct4/MbdVzObbmY7AJjZDOAooLLFpmzjOqExNbMvE70ZfyzZoERjeg+wl0Vnf9mCaKJwQ6pOtTG5ATjJoiP/dwf2IjqYp5aYk23c/Yx3g7kJONfdf1WpPMbzuEgT6esbzGwagJntQTSmq+PdiAbN7OB414LT2IzX8pxM5LmLmbUAJxDtr0tc1shjWk3m61NJx7R5TdZRfroUfyHan+tW4PH4evu4fAHw34l6twPPAi8TfZJtj8v3IHqzXQVcDWwZl28VL6+K1+/RQH09M857FfChVIzVwN6psguIDvp5gOiDwt559iPvvgKvJTrrxoNxv74FTCvjuE6wn7OJvpp8BLg/vpxVtjEFOoHHiI56Py8u+yJwzFhjQrS7yQDwKIkj2rNiFn0Zbz+BzwJ/TYzh/UT7mld9Hhd9mUBfj0s8L1cARydiLiCaJA4A3yH+FdtG7Ge8biFwZypeI4/pQUTvnX8FngdWJtpmvueUcUyb9aKfhBYRERERQbtSiIiIiIgAmhiLiIiIiACaGIuIiIiIAJoYi4iIiIgAmhiLiIiIiACaGIuIiIiIAJoYi4iIiIgA8P8UdaxAxdZiyQAAAABJRU5ErkJggg==\n",
      "text/plain": [
       "<Figure size 864x864 with 2 Axes>"
      ]
     },
     "metadata": {
      "needs_background": "light"
     },
     "output_type": "display_data"
    }
   ],
   "source": [
    "genenames = gdf['symbol'].values\n",
    "samplenames = [] #skipping names to replace in intaglio\n",
    "data = gdf.iloc[0:,7:37].values\n",
    "\n",
    "# log2matrix = np.vectorize(log2)\n",
    "# data = log2matrix(gdf.iloc[0:,7:37].values)\n",
    "\n",
    "fig, ax = plt.subplots(figsize=[12,12])\n",
    "im, cbar = heatmap(data, genenames, samplenames, ax=ax,\n",
    "                   cmap=\"inferno\", cbarlabel=\"fpkm\")\n",
    "# texts = annotate_heatmap(im, valfmt=\"{x:.1f} t\")\n",
    "\n",
    "plt.savefig(\"{0}.svg\".format(plotName), transparent=True, format=\"svg\")\n",
    "plt.show()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 35,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "image/png": "iVBORw0KGgoAAAANSUhEUgAACGgAAAEYCAYAAAAKilKRAAAABHNCSVQICAgIfAhkiAAAAAlwSFlzAAALEgAACxIB0t1+/AAAADh0RVh0U29mdHdhcmUAbWF0cGxvdGxpYiB2ZXJzaW9uMy4xLjAsIGh0dHA6Ly9tYXRwbG90bGliLm9yZy+17YcXAAAgAElEQVR4nOzdeZQldXn/8fcHGJDNAIpLosgioiHGbTRCXAAxIhr3BWURFSeo+HPBMWgSxR8al4lLjBqCqAiK+xbZjMqAIGJAjOsviCju4LApKKs8vz+qWtu2+3ZP37W6369z+tR01VNVT3vUz7n3PvdbqSokSZIkSZIkSZIkSZI0PBuMuwFJkiRJkiRJkiRJkqSlzgENSZIkSZIkSZIkSZKkIXNAQ5IkSZIkSZIkSZIkacgc0JAkSZIkSZIkSZIkSRoyBzQkSZIkSZIkSZIkSZKGbKNxNyBpOJI8EDgQ2Am4HnhtVZ033q4kSfpDw8irJC8CtqqqI/vvUJK01Ezaa6Uk2wMHA/9TVZ8aVx+SpKVp0nJPkqS5mFlaLlJV4+5BWraSFPDDqtp+wNd9KfBGINN2P2B6kCXZGtgb2K7d9b/A6VV13SB7kSR1X9fyKsklwF2qKnPVSJIm2zizZ9SS7AGsBd5XVQePqw9J0vh0OfeSbAM8DNgeuAX4Hs1rtmsGdQ9J0uQY8/uEGwAPBu4NbA78HDirqr43yF6kYfMRJ1LHJNk1yY+SnJVk01mO7wS8Afg1sC+wCfCnwNfa4yuSvAn4GfAR4F/an5OAHyd5zmj+EknSUmZeSZJGrd/smaX+fkl+nqSS7L3Inta25x+3mPMlSZrLuHMvySYzXrO9keY126eAnyV5ZftBmiRpmRtEZiXZE/gucAbwVuC1wHuAi5KcmuQuPe5/TZtv8/3ce4B/tjQnH3Eidc8+wJ3bnz8Hvjrj+GNohq/eVlWntvt+DpAkwKeBRwLXAP8GfIkm7B4IPB84JskWVfWWIf8dkqSlzbySJI3aorNnpiSPAj5M862sRUnyFGAP4ErgpYu9jiRJcxhb7iXZBPgczbeYr6AZzvgKsCFN9q0CXg3cM8lTymW8JWm56yuz2pz6FM3n2v9NM5hxCbAD8OL2+uck2b2qfjjL/ac+Dz9znj6vXeDfI/XFAQ2pez5IEzaXMPvE+07t9r9nOfZ4mg+7LgX2qqr/N+3YR5J8BDgb+OckH6mqnw6sa0nScmNeSZJGrZ/s+Z0kz6UZDtyAZlBwy/VtJMlmNN8iBlhdVZev7zUkSZrHOHPvKJrhjG8Be1fVZdOO/WeSo4HTgScBh7XXlyQtX4vOrCSbA++l+Uz7GODQ6YN/Sd4DfJJm5Y230ryvONNGAFW1x2L/AGmQXGJM6piq+llVPbyqnlNVt8xSMjXp/qtZzv0E8Chg/xkfdk0dPxf4EHAr4MkDbFuStMyYV5KkUesne6BZwSnJG4B3trueCVywyHZeQfPtsLNp3kyUJGmgxpV7Sf6EZujiFuCpM4Yzpnr7LjD1WMoj2lUSJUnLVJ+Z9RRgW+AnwGEzV2WqqhuBZwM3AX+bZJtZruGCBZooDmhIS8/UC55Zlw6sqlOq6vQe50+9ELv7QLuSJOkPmVeSpFHrmT3AE4GXATcAT6qq9y3qJsmOwOE0bxAeOvMNREmSRmRYubcXsClwZlV9Z66idon6HwF/CtxngdeWJC1PvTLrEuANwD9V1U2znVxVlwLfpHnU1l3/4MKJwxmaOP6XUtJMN467AUmSFsC8kiQN2seB9wPvnWdIcD5voVnl6XVV9e2BdCZJ0uAtNvd2bLffWEDt/wDbATuw+FWpJEnLWFWtBdYuoHRFu5352feKmYXSuDmgIWmmB7Tbr4y1C0mSejOvJEkD1a50cWA/10jyCOAxwA+AowbRlyRJw9BH7k190DXrt5hnmBqs33AR95EkaUGSbAncrf31hzMO+1m4Jo6POJEmVJIXJKkkNyd52rT927f7z5i27+B2XwHPaHevndqX5OAF3vNewFNplh88cY6aP0vy2iRfTrIuyY1JLkvywST3XOSfK0nqqEnNq2m1GyZ5TpIvJLk8yXVJvpfkmCT3Xs8/V5I0AcaRPQvsawXwr+2vz6+q6xZ5nT9L8sYk307ymyRXJjk3yRFJ/mSec1cmOTrJt5Jc1b5euzjJW5LcZjH9SJLGawJz76ftdseeVY2d2+26AdxXkjThxphZhwKbAP9TVT+dccwBDU0c/0spTaAkzwXeBtwMPL2qPjrPKZcCZ7b/vjtwe+DrwNXTjs91r1sB2wNPpnlO8vXAU6vqhllqnwn8O03QXQf8L8004l2B/YDHJXlEVX1x/r9SktR1k5pX0865PXAqzfOObwD+H803uO4KPAc4JMmbgb+vqt/O07skaQKMMnsW4UXALsB3gacneSOwNc0HWacAb6uqq3pdIMkTgfcCWwK/oFk+fjPgfsBfAYcn2a+qvjDjvI1o/nN5brvriraPzWly70XAY5PsVlWXDeBvlSSNwITm3tT7fg9PsvVc2ZZkF+AvaXo/fwD3lSRNsHFlVpKdgFe2v66ZpWTFtNqPAw8Ebgv8Cvgq8B7go+3KUtJIuIKGNGGSHAK8g2aZwKcsIMSoqtOqao+q2gM4rd39oql9VXXabOe104rX0Xxg9X+BbwEPqKpz57jV5sBVwDOBravqvlW1Erhde/6tgHcn8f9bJGmJm/C8mnIq8OfAy4DbVtV9quqvaHLrycBlNMMeb52vd0nS+I0yexbR258A/9D+ejfgAOAvgD+jeSzXkcB3kzy4x2VW0qwM9WNgb+AOVfXAqvpLmjcr30zzRuIpSXabce5v22PnAw8Dtq2qv6qqvwDuDHwG2AF4XZ9/qiRpRCY196rqh8B/0gwTvmO29wGTbAL8BxDglKq6pt/7SpIm17gyq30d9mlgC5rXQh+epWz6YgUPpRmg/wbNF5Ef0Z7z6SSbzXc/aVD8EFWaIO2STcfQhNgTq+qTQ77l/9BMCF7Z/r47sCbJNrMVV9XbgZ2r6rjp31iuqhur6lXAf9N8O2vlcNuWJI3TpOfVNH8B7FNVa6rq2qmdVfXbqvoY8CCaqfzDkjxkCH1LkgZkDNmzvg4Dph4/ci6wB82A+58Af0uz+uBtgZOT7DzbBYBdaYYQH1hVX5j+Da6qurKqDgdeCmwMvC/JhtOOF3AgsHtVnT7j3HXAwTSrTz1p+nmSpMnUgdz7PzSrNT0NOC3Jg5JsmmTLJPsAZ9F8CAbwhnE1KUkavnFlVpKNgU/SvI66Btivxwq5JwC7VdVtq+oBVXV/4DY0X0b+Jc1rtneNoG0JcEBDmhhJ9gfeTbP0+uOq6jPDvmdVvaiqVlbVbYB7AJ+gCaIvJNlyjnOunW1/a+qbzDsMtlNJ0qToSl613lpVZ/S47sXAa9tfXzKofiVJgzWO7Fkf7RuDL2p/PRZ4UFWdWVW/qapfVdVJNI/bOpXm28ZzfVB1C/CMeb5l/GaaIY6dabLwd6rqhqq6abaTqupKmkeebEkzKCJJmlCTnnvwu1U09qDJlofTDGT8hma5+FOB+7elZ1XVOePoUZI0fOPKrCQBjgf25Perdlw8W21V/ayqDpq5Em9V3VRVx9Hk2I00j6ncfbidSw0HNKQJkOSpwPuAG4DHVNWpo+6hqv63qp4EfAi4N80k/KyS3C7JQUn+JckHkvxnki8Aj21LNh1By5KkEetaXrGwyfcPtNu92xd3kqQJMgnZswC70ww9XA68cLZvbVXV9cDf0TyP+bFzrAJ1blV9q9eN2pUxTmx//ZvZapLskuQFSd6R5ENJTk6yFrhLW+LrNUmaUB3JPQDazPoLmhWcTgD+i+b11UuBXwBF87hJSdISNObMehPwVJoh94P6eYRXVZ0HvKf99aAB9CbNa6P5SyQN2a1p/s9/Q+CYqvqvMffzj8B+wAv4/beKAUhyK+BfgFXAitG3Jkkao87kVev6qrpovotU1c+TXA1sRbO04eUD7VKS1I9Jy565/Hm7/WxV/Wauoqr6cZLzgN1ovln82Rkl31jg/b7TbrebvjPJdjRLCz9igdeRJE2WruTe77QrN72//QEgyVHA7YD3z/y2siRpyRhbZiU5HHhx++vzq+pDA7jsx4FD+f0KUNJQuYKGNH6/Ap5FM+n3gnZJqLFpl4G6Crh9kjvMOPwJ4PnAz9rtXYFNqipVFeDVI21WkjRKXcoraL6tJUnqtonKnh62brc/XUDtj9vtbCtoLFq7IseXaYYzzgGeDNwR2Gja67UzB3lPSdLAdSX35pRkZ2A18GvgiDG3I0kanrFkVpL9gDXtr6+oqqMHdOkftdutBnQ9qSdX0JAmQFV9OMnWwL8DxyX5TVV9cowt3dxuf7fUe5KHAI8Efgjcu6qunuU8n2UsSUtYF/Jqmk2T3LWqvtfrAknuSPPi69fAFQPuT5LUpwnMntn8st3+yQJqp4Y5Zns9dc8F3u8e7fZH0/YdBvwp8CngCe2jUGby9ZokTbiO5F4vbwc2AV5TVQsZXJQkddSoMyvJXjSPVAmwpqpeN8DLT71WcmVdjYQraEgTop30+weawakPJRnosrRJNk/y1PY5xLN9kDVVtz2wLc0HVb+Ydug+7fbTcwxnANx9EL1KkiZXB/Jquucs4JZPa7efn+PDLEnSmA07ewbgf9rtHvNk1+bAA2lWebpglpLdkuy6gPtNZdf0ZYSnXq8dP1ueJdkY2HEB15YkjVkHcm9WSZ4M/A3NAOGbxtyOJGkERpVZSe4FfBLYGDi2ql62HudumuQ+85Q9pt2es8gWpfXigIY0Qarqn4E304TMJ5I8eICX/zPgROB5wH496l7Vbj9TVb+dtn9qxZ1tZzspyT2APfrsUZLUAROeV9O9OMkec10gyU7AP7W/vnk9+5QkjdCQs6dfXwYuBnYBnt2j7hXAlsBnq+qyWY5vALwvyZZzXaB93vJfAhcBJ0071PP1GvAMYNMevUmSJsiE594fSbIFv39NtbqqrhtnP5Kk0Rl2ZiW5C3AqcGvgo8Dfrce5u9EMDq5Ncuc5au4OPJ/mcS3v7rthaQEc0JAmz0uB44DNgJOS3H8QF62q7wLvaH89Nsn+STacOp7ktkmOBw4GrgVeOeMSa2m+6fXUJM+c/s2wJA8FPk/zpiTAikH0LEmaaJOaV9NdDJyWZHX7huHUNTZM8iTgbJrHm7y9qr44iP4lSUM1lOzpVzso+AKaN/T+Pcnh7YoVQPOhVZLX0wxoXAscPselLgbuBZyb5GEzXnNtk+RfgH8BbgQOrqqbp517ert9TZIHTDsvSZ4H/Csw9dgvX69JUjdMZO7N4UjgTsDZVfWRMfciSRq9oWRWks2A04A7Ap8FDqiqW9bjEl8FLqN5HOUZSR407dpJ8jSaVTO2AN5cVd8aRN/SfBzQkCZMuxztIcCnaSYCT0uy0GcRz+dwmm8lbwa8H7g8yQVJzgd+ChwIrAP+tqoumtHXBcDraP5/4z3AT5Ock+Qi4Azg68Br2vLNB9SvJGlCTWpezfBomm8Yv7G9xteSnEvzSJSPArcH3gK8aEB9S5KGaMjZ05eqOhV4FvBbmiGKK5J8pc2udcDfA1cAj66q78xxmfOBg4DtaQbgL03y5SRfp3lT8fD2GvtW1cyld/8NOJNmBY2vJLkwyTnApTSDj6/h90Mcvl6TpA6Y5Nybrn081wtpBhVfOOZ2JEljMMTMuh1w9/bf2wD/leSMeX4OntbXjcC+NJ9f7QicleSSJF8Gfkbz/uPWwNuAIwbQr7QgDmhIE6j9BtZTaVat2Ab4XJKdB3Ddm6pqf+DBNB94/YJmGd5dgG/QLBf/51V1xhyX+EdgFfAVmuVx7wlcA7wY+FuaN/8A7tBvr5KkyTfBeTXlp8B9gcNoPvTageabyVfTDBuurKqX9HhEiiRpwgwrewahqt4H3INmWOKnNK+X7g5cCBwF7FJVZ85zjQ8Cu9K8QfhL4N7AXYCv0TyW665V9YVZzruZZjDxdcC3gTsDdwO+CTymXXb48rbc12uS1BGTnHvTvJPmUVvHtV/wkiQtQyPIrPsDD13Az/Yz+voRsBvNo1HOpvlC2H2BG4ATgAdW1Qt9f1CjlGaoSZIkSZIkSZIkSZIkScPiChqSJEmSJEmSJEmSJElD5oCGJEkjkmSLJKckqSQHzFGzR3u818+RI25dkiRJkiRJkiRJfdpo3A1IkrQcJLkjcDJwnwWeshb44hzHzhhET5IkSZIkSZIkSRodBzQkSRqyJLsCpwB3Ar4APGwBp51eVa8ZamOSJEmSJEmSJEkaGQc0JEkaoiQrgc8DmwH7A7diYQMakiRJkiRJkiRJWkI2GHcDkiQtcd8Dfgw8vqo+NO5mJEmSJEmSJEmSNB6uoLGe9tlnnzrttNPG3YYkLScZdwP9qKqrk9ynqm4e9b3NLEkai07n1jiZW5I0FubWIplbkjQW5tYimVuSNBaz5pYDGuvp8ssvH3cLkqSOWeRwxu2THAjcCbgGOKeqLlifC5hZkqQuMbckSV1ibkmSusTckqTJ4YCGJEmT6bCZO5J8ATigqi4dQz+SJEmSJEmSJEnqwwbjbkCSJP2B3wKfAR4FbAPcCrgf8GHgYcDaJJvPdXKSVUnOT3L+unXrRtGvJEmSJEmSJEmSFsABDUmSJkhVnVVVj6mqU6rqqqq6oaouqKr9gOOAuwMv6XH+MVW1sqpWbrvttqNqW5IkSZIkSZIkSfNwQEOSpO5YDVwLPGPcjUiSJEmSJEmSJGn9OKAhSVJHVNXlwJeAnZJsMe5+JEmSJEmSJEmStHAOaEiS1C1XtttNxtqFJEmSJEmSJEmS1osDGpIkdctONI85uXK+QkmSJEmSJEmSJE0OBzQkSZoQSW6X5I1JDpnj+EOABwCnVFWNtjtJkiRJkiRJkiT1Y6NxN6BuWr16dc/ja9asGVEnkrSk3BZ4IXBjku9W1RenDiT5S+A44AbgqPG0tzydsudpPY/vu3afEXUiSeoyX0NJkqTFesTzrul5/LPv3HJEnUiS5uN7iZLm44CGJEkToqq+k+RQ4BhgbZKTge8AdwEeDxRwYFV9a4xtSpIkSZIkSZIkaREc0JAkaYJU1XuTfAs4HHgQ8AjgMuD9wJqqunCc/UmSJEmSJEmSJGlxHNCQJGmEquo4mkeV9Ko5D9hvFP1IkiRJkiRJkiRpNDYYdwOSJEmSJEmSJEmSJElLnQMakiRJkiRJkiRJkiRJQ+aAhiRJkiRJkiRJkiRJ0pA5oCFJkiRJkiSp05JskeSUJJXkgB51SXJAki8nuTLJVUnObfdllD1LkiRJWn4c0JAkSZIkSZLUWUnuCHwReOQCyo8GTgC2Bj4IfADYqt139LB6lCRJkiSAjcbdgCRJkiRJkiQtRpJdgVOAOwFfAB7Wo/YgYBXNUMYzq+qmdv8K4L3AqiRnVtWJQ29ckiRJ0rLkChqSJEmSJEmSOifJSuBLwB2B/YH396gN8CrgUuA5U8MZAO2/DwF+Dhw5xJYlSZIkLXMOaEiSJEmSJEnqou8BPwYeX1Ufmqf2AcCOwLFVdd3Mg1V1PfAeYOck9x94p5IkSZKEAxqSJEmSJEmSOqiqrgbuU1UnL6B8t3Z7eo+ate12974akyRJkqQ5OKAhSZIkSZIkqZOq6uYFlu7Ybi/sUXPRjFpJkiRJGigHNCRJkiRJkiQtdbdpt1f0qLl8Ru0fSLIqyflJzl+3bt1Am5MkSZK0PDigIUmSJEmSJGmp2waoqrphroKq+g1Qbe1sx4+pqpVVtXLbbbcdUpuSJEmSljIHNCRJkiRJkiQtB0mSHgc3BOY8LkmSJEn9ckBDkiRJkiRJ0lJ3ZbvdrEfN5jNqJUmSJGmgNhp3A5IkSZIkSZI0ZFe0262AX89Rs9WMWmlRHvG8a3oe/+w7txxRJ5IkSZo0rqAhSZIkSZIkaan7fru9W4+anWfUSpIkSdJAOaAhSZIkSZIkaak7p93u1aNmzxm1kiQNTZK7J7kxyYVJNpmjZkWSw5J8Lcmvkvwiydokjxp1v5KkwXBAQ5IkSZIkSdJSdx7NyhiHJNl05sF237OBi6rqvFE3J0lalv4VWAE8v6pumHkwycbAJ4F/A24B3gN8CrgrcFKSl4+wV0nSgDigIUmSJEmSJGlJq6oCXg3cATg2yYqpY+2/39UeO3IsDUqSlpUkjwf+BvhgVX1+jrIjgEcBbwBWVtWLqmoVcA/gdOC1Sf56JA1LkgZmo3E3oOXnlD1P63l837X7jKgTSZIkSZIkLRdVdXz7QdYq4H5JPgcE2BvYBTimqk4cZ4+SpKUvya2ANwG/BF4yR83mwGrga8DL20FDAKrq2iT7Az+gGSx8+LB7liQNjitoSJIkSZIkSVouDgUOAq4GDgD2b/99UHtMkqRhexmwA/APVXXpHDWPBrYA3jF9OGNKe97Hgb2S3H5onUqSBs4VNCRJkiRJkiR1XlUdBxw3T00BJ7Q/kiSNVJLtaB5dcj7w7z1Kd2u3p/eoWUszaPhA4NMDaVCSNHSuoCFJkiRJkiRJkiQN35uATYBDq+qWHnU7AjcAP+xRc9G0WklSRzigIUmSJEmSJEmSJA1Rkj2BJwE/AN6e5Jftz+eTPHpG+W2Aq+YZ4rh8Wq0kqSMc0JAkSZIkSZIkSZKG6/+2252AWwNfAL4N7Al8JsnrptVuA1w3z/V+Pa32jyRZleT8JOevW7du8V1LkgbKAQ1JkiRJ0h9JskWSU5JUkgN61CXJAUm+nOTKJFclObfdl1H2LEmSJEmTKMmDgQcBvwEOrKpdq+oJVbU7sBtwMXBEksdNO23DeS67Ua+DVXVMVa2sqpXbbrttP+1LkgaoswMaSR7Zvll4WZIbk/wgyTuT3HmOet80lCRJkqQFSHJH4IvAIxdQfjRwArA18EHgA8BW7b6jh9WjJEmSJHXIge32FVX1/ukHquq/gacBtwB/3+6+EthsnmtuPq1WktQRnRzQSHIUcArwV8DJwFuBrwOrgK8nudcsp/mmoSRJkiTNI8muwLnAvWiW3O1VexDN67APAPesqudX1WHAPdt9q5I8fcgtS5IkSdKkuw9wPfDvsx2sqvOAs4EHJNkUuAK49TxfMN6q3V4xyEYlScPVc/mjSZRkF+ClwFnAY6vqqmnHdgfOAI4F7j9t//Q3DZ9ZVTe1+1cA76V50/DMqjpxVH+HJEmSJE2aJCuBz9N8U2t/4FbAw+aoDfAq4FLgOVOvswCq6qYkhwB7AUcCvtaSJEmStJxtBVxWVTf2qLkEeAjNF42/D2wMbA/8YI76ndvt9wfToiRpFDq3gkZVXQjcFXji9OGM9tg5wCeAlUm2h/nfNAQOAX5O86ahJEmSJC1n3wN+DDy+qj40T+0DgB2BY6vqupkHq+p64D3AzknuP/O4JEmSJC0jVwC3SbJhj5rb0zzm5CrgnHbfXj3q92zrzx1Ih5KkkejcgAZAVf20qtbNcfiSdju1tJNvGkqSJEnSAlTV1cB9qurkBZTv1m5P71Gztt3u3ldjkiRJktRtZwBbAE+c7WCSHYCHAv/dfpZ1MnAtcNhsjzlJcof2WqdX1WXDalqSNHide8TJAvw5zXO8Lmx/X+ibhv9A86bhecNrTZIkSRq8FRfs1PP4Tfe9eESdaCmoqpsXWLpju72wR81FM2olSZIkaTl6O3AY8LYk36+q86cOtMMWJ9A8YvK1AFX16yRrgFcDr09yRFVVW78FzWMkN8HV4SWpc5bUgEaSZwN/C7x22moZvmkoSZIkSYN3m3Z7RY+ay2fUSpIkSdKyU1U/SbIf8GHgy0k+BXyX5rEmTwC2Bv6xqk6adtrraVaJfxmwd5IvApsB+wJ3Al5RVV8a4Z8hSRqATg9oJNkKeBFwW+CBwL2ANwOvnFbW95uGSVYBqwC22267PjqWJEmSpCVjG6Cq6oa5CqrqN0mqrf0jvtaSJEmStFxU1UlJ7gUcATwceCzwS+Bs4M1VdcaM+huTPB44FHgWcAjNCvLfAp47Y5hDktQRnR7QALYCXjXt928CZ1XVLdP29f2mYVUdAxwDsHLlyuq7a0mSJElaGpIkU0vtznJwQ+CPnpc8xddakiRJkpaTqvoezaDFQutvAv6t/ZEkLQEbjLuBflTVJcCGwJ2BRwFXAZ9M8uIZpUky55uC871pKEmSJEn6I1e228161Gw+o1aSJEmSJElatjo9oAFQVbdU1U+q6hRgL+A04I1JdmpLfNNQkiRJkgZv6jGSW/WomTrW65GTkiRJkiRJ0rLQ+QGN6arqt8AraR7d8sx2t28aSpIkSdLgfb/d3q1Hzc4zaiVJkiRJkqRla0kNaLS+2W7v0W5901CSJEmSBu+cdrtXj5o9Z9RKkiRJkiRJy1YnBzSS3KrH4S3a7S3t1jcNJUmSJGnwzqMZcj8kyaYzD7b7ng1cVFXnjbo5SZIkSZIkadJ0akAjyYZJPgP8PMlcjyx5ers9b9rWNw0lSZIkaYCqqoBXA3cAjk2yYupY++93tceOHEuDkiRJkiRJ0oTp1IBGVf0W+F9gK+DdSTaefjzJg4DXAr8EjmvP8U1DSZIkSRqCqjoeOIZmUP6bSf4tydtpHj25P3BMVZ04zh4lSZIkSZKkSbHRuBtYhFcB9wCeAFyc5DRgXbvvscCvgSdX1S+mTqiq45P8NbAKuF+SzwEB9gZ2wTcNJUmSJGmxDgXOBp4PHNDuuxA4CHj/uJqSJEmSJEmSJk3nBjSq6jdJHkfzbayDgb8FtgZ+ChwNrKmqH8xyqm8aSpIkSdJ6qKrjaFcn7FFTwAntjyRJkiRJkqQ5dG5AA6Cqbgbe1/4s9BzfNJQkSZIkSZIkSZIkSWOxwbgbkCRJkiRJkiRJkiRJWuoc0JAkSZIkSZIkSZIkSRoyBzQkSZIkSZIkSZIkSZKGzAENSZJGJMkWSU5JUkkO6FGXJAck+XKSK5NcleTcdl9G2bMkSZIkSZIkSZIGwwENSZJGIMkdgS8Cj1xA+dHACcDWwAeBDwBbtfuOHlaPkiRJkiRJkiRJGp6Nxt2AJElLXZJdgVOAOwFfAB7Wo/YgYBXNUMYzq+qmdv8K4L3AqiRnVtWJQ29ckiRJkiRJkiRJA+MKGpIkDVGSlcCXgDsC+wPv7y9/KBEAACAASURBVFEb4FXApcBzpoYzANp/HwL8HDhyiC1LkiRJkiRJkiRpCFxBQ5Kk4foe8GPgiKo6OcnBPWofAOwIvKaqrpt5sKquT/Ie4B+S3L+qzhtKx5I0i6/ec9eex+/3zW+PqBNJkiRJkiRJ6iZX0JAkaYiq6mrgPlV18gLKd2u3p/eoWdtud++rMUmSJEmSJEmSJI2UAxqSJA1ZVd28wNId2+2FPWoumlErSZIkSZIkSZKkDnBAQ5KkyXGbdntFj5rLZ9RKkiRJkiRJkiSpAxzQkCRpcmwDVFXdMFdBVf0GqLb2jyRZleT8JOevW7duSG1KkiRJkiRJkiRpfTmgIUnSZEmS9Di4ITDn8ao6pqpWVtXKbbfddigNSpIkSZIkSZIkaf05oCFJ0uS4st1u1qNm8xm1kiRJkiRJkiRJ6gAHNCRJmhxXtNutetRMHbuiR40kSZIkSZIkSZImzEbjbkCSJP3O99vt3YCfzlGz84xaSX1YccFOPY/fdN+LR9SJJEmSJEmSJGmpG8iARpKNgT2BuwO3BtKjvKrqqEHcV5KkxZjg3Dqn3e4FrJ2jZs8ZtZKkJW6Cc0uSpD9ibkmSusTckiSNWt8DGkmeCLwTuO3UrnlOKcAAkySNxYTn1nk0K2MckuSfq+q66QeTbAo8G7ioqs4bUU+SpDGa8NySJOkPmFuSpC4xtyRJ49DXgEaSBwMfBjYAvgWcBPwMuKb/1iRJGqxJz62qqiSvBt4HHJvk4Kq6CSDJCuBdwB2Aw8fYpiRpRCY9tyRJms7ckiR1ibklSRqXflfQeBnNROHhVfWWAfQjSdIwTXxuVdXxSf4aWAXcL8nnaHreG9gFOKaqThxnj5KkkZn43Oqi1atX9zy+Zs2aEXUiSUuOuSVJ6hJzS5I0Fv0OaPwVcJnhJUnqiK7k1qHA2cDzgQPafRcCBwHvH1dTkqSR60puSZIE5pYkqVvMLUnSWPQ7oLEFcMEgGpEkaQTGnltVdRxw3Dw1BZzQ/kiSlq+x55YkSevB3JIkdYm5JUkaiw36PP9i4HaDaESSpBEwtyRJXWJuSZK6xNySJHWJuSVJGot+BzSOB3ZKsucgmpEkacjMLUlSl5hbkqQuMbckSV1ibkmSxqLfAY1/Bc4FPp7k4CSbDKAnSZKGxdySJHWJuSVJ6hJzS5LUJeaWJGksNurn5Kq6MckjgP8A3g28KcnXgZ8BN859Wj27n/tKkrQY5pYkqUvMLUlSl5hbkqQuMbckSePS14BG65nA44EAWwN7zFNfgAEmSRoXc2uJWL16dc/ja9asGVEnkjRU5pYkqUvMLUlSl4w9t5K8Hvh74LdV9Uef2SUJsD/wfGCXttcLgbcDH6iqGmQ/kqTh62tAI8kTaJaBugU4GfgMzXThNf23JknSYJlbkqQuMbckSV1ibkmSumQScivJzsCLgZtpBi9mczSwimYo44M0QyJ7AycADwb+bvidSpIGqd8VNF5AEwb7VdXHBtCPJEnDZG5JkrrE3JIkdYm5JUnqkknIrbcC36EZCtl95sEkB9EMZ3wAeGZV3dTuXwG8F1iV5MyqOnF0LUuS+rVBn+ffF/iRL7okSR1hbkmSusTckiR1ibklSeqSseZWkkcD+wIvolnFY+bxAK8CLgWeMzWcAdD++xDg58CRo+hXkjQ4/Q5o3AT8dBCNSJI0AuaWJKlLzC1JUpeYW5KkLhlbbiXZGHgL8LGqOnOOsgcAOwLHVtV1Mw9W1fXAe4Cdk9x/aM1Kkgau3wGNc4FdkvR7HUmSRsHckiR1ibklSeoSc0uS1CXjzK3DgTsDL+tRs1u7Pb1Hzdp2+0ePR5EkTa5+g+dNwDbAKwbQiyRJw2ZuSZK6xNySJHWJuSVJ6pKx5FaSPwP+AXhzVf2gR+mO7fbCHjUXzaiVJHVAXwMaVbUWWA28MskxSXYYTFu9JfnTJO9IckmSG5NckeTUJI+ao35FksOSfC3Jr5L8IsnaueolSUvTuHJLkqTFMLckSV1ibkmSumSMufUvwDXAP89Td5t2e0WPmstn1EqSOmCjfk5OMrW00lXAs4FnJ/khcBnwR8/EalVVPayPe94L+DxwW+Bs4OPAtsCjgX2SvK6qXjGtfmPgE8CjgAtonsm1GfBI4KQkr6iq1y22H0lSd4wjtyRJWixzS5LUJeaWJKlLxvT51kOA/YBnVdW185Rv097vhrkKquo3Saqtne1+q4BVANttt93impYkDVxfAxrAHrPs2779mUst9mZJNgQ+TDOccUhVvXvasW2B/wJenuTUqjqrPXQEzXDGG4CXV1W19VsAnwZem+SLVfWlxfYlSeqMPWbZtz1Dyi1Jkvq0xyz7tsfckiRNpj1m2bc95pY0Mo943jU9j3/2nVuOqBOpE/aYZd/2DPfzrbcBXwWOW/hpydTnWnNcM3OdXFXHAMcArFy50syVpAnR74DGvYBrgVsG0MtC7AXsAnxy+nAGQFWtS/IC4CzgQOCsJJvTLFH1NaYNZ7T11ybZH/gBcCTw8NH8CZKkMRp1bkmS1A9zS5LUJeaWJKlLRp1bh7b3fPBcAxczXNluNwN+PUfN5jNqJUkd0NeARlV9c1CNLNB5wIOAn89x/Kvt9i7t9tHAFsA7Zgu8qro0yceBpyW5fVVdNuiGJUmTYwy5JUnSoplbkqQuMbckSV0yytxKsglwFPCRqjp7gadd0W63Yu4Bja1m1EqSOmCDfk5ul09a33MOW+z9qurqqvpSVX1/jpLbtduptdx2a7enz1I7ZS3Nfw4PXGxfkqRuGHVuSZLUD3NLktQl5pYkqUtGnFubAlsDT0lSM3+AhwIbTtv3ImDqc7C79bjuzu12rs/MJEkTqK8BDeAj6xNiSV4O/Guf9+xl/3Z7VrvdEbgB+GGPcy6aVitJWtomLbckSerF3JIkdYm5JUnqklHm1o3Ah3v8rANq2u8XAue05+7V47p7tttzetRIkiZMX484AR4PfCDJ06uq53O6krwGeAVNyAxckrsBLweuAt7T7r4NcNU8vV0+rXaua68CVgFst912/TcrSRqXicktSZIWwNySJHWJuSVJ6pKR5VZV/QbYr8f1zwAeVFX7TdsXmpUxDknyz1V13YxzNgWeDVxUVectpi9J0nj0u4LGZcCTgePbsJhVkn+lGZ64BTi0z3vOdv2tgf8EtgBeXlVTjzjZBrhuzhMbU8/u2maugqo6pqpWVtXKbbfdtu9+JUljMxG5JUnSAplbkqQuMbckSV0y0blVVQW8GrgDcGySFdN6WgG8qz125Kh6kiQNRr8raDwUOB14GnAzcPD0g22ovQt4Fs0STgdW1Uf7vOcfSLIZcDKwC3BCVf3HjJL5lqjq9z8DSVJ3jD23JElaD+aWJKlLzC1JUpdMfG5V1fFJ/ppmhff7JfkcEGBvms/EjqmqE0fZkySpf30NJ1TVd5M8FFgLHJjk5qo6BKB9dtf7gafSrFLxhKr6XL8NT9dOCX4c2A04lWY5p+muBG47z2U2n1YrSVrCxp1by93q1at7Hl+zZs2IOpGkbjC3JEldYm5JkrqkQ7l1KHA28HzggHbfhcBBND1Kkjqm79UjquriNsROB56Z5CbgRcDHgEcBVwGPqqpz+73XdEk2AI4H9gHOBJ5YVTfNKLsCuHWStMtBzWarabWSpCVuXLklSdJimFuSpC4xtyRJXTIpuVVVe/Q4VsAJ7Y8kaQnYYBAXqaof0CwHdQnNUksX04TXz4CHDCm83g7sB3wZeHRVXTdLzfeBjYHte1xn52m1kqRlYEy5JUnSophbkqQuMbckSV1ibkmSRq3vFTSmVNWPkjyEZjmouwLfAx5eVT8c1D2mJDkKeC5wAfDIqrp2jtJzgBcCewHvnqNmT+AWwJCVpGVklLklSTOtuGCnnsdvuu/FI+pk+fnqPXftefx+3/z2iDpZP+aWJKlLzC1JUpeYW5KkUeo5oNEG0vp6JfBG4B+BuyS5y8yCqvriIq471dP/aa/9LeBvquqXPcpPBq4FDkvynpmPOUlyB+CJwOlVddlie5IkTYZJzC1JkuZibkmSumSp5FaS44BnzFO2Q1VdMvxutNw94nnX9Dz+2XduOaJOpKVnqeSWJGnpmW8FjTOAmqdmNgE+2OP4hou4JkkOAN4KfJdmevGKXvVV9eska4BXA69PcsTUkEaSLYATgU2AIxfTjyRp4pzBBOWWNJtT9jyt5/F91+4zok4kTYAzMLckSd1xBksrt94C/GqOY1ePshFJ0lCcwdLKLUnSEjHfgMbxLC7AhuUomnA8Czg0yVx1V1fVW9t/vx54APAyYO8kXwQ2A/YF7gS8oqq+NNSuJUmjMmm5JUlSL+aWJKlLllpuvbmqfjLuJiRJQ7PUckuStET0HNCoqoNH1MdCTU1kPHueuh/SrLRBVd2Y5PHAocCzgEOA62kekfLcqjppSL1KkkZsAnNLkqQ5mVuSpC4xtyRJXWJuSZIm1XwraEyUqtp+kefdBPxb+yNJkiRJkiRJkiRJkjRSAx/QSHJbYDuapaN+VFVXDPoekiQNirklSeoSc0uS1CXmliSpS8wtSdIoDGxAI8n+wOHAvWbs/yrwpqr68KDu1WWrV6/ueXzNmjUj6kSSljdzS5LUJeaWJKlLOpJbK5M8DtgS+AlwWlWtG3NPkqQx6EhuSZKWiL4HNJLcCvgg8BggwLXA99rDdwVWAicmeQrw9Kq6od97SpK0WOaWJKlLzC1JUpd0LLc+OeP365O8tqpeM5ZuJEkj17HckiQtEYNYQeNY4LHAj2gmDD9dVTcDJNmoPfYm4HFt7YEDuKckSYtlbqkTTtnztJ7H9127z4g6kTRm5pYkqUu6kFtXAG8ATgS+C9wK2Bt4HXBUko2r6pWznZhkFbAKYLvtthtNt5KkYepCbkmSlpgN+jk5yUOApwMXASur6uNT4QVQVTdX1ceB+wPfB56e5KH93FOSpMUytyRJXWJuSZK6pCu5VVWHV9URVfWNqrq+qq6uqo8BuwPrgJcn2WGOc4+pqpVVtXLbbbcdad+SpMHqSm5JkpaeflfQOAQo4LlVdflcRVW1Lslzgf8CngOc2ed9paHxG8vSkmZuSVLHfPWeu/Y8fr9vfntEnYyFuSVJ6pJO51bb1z8BRwP7Az7qRJKWtk7nliSpu/od0HgIcFlVnT5fYVV9PsmlwIP6vKckSYtlbkla1lZcsFPP4zfd9+IRdaIFMrckSV2yFHLr0zQDGvcedyOSpKFbCrklSeqgfgc0bg9csB71lwD36fOekiQtlrklSeqSTuRWkuOAZ8xTtkNVXTL8biRJY9SJ3JrHle12k7F2IUkahaWQW5ImlF+SUi/9Dmj8EthiPeo3B67t856SJC2WuSVJ6pKu5dZbgF/NcezqUTYiSRqLruXWbKbeSf/RWLuQJI3CUsgtSVIH9TugcQHwsCRbVVXPN9ySbA3cAzijz3tKkrRY5pYkqUu6lltvrqqfjPH+kqTxmvjcSvIg4HDgeVX181lKVrfbk0bXlSRpTCY+t6T1ccqep/U8vu/afUbUiaT5bNDn+ccDK4CXLKD2JcCGwPv7vKckSYtlbkkaqhUX7NTzR1pP5pYkqUu6kFv7AI8DPpzk1tMPJHkBcDBwdlWdOuK+JEmj14XckiQtQX0NaFTVh4C1wBFJHj9XXZInAEcAX66q4/u5pyRJi2VuSZK6xNySJHVJR3LrKOBk4MHAxUn+I8kbknwFeBvwbeApI+5JkjQGHcktSdIS1NcjTpLcDvg74EPAR5OcABwLfKMtuSfwHOBA4ELg+Um2m++6VeVzHiVJA2duSZK6xNySJHVJF3Krqm5I8ljgWe3PU2m+Pf2/NI83eUdVXTeo+0mSJlcXckuStDT1NaABXApU++8AB7U/MwW4O80zveZTA+hLkqTZmFuSpC7pWm6tTPI4YEvgJ8BpVbVuSPeSJE2eTuRWVf0WeFf7I0lavjqRW5KkpWfOoEiyZVVdM8/5X+T3ASZJ0tiYW9L6e8Tzev9P5rPv3HJEnUjLzxLNrU/O+P36JK+tqtfMdUKSVcAqgO22m/fLaJKkMVmiuSVJWqLMLUnSJOs1yfcfSfYF7l5Vl85RcxBwfVX9YvCtSZK0XswtSUvOigt26nn8pvtePKJONARLKbeuAN4AnAh8F7gVsDfwOuCoJBtX1StnO7GqjgGOAVi5cqVvjkrS5FpKuSVJWvrMLUnSxNqgx7Hb0CxNe3mPmh8AHxxoR5IkLY65JUnqkiWTW1V1eFUdUVXfqKrrq+rqqvoYsDuwDnh5kh3G3KYkqT9LJrckScuCuSVJmli9BjSmpgp37FGT9keSpHEztyRJXbLkc6uq1gH/RLNy4/5jbkeS1J8ln1uSpCXF3JIkTaxejzg5CTgQ+HiSFwLnLeCZXZIkjcuSya0kxwHPmKdsh6q6ZPjdSNL6+eo9d+15/H7f/PaIOpl4Sya35vFp4Gjg3uNuZDk4Zc/Teh7fd+0+I+pE0hK0XHJLkrQ0mFuSpInVa0DjY8B/Ao8BPgeQzDpM+NAkv12Pe1ZV9bqvJEmLsRRz6y3Ar+Y4dvUoG5EkDdxSzK3ZXNluNxlrF5Kkfi2X3JIkLQ3mliRpYs0ZJFVVSZ4AHAQ8DtgJ2Jw/XPLpLsCN/H65KEmSxmKJ5tabq+on425Ck81vS0vdtERzazY7tdsfjbULSVJfllFuSZKWAHNLkjTJek76VdUtwHHtzx9Jcgvw5araa+CdSZK0nswtSVKXLIXcSvIg4HDgeVX181lKVrfbk0bXlUbNYUFpeVgKuSVJWj7MLUnSpHIppgm1evXqnsfXrFkzok4kSfr/7d15mKVnWSfg39PQLCFRtoByEbYIGQ0ZMYm4OwQXYthFBQ0CsgQIos5oKyDOJAIj2Kh4KRkJAhFCHFBAGJKABsPmGjowE4IEjMiioGzBAElokmf+OKeSolJ1qrqrzn7f1/VdX+q873f6+d7q1K+r+6n3A4ANnZjBT6Tdrqoe1N3XP5qrqp6e5HFJ3t3d50+pPpiYB5w6+rHmbz3jsAlVAjAfNPgBALCMttug8UdJ/mEnCgGACZBbAMyTecit5yS5T5IHJrm8ql6f5Iok90ty3yTvT/ITU6sOgEmah9wCgBVyC5gbuy8+cuT4/mMvn1Al7IRtNWh098/sVCEAMG5zmFvHV9XDkhyW5BNJ3tLdn55yTQBMyDzkVndfU1UPTfL44fHIJLuTfDCDx5u8uLuvmmKJAEzIPOQWAKyQWwBMi0ecAMDsesOaj6+uqud193OnUg3ADNt3zNEjx4+75NIJVbJ8uvvaJC8dHgAAAMAmPOoLlteuaRcAANzIZ5O8IMm3Jrllktsk+fEMdtF4TlX9+kYXVtUpVfWeqnrPpz9tsw0AAAAAAIBZYQcNAJgx3f2La166OsmfVtU7klya5JlV9Yru/sg6156Z5MwkOf7443vsxcJBeMCpV44cf+sZh02oEgBYbp5jDAAAAJOlQQMA5kR3f7qqfi3JHyQ5OYlHnQAAHCQNgwAAAMCkadAAgPnyxgwaNO4z7UIAAAAAAGA1zfAwmgYNWADCDpbK54bnm0+1CgAAAADggFTVnZL8apIHJrlTkiuT/H2S3+/uc9eZvzvJk5M8IcmRGTwK+dIkL1xvPovnvBPeMnL8pAtPnFAlwE7RoAEA82XlQeEfm2oVAMDc8hd8ACwi+QbMuqr61iQXJLl9kncneV2Sw5M8KMmJVfUb3f2sVfNvluT1GTRzXJzk5UkOSfIjSd5cVc/q7t+Y7F0AsF0aNABghlTV9yb5xSSndvcn15myZ3h+8+SqAgAAAAAOVlXdJMlrMmjOeGJ3v2zV2OFJ/jzJM6vq/O5+13DoGRk0Z7wgyTO7u4fzD83gMcjPq6p3dvdfTfBWANimXdMuYDuq6tCqOq+quqoePWJeVdWjq+pvqupzVfX5qvrb4Ws1yZoBYBMnJnlYktdU1detHqiqpyd5XJJ3d/f5U6gNAAAAADhw909yVJI3rG7OSJLu/nSSpw8//OkkqapbZfCDWu/NquaM4fwvJjk5yTVJTht75QDsqLndQaOqvjHJuUm+bQvT/yDJKUkuS/LHSTrJDyZ5VZLvy+D5XQAwC56T5D4ZdMdfXlWvT3JFkvsluW+S9yf5ialVBwAAAMBU7Dvm6JHjx11y6YQq4SBclOR7k6y3Y26S7Bue7zo8PyjJoUlevLo5Y0V3f6qqXpfkJ6vqjt39bztdMADjMZcNGlV1dJLzktw5yduS/MCIuY/JoDnj1Ul+prv3D1/fneQVSU6pqnd09zljLxwANtHd11TVQ5M8fng8MsnuJB/MoGv+xd191RRLBAAAAAAOQHdfkWTUo0juMDxfOTx/1/D8lyOuuTCDnTS+M4NHnsBcecCpV44cf+sZh02oEpisuWvQqKrjk1yQ5JAMgucW2aBBY/j4kv+R5FNJnrTSnJEk3b2/qp6YwbZSpyXRoAHATOjua5O8dHgAAABLyk9KA8DSOHl4ftfwfI8MHmHy0RHXfHjVXADmxNw1aCT5xyQfT/KM7j63qh43Yu59Mwim567308bdfXVVvTzJr1bVt3f3RWOpGAAAAAAAANaoqnsleWaSzyd5+fDl2yX5fHdfN+LSz6yau977npLBDvO5y13usjPFAsyBWW903zXVX/0gDLeB+rbuPncL07e6BVSSfPe2CgMAAAAAAIAtqqrbJHlTkkOTPLO7V575cNskmz3m+Eur5t5Id5/Z3cd39/GHH374jtQLwPbN4w4a6e6vbnHqyrZOl42YYwsoAIAZ5DmUAMDBmvWfmAIAqKpDkpyb5Kgkr+rul6yZcpNN3mIu/42P+eXv6mBnzN0OGgdoZVunz46YM3ILKAAAAAAAANgpVbU7yesy2An+/CRPWDPlc0kO2eRtbrVqLgBzYtG7626bpLv7mo0mdPeXq6qzwRZQied0sTh0NwLzYM+ePSPH9+7dO6FKAAAAAGBnVdWuJK9McmKSdyR5RHfvXzPts0m+rqqqu3uDt7r1qrkAzIlFb9BIkhoVYFV1kyQ16g26+8wkZybJ8ccfv1EQAgAAAADMPY3zAGP1+0keleRvkjyou69aZ84/JblZkrsl+cgG73PPVXMB5tayPaJy0R9xsrKt06htoGwBBQAAAAAAwFhV1XOSPDXJxUl+pLu/uMHUvx6e7z/i7U5Icl2Sv925CgEYt0XfQWNlW6dbJ/nSBnNsAQUzbPfFR44c33/s5ROqBAAAAAAADk5V/VySZyd5f5If7u4vjJh+bpIvJvnZqnr52l3iq+obkjwiyV9297+Nq2YAdt6iN2isbOt0ryT/ssEcW0ABAAAAAMAO8oNXcIOqenSSFyX5UJIf6u6RPzTc3V+qqr1JTk/y/Kp6xkqTRlUdmuScJDdPctpYCwdgxy16g8bqLaAu3GDOCWvmAgAAAAAAwE55TpJK8q4kT6mqjeZd0d0vGv7385PcN8kvJ/nBqnpnkkOSnJTkzkme1d1/NdaqAdhxi96gcVEGO2M8sar+Z3dftXqwqm6Z5AlJPtzdF02jQAAAAAAAABbaSkfGEzaZ99EMdtpId3+lqh6e5ClJHp/kiUmuzuARKU/t7jePqVYAxmihGzS6u6vq9CR/lOQPq+px3b0/Sapqd5KXJvmGJL84xTJnyp49e0aO7927d0KVAAAAAAAAzL/uvttBXrc/ye8NDwAWwEI3aCRJd7+yqr4nySlJjquqv8igU/EHkxyV5MzuPmeaNQIAAAAAAEzSvmOOHjl+3CWXTqgSAFgeC9+gMfSUJO9O8rQkjx6+dlmSxyQ5e1pFAQAAAAAAAADLYe4bNLr7rCRnbTKnk7xqeAAAAAAAAAAATNTcN2gsuz179owc37t374QqAQCArbOVLrPO91oAAAAA7LRd0y4AAAAAAAAAAGDRadAAAAAAAAAAABgzDRoAAAAAAAAAAGOmQQMAAAAAAAAAYMxuOu0CAAAAAAAAAICDt/viIzeds//Yy7f96+w75uiR48ddcum2f41FpkGDpXHeCW8ZOX7ShSdOqJLFttkX/534wr8oZj3AZr0+AAAAAAAAmCcaNAAAAAAAAAA24IeAgZ2ya9oFAAAAAAAAAAAsOjtoAAAAAAAAAAA7Zt8xR48cP+6SSydUyWzRoAFT8IBTrxw5/tYzDptQJQAAwHbs2bNn5PjevXsnVAkAAADA1uy++MiR4/uPvXxClSwfDRoAwELb7B/OEv94BgAAAAAAjN+uaRcAAAAAAAAAALDoNGgAAAAAAAAAAIyZR5wAbINndAEAAAAAAABboUEDAAAAAHbAvmOOHjl+3CWXTqgSAAAAZpFHnAAAAAAAAAAAjJkdNACWjJ/oAgAAAAAAYJ4syr9vadAAAIBN7L74yJHj+4+9fEKVAAAAAAAwrzRoAAAw1x5w6pUjx996xmETqgQAANhp553wlpHjJ1144oQqAQCA7dOgAQAAAMCG7CQFAAAAO2PXtAsAAAAAAAAAAFh0dtAAAAAAYCntO+bokePHXXLphCoBAABgGWjQAAAAABbGA069cuT4W884bEKVsEg0cgAAALATNGgAAAAAAAAAE7Fnz56R43v37p1QJQCTp0EDAIAd4SeWAQAAAABgYxo0gBvxD2xAopMdAGBe7b74yJHj+4+9fEKVAAAAAKtp0ABYh7/QBACA2aKRHAAAgI34npF5oUEDAICRfHMDAAAAAADbt2vaBQAAAAAAAAAALDo7aMAO8hPGAAAAAAAAAKxHgwYwFbsvPnLk+P5jL59QJYtj3zFHjxw/7pJLJ1QJAAAAALPKD5kBAEyPBg0AgB1w3glvGTl+0oUnTqgSAJhNsnL7NLoDALPMD5ABwOZ2TbuASaqqw6vqhVX1oaq6qqo+VlWvqap7T7s2AFhLbgEwT+QWsPviI0ceMEvkFgDzRG4BLI6l2UGjqo5I8vYk90hyfpI3JvnGJD+a5MFV9eDuftv0KgSAdwkS5QAAEV5JREFUG8gtAOaJ3IKts608TJ/cgsW21R2n7EzFvJBbLDvfQ7FolqZBI8lZSe6a5KHd/aaVF6vqyCTvSnJOVR3V3VdMqT5gG3xDxQI6K3ILgPlxVuQWC8pfBsJCOityC4D5cVbkFiwk/7a1nJaiQaOq/kuS+yf53dXhlSTdfXlVPTXJnyV5WpLnTaFEALie3AJgnsitnbVnz56R43v37p1QJQCLSW6xVeed8JaR4yddeOKEKgGWmdwCWDxL0aCR5FHD8+9tMP6mJJ9I8lNZ0ADzl3wAc2XpcwuAuTITueV7HpgffkqMKZuJ3AKALZJbW+D7QeBg7Dvm6JHjx11y6Vh+3WVp0PiuJB/r7nW/w+/urqq3Jzm5qm5tG6jZoEvdVrqwxOQWAPNEbgEcoGn9RSBJ5BYA80VuASyYZWnQuEeSv91kzoeTVJK7J3nv2CsCgI3JLQDmidxaUJrmgQUlt4CFpxFwocgtgAWz8A0aVbU7yWFJPrvJ1M8Mz7cbb0XANNlKl1kntwCYJ3ILGCffv7HT5JYt4GGc5BY7TW7B7PG1np1Q3T3tGsaqqu6Y5FNJXtHdjx8x77FJzkryyO5+7ZqxU5KcMvzwqCSXjafasbl9bgjoZeB+SazLRuZxXT7T3Uvz45nbzS2ZNZeW7Z6X7X63yrqsbx7XRW6tP09uLYZlu99kOe95K6zL+uZxXeTW+vPk1mJwv6ywNuubx3WRW+vPk1uLwf2SWJdR5nFt1s2thd9BY5WbbDK+4Vp095lJztzZcianqt7T3cdPu45Jcb8k1mUj1mWuHFRuyaz5s2z3vGz3u1XWZX3WZa7IrSWwbPebLOc9b4V1WZ91mStyawm4X1ZYm/VZl7kit5aA+yWxLqMs0trsmnYBE/C54fmQTebdas18AJgGuQXAPJFbAMwTuQXAPJFbAAto4Rs0unt/kiuT3HqTqSvjmz3LCwDGRm4BME/kFgDzRG4BME/kFsBiWvgGjaF/SnKvTebcM0kn+cj4y5m4ud3C6iC5XxLrshHrMh+WObeW8ffost3zst3vVlmX9VmX+SC3lsey3W+ynPe8FdZlfdZlPsit5eF+WWFt1mdd5oPcWh7ul8S6jLIwa1PdPe0axq6qzkjy1CTf1N2XrzNeST6W5D+6++hJ1wcAq8ktAOaJ3AJgnsgtAOaJ3AJYPMuyg8b/Hp5/boPxhya5c5JzJlMOAIwktwCYJ3ILgHkitwCYJ3ILYMEsxQ4aSVJVb0vyX5L8aHe/adXrRyZ5V5KbJDmqu6+YUokAcD25BcA8kVsAzBO5BcA8kVsAi2WZGjSOSPL2JPdIcl6SS5PcKcnDk1SSB3f326ZW4Iyoqrtl8Jyy07v7tKkWA7DE5NbWyC2A2SC3tkZuAcwGubU1cgtgNsitrZFbwLxYlkecpLs/nuQ7k/xWknsleXoGHYdvTnLfaYVXVb29qnrNcW1Vfb6q3l1Vv1BVN59GbdM2XIsLtjDvfsO5zx5+/OThx887gF/rkqq6pqpuu52aJ6GqbldVz6mq91XVl4fHZVX1+1V192nXNymr/t/5SlXd8QCu+8DK/2vjrG9WDNfpq9OugwMnt+aP3Fqf3JJZB0JuzS+5NX/k1vrkltw6EHJrfsmt+SO31ie35NaBkFvzS27NH7m1Prkltw7EIufW0jRoJEl3f7q7f6m779ndt+zuI7r7kd39/mmXluT0VccLkpyf5D8l+Z0kF1XV4dMrb+68NslXkpxcVbXZ5Ko6Nsm9k5zb3Z8bd3HbUVXHZNAd++wM7vFlSf5Xks8meVqSD1TVj02vwqnYneRntjKxqr4/yTePtxzYOXJracit5SGzWGhya2nIreUht1hocmtpyK3lIbdYaHJracit5SG3lthNp10ASZLr1ttuqaoOyeAL1KOSvDDJYydc11zq7s9X1XlJHpbk+5K8c5NLVtb1lWMtbJuqaneSP01yeJJHdvdr14yfnMEfeJbpOXPXJflSkidW1Qt682c2nTI8/3uSO4y1MlhscmsHya2lyS2ZBdMjt3aQ3JJbG5BbsHPk1g6SW3JrA3ILdo7c2kFyS25tQG4tmKXaQWPedPeXkzwxyZeTPKKqNNRs3dnD86NHTRqGwk9m0KF37riL2qYfymD7sj9cG15J0t2vTnL37t5026wF0knOSXJkkh8YNXG4vdcjkuxLcsn4S4PlI7e2RW4tPpkFM0ZubYvcWnxyC2aM3NoWubX45BbMGLm1LXJr8cmtJadBY8Z195eSfDDJrZLcfsrlzJM3Z9Bp9+ObPOPspAw69v64u/dPpLKDd/TwfPFGE4a/X5bNS4fnU0bOSh6T5Bar5gNjILcOmtxaDjILZozcOmhyaznILZgxcuugya3lILdgxsitgya3loPcWmIaNObDyvO5vjzVKuZId1+TwXZJt07yoBFT52L7p6Frh+cjp1rFjOnufRmE+sOqatTWTk/KYMuocyZSGCw3uXWA5NZykFkws+TWAZJby0FuwcySWwdIbi0HuQUzS24dILm1HOTWctOgMeOq6n5JjkhycXf/x5qxh1TVuVX16aq6pqo+WlUvr6qj132zG677vqr6k6r65PC6f66qP6iqu4645uur6jlVdVlVXVVV/1pVL6uqb9yRGx2PkdtAVdXtkjwwyQe7+6KJVXXwVmp8alV9x1QrmT0vTbI7yePWG6yq70vyLUle091XTrAuWDpya1vk1nKQWTBD5Na2yK3lILdghsitbZFby0FuwQyRW9sit5aD3FpSGjRmUFXdrKruWVW/kuRNSa5L8sxV4zetqrOTvDHJfZK8IclvJfm7JI9K8r6qesoG7/38JO/M4JlGfz687l1JfiLJpUnut841t0hyfpJfTfLhJL+bQVfX45P8fVXdZvt3PRbvTPKxJCfV4BlNa/1kkptlProL093vyuBzdmiSd1fVS6rq26dc1qx4dQYdhE+sqlpnfGWLKFtAwRjIrR0jt5aDzIIpk1s7Rm4tB7kFUya3dozcWg5yC6ZMbu0YubUc5Nay6m7HFI8kb0/SI45PJHnYmmteOBx7SZJbrBk7Isl7Mwi9E9eM/ezwuv+T5OvXjH19ktck+cpwzmmrxk4evvbsNdf8tySvS3LbMa1NJ7lgC/Put159w7HfGI49ZZ2xizLYVumIaf8+OIA1OWz4eb921e+Ry5L80rg+D7N6DP/f+eqqj182XI/7r5l3myRXJblk1WsXDL78Tf8+Jr1ODsd2D7k1cm3k1o1rllsts7azVg7Hdg+5NXJt5NaNa5ZbLbe2s1YOx3YPuTVybeTWjWuWWy23trNWDsd2D7k1cm3k1o1rllstt7azVot02EFjNnSS04fHc5PsT3Jlku9Nctfu/rOVicNtmn4+yfuTPK27r/6aN+r+eAbPpLoqyd5V1x2a5NeT/FuSk7v7C2uu+0KSn0rynnXqu8vw/P/WXPPb3f2I7v7cgd7wBL1qeP6abaCq6luSHJ/kwuGazYXuvrK7n5zBlkbPT/LRJPfK4HP9iao6dZr1TdmZw/Mpa15/TJJbRIch7CS5NT5yaznILJgsuTU+cms5yC2YLLk1PnJrOcgtmCy5NT5yaznIrSWkQWM2XNfdpw2PX8ugg+ywJPfp7mvXzH1YkpsmeV53f3W9N+vuf8mgW/DeVfXNw5d/IINuqxf1mmd9rbru2gw68tb6++H5N6vqQVV1swO5uWnq7g8keV+S766qu68aeuzwPBfbP63V3Zd19zO7+25JvieDrcJumeTFVfXYkRcvqO7+uwz+kPXwqrr9qqEnJbk6N/xhBtg+uTUmcms5yCyYOLk1JnJrOcgtmDi5NSZyaznILZg4uTUmcms5yK3lpEFjNp2e5D+SnF5Vt14zdq/h+eJN3mPf8HzUAV73N2tf6O63JXlGkntksH3U56vqgqr6+ao6bJP3244rk9x8C/NuMTxftcH42Ukqg62sUlW7Mug4/FIGW1jNte7+6+5+aJL/Pnzp16dZz5S9NIPnrj02Sarqe5IcneR13f35aRYGC05uDcitLZBb15NZMD1ya0BubYHcup7cgumRWwNyawvk1vXkFkyP3BqQW1sgt64nt5aMBo0Z1N2fyWB7n9sl+bWNpm317Q7yurU1vSDJPZP8QpI/T3LfJC9KclFV3fZg3nML/iXJnbYw787D8yc3GD8ng2eWrWwD9UPD9319d39pWxXOlucm+XiSu6zzB59lcXYGf5B50vDjJw/PtoCCMZJb15NbB2bZc0tmwZTIrevJrQMjt+QWTIXcup7cOjByS27BVMit68mtAyO35NZS0aAxu34ngy9GP1tV37Tq9Q8Nz8ducv3K+IfWnDe77js3Gujuj3b373b3w5PcIclvZtDB+CubvOfB+mCSu1fVEZvM+/7hed96g939ySR/meSoqvr23LD90x/tSJUzors7yb8OP6xp1jIt3X1Fkj/J4HP90CQ/luRD3f2O6VYGS0Fuya0Dsuy5JbNg6uSW3DogcktuwZTJLbl1QOSW3IIpk1ty64DILbm1bDRozKjuvjrJszPY0mbvqqE3JflqkmdV1U3Wu7aq7pTkkUku7e5/GL78tiRfSPILG23bNHy/Z2wwdsg69a1sNXTMVu7pILwugy/EezaaUFX3SPLjSd7f3ZeNeK+zh+dTM3jO2SeSXLhDdU5EDTymqm66wfjtktw7yT8u+ZZHKx2FZ2Xw3LI/nF4psDzkVhK59TXk1pbILJgSuZVEbn0NubUlcgumRG4lkVtfQ25tidyCKZFbSeTW15BbWyK3logGjdl2dpL3JXlYVZ2QJN39kSS/l+Q/J3lFVd1q9QVVdVSS85MckuSXV17v7i8mOS3JNyT5k6q6w5rrDk/yp0mOW1tEVT0ryWVVdc81Qw8anv/54G5vU3+cwf0/var+a1V9TdfcMLzOy+A5XusG7yqvS/LlJI/L4Avb2d193Y5XPF6/lUFX5NlVdejqgWF4vT7JrTLYCmppdfe7k3wgya2T7M+CdZLCjJNbcms1ubUJmQVTJ7fk1mpyaxNyC6ZObsmt1eTWJuQWTJ3cklurya1NyK3lsm6nErOhu6+rqj1J/iLJb1fVccMvur+SQRD9dJITqurcJJ9Lco8kD8mgK/Hp3X3emvd7UVXdNYPnbP1DVb0pgy2D7pxBGN0syWOSvHZNKf83yR2TvLeq3pBBd943J3lgkk8lecGO3/yg3mur6iEZ3n+SJ1XVBUm+mORbkpyYwe/hX+ruczd5ry8O7/dRw5deOY6ax+z5GWx39cgkD6iqN2fwHLM7J3lwksOSnN7dvmgPOgt/O8kbu/vfp13MDNhVVaeNGL+iu180qWJYXHJLbq0ht7ZGZt2Y3GIi5JbcWkNubY3cujG5xUTILbm1htzaGrl1Y3KLiZBbcmsNubU1cuvGFjO3utsxxSPJ25N8dZM55yfpJE9Y8/pDMuiw+0ySryT5WJKXJ7n3Ju93vww67j6V5JoMOgTPSHJEBsHYSU5bc823ZhBsn0pydZKPJHlJkiMnsEa3yqCDcF8GQX11ko9m0IH5HQfwPicN7+2iaX/et7EWuzJ4xti5w8/FVzIIsVcnue+065vwWmz4/06S2w5/n/zwBuMXZPhYs0U/huvUmxz/PO06HfNzyK0trZHcuuEe5FbLrINYK7nl2LFDbm1pjeTWDfcgt1puHcRayS3Hjh1ya0trJLduuAe51XLrINZKbjl27JBbW1ojuXXDPcitllsHsVYLmVs1vEEAAAAAAAAAAMZk17QLAAAAAAAAAABYdBo0AAAAAAAAAADGTIMGAAAAAAAAAMCYadAAAAAAAAAAABgzDRoAAAAAAAAAAGOmQQMAAAAAAAAAYMw0aAAAAAAAAAAAjJkGDQAAAAAAAACAMdOgAQAAAAAAAAAwZho0AAAAAAAAAADGTIMGAAAAAAAAAMCYadAAAAAAAAAAABiz/w8lxafE0HBKzgAAAABJRU5ErkJggg==\n",
      "text/plain": [
       "<Figure size 2160x288 with 5 Axes>"
      ]
     },
     "metadata": {
      "needs_background": "light"
     },
     "output_type": "display_data"
    }
   ],
   "source": [
    "plotName = 'test'\n",
    "\n",
    "# foxq2 = S specific;\n",
    "\n",
    "# genelist = ['wrb','get4'] # disrupts photoreceptors synapses (Danielle, 2016)\n",
    "# genelist = ['pappaa','igf2bp2b'] # In Miller, 2018 model pappaa cleaves igf-bp to increase igf1 availability. unfortunately, this is supposed to bind mRNA not IGF itself\n",
    "# genelist = ['gpx1a','gpx1b','gpx4a','gpx4b']\n",
    "# genelist = ['rgs9bp','RGS9BP']\n",
    "# genelist = ['chchd1','chchd10']\n",
    "# genelist = ['yipf6','yipf1','meis1b','pbx1a']\n",
    "# genelist = ['sema3fb','sema3fa','sema6d']\n",
    "# genelist = ['esama','esamb']\n",
    "# genelist = ['il34','il7r']\n",
    "# genelist = ['cxxc4','cxxc1a']\n",
    "# genelist = ['mbd3b','mbd3a']\n",
    "# genelist = ['cdr2a','cdr2l']\n",
    "# genelist = ['coro6','coro2a','coro1ca']\n",
    "# genelist = ['auts2a','auts2b']\n",
    "# genelist = ['plxnb1a','plxna2']\n",
    "# genelist = ['cux1a','cux1b','cux2b']\n",
    "# genelist = ['arih1','arih2','arih1l']\n",
    "# genelist = ['crhbp','crhr2','crhr1']\n",
    "# genelist = ['chn1','chn2']\n",
    "# genelist = ['robo1','robo2','robo3','robo4'] #nothing here\n",
    "# genelist = ['ntng1a','ntng2a','ntng2b'] #low counts but interesting. Netrins are important for RGC axons going towards optic disc\n",
    "# genelist = ['fat1a','fat1b','yap1','scrib','cc2d2a'] #fat1 important for renal development along yap1/hippo and scribble; cc2d2a involved in cilium development\n",
    "# genelist = ['c1qbp','c1qa','c1qb'] #keep this in mind. Beth Stevens work shows that microglia engulf synapses through activation by c1q; maybe c1qbp is produced in healthy cells (inside mitochondria) to keep c1q bound and gets released upon damage.\n",
    "# genelist = ['cc2d2a','rab8'] #Bahcman 2011: cc2d2a-/- causes Joubert syndrom. Important for cilium development by association with rab8. Not expressed in adults?\n",
    "\n",
    "# genelist = ['pcloa','pclob']\n",
    "# genelist = ['mtdha','mtdhb']\n",
    "# genelist = ['igsf9b','igsf21a','igsf21b']\n",
    "# genelist = ['rx3','lin9','igsf21b','foxq2','foxo3b','foxg1b','rxrga','smad5']\n",
    "# genelist = ['sall1a','sall1b','sall2','sall4']\n",
    "# genelist = ['rbm8a','rbm39a','rbm39b','rbmx','rbm22','rbm5']\n",
    "# genelist = ['rbm5','rbm4.1','rbm4.2']\n",
    "genelist = ['kif5ba','kif1aa','kif1ab','kif1bp'] #kif5-/- : Charcot-MarieTooth; Campbell 2013: kif5s in all retinal layers, but only kif5ba and bb are in ONL\n",
    "genelist = ['kif3a','kif3b','kif16ba','kif19','kif25'] #\n",
    "\n",
    "\n",
    "# genelist = ['pcdh2ab9','pcdh2g28','pcdh15a','pcdh15b','pcdh11','pcdh10a'] #interesting ones\n",
    "# plotName = 'pcdhs'\n",
    "\n",
    "# genelist = ['sema7a','sema6d','sema3fb','sema3fa','sema4ab','sema4ba'] #the good ones\n",
    "# genelist = ['sema6a','sema3aa','sema6e','sema5bb','sema4c','sema6bb','sema6dl','sema5ba'] # the inconsistent ones\n",
    "# genelist = ['sema6ba','sema5a','sema3c','sema3b','sema3e','sema3d','sema3h','sema3ab','sema4e','sema4aa','sema3gb','sema3ga','sema4bb','sema3bl','sema4ga','sema4gb'] # the negative ones\n",
    "\n",
    "# genelist = ['ncdn','flrt3'] \n",
    "# genelist = ['odf2a','odf2b']\n",
    "# genelist = ['syne1a','syne1b','syne2b','syne3']\n",
    "# genelist = ['rasgrf2a','rasgrf2b']\n",
    "# genelist = ['rasgrf2a','rasgrf2b','nmnat2','rbx1','nrf1','unc119.2','unc119b'] # involved in migration\n",
    "\n",
    "# genelist = ['syt5a','syt5b','syt1a','syt1b','syt14a']\n",
    "\n",
    "\n",
    "# genelist = ['myo19','myo16','myo7bb','myo5b','myo1cb','myo5aa','myo7aa','myo6a','myo6b']\n",
    "# genelist = ['fgfr1op','fgfr1op2','fgf1b','fgf13a','fgf11b']\n",
    "# genelist = ['prdm1b','fgfr1op2']\n",
    "# genelist = ['frmpd1a','gpsm2'] # Ulisse + Pahlberg: these proteins inteact with rod transducin \n",
    "\n",
    "# genelist = ['epha7','frmpd2','pkp4','chd7','hook1']\n",
    "# genelist = ['thrb','samd7','samd1a','samd11','dscaml1'] #samd11:https://www.ncbi.nlm.nih.gov/pubmed/27734943\n",
    "# genelist = ['thrb','nrl']\n",
    "# genelist = ['opn3','opn4.1','opn4a','opn4b','opn5','opn6a','opn9'] #\n",
    "            \n",
    "fH, axes = plt.subplots(1,len(genelist),sharey=False,figsize=[6*len(genelist),4])\n",
    "for i,g in enumerate(genelist,start=0):\n",
    "    n, h = plotBars(g)\n",
    "    plt.sca(axes[i])\n",
    "    pH = plt.bar(n, h, width=0.8, bottom=None, align='center', data=None,\n",
    "                 color=pC['plt'])\n",
    "#     if  i == 0:\n",
    "    formatFigure(g, fH, axes[i], pH)\n",
    "#     else:\n",
    "#         formatFigure_right(g, fH, axes[i], pH)\n",
    "# plt.savefig(\"{0}.svg\".format(plotName), transparent=True, format=\"svg\")\n",
    "fH.tight_layout()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "np.size(data)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "pwd"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "np.sqrt(2)*2"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "pwd"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "pC['plt'][20]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "colors"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Trying to draw counts for single gene by wedges in a circle \n",
    "> Prob not worth it, difficult to interpret"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "fig, ax = plt.subplots(figsize=[12,12])\n",
    "\n",
    "n = 30;\n",
    "r = 0;\n",
    "wdg_theta = 360/30;\n",
    "\n",
    "patches = []\n",
    "circles = []\n",
    "\n",
    "pM = matplotlib.patches.Wedge(0, np.mean(data[r,0:6]), wdg_theta * (0), (wdg_theta * (6)), edgecolor=\"none\", facecolor = pC['r'], alpha = 0.4)\n",
    "patches.append(pM)\n",
    "\n",
    "pM = matplotlib.patches.Wedge(0, np.mean(data[r,7:11]), wdg_theta * (6), (wdg_theta * (11)), edgecolor=\"none\", facecolor = pC['u'], alpha = 0.4)\n",
    "patches.append(pM)\n",
    "\n",
    "pM = matplotlib.patches.Wedge(0, np.mean(data[r,12:17]), wdg_theta * (11), (wdg_theta * (17)), edgecolor=\"none\", facecolor = pC['s'], alpha = 0.4)\n",
    "patches.append(pM)\n",
    "\n",
    "pM = matplotlib.patches.Wedge(0, np.mean(data[r,18:24]), wdg_theta * (17), (wdg_theta * (24)), edgecolor=\"none\", facecolor = pC['m'], alpha = 0.4)\n",
    "patches.append(pM)\n",
    "\n",
    "pM = matplotlib.patches.Wedge(0, np.mean(data[r,25:30]), wdg_theta * (24), (wdg_theta * (30)), edgecolor=\"none\", facecolor = pC['l'], alpha = 0.4)\n",
    "patches.append(pM)\n",
    "\n",
    "for i in range(n):\n",
    "    p = matplotlib.patches.Wedge(0, data[r,i], wdg_theta * (i), (wdg_theta * (i+1)) -2, edgecolor=\"none\", facecolor = pC['plt'][i], alpha = 0.4)\n",
    "    patches.append(p)\n",
    "\n",
    "for i in [0.001,.2,.4,.6,.8,1]:\n",
    "    c = matplotlib.patches.Circle((0,0), np.max(data[r,:])*i, edgecolor = '#000000', facecolor='none', alpha=0.59)\n",
    "    circles.append(c)\n",
    "\n",
    "plt.xlim(-np.max(data[r,:])*1, np.max(data[r,:])*1)\n",
    "plt.ylim(-np.max(data[r,:])*1, np.max(data[r,:])*1)\n",
    "\n",
    "\n",
    "circleColl = matplotlib.collections.PatchCollection(circles, match_original=True);\n",
    "ax.add_collection(circleColl)\n",
    "\n",
    "patchColl = matplotlib.collections.PatchCollection(patches, match_original=True);\n",
    "ax.add_collection(patchColl)\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "np.max(data[r,:])"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "np.max(data[4,i])"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "data"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "range(0,4)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "range?"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### From 2019, Whitaker ,..., Swaroop\n",
    "> genes involved in rod synapse size and localization\n",
    "\n",
    "> [link to biorXiv](https://www.biorxiv.org/content/biorxiv/early/2019/08/30/753012.full.pdf)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "plotName = \"test\"\n",
    "\n",
    "# gdf = findgenelist('dpf') # dpf3 -> not this one but dpf2 in all photoreceptors\n",
    "# gdf = findgenelist('epb') # epb4.1l2 -> not this one but some interesting patterns in this family\n",
    "# gdf = findgenelist('grtp') # grtp1 -> this one is interesting!\n",
    "# gdf = findgenelist('kcnj') # kncj14 -> squirrel too but not zebrafish\n",
    "# gdf = gdf.sort_values(by=[\"baseMean\"],ascending=False)\n",
    "\n",
    "# gdf = findgenelist('llgl') # llgl2 -> nope\n",
    "# gdf = findgenelist('rab') # rab28 -> nope; this family is too big\n",
    "# gdf = gdf[gdf.baseMean > 2]\n",
    "# gdf = gdf.sort_values(by=[\"baseMean\"],ascending=False)\n",
    "\n",
    "# gdf = findgenelist('rom1') #rom1 -> rod-specific; what about outer segments?\n",
    "\n",
    "# gdf = findgenelist('arl') # arl2bp -> nope but arl3l2 in all photoreceptors, and arl3l1 downregulated in uv and s. Both confirmed eye specific (https://zfin.org/ZDB-GENE-040426-1678; https://zfin.org/ZDB-GENE-040426-1649)\n",
    "# These genes are supponsed to interact with unc119c (https://www.ncbi.nlm.nih.gov/pubmed/27079236) and act as their GDF\n",
    "# gdf = gdf[gdf.baseMean > 5]\n",
    "# # gdf = gdf[gdf.baseMean < 300]\n",
    "# gdf = gdf.sort_values(by=[\"baseMean\"],ascending=False)\n",
    "\n",
    "# gdf = findgenelist('unc119') # unc119c is photoreceptor specific (https://www.ncbi.nlm.nih.gov/pubmed/27079236) and CoIPs with arl3l2\n",
    "\n",
    "\n",
    "gdf = findgenelist('ppar') #ppargc2\n",
    "# gdf = findgenelist('dpf')\n",
    "# gdf = findgenelist('dpf')\n",
    "# gdf = findgenelist('dpf')\n",
    "# gdf = findgenelist('dpf')\n",
    "# gdf = findgenelist('dpf')\n",
    "# gdf = findgenelist('dpf')\n",
    "# gdf = findgenelist('dpf')\n",
    "# gdf = findgenelist('dpf')\n",
    "# gdf = findgenelist('dpf')\n",
    "# gdf = findgenelist('dpf')\n",
    "# gdf = findgenelist('dpf')\n",
    "# gdf = findgenelist('dpf')\n",
    "# gdf = findgenelist('dpf')\n",
    "\n",
    "### https://www.sciencedirect.com/science/article/pii/S2211124718303218\n",
    "# gdf = findgenelist('frmpd') #frmpd2 interacts with lrit1 in cones \n",
    "# gdf = findgenelist('sdc') #sdcbp interacts with lrit 1\n",
    "# gdf = findgenelist('grip') #grip1 interacts with lrit1\n",
    "\n",
    "# gdf = findgenelist('dpf')\n",
    "# gdf = gdf.sort_values(by=[\"baseMean\"],ascending=False)\n",
    "# gdf = gdf.drop(np.arange(20,93))\n",
    "# gdf = gdf.drop([0])\n",
    "# gdf = gdf.sort_values(by=[\"symbol\"])\n",
    "# gdf = gdf[gdf.baseMean > .45]\n",
    "\n",
    "gdf\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "genenames = gdf['symbol'].values\n",
    "samplenames = [] #skipping names to replace in intaglio\n",
    "log2matrix = np.vectorize(log2)\n",
    "# data = log2matrix(gdf.iloc[0:,7:15].values)\n",
    "data = gdf.iloc[0:,7:37].values\n",
    "\n",
    "\n",
    "fig, ax = plt.subplots(figsize=[12,12])\n",
    "im, cbar = heatmap(data, genenames, samplenames, ax=ax,\n",
    "                   cmap=\"inferno\", cbarlabel=\"fpkm\")\n",
    "# texts = annotate_heatmap(im, valfmt=\"{x:.1f} t\")\n",
    "\n",
    "plt.savefig(\"{0}.svg\".format(plotName), transparent=True, format=\"svg\")\n",
    "plt.show()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# gdf = gdf[gdf.baseMean > .45]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.7.4"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 4
}
